{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "\n",
    "- [Introduction](#introduction)\n",
    "\n",
    "- [Links of Interest](#links)\n",
    "\n",
    "- [Part 1. Loading and Filtering Data](#part_1)\n",
    "    - [Setting input and output paths](#paths)\n",
    "    - [Reading in genes, annotations, and phenotype descriptions](#read_this_data)\n",
    "    - [KEGG biochemcial pathways](#read_in_kegg)\n",
    "    - [PlantCyc biochemcial pathways](#read_in_plantcyc)\n",
    "    - [Phenotype hierarchical subsets](#read_in_lloyd_meinke_subsets)\n",
    "    - [Phenotype hierarchical classes](#read_in_lloyd_meinke_classes)\n",
    "    - [Relating the group datasets to genes](#relating)\n",
    "    - [Filtering the datasets](#filtering)\n",
    "    \n",
    "- [Part 2. NLP Models](#part_2)\n",
    "    - [Word2Vec and Doc2Vec](#word2vec_doc2vec)\n",
    "    - [BERT and BioBERT](#bert_biobert)\n",
    "    - [Loading models](#load_models)\n",
    "\n",
    "- [Part 3. NLP Choices](#part_3)\n",
    "    - [Preprocessing descriptions](#preprocessing)\n",
    "    - [POS Tagging](#pos_tagging)\n",
    "    - [Reducing vocabulary size](#vocab)\n",
    "    - [Annotating with biological ontologies](#annotation)\n",
    "    - [Splitting into phene descriptions](#phenes)\n",
    "    \n",
    "- [Part 4. Generating Vectors and Distance Matrices](#part_4)\n",
    "    - [Defining methods to use](#methods)\n",
    "    - [Running all methods](#running)\n",
    "    - [Merging distances into an edgelist](#merging)\n",
    "    - [Adding edge information](#merging)\n",
    "    \n",
    "- [Part 6. Biological Questions](#part_5)\n",
    "    - [Distributions of distance values](#ks)\n",
    "    - [Within-group distance values](#within)\n",
    "    - [Predictions and AUC for shared pathways or interactions](#auc)\n",
    "    - [Tests for querying to recover related genes](#y)\n",
    "    - [Producing output summary table](#output)\n",
    "\n",
    "- [Part 5. Clustering Analysis](#part_6)\n",
    "    - [Topic modeling](#topic_modeling)\n",
    "    - [Agglomerative clustering](#clustering)\n",
    "    - [Phenologs for OMIM disease phenotypes](#phenologs)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"introduction\"></a>\n",
    "### Introduction: Text Mining Analysis of Phenotype Descriptions in Plants\n",
    "The purpose of this notebook is to evaluate what can be learned from a natural language processing approach to analyzing free-text descriptions of phenotype descriptions of plants. The approach is to generate pairwise distances matrices between a set of plant phenotype descriptions across different species, sourced from academic papers and online model organism databases. These pairwise distance matrices can be constructed using any vectorization method that can be applied to natural language. In this notebook, we specifically evaluate the use of n-gram and bag-of-words techniques, word and document embedding using Word2Vec and Doc2Vec, context-dependent word-embeddings using BERT and BioBERT, and ontology term annotations with automated annotation tools such as NOBLE Coder.\n",
    "\n",
    "Loading, manipulation, and filtering of the dataset of phenotype descriptions associated with genes across different plant species is largely handled through a Python package created for this purpose called OATS (Ontology Annotation and Text Similarity) which is available [here](https://github.com/irbraun/oats). Preprocessing of the descriptions, mapping the dataset to additional resources such as protein-protein interaction databases and biochemical pathway databases are handled in this notebook using that package as well. In the evaluation of each of these natural language processing approaches to analyzing this dataset of descriptions, we compare performance against a dataset generated through manual annotation of a similar dataset in Oellrich Walls et al. (2015) and against manual annotations with experimentally determined terms from the Gene Ontology (PO) and the Plant Ontology (PO).\n",
    "\n",
    "<a id=\"links\"></a>\n",
    "### Relevant links of interest:\n",
    "- Paper describing comparison of NLP and ontology annotation approaches to curation: [Braun, Lawrence-Dill (2019)](https://doi.org/10.3389/fpls.2019.01629)\n",
    "- Paper describing results of manual phenotype description curation: [Oellrich, Walls et al. (2015](https://plantmethods.biomedcentral.com/articles/10.1186/s13007-015-0053-y)\n",
    "- Plant databases with phenotype description text data available: [TAIR](https://www.arabidopsis.org/), [SGN](https://solgenomics.net/), [MaizeGDB](https://www.maizegdb.org/)\n",
    "- Python package for working with phenotype descriptions: [OATS](https://github.com/irbraun/oats)\n",
    "- Python package used for general NLP functions: [NLTK](https://www.nltk.org/), [Gensim](https://radimrehurek.com/gensim/auto_examples/index.html)\n",
    "- Python package used for working with biological ontologies: [Pronto](https://pronto.readthedocs.io/en/latest/)\n",
    "- Python package for loading pretrained BERT models: [PyTorch Pretrained BERT](https://pypi.org/project/pytorch-pretrained-bert/)\n",
    "- For BERT Models pretrained on PubMed and PMC: [BioBERT Paper](https://arxiv.org/abs/1901.08746), [BioBERT Models](https://github.com/naver/biobert-pretrained)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warming up PyWSD (takes ~10 secs)... took 4.652247667312622 secs.\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "import torch\n",
    "import itertools\n",
    "import argparse\n",
    "import multiprocessing as mp\n",
    "from collections import Counter, defaultdict\n",
    "from inspect import signature\n",
    "from scipy.stats import ks_2samp, hypergeom, pearsonr, spearmanr\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, auc\n",
    "from sklearn.model_selection import train_test_split, KFold\n",
    "from scipy import spatial, stats\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "from pytorch_pretrained_bert import BertTokenizer, BertModel, BertForMaskedLM\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer\n",
    "from sklearn.decomposition import NMF\n",
    "from sklearn.decomposition import LatentDirichletAllocation as LDA\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "\n",
    "from nltk.corpus import brown, stopwords\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "nltk.download('stopwords', quiet=True)\n",
    "nltk.download('wordnet', quiet=True)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.utils.utils import function_wrapper_with_duration\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.annotation.annotation import annotate_using_noble_coder\n",
    "from oats.distances import pairwise as pw\n",
    "from oats.distances.edgelists import merge_edgelists, make_undirected, remove_self_loops, subset_with_ids\n",
    "from oats.nlp.vocabulary import get_overrepresented_tokens, get_vocab_from_tokens\n",
    "from oats.nlp.vocabulary import reduce_vocab_connected_components, reduce_vocab_linares_pontes\n",
    "from oats.nlp.preprocess import concatenate_with_bar_delim\n",
    "\n",
    "from _utils import Method\n",
    "from _utils import IndexedGraph\n",
    "\n",
    "mpl.rcParams[\"figure.dpi\"] = 400\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_1\"></a>\n",
    "# Part 1. Loading and Filtering Data\n",
    "<a id=\"paths\"></a>\n",
    "### Setting up the input and output paths and summarizing output table\n",
    "This section defines some constants which are used for creating a uniquely named directory to contain all the outputs from running this instance of this notebook. The naming scheme is based on the time that the notebook is run. The other constants are used for specifying information in the output table about what the topic was for this notebook when it was run, such as looking at KEGG biochemical pathways or STRING protein-protein interaction data some other type of gene function grouping or hierarchy. These values are arbitrary and are just for keeping better notes about what the output of the notebook corresponds to. All the input and output file paths for loading datasets or models are also contained within this cell, so that if anything is moved the directories and file names should only have to be changed at this point and nowhere else further into the notebook. If additional files are added to the notebook cells they should be put here as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create and name an output directory according to when the notebooks or script was run.\n",
    "OUTPUT_DIR = os.path.join(\"../outputs\",datetime.datetime.now().strftime('%m_%d_%Y_h%Hm%Ms%S'))\n",
    "os.mkdir(OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dataset_filename = \"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\"          # The full dataset pickle.\n",
    "kegg_pathways_filename = \"../data/pickles/groupings_from_kegg_pathways.pickle\"                       # The pathway groupings from KEGG.\n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                         # The pahway groupings from Plant Metabolic Network.\n",
    "lloyd_subsets_filename = \"../data/pickles/groupings_from_lloyd_subsets.pickle\"                       # The functional subsets defined by Lloyd and Meinke (2012).\n",
    "lloyd_classes_filename = \"../data/pickles/groupings_from_lloyd_classes.pickle\"                       # The functional classes defined by Lloyd and Meinke (2012).\n",
    "background_corpus_filename = \"../data/corpus_related_files/untagged_text_corpora/background.txt\"     # Text file with background content.\n",
    "phenotypes_corpus_filename = \"../data/corpus_related_files/untagged_text_corpora/phenotypes_all.txt\" # Text file with specific content.\n",
    "doc2vec_pubmed_filename = \"../gensim/pubmed_dbow/doc2vec_2.bin\"                                      # File holding saved Doc2Vec model trained on PubMed.\n",
    "doc2vec_wikipedia_filename = \"../gensim/enwiki_dbow/doc2vec.bin\"                                     # File holding saved Doc2Vec model trained on Wikipedia.\n",
    "word2vec_model_filename = \"../gensim/wiki_sg/word2vec.bin\"                                           # File holding saved Word2Vec model trained on Wikipedia.\n",
    "go_filename = \"../ontologies/go.obo\"                                                                 # Gene Ontology file in OBO format.\n",
    "po_filename = \"../ontologies/po.obo\"                                                                 # Plant Ontology file in OBO format.\n",
    "pato_filename = \"../ontologies/pato.obo\"                                                             # Phenotype and Trait Ontology file in OBO format.\n",
    "noblecoder_jarfile_path = \"../lib/NobleCoder-1.0.jar\"                                                # Jar for NOBLE Coder annotation tool.\n",
    "biobert_pmc_path = \"../gensim/biobert_v1.0_pmc/pytorch_model\"                                        # Path for PyTorch BioBERT model.\n",
    "biobert_pubmed_path = \"../gensim/biobert_v1.0_pubmed/pytorch_model\"                                  # Path for PyTorch BioBERT model.\n",
    "biobert_pubmed_pmc_path = \"../gensim/biobert_v1.0_pubmed_pmc/pytorch_model\"                          # Path for PyTorch BioBERT model.\n",
    "panther_to_omim_filename = \"../data/orthology_related_files/ath_to_hsa/pantherdb_omim_df.csv\"        # File with mappings to human orthologs and disease phenotypes.\n",
    "pppn_edgelist_path = \"../data/supplemental_files_oellrich_walls/13007_2015_53_MOESM9_ESM.txt\"\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"read_this_data\"></a>\n",
    "### Reading in the dataset of genes and their associated phenotype descriptions and annotations\n",
    "The actual choice of pickle that is loaded here depends on the objective that is being used for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6364</td>\n",
       "      <td>3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1406</td>\n",
       "      <td>811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7999</td>\n",
       "      <td>4839</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath       6364                 3813\n",
       "1     gmx         30                   24\n",
       "2     mtr         37                   36\n",
       "3     osa         92                   85\n",
       "4     sly         70                   70\n",
       "5     zma       1406                  811\n",
       "6   total       7999                 4839"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = load_from_pickle(dataset_filename)\n",
    "dataset.filter_has_description()\n",
    "dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"relating\"></a>\n",
    "### Relating the dataset of genes to the dataset of groupings or categories\n",
    "This section generates tables that indicate how the genes present in the dataset were mapped to the defined pathways or groups. This includes a summary table that indicates how many genes by species were succcessfully mapped to atleast one pathway or group, as well as a more detailed table describing how many genes from each species were mapped to each particular pathway or group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_in_groupings_object_and_write_summary_tables(dataset, groupings_filename, name):\n",
    "\n",
    "    # Load the groupings object.\n",
    "    groups = load_from_pickle(groupings_filename)\n",
    "    id_to_group_ids = groups.get_id_to_group_ids_dict(dataset.get_gene_dictionary())\n",
    "    group_id_to_ids = groups.get_group_id_to_ids_dict(dataset.get_gene_dictionary())\n",
    "    group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "    groups.to_csv(os.path.join(OUTPUT_DIR,\"part_1_{}_groupings.csv\".format(name)))\n",
    "\n",
    "    # Generate a table describing how many of the genes input from each species map to atleast one group.\n",
    "    summary = defaultdict(dict)\n",
    "    species_dict = dataset.get_species_dictionary()\n",
    "    for species in dataset.get_species():\n",
    "        summary[species][\"input\"] = len([x for x in dataset.get_ids() if species_dict[x]==species])\n",
    "        summary[species][\"mapped\"] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "    table = pd.DataFrame(summary).transpose()\n",
    "    table.loc[\"total\"]= table.sum()\n",
    "    table[\"fraction\"] = table.apply(lambda row: \"{:0.4f}\".format(row[\"mapped\"]/row[\"input\"]), axis=1)\n",
    "    table = table.reset_index(inplace=False)\n",
    "    table = table.rename({\"index\":\"species\"}, axis=\"columns\")\n",
    "    table.to_csv(os.path.join(OUTPUT_DIR,\"part_1_{}_mappings_summary.csv\".format(name)), index=False)\n",
    "\n",
    "    \n",
    "    # Generate a table describing how many genes from each species map to which particular group.\n",
    "    summary = defaultdict(dict)\n",
    "    for group_id,ids in group_id_to_ids.items():\n",
    "        summary[group_id].update({species:len([x for x in ids if species_dict[x]==species]) for species in dataset.get_species()})\n",
    "        summary[group_id][\"total\"] = len([x for x in ids])\n",
    "    table = pd.DataFrame(summary).transpose()\n",
    "    table = table.sort_values(by=\"total\", ascending=False)\n",
    "    table = table.reset_index(inplace=False)\n",
    "    table = table.rename({\"index\":\"pathway_id\"}, axis=\"columns\")\n",
    "    table[\"pathway_name\"] = table[\"pathway_id\"].map(groups.get_long_name)\n",
    "    table.loc[\"total\"] = table.sum()\n",
    "    table.loc[\"total\",\"pathway_id\"] = \"total\"\n",
    "    table.loc[\"total\",\"pathway_name\"] = \"total\"\n",
    "    table = table[table.columns.tolist()[-1:] + table.columns.tolist()[:-1]]\n",
    "    table.to_csv(os.path.join(OUTPUT_DIR,\"part_1_{}_mappings_by_group.csv\".format(name)), index=False)\n",
    "    \n",
    "    \n",
    "    # What are the similarites between the groups for the genes present in this dataset?\n",
    "    group_sims = defaultdict(dict)\n",
    "    for group_id_1,ids_1 in group_id_to_ids.items():\n",
    "        for group_id_2,ids_2 in group_id_to_ids.items():\n",
    "            jaccard_sim = len(set(ids_1).intersection(set(ids_2)))/len(set(ids_1).union(set(ids_2)))\n",
    "            group_sims[group_id_1][group_id_2] = jaccard_sim\n",
    "    table = pd.DataFrame(group_sims)\n",
    "    \n",
    "    \n",
    "    # Changing the order of the Lloyd, Meinke phenotype subsets to match other figures for consistency, special case.\n",
    "    if name == \"subsets\":\n",
    "        filename = \"../data/group_related_files/lloyd/lloyd_function_hierarchy_irb_cleaned.csv\"\n",
    "        lmtm_df = pd.read_csv(filename)    \n",
    "        subsets_in_order = [col for col in lmtm_df[\"Subset Symbol\"].values if col in table.columns]\n",
    "        table = table[subsets_in_order]\n",
    "        table = table.reindex(subsets_in_order)\n",
    "        \n",
    "        \n",
    "    # Formatting the column names for this table correctly and outputting to a file.\n",
    "    table = table.reset_index(drop=False).rename({\"index\":\"group\"},axis=1).reset_index(drop=False).rename({\"index\":\"order\"},axis=1)\n",
    "    table.to_csv(os.path.join(OUTPUT_DIR,\"part_1_{}_similarity_matrix.csv\".format(name)), index=False)\n",
    "    \n",
    "\n",
    "    # Returning the groupings object and the list of IDs for genes that were mapped to one or more groups.\n",
    "    return(groups, group_mapped_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathways from KEGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>ncbi_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>ko_number</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>hkl3|hexokinase-like 3</td>\n",
       "      <td>at4g37840</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00844</td>\n",
       "      <td>EC:2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>hxk3|hexokinase 3</td>\n",
       "      <td>at1g47840</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00844</td>\n",
       "      <td>EC:2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>hxk2|hexokinase 2</td>\n",
       "      <td>at2g19860</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00844</td>\n",
       "      <td>EC:2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>hxk1|hexokinase 1</td>\n",
       "      <td>at4g29130</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00844</td>\n",
       "      <td>EC:2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>hkl1|hexokinase-like 1</td>\n",
       "      <td>at1g50460</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00844</td>\n",
       "      <td>EC:2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>athxk4|hexokinase</td>\n",
       "      <td>at3g20040</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00844</td>\n",
       "      <td>EC:2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pgi1|phosphoglucose isomerase 1</td>\n",
       "      <td>at4g24620</td>\n",
       "      <td></td>\n",
       "      <td>KO:K01810</td>\n",
       "      <td>EC:5.3.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>sugar isomerase (sis) family protein</td>\n",
       "      <td>at5g42740</td>\n",
       "      <td></td>\n",
       "      <td>KO:K01810</td>\n",
       "      <td>EC:5.3.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk6|phosphofructokinase 6</td>\n",
       "      <td>at4g32840</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk7|phosphofructokinase 7</td>\n",
       "      <td>at5g56630</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk4|phosphofructokinase 4</td>\n",
       "      <td>at5g61580</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk1|phosphofructokinase 1</td>\n",
       "      <td>at4g29220</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk5|phosphofructokinase 5</td>\n",
       "      <td>at2g22480</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk3|phosphofructokinase 3</td>\n",
       "      <td>at4g26270</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>pfk2|phosphofructokinase 2</td>\n",
       "      <td>at5g47810</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00850</td>\n",
       "      <td>EC:2.7.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>phosphofructokinase family protein</td>\n",
       "      <td>at1g76550</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00895</td>\n",
       "      <td>EC:2.7.1.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>mee51|phosphofructokinase family protein</td>\n",
       "      <td>at4g04040</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00895</td>\n",
       "      <td>EC:2.7.1.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>phosphofructokinase family protein</td>\n",
       "      <td>at1g12000</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00895</td>\n",
       "      <td>EC:2.7.1.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>phosphofructokinase family protein</td>\n",
       "      <td>at1g20950</td>\n",
       "      <td></td>\n",
       "      <td>KO:K00895</td>\n",
       "      <td>EC:2.7.1.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>ko00010</td>\n",
       "      <td>path:ath00010\\tGlycolysis / Gluconeogenesis - ...</td>\n",
       "      <td>fbp|inositol monophosphatase family protein</td>\n",
       "      <td>at1g43670</td>\n",
       "      <td></td>\n",
       "      <td>KO:K03841</td>\n",
       "      <td>EC:3.1.3.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species pathway_id                                       pathway_name                                   gene_names    ncbi_id uniprot_id  ko_number    ec_number\n",
       "0      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                       hkl3|hexokinase-like 3  at4g37840             KO:K00844   EC:2.7.1.1\n",
       "1      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                            hxk3|hexokinase 3  at1g47840             KO:K00844   EC:2.7.1.1\n",
       "2      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                            hxk2|hexokinase 2  at2g19860             KO:K00844   EC:2.7.1.1\n",
       "3      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                            hxk1|hexokinase 1  at4g29130             KO:K00844   EC:2.7.1.1\n",
       "4      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                       hkl1|hexokinase-like 1  at1g50460             KO:K00844   EC:2.7.1.1\n",
       "5      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                            athxk4|hexokinase  at3g20040             KO:K00844   EC:2.7.1.1\n",
       "6      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...              pgi1|phosphoglucose isomerase 1  at4g24620             KO:K01810   EC:5.3.1.9\n",
       "7      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...         sugar isomerase (sis) family protein  at5g42740             KO:K01810   EC:5.3.1.9\n",
       "8      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk6|phosphofructokinase 6  at4g32840             KO:K00850  EC:2.7.1.11\n",
       "9      ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk7|phosphofructokinase 7  at5g56630             KO:K00850  EC:2.7.1.11\n",
       "10     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk4|phosphofructokinase 4  at5g61580             KO:K00850  EC:2.7.1.11\n",
       "11     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk1|phosphofructokinase 1  at4g29220             KO:K00850  EC:2.7.1.11\n",
       "12     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk5|phosphofructokinase 5  at2g22480             KO:K00850  EC:2.7.1.11\n",
       "13     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk3|phosphofructokinase 3  at4g26270             KO:K00850  EC:2.7.1.11\n",
       "14     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...                   pfk2|phosphofructokinase 2  at5g47810             KO:K00850  EC:2.7.1.11\n",
       "15     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...           phosphofructokinase family protein  at1g76550             KO:K00895  EC:2.7.1.90\n",
       "16     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...     mee51|phosphofructokinase family protein  at4g04040             KO:K00895  EC:2.7.1.90\n",
       "17     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...           phosphofructokinase family protein  at1g12000             KO:K00895  EC:2.7.1.90\n",
       "18     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...           phosphofructokinase family protein  at1g20950             KO:K00895  EC:2.7.1.90\n",
       "19     ath    ko00010  path:ath00010\\tGlycolysis / Gluconeogenesis - ...  fbp|inositol monophosphatase family protein  at1g43670             KO:K03841  EC:3.1.3.11"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Readin in the dataset of groupings for pathways in KEGG.\n",
    "kegg_groups, kegg_mapped_ids = read_in_groupings_object_and_write_summary_tables(dataset, kegg_pathways_filename, \"kegg\")\n",
    "kegg_groups.to_pandas().head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathways from PlantCyc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at1g52400-monomer|abscisic acid glucose ester ...</td>\n",
       "      <td>EC-3.2.1.175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at4g15550-monomer|abscisate &amp;beta;-glucosyltra...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at4g15260-monomer|abscisate &amp;beta;-glucosyltra...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at3g21790-monomer|abscisate &amp;beta;-glucosyltra...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at3g21760-monomer|abscisate &amp;beta;-glucosyltra...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at2g23210-monomer|abscisate &amp;beta;-glucosyltra...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at1g05530-monomer|abscisic acid glucosyltransf...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at1g05560-monomer|abscisic acid glucosyltransf...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at4g34138-monomer|abscisic acid glucosyltransf...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at2g23250-monomer|abscisic acid glucosyltransf...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at2g23260-monomer|abscisic acid glucosyltransf...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWY-5272</td>\n",
       "      <td>abscisic acid degradation by glucosylation</td>\n",
       "      <td>at3g21780-monomer|abscisic acid glycosyltransf...</td>\n",
       "      <td>EC-2.4.1.263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at1g09420-monomer|glucose-6-phosphate dehydrog...</td>\n",
       "      <td>EC-1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at5g35790-monomer|glucose-6-phosphate 1-dehydr...</td>\n",
       "      <td>EC-1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at5g13110-monomer|glucose-6-phosphate 1-dehydr...</td>\n",
       "      <td>EC-1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at5g40760-monomer|glucose-6-phosphate 1-dehydr...</td>\n",
       "      <td>EC-1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at3g27300-monomer|glucose-6-phosphate 1-dehydr...</td>\n",
       "      <td>EC-1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at1g24280-monomer|glucose-6-phosphate 1-dehydr...</td>\n",
       "      <td>EC-1.1.1.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at5g41670-monomer|phosphogluconate dehydrogena...</td>\n",
       "      <td>EC-1.1.1.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>OXIDATIVEPENT-PWY</td>\n",
       "      <td>pentose phosphate pathway (oxidative branch) I</td>\n",
       "      <td>at1g64190-monomer|phosphogluconate dehydrogena...</td>\n",
       "      <td>EC-1.1.1.44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species         pathway_id                                    pathway_name                                         gene_names     ec_number\n",
       "0      ath           PWY-5272      abscisic acid degradation by glucosylation  at1g52400-monomer|abscisic acid glucose ester ...  EC-3.2.1.175\n",
       "1      ath           PWY-5272      abscisic acid degradation by glucosylation  at4g15550-monomer|abscisate &beta;-glucosyltra...  EC-2.4.1.263\n",
       "2      ath           PWY-5272      abscisic acid degradation by glucosylation  at4g15260-monomer|abscisate &beta;-glucosyltra...  EC-2.4.1.263\n",
       "3      ath           PWY-5272      abscisic acid degradation by glucosylation  at3g21790-monomer|abscisate &beta;-glucosyltra...  EC-2.4.1.263\n",
       "4      ath           PWY-5272      abscisic acid degradation by glucosylation  at3g21760-monomer|abscisate &beta;-glucosyltra...  EC-2.4.1.263\n",
       "5      ath           PWY-5272      abscisic acid degradation by glucosylation  at2g23210-monomer|abscisate &beta;-glucosyltra...  EC-2.4.1.263\n",
       "6      ath           PWY-5272      abscisic acid degradation by glucosylation  at1g05530-monomer|abscisic acid glucosyltransf...  EC-2.4.1.263\n",
       "7      ath           PWY-5272      abscisic acid degradation by glucosylation  at1g05560-monomer|abscisic acid glucosyltransf...  EC-2.4.1.263\n",
       "8      ath           PWY-5272      abscisic acid degradation by glucosylation  at4g34138-monomer|abscisic acid glucosyltransf...  EC-2.4.1.263\n",
       "9      ath           PWY-5272      abscisic acid degradation by glucosylation  at2g23250-monomer|abscisic acid glucosyltransf...  EC-2.4.1.263\n",
       "10     ath           PWY-5272      abscisic acid degradation by glucosylation  at2g23260-monomer|abscisic acid glucosyltransf...  EC-2.4.1.263\n",
       "11     ath           PWY-5272      abscisic acid degradation by glucosylation  at3g21780-monomer|abscisic acid glycosyltransf...  EC-2.4.1.263\n",
       "12     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at1g09420-monomer|glucose-6-phosphate dehydrog...   EC-1.1.1.49\n",
       "13     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at5g35790-monomer|glucose-6-phosphate 1-dehydr...   EC-1.1.1.49\n",
       "14     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at5g13110-monomer|glucose-6-phosphate 1-dehydr...   EC-1.1.1.49\n",
       "15     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at5g40760-monomer|glucose-6-phosphate 1-dehydr...   EC-1.1.1.49\n",
       "16     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at3g27300-monomer|glucose-6-phosphate 1-dehydr...   EC-1.1.1.49\n",
       "17     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at1g24280-monomer|glucose-6-phosphate 1-dehydr...   EC-1.1.1.49\n",
       "18     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at5g41670-monomer|phosphogluconate dehydrogena...   EC-1.1.1.44\n",
       "19     ath  OXIDATIVEPENT-PWY  pentose phosphate pathway (oxidative branch) I  at1g64190-monomer|phosphogluconate dehydrogena...   EC-1.1.1.44"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading in the dataset of groupings for pathways in PlantCyc.\n",
    "pmn_groups, pmn_mapped_ids = read_in_groupings_object_and_write_summary_tables(dataset, pmn_pathways_filename, \"pmn\")\n",
    "pmn_groups.to_pandas().head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype subsets from Lloyd and Meinke (2012)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>group_id</th>\n",
       "      <th>gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>FSM</td>\n",
       "      <td>at1g01030|nga3|top1|ngatha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>EMB|FSM|OVP|SRF</td>\n",
       "      <td>at1g01040|sus1|dcl1|sin1|caf|abnormal suspensor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>CDR|LIT</td>\n",
       "      <td>at1g01060|lhy|late elongated hypocotyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>IST|WAT</td>\n",
       "      <td>at1g01120|kcs1|3-ketoacyl-coa synthase defective</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>OVP|SRF</td>\n",
       "      <td>at1g01280|cyp703a2|cytochrome p450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>EMB</td>\n",
       "      <td>at1g01370|cenh3|centromere-specific histone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>CHS</td>\n",
       "      <td>at1g01460|pipk11|phosphatidylinositol phosphat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>NLS|GRS|IST</td>\n",
       "      <td>at1g01480|acs2|aminocyclopropane carboxylate s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>LEF|FSM</td>\n",
       "      <td>at1g01510|an|angustifolia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>SRL|ROT|LEF|MSL|STT|RTH|TCM|TMP</td>\n",
       "      <td>at1g01550|bps1|bypass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>SRF</td>\n",
       "      <td>at1g01690|prd3|putative recombination initiati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>TMP</td>\n",
       "      <td>at1g01860|pfc1|paleface</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>ROT</td>\n",
       "      <td>at1g01950|ark2|armadillo repeat kinesin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>OVP</td>\n",
       "      <td>at1g02050|lap6|pksa|less adhesive pollen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>SRF</td>\n",
       "      <td>at1g02065|spl8|squamosa promoter binding prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>PIG|LIT</td>\n",
       "      <td>at1g02090|fus5|cop15|fusca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>MSL|PTH</td>\n",
       "      <td>at1g02120|vad1|vascular-associated death</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>GAM</td>\n",
       "      <td>at1g02140|hap1|mago|mee63|hapless</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>IST|FSM|WAT</td>\n",
       "      <td>at1g02205|cer1|eceriferum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>PIG</td>\n",
       "      <td>at1g02280|ppi1|toc33|plastid protein import</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species                         group_id                                         gene_names\n",
       "0      ath                              FSM                         at1g01030|nga3|top1|ngatha\n",
       "1      ath                  EMB|FSM|OVP|SRF    at1g01040|sus1|dcl1|sin1|caf|abnormal suspensor\n",
       "2      ath                          CDR|LIT             at1g01060|lhy|late elongated hypocotyl\n",
       "3      ath                          IST|WAT   at1g01120|kcs1|3-ketoacyl-coa synthase defective\n",
       "4      ath                          OVP|SRF                 at1g01280|cyp703a2|cytochrome p450\n",
       "5      ath                              EMB        at1g01370|cenh3|centromere-specific histone\n",
       "6      ath                              CHS  at1g01460|pipk11|phosphatidylinositol phosphat...\n",
       "7      ath                      NLS|GRS|IST  at1g01480|acs2|aminocyclopropane carboxylate s...\n",
       "8      ath                          LEF|FSM                          at1g01510|an|angustifolia\n",
       "9      ath  SRL|ROT|LEF|MSL|STT|RTH|TCM|TMP                              at1g01550|bps1|bypass\n",
       "10     ath                              SRF  at1g01690|prd3|putative recombination initiati...\n",
       "11     ath                              TMP                            at1g01860|pfc1|paleface\n",
       "12     ath                              ROT            at1g01950|ark2|armadillo repeat kinesin\n",
       "13     ath                              OVP           at1g02050|lap6|pksa|less adhesive pollen\n",
       "14     ath                              SRF  at1g02065|spl8|squamosa promoter binding prote...\n",
       "15     ath                          PIG|LIT                         at1g02090|fus5|cop15|fusca\n",
       "16     ath                          MSL|PTH           at1g02120|vad1|vascular-associated death\n",
       "17     ath                              GAM                  at1g02140|hap1|mago|mee63|hapless\n",
       "18     ath                      IST|FSM|WAT                          at1g02205|cer1|eceriferum\n",
       "19     ath                              PIG        at1g02280|ppi1|toc33|plastid protein import"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading in the datasets of phenotype subset classifications from the Lloyd, Meinke 2012 paper.\n",
    "phe_subsets_groups, subsets_mapped_ids = read_in_groupings_object_and_write_summary_tables(dataset, lloyd_subsets_filename, \"subsets\")\n",
    "phe_subsets_groups.to_pandas().head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype subsets from Lloyd and Meinke (2012)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>group_id</th>\n",
       "      <th>gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>R</td>\n",
       "      <td>at1g01030|nga3|top1|ngatha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>S</td>\n",
       "      <td>at1g01040|sus1|dcl1|sin1|caf|abnormal suspensor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>T</td>\n",
       "      <td>at1g01060|lhy|late elongated hypocotyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g01120|kcs1|3-ketoacyl-coa synthase defective</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>R</td>\n",
       "      <td>at1g01280|cyp703a2|cytochrome p450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>S</td>\n",
       "      <td>at1g01370|cenh3|centromere-specific histone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>H</td>\n",
       "      <td>at1g01460|pipk11|phosphatidylinositol phosphat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g01480|acs2|aminocyclopropane carboxylate s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g01510|an|angustifolia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>L</td>\n",
       "      <td>at1g01550|bps1|bypass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>R</td>\n",
       "      <td>at1g01690|prd3|putative recombination initiati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>P</td>\n",
       "      <td>at1g01860|pfc1|paleface</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g01950|ark2|armadillo repeat kinesin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>R</td>\n",
       "      <td>at1g02050|lap6|pksa|less adhesive pollen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>R</td>\n",
       "      <td>at1g02065|spl8|squamosa promoter binding prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g02090|fus5|cop15|fusca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g02120|vad1|vascular-associated death</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>G</td>\n",
       "      <td>at1g02140|hap1|mago|mee63|hapless</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g02205|cer1|eceriferum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>V</td>\n",
       "      <td>at1g02280|ppi1|toc33|plastid protein import</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species group_id                                         gene_names\n",
       "0      ath        R                         at1g01030|nga3|top1|ngatha\n",
       "1      ath        S    at1g01040|sus1|dcl1|sin1|caf|abnormal suspensor\n",
       "2      ath        T             at1g01060|lhy|late elongated hypocotyl\n",
       "3      ath        V   at1g01120|kcs1|3-ketoacyl-coa synthase defective\n",
       "4      ath        R                 at1g01280|cyp703a2|cytochrome p450\n",
       "5      ath        S        at1g01370|cenh3|centromere-specific histone\n",
       "6      ath        H  at1g01460|pipk11|phosphatidylinositol phosphat...\n",
       "7      ath        V  at1g01480|acs2|aminocyclopropane carboxylate s...\n",
       "8      ath        V                          at1g01510|an|angustifolia\n",
       "9      ath        L                              at1g01550|bps1|bypass\n",
       "10     ath        R  at1g01690|prd3|putative recombination initiati...\n",
       "11     ath        P                            at1g01860|pfc1|paleface\n",
       "12     ath        V            at1g01950|ark2|armadillo repeat kinesin\n",
       "13     ath        R           at1g02050|lap6|pksa|less adhesive pollen\n",
       "14     ath        R  at1g02065|spl8|squamosa promoter binding prote...\n",
       "15     ath        V                         at1g02090|fus5|cop15|fusca\n",
       "16     ath        V           at1g02120|vad1|vascular-associated death\n",
       "17     ath        G                  at1g02140|hap1|mago|mee63|hapless\n",
       "18     ath        V                          at1g02205|cer1|eceriferum\n",
       "19     ath        V        at1g02280|ppi1|toc33|plastid protein import"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading in the datasets of phenotype class classifications from the Lloyd, Meinke 2012 paper.\n",
    "phe_classes_groups, classes_mapped_ids = read_in_groupings_object_and_write_summary_tables(dataset, lloyd_classes_filename, \"classes\")\n",
    "phe_classes_groups.to_pandas().head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"filtering\"></a>\n",
    "### Relating pairs of genes to information about network edges from other sources\n",
    "This is done to only include genes (and the corresponding phenotype descriptions and annotations) which are useful for the current analysis. In this case we want to only retain genes that are mentioned atleast one time in the STRING database for a given species. If a gene is not mentioned at all in STRING, there is no information available for whether or not it interacts with any other proteins in the dataset so choose to not include it in the analysis. Only genes that have atleast one true positive are included because these are the only ones for which the missing information (negatives) is meaningful. This should be run instead of the subsequent cell, or the other way around, based on whether or not protein-protein interactions is the prediction goal for the current analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EQ-based similarities from Oellrich, Walls et al., (2015)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5959</td>\n",
       "      <td>1799</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5959</td>\n",
       "      <td>1840</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5959</td>\n",
       "      <td>443</td>\n",
       "      <td>0.926471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5959</td>\n",
       "      <td>1766</td>\n",
       "      <td>0.516393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5959</td>\n",
       "      <td>1838</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5959</td>\n",
       "      <td>2035</td>\n",
       "      <td>0.954545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5959</td>\n",
       "      <td>2059</td>\n",
       "      <td>0.417219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5959</td>\n",
       "      <td>1795</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5959</td>\n",
       "      <td>1894</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5959</td>\n",
       "      <td>6227</td>\n",
       "      <td>0.900000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   from    to     value\n",
       "0  5959  1799  1.000000\n",
       "1  5959  1840  1.000000\n",
       "2  5959   443  0.926471\n",
       "3  5959  1766  0.516393\n",
       "4  5959  1838  1.000000\n",
       "5  5959  2035  0.954545\n",
       "6  5959  2059  0.417219\n",
       "7  5959  1795  1.000000\n",
       "8  5959  1894  1.000000\n",
       "9  5959  6227  0.900000"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ow_edgelist = AnyInteractions(dataset.get_name_to_id_dictionary(), pppn_edgelist_path)\n",
    "ow_edgelist.df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein associations from STRING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>known_associations</th>\n",
       "      <th>predicted_associations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>73.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>73.0</td>\n",
       "      <td>948.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>73.0</td>\n",
       "      <td>4925.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>73.0</td>\n",
       "      <td>4488.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>73.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>73.0</td>\n",
       "      <td>4275.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>73.0</td>\n",
       "      <td>473.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>73.0</td>\n",
       "      <td>5461.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>73.0</td>\n",
       "      <td>5818.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>73.0</td>\n",
       "      <td>4542.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     from      to  known_associations  predicted_associations\n",
       "346  73.0    49.0                 161                       0\n",
       "348  73.0   948.0                   0                       0\n",
       "351  73.0  4925.0                   0                       0\n",
       "352  73.0  4488.0                   0                       0\n",
       "353  73.0   132.0                   0                       0\n",
       "355  73.0  4275.0                   0                       0\n",
       "357  73.0   473.0                   0                       0\n",
       "358  73.0  5461.0                   0                       0\n",
       "359  73.0  5818.0                   0                       0\n",
       "360  73.0  4542.0                   0                       0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # Maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # Tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # Medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # Rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # Soybean\n",
    "]\n",
    "genes = dataset.get_gene_dictionary()\n",
    "string_edgelist = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "string_edgelist.df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Orthologous genes from PANTHER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>5730.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179180</th>\n",
       "      <td>1898.0</td>\n",
       "      <td>1898.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190004</th>\n",
       "      <td>3038.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192404</th>\n",
       "      <td>1685.0</td>\n",
       "      <td>1867.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198820</th>\n",
       "      <td>3864.0</td>\n",
       "      <td>4708.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413820</th>\n",
       "      <td>3880.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415020</th>\n",
       "      <td>2564.0</td>\n",
       "      <td>1155.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431292</th>\n",
       "      <td>3652.0</td>\n",
       "      <td>2225.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470160</th>\n",
       "      <td>2552.0</td>\n",
       "      <td>575.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470668</th>\n",
       "      <td>1017.0</td>\n",
       "      <td>1130.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          from      to  value\n",
       "692     5730.0    23.0    1.0\n",
       "179180  1898.0  1898.0    1.0\n",
       "190004  3038.0  2411.0    1.0\n",
       "192404  1685.0  1867.0    1.0\n",
       "198820  3864.0  4708.0    1.0\n",
       "413820  3880.0   427.0    1.0\n",
       "415020  2564.0  1155.0    1.0\n",
       "431292  3652.0  2225.0    1.0\n",
       "470160  2552.0   575.0    1.0\n",
       "470668  1017.0  1130.0    1.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panther_edgelist = AnyInteractions(dataset.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "panther_edgelist.df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsetting the dataset to include only genes with relevance to any of the biological questions\n",
    "This is done to only include genes (and the corresponding phenotype descriptions and annotations) which are useful for the current analysis. In this case we want to only retain genes that are mapped to atleast one pathway in whatever the source of pathway membership we are using is (KEGG, Plant Metabolic Network, etc). This is because for these genes, it will be impossible to correctly predict their pathway membership, and we have no evidence that they belong or do not belong in certain pathways so they can not be identified as being true or false negatives in any case. This should not actually be necessary if the dataset used to start the notebook analysis has already be subset for just the genes that either have pathway information of phenotype classification information, this should just be used to double check that the numbers make sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the list of all the IDs in this dataset that have any relevant mapping at all to the biological questions.\n",
    "ids_with_any_mapping = list(set(flatten([\n",
    "    kegg_mapped_ids,\n",
    "    pmn_mapped_ids,\n",
    "    subsets_mapped_ids,\n",
    "    classes_mapped_ids,\n",
    "    string_edgelist.ids,\n",
    "    panther_edgelist.ids\n",
    "])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the list of all the IDs in this dataset that have all of types of curated values we want to look at. \n",
    "annots = dataset.get_annotations_dictionary()\n",
    "go_mapped_ids = [i for i in dataset.get_ids() if \"GO\" in annots[i]]\n",
    "po_mapped_ids = [i for i in dataset.get_ids() if \"PO\" in annots[i]]\n",
    "ids_with_all_annotations = list(set(flatten([\n",
    "    go_mapped_ids,\n",
    "    po_mapped_ids,\n",
    "    ow_edgelist.ids\n",
    "])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>446</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mtr</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>osa</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sly</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zma</td>\n",
       "      <td>38</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>total</td>\n",
       "      <td>500</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath        446                  387\n",
       "1     mtr          1                    1\n",
       "2     osa         13                   12\n",
       "3     sly          2                    2\n",
       "4     zma         38                   37\n",
       "5   total        500                  439"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.filter_with_ids(ids_with_any_mapping)\n",
    "dataset.filter_random_k(500)\n",
    "dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_2\"></a>\n",
    "# Part 2. NLP Models\n",
    "\n",
    "\n",
    "<a id=\"word2vec_doc2vec\"></a>\n",
    "### Word2Vec and Doc2Vec\n",
    "Word2Vec is a word embedding technique using a neural network trained on a so-called *false task*, namely either predicting a missing word from within a sequence of context words drawn from a sentence or phrase, or predicting which contexts words surround some given input word drawn from a sentence or phrase. Each of these tasks are supervised (the correct answer is fixed and known), but can be generated from unlabelled text data such as a collection of books or wikipedia articles, meaning that even though the task itself is supervised the training data can be generated automatically, enabling the creation of enormous training sets. The internal representation for particular words learned during the training process contain semantically informative features related to that given word, and can therefore be used as embeddings used downstream for tasks such as finding similarity between words or as input into additional models. Doc2Vec is an extension of this technique that determines vector embeddings for entire documents (strings containing multiple words, could be sentences, paragraphs, or documents).\n",
    "\n",
    "\n",
    "<a id=\"bert_biobert\"></a>\n",
    "### BERT and BioBERT\n",
    "BERT ('Bidirectional Encoder Representations from Transformers') is another neueral network-based model trained on two different false tasks, namely predicting the subsequent sentence given some input sentence, or predicting the identity of a set of words masked from an input sentence. Like Word2Vec, this architecture can be used to generate vector embeddings for a particular input word by extracting values from a subset of the encoder layers that correspond to that input word. Practically, a major difference is that because the input word is input in the context of its surrounding sentence, the embedding reflects the meaning of a particular word in a particular context (such as the difference in the meaning of *root* in the phrases *plant root* and *root of the problem*. BioBERT refers to a set of BERT models which have been finetuned on the PubMed and PMC corpora. See the list of relevant links for the publications and pages associated with these models.\n",
    "\n",
    "<a id=\"load_models\"></a>\n",
    "### Loading trained and saved models\n",
    "Versions of the architectures discussed above which have been saved as trained models are loaded here. Some of these models are loaded as pretrained models from the work of other groups, and some were trained on data specific to this notebook and loaded here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Files and models related to the machine learning text embedding methods used here.\n",
    "doc2vec_wiki_model = gensim.models.Doc2Vec.load(doc2vec_wikipedia_filename)\n",
    "doc2vec_pubmed_model = gensim.models.Doc2Vec.load(doc2vec_pubmed_filename)\n",
    "word2vec_model = gensim.models.Word2Vec.load(word2vec_model_filename)\n",
    "bert_tokenizer_base = BertTokenizer.from_pretrained('bert-base-uncased')\n",
    "bert_tokenizer_pmc = BertTokenizer.from_pretrained(biobert_pmc_path)\n",
    "bert_tokenizer_pubmed = BertTokenizer.from_pretrained(biobert_pubmed_path)\n",
    "bert_tokenizer_pubmed_pmc = BertTokenizer.from_pretrained(biobert_pubmed_pmc_path)\n",
    "bert_model_base = BertModel.from_pretrained('bert-base-uncased')\n",
    "bert_model_pmc = BertModel.from_pretrained(biobert_pmc_path)\n",
    "bert_model_pubmed = BertModel.from_pretrained(biobert_pubmed_path)\n",
    "bert_model_pubmed_pmc = BertModel.from_pretrained(biobert_pubmed_pmc_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading in the descriptions from hand-picked dataset of phenotype pairs\n",
    "See the other notebook for the creation of this dataset. This is included in this notebook instead of a separated notebook because we want the treatment of the individual phenotype text instances to be the same as is done for the descriptions from the real dataset of plant phenotypes. The list of computational approaches being evaluated for this task is the same in both cases so all of the cells between the point where the descriptions are read in and when the distance matrices are found using all those methods are the same for this task as any of the biological questions that this notebook is focused on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'long hypocotyl under far-red light; partially unfolded cotyledons under far-red light',\n",
       " 1: 'embryo defective; seedling defective',\n",
       " 2: 'increased lateral root number; late flowering',\n",
       " 3: 'embryo defective-globular; embryo defective; globular',\n",
       " 4: 'abnormal microtubule development; abnormal pavement cell morphology',\n",
       " 5: 'albino; dwarf',\n",
       " 6: 'pale green leaves; reduced rosette size',\n",
       " 7: 'pale green seedlings; pale green seeds',\n",
       " 8: 'embryo defective; seedling defective',\n",
       " 9: 'large rosette leaves',\n",
       " 10: 'chlorotic leaves; dwarf',\n",
       " 11: 'abnormal cotyledon morphology; collapsed root tip',\n",
       " 12: 'dwarf; fragile inflorescence stems',\n",
       " 13: 'anther defects; dwarf',\n",
       " 14: 'curled leaves; increased rosette leaf number',\n",
       " 15: 'dark green; dwarf',\n",
       " 16: 'resistant to anthranilate analogs (herbicide); seedling lethal without exogenous tryptophan',\n",
       " 17: 'elevated cobalt levels in shoots; sensitive to cobalt',\n",
       " 18: 'dwarf; sensitive to oxidative stress',\n",
       " 19: 'altered response to wounding; insensitive to jasmonate',\n",
       " 20: 'sensitive to zinc',\n",
       " 21: 'embryo defective (inferred); female gametophyte defective',\n",
       " 22: 'pale green; slow growth',\n",
       " 23: 'complete female gametophyte defective; male gametophyte defective',\n",
       " 24: 'abnormal flower morphology; abnormal leaf morphology',\n",
       " 25: 'right-handed helical growth in the root and other rapidly elongating organs; strong severity of cell expansion defects',\n",
       " 26: 'late flowering; late flowering; strong influence of short days on flowering time; small but significant effect of vernalization on flowering time; narrow leaves, increased number of cauline leaves; flowering occurs about 13 days later than wild type',\n",
       " 27: 'sensitive to avirulent pseudomonas syringae; sensitive to bth (sa mimic)',\n",
       " 28: 'abnormal flower morphology; abnormal leaf morphology',\n",
       " 29: 'decreased potassium permeability of the plasma membrane in leaf mesophyll cell layers',\n",
       " 30: 'altered fatty acid composition; altered systemic acquired resistance response',\n",
       " 31: 'abnormal flower morphology; abnormal leaf morphology',\n",
       " 32: 'embryo defective-globular; embryo defective; globular',\n",
       " 33: 'elevated starch levels',\n",
       " 34: 'embryo defective-cotyledon; embryo defective; cotyledon',\n",
       " 35: 'embryo defective-preglobular / globular; embryo defective; preglobular / globular',\n",
       " 36: 'abnormal mitochondria morphology; abnormal peroxisome morphology',\n",
       " 37: 'complete sterility; defects in meiosis',\n",
       " 38: 'albino seedlings; albino seeds',\n",
       " 39: 'embryo defective-globular; embryo defective; globular',\n",
       " 40: 'embryo defective-cotyledon; embryo defective; cotyledon',\n",
       " 41: 'embryo defective-cotyledon; embryo defective; cotyledon',\n",
       " 42: 'sensitive to manganese starvation; under mn-deficient conditions, roots are shorter, cell elongation was inhibited, and root hair elongation is markedly compromised',\n",
       " 43: 'completely sterile; dwarf',\n",
       " 44: 'decreased stomatal closure; decreased stomatal closure; no other phenotypes detected',\n",
       " 45: 'reduced nitrate uptake; resistant to chlorate',\n",
       " 46: 'abnormal cell plates; abnormal cell walls',\n",
       " 47: 'embryo defective-preglobular; embryo defective; preglobular',\n",
       " 48: 'abnormal pith cell morphology; dwarf',\n",
       " 49: 'low penetrance of heart-shaped cotyledons; semi-dominant; cotyledons fuse along one side at very low frequency (approx 0.5 %), slight fusion of sepals, stamen fusion at very low frequency',\n",
       " 50: 'sensitive to uv-b light',\n",
       " 51: 'embryo defective-cotyledon; embryo defective; cotyledon',\n",
       " 52: 'insensitive to cytokinin; short roots',\n",
       " 53: 'embryo defective-transition; embryo defective; transition',\n",
       " 54: 'albino seedlings; pale green seedlings',\n",
       " 55: 'homozygous plants have small and hyponastic cotyledons, some seedlings lack the primary root; 80% show early seedling lethality; epinastic cotyledons; wavy leaf margins; increased number of axillary shoots; reduced fertility',\n",
       " 56: 'blunt siliques; short inflorescence stems',\n",
       " 57: 'anthocyanin accumulation; dark-grown seedlings are de-etiolated',\n",
       " 58: 'abnormal rosette leaf morphology; decreased leaf number',\n",
       " 59: 'abnormal leaf morphology; dwarf',\n",
       " 60: 'reduced lateral root formation',\n",
       " 61: 'abnormal chloroplast morphology; albino',\n",
       " 62: 'altered response to red:far-red light; decreased branching',\n",
       " 63: 'fewer lateral roots; semi-dwarf',\n",
       " 64: 'abnormal flower morphology; blister-like outgrowths on cotyledons',\n",
       " 65: 'elongated hypocotyl in far-red light',\n",
       " 66: 'dwarf; pale green rosette',\n",
       " 67: 'this mutant has about 1/10 of the ornithine aminotransferase activity detected in wild type plants, but it has no other obvious phenotypic defects when grown under greenhouse conditions; in response to salt-stress, the mutant appears to produce wild-type levels of proline',\n",
       " 68: 'insensitive to brassinosteroids; semi-dwarf',\n",
       " 69: 'defects in meiosis; severely reduced fertility',\n",
       " 70: 'mutants did not grow on medium  containing 2.5 mm or 5 mm of l-glutamine as the sole nitrogen source',\n",
       " 71: 'embryo defective; developmental arrest of mutant embryos',\n",
       " 72: 'pigment defective embryo',\n",
       " 73: 'female gametophyte defective; rare embryo defective (inferred)',\n",
       " 74: 'embryo defective; root meristemless; the mutant exhibits a significant flg22-induced reduction in bacterial growth, indicating that the gene is most probably not required for flg22-induced bacterial resistance',\n",
       " 75: 'delayed germination; germination of the mutants exhibited hypersensitivity to exogenously applied abscisic acid and paclobutrazol, an inhibitor of gibberellin biosynthesis',\n",
       " 76: 'the morphology of mutant plants was similar and to some extent weaker than ssm (cs68717) plants; ectopic occurrence of myrosin cells is found in the leaves (increased number of myrosin cells)',\n",
       " 77: 'sensitive to oxidative stress',\n",
       " 78: 'elevated threonine levels; short roots',\n",
       " 79: 'chlorate resistance due to a reduced uptake of chlorate; chlorate resistant due to reduced chlorate uptake',\n",
       " 80: 'late flowering; short leaf blades',\n",
       " 81: 'slow growth',\n",
       " 82: 'embryo defective; developmental arrest of mutant embryos occurs at cotyledon stage',\n",
       " 83: 'low hydroxyl fatty acid levels in seeds',\n",
       " 84: 'embryo defective-globular; embryo defective; globular',\n",
       " 85: 'altered seed storage composition, reduced oil and protein body contents with concurrent accumulation of starch; developing embryos have white colored appearance; cotyledons and hypocotyl cells contain no recognizable protein bodies and few oil bodies; starch granules, membrane stacks, vesicles, and vacuoles are present in cells (all of which are absent in wt cells); oil bodies contain higher electron density substances than wt; mature seeds shrink upon desiccation (as a consequence of insufficient deposition of storage molecules) and are lethal; kanamycin resistant',\n",
       " 86: 'abnormal mitochondria morphology; abnormal peroxisome morphology',\n",
       " 87: 'chromosomes do not synapse during both male and female meiosis; there is a total absence of chiasmata',\n",
       " 88: 'pale green; seedling lethal',\n",
       " 89: 'embryo defective; leafy cotyledons',\n",
       " 90: 'homozygotes are embryo lethal; homozygotes are embryo lethal',\n",
       " 91: 'embryo defective-globular; embryo defective; globular',\n",
       " 92: 'etiolated seedlings of both wild type and mtk mutants have very low emission of ethylene; the eto3 mutant on the other hand has much higher emission of ethylene',\n",
       " 93: 'dwarf; pale green leaves',\n",
       " 94: 'decreased stomatal density; reduced number of stomata and non protruding cells on hypocotyl epidermis',\n",
       " 95: 'resistant to tunicamycin; small seedlings',\n",
       " 96: 'resistant to tunicamycin; small seedlings',\n",
       " 97: 'crinkled leaves; embryo lethal',\n",
       " 98: 'dwarf in the dark; no apical hook in the dark',\n",
       " 99: 'embryo defective; variegated seedlings'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This would only impact the notebook if run in the absence of the subsequent cells, never matters in the script.\n",
    "num_pairs = 50\n",
    "mupdata = pd.read_csv(\"../data/corpus_related_files/phenotype_pairs/scored.csv\")\n",
    "assert num_pairs == mupdata.shape[0]\n",
    "descriptions = mupdata[\"Phenotype 1\"].values.tolist()\n",
    "descriptions.extend(mupdata[\"Phenotype 2\"].values.tolist())\n",
    "descriptions = {i:description for i,description in enumerate(descriptions)}\n",
    "pair_to_score = {(i,i+num_pairs):s for i,s in enumerate(mupdata[\"Score\"].values)}\n",
    "descriptions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_3\"></a>\n",
    "# Part 3. NLP Choices\n",
    "\n",
    "<a id=\"preprocessing\"></a>\n",
    "### Preprocessing text descriptions\n",
    "The preprocessing methods applied to the phenotype descriptions are a choice which impacts the subsequent vectorization and similarity methods which construct the pairwise distance matrix from each of these descriptions. The preprocessing methods that make sense are also highly dependent on the vectorization method or embedding method that is to be applied. For example, stemming (which is part of the full proprocessing done below using the Gensim preprocessing function) is useful for the n-grams and bag-of-words methods but not for the document embeddings methods which need each token to be in the vocabulary that was constructed and used when the model was trained. For this reason, embedding methods with pretrained models where the vocabulary is fixed should have a lighter degree of preprocessing not involving stemming or lemmatization but should involve things like removal of non-alphanumerics and normalizing case. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain a mapping between IDs and the raw text descriptions associated with that ID from the dataset.\n",
    "descriptions = dataset.get_description_dictionary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary of phene descriptions and a dictionary to convert back to the phenotype/gene IDs.\n",
    "phenotype_ids = list(descriptions.keys())\n",
    "phenes = {}\n",
    "phene_id_to_id = {}\n",
    "phene_id = max(phenotype_ids)+1\n",
    "for i,phene_list in {i:sent_tokenize(d) for i,d in descriptions.items()}.items():\n",
    "    for phene in phene_list:\n",
    "        phenes[phene_id] = phene\n",
    "        phene_id_to_id[phene_id] = i\n",
    "        phene_id = phene_id+1\n",
    "phene_ids = list(phenes.keys())\n",
    "\n",
    "# Update the dictionary of descriptions to also include the individualy parsed phenes.\n",
    "descriptions.update(phenes)\n",
    "assert len(descriptions) == len(phenotype_ids)+len(phene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Applying canned prepreprocessing approaches to the descriptions.\n",
    "processed = defaultdict(dict)\n",
    "processed[\"simple\"] = {i:\" \".join(simple_preprocess(d)) for i,d in descriptions.items()}\n",
    "processed[\"simple_no_stops\"] = {i:remove_stopwords(\" \".join(simple_preprocess(d))) for i,d in descriptions.items()}\n",
    "processed[\"full\"] = {i:\" \".join(preprocess_string(d)) for i,d in descriptions.items()}\n",
    "\n",
    "# Set of stopwords, used later for checking it tokens in a list are stopwords or not.\n",
    "stop_words = set(stopwords.words('english')) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"pos_tagging\"></a>\n",
    "### POS tagging the phenotype descriptions for nouns and adjectives\n",
    "Note that preprocessing of the descriptions should be done after part-of-speech tagging, because tokens that are removed during preprocessing before n-gram analysis contain information that the parser needs to accurately call parts-of-speech. This step should be done on the raw descriptions and then the resulting bags of words can be subset using additional preprocesssing steps before input in one of the vectorization methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_pos_tokens = lambda text,pos: \" \".join([t[0] for t in nltk.pos_tag(word_tokenize(text)) if t[1].lower()==pos.lower()])\n",
    "processed[\"nouns\"] =  {i:get_pos_tokens(d,\"NN\") for i,d in descriptions.items()}\n",
    "processed[\"nouns_full\"] = {i:\" \".join(preprocess_string(d)) for i,d in processed[\"nouns\"].items()}\n",
    "processed[\"nouns_simple\"] = {i:\" \".join(simple_preprocess(d)) for i,d in processed[\"nouns\"].items()}\n",
    "processed[\"adjectives\"] =  {i:get_pos_tokens(d,\"JJ\") for i,d in descriptions.items()}\n",
    "processed[\"adjectives_full\"] = {i:\" \".join(preprocess_string(d)) for i,d in processed[\"adjectives\"].items()}\n",
    "processed[\"adjectives_simple\"] = {i:\" \".join(simple_preprocess(d)) for i,d in processed[\"adjectives\"].items()}\n",
    "processed[\"nouns_adjectives\"] = {i:\"{} {}\".format(processed[\"nouns\"][i],processed[\"adjectives\"][i]) for i in descriptions.keys()}\n",
    "processed[\"nouns_adjectives_full\"] = {i:\"{} {}\".format(processed[\"nouns_full\"][i],processed[\"adjectives_full\"][i]) for i in descriptions.keys()}\n",
    "processed[\"nouns_adjectives_simple\"] = {i:\"{} {}\".format(processed[\"nouns_simple\"][i],processed[\"adjectives_simple\"][i]) for i in descriptions.keys()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reducing vocabulary size based on identifying important words\n",
    "These approcahes for reducing the vocabulary size of the dataset work by identifying which words in the descriptions are likely to be the most important for identifying differences between the phenotypes and meaning of the descriptions. One approach is to determine which words occur at a higher rate in text of interest such as articles about plant phenotypes as compared to their rates in more general texts such as a corpus of news articles. These approaches do not create modified versions of the descriptions but rather provide vocabulary objects that can be passed to the sklearn vectorizer or constructors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create ontology objects for all the biological ontologies being used.\n",
    "pato = Ontology(pato_filename)\n",
    "po = Ontology(po_filename)\n",
    "go = Ontology(go_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting sets of tokens that are part of bio ontology term labels or synonyms.\n",
    "bio_ontology_tokens = list(set(po.tokens()).union(set(go.tokens())))\n",
    "bio_ontology_tokens = [t for t in bio_ontology_tokens if t not in stop_words]\n",
    "bio_ontology_tokens_simple = flatten([simple_preprocess(t) for t in bio_ontology_tokens])\n",
    "bio_ontology_tokens_full = flatten([preprocess_string(t) for t in bio_ontology_tokens])\n",
    "with open(os.path.join(OUTPUT_DIR,\"part_3_bio_ontology_vocab_size_{}.txt\".format(len(bio_ontology_tokens))),\"w\") as f:\n",
    "    f.write(\" \".join(bio_ontology_tokens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting sets of tokens that are overprepresented in plant phenotype papers as compared to some background corpus.\n",
    "maximum_number_of_tokens = 10000\n",
    "background_corpus = open(background_corpus_filename,\"r\").read()\n",
    "phenotypes_corpus = open(phenotypes_corpus_filename,\"r\").read()\n",
    "ppp_overrepresented_tokens = get_overrepresented_tokens(phenotypes_corpus, background_corpus, max_features=maximum_number_of_tokens)\n",
    "ppp_overrepresented_tokens = [t for t in ppp_overrepresented_tokens if t not in stop_words]\n",
    "ppp_overrepresented_tokens_simple = flatten([simple_preprocess(t) for t in ppp_overrepresented_tokens])\n",
    "ppp_overrepresented_tokens_full = flatten([preprocess_string(t) for t in ppp_overrepresented_tokens])\n",
    "with open(os.path.join(OUTPUT_DIR,\"part_3_plant_phenotype_vocab_size_{}.txt\".format(len(ppp_overrepresented_tokens))), \"w\") as f:\n",
    "    f.write(\" \".join(ppp_overrepresented_tokens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generating processed description entries by subsetting tokens to only include ones from these vocabularies.\n",
    "ppp_overrepresented_tokens_full_set = set(ppp_overrepresented_tokens_full)\n",
    "bio_ontology_tokens_full_set = set(bio_ontology_tokens_full)\n",
    "processed[\"plant_overrepresented_tokens\"] = {i:\" \".join([token for token in word_tokenize(text) if token in ppp_overrepresented_tokens_full_set]) for i,text in processed[\"full\"].items()}\n",
    "processed[\"bio_ontology_tokens\"] = {i:\" \".join([token for token in word_tokenize(text) if token in bio_ontology_tokens_full_set]) for i,text in processed[\"full\"].items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"vocab\"></a>\n",
    "### Reducing the vocabulary size using a word distance matrix\n",
    "These approaches for reducing the vocabulary size of the dataset work by replacing multiple words that occur throughout the dataset of descriptions with an identical word that is representative of this larger group of words. The total number of unique words across all descriptions is therefore reduced, and when observing n-gram overlaps between vector representations of these descriptions, overlaps will now occur between descriptions that included different but similar words. These methods work by actually generating versions of these descriptions that have the word replacements present. The returned objects for these methods are the revised description dictionary, a dictionary mapping tokens in the full vocabulary to tokens in the reduced vocabulary, and a dictionary mapping tokens in the reduced vocabulary to a list of tokens in the full vocabulary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate a pairwise distance matrix object using the oats subpackage, and create an appropriately shaped matrix,\n",
    "# making sure that the tokens list is in the same order as the indices representing each word in the distance matrix.\n",
    "# This is currently triviala because the IDs that are used are ordered integers 0 to n, but this might not always be\n",
    "# the case so it's not directly assumed here.\n",
    "tokens = list(set([w for w in flatten(d.split() for d in processed[\"simple\"].values())]))\n",
    "tokens_dict = {i:w for i,w in enumerate(tokens)}\n",
    "graph = pw.pairwise_square_word2vec(word2vec_model, tokens_dict, \"cosine\")\n",
    "distance_matrix = graph.array\n",
    "tokens = [tokens_dict[graph.index_to_id[index]] for index in np.arange(distance_matrix.shape[0])]\n",
    "\n",
    "# The other argument that the Linares Pontes algorithm needs is a value for n, see paper or description above.\n",
    "n = 3\n",
    "processed[\"linares_pontes\"], reduce_lp, unreduce_lp = reduce_vocab_linares_pontes(processed[\"simple\"], tokens, distance_matrix, n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"annotation\"></a>\n",
    "### Annotating descriptions with ontology terms\n",
    "This section generates dictionaries that map gene IDs from the dataset to lists of strings, where those strings are ontology term IDs. How the term IDs are found for each gene entry with its corresponding phenotype description depends on the cell below. Firstly, the terms are found by using the NOBLE Coder annotation tool through these wrapper functions to identify the terms by looking for instances of the term's label or synonyms in the actual text of the phenotype descriptions. Secondly, the next cell just draws the terms directly from the dataset itself. In this case, these are high-confidence annotations done by curators for a comparison against what can be accomplished through computational analysis of the text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the NOBLE Coder annotator over the raw input text descriptions, which handles things like case normalization.\n",
    "direct_annots_nc_go_precise = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"go\", precise=1)\n",
    "direct_annots_nc_go_partial = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"go\", precise=0)\n",
    "direct_annots_nc_po_precise = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"po\", precise=1)\n",
    "direct_annots_nc_po_partial = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"po\", precise=0)\n",
    "direct_annots_nc_pato_precise = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"pato\", precise=1)\n",
    "direct_annots_nc_pato_partial = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"pato\", precise=0)\n",
    "\n",
    "# Use the ontology hierarchies to add terms that are inherited by the terms that were annotated to the text.\n",
    "inherited_annots_nc_go_precise = {i:go.inherited(term_id_list) for i,term_id_list in direct_annots_nc_go_precise.items()}\n",
    "inherited_annots_nc_go_partial = {i:go.inherited(term_id_list) for i,term_id_list in direct_annots_nc_go_partial.items()}\n",
    "inherited_annots_nc_po_precise = {i:po.inherited(term_id_list) for i,term_id_list in direct_annots_nc_po_precise.items()}\n",
    "inherited_annots_nc_po_partial = {i:po.inherited(term_id_list) for i,term_id_list in direct_annots_nc_po_partial.items()}\n",
    "inherited_annots_nc_pato_precise = {i:pato.inherited(term_id_list) for i,term_id_list in direct_annots_nc_pato_precise.items()}\n",
    "inherited_annots_nc_pato_partial = {i:pato.inherited(term_id_list) for i,term_id_list in direct_annots_nc_pato_partial.items()}\n",
    "\n",
    "# Merge the ontology term annotations for each descritpion into a single dictionary for the precise and partial levels.\n",
    "all_precise_annotations = {i:flatten([inherited_annots_nc_go_precise[i],inherited_annots_nc_po_precise[i],inherited_annots_nc_pato_precise[i]]) for i in descriptions.keys()}\n",
    "all_partial_annotations = {i:flatten([inherited_annots_nc_go_partial[i],inherited_annots_nc_po_partial[i],inherited_annots_nc_pato_partial[i]]) for i in descriptions.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Treating these sets of inherited ontology terms as tokens so that they can be used as n-grams.\n",
    "processed[\"precise_annotations\"] = {i:\" \".join(annots) for i,annots in all_precise_annotations.items()}\n",
    "processed[\"partial_annotations\"] = {i:\" \".join(annots) for i,annots in all_partial_annotations.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create description strings with all ontology term anntotations concatenated to the end of the descriptions.\n",
    "processed[\"simple_plus_precise_annotations\"] = {i:\" \".join(flatten([text,all_precise_annotations[i]])) for i,text in processed[\"simple\"].items()}\n",
    "processed[\"simple_plus_partial_annotations\"] = {i:\" \".join(flatten([text,all_partial_annotations[i]])) for i,text in processed[\"simple\"].items()}\n",
    "processed[\"full_plus_precise_annotations\"] = {i:\" \".join(flatten([text,all_precise_annotations[i]])) for i,text in processed[\"full\"].items()}\n",
    "processed[\"full_plus_partial_annotations\"] = {i:\" \".join(flatten([text,all_partial_annotations[i]])) for i,text in processed[\"full\"].items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create ontology term annotations dictionaries for all the high confidence annotations present in the dataset.\n",
    "curated_go_annotations = dataset.get_annotations_dictionary(\"go\")\n",
    "curated_po_annotations = dataset.get_annotations_dictionary(\"po\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"phenes\"></a>\n",
    "### Splitting the descriptions into individual phenes\n",
    "As a preprocessing step, split into a new set of descriptions that's larger. Note that phenotypes are split into phenes, and the phenes that are identical are retained as separate entries in the dataset. This makes the distance matrix calculation more needlessly expensive, because vectors need to be found for the same string more than once, but it simplifies converting the edgelist back to having IDs that reference the genes (full phenotypes) instead of the smaller phenes. If anything, that problem should be addressed in the pairwise functions, not here. (The package should handle it, not when creating input data for those methods)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform the processed description dictionaries back into referencing either just phenotypes or phenes.\n",
    "descriptions = dataset.get_description_dictionary()\n",
    "processes = list(processed.keys())\n",
    "unmerged = defaultdict(dict)\n",
    "for process,di in processed.items():\n",
    "    unmerged[\"{}_phenes\".format(process)] = {i:text for i,text in di.items() if i in phene_ids}\n",
    "    unmerged[process] = {i:text for i,text in di.items() if i in phenotype_ids}\n",
    "processed = unmerged\n",
    "\n",
    "# Checking to make sure the size of each dictionary is as expected.\n",
    "for process in processes:\n",
    "    assert len(phenotype_ids) == len(processed[process].keys())\n",
    "    assert len(phene_ids) == len(processed[\"{}_phenes\".format(process)].keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_4\"></a>\n",
    "# Part 4. Generating vector representations and pairwise distances matrices\n",
    "This section uses the text descriptions, preprocessed text descriptions, or ontology term annotations created or read in the previous sections to generate a vector representation for each gene and build a pairwise distance matrix for the whole dataset. Each method specified is a unique combination of a method of vectorization (bag-of-words, n-grams, document embedding model, etc) and distance metric (Euclidean, Jaccard, cosine, etc) applied to those vectors in constructing the pairwise matrix. The method of vectorization here is equivalent to feature selection, so the task is to figure out which type of vectors will encode features that are useful (n-grams, full words, only words from a certain vocabulary, etc).\n",
    "\n",
    "<a id=\"methods\"></a>\n",
    "### Specifying a list of NLP methods to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "methods = [ \n",
    "    # Methods that use neural networks to generate embeddings.\n",
    "    Method(\"Doc2Vec\", \"Wikipedia,Size=300\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_wiki_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Mean\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Max\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use topic models to generate embeddings.\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=20\", pw.pairwise_square_topic_model, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"num_topics\":20, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=50\", pw.pairwise_square_topic_model, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"num_topics\":50, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=100\", pw.pairwise_square_topic_model, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"num_topics\":100, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=200\", pw.pairwise_square_topic_model, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"num_topics\":200, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use variations on the n-grams approach with full preprocessing (includes stemming).\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"full\"], \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "    Method(\"N-Grams\", \"Full,Nouns,Adjectives,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"nouns_adjectives_full\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Linares_Pontes,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"linares_pontes\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Precise_Annotations,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"full_plus_precise_annotations\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Partial_Annotations,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"full_plus_partial_annotations\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use terms inferred from automated annotation of the text.\n",
    "    Method(\"NOBLE Coder\", \"Precise\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"precise_annotations\"], \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"metric\":\"jaccard\", \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"NOBLE Coder\", \"Partial\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"partial_annotations\"], \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"metric\":\"jaccard\", \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"NOBLE Coder\", \"Precise,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"precise_annotations\"], \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"metric\":\"cosine\", \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"NOBLE Coder\", \"Partial,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"partial_annotations\"], \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"metric\":\"cosine\", \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use terms assigned by humans that are present in the dataset.\n",
    "    Method(\"GO\", \"None\", pw.pairwise_square_annotations, {\"ids_to_annotations\":curated_go_annotations, \"ontology\":go, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"PO\", \"None\", pw.pairwise_square_annotations, {\"ids_to_annotations\":curated_po_annotations, \"ontology\":po, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"tfidf\":False}, spatial.distance.jaccard),\n",
    "\n",
    "    \n",
    "    # Split into phene descriptions\n",
    "    \n",
    "    \n",
    "    # Methods that use neural networks to generate embeddings.\n",
    "    Method(\"Doc2Vec Phenes\", \"Wikipedia,Size=300\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_wiki_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec Phenes\", \"Wikipedia,Size=300,Mean\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec Phenes\", \"Wikipedia,Size=300,Max\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams Phenes\", \"Linares_Pontes,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":processed[\"linares_pontes_phenes\"], \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"running\"></a>\n",
    "### Running all of the methods to generate distance matrices\n",
    "Notes- Instead of passing in similarity function like cosine distance that will get evaluated for every possible i,j pair of vetors that are created (this is very big when splitting by phenes), don't use a specific similarity function, but instead let the object use a KNN classifier. pass in some limit for k like 100. then the object uses some more efficient (not brute force) algorithm to set the similarity of some vector v to its 100 nearest neighbors as those 100 probabilities, and sets everything else to 0. This would need to be implemented as a matching but separate function from the get_square_matrix_from_vectors thing. And then this would need to be noted in the similarity function that was used for these in the big table of methods. This won't work because the faster (not brute force algorithms) are not for sparse vectors like n-grams, and the non-sparse embeddings aren't really the problem here because those vectors are relatively much short, even when concatenating BERT encoder layers thats only up to around length of ~1000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doc2Vec:Wikipedia,Size=300                                             00:00:02\n",
      "Word2Vec:Wikipedia,Size=300,Mean                                       00:00:00\n",
      "Word2Vec:Wikipedia,Size=300,Max                                        00:00:00\n",
      "Topic Models:NMF,Full,Topics=20                                        00:00:00\n",
      "Topic Models:NMF,Full,Topics=50                                        00:00:02\n",
      "Topic Models:NMF,Full,Topics=100                                       00:00:08\n",
      "Topic Models:NMF,Full,Topics=200                                       00:00:32\n",
      "N-Grams:Full,Words,1-grams                                             00:00:00\n",
      "N-Grams:Full,Words,1-grams,Binary                                      00:00:00\n",
      "N-Grams:Full,Words,1-grams,TFIDF                                       00:00:00\n",
      "N-Grams:Full,Words,1-grams,Binary,TFIDF                                00:00:00\n",
      "N-Grams:Full,Nouns,Adjectives,1-grams                                  00:00:00\n",
      "N-Grams:Linares_Pontes,Words,1-grams                                   00:00:00\n",
      "N-Grams:Full,Precise_Annotations,Words,1-grams                         00:00:00\n",
      "N-Grams:Full,Partial_Annotations,Words,1-grams                         00:00:02\n",
      "NOBLE Coder:Precise                                                    00:00:00\n",
      "NOBLE Coder:Partial                                                    00:00:03\n",
      "NOBLE Coder:Precise,TFIDF                                              00:00:00\n",
      "NOBLE Coder:Partial,TFIDF                                              00:00:04\n",
      "GO:None                                                                00:00:00\n",
      "PO:None                                                                00:00:00\n",
      "Doc2Vec Phenes:Wikipedia,Size=300                                      00:00:07\n",
      "Word2Vec Phenes:Wikipedia,Size=300,Mean                                00:00:05\n",
      "Word2Vec Phenes:Wikipedia,Size=300,Max                                 00:00:04\n",
      "N-Grams Phenes:Linares_Pontes,Words,1-grams                            00:00:07\n"
     ]
    }
   ],
   "source": [
    "# Generate all the pairwise distance matrices (not in parallel).\n",
    "graphs = {}\n",
    "names = []\n",
    "durations = []\n",
    "for method in methods:\n",
    "    graph,duration = function_wrapper_with_duration(function=method.function, args=method.kwargs)\n",
    "    graphs[method.name_with_hyperparameters] = graph\n",
    "    names.append(method.name_with_hyperparameters)\n",
    "    durations.append(to_hms(duration))\n",
    "    print(\"{:70} {}\".format(method.name_with_hyperparameters,to_hms(duration)))\n",
    "durations_df = pd.DataFrame({\"method\":names,\"duration\":durations})\n",
    "durations_df.to_csv(os.path.join(OUTPUT_DIR,\"part_4_durations.csv\"), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"merging\"></a>\n",
    "### Merging all of the distance matrices into a single dataframe specifying edges\n",
    "This section also handles replacing IDs from the individual methods that are references individual phenes that are part of a larger phenotype, and replacing those IDs with IDs referencing the full phenotypes (one-to-one relationship between phenotypes and genes). In this case, the minimum distance found between any two phenes from those two phenotypes represents the distance between that pair of phenotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams\n",
       "1    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000\n",
       "2    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073\n",
       "3    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018\n",
       "4    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782\n",
       "5    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036\n",
       "6      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505\n",
       "7    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938\n",
       "8    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089\n",
       "9    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230\n",
       "10   216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447\n",
       "11   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094\n",
       "12   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275\n",
       "13   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454\n",
       "14   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508\n",
       "15   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150\n",
       "16   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000\n",
       "17   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000\n",
       "18   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681\n",
       "19   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772\n",
       "20   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging all the edgelists together.\n",
    "metric_dict = {method.name_with_hyperparameters:method.metric for method in methods}\n",
    "tags_dict = {method.name_with_hyperparameters:method.tag for method in methods}\n",
    "names = list(graphs.keys())\n",
    "edgelists = {k:v.edgelist for k,v in graphs.items()}\n",
    "\n",
    "# Modify the edgelists for the methods that were using a phene split.\n",
    "for name,edgelist in edgelists.items():\n",
    "    # Converting phene IDs back to phenotype (gene) IDs where applicable.\n",
    "    if \"phene\" in tags_dict[name]:\n",
    "        edgelist[\"from\"] = edgelist[\"from\"].map(lambda x: phene_id_to_id[x])\n",
    "        edgelist[\"to\"] = edgelist[\"to\"].map(lambda x: phene_id_to_id[x])\n",
    "        edgelist = edgelist.groupby([\"from\",\"to\"], as_index=False).min()\n",
    "    # Making sure the edges are listed with the nodes sorted consistently.\n",
    "    cond = edgelist[\"from\"] > edgelist[\"to\"]\n",
    "    edgelist.loc[cond, ['from', 'to']] = edgelist.loc[cond, ['to', 'from']].values\n",
    "    edgelists[name] = edgelist\n",
    "\n",
    "# Do the merge step and remove self edges from the full dataframe.\n",
    "df = merge_edgelists(edgelists, default_value=1.000)\n",
    "df = remove_self_loops(df)\n",
    "df[\"from\"] = df[\"from\"].astype(\"int64\")\n",
    "df[\"to\"] = df[\"to\"].astype(\"int64\")\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ensemble\"></a>\n",
    "### Combining multiple distances measurements into summarizing distance values\n",
    "The purpose of this section is to iteratively train models on subsections of the dataset using simple regression or machine learning approaches to predict a value from zero to one indicating indicating how likely is it that two genes share atleast one of the specified groups in common. The information input to these models is the distance scores provided by each method in some set of all the methods used in this notebook. The purpose is to see whether or not a function of these similarity scores specifically trained to the task of predicting common groupings is better able to used the distance metric information to report a score for this task."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean\n",
       "1    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313\n",
       "2    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279\n",
       "3    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555\n",
       "4    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239\n",
       "5    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044\n",
       "6      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266\n",
       "7    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195\n",
       "8    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494\n",
       "9    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518\n",
       "10   216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178\n",
       "11   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760\n",
       "12   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626\n",
       "13   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244\n",
       "14   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487\n",
       "15   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812\n",
       "16   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507\n",
       "17   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420\n",
       "18   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773\n",
       "19   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061\n",
       "20   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the average distance percentile as a means of combining multiple scores.\n",
    "name = \"Mean\"\n",
    "names_to_use_for_mean = [name for name in names if not name in [\"GO:None\",\"PO:None\"]]\n",
    "df[name] = df[names_to_use_for_mean].rank(pct=True).mean(axis=1)\n",
    "names.append(name)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doc2Vec:Wikipedia,Size=300\n",
      "Word2Vec:Wikipedia,Size=300,Mean\n",
      "Word2Vec:Wikipedia,Size=300,Max\n",
      "Topic Models:NMF,Full,Topics=20\n",
      "Topic Models:NMF,Full,Topics=50\n",
      "Topic Models:NMF,Full,Topics=100\n",
      "Topic Models:NMF,Full,Topics=200\n",
      "N-Grams:Full,Words,1-grams\n",
      "N-Grams:Full,Words,1-grams,Binary\n",
      "N-Grams:Full,Words,1-grams,TFIDF\n",
      "N-Grams:Full,Words,1-grams,Binary,TFIDF\n",
      "N-Grams:Full,Nouns,Adjectives,1-grams\n",
      "N-Grams:Linares_Pontes,Words,1-grams\n",
      "N-Grams:Full,Precise_Annotations,Words,1-grams\n",
      "N-Grams:Full,Partial_Annotations,Words,1-grams\n",
      "NOBLE Coder:Precise\n",
      "NOBLE Coder:Partial\n",
      "NOBLE Coder:Precise,TFIDF\n",
      "NOBLE Coder:Partial,TFIDF\n",
      "GO:None\n",
      "PO:None\n",
      "Doc2Vec Phenes:Wikipedia,Size=300\n",
      "Word2Vec Phenes:Wikipedia,Size=300,Mean\n",
      "Word2Vec Phenes:Wikipedia,Size=300,Max\n",
      "N-Grams Phenes:Linares_Pontes,Words,1-grams\n",
      "Mean\n"
     ]
    }
   ],
   "source": [
    "# Normalizing all of the array representations of the graphs so they can be combined. Then this version of the arrays\n",
    "# should be used by any other cells that need all of the arrays, rather than the arrays accessed from the graph\n",
    "# objects. This is necessary for this analysis because some of the graph objects refer to phene datasets not\n",
    "# phenotype datasets.\n",
    "name_to_array = {}\n",
    "ids = list(descriptions.keys())\n",
    "n = len(descriptions)\n",
    "id_to_array_index = {i:idx for idx,i in enumerate(ids)}\n",
    "array_index_to_id = {idx:i for i,idx in id_to_array_index.items()}\n",
    "for name in names:\n",
    "    print(name)\n",
    "    idx = list(df.columns).index(name)+1\n",
    "    arr = np.ones((n, n))\n",
    "    for row in df.itertuples():\n",
    "        arr[id_to_array_index[row[1]]][id_to_array_index[row[2]]] = row[idx]\n",
    "        arr[id_to_array_index[row[2]]][id_to_array_index[row[1]]] = row[idx]\n",
    "    np.fill_diagonal(arr, 0.000) \n",
    "    name_to_array[name] = arr    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding correlations between human and computational approaches for hand-picked phenotype pairs\n",
    "This is only meant to be run in the context of the notebook, and should never be used in the script "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "OBJECTIVE = 1\n",
    "if OBJECTIVE == 2:\n",
    "#if OBJECTIVE not in (\"phenotypes\", \"pathways\", \"interactions\", \"orthologs\"):\n",
    "    small_table = defaultdict(dict)\n",
    "    for name in names:\n",
    "        values = []\n",
    "        scores = []\n",
    "        for tup,score in pair_to_score.items():\n",
    "            i = id_to_array_index[tup[0]]\n",
    "            j = id_to_array_index[tup[1]]\n",
    "            value = 1 - name_to_array[name][i,j]\n",
    "            values.append(value)\n",
    "            scores.append(score)\n",
    "        rho,pval = spearmanr(values,scores)\n",
    "        small_table[name] = {\"rho\":rho,\"pval\":pval}\n",
    "    pd.DataFrame(small_table).transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_5\"></a>\n",
    "# Part 5. Biological Questions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean\n",
       "1    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313\n",
       "2    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279\n",
       "3    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555\n",
       "4    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239\n",
       "5    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044\n",
       "6      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266\n",
       "7    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195\n",
       "8    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494\n",
       "9    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518\n",
       "10   216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178\n",
       "11   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760\n",
       "12   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626\n",
       "13   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244\n",
       "14   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487\n",
       "15   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812\n",
       "16   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507\n",
       "17   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420\n",
       "18   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773\n",
       "19   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061\n",
       "20   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"pathway_objective\"></a>\n",
    "### Using shared pathway membership (PlantCyc and KEGG) as the objective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair mapped to a pathway resource.\n",
    "pathway_mapped_ids = set(kegg_mapped_ids+pmn_mapped_ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in pathway_mapped_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in pathway_mapped_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"pathways\"] = -1\n",
    "id_to_kegg_group_ids, kegg_group_id_to_ids = kegg_groups.get_groupings_for_dataset(dataset)\n",
    "id_to_pmn_group_ids, pmn_group_id_to_ids = pmn_groups.get_groupings_for_dataset(dataset)\n",
    "id_to_group_ids = {i:flatten([id_to_kegg_group_ids[i],id_to_pmn_group_ids[i]]) for i in dataset.get_ids()}\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"pathways\"] = df[[\"from\",\"to\"]].apply(lambda x: len(set(id_to_group_ids[x[\"from\"]]).intersection(set(id_to_group_ids[x[\"to\"]])))>0, axis=1)*1\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\",\"pair_is_valid\"], axis=\"columns\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair mapped to a pathway resource.\n",
    "pathway_mapped_ids = set(kegg_mapped_ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in pathway_mapped_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in pathway_mapped_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"kegg_only\"] = -1\n",
    "id_to_kegg_group_ids, kegg_group_id_to_ids = kegg_groups.get_groupings_for_dataset(dataset)\n",
    "id_to_pmn_group_ids, pmn_group_id_to_ids = pmn_groups.get_groupings_for_dataset(dataset)\n",
    "id_to_group_ids = {i:flatten([id_to_kegg_group_ids[i],id_to_pmn_group_ids[i]]) for i in dataset.get_ids()}\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"kegg_only\"] = df[[\"from\",\"to\"]].apply(lambda x: len(set(id_to_group_ids[x[\"from\"]]).intersection(set(id_to_group_ids[x[\"to\"]])))>0, axis=1)*1\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\",\"pair_is_valid\"], axis=\"columns\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "      <th>known</th>\n",
       "      <th>predicted</th>\n",
       "      <th>orthologs</th>\n",
       "      <th>eqs_distance</th>\n",
       "      <th>curated</th>\n",
       "      <th>same</th>\n",
       "      <th>pathways_kegg_only</th>\n",
       "      <th>pathways_plantcyc_only</th>\n",
       "      <th>kegg_only</th>\n",
       "      <th>pmn_only</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets  known  predicted  orthologs  eqs_distance  curated   same  pathways_kegg_only  pathways_plantcyc_only  kegg_only  pmn_only\n",
       "0    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0   -1.0       -1.0       -1.0           1.0     True   True                  -1                      -1         -1        -1\n",
       "1    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0    0.0        0.0       -1.0           1.0     True   True                  -1                      -1         -1        -1\n",
       "2    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "3    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0   -1.0       -1.0       -1.0           1.0     True   True                   0                      -1          0        -1\n",
       "4    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "5      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1    0.0        0.0       -1.0           1.0     True  False                  -1                      -1         -1        -1\n",
       "6    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0   -1.0       -1.0       -1.0           1.0     True   True                   0                      -1          0        -1\n",
       "7    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "8    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "9    216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0    0.0        0.0       -1.0           1.0     True   True                  -1                      -1         -1        -1\n",
       "10   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760         0        0    0.0        0.0       -1.0           1.0     True   True                   0                      -1          0        -1\n",
       "11   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626        -1       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "12   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244         0        0    0.0        0.0       -1.0           1.0     True   True                  -1                      -1         -1        -1\n",
       "13   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487        -1       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "14   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812        -1        0    0.0        0.0       -1.0           1.0     True   True                  -1                      -1         -1        -1\n",
       "15   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507        -1       -1    0.0        0.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "16   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420        -1        0   -1.0       -1.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "17   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773        -1        0   -1.0       -1.0       -1.0          -1.0    False   True                  -1                      -1         -1        -1\n",
       "18   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061        -1        0   -1.0       -1.0       -1.0           0.9     True   True                  -1                      -1         -1        -1\n",
       "19   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549         0       -1    0.0        0.0       -1.0          -1.0    False   True                   0                      -1          0        -1"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair mapped to a pathway resource.\n",
    "pathway_mapped_ids = set(pmn_mapped_ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in pathway_mapped_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in pathway_mapped_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"pmn_only\"] = -1\n",
    "id_to_kegg_group_ids, kegg_group_id_to_ids = kegg_groups.get_groupings_for_dataset(dataset)\n",
    "id_to_pmn_group_ids, pmn_group_id_to_ids = pmn_groups.get_groupings_for_dataset(dataset)\n",
    "id_to_group_ids = {i:flatten([id_to_kegg_group_ids[i],id_to_pmn_group_ids[i]]) for i in dataset.get_ids()}\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"pmn_only\"] = df[[\"from\",\"to\"]].apply(lambda x: len(set(id_to_group_ids[x[\"from\"]]).intersection(set(id_to_group_ids[x[\"to\"]])))>0, axis=1)*1\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\",\"pair_is_valid\"], axis=\"columns\", inplace=True)\n",
    "\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"subset_objective\"></a>\n",
    "### Using shared phenotype classification (Lloyd and Meinke et al., 2012) as the objective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets\n",
       "1    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0\n",
       "2    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0\n",
       "3    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1\n",
       "4    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0\n",
       "5    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1\n",
       "6      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1\n",
       "7    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0\n",
       "8    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1\n",
       "9    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1\n",
       "10   216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0\n",
       "11   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760         0        0\n",
       "12   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626        -1       -1\n",
       "13   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244         0        0\n",
       "14   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487        -1       -1\n",
       "15   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812        -1        0\n",
       "16   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507        -1       -1\n",
       "17   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420        -1        0\n",
       "18   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773        -1        0\n",
       "19   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061        -1        0\n",
       "20   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549         0       -1"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair are mapped to a phenotype classification.\n",
    "relevant_ids = set(subsets_mapped_ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in relevant_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in relevant_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"subsets\"] = -1\n",
    "id_to_group_ids,_ = phe_subsets_groups.get_groupings_for_dataset(dataset)\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"subsets\"] = df[[\"from\",\"to\"]].apply(lambda x: len(set(id_to_group_ids[x[\"from\"]]).intersection(set(id_to_group_ids[x[\"to\"]])))>0, axis=1)*1\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\", \"pair_is_valid\"], axis=\"columns\", inplace=True)\n",
    "\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"association_objective\"></a>\n",
    "### Using protein assocations (STRING) as the objective "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "      <th>known</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets  known  predicted\n",
       "0    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0   -1.0       -1.0\n",
       "1    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0    0.0        0.0\n",
       "2    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1    0.0        0.0\n",
       "3    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0   -1.0       -1.0\n",
       "4    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1    0.0        0.0\n",
       "5      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1    0.0        0.0\n",
       "6    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0   -1.0       -1.0\n",
       "7    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1    0.0        0.0\n",
       "8    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1    0.0        0.0\n",
       "9    216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0    0.0        0.0\n",
       "10   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760         0        0    0.0        0.0\n",
       "11   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626        -1       -1    0.0        0.0\n",
       "12   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244         0        0    0.0        0.0\n",
       "13   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487        -1       -1    0.0        0.0\n",
       "14   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812        -1        0    0.0        0.0\n",
       "15   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507        -1       -1    0.0        0.0\n",
       "16   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420        -1        0   -1.0       -1.0\n",
       "17   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773        -1        0   -1.0       -1.0\n",
       "18   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061        -1        0   -1.0       -1.0\n",
       "19   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549         0       -1    0.0        0.0"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair are mapped to a phenotype classification.\n",
    "relevant_ids = set(string_edgelist.ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in relevant_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in relevant_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"known\"] = -1\n",
    "df[\"predicted\"] = -1\n",
    "df = df.merge(right=string_edgelist.df, how=\"left\", on=[\"from\",\"to\"])\n",
    "df[\"known_associations\"].fillna(value=0, inplace=True)\n",
    "df[\"predicted_associations\"].fillna(value=0, inplace=True)\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"known\"] = df[\"known_associations\"]\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"predicted\"] = df[\"predicted_associations\"]\n",
    "\n",
    "# Convert all the positive values from string on range 0 to arbitrary n to be equal to 1.\n",
    "df.loc[df[\"known\"] >= 1, \"known\"] = 1 \n",
    "df.loc[df[\"predicted\"] >= 1, \"predicted\"] = 1 \n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\",\"pair_is_valid\",\"known_associations\",\"predicted_associations\"], axis=\"columns\", inplace=True)\n",
    "\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ortholog_objective\"></a>\n",
    "### Using orthology between genes (PANTHER) as the objective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "      <th>known</th>\n",
       "      <th>predicted</th>\n",
       "      <th>orthologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets  known  predicted  orthologs\n",
       "0    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0   -1.0       -1.0       -1.0\n",
       "1    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0    0.0        0.0       -1.0\n",
       "2    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1    0.0        0.0       -1.0\n",
       "3    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0   -1.0       -1.0       -1.0\n",
       "4    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1    0.0        0.0       -1.0\n",
       "5      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1    0.0        0.0       -1.0\n",
       "6    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0   -1.0       -1.0       -1.0\n",
       "7    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1    0.0        0.0       -1.0\n",
       "8    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1    0.0        0.0       -1.0\n",
       "9    216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0    0.0        0.0       -1.0\n",
       "10   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760         0        0    0.0        0.0       -1.0\n",
       "11   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626        -1       -1    0.0        0.0       -1.0\n",
       "12   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244         0        0    0.0        0.0       -1.0\n",
       "13   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487        -1       -1    0.0        0.0       -1.0\n",
       "14   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812        -1        0    0.0        0.0       -1.0\n",
       "15   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507        -1       -1    0.0        0.0       -1.0\n",
       "16   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420        -1        0   -1.0       -1.0       -1.0\n",
       "17   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773        -1        0   -1.0       -1.0       -1.0\n",
       "18   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061        -1        0   -1.0       -1.0       -1.0\n",
       "19   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549         0       -1    0.0        0.0       -1.0"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair are mapped to a phenotype classification.\n",
    "relevant_ids = set(panther_edgelist.ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in relevant_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in relevant_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"orthologs\"] = -1\n",
    "df = df.merge(right=panther_edgelist.df, how=\"left\", on=[\"from\",\"to\"])\n",
    "df[\"value\"].fillna(value=0, inplace=True)\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"orthologs\"] = df[\"value\"]\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\",\"pair_is_valid\",\"value\"], axis=\"columns\", inplace=True)\n",
    "\n",
    "df.head(20)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Curator-derived similarity values from Oellrich, Walls et al., 2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "      <th>known</th>\n",
       "      <th>predicted</th>\n",
       "      <th>orthologs</th>\n",
       "      <th>eqs_distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets  known  predicted  orthologs  eqs_distance\n",
       "0    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0   -1.0       -1.0       -1.0           1.0\n",
       "1    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0    0.0        0.0       -1.0           1.0\n",
       "2    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1    0.0        0.0       -1.0          -1.0\n",
       "3    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0   -1.0       -1.0       -1.0           1.0\n",
       "4    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1    0.0        0.0       -1.0          -1.0\n",
       "5      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1    0.0        0.0       -1.0           1.0\n",
       "6    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0   -1.0       -1.0       -1.0           1.0\n",
       "7    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1    0.0        0.0       -1.0          -1.0\n",
       "8    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1    0.0        0.0       -1.0          -1.0\n",
       "9    216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0    0.0        0.0       -1.0           1.0\n",
       "10   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760         0        0    0.0        0.0       -1.0           1.0\n",
       "11   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626        -1       -1    0.0        0.0       -1.0          -1.0\n",
       "12   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244         0        0    0.0        0.0       -1.0           1.0\n",
       "13   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487        -1       -1    0.0        0.0       -1.0          -1.0\n",
       "14   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812        -1        0    0.0        0.0       -1.0           1.0\n",
       "15   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507        -1       -1    0.0        0.0       -1.0          -1.0\n",
       "16   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420        -1        0   -1.0       -1.0       -1.0          -1.0\n",
       "17   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773        -1        0   -1.0       -1.0       -1.0          -1.0\n",
       "18   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061        -1        0   -1.0       -1.0       -1.0           0.9\n",
       "19   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549         0       -1    0.0        0.0       -1.0          -1.0"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair are mapped to all the curation types.\n",
    "relevant_ids = set(ow_edgelist.ids)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in relevant_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in relevant_ids)\n",
    "df[\"pair_is_valid\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "\n",
    "# Add a column giving the actual target output value for this biological task, with -1 for the irrelevant rows.\n",
    "df[\"eqs_distance\"] = -1\n",
    "df = df.merge(right=ow_edgelist.df, how=\"left\", on=[\"from\",\"to\"])\n",
    "df[\"value\"].fillna(value=0, inplace=True)\n",
    "df.loc[(df[\"pair_is_valid\"]==True),\"eqs_distance\"] = 1-df[\"value\"]\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\",\"pair_is_valid\",\"value\"], axis=\"columns\", inplace=True)\n",
    "\n",
    "df.head(20)   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking whether gene pairs are considered curated or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "      <th>known</th>\n",
       "      <th>predicted</th>\n",
       "      <th>orthologs</th>\n",
       "      <th>eqs_distance</th>\n",
       "      <th>curated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>216</td>\n",
       "      <td>2024</td>\n",
       "      <td>0.428118</td>\n",
       "      <td>0.112537</td>\n",
       "      <td>0.085785</td>\n",
       "      <td>0.680843</td>\n",
       "      <td>0.841399</td>\n",
       "      <td>0.917374</td>\n",
       "      <td>0.590837</td>\n",
       "      <td>0.807489</td>\n",
       "      <td>0.938272</td>\n",
       "      <td>0.919413</td>\n",
       "      <td>0.957781</td>\n",
       "      <td>0.861796</td>\n",
       "      <td>0.350995</td>\n",
       "      <td>0.312535</td>\n",
       "      <td>0.224063</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.826178</td>\n",
       "      <td>0.903649</td>\n",
       "      <td>0.851012</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.315223</td>\n",
       "      <td>0.159200</td>\n",
       "      <td>0.111529</td>\n",
       "      <td>0.278094</td>\n",
       "      <td>0.201760</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>216</td>\n",
       "      <td>5052</td>\n",
       "      <td>0.424170</td>\n",
       "      <td>0.139413</td>\n",
       "      <td>0.091400</td>\n",
       "      <td>0.773170</td>\n",
       "      <td>0.873296</td>\n",
       "      <td>0.943691</td>\n",
       "      <td>0.924206</td>\n",
       "      <td>0.826771</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>0.911288</td>\n",
       "      <td>0.859676</td>\n",
       "      <td>0.787488</td>\n",
       "      <td>0.425466</td>\n",
       "      <td>0.199714</td>\n",
       "      <td>0.108857</td>\n",
       "      <td>0.626866</td>\n",
       "      <td>0.659009</td>\n",
       "      <td>0.621734</td>\n",
       "      <td>0.654900</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>0.927835</td>\n",
       "      <td>0.335364</td>\n",
       "      <td>0.213332</td>\n",
       "      <td>0.118058</td>\n",
       "      <td>0.349275</td>\n",
       "      <td>0.149626</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>216</td>\n",
       "      <td>5610</td>\n",
       "      <td>0.369151</td>\n",
       "      <td>0.263501</td>\n",
       "      <td>0.129978</td>\n",
       "      <td>0.776643</td>\n",
       "      <td>0.962174</td>\n",
       "      <td>0.915003</td>\n",
       "      <td>0.888912</td>\n",
       "      <td>0.906839</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.964141</td>\n",
       "      <td>0.956418</td>\n",
       "      <td>0.949035</td>\n",
       "      <td>0.807121</td>\n",
       "      <td>0.129815</td>\n",
       "      <td>0.055521</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.726462</td>\n",
       "      <td>0.854291</td>\n",
       "      <td>0.742554</td>\n",
       "      <td>0.910256</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>0.336720</td>\n",
       "      <td>0.322934</td>\n",
       "      <td>0.165499</td>\n",
       "      <td>0.746454</td>\n",
       "      <td>0.249244</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>216</td>\n",
       "      <td>6171</td>\n",
       "      <td>0.446861</td>\n",
       "      <td>0.345385</td>\n",
       "      <td>0.229512</td>\n",
       "      <td>0.981465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.985814</td>\n",
       "      <td>0.996383</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.520743</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.247915</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.989290</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.981289</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.728261</td>\n",
       "      <td>0.361544</td>\n",
       "      <td>0.331911</td>\n",
       "      <td>0.213662</td>\n",
       "      <td>0.444508</td>\n",
       "      <td>0.615487</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>201</td>\n",
       "      <td>216</td>\n",
       "      <td>0.596968</td>\n",
       "      <td>0.205333</td>\n",
       "      <td>0.105997</td>\n",
       "      <td>0.917820</td>\n",
       "      <td>0.989897</td>\n",
       "      <td>0.998883</td>\n",
       "      <td>0.976154</td>\n",
       "      <td>0.941253</td>\n",
       "      <td>0.953125</td>\n",
       "      <td>0.980870</td>\n",
       "      <td>0.970181</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>0.594901</td>\n",
       "      <td>0.287441</td>\n",
       "      <td>0.194935</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>0.875923</td>\n",
       "      <td>0.907908</td>\n",
       "      <td>0.903044</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.435394</td>\n",
       "      <td>0.194783</td>\n",
       "      <td>0.129693</td>\n",
       "      <td>0.514150</td>\n",
       "      <td>0.361812</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>216</td>\n",
       "      <td>4451</td>\n",
       "      <td>0.593152</td>\n",
       "      <td>0.488502</td>\n",
       "      <td>0.474958</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.273178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963589</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.946480</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.369786</td>\n",
       "      <td>0.578140</td>\n",
       "      <td>0.479952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.785507</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>216</td>\n",
       "      <td>805</td>\n",
       "      <td>0.538554</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.435646</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.738420</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>216</td>\n",
       "      <td>1217</td>\n",
       "      <td>0.439377</td>\n",
       "      <td>0.280116</td>\n",
       "      <td>0.164259</td>\n",
       "      <td>0.975321</td>\n",
       "      <td>0.909093</td>\n",
       "      <td>0.993256</td>\n",
       "      <td>0.999985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.336211</td>\n",
       "      <td>0.271480</td>\n",
       "      <td>0.913793</td>\n",
       "      <td>0.933396</td>\n",
       "      <td>0.922456</td>\n",
       "      <td>0.947386</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.343750</td>\n",
       "      <td>0.252482</td>\n",
       "      <td>0.167218</td>\n",
       "      <td>0.619681</td>\n",
       "      <td>0.506773</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>216</td>\n",
       "      <td>2416</td>\n",
       "      <td>0.322017</td>\n",
       "      <td>0.228381</td>\n",
       "      <td>0.206814</td>\n",
       "      <td>0.817604</td>\n",
       "      <td>0.918954</td>\n",
       "      <td>0.788922</td>\n",
       "      <td>0.635946</td>\n",
       "      <td>0.830411</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>0.857558</td>\n",
       "      <td>0.866754</td>\n",
       "      <td>0.974802</td>\n",
       "      <td>0.825326</td>\n",
       "      <td>0.343586</td>\n",
       "      <td>0.177307</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.889422</td>\n",
       "      <td>0.822515</td>\n",
       "      <td>0.945946</td>\n",
       "      <td>0.880435</td>\n",
       "      <td>0.125491</td>\n",
       "      <td>0.258668</td>\n",
       "      <td>0.180932</td>\n",
       "      <td>0.683772</td>\n",
       "      <td>0.249061</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>216</td>\n",
       "      <td>4270</td>\n",
       "      <td>0.470163</td>\n",
       "      <td>0.228094</td>\n",
       "      <td>0.140422</td>\n",
       "      <td>0.980815</td>\n",
       "      <td>0.999690</td>\n",
       "      <td>0.987467</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.801536</td>\n",
       "      <td>0.292523</td>\n",
       "      <td>0.205338</td>\n",
       "      <td>0.867647</td>\n",
       "      <td>0.927024</td>\n",
       "      <td>0.923473</td>\n",
       "      <td>0.945773</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.401946</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>0.165360</td>\n",
       "      <td>0.768092</td>\n",
       "      <td>0.542549</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets  known  predicted  orthologs  eqs_distance  curated\n",
       "0    216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0   -1.0       -1.0       -1.0           1.0     True\n",
       "1    216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0    0.0        0.0       -1.0           1.0     True\n",
       "2    216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1    0.0        0.0       -1.0          -1.0    False\n",
       "3    216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0   -1.0       -1.0       -1.0           1.0     True\n",
       "4    216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1    0.0        0.0       -1.0          -1.0    False\n",
       "5      4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1    0.0        0.0       -1.0           1.0     True\n",
       "6    216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0   -1.0       -1.0       -1.0           1.0     True\n",
       "7    216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1    0.0        0.0       -1.0          -1.0    False\n",
       "8    216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1    0.0        0.0       -1.0          -1.0    False\n",
       "9    216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0    0.0        0.0       -1.0           1.0     True\n",
       "10   216  2024                    0.428118                          0.112537                         0.085785                         0.680843                         0.841399                          0.917374                          0.590837                    0.807489                           0.938272                          0.919413                                 0.957781                               0.861796                              0.350995                                        0.312535                                        0.224063             0.800000             0.826178                   0.903649                   0.851012  1.000000  0.010870                           0.315223                                 0.159200                                0.111529                                     0.278094  0.201760         0        0    0.0        0.0       -1.0           1.0     True\n",
       "11   216  5052                    0.424170                          0.139413                         0.091400                         0.773170                         0.873296                          0.943691                          0.924206                    0.826771                           0.865672                          0.911288                                 0.859676                               0.787488                              0.425466                                        0.199714                                        0.108857             0.626866             0.659009                   0.621734                   0.654900  0.952381  0.927835                           0.335364                                 0.213332                                0.118058                                     0.349275  0.149626        -1       -1    0.0        0.0       -1.0          -1.0    False\n",
       "12   216  5610                    0.369151                          0.263501                         0.129978                         0.776643                         0.962174                          0.915003                          0.888912                    0.906839                           0.948718                          0.964141                                 0.956418                               0.949035                              0.807121                                        0.129815                                        0.055521             0.774194             0.726462                   0.854291                   0.742554  0.910256  0.054348                           0.336720                                 0.322934                                0.165499                                     0.746454  0.249244         0        0    0.0        0.0       -1.0           1.0     True\n",
       "13   216  6171                    0.446861                          0.345385                         0.229512                         0.981465                         1.000000                          0.985814                          0.996383                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.520743                                        1.000000                                        0.247915             1.000000             0.989290                   1.000000                   0.981289  0.972973  0.728261                           0.361544                                 0.331911                                0.213662                                     0.444508  0.615487        -1       -1    0.0        0.0       -1.0          -1.0    False\n",
       "14   201   216                    0.596968                          0.205333                         0.105997                         0.917820                         0.989897                          0.998883                          0.976154                    0.941253                           0.953125                          0.980870                                 0.970181                               0.943656                              0.594901                                        0.287441                                        0.194935             0.806452             0.875923                   0.907908                   0.903044  0.926471  0.042105                           0.435394                                 0.194783                                0.129693                                     0.514150  0.361812        -1        0    0.0        0.0       -1.0           1.0     True\n",
       "15   216  4451                    0.593152                          0.488502                         0.474958                         1.000000                         1.000000                          1.000000                          0.999997                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        1.000000                                        0.273178             1.000000             0.963589                   1.000000                   0.946480  0.945946  0.880435                           0.369786                                 0.578140                                0.479952                                     1.000000  0.785507        -1       -1    0.0        0.0       -1.0          -1.0    False\n",
       "16   216   805                    0.538554                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.435646                                 0.588919                                0.803272                                     1.000000  0.738420        -1        0   -1.0       -1.0       -1.0          -1.0    False\n",
       "17   216  1217                    0.439377                          0.280116                         0.164259                         0.975321                         0.909093                          0.993256                          0.999985                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.687500                                        0.336211                                        0.271480             0.913793             0.933396                   0.922456                   0.947386  0.955556  1.000000                           0.343750                                 0.252482                                0.167218                                     0.619681  0.506773        -1        0   -1.0       -1.0       -1.0          -1.0    False\n",
       "18   216  2416                    0.322017                          0.228381                         0.206814                         0.817604                         0.918954                          0.788922                          0.635946                    0.830411                           0.942857                          0.857558                                 0.866754                               0.974802                              0.825326                                        0.343586                                        0.177307             0.847458             0.842697                   0.889422                   0.822515  0.945946  0.880435                           0.125491                                 0.258668                                0.180932                                     0.683772  0.249061        -1        0   -1.0       -1.0       -1.0           0.9     True\n",
       "19   216  4270                    0.470163                          0.228094                         0.140422                         0.980815                         0.999690                          0.987467                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.801536                                        0.292523                                        0.205338             0.867647             0.927024                   0.923473                   0.945773  0.972973  0.090000                           0.401946                                 0.322728                                0.165360                                     0.768092  0.542549         0       -1    0.0        0.0       -1.0          -1.0    False"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column that indicates whether or not both genes of the pair are mapped to all the curation types.\n",
    "relevant_ids = set(ids_with_all_annotations)\n",
    "df[\"from_is_valid\"] = df[\"from\"].map(lambda x: x in relevant_ids)\n",
    "df[\"to_is_valid\"] = df[\"to\"].map(lambda x: x in relevant_ids)\n",
    "df[\"curated\"] = df[\"from_is_valid\"]*df[\"to_is_valid\"]\n",
    "df.drop(labels=[\"from_is_valid\",\"to_is_valid\"], axis=\"columns\", inplace=True)\n",
    "\n",
    "df.head(20)   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking whether gene pairs are intraspecies or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "      <th>pathways</th>\n",
       "      <th>subsets</th>\n",
       "      <th>known</th>\n",
       "      <th>predicted</th>\n",
       "      <th>orthologs</th>\n",
       "      <th>eqs_distance</th>\n",
       "      <th>curated</th>\n",
       "      <th>same</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>216</td>\n",
       "      <td>796</td>\n",
       "      <td>0.536370</td>\n",
       "      <td>0.616745</td>\n",
       "      <td>0.878329</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.257021</td>\n",
       "      <td>0.280834</td>\n",
       "      <td>0.885246</td>\n",
       "      <td>0.941406</td>\n",
       "      <td>0.915908</td>\n",
       "      <td>0.920537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945652</td>\n",
       "      <td>0.427748</td>\n",
       "      <td>0.588919</td>\n",
       "      <td>0.803272</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.737313</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>216</td>\n",
       "      <td>1422</td>\n",
       "      <td>0.433818</td>\n",
       "      <td>0.346134</td>\n",
       "      <td>0.165990</td>\n",
       "      <td>0.999837</td>\n",
       "      <td>0.993298</td>\n",
       "      <td>0.885720</td>\n",
       "      <td>0.999828</td>\n",
       "      <td>0.973302</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.987683</td>\n",
       "      <td>0.980345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.651355</td>\n",
       "      <td>0.531820</td>\n",
       "      <td>0.129690</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.837121</td>\n",
       "      <td>0.961398</td>\n",
       "      <td>0.861081</td>\n",
       "      <td>0.895522</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.348839</td>\n",
       "      <td>0.351609</td>\n",
       "      <td>0.174732</td>\n",
       "      <td>0.578073</td>\n",
       "      <td>0.435279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216</td>\n",
       "      <td>4105</td>\n",
       "      <td>0.458176</td>\n",
       "      <td>0.236643</td>\n",
       "      <td>0.133404</td>\n",
       "      <td>0.895387</td>\n",
       "      <td>0.929673</td>\n",
       "      <td>0.993294</td>\n",
       "      <td>0.980822</td>\n",
       "      <td>0.922791</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0.971135</td>\n",
       "      <td>0.959797</td>\n",
       "      <td>0.880477</td>\n",
       "      <td>0.872422</td>\n",
       "      <td>0.324863</td>\n",
       "      <td>0.182087</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.890090</td>\n",
       "      <td>0.926680</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.406371</td>\n",
       "      <td>0.344659</td>\n",
       "      <td>0.153674</td>\n",
       "      <td>0.811018</td>\n",
       "      <td>0.383555</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>216</td>\n",
       "      <td>402</td>\n",
       "      <td>0.592512</td>\n",
       "      <td>0.625898</td>\n",
       "      <td>0.472728</td>\n",
       "      <td>0.957836</td>\n",
       "      <td>0.905239</td>\n",
       "      <td>0.754834</td>\n",
       "      <td>0.816480</td>\n",
       "      <td>0.909464</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.934251</td>\n",
       "      <td>0.908423</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.948969</td>\n",
       "      <td>0.413385</td>\n",
       "      <td>0.290820</td>\n",
       "      <td>0.921875</td>\n",
       "      <td>0.941032</td>\n",
       "      <td>0.966350</td>\n",
       "      <td>0.933232</td>\n",
       "      <td>0.976744</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.485366</td>\n",
       "      <td>0.623417</td>\n",
       "      <td>0.446650</td>\n",
       "      <td>0.781782</td>\n",
       "      <td>0.530239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>216</td>\n",
       "      <td>4713</td>\n",
       "      <td>0.410375</td>\n",
       "      <td>0.147384</td>\n",
       "      <td>0.103271</td>\n",
       "      <td>0.891950</td>\n",
       "      <td>0.892315</td>\n",
       "      <td>0.902685</td>\n",
       "      <td>0.664281</td>\n",
       "      <td>0.923654</td>\n",
       "      <td>0.966102</td>\n",
       "      <td>0.969605</td>\n",
       "      <td>0.971432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.587756</td>\n",
       "      <td>0.166263</td>\n",
       "      <td>0.206348</td>\n",
       "      <td>0.747368</td>\n",
       "      <td>0.844780</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>0.850401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803922</td>\n",
       "      <td>0.334199</td>\n",
       "      <td>0.272367</td>\n",
       "      <td>0.147673</td>\n",
       "      <td>0.622036</td>\n",
       "      <td>0.264044</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>216</td>\n",
       "      <td>0.455360</td>\n",
       "      <td>0.182428</td>\n",
       "      <td>0.108639</td>\n",
       "      <td>0.867171</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>0.997674</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.982922</td>\n",
       "      <td>0.963770</td>\n",
       "      <td>0.919863</td>\n",
       "      <td>0.482446</td>\n",
       "      <td>0.280443</td>\n",
       "      <td>0.038583</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.654663</td>\n",
       "      <td>0.835842</td>\n",
       "      <td>0.655664</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.336356</td>\n",
       "      <td>0.232288</td>\n",
       "      <td>0.165308</td>\n",
       "      <td>0.372505</td>\n",
       "      <td>0.270266</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>216</td>\n",
       "      <td>1836</td>\n",
       "      <td>0.480718</td>\n",
       "      <td>0.302177</td>\n",
       "      <td>0.226400</td>\n",
       "      <td>0.999457</td>\n",
       "      <td>0.976573</td>\n",
       "      <td>0.999339</td>\n",
       "      <td>0.996973</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.817884</td>\n",
       "      <td>0.422911</td>\n",
       "      <td>0.286628</td>\n",
       "      <td>0.949153</td>\n",
       "      <td>0.914345</td>\n",
       "      <td>0.973037</td>\n",
       "      <td>0.936882</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>0.317276</td>\n",
       "      <td>0.266961</td>\n",
       "      <td>0.216232</td>\n",
       "      <td>0.661938</td>\n",
       "      <td>0.556195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216</td>\n",
       "      <td>5758</td>\n",
       "      <td>0.348328</td>\n",
       "      <td>0.165209</td>\n",
       "      <td>0.120374</td>\n",
       "      <td>0.684373</td>\n",
       "      <td>0.841409</td>\n",
       "      <td>0.803934</td>\n",
       "      <td>0.995512</td>\n",
       "      <td>0.794684</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.844979</td>\n",
       "      <td>0.860409</td>\n",
       "      <td>0.853615</td>\n",
       "      <td>0.612965</td>\n",
       "      <td>0.203988</td>\n",
       "      <td>0.175869</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.640049</td>\n",
       "      <td>0.627956</td>\n",
       "      <td>0.553247</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.354914</td>\n",
       "      <td>0.237185</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.651089</td>\n",
       "      <td>0.170494</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>216</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.484129</td>\n",
       "      <td>0.289102</td>\n",
       "      <td>0.291561</td>\n",
       "      <td>0.489068</td>\n",
       "      <td>0.735018</td>\n",
       "      <td>0.995637</td>\n",
       "      <td>0.749639</td>\n",
       "      <td>0.871963</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>0.945263</td>\n",
       "      <td>0.961880</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.675240</td>\n",
       "      <td>0.308672</td>\n",
       "      <td>0.508967</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.991935</td>\n",
       "      <td>0.975562</td>\n",
       "      <td>0.986666</td>\n",
       "      <td>0.903226</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>0.360427</td>\n",
       "      <td>0.370360</td>\n",
       "      <td>0.299303</td>\n",
       "      <td>0.707230</td>\n",
       "      <td>0.482518</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>216</td>\n",
       "      <td>1613</td>\n",
       "      <td>0.454203</td>\n",
       "      <td>0.450065</td>\n",
       "      <td>0.224581</td>\n",
       "      <td>0.769878</td>\n",
       "      <td>0.923144</td>\n",
       "      <td>0.942272</td>\n",
       "      <td>0.996709</td>\n",
       "      <td>0.884186</td>\n",
       "      <td>0.972973</td>\n",
       "      <td>0.902283</td>\n",
       "      <td>0.952045</td>\n",
       "      <td>0.874447</td>\n",
       "      <td>0.920050</td>\n",
       "      <td>0.270622</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.836066</td>\n",
       "      <td>0.883069</td>\n",
       "      <td>0.770878</td>\n",
       "      <td>0.866102</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>0.409886</td>\n",
       "      <td>0.458758</td>\n",
       "      <td>0.233068</td>\n",
       "      <td>0.646447</td>\n",
       "      <td>0.361178</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   from    to  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean  pathways  subsets  known  predicted  orthologs  eqs_distance  curated   same\n",
       "0   216   796                    0.536370                          0.616745                         0.878329                         0.999791                         1.000000                          1.000000                          1.000000                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              1.000000                                        0.257021                                        0.280834             0.885246             0.941406                   0.915908                   0.920537  1.000000  0.945652                           0.427748                                 0.588919                                0.803272                                     1.000000  0.737313        -1        0   -1.0       -1.0       -1.0           1.0     True   True\n",
       "1   216  1422                    0.433818                          0.346134                         0.165990                         0.999837                         0.993298                          0.885720                          0.999828                    0.973302                           0.977273                          0.987683                                 0.980345                               1.000000                              0.651355                                        0.531820                                        0.129690             0.921875             0.837121                   0.961398                   0.861081  0.895522  0.869565                           0.348839                                 0.351609                                0.174732                                     0.578073  0.435279         0        0    0.0        0.0       -1.0           1.0     True   True\n",
       "2   216  4105                    0.458176                          0.236643                         0.133404                         0.895387                         0.929673                          0.993294                          0.980822                    0.922791                           0.948718                          0.971135                                 0.959797                               0.880477                              0.872422                                        0.324863                                        0.182087             0.833333             0.890090                   0.926680                   0.924508  0.947368  0.869565                           0.406371                                 0.344659                                0.153674                                     0.811018  0.383555        -1       -1    0.0        0.0       -1.0          -1.0    False   True\n",
       "3   216   402                    0.592512                          0.625898                         0.472728                         0.957836                         0.905239                          0.754834                          0.816480                    0.909464                           0.967742                          0.934251                                 0.908423                               0.830969                              0.948969                                        0.413385                                        0.290820             0.921875             0.941032                   0.966350                   0.933232  0.976744  1.000000                           0.485366                                 0.623417                                0.446650                                     0.781782  0.530239         0        0   -1.0       -1.0       -1.0           1.0     True   True\n",
       "4   216  4713                    0.410375                          0.147384                         0.103271                         0.891950                         0.892315                          0.902685                          0.664281                    0.923654                           0.966102                          0.969605                                 0.971432                               1.000000                              0.587756                                        0.166263                                        0.206348             0.747368             0.844780                   0.821321                   0.850401  1.000000  0.803922                           0.334199                                 0.272367                                0.147673                                     0.622036  0.264044        -1       -1    0.0        0.0       -1.0          -1.0    False   True\n",
       "5     4   216                    0.455360                          0.182428                         0.108639                         0.867171                         0.945635                          0.999989                          0.997674                    0.952853                           0.947368                          0.982922                                 0.963770                               0.919863                              0.482446                                        0.280443                                        0.038583             0.764706             0.654663                   0.835842                   0.655664  1.000000  0.757576                           0.336356                                 0.232288                                0.165308                                     0.372505  0.270266        -1       -1    0.0        0.0       -1.0           1.0     True  False\n",
       "6   216  1836                    0.480718                          0.302177                         0.226400                         0.999457                         0.976573                          0.999339                          0.996973                    1.000000                           1.000000                          1.000000                                 1.000000                               1.000000                              0.817884                                        0.422911                                        0.286628             0.949153             0.914345                   0.973037                   0.936882  0.941176  0.956522                           0.317276                                 0.266961                                0.216232                                     0.661938  0.556195         0        0   -1.0       -1.0       -1.0           1.0     True   True\n",
       "7   216  5758                    0.348328                          0.165209                         0.120374                         0.684373                         0.841409                          0.803934                          0.995512                    0.794684                           0.900000                          0.844979                                 0.860409                               0.853615                              0.612965                                        0.203988                                        0.175869             0.666667             0.640049                   0.627956                   0.553247  1.000000  1.000000                           0.354914                                 0.237185                                0.129197                                     0.651089  0.170494         0       -1    0.0        0.0       -1.0          -1.0    False   True\n",
       "8   216  4488                    0.484129                          0.289102                         0.291561                         0.489068                         0.735018                          0.995637                          0.749639                    0.871963                           0.969697                          0.945263                                 0.961880                               1.000000                              0.675240                                        0.308672                                        0.508967             0.947368             0.991935                   0.975562                   0.986666  0.903226  0.141509                           0.360427                                 0.370360                                0.299303                                     0.707230  0.482518        -1       -1    0.0        0.0       -1.0          -1.0    False   True\n",
       "9   216  1613                    0.454203                          0.450065                         0.224581                         0.769878                         0.923144                          0.942272                          0.996709                    0.884186                           0.972973                          0.902283                                 0.952045                               0.874447                              0.920050                                        0.270622                                        0.042017             0.836066             0.883069                   0.770878                   0.866102  0.894737  0.823529                           0.409886                                 0.458758                                0.233068                                     0.646447  0.361178        -1        0    0.0        0.0       -1.0           1.0     True   True"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_dict = dataset.get_species_dictionary()\n",
    "df[\"same\"] = df[[\"from\",\"to\"]].apply(lambda x: species_dict[x[\"from\"]]==species_dict[x[\"to\"]],axis=1)\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking to make sure that the number of genes and pairs matches what is expected at this point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining a nested dictionary with shape dict[curated][question][species][approach][metric] --> value.\n",
    "curated = [True,False]\n",
    "species = [\"intra\",\"inter\",\"both\"]\n",
    "question = [\"subsets\", \"pathways\", \"known\", \"predicted\", \"orthologs\"]\n",
    "tables = defaultdict(dict)\n",
    "for c,q in itertools.product(curated,question): \n",
    "    tables[c][q] = defaultdict(dict)\n",
    "for c,q,s in itertools.product(curated,question,species): \n",
    "    tables[c][q][s] = defaultdict(dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making the tables now\n"
     ]
    }
   ],
   "source": [
    "print(\"Making the tables now\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the value of *n* for each type of iteration through a subset of the dataset?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>question</th>\n",
       "      <th>curated</th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "      <th>num_pairs</th>\n",
       "      <th>positive_fraction</th>\n",
       "      <th>negative_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>intra</td>\n",
       "      <td>215</td>\n",
       "      <td>10867</td>\n",
       "      <td>21890</td>\n",
       "      <td>32757</td>\n",
       "      <td>0.331746</td>\n",
       "      <td>0.668254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>inter</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>both</td>\n",
       "      <td>215</td>\n",
       "      <td>10867</td>\n",
       "      <td>21890</td>\n",
       "      <td>32757</td>\n",
       "      <td>0.331746</td>\n",
       "      <td>0.668254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pathways</td>\n",
       "      <td>true</td>\n",
       "      <td>intra</td>\n",
       "      <td>102</td>\n",
       "      <td>6066</td>\n",
       "      <td>5122</td>\n",
       "      <td>11188</td>\n",
       "      <td>0.542188</td>\n",
       "      <td>0.457812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pathways</td>\n",
       "      <td>true</td>\n",
       "      <td>inter</td>\n",
       "      <td>105</td>\n",
       "      <td>11</td>\n",
       "      <td>301</td>\n",
       "      <td>312</td>\n",
       "      <td>0.035256</td>\n",
       "      <td>0.964744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pathways</td>\n",
       "      <td>true</td>\n",
       "      <td>both</td>\n",
       "      <td>105</td>\n",
       "      <td>6077</td>\n",
       "      <td>5423</td>\n",
       "      <td>11500</td>\n",
       "      <td>0.528435</td>\n",
       "      <td>0.471565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>known</td>\n",
       "      <td>true</td>\n",
       "      <td>intra</td>\n",
       "      <td>191</td>\n",
       "      <td>6058</td>\n",
       "      <td>19928</td>\n",
       "      <td>25986</td>\n",
       "      <td>0.233126</td>\n",
       "      <td>0.766874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>known</td>\n",
       "      <td>true</td>\n",
       "      <td>inter</td>\n",
       "      <td>191</td>\n",
       "      <td>0</td>\n",
       "      <td>1845</td>\n",
       "      <td>1845</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>known</td>\n",
       "      <td>true</td>\n",
       "      <td>both</td>\n",
       "      <td>191</td>\n",
       "      <td>6058</td>\n",
       "      <td>21773</td>\n",
       "      <td>27831</td>\n",
       "      <td>0.217671</td>\n",
       "      <td>0.782329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>predicted</td>\n",
       "      <td>true</td>\n",
       "      <td>intra</td>\n",
       "      <td>191</td>\n",
       "      <td>5381</td>\n",
       "      <td>20605</td>\n",
       "      <td>25986</td>\n",
       "      <td>0.207073</td>\n",
       "      <td>0.792927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>predicted</td>\n",
       "      <td>true</td>\n",
       "      <td>inter</td>\n",
       "      <td>191</td>\n",
       "      <td>0</td>\n",
       "      <td>1845</td>\n",
       "      <td>1845</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>predicted</td>\n",
       "      <td>true</td>\n",
       "      <td>both</td>\n",
       "      <td>191</td>\n",
       "      <td>5381</td>\n",
       "      <td>22450</td>\n",
       "      <td>27831</td>\n",
       "      <td>0.193346</td>\n",
       "      <td>0.806654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>orthologs</td>\n",
       "      <td>true</td>\n",
       "      <td>intra</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>1232</td>\n",
       "      <td>1232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>orthologs</td>\n",
       "      <td>true</td>\n",
       "      <td>inter</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>489</td>\n",
       "      <td>489</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>orthologs</td>\n",
       "      <td>true</td>\n",
       "      <td>both</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>1721</td>\n",
       "      <td>1721</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>subsets</td>\n",
       "      <td>false</td>\n",
       "      <td>intra</td>\n",
       "      <td>281</td>\n",
       "      <td>12897</td>\n",
       "      <td>36348</td>\n",
       "      <td>49245</td>\n",
       "      <td>0.261895</td>\n",
       "      <td>0.738105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>subsets</td>\n",
       "      <td>false</td>\n",
       "      <td>inter</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>subsets</td>\n",
       "      <td>false</td>\n",
       "      <td>both</td>\n",
       "      <td>281</td>\n",
       "      <td>12897</td>\n",
       "      <td>36348</td>\n",
       "      <td>49245</td>\n",
       "      <td>0.261895</td>\n",
       "      <td>0.738105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>pathways</td>\n",
       "      <td>false</td>\n",
       "      <td>intra</td>\n",
       "      <td>203</td>\n",
       "      <td>7152</td>\n",
       "      <td>17049</td>\n",
       "      <td>24201</td>\n",
       "      <td>0.295525</td>\n",
       "      <td>0.704475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pathways</td>\n",
       "      <td>false</td>\n",
       "      <td>inter</td>\n",
       "      <td>205</td>\n",
       "      <td>75</td>\n",
       "      <td>3497</td>\n",
       "      <td>3572</td>\n",
       "      <td>0.020997</td>\n",
       "      <td>0.979003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pathways</td>\n",
       "      <td>false</td>\n",
       "      <td>both</td>\n",
       "      <td>205</td>\n",
       "      <td>7227</td>\n",
       "      <td>20546</td>\n",
       "      <td>27773</td>\n",
       "      <td>0.260217</td>\n",
       "      <td>0.739783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>known</td>\n",
       "      <td>false</td>\n",
       "      <td>intra</td>\n",
       "      <td>372</td>\n",
       "      <td>7977</td>\n",
       "      <td>66121</td>\n",
       "      <td>74098</td>\n",
       "      <td>0.107655</td>\n",
       "      <td>0.892345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>known</td>\n",
       "      <td>false</td>\n",
       "      <td>inter</td>\n",
       "      <td>373</td>\n",
       "      <td>0</td>\n",
       "      <td>6819</td>\n",
       "      <td>6819</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>known</td>\n",
       "      <td>false</td>\n",
       "      <td>both</td>\n",
       "      <td>373</td>\n",
       "      <td>7977</td>\n",
       "      <td>72940</td>\n",
       "      <td>80917</td>\n",
       "      <td>0.098582</td>\n",
       "      <td>0.901418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>predicted</td>\n",
       "      <td>false</td>\n",
       "      <td>intra</td>\n",
       "      <td>372</td>\n",
       "      <td>6380</td>\n",
       "      <td>67718</td>\n",
       "      <td>74098</td>\n",
       "      <td>0.086102</td>\n",
       "      <td>0.913898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>predicted</td>\n",
       "      <td>false</td>\n",
       "      <td>inter</td>\n",
       "      <td>373</td>\n",
       "      <td>0</td>\n",
       "      <td>6819</td>\n",
       "      <td>6819</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>predicted</td>\n",
       "      <td>false</td>\n",
       "      <td>both</td>\n",
       "      <td>373</td>\n",
       "      <td>6380</td>\n",
       "      <td>74537</td>\n",
       "      <td>80917</td>\n",
       "      <td>0.078846</td>\n",
       "      <td>0.921154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>intra</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>2386</td>\n",
       "      <td>2386</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>inter</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>2088</td>\n",
       "      <td>2089</td>\n",
       "      <td>0.000479</td>\n",
       "      <td>0.999521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>both</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>4474</td>\n",
       "      <td>4475</td>\n",
       "      <td>0.000223</td>\n",
       "      <td>0.999777</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     question curated species  num_genes  positive  negative  num_pairs  positive_fraction  negative_fraction\n",
       "0     subsets    true   intra        215     10867     21890      32757           0.331746           0.668254\n",
       "1     subsets    true   inter          0         0         0          0                NaN                NaN\n",
       "2     subsets    true    both        215     10867     21890      32757           0.331746           0.668254\n",
       "3    pathways    true   intra        102      6066      5122      11188           0.542188           0.457812\n",
       "4    pathways    true   inter        105        11       301        312           0.035256           0.964744\n",
       "5    pathways    true    both        105      6077      5423      11500           0.528435           0.471565\n",
       "6       known    true   intra        191      6058     19928      25986           0.233126           0.766874\n",
       "7       known    true   inter        191         0      1845       1845           0.000000           1.000000\n",
       "8       known    true    both        191      6058     21773      27831           0.217671           0.782329\n",
       "9   predicted    true   intra        191      5381     20605      25986           0.207073           0.792927\n",
       "10  predicted    true   inter        191         0      1845       1845           0.000000           1.000000\n",
       "11  predicted    true    both        191      5381     22450      27831           0.193346           0.806654\n",
       "12  orthologs    true   intra         36         0      1232       1232           0.000000           1.000000\n",
       "13  orthologs    true   inter         37         0       489        489           0.000000           1.000000\n",
       "14  orthologs    true    both         37         0      1721       1721           0.000000           1.000000\n",
       "15    subsets   false   intra        281     12897     36348      49245           0.261895           0.738105\n",
       "16    subsets   false   inter          0         0         0          0                NaN                NaN\n",
       "17    subsets   false    both        281     12897     36348      49245           0.261895           0.738105\n",
       "18   pathways   false   intra        203      7152     17049      24201           0.295525           0.704475\n",
       "19   pathways   false   inter        205        75      3497       3572           0.020997           0.979003\n",
       "20   pathways   false    both        205      7227     20546      27773           0.260217           0.739783\n",
       "21      known   false   intra        372      7977     66121      74098           0.107655           0.892345\n",
       "22      known   false   inter        373         0      6819       6819           0.000000           1.000000\n",
       "23      known   false    both        373      7977     72940      80917           0.098582           0.901418\n",
       "24  predicted   false   intra        372      6380     67718      74098           0.086102           0.913898\n",
       "25  predicted   false   inter        373         0      6819       6819           0.000000           1.000000\n",
       "26  predicted   false    both        373      6380     74537      80917           0.078846           0.921154\n",
       "27  orthologs   false   intra         80         0      2386       2386           0.000000           1.000000\n",
       "28  orthologs   false   inter         81         1      2088       2089           0.000479           0.999521\n",
       "29  orthologs   false    both         81         1      4474       4475           0.000223           0.999777"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_idx_lists = []\n",
    "subset_properties = []\n",
    "table_lists = defaultdict(list)\n",
    "for c,q,s in itertools.product(curated,question,species):\n",
    "    \n",
    "    # Remembering what the properties for this particular subset are.\n",
    "    subset_properties.append((c,q,s))\n",
    "    \n",
    "    # Subsetting the dataframe to the rows (gene pairs) that are relevant for this particular biological question.\n",
    "    subset = df[df[q] != -1]\n",
    "    if c:\n",
    "        subset = subset[subset[\"curated\"] == True]\n",
    "        \n",
    "        \n",
    "    # Subsetting the dataframe to the rows (gene pairs) where both genes are from the same or different species.\n",
    "    if s == \"intra\":\n",
    "        subset = subset[subset[\"same\"] == True]\n",
    "    elif s == \"inter\":\n",
    "        subset = subset[subset[\"same\"] == False]\n",
    "        \n",
    "    subset_idx_lists.append(subset.index.to_list())\n",
    "    \n",
    "    # Adding values to the table that are specific to this biological question.\n",
    "    counts = Counter(subset[q].values)\n",
    "    \n",
    "    table_lists[\"question\"].append(q.lower())\n",
    "    table_lists[\"curated\"].append(str(c).lower())\n",
    "    table_lists[\"species\"].append(s.lower())\n",
    "    table_lists[\"num_genes\"].append(len(set(subset[\"to\"].values).union(set(subset[\"from\"].values))))\n",
    "    table_lists[\"positive\"].append(counts[1])\n",
    "    table_lists[\"negative\"].append(counts[0])\n",
    "    #table_lists[\"class_ratio\"].append(\"{:0.4f}\".format(counts[1]/counts[0]))\n",
    "\n",
    "pairs_table = pd.DataFrame(table_lists)  \n",
    "pairs_table[\"num_pairs\"] = pairs_table[\"positive\"]+pairs_table[\"negative\"]\n",
    "pairs_table[\"positive_fraction\"] = pairs_table[\"positive\"] / pairs_table[\"num_pairs\"]\n",
    "pairs_table[\"negative_fraction\"] = pairs_table[\"negative\"] / pairs_table[\"num_pairs\"]\n",
    "pairs_table.to_csv(os.path.join(OUTPUT_DIR,\"part_5_biological_question_n_values.csv\"), index=False)\n",
    "pairs_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>question_1</th>\n",
       "      <th>question_2</th>\n",
       "      <th>num_pairs_1</th>\n",
       "      <th>num_pairs_2</th>\n",
       "      <th>num_overlap</th>\n",
       "      <th>sim_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subsets</td>\n",
       "      <td>pathways</td>\n",
       "      <td>49245</td>\n",
       "      <td>27773</td>\n",
       "      <td>14637</td>\n",
       "      <td>0.651563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pathways</td>\n",
       "      <td>known</td>\n",
       "      <td>27773</td>\n",
       "      <td>80917</td>\n",
       "      <td>17703</td>\n",
       "      <td>0.542581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pathways</td>\n",
       "      <td>predicted</td>\n",
       "      <td>27773</td>\n",
       "      <td>80917</td>\n",
       "      <td>17703</td>\n",
       "      <td>0.536508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subsets</td>\n",
       "      <td>known</td>\n",
       "      <td>49245</td>\n",
       "      <td>80917</td>\n",
       "      <td>28357</td>\n",
       "      <td>0.451908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subsets</td>\n",
       "      <td>predicted</td>\n",
       "      <td>49245</td>\n",
       "      <td>80917</td>\n",
       "      <td>28357</td>\n",
       "      <td>0.366745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>known</td>\n",
       "      <td>predicted</td>\n",
       "      <td>80917</td>\n",
       "      <td>80917</td>\n",
       "      <td>80917</td>\n",
       "      <td>0.163829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>subsets</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>49245</td>\n",
       "      <td>4475</td>\n",
       "      <td>1153</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pathways</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>27773</td>\n",
       "      <td>4475</td>\n",
       "      <td>1383</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>known</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>80917</td>\n",
       "      <td>4475</td>\n",
       "      <td>2163</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>predicted</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>80917</td>\n",
       "      <td>4475</td>\n",
       "      <td>2163</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  question_1 question_2  num_pairs_1  num_pairs_2  num_overlap  sim_overlap\n",
       "0    subsets   pathways        49245        27773        14637     0.651563\n",
       "1   pathways      known        27773        80917        17703     0.542581\n",
       "2   pathways  predicted        27773        80917        17703     0.536508\n",
       "3    subsets      known        49245        80917        28357     0.451908\n",
       "4    subsets  predicted        49245        80917        28357     0.366745\n",
       "5      known  predicted        80917        80917        80917     0.163829\n",
       "6    subsets  orthologs        49245         4475         1153     0.000000\n",
       "7   pathways  orthologs        27773         4475         1383     0.000000\n",
       "8      known  orthologs        80917         4475         2163     0.000000\n",
       "9  predicted  orthologs        80917         4475         2163     0.000000"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking more at the distributions of target values for each of the biological questions.\n",
    "from scipy.spatial.distance import jaccard\n",
    "row_tuples = []\n",
    "for q1,q2 in itertools.combinations(question, 2):\n",
    "    q1_subset = df[df[q1] != -1]\n",
    "    q2_subset = df[df[q2] != -1]\n",
    "    overlap_subset  = q1_subset[q1_subset[q2] != -1]\n",
    "    q1_num_pairs = q1_subset.shape[0]\n",
    "    q2_num_pairs = q2_subset.shape[0]\n",
    "    overlap_size = overlap_subset.shape[0]\n",
    "    overlap_sim = 1-jaccard(overlap_subset[q1].values, overlap_subset[q2].values)\n",
    "    row_tuples.append((q1, q2, q1_num_pairs, q2_num_pairs, overlap_size, overlap_sim))\n",
    "question_overlaps_table = pd.DataFrame(row_tuples)\n",
    "question_overlaps_table.columns = [\"question_1\", \"question_2\", \"num_pairs_1\", \"num_pairs_2\", \"num_overlap\", \"sim_overlap\"]\n",
    "question_overlaps_table.sort_values(by=\"sim_overlap\", ascending=False, inplace=True)\n",
    "question_overlaps_table.reset_index(inplace=True, drop=True)\n",
    "question_overlaps_table.to_csv(os.path.join(OUTPUT_DIR,\"part_5_biological_question_overlaps.csv\"), index=False)\n",
    "question_overlaps_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ks\"></a>\n",
    "### Do the edges joining genes that share a group, pathway, or interaction come from a different distribution?\n",
    "The purpose of this section is to visualize kernel estimates for the distributions of distance or similarity scores generated by each of the methods tested for measuring semantic similarity or generating vector representations of the phenotype descriptions. Ideally, better methods should show better separation betwene the distributions for distance values between two genes involved in a common specified group or two genes that are not. Additionally, a statistical test is used to check whether these two distributions are significantly different from each other or not, although this is a less informative measure than the other tests used in subsequent sections, because it does not address how useful these differences in the distributions actually are for making predictions about group membership."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with all kolmogorov–smirnov tests\n"
     ]
    }
   ],
   "source": [
    "for properties,idxs in zip(subset_properties, subset_idx_lists):\n",
    "    \n",
    "    # Remember the properties for this subset being looked at, and subset the dataframe accordingly.\n",
    "    c,q,s = properties\n",
    "    subset = df.loc[idxs]\n",
    "        \n",
    "    # Check that this subsetting leaves a valid dataset with both positive and negatives samples.\n",
    "    class_values = pd.unique(subset[q].values)\n",
    "    if not (len(class_values)==2 and 0 in class_values and 1 in class_values):\n",
    "        continue\n",
    "    \n",
    "    # Use Kolmogorov-Smirnov test to see if edges between genes that share a group come from a distinct distribution.\n",
    "    ppi_pos_dict = {name:(subset[subset[q] > 0.00][name].values) for name in names}\n",
    "    ppi_neg_dict = {name:(subset[subset[q] == 0.00][name].values) for name in names}\n",
    "    for name in names:\n",
    "        stat,p = ks_2samp(ppi_pos_dict[name],ppi_neg_dict[name])\n",
    "        pos_mean = np.average(ppi_pos_dict[name])\n",
    "        neg_mean = np.average(ppi_neg_dict[name])\n",
    "        pos_n = len(ppi_pos_dict[name])\n",
    "        neg_n = len(ppi_neg_dict[name])\n",
    "        \n",
    "        tables[c][q][s][name].update({\"mean_1\":pos_mean, \"mean_0\":neg_mean, \"n_1\":pos_n, \"n_0\":neg_n})\n",
    "        tables[c][q][s][name].update({\"ks\":stat, \"ks_pval\":p})\n",
    "\n",
    "    # Show the kernel estimates for each distribution of weights for each method.\n",
    "    #num_plots, plots_per_row, row_width, row_height = (len(names), 4, 14, 3)\n",
    "    #fig,axs = plt.subplots(math.ceil(num_plots/plots_per_row), plots_per_row, squeeze=False)\n",
    "    #for name,ax in zip(names,axs.flatten()):\n",
    "    #    ax.set_title(name)\n",
    "    #    ax.set_xlabel(\"value\")\n",
    "    #    ax.set_ylabel(\"density\")\n",
    "    #    sns.kdeplot(ppi_pos_dict[name], color=\"black\", shade=False, alpha=1.0, ax=ax)\n",
    "    #    sns.kdeplot(ppi_neg_dict[name], color=\"black\", shade=True, alpha=0.1, ax=ax) \n",
    "    #fig.set_size_inches(row_width, row_height*math.ceil(num_plots/plots_per_row))\n",
    "    #fig.tight_layout()\n",
    "    #fig.savefig(os.path.join(OUTPUT_DIR,\"part_6_kernel_density.png\"),dpi=400)\n",
    "    #plt.close()\n",
    "    \n",
    "print(\"done with all kolmogorov–smirnov tests\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"within\"></a>\n",
    "### Looking at within-group or within-pathway distances in each graph\n",
    "The purpose of this section is to determine which methods generated graphs which tightly group genes which share common pathways or group membership with one another. In order to compare across different methods where the distance value distributions are different, the mean distance values for each group for each method are convereted to percentile scores. Lower percentile scores indicate that the average distance value between any two genes that belong to that group is lower than most of the distance values in the entire distribution for that method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done generating mean within-group distance values\n"
     ]
    }
   ],
   "source": [
    "# What are the different groupings we are interested in for these mean within-group distance tables?\n",
    "grouping_objects = [kegg_groups, pmn_groups, phe_subsets_groups]\n",
    "grouping_names = [\"kegg_only\",\"pmn_only\",\"subsets\"]\n",
    "for (groups,q) in zip(grouping_objects,grouping_names):\n",
    "\n",
    "    # Only look at gene pairs where both are relevant to the given biological question.\n",
    "    subset = df[df[q] != -1]\n",
    "    \n",
    "    # The grouping dictionaries for this particular biological question.    \n",
    "    id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(dataset)\n",
    "\n",
    "    # Get all the average within-group distance values for each approach.\n",
    "    group_ids = list(group_id_to_ids.keys())\n",
    "    graph = IndexedGraph(subset)\n",
    "    within_percentiles_dict = defaultdict(lambda: defaultdict(list))\n",
    "    all_weights_dict = {}\n",
    "    for name in names:\n",
    "        for group in group_ids:\n",
    "            within_ids = group_id_to_ids[group]\n",
    "            within_pairs = [(i,j) for i,j in itertools.permutations(within_ids,2)]\n",
    "            mean_weight = np.mean((graph.get_values(within_pairs, kind=name)))\n",
    "            within_percentiles_dict[name][group] = stats.percentileofscore(subset[name].values, mean_weight, kind=\"rank\")\n",
    "\n",
    "    # Generating a dataframe of percentiles of the mean in-group distance scores.\n",
    "    within_dist_data = pd.DataFrame(within_percentiles_dict)\n",
    "    within_dist_data = within_dist_data.dropna(axis=0, inplace=False)\n",
    "    within_dist_data = within_dist_data.round(4)\n",
    "\n",
    "    # Adding relevant information to this dataframe and saving.\n",
    "    # Defining mean_group_rank: the average of the individual rank given to this pathway by each approach.\n",
    "    # Defining mean_avg_pair_percentile: the average across all approaches of the average distance percentile for each gene pair.\n",
    "    within_dist_data[\"mean_group_rank\"] = within_dist_data.rank().mean(axis=1)\n",
    "    within_dist_data[\"mean_avg_pair_percentile\"] = within_dist_data.mean(axis=1)\n",
    "    within_dist_data.sort_values(by=\"mean_avg_pair_percentile\", inplace=True)\n",
    "    within_dist_data.reset_index(inplace=True)\n",
    "    within_dist_data[\"group_id\"] = within_dist_data[\"index\"]\n",
    "    within_dist_data[\"full_name\"] = within_dist_data[\"group_id\"].apply(lambda x: groups.get_long_name(x))\n",
    "    within_dist_data[\"n\"] = within_dist_data[\"group_id\"].apply(lambda x: len(group_id_to_ids[x]))\n",
    "    within_dist_data = within_dist_data[flatten([\"group_id\",\"full_name\",\"n\",\"mean_avg_pair_percentile\",\"mean_group_rank\",names])]\n",
    "    within_dist_data.to_csv(os.path.join(OUTPUT_DIR,\"part_5_{}_within_distances.csv\".format(question)), index=False)\n",
    "    within_dist_data.head(5)\n",
    "\n",
    "print(\"done generating mean within-group distance values\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"auc\"></a>\n",
    "### Predicting whether two genes belong to the same group, pathway, or share an interaction\n",
    "The purpose of this section is to see if whether or not two genes share atleast one common pathway can be predicted from the distance scores assigned using analysis of text similarity. The evaluation of predictability is done by reporting a precision and recall curve for each method, as well as remembering the area under the curve, and ratio between the area under the curve and the baseline (expected area when guessing randomly) for each method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bootstrap(fraction, num_iterations, y_true, y_prob):\n",
    "    # Run the desired number of bootstrap iterations over the full population of predictions and return st devs.\n",
    "    scores = pd.DataFrame([bootstrap_iteration(fraction, y_true, y_prob) for i in range(num_iterations)])\n",
    "    standard_deviations = {\n",
    "        \"f_1_max_std\": np.std(scores[\"f_1_max\"].values),\n",
    "        \"f_2_max_std\": np.std(scores[\"f_2_max\"].values),\n",
    "        \"f_point5_max_std\": np.std(scores[\"f_point5_max\"].values)}\n",
    "    return(standard_deviations)\n",
    "\n",
    "\n",
    "def bootstrap_iteration(fraction, y_true, y_prob):\n",
    "    assert len(y_true) == len(y_prob)\n",
    "    # Subset the total population of predictions using the provided fraction.\n",
    "    num_predictions = len(y_true)\n",
    "    bootstrapping_fraction = fraction\n",
    "    num_to_retain = int(np.ceil(num_predictions*bootstrapping_fraction))\n",
    "    idx = np.random.choice(np.arange(num_predictions), num_to_retain, replace=False)\n",
    "    y_true_sample = y_true[idx]\n",
    "    y_prob_sample = y_prob[idx]\n",
    "    \n",
    "    # Calculate any desired metrics using just that subset.\n",
    "    n_pos, n_neg = Counter(y_true_sample)[1], Counter(y_true_sample)[0]\n",
    "    precision, recall, thresholds = precision_recall_curve(y_true_sample, y_prob_sample)\n",
    "    baseline = Counter(y_true_sample)[1]/len(y_true_sample) \n",
    "    area = auc(recall, precision)\n",
    "    auc_to_baseline_auc_ratio = area/baseline\n",
    "    \n",
    "    # Find the maximum Fß score for different values of ß.  \n",
    "    f_beta = lambda pr,re,beta: [((1+beta**2)*p*r)/((((beta**2)*p)+r)) for p,r in zip(pr,re)]\n",
    "    f_1_scores = f_beta(precision,recall,beta=1)\n",
    "    f_2_scores = f_beta(precision,recall,beta=2)\n",
    "    f_point5_scores = f_beta(precision,recall,beta=0.5)\n",
    "    \n",
    "    # Create a dictionary of those metric values to return.\n",
    "    scores={\"f_1_max\":np.nanmax(f_1_scores),\"f_2_max\":np.nanmax(f_2_scores),\"f_point5_max\":np.nanmax(f_point5_scores)}\n",
    "    return(scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with finding precision and recall values for each approach\n"
     ]
    }
   ],
   "source": [
    "for properties,idxs in zip(subset_properties, subset_idx_lists):\n",
    "    \n",
    "    # Remember the properties for this subset being looked at, and subset the dataframe accordingly.\n",
    "    c,q,s = properties\n",
    "    subset = df.loc[idxs]\n",
    "\n",
    "\n",
    "\n",
    "    # Check that this subsetting leaves a valid dataset with both positive and negatives samples.\n",
    "    class_values = pd.unique(subset[q].values)\n",
    "    if not (len(class_values)==2 and 0 in class_values and 1 in class_values):\n",
    "        continue\n",
    "\n",
    "\n",
    "\n",
    "    y_true_dict = {name:subset[q].values for name in names}       #just added .values here...\n",
    "    y_prob_dict = {name:(1 - subset[name].values) for name in names}\n",
    "    num_plots, plots_per_row, row_width, row_height = (len(names), 4, 14, 3)\n",
    "    fig,axs = plt.subplots(math.ceil(num_plots/plots_per_row), plots_per_row, squeeze=False)\n",
    "    for name,ax in zip(names, axs.flatten()):\n",
    "\n",
    "        # Obtaining the values and metrics.\n",
    "        y_true, y_prob = y_true_dict[name], y_prob_dict[name]\n",
    "        n_pos, n_neg = Counter(y_true)[1], Counter(y_true)[0]\n",
    "        precision, recall, thresholds = precision_recall_curve(y_true, y_prob)\n",
    "        baseline_auc = Counter(y_true)[1]/len(y_true) \n",
    "        area = auc(recall, precision)\n",
    "        auc_to_baseline_auc_ratio = area/baseline_auc\n",
    "        # The baseline F1 max has a precision of the ratio of positives to all samples and a recall of 1.\n",
    "        # This is because a random classifier achieves that precision at all recall values, so recall is maximized to\n",
    "        # find the maximum F1 value that can be expected due to random chance.\n",
    "        baseline_f1_max = (2*baseline_auc*1)/(baseline_auc+1)\n",
    "\n",
    "        #TABLE[name].update({\"auc\":area,\"ratio\":auc_to_baseline_auc_ratio, \"baseline\":baseline_f1_max, })\n",
    "        tables[c][q][s][name].update({\"auc\":area,\"ratio\":auc_to_baseline_auc_ratio, \"baseline\":baseline_f1_max, })\n",
    "\n",
    "\n",
    "        # Find the maximum Fß score for different values of ß.  \n",
    "        f_beta = lambda pr,re,beta: [((1+beta**2)*p*r)/((((beta**2)*p)+r)) for p,r in zip(pr,re)]\n",
    "        f_1_scores = f_beta(precision,recall,beta=1)\n",
    "        f_2_scores = f_beta(precision,recall,beta=2)\n",
    "        f_point5_scores = f_beta(precision,recall,beta=0.5)\n",
    "        f_1_max, f_1_std = np.nanmax(f_1_scores), np.std(f_1_scores)\n",
    "        f_2_max, f_2_std = np.nanmax(f_2_scores), np.std(f_2_scores)\n",
    "        f_point5_max, f_point5_std = np.nanmax(f_point5_scores), np.std(f_point5_scores)\n",
    "\n",
    "        # Find the standard deviation of each metric when subsampling the dataset of predictions for each method.\n",
    "        bootstrap_fraction = 0.5\n",
    "        bootstrap_iterations = 2\n",
    "        bootstrapped_std_dict = bootstrap(bootstrap_fraction, bootstrap_iterations, y_true, y_prob)\n",
    "\n",
    "        tables[c][q][s][name].update({\"f1_max\":f_1_max, \"f5_max\":f_point5_max, \"f2_max\":f_2_max})\n",
    "        #TABLE[name].update({\"f1_std\":f_1_std, \"f5_std\":f_point5_std, \"f2_std\":f_2_std})\n",
    "        tables[c][q][s][name].update({\"f1_std\":bootstrapped_std_dict[\"f_1_max_std\"], \n",
    "                            \"f5_std\":bootstrapped_std_dict[\"f_point5_max_std\"], \n",
    "                            \"f2_std\":bootstrapped_std_dict[\"f_2_max_std\"]}) \n",
    "\n",
    "        # Producing the precision recall curve.\n",
    "        #step_kwargs = ({'step': 'post'} if 'step' in signature(plt.fill_between).parameters else {})\n",
    "        #ax.step(recall, precision, color='black', alpha=0.2, where='post')\n",
    "        #ax.fill_between(recall, precision, alpha=0.7, color='black', **step_kwargs)\n",
    "        #ax.axhline(baseline_auc, linestyle=\"--\", color=\"lightgray\")\n",
    "        #ax.set_xlabel('Recall')\n",
    "        #ax.set_ylabel('Precision')\n",
    "        #ax.set_ylim([0.0, 1.05])\n",
    "        #ax.set_xlim([0.0, 1.0])\n",
    "        #ax.set_title(\"PR {0} (Baseline={1:0.3f})\".format(name, baseline_auc))\n",
    "\n",
    "    #fig.set_size_inches(row_width, row_height*math.ceil(num_plots/plots_per_row))\n",
    "    #fig.tight_layout()\n",
    "    #fig.savefig(os.path.join(OUTPUT_DIR,\"part_5_prcurve_shared.png\"),dpi=400)\n",
    "    plt.close()\n",
    "\n",
    "print(\"done with finding precision and recall values for each approach\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"y\"></a>\n",
    "### Are genes in the same group or pathway ranked higher with respect to individual nodes?\n",
    "This is a way of statistically seeing if for some value k, the graph ranks more edges from some particular gene to any other gene that it has a true protein-protein interaction with higher or equal to rank k, than we would expect due to random chance. This way of looking at the problem helps to be less ambiguous than the previous methods, because it gets at the core of how this would actually be used. In other words, we don't really care how much true information we're missing as long as we're still able to pick up some new useful information by building these networks, so even though we could be missing a lot, what's going on at the very top of the results? These results should be comparable to very strictly thresholding the network and saying that the remaining edges are our guesses at interactions. This is comparable to just looking at the far left-hand side of the precision recall curves, but just quantifies it slightly differently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done skipping the y thing\n"
     ]
    }
   ],
   "source": [
    "skip_this = True\n",
    "if skip_this:\n",
    "    \n",
    "    # When the edgelist is generated above, only the lower triangle of the pairwise matrix is retained for edges in the \n",
    "    # graph. This means that in terms of the indices of each node, only the (i,j) node is listed in the edge list where\n",
    "    # i is less than j. This makes sense because the graph that's specified is assumed to already be undirected. However\n",
    "    # in order to be able to easily subset the edgelist by a single column to obtain rows that correspond to all edges\n",
    "    # connected to a particular node, this method will double the number of rows to include both (i,j) and (j,i) edges.\n",
    "    df = make_undirected(df)\n",
    "\n",
    "    # What's the number of functional partners ranked k or higher in terms of phenotypic description similarity for \n",
    "    # each gene? Also figure out the maximum possible number of functional partners that could be theoretically\n",
    "    # recovered in this dataset if recovered means being ranked as k or higher here.\n",
    "    k = 10      # The threshold of interest for gene ranks.\n",
    "    n = 100     # Number of Monte Carlo simulation iterations to complete.\n",
    "    df[list(names)] = df.groupby(\"from\")[list(names)].rank()\n",
    "    ys = df[df[\"shared\"]==1][list(names)].apply(lambda s: len([x for x in s if x<=k]))\n",
    "    ymax = sum(df.groupby(\"from\")[\"shared\"].apply(lambda s: min(len([x for x in s if x==1]),k)))\n",
    "\n",
    "    # Monte Carlo simulation to see what the probability is of achieving each y-value by just randomly pulling k \n",
    "    # edges for each gene rather than taking the top k ones that the similarity methods specifies when ranking.\n",
    "    ysims = [sum(df.groupby(\"from\")[\"shared\"].apply(lambda s: len([x for x in s.sample(k) if x>0.00]))) for i in range(n)]\n",
    "    for name in names:\n",
    "        pvalue = len([ysim for ysim in ysims if ysim>=ys[name]])/float(n)\n",
    "        TABLE[name].update({\"y\":ys[name], \"y_max\":ymax, \"y_ratio\":ys[name]/ymax, \"y_pval\":pvalue})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"mean\"></a>\n",
    "### Predicting biochemical pathway or group membership based on mean vectors\n",
    "This section looks at how well the biochemical pathways that a particular gene is a member of can be predicted based on the similarity between the vector representation of the phenotype descriptions for that gene and the average vector for all the vector representations of phenotypes asociated with genes that belong to that particular pathway. In calculating the average vector for a given biochemical pathway, the vector corresponding to the gene that is currently being classified is not accounted for, to avoid overestimating the performance by including information about the ground truth during classification. This leads to missing information in the case of biochemical pathways that have only one member. This can be accounted for by only limiting the overall dataset to only include genes that belong to pathways that have atleast two genes mapped to them, and only including those pathways, or by removing the missing values before calculating the performance metrics below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get the list of methods to look at, and a mapping between each method and the correct similarity metric to apply.\n",
    "# vector_dicts = {k:v.vector_dictionary for k,v in graphs.items()}\n",
    "# names = list(vector_dicts.keys())\n",
    "# group_id_to_ids = groups.get_group_id_to_ids_dict(dataset.get_gene_dictionary())\n",
    "# valid_group_ids = [group for group,id_list in group_id_to_ids.items() if len(id_list)>1]\n",
    "# valid_ids = [i for i in dataset.get_ids() if len(set(valid_group_ids).intersection(set(id_to_group_ids[i])))>0]\n",
    "# pred_dict = defaultdict(lambda: defaultdict(dict))\n",
    "# true_dict = defaultdict(lambda: defaultdict(dict))\n",
    "# for name in names:\n",
    "#     for group in valid_group_ids:\n",
    "#         ids = group_id_to_ids[group]\n",
    "#         for identifier in valid_ids:\n",
    "#             # What's the mean vector of this group, without this particular one that we're trying to classify.\n",
    "#             vectors = np.array([vector_dicts[name][some_id] for some_id in ids if not some_id==identifier])\n",
    "#             mean_vector = vectors.mean(axis=0)\n",
    "#             this_vector = vector_dicts[name][identifier]\n",
    "#             pred_dict[name][identifier][group] = 1-metric_dict[name](mean_vector, this_vector)\n",
    "#             true_dict[name][identifier][group] = (identifier in group_id_to_ids[group])*1                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# num_plots, plots_per_row, row_width, row_height = (len(names), 4, 14, 3)\n",
    "# fig,axs = plt.subplots(math.ceil(num_plots/plots_per_row), plots_per_row, squeeze=False)\n",
    "# for name,ax in zip(names, axs.flatten()):\n",
    "#     \n",
    "#     # Obtaining the values and metrics.\n",
    "#     y_true = pd.DataFrame(true_dict[name]).as_matrix().flatten()\n",
    "#     y_prob = pd.DataFrame(pred_dict[name]).as_matrix().flatten()\n",
    "#     n_pos, n_neg = Counter(y_true)[1], Counter(y_true)[0]\n",
    "#     precision, recall, thresholds = precision_recall_curve(y_true, y_prob)\n",
    "#     baseline = Counter(y_true)[1]/len(y_true) \n",
    "#     area = auc(recall, precision)\n",
    "#     auc_to_baseline_auc_ratio = area/baseline\n",
    "#     TABLE[name].update({\"mean_auc\":area, \"mean_baseline\":baseline, \"mean_ratio\":auc_to_baseline_auc_ratio})\n",
    "# \n",
    "#     # Producing the precision recall curve.\n",
    "#     step_kwargs = ({'step': 'post'} if 'step' in signature(plt.fill_between).parameters else {})\n",
    "#     ax.step(recall, precision, color='black', alpha=0.2, where='post')\n",
    "#     ax.fill_between(recall, precision, alpha=0.7, color='black', **step_kwargs)\n",
    "#     ax.axhline(baseline, linestyle=\"--\", color=\"lightgray\")\n",
    "#     ax.set_xlabel('Recall')\n",
    "#     ax.set_ylabel('Precision')\n",
    "#     ax.set_ylim([0.0, 1.05])\n",
    "#     ax.set_xlim([0.0, 1.0])\n",
    "#     ax.set_title(\"PR {0} (Baseline={1:0.3f})\".format(name[:10], baseline))\n",
    "#     \n",
    "# fig.set_size_inches(row_width, row_height*math.ceil(num_plots/plots_per_row))\n",
    "# fig.tight_layout()\n",
    "# fig.savefig(os.path.join(OUTPUT_DIR,\"part_6_prcurve_mean_classifier.png\"),dpi=400)\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicting biochemical pathway membership based on mean similarity values\n",
    "This section looks at how well the biochemical pathways that a particular gene is a member of can be predicted based on the average similarity between the vector representationt of the phenotype descriptions for that gene and each of the vector representations for other phenotypes associated with genes that belong to that particular pathway. In calculating the average similarity to other genes from a given biochemical pathway, the gene that is currently being classified is not accounted for, to avoid overestimating the performance by including information about the ground truth during classification. This leads to missing information in the case of biochemical pathways that have only one member. This can be accounted for by only limiting the overall dataset to only include genes that belong to pathways that have atleast two genes mapped to them, and only including those pathways, or by removing the missing values before calculating the performance metrics below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicting biochemical pathway or group membership with KNN classifier\n",
    "This section looks at how well the group(s) or biochemical pathway(s) that a particular gene belongs to can be predicted based on a KNN classifier generated using every other gene. For this section, only the groups or pathways which contain more than one gene, and the genes mapped to those groups or pathways, are of interest. This is because for other genes, if we consider them then it will be true that that gene belongs to that group in the target vector, but the KNN classifier could never predict this because when that gene is held out, nothing could provide a vote for that group, because there are zero genes available to be members of the K nearest neighbors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"output\"></a>\n",
    "### Summarizing the results for this notebook\n",
    "Write a large table of results to an output file. Columns are generally metrics and rows are generally methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Method</th>\n",
       "      <th>Species</th>\n",
       "      <th>Objective</th>\n",
       "      <th>Curated</th>\n",
       "      <th>Hyperparameters</th>\n",
       "      <th>Group</th>\n",
       "      <th>Order</th>\n",
       "      <th>mean_1</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>n_1</th>\n",
       "      <th>n_0</th>\n",
       "      <th>ks</th>\n",
       "      <th>ks_pval</th>\n",
       "      <th>auc</th>\n",
       "      <th>ratio</th>\n",
       "      <th>baseline</th>\n",
       "      <th>f1_max</th>\n",
       "      <th>f5_max</th>\n",
       "      <th>f2_max</th>\n",
       "      <th>f1_std</th>\n",
       "      <th>f5_std</th>\n",
       "      <th>f2_std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Doc2Vec</td>\n",
       "      <td>intra</td>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>Wikipedia,Size=300</td>\n",
       "      <td>NLP</td>\n",
       "      <td>0</td>\n",
       "      <td>0.518946</td>\n",
       "      <td>0.444828</td>\n",
       "      <td>10867.0</td>\n",
       "      <td>21890.0</td>\n",
       "      <td>0.404437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.262805</td>\n",
       "      <td>0.792186</td>\n",
       "      <td>0.498212</td>\n",
       "      <td>0.498337</td>\n",
       "      <td>0.383051</td>\n",
       "      <td>0.712899</td>\n",
       "      <td>0.001747</td>\n",
       "      <td>0.001649</td>\n",
       "      <td>0.001433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Word2Vec</td>\n",
       "      <td>intra</td>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>Wikipedia,Size=300,Mean</td>\n",
       "      <td>NLP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.232176</td>\n",
       "      <td>0.349355</td>\n",
       "      <td>10867.0</td>\n",
       "      <td>21890.0</td>\n",
       "      <td>0.343093</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.524663</td>\n",
       "      <td>1.581522</td>\n",
       "      <td>0.498212</td>\n",
       "      <td>0.552898</td>\n",
       "      <td>0.609972</td>\n",
       "      <td>0.748872</td>\n",
       "      <td>0.001945</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0.001098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Word2Vec</td>\n",
       "      <td>intra</td>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>Wikipedia,Size=300,Max</td>\n",
       "      <td>NLP</td>\n",
       "      <td>2</td>\n",
       "      <td>0.255199</td>\n",
       "      <td>0.292205</td>\n",
       "      <td>10867.0</td>\n",
       "      <td>21890.0</td>\n",
       "      <td>0.297242</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.507730</td>\n",
       "      <td>1.530478</td>\n",
       "      <td>0.498212</td>\n",
       "      <td>0.536380</td>\n",
       "      <td>0.577138</td>\n",
       "      <td>0.715067</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.005069</td>\n",
       "      <td>0.002084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Topic Models</td>\n",
       "      <td>intra</td>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>NMF,Full,Topics=20</td>\n",
       "      <td>NLP</td>\n",
       "      <td>3</td>\n",
       "      <td>0.666534</td>\n",
       "      <td>0.853653</td>\n",
       "      <td>10867.0</td>\n",
       "      <td>21890.0</td>\n",
       "      <td>0.429723</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.546983</td>\n",
       "      <td>1.648801</td>\n",
       "      <td>0.498212</td>\n",
       "      <td>0.632343</td>\n",
       "      <td>0.591594</td>\n",
       "      <td>0.789329</td>\n",
       "      <td>0.003411</td>\n",
       "      <td>0.002796</td>\n",
       "      <td>0.001420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Topic Models</td>\n",
       "      <td>intra</td>\n",
       "      <td>subsets</td>\n",
       "      <td>true</td>\n",
       "      <td>NMF,Full,Topics=50</td>\n",
       "      <td>NLP</td>\n",
       "      <td>4</td>\n",
       "      <td>0.733366</td>\n",
       "      <td>0.905444</td>\n",
       "      <td>10867.0</td>\n",
       "      <td>21890.0</td>\n",
       "      <td>0.598089</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.600420</td>\n",
       "      <td>1.809879</td>\n",
       "      <td>0.498212</td>\n",
       "      <td>0.718605</td>\n",
       "      <td>0.646270</td>\n",
       "      <td>0.832566</td>\n",
       "      <td>0.001392</td>\n",
       "      <td>0.002379</td>\n",
       "      <td>0.000511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>Doc2Vec Phenes</td>\n",
       "      <td>both</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>Wikipedia,Size=300</td>\n",
       "      <td>NLP</td>\n",
       "      <td>21</td>\n",
       "      <td>0.287475</td>\n",
       "      <td>0.251211</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4474.0</td>\n",
       "      <td>0.535092</td>\n",
       "      <td>0.930056</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>1.075204</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000961</td>\n",
       "      <td>0.000601</td>\n",
       "      <td>0.002398</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>Word2Vec Phenes</td>\n",
       "      <td>both</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>Wikipedia,Size=300,Mean</td>\n",
       "      <td>NLP</td>\n",
       "      <td>22</td>\n",
       "      <td>0.377072</td>\n",
       "      <td>0.260887</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4474.0</td>\n",
       "      <td>0.715467</td>\n",
       "      <td>0.569385</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.698564</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>0.000390</td>\n",
       "      <td>0.001559</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>Word2Vec Phenes</td>\n",
       "      <td>both</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>Wikipedia,Size=300,Max</td>\n",
       "      <td>NLP</td>\n",
       "      <td>23</td>\n",
       "      <td>0.231882</td>\n",
       "      <td>0.188057</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4474.0</td>\n",
       "      <td>0.744077</td>\n",
       "      <td>0.512179</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>0.671922</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000600</td>\n",
       "      <td>0.000375</td>\n",
       "      <td>0.001500</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>N-Grams Phenes</td>\n",
       "      <td>both</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>Linares_Pontes,Words,1-grams</td>\n",
       "      <td>NLP</td>\n",
       "      <td>24</td>\n",
       "      <td>0.764298</td>\n",
       "      <td>0.559057</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4474.0</td>\n",
       "      <td>0.661153</td>\n",
       "      <td>0.677989</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.753113</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.001681</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>Mean</td>\n",
       "      <td>both</td>\n",
       "      <td>orthologs</td>\n",
       "      <td>false</td>\n",
       "      <td>None</td>\n",
       "      <td>NLP</td>\n",
       "      <td>25</td>\n",
       "      <td>0.558944</td>\n",
       "      <td>0.400241</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4474.0</td>\n",
       "      <td>0.724184</td>\n",
       "      <td>0.551955</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.690373</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>0.001541</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>520 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Method Species  Objective Curated               Hyperparameters Group  Order    mean_1    mean_0      n_1      n_0        ks   ks_pval       auc     ratio  baseline    f1_max    f5_max    f2_max    f1_std    f5_std    f2_std\n",
       "0            Doc2Vec   intra    subsets    true            Wikipedia,Size=300   NLP      0  0.518946  0.444828  10867.0  21890.0  0.404437  0.000000  0.262805  0.792186  0.498212  0.498337  0.383051  0.712899  0.001747  0.001649  0.001433\n",
       "1           Word2Vec   intra    subsets    true       Wikipedia,Size=300,Mean   NLP      1  0.232176  0.349355  10867.0  21890.0  0.343093  0.000000  0.524663  1.581522  0.498212  0.552898  0.609972  0.748872  0.001945  0.001896  0.001098\n",
       "2           Word2Vec   intra    subsets    true        Wikipedia,Size=300,Max   NLP      2  0.255199  0.292205  10867.0  21890.0  0.297242  0.000000  0.507730  1.530478  0.498212  0.536380  0.577138  0.715067  0.004421  0.005069  0.002084\n",
       "3       Topic Models   intra    subsets    true            NMF,Full,Topics=20   NLP      3  0.666534  0.853653  10867.0  21890.0  0.429723  0.000000  0.546983  1.648801  0.498212  0.632343  0.591594  0.789329  0.003411  0.002796  0.001420\n",
       "4       Topic Models   intra    subsets    true            NMF,Full,Topics=50   NLP      4  0.733366  0.905444  10867.0  21890.0  0.598089  0.000000  0.600420  1.809879  0.498212  0.718605  0.646270  0.832566  0.001392  0.002379  0.000511\n",
       "..               ...     ...        ...     ...                           ...   ...    ...       ...       ...      ...      ...       ...       ...       ...       ...       ...       ...       ...       ...       ...       ...       ...\n",
       "515   Doc2Vec Phenes    both  orthologs   false            Wikipedia,Size=300   NLP     21  0.287475  0.251211      1.0   4474.0  0.535092  0.930056  0.000240  1.075204  0.000447  0.000961  0.000601  0.002398  0.000007  0.000004  0.000016\n",
       "516  Word2Vec Phenes    both  orthologs   false       Wikipedia,Size=300,Mean   NLP     22  0.377072  0.260887      1.0   4474.0  0.715467  0.569385  0.000156  0.698564  0.000447  0.000624  0.000390  0.001559  0.000006  0.000004  0.000015\n",
       "517  Word2Vec Phenes    both  orthologs   false        Wikipedia,Size=300,Max   NLP     23  0.231882  0.188057      1.0   4474.0  0.744077  0.512179  0.000150  0.671922  0.000447  0.000600  0.000375  0.001500  0.000003  0.000002  0.000006\n",
       "518   N-Grams Phenes    both  orthologs   false  Linares_Pontes,Words,1-grams   NLP     24  0.764298  0.559057      1.0   4474.0  0.661153  0.677989  0.000168  0.753113  0.000447  0.000673  0.000421  0.001681  0.000006  0.000003  0.000014\n",
       "519             Mean    both  orthologs   false                          None   NLP     25  0.558944  0.400241      1.0   4474.0  0.724184  0.551955  0.000154  0.690373  0.000447  0.000617  0.000386  0.001541  0.000000  0.000000  0.000000\n",
       "\n",
       "[520 rows x 22 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_dfs = []\n",
    "for s,c,q in itertools.product(species,curated,question):\n",
    "    TABLE = tables[c][q][s]\n",
    "    results = pd.DataFrame(TABLE).transpose()\n",
    "    columns = flatten([\"Species\", \"Objective\",\"Curated\",\"Hyperparameters\",\"Group\",\"Order\",results.columns])\n",
    "    results[\"Hyperparameters\"] = \"\"\n",
    "    results[\"Group\"] = \"NLP\"\n",
    "    results[\"Order\"] = np.arange(results.shape[0])\n",
    "    results[\"Species\"] = s.lower()\n",
    "    results[\"Objective\"] = q.lower()\n",
    "    results[\"Curated\"] = str(c).lower()\n",
    "    results = results[columns]\n",
    "    results.reset_index(inplace=True)\n",
    "    results = results.rename({\"index\":\"Method\"}, axis=\"columns\")\n",
    "    hyperparam_sep = \":\"\n",
    "    results[\"Hyperparameters\"] = results[\"Method\"].map(lambda x: x.split(hyperparam_sep)[1] if hyperparam_sep in x else \"None\")\n",
    "    results[\"Method\"] = results[\"Method\"].map(lambda x: x.split(hyperparam_sep)[0])\n",
    "    result_dfs.append(results)\n",
    "\n",
    "results = pd.concat(result_dfs)\n",
    "results.reset_index(inplace=True, drop=True)\n",
    "results.to_csv(os.path.join(OUTPUT_DIR,\"part_5_full_table.csv\"), index=False)\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_6\"></a>\n",
    "# Part 6. Clustering Analysis\n",
    "The purpose of this section is to look at different ways that the embeddings obtained for the dataset of phenotype descriptions can be used to cluster or organize the genes to which those phenotypes are mapped into subgroups or representations. These approaches include generating topic models from the data, and doing agglomerative clustering to find clusters to which each gene belongs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Objects initially created in previous sections that are used by this section.\n",
    "groups = phe_subsets_groups\n",
    "id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(dataset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"topic_modeling\"></a>\n",
    "### Approach 1: Topic modeling based on n-grams with a reduced vocabulary\n",
    "Topic modelling learns a set of word probability distributions from the dataset of text descriptions, which represent distinct topics which are present in the dataset. Each text description can then be represented as a discrete probability distribution over the learned topics based on the probability that a given piece of text belongs to each particular topics. This is a form of data reduction because a high dimensionsal bag-of-words can be represented as a vector of *k* probabilities where *k* is the number of topics. The main advantages of topic modelling over clustering is that topic modelling provides soft classifications that can be additionally interpreted, rather than hard classifications into a single cluster. Topic models are also explainable, because the word probability distributions for that topic can be used to determine which words are most representative of any given topic. One problem with topic modelling is that is uses the n-grams embeddings to semantic similarity between different words is not accounted for. To help alleviate this, this section uses implementations of some existing algorithms to compress the vocabulary as a preprocessing step based on word distance matrices generated using word embeddings.\n",
    "\n",
    "Topic models define topics present in a dataset of texts as word or n-gram probability distributions. These models represent each instance of text then as being composed of or generated as as mixture of these topics. The vector for each text that indicates which fraction of that text is generated by a each topic is of length *n* where *n* is the number of topics, and can be used as a reduced dimensionality of the text, with a much smaller vector length than the n-grams embedding itself. Therefore we can build a topic model of the data with 100 topics for example in order to then represent each description in the dataset as a a vector of length 100. This section constructs topic models from the n-gram representations of the dataset and selects different values for the number of topics in order to find a value that works well during the grid search over the training dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>topic</th>\n",
       "      <th>GAM</th>\n",
       "      <th>GEM</th>\n",
       "      <th>EMG</th>\n",
       "      <th>MGD</th>\n",
       "      <th>EMB</th>\n",
       "      <th>MSD</th>\n",
       "      <th>NHM</th>\n",
       "      <th>SRL</th>\n",
       "      <th>GER</th>\n",
       "      <th>NLS</th>\n",
       "      <th>PIG</th>\n",
       "      <th>GRS</th>\n",
       "      <th>ROT</th>\n",
       "      <th>LEF</th>\n",
       "      <th>IST</th>\n",
       "      <th>ARC</th>\n",
       "      <th>MSL</th>\n",
       "      <th>FSM</th>\n",
       "      <th>OVP</th>\n",
       "      <th>SRF</th>\n",
       "      <th>SSC</th>\n",
       "      <th>FLT</th>\n",
       "      <th>SEN</th>\n",
       "      <th>MTM</th>\n",
       "      <th>STT</th>\n",
       "      <th>RTH</th>\n",
       "      <th>TCM</th>\n",
       "      <th>CUL</th>\n",
       "      <th>PRA</th>\n",
       "      <th>CPR</th>\n",
       "      <th>WAT</th>\n",
       "      <th>TMP</th>\n",
       "      <th>LIT</th>\n",
       "      <th>MEC</th>\n",
       "      <th>MPH</th>\n",
       "      <th>NUT</th>\n",
       "      <th>HRM</th>\n",
       "      <th>CHS</th>\n",
       "      <th>MCH</th>\n",
       "      <th>PTH</th>\n",
       "      <th>OBI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.076284</td>\n",
       "      <td>0.003959</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016848</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.283903e-03</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000581</td>\n",
       "      <td>0.002419</td>\n",
       "      <td>0.000556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>0.005617</td>\n",
       "      <td>0.002117</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003597</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>0.002384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>0.010256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.082733</td>\n",
       "      <td>0.009446</td>\n",
       "      <td>0.001537</td>\n",
       "      <td>0.017872</td>\n",
       "      <td>0.000339</td>\n",
       "      <td>0.009721</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.593859e-03</td>\n",
       "      <td>0.003191</td>\n",
       "      <td>0.005394</td>\n",
       "      <td>0.002710</td>\n",
       "      <td>0.003352</td>\n",
       "      <td>0.001981</td>\n",
       "      <td>0.001919</td>\n",
       "      <td>0.011893</td>\n",
       "      <td>0.002803</td>\n",
       "      <td>0.005643</td>\n",
       "      <td>0.014627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033349</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.001743</td>\n",
       "      <td>0.002580</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007809</td>\n",
       "      <td>0.002025</td>\n",
       "      <td>0.004153</td>\n",
       "      <td>0.008524</td>\n",
       "      <td>0.001489</td>\n",
       "      <td>4.526762e-03</td>\n",
       "      <td>0.001240</td>\n",
       "      <td>0.011141</td>\n",
       "      <td>0.000929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013736</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>0.004185</td>\n",
       "      <td>0.007506</td>\n",
       "      <td>0.001041</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.027227</td>\n",
       "      <td>0.011609</td>\n",
       "      <td>0.095080</td>\n",
       "      <td>0.001010</td>\n",
       "      <td>0.013777</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.126438e-02</td>\n",
       "      <td>0.003949</td>\n",
       "      <td>0.003779</td>\n",
       "      <td>0.009216</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>0.004516</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.000801</td>\n",
       "      <td>0.002013</td>\n",
       "      <td>0.001390</td>\n",
       "      <td>0.001884</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001144</td>\n",
       "      <td>0.000866</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001096</td>\n",
       "      <td>0.003110</td>\n",
       "      <td>0.000986</td>\n",
       "      <td>0.003070</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>4.224044e-04</td>\n",
       "      <td>0.001010</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001212</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.002581</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.001977</td>\n",
       "      <td>0.000433</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008291</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>6.327772e-04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000435</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.014415</td>\n",
       "      <td>0.020483</td>\n",
       "      <td>0.050040</td>\n",
       "      <td>0.022687</td>\n",
       "      <td>0.005832</td>\n",
       "      <td>0.053081</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.551137e-02</td>\n",
       "      <td>0.036639</td>\n",
       "      <td>0.019778</td>\n",
       "      <td>0.030213</td>\n",
       "      <td>0.023450</td>\n",
       "      <td>0.020206</td>\n",
       "      <td>0.016841</td>\n",
       "      <td>0.021650</td>\n",
       "      <td>0.020645</td>\n",
       "      <td>0.037356</td>\n",
       "      <td>0.025890</td>\n",
       "      <td>0.009345</td>\n",
       "      <td>0.015055</td>\n",
       "      <td>0.022114</td>\n",
       "      <td>0.032296</td>\n",
       "      <td>0.020620</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019487</td>\n",
       "      <td>0.028734</td>\n",
       "      <td>0.025016</td>\n",
       "      <td>0.026165</td>\n",
       "      <td>0.040020</td>\n",
       "      <td>1.555394e-02</td>\n",
       "      <td>0.016904</td>\n",
       "      <td>0.038592</td>\n",
       "      <td>0.026897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020284</td>\n",
       "      <td>0.038561</td>\n",
       "      <td>0.015501</td>\n",
       "      <td>0.008961</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022707</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>0.005669</td>\n",
       "      <td>0.006224</td>\n",
       "      <td>0.015926</td>\n",
       "      <td>0.004878</td>\n",
       "      <td>0.009833</td>\n",
       "      <td>0.037933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.633048e-03</td>\n",
       "      <td>0.003357</td>\n",
       "      <td>0.005175</td>\n",
       "      <td>0.002612</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.001471</td>\n",
       "      <td>0.001035</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.002164</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001169</td>\n",
       "      <td>0.004181</td>\n",
       "      <td>0.001537</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004878</td>\n",
       "      <td>0.002116</td>\n",
       "      <td>0.006437</td>\n",
       "      <td>0.002661</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.004878</td>\n",
       "      <td>0.003224</td>\n",
       "      <td>0.001837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005854</td>\n",
       "      <td>0.002118</td>\n",
       "      <td>0.001729</td>\n",
       "      <td>0.001239</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002191</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>40</td>\n",
       "      <td>0.000330</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.008155</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.001066</td>\n",
       "      <td>0.024790</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.178688e-03</td>\n",
       "      <td>0.001284</td>\n",
       "      <td>0.009390</td>\n",
       "      <td>0.002173</td>\n",
       "      <td>0.003544</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.002311</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002578</td>\n",
       "      <td>0.001580</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.002033</td>\n",
       "      <td>0.003872</td>\n",
       "      <td>0.013274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002595</td>\n",
       "      <td>0.002372</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.010055</td>\n",
       "      <td>0.009616</td>\n",
       "      <td>2.023576e-03</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.002958</td>\n",
       "      <td>0.006279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012581</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.004230</td>\n",
       "      <td>0.000668</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>39</td>\n",
       "      <td>0.000926</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.001930</td>\n",
       "      <td>0.001523</td>\n",
       "      <td>0.002631</td>\n",
       "      <td>0.006285</td>\n",
       "      <td>0.035787</td>\n",
       "      <td>1.602167e-02</td>\n",
       "      <td>0.001013</td>\n",
       "      <td>0.001207</td>\n",
       "      <td>0.017470</td>\n",
       "      <td>0.004987</td>\n",
       "      <td>0.000556</td>\n",
       "      <td>0.002134</td>\n",
       "      <td>0.001168</td>\n",
       "      <td>0.000494</td>\n",
       "      <td>0.005958</td>\n",
       "      <td>0.000651</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000627</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004473</td>\n",
       "      <td>0.007775</td>\n",
       "      <td>0.001283</td>\n",
       "      <td>0.003527</td>\n",
       "      <td>0.001498</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002133</td>\n",
       "      <td>0.001712</td>\n",
       "      <td>0.007679</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001823</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>0.000752</td>\n",
       "      <td>0.000578</td>\n",
       "      <td>0.000882</td>\n",
       "      <td>0.000651</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>0.006671</td>\n",
       "      <td>0.012729</td>\n",
       "      <td>0.021005</td>\n",
       "      <td>0.025876</td>\n",
       "      <td>0.009524</td>\n",
       "      <td>0.000409</td>\n",
       "      <td>0.157451</td>\n",
       "      <td>1.304053e-02</td>\n",
       "      <td>0.028726</td>\n",
       "      <td>0.021890</td>\n",
       "      <td>0.017524</td>\n",
       "      <td>0.020969</td>\n",
       "      <td>0.008290</td>\n",
       "      <td>0.009959</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.017603</td>\n",
       "      <td>0.042049</td>\n",
       "      <td>0.016916</td>\n",
       "      <td>0.005305</td>\n",
       "      <td>0.009430</td>\n",
       "      <td>0.018462</td>\n",
       "      <td>0.013113</td>\n",
       "      <td>0.008458</td>\n",
       "      <td>0.036994</td>\n",
       "      <td>0.011636</td>\n",
       "      <td>0.024320</td>\n",
       "      <td>0.007580</td>\n",
       "      <td>0.022725</td>\n",
       "      <td>0.008893</td>\n",
       "      <td>2.897721e-03</td>\n",
       "      <td>0.041524</td>\n",
       "      <td>0.031675</td>\n",
       "      <td>0.019677</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012537</td>\n",
       "      <td>0.033642</td>\n",
       "      <td>0.010014</td>\n",
       "      <td>0.016361</td>\n",
       "      <td>0.004657</td>\n",
       "      <td>0.020980</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>0.012837</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0.012889</td>\n",
       "      <td>0.004663</td>\n",
       "      <td>0.016436</td>\n",
       "      <td>0.018711</td>\n",
       "      <td>0.071562</td>\n",
       "      <td>1.001991e-01</td>\n",
       "      <td>0.003741</td>\n",
       "      <td>0.005929</td>\n",
       "      <td>0.034533</td>\n",
       "      <td>0.028327</td>\n",
       "      <td>0.005865</td>\n",
       "      <td>0.004795</td>\n",
       "      <td>0.000609</td>\n",
       "      <td>0.005205</td>\n",
       "      <td>0.031796</td>\n",
       "      <td>0.002705</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001928</td>\n",
       "      <td>0.003997</td>\n",
       "      <td>0.001540</td>\n",
       "      <td>0.038092</td>\n",
       "      <td>0.044936</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.004663</td>\n",
       "      <td>0.005295</td>\n",
       "      <td>0.007216</td>\n",
       "      <td>0.004011</td>\n",
       "      <td>4.014630e-04</td>\n",
       "      <td>0.004663</td>\n",
       "      <td>0.004046</td>\n",
       "      <td>0.012700</td>\n",
       "      <td>0.076252</td>\n",
       "      <td>0.006515</td>\n",
       "      <td>0.003170</td>\n",
       "      <td>0.019642</td>\n",
       "      <td>0.002641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005840</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>32</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.354751e-03</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002345</td>\n",
       "      <td>0.001541</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000457</td>\n",
       "      <td>0.001004</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001622</td>\n",
       "      <td>0.001450</td>\n",
       "      <td>0.001006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001449</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000580</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.299274e-04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "      <td>0.001856</td>\n",
       "      <td>0.003125</td>\n",
       "      <td>0.001477</td>\n",
       "      <td>0.002604</td>\n",
       "      <td>0.001363</td>\n",
       "      <td>0.004369</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.644227e-03</td>\n",
       "      <td>0.024748</td>\n",
       "      <td>0.004027</td>\n",
       "      <td>0.001440</td>\n",
       "      <td>0.001425</td>\n",
       "      <td>0.004336</td>\n",
       "      <td>0.002683</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.003090</td>\n",
       "      <td>0.001612</td>\n",
       "      <td>0.003374</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003964</td>\n",
       "      <td>0.003751</td>\n",
       "      <td>0.006314</td>\n",
       "      <td>0.008872</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.002604</td>\n",
       "      <td>0.002164</td>\n",
       "      <td>0.005246</td>\n",
       "      <td>0.002281</td>\n",
       "      <td>6.725981e-03</td>\n",
       "      <td>0.010451</td>\n",
       "      <td>0.010825</td>\n",
       "      <td>0.000976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003125</td>\n",
       "      <td>0.003273</td>\n",
       "      <td>0.011014</td>\n",
       "      <td>0.023511</td>\n",
       "      <td>0.006808</td>\n",
       "      <td>0.002148</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.515093e-04</td>\n",
       "      <td>0.000703</td>\n",
       "      <td>0.003583</td>\n",
       "      <td>0.001458</td>\n",
       "      <td>0.001180</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.002292</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.000881</td>\n",
       "      <td>0.000325</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002600</td>\n",
       "      <td>0.001079</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001265</td>\n",
       "      <td>0.000813</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000595</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>2.323182e-03</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001184</td>\n",
       "      <td>0.000178</td>\n",
       "      <td>0.000846</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>41</td>\n",
       "      <td>0.001245</td>\n",
       "      <td>0.002501</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001280</td>\n",
       "      <td>0.004953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.134306e-03</td>\n",
       "      <td>0.000925</td>\n",
       "      <td>0.146578</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0.022830</td>\n",
       "      <td>0.035425</td>\n",
       "      <td>0.009414</td>\n",
       "      <td>0.079916</td>\n",
       "      <td>0.034362</td>\n",
       "      <td>0.000431</td>\n",
       "      <td>0.006022</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.008752</td>\n",
       "      <td>0.001097</td>\n",
       "      <td>0.030356</td>\n",
       "      <td>0.009425</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020543</td>\n",
       "      <td>0.009382</td>\n",
       "      <td>0.025922</td>\n",
       "      <td>0.023424</td>\n",
       "      <td>0.004478</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001280</td>\n",
       "      <td>0.000768</td>\n",
       "      <td>0.104747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001535</td>\n",
       "      <td>0.017089</td>\n",
       "      <td>0.034239</td>\n",
       "      <td>0.010471</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006945</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>34</td>\n",
       "      <td>0.009407</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000619</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.785738e-04</td>\n",
       "      <td>0.001025</td>\n",
       "      <td>0.021810</td>\n",
       "      <td>0.014261</td>\n",
       "      <td>0.032747</td>\n",
       "      <td>0.016591</td>\n",
       "      <td>0.008171</td>\n",
       "      <td>0.013460</td>\n",
       "      <td>0.031059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013545</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014248</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017964</td>\n",
       "      <td>0.023927</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013393</td>\n",
       "      <td>0.002675</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003574</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>0.010891</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>0.003801</td>\n",
       "      <td>0.007603</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.006336</td>\n",
       "      <td>0.002225</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.754604e-03</td>\n",
       "      <td>0.004224</td>\n",
       "      <td>0.014215</td>\n",
       "      <td>0.009182</td>\n",
       "      <td>0.014836</td>\n",
       "      <td>0.019847</td>\n",
       "      <td>0.009649</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>0.042169</td>\n",
       "      <td>0.020935</td>\n",
       "      <td>0.024689</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014783</td>\n",
       "      <td>0.011486</td>\n",
       "      <td>0.024659</td>\n",
       "      <td>0.027236</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004388</td>\n",
       "      <td>0.041348</td>\n",
       "      <td>0.003501</td>\n",
       "      <td>0.010415</td>\n",
       "      <td>0.003456</td>\n",
       "      <td>4.047169e-08</td>\n",
       "      <td>0.006336</td>\n",
       "      <td>0.003801</td>\n",
       "      <td>0.007698</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007603</td>\n",
       "      <td>0.002715</td>\n",
       "      <td>0.014731</td>\n",
       "      <td>0.001583</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003009</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000285</td>\n",
       "      <td>0.002642</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001187</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000354</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.004753</td>\n",
       "      <td>0.000539</td>\n",
       "      <td>0.013182</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>0.011511</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005075</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007924</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.142219e-05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011961</td>\n",
       "      <td>0.002652</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.005532</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>0.016690</td>\n",
       "      <td>0.018714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000815</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.002547</td>\n",
       "      <td>0.002099</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004848</td>\n",
       "      <td>0.003228</td>\n",
       "      <td>0.009803</td>\n",
       "      <td>0.003669</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.111841e-05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001646</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.766100e-04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001853</td>\n",
       "      <td>0.000803</td>\n",
       "      <td>0.005153</td>\n",
       "      <td>0.001889</td>\n",
       "      <td>0.004320</td>\n",
       "      <td>0.004399</td>\n",
       "      <td>0.005958</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.009022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001521</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002898</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005527</td>\n",
       "      <td>0.003264</td>\n",
       "      <td>0.001602</td>\n",
       "      <td>9.231195e-05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001990</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000878</td>\n",
       "      <td>0.004160</td>\n",
       "      <td>0.000395</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>27</td>\n",
       "      <td>0.000987</td>\n",
       "      <td>0.001975</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.001646</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.459350e-03</td>\n",
       "      <td>0.004134</td>\n",
       "      <td>0.004491</td>\n",
       "      <td>0.002061</td>\n",
       "      <td>0.001200</td>\n",
       "      <td>0.011568</td>\n",
       "      <td>0.009638</td>\n",
       "      <td>0.002572</td>\n",
       "      <td>0.015486</td>\n",
       "      <td>0.004530</td>\n",
       "      <td>0.013080</td>\n",
       "      <td>0.016956</td>\n",
       "      <td>0.009992</td>\n",
       "      <td>0.002745</td>\n",
       "      <td>0.012754</td>\n",
       "      <td>0.002542</td>\n",
       "      <td>0.011361</td>\n",
       "      <td>0.001734</td>\n",
       "      <td>0.003202</td>\n",
       "      <td>0.012036</td>\n",
       "      <td>0.003393</td>\n",
       "      <td>0.005065</td>\n",
       "      <td>1.070163e-03</td>\n",
       "      <td>0.003051</td>\n",
       "      <td>0.002937</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001975</td>\n",
       "      <td>0.000705</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>0.001906</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001417</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>0.035476</td>\n",
       "      <td>0.034458</td>\n",
       "      <td>0.009730</td>\n",
       "      <td>0.077813</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.126341e-03</td>\n",
       "      <td>0.003832</td>\n",
       "      <td>0.004252</td>\n",
       "      <td>0.001099</td>\n",
       "      <td>0.001403</td>\n",
       "      <td>0.001735</td>\n",
       "      <td>0.007494</td>\n",
       "      <td>0.002437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020219</td>\n",
       "      <td>0.202983</td>\n",
       "      <td>0.036619</td>\n",
       "      <td>0.005260</td>\n",
       "      <td>0.010867</td>\n",
       "      <td>0.026271</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000774</td>\n",
       "      <td>0.006444</td>\n",
       "      <td>0.002850</td>\n",
       "      <td>0.028631</td>\n",
       "      <td>0.003391</td>\n",
       "      <td>6.505063e-05</td>\n",
       "      <td>0.005748</td>\n",
       "      <td>0.003449</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007221</td>\n",
       "      <td>0.002463</td>\n",
       "      <td>0.001730</td>\n",
       "      <td>0.003593</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002722</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>0.010787</td>\n",
       "      <td>0.030170</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.006240</td>\n",
       "      <td>0.000843</td>\n",
       "      <td>0.013298</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.075898e-03</td>\n",
       "      <td>0.044442</td>\n",
       "      <td>0.007392</td>\n",
       "      <td>0.038069</td>\n",
       "      <td>0.001705</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.001683</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.007506</td>\n",
       "      <td>0.011687</td>\n",
       "      <td>0.007802</td>\n",
       "      <td>0.236762</td>\n",
       "      <td>0.002000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031547</td>\n",
       "      <td>0.060401</td>\n",
       "      <td>0.002674</td>\n",
       "      <td>0.007568</td>\n",
       "      <td>0.020358</td>\n",
       "      <td>8.830169e-04</td>\n",
       "      <td>0.046192</td>\n",
       "      <td>0.028185</td>\n",
       "      <td>0.004211</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007488</td>\n",
       "      <td>0.003326</td>\n",
       "      <td>0.021032</td>\n",
       "      <td>0.009091</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003241</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>38</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>0.007775</td>\n",
       "      <td>0.007967</td>\n",
       "      <td>0.004277</td>\n",
       "      <td>0.023796</td>\n",
       "      <td>0.006894</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.072893e-02</td>\n",
       "      <td>0.015666</td>\n",
       "      <td>0.012153</td>\n",
       "      <td>0.002013</td>\n",
       "      <td>0.022173</td>\n",
       "      <td>0.049573</td>\n",
       "      <td>0.031808</td>\n",
       "      <td>0.020901</td>\n",
       "      <td>0.031114</td>\n",
       "      <td>0.015425</td>\n",
       "      <td>0.016782</td>\n",
       "      <td>0.013980</td>\n",
       "      <td>0.003123</td>\n",
       "      <td>0.006981</td>\n",
       "      <td>0.003016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.223901</td>\n",
       "      <td>0.004081</td>\n",
       "      <td>0.006659</td>\n",
       "      <td>0.077016</td>\n",
       "      <td>0.019907</td>\n",
       "      <td>0.002568</td>\n",
       "      <td>6.208051e-03</td>\n",
       "      <td>0.004277</td>\n",
       "      <td>0.002825</td>\n",
       "      <td>0.005709</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005132</td>\n",
       "      <td>0.027330</td>\n",
       "      <td>0.026134</td>\n",
       "      <td>0.001764</td>\n",
       "      <td>0.117402</td>\n",
       "      <td>0.002949</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>33</td>\n",
       "      <td>0.008696</td>\n",
       "      <td>0.015462</td>\n",
       "      <td>0.000928</td>\n",
       "      <td>0.012578</td>\n",
       "      <td>0.010429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042321</td>\n",
       "      <td>1.461037e-02</td>\n",
       "      <td>0.010402</td>\n",
       "      <td>0.032544</td>\n",
       "      <td>0.019463</td>\n",
       "      <td>0.045588</td>\n",
       "      <td>0.009441</td>\n",
       "      <td>0.121918</td>\n",
       "      <td>0.019408</td>\n",
       "      <td>0.055733</td>\n",
       "      <td>0.068582</td>\n",
       "      <td>0.066080</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>0.051806</td>\n",
       "      <td>0.021616</td>\n",
       "      <td>0.066942</td>\n",
       "      <td>0.060691</td>\n",
       "      <td>0.172685</td>\n",
       "      <td>0.038025</td>\n",
       "      <td>0.026868</td>\n",
       "      <td>0.051846</td>\n",
       "      <td>0.013403</td>\n",
       "      <td>0.033000</td>\n",
       "      <td>2.704980e-03</td>\n",
       "      <td>0.014169</td>\n",
       "      <td>0.020728</td>\n",
       "      <td>0.020828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031882</td>\n",
       "      <td>0.021745</td>\n",
       "      <td>0.016628</td>\n",
       "      <td>0.015850</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008990</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>5</td>\n",
       "      <td>0.016485</td>\n",
       "      <td>0.033778</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.026544</td>\n",
       "      <td>0.007079</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.187953e-02</td>\n",
       "      <td>0.017696</td>\n",
       "      <td>0.023896</td>\n",
       "      <td>0.006809</td>\n",
       "      <td>0.005603</td>\n",
       "      <td>0.102423</td>\n",
       "      <td>0.007465</td>\n",
       "      <td>0.013933</td>\n",
       "      <td>0.017159</td>\n",
       "      <td>0.005944</td>\n",
       "      <td>0.014655</td>\n",
       "      <td>0.001666</td>\n",
       "      <td>0.015445</td>\n",
       "      <td>0.032030</td>\n",
       "      <td>0.011790</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006495</td>\n",
       "      <td>0.178102</td>\n",
       "      <td>0.041871</td>\n",
       "      <td>0.017651</td>\n",
       "      <td>0.026002</td>\n",
       "      <td>5.718955e-03</td>\n",
       "      <td>0.026544</td>\n",
       "      <td>0.021942</td>\n",
       "      <td>0.015690</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031853</td>\n",
       "      <td>0.041886</td>\n",
       "      <td>0.081145</td>\n",
       "      <td>0.010543</td>\n",
       "      <td>0.067659</td>\n",
       "      <td>0.029393</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>26</td>\n",
       "      <td>0.006005</td>\n",
       "      <td>0.008950</td>\n",
       "      <td>0.005470</td>\n",
       "      <td>0.027712</td>\n",
       "      <td>0.009344</td>\n",
       "      <td>0.047272</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.482151e-03</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.050869</td>\n",
       "      <td>0.007518</td>\n",
       "      <td>0.028166</td>\n",
       "      <td>0.045997</td>\n",
       "      <td>0.018194</td>\n",
       "      <td>0.034555</td>\n",
       "      <td>0.012405</td>\n",
       "      <td>0.003502</td>\n",
       "      <td>0.005605</td>\n",
       "      <td>0.020393</td>\n",
       "      <td>0.022097</td>\n",
       "      <td>0.006393</td>\n",
       "      <td>0.009852</td>\n",
       "      <td>0.018963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076273</td>\n",
       "      <td>0.032824</td>\n",
       "      <td>0.092447</td>\n",
       "      <td>0.030131</td>\n",
       "      <td>0.027008</td>\n",
       "      <td>7.435886e-04</td>\n",
       "      <td>0.007459</td>\n",
       "      <td>0.020401</td>\n",
       "      <td>0.015654</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025106</td>\n",
       "      <td>0.005834</td>\n",
       "      <td>0.008004</td>\n",
       "      <td>0.011704</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.014353</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>15</td>\n",
       "      <td>0.004169</td>\n",
       "      <td>0.005005</td>\n",
       "      <td>0.005240</td>\n",
       "      <td>0.004133</td>\n",
       "      <td>0.014199</td>\n",
       "      <td>0.002046</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.326769e-03</td>\n",
       "      <td>0.002987</td>\n",
       "      <td>0.007304</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.000899</td>\n",
       "      <td>0.001433</td>\n",
       "      <td>0.002589</td>\n",
       "      <td>0.001633</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.004802</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001439</td>\n",
       "      <td>0.003765</td>\n",
       "      <td>0.002645</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.004133</td>\n",
       "      <td>0.015999</td>\n",
       "      <td>0.002619</td>\n",
       "      <td>0.002654</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.004393</td>\n",
       "      <td>0.002721</td>\n",
       "      <td>0.014630</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005664</td>\n",
       "      <td>0.001815</td>\n",
       "      <td>0.001538</td>\n",
       "      <td>0.000986</td>\n",
       "      <td>0.001104</td>\n",
       "      <td>0.001880</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003538</td>\n",
       "      <td>1.291828e-03</td>\n",
       "      <td>0.000384</td>\n",
       "      <td>0.003662</td>\n",
       "      <td>0.010621</td>\n",
       "      <td>0.001675</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.002371</td>\n",
       "      <td>0.001310</td>\n",
       "      <td>0.000423</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001809</td>\n",
       "      <td>0.001762</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001087</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.016823</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003170</td>\n",
       "      <td>0.000670</td>\n",
       "      <td>0.012517</td>\n",
       "      <td>0.000330</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.002217</td>\n",
       "      <td>0.001311</td>\n",
       "      <td>0.002016</td>\n",
       "      <td>0.000866</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>0.001404</td>\n",
       "      <td>0.002809</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002341</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.680286e-03</td>\n",
       "      <td>0.002942</td>\n",
       "      <td>0.003386</td>\n",
       "      <td>0.005035</td>\n",
       "      <td>0.003413</td>\n",
       "      <td>0.003645</td>\n",
       "      <td>0.001932</td>\n",
       "      <td>0.000651</td>\n",
       "      <td>0.001947</td>\n",
       "      <td>0.004011</td>\n",
       "      <td>0.004239</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002688</td>\n",
       "      <td>0.002006</td>\n",
       "      <td>0.001751</td>\n",
       "      <td>0.005783</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002608</td>\n",
       "      <td>0.001621</td>\n",
       "      <td>0.003338</td>\n",
       "      <td>0.027717</td>\n",
       "      <td>8.176311e-03</td>\n",
       "      <td>0.003380</td>\n",
       "      <td>0.003077</td>\n",
       "      <td>0.001395</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004075</td>\n",
       "      <td>0.011763</td>\n",
       "      <td>0.002746</td>\n",
       "      <td>0.008768</td>\n",
       "      <td>0.004182</td>\n",
       "      <td>0.001177</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003143</td>\n",
       "      <td>0.000503</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.001877</td>\n",
       "      <td>0.001038</td>\n",
       "      <td>0.005421</td>\n",
       "      <td>0.001063</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001290</td>\n",
       "      <td>0.003589</td>\n",
       "      <td>0.005892</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003298</td>\n",
       "      <td>0.001490</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.001439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.943328e-03</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000417</td>\n",
       "      <td>0.001364</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002935</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>37</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.003827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.524882e-03</td>\n",
       "      <td>0.015955</td>\n",
       "      <td>0.004685</td>\n",
       "      <td>0.005406</td>\n",
       "      <td>0.006242</td>\n",
       "      <td>0.008580</td>\n",
       "      <td>0.003907</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007945</td>\n",
       "      <td>0.000597</td>\n",
       "      <td>0.001904</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002365</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001624</td>\n",
       "      <td>0.004639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000357</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001042</td>\n",
       "      <td>0.003340</td>\n",
       "      <td>9.888114e-02</td>\n",
       "      <td>0.008646</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.010604</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008096</td>\n",
       "      <td>0.006269</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>16</td>\n",
       "      <td>0.004124</td>\n",
       "      <td>0.002754</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027103</td>\n",
       "      <td>0.000897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.051166e-02</td>\n",
       "      <td>0.042941</td>\n",
       "      <td>0.003175</td>\n",
       "      <td>0.005592</td>\n",
       "      <td>0.005191</td>\n",
       "      <td>0.001768</td>\n",
       "      <td>0.003109</td>\n",
       "      <td>0.032932</td>\n",
       "      <td>0.007879</td>\n",
       "      <td>0.008882</td>\n",
       "      <td>0.006724</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007366</td>\n",
       "      <td>0.034191</td>\n",
       "      <td>0.006123</td>\n",
       "      <td>0.002505</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018612</td>\n",
       "      <td>0.001311</td>\n",
       "      <td>0.003277</td>\n",
       "      <td>0.046634</td>\n",
       "      <td>1.420022e-02</td>\n",
       "      <td>0.094450</td>\n",
       "      <td>0.074726</td>\n",
       "      <td>0.001501</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002754</td>\n",
       "      <td>0.020732</td>\n",
       "      <td>0.003674</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001110</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>22</td>\n",
       "      <td>0.004912</td>\n",
       "      <td>0.009711</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009222</td>\n",
       "      <td>0.001320</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.064638</td>\n",
       "      <td>1.939057e-02</td>\n",
       "      <td>0.009287</td>\n",
       "      <td>0.030398</td>\n",
       "      <td>0.028433</td>\n",
       "      <td>0.023836</td>\n",
       "      <td>0.022357</td>\n",
       "      <td>0.018828</td>\n",
       "      <td>0.022895</td>\n",
       "      <td>0.042368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022558</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.002395</td>\n",
       "      <td>0.011749</td>\n",
       "      <td>0.020922</td>\n",
       "      <td>0.022498</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000171</td>\n",
       "      <td>0.052921</td>\n",
       "      <td>0.003497</td>\n",
       "      <td>0.016938</td>\n",
       "      <td>0.004414</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.015845</td>\n",
       "      <td>0.023615</td>\n",
       "      <td>0.129122</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.038167</td>\n",
       "      <td>0.005815</td>\n",
       "      <td>0.018737</td>\n",
       "      <td>0.002750</td>\n",
       "      <td>0.000735</td>\n",
       "      <td>0.003660</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>30</td>\n",
       "      <td>0.003264</td>\n",
       "      <td>0.002760</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002300</td>\n",
       "      <td>0.006349</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041479</td>\n",
       "      <td>8.567207e-02</td>\n",
       "      <td>0.001533</td>\n",
       "      <td>0.002003</td>\n",
       "      <td>0.035251</td>\n",
       "      <td>0.015872</td>\n",
       "      <td>0.007740</td>\n",
       "      <td>0.001336</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016907</td>\n",
       "      <td>0.000985</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000552</td>\n",
       "      <td>0.016586</td>\n",
       "      <td>0.000690</td>\n",
       "      <td>0.019555</td>\n",
       "      <td>0.024960</td>\n",
       "      <td>0.001672</td>\n",
       "      <td>0.005087</td>\n",
       "      <td>0.002768</td>\n",
       "      <td>0.001254</td>\n",
       "      <td>0.001331</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002300</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001687</td>\n",
       "      <td>0.164228</td>\n",
       "      <td>0.002759</td>\n",
       "      <td>0.000986</td>\n",
       "      <td>0.007534</td>\n",
       "      <td>0.000531</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>0.000986</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>4</td>\n",
       "      <td>0.004799</td>\n",
       "      <td>0.009599</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007999</td>\n",
       "      <td>0.018042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.114848</td>\n",
       "      <td>8.276721e-02</td>\n",
       "      <td>0.016489</td>\n",
       "      <td>0.025787</td>\n",
       "      <td>0.127823</td>\n",
       "      <td>0.073442</td>\n",
       "      <td>0.017985</td>\n",
       "      <td>0.018367</td>\n",
       "      <td>0.045706</td>\n",
       "      <td>0.015974</td>\n",
       "      <td>0.006349</td>\n",
       "      <td>0.007365</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017991</td>\n",
       "      <td>0.013022</td>\n",
       "      <td>0.017447</td>\n",
       "      <td>0.019970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021521</td>\n",
       "      <td>0.021001</td>\n",
       "      <td>0.011286</td>\n",
       "      <td>0.043040</td>\n",
       "      <td>0.004486</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.007999</td>\n",
       "      <td>0.004935</td>\n",
       "      <td>0.020575</td>\n",
       "      <td>0.311681</td>\n",
       "      <td>0.009599</td>\n",
       "      <td>0.007034</td>\n",
       "      <td>0.015030</td>\n",
       "      <td>0.006418</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003428</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.472932e-04</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000850</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001244</td>\n",
       "      <td>0.000843</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.001734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006098</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.004848</td>\n",
       "      <td>0.000753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012195</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003553</td>\n",
       "      <td>0.003551</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000365</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.532819e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000197</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000564</td>\n",
       "      <td>0.001373</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.044953e-05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001025</td>\n",
       "      <td>0.000844</td>\n",
       "      <td>0.002682</td>\n",
       "      <td>0.001442</td>\n",
       "      <td>0.002231</td>\n",
       "      <td>0.000396</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>10</td>\n",
       "      <td>0.005533</td>\n",
       "      <td>0.005625</td>\n",
       "      <td>0.001531</td>\n",
       "      <td>0.003437</td>\n",
       "      <td>0.000827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002604</td>\n",
       "      <td>2.052996e-03</td>\n",
       "      <td>0.003475</td>\n",
       "      <td>0.005411</td>\n",
       "      <td>0.002003</td>\n",
       "      <td>0.001900</td>\n",
       "      <td>0.005424</td>\n",
       "      <td>0.001481</td>\n",
       "      <td>0.001467</td>\n",
       "      <td>0.001522</td>\n",
       "      <td>0.003901</td>\n",
       "      <td>0.001611</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>0.003702</td>\n",
       "      <td>0.002946</td>\n",
       "      <td>0.003591</td>\n",
       "      <td>0.011404</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005850</td>\n",
       "      <td>0.003795</td>\n",
       "      <td>0.005907</td>\n",
       "      <td>0.001996</td>\n",
       "      <td>4.161423e-03</td>\n",
       "      <td>0.003437</td>\n",
       "      <td>0.012707</td>\n",
       "      <td>0.002839</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004124</td>\n",
       "      <td>0.008849</td>\n",
       "      <td>0.011859</td>\n",
       "      <td>0.034071</td>\n",
       "      <td>0.023007</td>\n",
       "      <td>0.009162</td>\n",
       "      <td>0.016725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000781</td>\n",
       "      <td>0.001561</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001301</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.591217e-04</td>\n",
       "      <td>0.001224</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>0.000754</td>\n",
       "      <td>0.001077</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.000558</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002639</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>0.001126</td>\n",
       "      <td>0.004647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.001301</td>\n",
       "      <td>0.001696</td>\n",
       "      <td>0.002146</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>6.148411e-03</td>\n",
       "      <td>0.009971</td>\n",
       "      <td>0.007749</td>\n",
       "      <td>0.001315</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011750</td>\n",
       "      <td>0.004447</td>\n",
       "      <td>0.006172</td>\n",
       "      <td>0.019605</td>\n",
       "      <td>0.007488</td>\n",
       "      <td>0.001688</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>31</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.001832</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002980</td>\n",
       "      <td>0.002621</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.005649</td>\n",
       "      <td>0.004048</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003485</td>\n",
       "      <td>0.009683</td>\n",
       "      <td>0.007611</td>\n",
       "      <td>0.001176</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>0.000597</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001876</td>\n",
       "      <td>6.509566e-04</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.001848</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024381</td>\n",
       "      <td>0.013445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>0.015282</td>\n",
       "      <td>0.011712</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.010910</td>\n",
       "      <td>0.003207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.104593e-03</td>\n",
       "      <td>0.016116</td>\n",
       "      <td>0.019700</td>\n",
       "      <td>0.017369</td>\n",
       "      <td>0.032802</td>\n",
       "      <td>0.019810</td>\n",
       "      <td>0.018015</td>\n",
       "      <td>0.008086</td>\n",
       "      <td>0.023357</td>\n",
       "      <td>0.014497</td>\n",
       "      <td>0.021824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013619</td>\n",
       "      <td>0.016497</td>\n",
       "      <td>0.020153</td>\n",
       "      <td>0.000817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004658</td>\n",
       "      <td>0.011923</td>\n",
       "      <td>0.015948</td>\n",
       "      <td>0.011164</td>\n",
       "      <td>0.014592</td>\n",
       "      <td>1.063190e-02</td>\n",
       "      <td>0.030816</td>\n",
       "      <td>0.023968</td>\n",
       "      <td>0.018736</td>\n",
       "      <td>0.036759</td>\n",
       "      <td>0.011712</td>\n",
       "      <td>0.023837</td>\n",
       "      <td>0.016833</td>\n",
       "      <td>0.023543</td>\n",
       "      <td>0.033491</td>\n",
       "      <td>0.007618</td>\n",
       "      <td>0.138560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    order  topic       GAM       GEM       EMG       MGD       EMB       MSD       NHM           SRL       GER       NLS       PIG       GRS       ROT       LEF       IST       ARC       MSL       FSM       OVP       SRF       SSC       FLT       SEN       MTM       STT       RTH       TCM       CUL       PRA           CPR       WAT       TMP       LIT       MEC       MPH       NUT       HRM       CHS       MCH       PTH       OBI\n",
       "0       0      3  0.076284  0.003959  0.000000  0.016848  0.000044  0.000000  0.000000  4.283903e-03  0.000000  0.000581  0.002419  0.000556  0.000000  0.000380  0.005617  0.002117  0.000000  0.000000  0.000000  0.003597  0.000000  0.000834  0.002384  0.000000  0.000000  0.000824  0.000000  0.000000  0.000000  0.000000e+00  0.000245  0.010256  0.000000  0.000000  0.000000  0.000000  0.000834  0.000076  0.000000  0.000000  0.000000\n",
       "1       1     24  0.082733  0.009446  0.001537  0.017872  0.000339  0.009721  0.000000  1.593859e-03  0.003191  0.005394  0.002710  0.003352  0.001981  0.001919  0.011893  0.002803  0.005643  0.014627  0.000000  0.033349  0.001080  0.001743  0.002580  0.000000  0.007809  0.002025  0.004153  0.008524  0.001489  4.526762e-03  0.001240  0.011141  0.000929  0.000000  0.013736  0.000334  0.001749  0.004185  0.007506  0.001041  0.000000\n",
       "2       2      1  0.027227  0.011609  0.095080  0.001010  0.013777  0.000000  0.000000  2.126438e-02  0.003949  0.003779  0.009216  0.000545  0.004516  0.000461  0.000801  0.002013  0.001390  0.001884  0.000000  0.001144  0.000866  0.000306  0.000000  0.000000  0.001096  0.003110  0.000986  0.003070  0.000567  4.224044e-04  0.001010  0.000606  0.000383  0.000000  0.001212  0.000689  0.002581  0.000353  0.001977  0.000433  0.000000\n",
       "3       3     17  0.000152  0.000000  0.000379  0.000000  0.008291  0.000000  0.000281  6.327772e-04  0.000000  0.000435  0.000238  0.000043  0.000000  0.000166  0.000000  0.000073  0.000179  0.000062  0.000000  0.000000  0.000000  0.000398  0.000106  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000e+00  0.000000  0.000000  0.000040  0.000000  0.000000  0.000000  0.000037  0.000000  0.000000  0.000000  0.000000\n",
       "4       4      2  0.014415  0.020483  0.050040  0.022687  0.005832  0.053081  0.000000  1.551137e-02  0.036639  0.019778  0.030213  0.023450  0.020206  0.016841  0.021650  0.020645  0.037356  0.025890  0.009345  0.015055  0.022114  0.032296  0.020620  0.000000  0.019487  0.028734  0.025016  0.026165  0.040020  1.555394e-02  0.016904  0.038592  0.026897  0.000000  0.020284  0.038561  0.015501  0.008961  0.000000  0.022707  0.000000\n",
       "5       5      9  0.005669  0.006224  0.015926  0.004878  0.009833  0.037933  0.000000  3.633048e-03  0.003357  0.005175  0.002612  0.001000  0.001471  0.001035  0.000106  0.000213  0.000135  0.002164  0.000000  0.001169  0.004181  0.001537  0.000000  0.000000  0.000000  0.004878  0.002116  0.006437  0.002661  0.000000e+00  0.004878  0.003224  0.001837  0.000000  0.005854  0.002118  0.001729  0.001239  0.000000  0.002191  0.000000\n",
       "6       6     40  0.000330  0.000185  0.008155  0.000154  0.001066  0.024790  0.000000  1.178688e-03  0.001284  0.009390  0.002173  0.003544  0.000040  0.002311  0.000000  0.002578  0.001580  0.000066  0.000065  0.001275  0.002033  0.003872  0.013274  0.000000  0.002595  0.002372  0.000066  0.010055  0.009616  2.023576e-03  0.000154  0.002958  0.006279  0.000000  0.012581  0.000066  0.004230  0.000668  0.000000  0.000128  0.000000\n",
       "7       7     39  0.000926  0.001822  0.001930  0.001523  0.002631  0.006285  0.035787  1.602167e-02  0.001013  0.001207  0.017470  0.004987  0.000556  0.002134  0.001168  0.000494  0.005958  0.000651  0.000000  0.000627  0.030043  0.000553  0.000000  0.000000  0.004473  0.007775  0.001283  0.003527  0.001498  0.000000e+00  0.002133  0.001712  0.007679  0.000000  0.001823  0.001115  0.000752  0.000578  0.000882  0.000651  0.000000\n",
       "8       8     25  0.006671  0.012729  0.021005  0.025876  0.009524  0.000409  0.157451  1.304053e-02  0.028726  0.021890  0.017524  0.020969  0.008290  0.009959  0.018868  0.017603  0.042049  0.016916  0.005305  0.009430  0.018462  0.013113  0.008458  0.036994  0.011636  0.024320  0.007580  0.022725  0.008893  2.897721e-03  0.041524  0.031675  0.019677  0.000000  0.012537  0.033642  0.010014  0.016361  0.004657  0.020980  0.000000\n",
       "9       9     36  0.012837  0.007225  0.012889  0.004663  0.016436  0.018711  0.071562  1.001991e-01  0.003741  0.005929  0.034533  0.028327  0.005865  0.004795  0.000609  0.005205  0.031796  0.002705  0.000000  0.001928  0.003997  0.001540  0.038092  0.044936  0.000118  0.004663  0.005295  0.007216  0.004011  4.014630e-04  0.004663  0.004046  0.012700  0.076252  0.006515  0.003170  0.019642  0.002641  0.000000  0.005840  0.000000\n",
       "10     10     32  0.000112  0.000000  0.000000  0.000000  0.000129  0.000000  0.000000  3.354751e-03  0.000000  0.002345  0.001541  0.000056  0.000073  0.000457  0.001004  0.000376  0.000000  0.001622  0.001450  0.001006  0.000000  0.001449  0.000000  0.000000  0.000168  0.000000  0.000580  0.000000  0.000000  7.299274e-04  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000117  0.000000  0.000134  0.000000\n",
       "11     11     20  0.001856  0.003125  0.001477  0.002604  0.001363  0.004369  0.000000  1.644227e-03  0.024748  0.004027  0.001440  0.001425  0.004336  0.002683  0.000044  0.003090  0.001612  0.003374  0.000000  0.003964  0.003751  0.006314  0.008872  0.000000  0.000062  0.002604  0.002164  0.005246  0.002281  6.725981e-03  0.010451  0.010825  0.000976  0.000000  0.003125  0.003273  0.011014  0.023511  0.006808  0.002148  0.000000\n",
       "12     12     13  0.000000  0.000000  0.000000  0.000000  0.000156  0.000000  0.000000  2.515093e-04  0.000703  0.003583  0.001458  0.001180  0.000141  0.002292  0.000016  0.000715  0.000931  0.000881  0.000325  0.000302  0.000000  0.002600  0.001079  0.000000  0.001265  0.000813  0.000000  0.000595  0.000041  2.323182e-03  0.000000  0.000000  0.000026  0.000000  0.001184  0.000178  0.000846  0.000242  0.000000  0.000018  0.000000\n",
       "13     13     41  0.001245  0.002501  0.000000  0.001280  0.004953  0.000000  0.000000  5.134306e-03  0.000925  0.146578  0.016716  0.022830  0.035425  0.009414  0.079916  0.034362  0.000431  0.006022  0.000697  0.008752  0.001097  0.030356  0.009425  0.000000  0.020543  0.009382  0.025922  0.023424  0.004478  0.000000e+00  0.001280  0.000768  0.104747  0.000000  0.001535  0.017089  0.034239  0.010471  0.000000  0.006945  0.000000\n",
       "14     14     34  0.009407  0.000000  0.000000  0.000000  0.000619  0.000000  0.000000  2.785738e-04  0.001025  0.021810  0.014261  0.032747  0.016591  0.008171  0.013460  0.031059  0.000000  0.013545  0.000000  0.014248  0.000000  0.017964  0.023927  0.000000  0.018750  0.000000  0.013393  0.002675  0.000809  0.000000e+00  0.000000  0.000000  0.003574  0.000000  0.000000  0.000386  0.010891  0.000758  0.000000  0.000305  0.000000\n",
       "15     15     11  0.003801  0.007603  0.000652  0.006336  0.002225  0.000000  0.000000  5.754604e-03  0.004224  0.014215  0.009182  0.014836  0.019847  0.009649  0.003366  0.042169  0.020935  0.024689  0.000000  0.014783  0.011486  0.024659  0.027236  0.000000  0.004388  0.041348  0.003501  0.010415  0.003456  4.047169e-08  0.006336  0.003801  0.007698  0.000000  0.007603  0.002715  0.014731  0.001583  0.000000  0.003009  0.000000\n",
       "16     16      7  0.000000  0.000000  0.000421  0.000000  0.000285  0.002642  0.000000  0.000000e+00  0.001187  0.000097  0.000000  0.000354  0.000088  0.004753  0.000539  0.013182  0.003465  0.011511  0.000000  0.005075  0.000000  0.007924  0.000095  0.000000  0.000066  0.000000  0.000145  0.000306  0.000000  0.000000e+00  0.000000  0.000000  0.000042  0.000000  0.000133  0.000006  0.000085  0.000000  0.000000  0.000115  0.000000\n",
       "17     17     19  0.000818  0.000000  0.000000  0.000000  0.000137  0.000000  0.000000  2.142219e-05  0.000000  0.011961  0.002652  0.000084  0.005532  0.000608  0.016690  0.018714  0.000000  0.000815  0.000084  0.002547  0.002099  0.000233  0.000000  0.000000  0.004848  0.003228  0.009803  0.003669  0.000000  3.111841e-05  0.000000  0.000000  0.000333  0.000000  0.001646  0.000000  0.001115  0.000172  0.000000  0.000044  0.000000\n",
       "18     18     12  0.000000  0.000000  0.000000  0.000000  0.000818  0.000000  0.000000  3.766100e-04  0.000000  0.001853  0.000803  0.005153  0.001889  0.004320  0.004399  0.005958  0.000100  0.009022  0.000000  0.002928  0.000000  0.001521  0.000000  0.000000  0.002898  0.000000  0.005527  0.003264  0.001602  9.231195e-05  0.000000  0.000000  0.000000  0.000000  0.000000  0.001990  0.000063  0.000878  0.004160  0.000395  0.000000\n",
       "19     19     27  0.000987  0.001975  0.000115  0.001646  0.000916  0.000000  0.000000  2.459350e-03  0.004134  0.004491  0.002061  0.001200  0.011568  0.009638  0.002572  0.015486  0.004530  0.013080  0.016956  0.009992  0.002745  0.012754  0.002542  0.011361  0.001734  0.003202  0.012036  0.003393  0.005065  1.070163e-03  0.003051  0.002937  0.000629  0.000000  0.001975  0.000705  0.007300  0.001906  0.000000  0.001417  0.000000\n",
       "20     20     21  0.035476  0.034458  0.009730  0.077813  0.000791  0.000000  0.000000  2.126341e-03  0.003832  0.004252  0.001099  0.001403  0.001735  0.007494  0.002437  0.000000  0.000000  0.020219  0.202983  0.036619  0.005260  0.010867  0.026271  0.000000  0.000774  0.006444  0.002850  0.028631  0.003391  6.505063e-05  0.005748  0.003449  0.002257  0.000000  0.007221  0.002463  0.001730  0.003593  0.000000  0.002722  0.000000\n",
       "21     21     14  0.010787  0.030170  0.000099  0.006240  0.000843  0.013298  0.000000  6.075898e-03  0.044442  0.007392  0.038069  0.001705  0.001765  0.001683  0.000182  0.000908  0.001125  0.007506  0.011687  0.007802  0.236762  0.002000  0.000000  0.000000  0.031547  0.060401  0.002674  0.007568  0.020358  8.830169e-04  0.046192  0.028185  0.004211  0.000000  0.007488  0.003326  0.021032  0.009091  0.000000  0.003241  0.000000\n",
       "22     22     38  0.005244  0.007775  0.007967  0.004277  0.023796  0.006894  0.000000  4.072893e-02  0.015666  0.012153  0.002013  0.022173  0.049573  0.031808  0.020901  0.031114  0.015425  0.016782  0.013980  0.003123  0.006981  0.003016  0.000000  0.223901  0.004081  0.006659  0.077016  0.019907  0.002568  6.208051e-03  0.004277  0.002825  0.005709  0.000000  0.005132  0.027330  0.026134  0.001764  0.117402  0.002949  0.000000\n",
       "23     23     33  0.008696  0.015462  0.000928  0.012578  0.010429  0.000000  0.042321  1.461037e-02  0.010402  0.032544  0.019463  0.045588  0.009441  0.121918  0.019408  0.055733  0.068582  0.066080  0.001190  0.051806  0.021616  0.066942  0.060691  0.172685  0.038025  0.026868  0.051846  0.013403  0.033000  2.704980e-03  0.014169  0.020728  0.020828  0.000000  0.031882  0.021745  0.016628  0.015850  0.000000  0.008990  0.000000\n",
       "24     24      5  0.016485  0.033778  0.000260  0.026544  0.007079  0.000000  0.000000  1.187953e-02  0.017696  0.023896  0.006809  0.005603  0.102423  0.007465  0.013933  0.017159  0.005944  0.014655  0.001666  0.015445  0.032030  0.011790  0.000040  0.000000  0.006495  0.178102  0.041871  0.017651  0.026002  5.718955e-03  0.026544  0.021942  0.015690  0.000000  0.031853  0.041886  0.081145  0.010543  0.067659  0.029393  0.000000\n",
       "25     25     26  0.006005  0.008950  0.005470  0.027712  0.009344  0.047272  0.000000  9.482151e-03  0.005495  0.050869  0.007518  0.028166  0.045997  0.018194  0.034555  0.012405  0.003502  0.005605  0.020393  0.022097  0.006393  0.009852  0.018963  0.000000  0.076273  0.032824  0.092447  0.030131  0.027008  7.435886e-04  0.007459  0.020401  0.015654  0.000000  0.025106  0.005834  0.008004  0.011704  0.000372  0.014353  0.000000\n",
       "26     26     15  0.004169  0.005005  0.005240  0.004133  0.014199  0.002046  0.000000  8.326769e-03  0.002987  0.007304  0.000756  0.000899  0.001433  0.002589  0.001633  0.000159  0.000074  0.004802  0.000000  0.001439  0.003765  0.002645  0.000000  0.000000  0.000353  0.004133  0.015999  0.002619  0.002654  0.000000e+00  0.004393  0.002721  0.014630  0.000000  0.005664  0.001815  0.001538  0.000986  0.001104  0.001880  0.000000\n",
       "27     27     28  0.000000  0.000000  0.000000  0.000000  0.004789  0.000000  0.003538  1.291828e-03  0.000384  0.003662  0.010621  0.001675  0.002473  0.001174  0.002371  0.001310  0.000423  0.001869  0.000000  0.001552  0.000000  0.001809  0.001762  0.000000  0.001087  0.000000  0.000003  0.016823  0.000181  0.000000e+00  0.000000  0.003170  0.000670  0.012517  0.000330  0.000689  0.002217  0.001311  0.002016  0.000866  0.000000\n",
       "28     28      0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000e+00  0.000000  0.000026  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000020  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000056  0.000000e+00  0.000001  0.000000  0.000000  0.000000  0.000000  0.000044  0.000000  0.000004  0.000000  0.000005  0.000000\n",
       "29     29     29  0.001404  0.002809  0.000000  0.002341  0.001369  0.000000  0.000000  5.680286e-03  0.002942  0.003386  0.005035  0.003413  0.003645  0.001932  0.000651  0.001947  0.004011  0.004239  0.000000  0.002688  0.002006  0.001751  0.005783  0.000000  0.000000  0.002608  0.001621  0.003338  0.027717  8.176311e-03  0.003380  0.003077  0.001395  0.000000  0.004075  0.011763  0.002746  0.008768  0.004182  0.001177  0.000000\n",
       "30     30      8  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000e+00  0.000000  0.003143  0.000503  0.000499  0.001877  0.001038  0.005421  0.001063  0.000000  0.001290  0.003589  0.005892  0.000000  0.003298  0.001490  0.000000  0.000000  0.000000  0.000415  0.001439  0.000000  7.943328e-03  0.000000  0.000000  0.001978  0.000000  0.000000  0.000417  0.001364  0.000163  0.000000  0.002935  0.000000\n",
       "31     31     37  0.000000  0.000000  0.000000  0.000084  0.003827  0.000000  0.000000  2.524882e-03  0.015955  0.004685  0.005406  0.006242  0.008580  0.003907  0.000000  0.007945  0.000597  0.001904  0.000000  0.002365  0.000000  0.001624  0.004639  0.000000  0.000357  0.000000  0.000000  0.001042  0.003340  9.888114e-02  0.008646  0.005747  0.010604  0.000000  0.000715  0.000000  0.008096  0.006269  0.000000  0.000496  0.000000\n",
       "32     32     16  0.004124  0.002754  0.000000  0.027103  0.000897  0.000000  0.000000  1.051166e-02  0.042941  0.003175  0.005592  0.005191  0.001768  0.003109  0.032932  0.007879  0.008882  0.006724  0.000000  0.007366  0.034191  0.006123  0.002505  0.000000  0.000000  0.018612  0.001311  0.003277  0.046634  1.420022e-02  0.094450  0.074726  0.001501  0.000000  0.002754  0.020732  0.003674  0.001221  0.000000  0.001110  0.000000\n",
       "33     33     22  0.004912  0.009711  0.000000  0.009222  0.001320  0.000000  0.064638  1.939057e-02  0.009287  0.030398  0.028433  0.023836  0.022357  0.018828  0.022895  0.042368  0.000000  0.022558  0.000010  0.002395  0.011749  0.020922  0.022498  0.000000  0.000171  0.052921  0.003497  0.016938  0.004414  0.000000e+00  0.015845  0.023615  0.129122  0.000000  0.038167  0.005815  0.018737  0.002750  0.000735  0.003660  0.000000\n",
       "34     34     30  0.003264  0.002760  0.000000  0.002300  0.006349  0.000000  0.041479  8.567207e-02  0.001533  0.002003  0.035251  0.015872  0.007740  0.001336  0.000000  0.000000  0.016907  0.000985  0.000000  0.000552  0.016586  0.000690  0.019555  0.024960  0.001672  0.005087  0.002768  0.001254  0.001331  0.000000e+00  0.002300  0.001380  0.001687  0.164228  0.002759  0.000986  0.007534  0.000531  0.000307  0.000986  0.000000\n",
       "35     35      4  0.004799  0.009599  0.000000  0.007999  0.018042  0.000000  0.114848  8.276721e-02  0.016489  0.025787  0.127823  0.073442  0.017985  0.018367  0.045706  0.015974  0.006349  0.007365  0.000000  0.017991  0.013022  0.017447  0.019970  0.000000  0.021521  0.021001  0.011286  0.043040  0.004486  0.000000e+00  0.007999  0.004935  0.020575  0.311681  0.009599  0.007034  0.015030  0.006418  0.000000  0.003428  0.000000\n",
       "36     36     18  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  7.472932e-04  0.000194  0.000000  0.000850  0.000625  0.000000  0.001244  0.000843  0.000002  0.000003  0.001734  0.000000  0.000000  0.000249  0.000000  0.000000  0.000000  0.006098  0.000000  0.000000  0.000000  0.000000  0.000000e+00  0.004848  0.000753  0.000000  0.000000  0.012195  0.000000  0.003553  0.003551  0.000000  0.000365  0.000000\n",
       "37     37     35  0.000000  0.000220  0.000000  0.000000  0.000009  0.000000  0.000000  9.532819e-07  0.000000  0.000197  0.000066  0.000564  0.001373  0.000536  0.000000  0.001207  0.000000  0.000000  0.000000  0.000042  0.000000  0.000000  0.000110  0.000000  0.000031  0.000608  0.000011  0.000157  0.000000  5.044953e-05  0.000000  0.000199  0.000953  0.000000  0.001025  0.000844  0.002682  0.001442  0.002231  0.000396  0.000000\n",
       "38     38     10  0.005533  0.005625  0.001531  0.003437  0.000827  0.000000  0.002604  2.052996e-03  0.003475  0.005411  0.002003  0.001900  0.005424  0.001481  0.001467  0.001522  0.003901  0.001611  0.000458  0.003702  0.002946  0.003591  0.011404  0.000000  0.000000  0.005850  0.003795  0.005907  0.001996  4.161423e-03  0.003437  0.012707  0.002839  0.000000  0.004124  0.008849  0.011859  0.034071  0.023007  0.009162  0.016725\n",
       "39     39      6  0.000781  0.001561  0.000000  0.001301  0.000182  0.000000  0.000000  4.591217e-04  0.001224  0.001670  0.000754  0.001077  0.002113  0.000269  0.000089  0.000000  0.000505  0.000558  0.000000  0.002639  0.001115  0.001126  0.004647  0.000000  0.000124  0.001301  0.001696  0.002146  0.001552  6.148411e-03  0.009971  0.007749  0.001315  0.000000  0.011750  0.004447  0.006172  0.019605  0.007488  0.001688  0.000000\n",
       "40     40     31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000e+00  0.000160  0.000087  0.001832  0.000000  0.002980  0.002621  0.000093  0.005649  0.004048  0.000000  0.000000  0.000000  0.000000  0.003485  0.009683  0.007611  0.001176  0.001511  0.000597  0.000000  0.001876  6.509566e-04  0.001894  0.000163  0.000000  0.000000  0.000000  0.000000  0.000153  0.001848  0.000000  0.024381  0.013445\n",
       "41     41     23  0.015282  0.011712  0.000130  0.010910  0.003207  0.000000  0.000000  9.104593e-03  0.016116  0.019700  0.017369  0.032802  0.019810  0.018015  0.008086  0.023357  0.014497  0.021824  0.000000  0.013619  0.016497  0.020153  0.000817  0.000000  0.004658  0.011923  0.015948  0.011164  0.014592  1.063190e-02  0.030816  0.023968  0.018736  0.036759  0.011712  0.023837  0.016833  0.023543  0.033491  0.007618  0.138560"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What to use as the list of texts from which to actually generate the topic model, one of the processed ones above.\n",
    "texts = list(descriptions.values())\n",
    "texts = list(processed[\"simple\"].values())\n",
    "texts = list(processed[\"full\"].values())\n",
    "\n",
    "\n",
    "# Creating and fitting the topic model, either NFM or LDA.\n",
    "number_of_topics = 42\n",
    "seed = 0\n",
    "vectorizer = TfidfVectorizer(max_features=10000, stop_words=\"english\", max_df=0.95, min_df=2, lowercase=False)\n",
    "features = vectorizer.fit_transform(texts)\n",
    "cls = NMF(n_components=number_of_topics, random_state=seed)\n",
    "cls.fit(features)\n",
    "\n",
    "# Function for retrieving the topic vectors for a list of text descriptions.\n",
    "def get_topic_embeddings(texts, model, vectorizer):\n",
    "    ngrams_vectors = vectorizer.transform(texts).toarray()\n",
    "    topic_vectors = model.transform(ngrams_vectors)\n",
    "    return(topic_vectors)\n",
    "    \n",
    "# Create the dataframe containing the average score assigned to each topic for the genes from each subset.\n",
    "group_to_topic_vector = {}\n",
    "for group_id,ids in group_id_to_ids.items():\n",
    "    texts = [descriptions[i] for i in ids]\n",
    "    topic_vectors = get_topic_embeddings(texts, cls, vectorizer)\n",
    "    mean_topic_vector = np.mean(topic_vectors, axis=0)\n",
    "    group_to_topic_vector[group_id] = mean_topic_vector\n",
    "    \n",
    "tm_df = pd.DataFrame(group_to_topic_vector)\n",
    "\n",
    "# Changing the order of the Lloyd, Meinke phenotype subsets to match other figures for consistency.\n",
    "filename = \"../data/group_related_files/lloyd/lloyd_function_hierarchy_irb_cleaned.csv\"\n",
    "lmtm_df = pd.read_csv(filename)    \n",
    "columns_in_order = [col for col in lmtm_df[\"Subset Symbol\"].values if col in tm_df.columns]\n",
    "tm_df = tm_df[columns_in_order]\n",
    "    \n",
    "# Reordering so consistency with the curated subsets can be checked by looking at the diagonal.\n",
    "tm_df[\"idxmax\"] = tm_df.idxmax(axis = 1)\n",
    "tm_df[\"idxmax\"] = tm_df[\"idxmax\"].apply(lambda x: tm_df.columns.get_loc(x))\n",
    "tm_df = tm_df.sort_values(by=\"idxmax\")\n",
    "tm_df.drop(columns=[\"idxmax\"], inplace=True)\n",
    "tm_df = tm_df.reset_index(drop=False).rename({\"index\":\"topic\"},axis=1).reset_index(drop=False).rename({\"index\":\"order\"},axis=1)\n",
    "tm_df.to_csv(os.path.join(OUTPUT_DIR,\"part_6_topic_modeling.csv\"), index=False)\n",
    "tm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0: visibl phenotyp height kanamycin averag \n",
      "1: embryo defect transit infer development \n",
      "2: type wild plant compar mutant \n",
      "3: gametophyt femal defect male complet \n",
      "4: pale green slow yellow semi \n",
      "5: root later long respons hair \n",
      "6: sensit aluminum salt stress agent \n",
      "7: flower earli late senesc slightli \n",
      "8: reduc fertil siliqu length small \n",
      "9: globular embryo defect arrest stage \n",
      "10: resist kanamycin stress diseas salt \n",
      "11: short hair pedicel height develop \n",
      "12: abnorm morpholog trichom develop floral \n",
      "13: leav rosett curl serrat brittl \n",
      "14: seed coat mucilag brown testa \n",
      "15: cotyledon embryo shape vein bodi \n",
      "16: low acid temperatur rate fatti \n",
      "17: preglobular embryo defect homozygot embryon \n",
      "18: aba stomat closur respons insensit \n",
      "19: infloresc stem humid bright glossi \n",
      "20: germin stress rate salt exogen \n",
      "21: pollen tube tetrad viabl format \n",
      "22: light red long etiol hypocotyl \n",
      "23: growth plant condit slow normal \n",
      "24: anther steril male meiosi defect \n",
      "25: mutant doubl singl indistinguish grown \n",
      "26: cell wall divis death gala \n",
      "27: increas number densiti acid later \n",
      "28: delai chloroplast accumul biogenesi green \n",
      "29: level elev oxid starch glucosinol \n",
      "30: infer pigment lethal albino seedl \n",
      "31: suscept pathogen pseudomona infect syringa \n",
      "32: tassel gener term ear defect \n",
      "33: leaf narrow surfac glossi midrib \n",
      "34: dwarf gibberellin dwarfism associ null \n",
      "35: phenotyp observ obviou describ relat \n",
      "36: homozyg lethal state mutat standard \n",
      "37: fluoresc illumin post mediat plastoquinon \n",
      "38: shoot meristem apic larger differenti \n",
      "39: seedl yellow white palegreen viresc \n",
      "40: kernel small textur acid irregular \n",
      "41: hypocotyl ethylen dark slightli longer \n"
     ]
    }
   ],
   "source": [
    "# Describing what the most representative tokens for each topic in the model are.\n",
    "num_top_words = 5\n",
    "feature_names = vectorizer.get_feature_names()\n",
    "for i,topic_vec in enumerate(cls.components_):\n",
    "    print(i,end=\": \")\n",
    "    for fid in topic_vec.argsort()[-1:-num_top_words-1:-1]:\n",
    "        word = feature_names[fid]\n",
    "        # The next line is applicable if words in the topic model are actually a function of the words in the texts.\n",
    "        #word = \" \".join(unreduce[word])\n",
    "        print(word, end=\" \")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"clustering\"></a>\n",
    "### Approach 2: Agglomerative clustering and comparison to predefined groups\n",
    "This clustering approach uses agglomerative clustering to cluster the genes into a fixed number of clusters based off the distances between their embedding representations using all of the above methods. Clustering into a fixed number of clusters allows for clustering into a similar number of groups as a present in some existing grouping of the data, such as phenotype categories or biochemical pathways, and then determining if the clusters obtained are at all similar to the groupings that already exist. Agglomerative clustering is used here in order to use an arbitrary predefined distance matrix, in this case the matrix being used is the mean distance percentiles from each of the different approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Generate the numpy array where values are mean distance percentiles between all the methods.\n",
    "mean_pct_array = name_to_array[\"Mean\"]\n",
    "to_id = array_index_to_id\n",
    "\n",
    "# Do agglomerative clustering based on that distance matrix.\n",
    "number_of_clusters = 42\n",
    "ac = AgglomerativeClustering(n_clusters=number_of_clusters, linkage=\"complete\", affinity=\"precomputed\")\n",
    "clustering = ac.fit(mean_pct_array)\n",
    "id_to_cluster = {}\n",
    "cluster_to_ids = defaultdict(list)\n",
    "for idx,c in enumerate(clustering.labels_):\n",
    "    id_to_cluster[to_id[idx]] = c\n",
    "    cluster_to_ids[c].append(to_id[idx])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>cluster</th>\n",
       "      <th>GAM</th>\n",
       "      <th>GEM</th>\n",
       "      <th>EMG</th>\n",
       "      <th>MGD</th>\n",
       "      <th>EMB</th>\n",
       "      <th>MSD</th>\n",
       "      <th>NHM</th>\n",
       "      <th>SRL</th>\n",
       "      <th>GER</th>\n",
       "      <th>NLS</th>\n",
       "      <th>PIG</th>\n",
       "      <th>GRS</th>\n",
       "      <th>ROT</th>\n",
       "      <th>LEF</th>\n",
       "      <th>IST</th>\n",
       "      <th>ARC</th>\n",
       "      <th>MSL</th>\n",
       "      <th>FSM</th>\n",
       "      <th>OVP</th>\n",
       "      <th>SRF</th>\n",
       "      <th>SSC</th>\n",
       "      <th>FLT</th>\n",
       "      <th>SEN</th>\n",
       "      <th>MTM</th>\n",
       "      <th>STT</th>\n",
       "      <th>RTH</th>\n",
       "      <th>TCM</th>\n",
       "      <th>CUL</th>\n",
       "      <th>PRA</th>\n",
       "      <th>CPR</th>\n",
       "      <th>WAT</th>\n",
       "      <th>TMP</th>\n",
       "      <th>LIT</th>\n",
       "      <th>MEC</th>\n",
       "      <th>MPH</th>\n",
       "      <th>NUT</th>\n",
       "      <th>HRM</th>\n",
       "      <th>CHS</th>\n",
       "      <th>MCH</th>\n",
       "      <th>PTH</th>\n",
       "      <th>OBI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>32</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>35</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.851064</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>24</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>33</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.391304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>26</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.172414</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>39</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>37</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>34</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.135135</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.310345</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>29</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>21</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.135135</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1875</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1875</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.216216</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>8</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.2500</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>40</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    order  cluster  GAM  GEM    EMG       MGD       EMB  MSD  NHM       SRL       GER       NLS       PIG       GRS       ROT       LEF       IST       ARC    MSL       FSM   OVP   SRF       SSC   FLT       SEN  MTM  STT       RTH       TCM       CUL       PRA       CPR       WAT  TMP     LIT  MEC  MPH       NUT   HRM       CHS   MCH       PTH  OBI\n",
       "0       0        1  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "1       1       23  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "2       2       32  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "3       3       17  0.6  0.6  0.375  0.333333  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "4       4       12  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "5       5       31  0.0  0.0  0.125  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "6       6       35  0.0  0.0  0.250  0.000000  0.851064  0.5  0.0  0.176471  0.000000  0.066667  0.027027  0.033333  0.086957  0.034483  0.000000  0.000000  0.000  0.071429  0.00  0.04  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.071429  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.05  0.000000  0.25  0.071429  0.0\n",
       "7       7       18  0.0  0.0  0.125  0.000000  0.000000  0.5  0.0  0.000000  0.000000  0.133333  0.000000  0.133333  0.086957  0.034483  0.142857  0.000000  0.125  0.000000  0.00  0.04  0.000000  0.00  0.000000  0.0  0.5  0.000000  0.285714  0.090909  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "8       8       24  0.0  0.0  0.000  0.000000  0.000000  0.0  0.5  0.058824  0.111111  0.000000  0.081081  0.066667  0.043478  0.034483  0.000000  0.083333  0.125  0.000000  0.00  0.00  0.000000  0.10  0.142857  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0625  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "9       9       10  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.176471  0.000000  0.000000  0.081081  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "10     10       19  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.333333  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.05  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "11     11       33  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.111111  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "12     12       25  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.066667  0.081081  0.166667  0.043478  0.068966  0.000000  0.083333  0.000  0.071429  0.00  0.04  0.000000  0.10  0.000000  0.0  0.0  0.000000  0.071429  0.090909  0.000000  0.000000  0.000000  0.0  0.0625  0.0  0.0  0.071429  0.00  0.038462  0.00  0.000000  0.0\n",
       "13     13       27  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.391304  0.000000  0.071429  0.083333  0.000  0.000000  0.00  0.04  0.000000  0.00  0.000000  0.0  0.0  0.333333  0.214286  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.05  0.000000  0.00  0.000000  0.0\n",
       "14     14       26  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.027027  0.066667  0.000000  0.172414  0.000000  0.083333  0.000  0.142857  0.00  0.12  0.000000  0.05  0.142857  0.0  0.0  0.000000  0.071429  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.038462  0.00  0.000000  0.0\n",
       "15     15        9  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.034483  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "16     16       11  0.0  0.0  0.000  0.000000  0.021277  0.0  0.0  0.000000  0.000000  0.000000  0.027027  0.066667  0.000000  0.034483  0.000000  0.000000  0.125  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.090909  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "17     17       39  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.125  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.038462  0.00  0.000000  0.0\n",
       "18     18       37  0.3  0.2  0.125  0.333333  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.034483  0.000000  0.000000  0.000  0.142857  0.75  0.32  0.000000  0.05  0.142857  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.1  0.0000  0.0  0.2  0.000000  0.00  0.038462  0.00  0.000000  0.0\n",
       "19     19       22  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.25  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.090909  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.038462  0.00  0.000000  0.0\n",
       "20     20       34  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.133333  0.000000  0.000000  0.000000  0.068966  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.15  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "21     21       30  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.15  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "22     22       28  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.058824  0.000000  0.000000  0.000000  0.033333  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.285714  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "23     23        0  0.0  0.0  0.000  0.000000  0.000000  0.0  0.5  0.000000  0.000000  0.066667  0.135135  0.100000  0.000000  0.310345  0.428571  0.250000  0.125  0.357143  0.00  0.12  0.428571  0.15  0.142857  1.0  0.1  0.166667  0.142857  0.090909  0.000000  0.000000  0.166667  0.0  0.1250  0.0  0.2  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "24     24       29  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.2  0.000000  0.071429  0.000000  0.000000  0.000000  0.166667  0.0  0.0000  0.0  0.2  0.071429  0.10  0.038462  0.00  0.000000  0.0\n",
       "25     25        3  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.1  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0625  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "26     26       16  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.166667  0.000000  0.000000  0.090909  0.000000  0.000000  0.1  0.0000  0.0  0.0  0.000000  0.00  0.038462  0.00  0.071429  0.0\n",
       "27     27       41  0.0  0.0  0.000  0.000000  0.021277  0.0  0.0  0.000000  0.000000  0.000000  0.027027  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.090909  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "28     28       21  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.090909  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "29     29        7  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.272727  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "30     30       38  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.027027  0.000000  0.000000  0.000000  0.000000  0.000000  0.125  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.090909  0.272727  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.142857  0.00  0.000000  0.00  0.000000  0.0\n",
       "31     31       15  0.0  0.0  0.000  0.166667  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.027027  0.000000  0.000000  0.000000  0.214286  0.000000  0.000  0.071429  0.00  0.20  0.000000  0.00  0.000000  0.0  0.1  0.000000  0.000000  0.000000  0.000000  0.333333  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "32     32        4  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.117647  0.000000  0.066667  0.135135  0.100000  0.000000  0.000000  0.000000  0.000000  0.125  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.181818  0.333333  0.000000  0.2  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.142857  0.0\n",
       "33     33       14  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.111111  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.142857  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.166667  0.333333  0.3  0.0000  0.0  0.2  0.214286  0.10  0.192308  0.25  0.000000  0.0\n",
       "34     34       20  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.166667  0.0  0.0000  0.0  0.0  0.000000  0.00  0.115385  0.00  0.000000  0.0\n",
       "35     35       13  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.033333  0.000000  0.034483  0.000000  0.083333  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.1875  0.0  0.0  0.000000  0.05  0.000000  0.00  0.000000  0.0\n",
       "36     36        6  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.066667  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.1875  0.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "37     37        5  0.0  0.0  0.000  0.000000  0.021277  0.0  0.0  0.235294  0.000000  0.000000  0.216216  0.066667  0.043478  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.142857  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  1.0  0.0  0.000000  0.00  0.000000  0.00  0.000000  0.0\n",
       "38     38        8  0.1  0.2  0.000  0.166667  0.085106  0.0  0.0  0.117647  0.333333  0.400000  0.108108  0.100000  0.304348  0.068966  0.142857  0.250000  0.000  0.142857  0.00  0.08  0.285714  0.20  0.142857  0.0  0.0  0.333333  0.071429  0.363636  0.090909  0.166667  0.166667  0.3  0.2500  0.0  0.2  0.428571  0.50  0.153846  0.25  0.142857  0.0\n",
       "39     39       36  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.034483  0.000000  0.000000  0.125  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.071429  0.05  0.269231  0.25  0.285714  0.0\n",
       "40     40        2  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0000  0.0  0.0  0.000000  0.00  0.000000  0.00  0.285714  0.0\n",
       "41     41       40  0.0  0.0  0.000  0.000000  0.000000  0.0  0.0  0.058824  0.000000  0.000000  0.000000  0.033333  0.000000  0.034483  0.000000  0.083333  0.000  0.000000  0.00  0.00  0.000000  0.00  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.0625  0.0  0.0  0.000000  0.10  0.000000  0.00  0.000000  1.0"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the dataframe containing the average score assigned to each topic for the genes from each subset.\n",
    "group_to_cluster_vector = {}\n",
    "for group_id,ids in group_id_to_ids.items():\n",
    "    \n",
    "    mean_cluster_vector = np.zeros(number_of_clusters)\n",
    "    for i in ids:\n",
    "        cluster = id_to_cluster[i]\n",
    "        mean_cluster_vector[cluster] = mean_cluster_vector[cluster]+1\n",
    "    mean_cluster_vector = mean_cluster_vector/mean_cluster_vector.sum(axis=0,keepdims=1)\n",
    "    group_to_cluster_vector[group_id] = mean_cluster_vector\n",
    "    \n",
    "ac_df = pd.DataFrame(group_to_cluster_vector)\n",
    "\n",
    "# Changing the order of the Lloyd, Meinke phenotype subsets to match other figures for consistency.\n",
    "filename = \"../data/group_related_files/lloyd/lloyd_function_hierarchy_irb_cleaned.csv\"\n",
    "lmtm_df = pd.read_csv(filename)    \n",
    "columns_in_order = [col for col in lmtm_df[\"Subset Symbol\"].values if col in ac_df.columns]\n",
    "ac_df = ac_df[columns_in_order]\n",
    "\n",
    "# Reordering so consistency with the curated subsets can be checked by looking at the diagonal.\n",
    "ac_df[\"idxmax\"] = ac_df.idxmax(axis = 1)\n",
    "ac_df[\"idxmax\"] = ac_df[\"idxmax\"].apply(lambda x: ac_df.columns.get_loc(x))\n",
    "ac_df = ac_df.sort_values(by=\"idxmax\")\n",
    "ac_df.drop(columns=[\"idxmax\"], inplace=True)\n",
    "ac_df = ac_df.reset_index(drop=False).rename({\"index\":\"cluster\"},axis=1).reset_index(drop=False).rename({\"index\":\"order\"},axis=1)\n",
    "ac_df.to_csv(os.path.join(OUTPUT_DIR,\"part_6_agglomerative_clustering.csv\"), index=False)\n",
    "ac_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"phenologs\"></a>\n",
    "### Approach 3: Looking for phenolog relationships between clusters and OMIM disease phenotypes\n",
    "This section produces a table of values that provides a score for the a particular pair of a cluster found for this dataset of plant genes and a disease phenotype. Currently the value indicates the fraction of the plant genes in that cluster that have orthologs associated with that disease phenotype. This should be replaced or supplemented with a p-value for evaluating the significance of this value given the distribution of genes and their mappings to all of the disease phenotypes. All the rows from the input dataframe containing the PantherDB and OMIM information where the ID from this dataset is not known or the mapping to a phenotype was unsuccessful are removed at this step, fix this if the metric for evaluating cluster to phenotype phenolog mappings need this information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_identifier</th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "      <th>gene_mim_number</th>\n",
       "      <th>phenotype_mim_number</th>\n",
       "      <th>phenotype_mim_name</th>\n",
       "      <th>id</th>\n",
       "      <th>compressed_phenotype_mim_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02090</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 1</td>\n",
       "      <td>1699</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02090</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 3B</td>\n",
       "      <td>1699</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02580</td>\n",
       "      <td>TPK1</td>\n",
       "      <td>606370.0</td>\n",
       "      <td>614458</td>\n",
       "      <td>Thiamine metabolism dysfunction syndrome 5 (ep...</td>\n",
       "      <td>611</td>\n",
       "      <td>Thiamine metabolism dysfunction syndrome 5 (ep...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02580</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>601573.0</td>\n",
       "      <td>277590</td>\n",
       "      <td>Weaver syndrome</td>\n",
       "      <td>611</td>\n",
       "      <td>Weaver syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02580</td>\n",
       "      <td>PEX6</td>\n",
       "      <td>601498.0</td>\n",
       "      <td>616617</td>\n",
       "      <td>Heimler syndrome 2</td>\n",
       "      <td>611</td>\n",
       "      <td>Heimler syndrome 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    species gene_identifier human_ortholog_gene_symbol  gene_mim_number  phenotype_mim_number                                 phenotype_mim_name    id                      compressed_phenotype_mim_name\n",
       "185   ARATH       AT1G02090                     KIF21A         608283.0                135700     Fibrosis of extraocular muscles, congenital, 1  1699                    Fibrosis of extraocular muscles\n",
       "186   ARATH       AT1G02090                     KIF21A         608283.0                135700    Fibrosis of extraocular muscles, congenital, 3B  1699                    Fibrosis of extraocular muscles\n",
       "286   ARATH       AT1G02580                       TPK1         606370.0                614458  Thiamine metabolism dysfunction syndrome 5 (ep...   611  Thiamine metabolism dysfunction syndrome 5 (ep...\n",
       "288   ARATH       AT1G02580                       EZH2         601573.0                277590                                    Weaver syndrome   611                                    Weaver syndrome\n",
       "289   ARATH       AT1G02580                       PEX6         601498.0                616617                                 Heimler syndrome 2   611                                 Heimler syndrome 2"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the dataframe mapping plant genes --> human orthologs --> disease phenotypes.\n",
    "omim_df = pd.read_csv(panther_to_omim_filename)\n",
    "# Add a column that indicates which ID in the dataset those plant genes refer to, for mapping to phenotypes.\n",
    "name_to_id = dataset.get_name_to_id_dictionary()\n",
    "omim_df[\"id\"] = omim_df[\"gene_identifier\"].map(lambda x: name_to_id.get(x,None))\n",
    "omim_df = omim_df.dropna(subset=[\"id\",\"phenotype_mim_name\"], inplace=False)\n",
    "omim_df[\"phenotype_mim_name\"] = omim_df[\"phenotype_mim_name\"].astype(str)\n",
    "omim_df[\"compressed_phenotype_mim_name\"] = omim_df[\"phenotype_mim_name\"].map(lambda x: x.split(\",\")[0])\n",
    "omim_df[\"id\"] = omim_df[\"id\"].astype(\"int64\")\n",
    "omim_df[\"phenotype_mim_number\"] = omim_df[\"phenotype_mim_number\"].astype(\"int64\")\n",
    "# Generate mappings between the IDs in this dataset and disease phenotypes or orthologous genes.\n",
    "id_to_mim_phenotype_names = defaultdict(list)\n",
    "for i,p in zip(omim_df[\"id\"].values,omim_df[\"compressed_phenotype_mim_name\"].values):\n",
    "    id_to_mim_phenotype_names[i].append(p)\n",
    "id_to_human_gene_symbols = defaultdict(list)\n",
    "for i,s in zip(omim_df[\"id\"].values,omim_df[\"human_ortholog_gene_symbol\"].values):\n",
    "    id_to_human_gene_symbols[i].append(s)\n",
    "omim_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "compressed_phenotype_mim_name\n",
       "3-Methylcrotonyl-CoA carboxylase 1 deficiency     3\n",
       "3-hydroxyisobutryl-CoA hydrolase deficiency       1\n",
       "3-methylglutaconic aciduria                       4\n",
       "5-fluorouracil toxicity                           1\n",
       "?Aniridia 2                                       2\n",
       "                                                 ..\n",
       "{Pulmonary fibrosis and/or bone marrow failure    1\n",
       "{Rhabdoid tumor predisposition syndrome 2}        3\n",
       "{Rheumatoid arthritis                             3\n",
       "{Yao syndrome}                                    1\n",
       "{von Hippel-Lindau syndrome                       2\n",
       "Length: 504, dtype: int64"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many genes in our dataset map to orthologs that map to the same OMIM phenotype?\n",
    "omim_df.groupby(\"compressed_phenotype_mim_name\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>7</th>\n",
       "      <th>16</th>\n",
       "      <th>21</th>\n",
       "      <th>0</th>\n",
       "      <th>37</th>\n",
       "      <th>36</th>\n",
       "      <th>39</th>\n",
       "      <th>9</th>\n",
       "      <th>17</th>\n",
       "      <th>4</th>\n",
       "      <th>8</th>\n",
       "      <th>10</th>\n",
       "      <th>3</th>\n",
       "      <th>40</th>\n",
       "      <th>25</th>\n",
       "      <th>14</th>\n",
       "      <th>18</th>\n",
       "      <th>11</th>\n",
       "      <th>24</th>\n",
       "      <th>6</th>\n",
       "      <th>28</th>\n",
       "      <th>26</th>\n",
       "      <th>38</th>\n",
       "      <th>29</th>\n",
       "      <th>12</th>\n",
       "      <th>5</th>\n",
       "      <th>2</th>\n",
       "      <th>27</th>\n",
       "      <th>13</th>\n",
       "      <th>15</th>\n",
       "      <th>1</th>\n",
       "      <th>20</th>\n",
       "      <th>33</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>41</th>\n",
       "      <th>22</th>\n",
       "      <th>19</th>\n",
       "      <th>23</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Fibrosis of extraocular muscles</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thiamine metabolism dysfunction syndrome 5 (episodic encephalopathy type)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Weaver syndrome</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Heimler syndrome 2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peroxisome biogenesis disorder 4A (Zellweger)</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    34  35  7   16  21  0   37  36  39  9   17  4   8   10  3   40  25  14  18  11  24  6   28  26  38  29  12  5   2   27  13  15  1   20  33  30  31  32  41  22  19  23\n",
       "Fibrosis of extraocular muscles                      0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Thiamine metabolism dysfunction syndrome 5 (epi...   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Weaver syndrome                                      0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Heimler syndrome 2                                   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Peroxisome biogenesis disorder 4A (Zellweger)        0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenolog_x_dict = defaultdict(dict)\n",
    "phenolog_p_dict = defaultdict(dict)\n",
    "candidate_genes_dict = defaultdict(dict)\n",
    "phenotypes = pd.unique(omim_df[\"compressed_phenotype_mim_name\"].values)\n",
    "clusters = list(cluster_to_ids.keys())\n",
    "for cluster,phenotype in itertools.product(clusters,phenotypes):\n",
    "    \n",
    "    # What are the candidate genes predicted if this phenolog pairing is real?\n",
    "    ids = cluster_to_ids[cluster]\n",
    "    candidate_genes_dict[cluster][phenotype] = list(set(flatten([id_to_human_gene_symbols[i] for i in ids if phenotype not in id_to_mim_phenotype_names.get(i,[])])))\n",
    "\n",
    "    # What is the p-value for this phenolog pairing?\n",
    "    # The size of the population (genes in the dataset).\n",
    "    M = len(id_to_cluster.keys())\n",
    "    # The number of elements we draw without replacement (genes in the cluster).\n",
    "    N = len(cluster_to_ids[cluster])     \n",
    "    # The number of available successes in the population (genes that map to orthologs that map to this phenotype).\n",
    "    n = len([i for i in id_to_cluster.keys() if phenotype in id_to_mim_phenotype_names.get(i,[])])\n",
    "    # The number of successes drawn (genes in this cluster that map to orthologs that map to this phenotype).\n",
    "    x = list(set(flatten([id_to_mim_phenotype_names.get(i,[]) for i in ids]))).count(phenotype)\n",
    "    prob = 1-hypergeom.cdf(x-1, M, n, N) # Equivalent to prob = 1-sum([hypergeom.pmf(x_i, M, n, N) for x_i in range(0,x)])\n",
    "    phenolog_x_dict[cluster][phenotype] = x\n",
    "    phenolog_p_dict[cluster][phenotype] = prob\n",
    "    \n",
    "\n",
    "# Convert the dictionary to a table of values with cluster and phenotype as the rows and columns.\n",
    "phenolog_matrix = pd.DataFrame(phenolog_x_dict)        \n",
    "phenolog_matrix.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>omim_phenotype_name</th>\n",
       "      <th>cluster</th>\n",
       "      <th>size</th>\n",
       "      <th>x</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_adjusted</th>\n",
       "      <th>candidate_gene_symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4217</th>\n",
       "      <td>Smith-McCort dysplasia 2</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MOGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18335</th>\n",
       "      <td>{Endometrial cancer</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18076</th>\n",
       "      <td>Immunodeficiency</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CERS1|PSMC3IP|CERS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18075</th>\n",
       "      <td>Combined oxidative phosphorylation deficiency 39</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CERS1|PSMC3IP|CERS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9180</th>\n",
       "      <td>Muscular dystrophy-dystroglycanopathy (limb-gi...</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ABCB7|GBE1|CTDP1|ALG3|MSH5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9179</th>\n",
       "      <td>Muscular dystrophy-dystroglycanopathy (congeni...</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ABCB7|GBE1|CTDP1|ALG3|MSH5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6445</th>\n",
       "      <td>Pontocerebellar hypoplasia type 2B</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6446</th>\n",
       "      <td>Neurodevelopmental disorder with dysmorphic fa...</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9178</th>\n",
       "      <td>Muscular dystrophy-dystroglycanopathy (congeni...</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ABCB7|GBE1|CTDP1|ALG3|MSH5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18215</th>\n",
       "      <td>Microphthalmia</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12860</th>\n",
       "      <td>?Immunodeficiency 61</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10884</th>\n",
       "      <td>[Gamma-glutamyltransferase</td>\n",
       "      <td>28</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19032</th>\n",
       "      <td>Pyruvate kinase deficiency</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12859</th>\n",
       "      <td>Glomerulosclerosis</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19035</th>\n",
       "      <td>Renal-hepatic-pancreatic dysplasia 1</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10885</th>\n",
       "      <td>?Glutathioninuria</td>\n",
       "      <td>28</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19031</th>\n",
       "      <td>Adenosine triphosphate</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17678</th>\n",
       "      <td>Ovarian dysgenesis 3</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GFM2|PBX1|TUSC3|MAGT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19033</th>\n",
       "      <td>Meckel syndrome 7</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19034</th>\n",
       "      <td>Nephronophthisis 3</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4219</th>\n",
       "      <td>{Asthma</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MOGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18320</th>\n",
       "      <td>Mismatch repair cancer syndrome</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15784</th>\n",
       "      <td>Lethal congenital contractural syndrome 3</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|LCT|PPT1|DPYD|OPA1|MAT1A|DLD|KL|WASF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15793</th>\n",
       "      <td>Methionine adenosyltransferase deficiency</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15794</th>\n",
       "      <td>5-fluorouracil toxicity</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11896</th>\n",
       "      <td>Retinitis pigmentosa 68</td>\n",
       "      <td>38</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RNF6|CYP19A1|XDH|COG7|DUOX2|CYP24A1|CYP27A1|SL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15795</th>\n",
       "      <td>Dihydropyrimidine dehydrogenase deficiency</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15796</th>\n",
       "      <td>Dihydrolipoamide dehydrogenase deficiency</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15780</th>\n",
       "      <td>Muscular dystrophy-dystroglycanopathy (limb-gi...</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|LCT|PPT1|DPYD|OPA1|MAT1A|DLD|KL|WASF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15781</th>\n",
       "      <td>Adams-Oliver syndrome 4</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LYST|DNM2|LCT|PPT1|DPYD|OPA1|MAT1A|DLD|KL|WASF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     omim_phenotype_name cluster  size  x   p_value  p_adjusted                             candidate_gene_symbols\n",
       "4217                            Smith-McCort dysplasia 2       9     2  1  0.004000         1.0                                               MOGS\n",
       "18335                                {Endometrial cancer      32     1  1  0.006000         1.0                                                   \n",
       "18076                                   Immunodeficiency      31     3  1  0.006000         1.0                                CERS1|PSMC3IP|CERS3\n",
       "18075   Combined oxidative phosphorylation deficiency 39      31     3  1  0.006000         1.0                                CERS1|PSMC3IP|CERS3\n",
       "9180   Muscular dystrophy-dystroglycanopathy (limb-gi...      11     4  1  0.008000         1.0                         ABCB7|GBE1|CTDP1|ALG3|MSH5\n",
       "9179   Muscular dystrophy-dystroglycanopathy (congeni...      11     4  1  0.008000         1.0                         ABCB7|GBE1|CTDP1|ALG3|MSH5\n",
       "6445                  Pontocerebellar hypoplasia type 2B      10     4  1  0.008000         1.0                                                   \n",
       "6446   Neurodevelopmental disorder with dysmorphic fa...      10     4  1  0.008000         1.0                                                   \n",
       "9178   Muscular dystrophy-dystroglycanopathy (congeni...      11     4  1  0.008000         1.0                         ABCB7|GBE1|CTDP1|ALG3|MSH5\n",
       "18215                                     Microphthalmia      32     1  1  0.008000         1.0                                                   \n",
       "12860                               ?Immunodeficiency 61      12     3  1  0.011976         1.0                                                   \n",
       "10884                         [Gamma-glutamyltransferase      28     3  1  0.011976         1.0                                                   \n",
       "19032                         Pyruvate kinase deficiency      41     3  1  0.011976         1.0           NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1\n",
       "12859                                 Glomerulosclerosis      12     3  1  0.011976         1.0                                                   \n",
       "19035               Renal-hepatic-pancreatic dysplasia 1      41     3  1  0.011976         1.0           NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1\n",
       "10885                                  ?Glutathioninuria      28     3  1  0.011976         1.0                                                   \n",
       "19031                             Adenosine triphosphate      41     3  1  0.011976         1.0           NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1\n",
       "17678                               Ovarian dysgenesis 3      31     3  1  0.011976         1.0                              GFM2|PBX1|TUSC3|MAGT1\n",
       "19033                                  Meckel syndrome 7      41     3  1  0.011976         1.0           NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1\n",
       "19034                                 Nephronophthisis 3      41     3  1  0.011976         1.0           NPC1L1|NPC1|CYB5R3|MAPKAPK3|ACACA|UGT1A1\n",
       "4219                                             {Asthma       9     2  1  0.011976         1.0                                               MOGS\n",
       "18320                    Mismatch repair cancer syndrome      32     1  1  0.012000         1.0                                                   \n",
       "15784          Lethal congenital contractural syndrome 3       1     7  1  0.014000         1.0  LYST|DNM2|LCT|PPT1|DPYD|OPA1|MAT1A|DLD|KL|WASF...\n",
       "15793          Methionine adenosyltransferase deficiency       1     7  1  0.014000         1.0  LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...\n",
       "15794                            5-fluorouracil toxicity       1     7  1  0.014000         1.0  LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...\n",
       "11896                            Retinitis pigmentosa 68      38     7  1  0.014000         1.0  RNF6|CYP19A1|XDH|COG7|DUOX2|CYP24A1|CYP27A1|SL...\n",
       "15795         Dihydropyrimidine dehydrogenase deficiency       1     7  1  0.014000         1.0  LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...\n",
       "15796          Dihydrolipoamide dehydrogenase deficiency       1     7  1  0.014000         1.0  LYST|DNM2|VCP|EOGT|LCT|PPT1|PIP5K1C|OPA1|POMGN...\n",
       "15780  Muscular dystrophy-dystroglycanopathy (limb-gi...       1     7  1  0.014000         1.0  LYST|DNM2|LCT|PPT1|DPYD|OPA1|MAT1A|DLD|KL|WASF...\n",
       "15781                            Adams-Oliver syndrome 4       1     7  1  0.014000         1.0  LYST|DNM2|LCT|PPT1|DPYD|OPA1|MAT1A|DLD|KL|WASF..."
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Produce a melted version of the phenolog matrix sorted by value and including predicted candidate genes.\n",
    "phenolog_matrix_reset = phenolog_matrix.reset_index(drop=False).rename({\"index\":\"omim_phenotype_name\"}, axis=\"columns\")\n",
    "phenolog_df = pd.melt(phenolog_matrix_reset, id_vars=[\"omim_phenotype_name\"], value_vars=phenolog_matrix.columns[1:], var_name=\"cluster\", value_name=\"x\")\n",
    "# What other information should be present in this melted phenologs matrix?\n",
    "phenolog_df[\"size\"] = phenolog_df[\"cluster\"].map(lambda x: len(cluster_to_ids[x]))\n",
    "phenolog_df[\"candidate_gene_symbols\"] = np.vectorize(lambda x,y: concatenate_with_bar_delim(*candidate_genes_dict[x][y]))(phenolog_df[\"cluster\"], phenolog_df[\"omim_phenotype_name\"])\n",
    "phenolog_df[\"p_value\"] = np.vectorize(lambda x,y: phenolog_p_dict[x][y])(phenolog_df[\"cluster\"], phenolog_df[\"omim_phenotype_name\"])\n",
    "phenolog_df[\"p_adjusted\"] = multipletests(phenolog_df[\"p_value\"].values, method='bonferroni')[1]\n",
    "phenolog_df.sort_values(by=[\"p_value\"], inplace=True, ascending=True)\n",
    "phenolog_df = phenolog_df[[\"omim_phenotype_name\", \"cluster\", \"size\", \"x\", \"p_value\", \"p_adjusted\", \"candidate_gene_symbols\"]]\n",
    "phenolog_df.to_csv(os.path.join(OUTPUT_DIR,\"part_6_phenologs.csv\"), index=False)\n",
    "phenolog_df.head(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Approach 4: Agglomerative clustering and sillhouette scores for each NLP method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Max</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=20</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=50</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=100</th>\n",
       "      <th>Topic Models:NMF,Full,Topics=200</th>\n",
       "      <th>N-Grams:Full,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,TFIDF</th>\n",
       "      <th>N-Grams:Full,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>N-Grams:Full,Nouns,Adjectives,1-grams</th>\n",
       "      <th>N-Grams:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Precise_Annotations,Words,1-grams</th>\n",
       "      <th>N-Grams:Full,Partial_Annotations,Words,1-grams</th>\n",
       "      <th>NOBLE Coder:Precise</th>\n",
       "      <th>NOBLE Coder:Partial</th>\n",
       "      <th>NOBLE Coder:Precise,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec Phenes:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Mean</th>\n",
       "      <th>Word2Vec Phenes:Wikipedia,Size=300,Max</th>\n",
       "      <th>N-Grams Phenes:Linares_Pontes,Words,1-grams</th>\n",
       "      <th>Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>0.111449</td>\n",
       "      <td>0.218015</td>\n",
       "      <td>0.341037</td>\n",
       "      <td>0.544834</td>\n",
       "      <td>0.188647</td>\n",
       "      <td>0.183020</td>\n",
       "      <td>0.152741</td>\n",
       "      <td>0.119528</td>\n",
       "      <td>0.088996</td>\n",
       "      <td>0.099786</td>\n",
       "      <td>0.089147</td>\n",
       "      <td>0.145943</td>\n",
       "      <td>0.276095</td>\n",
       "      <td>0.283791</td>\n",
       "      <td>0.412045</td>\n",
       "      <td>0.198858</td>\n",
       "      <td>0.142067</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.109463</td>\n",
       "      <td>0.319382</td>\n",
       "      <td>0.564720</td>\n",
       "      <td>0.186704</td>\n",
       "      <td>0.279598</td>\n",
       "      <td>0.315499</td>\n",
       "      <td>0.265143</td>\n",
       "      <td>0.197958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>0.120581</td>\n",
       "      <td>0.225574</td>\n",
       "      <td>0.335043</td>\n",
       "      <td>0.512995</td>\n",
       "      <td>0.240044</td>\n",
       "      <td>0.217882</td>\n",
       "      <td>0.170117</td>\n",
       "      <td>0.133973</td>\n",
       "      <td>0.096264</td>\n",
       "      <td>0.104005</td>\n",
       "      <td>0.093017</td>\n",
       "      <td>0.161037</td>\n",
       "      <td>0.278569</td>\n",
       "      <td>0.235299</td>\n",
       "      <td>0.410044</td>\n",
       "      <td>0.207532</td>\n",
       "      <td>0.145981</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.125490</td>\n",
       "      <td>0.330643</td>\n",
       "      <td>0.570454</td>\n",
       "      <td>0.230586</td>\n",
       "      <td>0.292356</td>\n",
       "      <td>0.335228</td>\n",
       "      <td>0.274763</td>\n",
       "      <td>0.222490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>0.130072</td>\n",
       "      <td>0.230533</td>\n",
       "      <td>0.329656</td>\n",
       "      <td>0.515729</td>\n",
       "      <td>0.290730</td>\n",
       "      <td>0.245884</td>\n",
       "      <td>0.178295</td>\n",
       "      <td>0.148342</td>\n",
       "      <td>0.101041</td>\n",
       "      <td>0.110101</td>\n",
       "      <td>0.096648</td>\n",
       "      <td>0.159462</td>\n",
       "      <td>0.263483</td>\n",
       "      <td>0.237463</td>\n",
       "      <td>0.322416</td>\n",
       "      <td>0.218477</td>\n",
       "      <td>0.152849</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.136101</td>\n",
       "      <td>0.332342</td>\n",
       "      <td>0.577437</td>\n",
       "      <td>0.242324</td>\n",
       "      <td>0.317798</td>\n",
       "      <td>0.316644</td>\n",
       "      <td>0.280276</td>\n",
       "      <td>0.212201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32</td>\n",
       "      <td>0.134597</td>\n",
       "      <td>0.225221</td>\n",
       "      <td>0.313976</td>\n",
       "      <td>0.502566</td>\n",
       "      <td>0.346565</td>\n",
       "      <td>0.272363</td>\n",
       "      <td>0.207765</td>\n",
       "      <td>0.158469</td>\n",
       "      <td>0.103876</td>\n",
       "      <td>0.113537</td>\n",
       "      <td>0.101670</td>\n",
       "      <td>0.167928</td>\n",
       "      <td>0.262828</td>\n",
       "      <td>0.234073</td>\n",
       "      <td>0.371505</td>\n",
       "      <td>0.220331</td>\n",
       "      <td>0.164366</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.163207</td>\n",
       "      <td>0.338161</td>\n",
       "      <td>0.588521</td>\n",
       "      <td>0.246228</td>\n",
       "      <td>0.335157</td>\n",
       "      <td>0.289350</td>\n",
       "      <td>0.289456</td>\n",
       "      <td>0.210839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36</td>\n",
       "      <td>0.136947</td>\n",
       "      <td>0.219404</td>\n",
       "      <td>0.298411</td>\n",
       "      <td>0.501666</td>\n",
       "      <td>0.370164</td>\n",
       "      <td>0.289452</td>\n",
       "      <td>0.222839</td>\n",
       "      <td>0.163598</td>\n",
       "      <td>0.111255</td>\n",
       "      <td>0.123295</td>\n",
       "      <td>0.107448</td>\n",
       "      <td>0.177537</td>\n",
       "      <td>0.261080</td>\n",
       "      <td>0.233573</td>\n",
       "      <td>0.363496</td>\n",
       "      <td>0.222286</td>\n",
       "      <td>0.166248</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.171188</td>\n",
       "      <td>0.344131</td>\n",
       "      <td>0.589310</td>\n",
       "      <td>0.254740</td>\n",
       "      <td>0.327207</td>\n",
       "      <td>0.296632</td>\n",
       "      <td>0.293667</td>\n",
       "      <td>0.208027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40</td>\n",
       "      <td>0.139424</td>\n",
       "      <td>0.229064</td>\n",
       "      <td>0.284544</td>\n",
       "      <td>0.504362</td>\n",
       "      <td>0.414206</td>\n",
       "      <td>0.308702</td>\n",
       "      <td>0.230935</td>\n",
       "      <td>0.169062</td>\n",
       "      <td>0.113839</td>\n",
       "      <td>0.128191</td>\n",
       "      <td>0.116499</td>\n",
       "      <td>0.188847</td>\n",
       "      <td>0.262360</td>\n",
       "      <td>0.223171</td>\n",
       "      <td>0.350074</td>\n",
       "      <td>0.226538</td>\n",
       "      <td>0.169389</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.184442</td>\n",
       "      <td>0.349342</td>\n",
       "      <td>0.602143</td>\n",
       "      <td>0.265526</td>\n",
       "      <td>0.327124</td>\n",
       "      <td>0.304860</td>\n",
       "      <td>0.290711</td>\n",
       "      <td>0.219085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>44</td>\n",
       "      <td>0.142921</td>\n",
       "      <td>0.227454</td>\n",
       "      <td>0.219824</td>\n",
       "      <td>0.509531</td>\n",
       "      <td>0.452335</td>\n",
       "      <td>0.330700</td>\n",
       "      <td>0.246308</td>\n",
       "      <td>0.171892</td>\n",
       "      <td>0.116414</td>\n",
       "      <td>0.134430</td>\n",
       "      <td>0.118322</td>\n",
       "      <td>0.197694</td>\n",
       "      <td>0.264310</td>\n",
       "      <td>0.200789</td>\n",
       "      <td>0.347361</td>\n",
       "      <td>0.230504</td>\n",
       "      <td>0.174017</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.193577</td>\n",
       "      <td>0.351535</td>\n",
       "      <td>0.602000</td>\n",
       "      <td>0.272019</td>\n",
       "      <td>0.335583</td>\n",
       "      <td>0.310045</td>\n",
       "      <td>0.289071</td>\n",
       "      <td>0.232856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>48</td>\n",
       "      <td>0.140025</td>\n",
       "      <td>0.232583</td>\n",
       "      <td>0.219534</td>\n",
       "      <td>0.512629</td>\n",
       "      <td>0.482208</td>\n",
       "      <td>0.356072</td>\n",
       "      <td>0.258147</td>\n",
       "      <td>0.173398</td>\n",
       "      <td>0.119572</td>\n",
       "      <td>0.140521</td>\n",
       "      <td>0.122365</td>\n",
       "      <td>0.201090</td>\n",
       "      <td>0.246159</td>\n",
       "      <td>0.204340</td>\n",
       "      <td>0.343282</td>\n",
       "      <td>0.235898</td>\n",
       "      <td>0.176611</td>\n",
       "      <td>0.016717</td>\n",
       "      <td>0.199486</td>\n",
       "      <td>0.355267</td>\n",
       "      <td>0.606176</td>\n",
       "      <td>0.278888</td>\n",
       "      <td>0.350051</td>\n",
       "      <td>0.311840</td>\n",
       "      <td>0.292557</td>\n",
       "      <td>0.233261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>52</td>\n",
       "      <td>0.142208</td>\n",
       "      <td>0.241484</td>\n",
       "      <td>0.222807</td>\n",
       "      <td>0.516074</td>\n",
       "      <td>0.478101</td>\n",
       "      <td>0.369188</td>\n",
       "      <td>0.268344</td>\n",
       "      <td>0.177673</td>\n",
       "      <td>0.120130</td>\n",
       "      <td>0.146656</td>\n",
       "      <td>0.125271</td>\n",
       "      <td>0.209408</td>\n",
       "      <td>0.249790</td>\n",
       "      <td>0.207168</td>\n",
       "      <td>0.318069</td>\n",
       "      <td>0.233944</td>\n",
       "      <td>0.180250</td>\n",
       "      <td>0.018367</td>\n",
       "      <td>0.210171</td>\n",
       "      <td>0.358777</td>\n",
       "      <td>0.598570</td>\n",
       "      <td>0.279232</td>\n",
       "      <td>0.357229</td>\n",
       "      <td>0.327182</td>\n",
       "      <td>0.290027</td>\n",
       "      <td>0.238097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>56</td>\n",
       "      <td>0.145246</td>\n",
       "      <td>0.248364</td>\n",
       "      <td>0.217597</td>\n",
       "      <td>0.512095</td>\n",
       "      <td>0.481746</td>\n",
       "      <td>0.388429</td>\n",
       "      <td>0.274894</td>\n",
       "      <td>0.180903</td>\n",
       "      <td>0.128455</td>\n",
       "      <td>0.150693</td>\n",
       "      <td>0.128101</td>\n",
       "      <td>0.212776</td>\n",
       "      <td>0.229215</td>\n",
       "      <td>0.204183</td>\n",
       "      <td>0.312741</td>\n",
       "      <td>0.242728</td>\n",
       "      <td>0.184838</td>\n",
       "      <td>0.051989</td>\n",
       "      <td>0.211497</td>\n",
       "      <td>0.361019</td>\n",
       "      <td>0.556714</td>\n",
       "      <td>0.288326</td>\n",
       "      <td>0.368442</td>\n",
       "      <td>0.331639</td>\n",
       "      <td>0.302346</td>\n",
       "      <td>0.236594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    n  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean  Word2Vec:Wikipedia,Size=300,Max  Topic Models:NMF,Full,Topics=20  Topic Models:NMF,Full,Topics=50  Topic Models:NMF,Full,Topics=100  Topic Models:NMF,Full,Topics=200  N-Grams:Full,Words,1-grams  N-Grams:Full,Words,1-grams,Binary  N-Grams:Full,Words,1-grams,TFIDF  N-Grams:Full,Words,1-grams,Binary,TFIDF  N-Grams:Full,Nouns,Adjectives,1-grams  N-Grams:Linares_Pontes,Words,1-grams  N-Grams:Full,Precise_Annotations,Words,1-grams  N-Grams:Full,Partial_Annotations,Words,1-grams  NOBLE Coder:Precise  NOBLE Coder:Partial  NOBLE Coder:Precise,TFIDF  NOBLE Coder:Partial,TFIDF   GO:None   PO:None  Doc2Vec Phenes:Wikipedia,Size=300  Word2Vec Phenes:Wikipedia,Size=300,Mean  Word2Vec Phenes:Wikipedia,Size=300,Max  N-Grams Phenes:Linares_Pontes,Words,1-grams      Mean\n",
       "0  20                    0.111449                          0.218015                         0.341037                         0.544834                         0.188647                          0.183020                          0.152741                    0.119528                           0.088996                          0.099786                                 0.089147                               0.145943                              0.276095                                        0.283791                                        0.412045             0.198858             0.142067                   0.016717                   0.109463  0.319382  0.564720                           0.186704                                 0.279598                                0.315499                                     0.265143  0.197958\n",
       "1  24                    0.120581                          0.225574                         0.335043                         0.512995                         0.240044                          0.217882                          0.170117                    0.133973                           0.096264                          0.104005                                 0.093017                               0.161037                              0.278569                                        0.235299                                        0.410044             0.207532             0.145981                   0.016717                   0.125490  0.330643  0.570454                           0.230586                                 0.292356                                0.335228                                     0.274763  0.222490\n",
       "2  28                    0.130072                          0.230533                         0.329656                         0.515729                         0.290730                          0.245884                          0.178295                    0.148342                           0.101041                          0.110101                                 0.096648                               0.159462                              0.263483                                        0.237463                                        0.322416             0.218477             0.152849                   0.016717                   0.136101  0.332342  0.577437                           0.242324                                 0.317798                                0.316644                                     0.280276  0.212201\n",
       "3  32                    0.134597                          0.225221                         0.313976                         0.502566                         0.346565                          0.272363                          0.207765                    0.158469                           0.103876                          0.113537                                 0.101670                               0.167928                              0.262828                                        0.234073                                        0.371505             0.220331             0.164366                   0.016717                   0.163207  0.338161  0.588521                           0.246228                                 0.335157                                0.289350                                     0.289456  0.210839\n",
       "4  36                    0.136947                          0.219404                         0.298411                         0.501666                         0.370164                          0.289452                          0.222839                    0.163598                           0.111255                          0.123295                                 0.107448                               0.177537                              0.261080                                        0.233573                                        0.363496             0.222286             0.166248                   0.016717                   0.171188  0.344131  0.589310                           0.254740                                 0.327207                                0.296632                                     0.293667  0.208027\n",
       "5  40                    0.139424                          0.229064                         0.284544                         0.504362                         0.414206                          0.308702                          0.230935                    0.169062                           0.113839                          0.128191                                 0.116499                               0.188847                              0.262360                                        0.223171                                        0.350074             0.226538             0.169389                   0.016717                   0.184442  0.349342  0.602143                           0.265526                                 0.327124                                0.304860                                     0.290711  0.219085\n",
       "6  44                    0.142921                          0.227454                         0.219824                         0.509531                         0.452335                          0.330700                          0.246308                    0.171892                           0.116414                          0.134430                                 0.118322                               0.197694                              0.264310                                        0.200789                                        0.347361             0.230504             0.174017                   0.016717                   0.193577  0.351535  0.602000                           0.272019                                 0.335583                                0.310045                                     0.289071  0.232856\n",
       "7  48                    0.140025                          0.232583                         0.219534                         0.512629                         0.482208                          0.356072                          0.258147                    0.173398                           0.119572                          0.140521                                 0.122365                               0.201090                              0.246159                                        0.204340                                        0.343282             0.235898             0.176611                   0.016717                   0.199486  0.355267  0.606176                           0.278888                                 0.350051                                0.311840                                     0.292557  0.233261\n",
       "8  52                    0.142208                          0.241484                         0.222807                         0.516074                         0.478101                          0.369188                          0.268344                    0.177673                           0.120130                          0.146656                                 0.125271                               0.209408                              0.249790                                        0.207168                                        0.318069             0.233944             0.180250                   0.018367                   0.210171  0.358777  0.598570                           0.279232                                 0.357229                                0.327182                                     0.290027  0.238097\n",
       "9  56                    0.145246                          0.248364                         0.217597                         0.512095                         0.481746                          0.388429                          0.274894                    0.180903                           0.128455                          0.150693                                 0.128101                               0.212776                              0.229215                                        0.204183                                        0.312741             0.242728             0.184838                   0.051989                   0.211497  0.361019  0.556714                           0.288326                                 0.368442                                0.331639                                     0.302346  0.236594"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics.cluster import silhouette_score\n",
    "# Note that homogeneity scores don't fit for evaluating how close the clustering is to pathway membership, etc.\n",
    "# This is because genes can be assigned to more than one pathway, metric would have to be changed to account for this.\n",
    "# So all this section does is determines which values of n_clusters provide good clustering results for each matrix.\n",
    "n_clusters_silhouette_scores = defaultdict(dict)\n",
    "min_n_clusters = 20\n",
    "max_n_clusters = 80\n",
    "step_size = 4\n",
    "number_of_clusters = np.arange(min_n_clusters, max_n_clusters, step_size)\n",
    "for n in number_of_clusters:\n",
    "    for name in names:\n",
    "        distance_matrix = name_to_array[name]\n",
    "        #to_id = array_index_to_id\n",
    "        ac = AgglomerativeClustering(n_clusters=n, linkage=\"complete\", affinity=\"precomputed\")\n",
    "        clustering = ac.fit(distance_matrix)\n",
    "        sil_score = silhouette_score(distance_matrix, clustering.labels_, metric=\"precomputed\")\n",
    "        n_clusters_silhouette_scores[name][n] = sil_score\n",
    "sil_df = pd.DataFrame(n_clusters_silhouette_scores).reset_index(drop=False).rename({\"index\":\"n\"},axis=\"columns\")\n",
    "sil_df.to_csv(os.path.join(OUTPUT_DIR,\"part_6_silhouette_scores_by_n.csv\"), index=False)\n",
    "sil_df.head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
