{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "\n",
    "- [Introduction](#introduction)\n",
    "\n",
    "- [Links of Interest](#links)\n",
    "\n",
    "- [Part 1. Loading and Filtering Data](#part_1)\n",
    "    - [Setting input and output paths](#paths)\n",
    "    - [Reading in datasets of phenotype descriptions](#read_this_data)\n",
    "    - [Reading in datasets of groupings](#read_other_data) (biochemical pathways, functional groups, etc.)\n",
    "    - [Relating the datasets](#relating)\n",
    "    - [Filtering the datasets](#filtering)\n",
    "    \n",
    "- [Part 2. NLP Models](#part_2)\n",
    "    - [Word2Vec and Doc2Vec](#word2vec_doc2vec)\n",
    "    - [BERT and BioBERT](#bert_biobert)\n",
    "    - [Loading models](#load_models)\n",
    "\n",
    "- [Part 3. NLP Choices](#part_3)\n",
    "    - [Preprocessing descriptions](#preprocessing)\n",
    "    - [POS Tagging](#pos_tagging)\n",
    "    - [Reducing vocabulary size](#vocab)\n",
    "    - [Annotating with biological ontologies](#annotation)\n",
    "    - [Splitting into phene descriptions](#phenes)\n",
    "    \n",
    "- [Part 4. Generating Vectors and Distance Matrices](#part_4)\n",
    "    - [Defining methods to use](#methods)\n",
    "    - [Running all methods](#running)\n",
    "    - [Merging distances into an edgelist](#merging)\n",
    "    - [Adding edge information](#merging)\n",
    "\n",
    "- [Part 5. Clustering Analysis](#part_5)\n",
    "    - [Topic modeling](#topic_modeling)\n",
    "    - [Agglomerative clustering](#clustering)\n",
    "    - [Phenologs for OMIM disease phenotypes](#phenologs)\n",
    "    \n",
    "- [Part 6. Supervised Tasks](#part_6)\n",
    "    - [Distributions of distance values](#ks)\n",
    "    - [Within-group distance values](#within)\n",
    "    - [Predictions and AUC for shared pathways or interactions](#auc)\n",
    "    - [Tests for querying to recover related genes](#y)\n",
    "    - [Producing output summary table](#output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"introduction\"></a>\n",
    "### Introduction: Text Mining Analysis of Phenotype Descriptions in Plants\n",
    "The purpose of this notebook is to evaluate what can be learned from a natural language processing approach to analyzing free-text descriptions of phenotype descriptions of plants. The approach is to generate pairwise distances matrices between a set of plant phenotype descriptions across different species, sourced from academic papers and online model organism databases. These pairwise distance matrices can be constructed using any vectorization method that can be applied to natural language. In this notebook, we specifically evaluate the use of n-gram and bag-of-words techniques, word and document embedding using Word2Vec and Doc2Vec, context-dependent word-embeddings using BERT and BioBERT, and ontology term annotations with automated annotation tools such as NOBLE Coder.\n",
    "\n",
    "Loading, manipulation, and filtering of the dataset of phenotype descriptions associated with genes across different plant species is largely handled through a Python package created for this purpose called OATS (Ontology Annotation and Text Similarity) which is available [here](https://github.com/irbraun/oats). Preprocessing of the descriptions, mapping the dataset to additional resources such as protein-protein interaction databases and biochemical pathway databases are handled in this notebook using that package as well. In the evaluation of each of these natural language processing approaches to analyzing this dataset of descriptions, we compare performance against a dataset generated through manual annotation of a similar dataset in Oellrich Walls et al. (2015) and against manual annotations with experimentally determined terms from the Gene Ontology (PO) and the Plant Ontology (PO).\n",
    "\n",
    "<a id=\"links\"></a>\n",
    "### Relevant links of interest:\n",
    "- Paper describing comparison of NLP and ontology annotation approaches to curation: [Braun, Lawrence-Dill (2019)](https://doi.org/10.3389/fpls.2019.01629)\n",
    "- Paper describing results of manual phenotype description curation: [Oellrich, Walls et al. (2015](https://plantmethods.biomedcentral.com/articles/10.1186/s13007-015-0053-y)\n",
    "- Plant databases with phenotype description text data available: [TAIR](https://www.arabidopsis.org/), [SGN](https://solgenomics.net/), [MaizeGDB](https://www.maizegdb.org/)\n",
    "- Python package for working with phenotype descriptions: [OATS](https://github.com/irbraun/oats)\n",
    "- Python package used for general NLP functions: [NLTK](https://www.nltk.org/), [Gensim](https://radimrehurek.com/gensim/auto_examples/index.html)\n",
    "- Python package used for working with biological ontologies: [Pronto](https://pronto.readthedocs.io/en/latest/)\n",
    "- Python package for loading pretrained BERT models: [PyTorch Pretrained BERT](https://pypi.org/project/pytorch-pretrained-bert/)\n",
    "- For BERT Models pretrained on PubMed and PMC: [BioBERT Paper](https://arxiv.org/abs/1901.08746), [BioBERT Models](https://github.com/naver/biobert-pretrained)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "import torch\n",
    "import itertools\n",
    "import multiprocessing as mp\n",
    "from collections import Counter, defaultdict\n",
    "from inspect import signature\n",
    "from scipy.stats import ks_2samp, hypergeom\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, auc\n",
    "from sklearn.model_selection import train_test_split, KFold\n",
    "from scipy import spatial, stats\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from pytorch_pretrained_bert import BertTokenizer, BertModel, BertForMaskedLM\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer\n",
    "from sklearn.decomposition import NMF\n",
    "from sklearn.decomposition import LatentDirichletAllocation as LDA\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.datasets.dataset import Dataset\n",
    "from oats.datasets.groupings import Groupings\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.datasets.string import String\n",
    "from oats.datasets.edges import Edges\n",
    "from oats.annotation.annotation import annotate_using_noble_coder\n",
    "from oats.graphs import pairwise as pw\n",
    "from oats.graphs.editing import merge_edgelists, make_undirected, remove_self_loops, subset_edgelist_with_ids\n",
    "from oats.graphs.indexed import IndexedGraph\n",
    "from oats.graphs.weighting import train_logistic_regression_model, apply_logistic_regression_model\n",
    "from oats.graphs.weighting import train_random_forest_model, apply_random_forest_model\n",
    "from oats.nlp.vocabulary import get_overrepresented_tokens, get_vocabulary_from_tokens\n",
    "from oats.nlp.vocabulary import reduce_vocabulary_connected_components, reduce_vocabulary_linares_pontes\n",
    "from oats.utils.utils import function_wrapper_with_duration\n",
    "from oats.nlp.preprocess import concatenate_with_bar_delim\n",
    "\n",
    "from _utils import Method\n",
    "\n",
    "mpl.rcParams[\"figure.dpi\"] = 400\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_1\"></a>\n",
    "# Part 1. Loading and Filtering Data\n",
    "<a id=\"paths\"></a>\n",
    "### Setting up the input and output paths and summarizing output table\n",
    "This section defines some constants which are used for creating a uniquely named directory to contain all the outputs from running this instance of this notebook. The naming scheme is based on the time that the notebook is run. The other constants are used for specifying information in the output table about what the topic was for this notebook when it was run, such as looking at KEGG biochemical pathways or STRING protein-protein interaction data some other type of gene function grouping or hierarchy. These values are arbitrary and are just for keeping better notes about what the output of the notebook corresponds to. All the input and output file paths for loading datasets or models are also contained within this cell, so that if anything is moved the directories and file names should only have to be changed at this point and nowhere else further into the notebook. If additional files are added to the notebook cells they should be put here as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# The summarizing output dictionary has the shape TABLE[method][metric] --> value.\n",
    "NOTEBOOK_TAGS = {\"kegg\":False, \"pmn\":False, \"classes\":False, \"subsets\":True, \"training\":False}\n",
    "TOPIC = \"Biochemical Pathways\"\n",
    "DATA = \"Filtered\"\n",
    "TABLE = defaultdict(dict)\n",
    "OUTPUT_DIR = os.path.join(\"../outputs\",datetime.datetime.now().strftime('%m_%d_%Y_h%Hm%Ms%S'))\n",
    "os.mkdir(OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "testing_dataset_filename = \"../data/pickles/text_plus_annotations_dataset.pickle\"                    # The full dataset pickle.\n",
    "training_dataset_filename = \"../data/pickles/training_dataset.pickle\"                                # The full dataset pickle.\n",
    "kegg_pathways_filename = \"../data/pickles/kegg_pathways.pickle\"                                      # The pathway groupings from KEGG.\n",
    "pmn_pathways_filename = \"../data/pickles/pmn_pathways.pickle\"                                        # The pahway groupings from Plant Metabolic Network.\n",
    "lloyd_subsets_filename = \"../data/pickles/lloyd_subsets.pickle\"                                      # The functional subsets defined by Lloyd and Meinke (2012).\n",
    "lloyd_classes_filename = \"../data/pickles/lloyd_classes.pickle\"                                      # The functional classes defined by Lloyd and Meinke (2012).\n",
    "background_corpus_filename = \"../data/corpus_related_files/untagged_text_corpora/background.txt\"     # Text file with background content.\n",
    "phenotypes_corpus_filename = \"../data/corpus_related_files/untagged_text_corpora/phenotypes_all.txt\" # Text file with specific content.\n",
    "doc2vec_pubmed_filename = \"../gensim/pubmed_dbow/doc2vec_2.bin\"                                      # File holding saved Doc2Vec model.\n",
    "doc2vec_wikipedia_filename = \"../gensim/enwiki_dbow/doc2vec.bin\"                                     # File holding saved Doc2Vec model.\n",
    "word2vec_model_filename = \"../gensim/wiki_sg/word2vec.bin\"                                           # File holding saved Word2Vec model.\n",
    "ontology_filename = \"../ontologies/mo.obo\"                                                           # Ontology file in OBO format.\n",
    "noblecoder_jarfile_path = \"../lib/NobleCoder-1.0.jar\"                                                # Jar for NOBLE Coder tool.\n",
    "biobert_pmc_path = \"../gensim/biobert_v1.0_pmc/pytorch_model\"                                        # Path for PyTorch BioBERT model.\n",
    "biobert_pubmed_path = \"../gensim/biobert_v1.0_pubmed/pytorch_model\"                                  # Path for PyTorch BioBERT model.\n",
    "biobert_pubmed_pmc_path = \"../gensim/biobert_v1.0_pubmed_pmc/pytorch_model\"                          # Path for PyTorch BioBERT model.\n",
    "panther_to_omim_filename = \"../data/orthology_related_files/pantherdb_omim_df.csv\"                   # File with mappings to human orthologs and disease phenotypes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"read_this_data\"></a>\n",
    "### Reading in the dataset of genes and their associated phenotype descriptions and annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ath</td>\n",
       "      <td>At3g49600|UBP26|AT3G49600|SUP32|ATUBP26|ubiqui...</td>\n",
       "      <td>50% defective seeds. Low penetrance of endospe...</td>\n",
       "      <td>GO:0005730|GO:0048316|PO:0000013|PO:0000037|PO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4346</th>\n",
       "      <td>4346</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74380|XXT5|xyloglucan xylosyltransferase 5...</td>\n",
       "      <td>Abnormal roothairs. Reduction in xyloglucan le...</td>\n",
       "      <td>GO:0005794|GO:0048767|GO:0005515|GO:0000139|GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4347</th>\n",
       "      <td>4347</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74450|AT1G74450.1|F1M20.13|F1M20_13</td>\n",
       "      <td>No visible phenotype.</td>\n",
       "      <td>GO:0003674|GO:0008150|PO:0000013|PO:0000037|PO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4348</th>\n",
       "      <td>4348</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74560|AT2G03440|NRP1|NAP1-related protein ...</td>\n",
       "      <td>mutants did not show any phenotype under in vi...</td>\n",
       "      <td>GO:0005634|GO:0005829|GO:0046686|GO:0003682|GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4349</th>\n",
       "      <td>4349</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74660|MIF1|mini zinc finger 1|F1M20.34|F1M...</td>\n",
       "      <td>Constitutive overexpression of MIF1 caused dra...</td>\n",
       "      <td>GO:0048509|GO:0045892|GO:0009640|GO:0003677|GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4350</th>\n",
       "      <td>4350</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74730|RIQ2|F25A4.30|F25A4_30</td>\n",
       "      <td>Reduced NPQ, affected organization of light-ha...</td>\n",
       "      <td>GO:0009535|GO:0009534|GO:0003674|GO:0009507|GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4351</th>\n",
       "      <td>4351</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74740|CPK30|CDPK1A|ATCPK30|calcium-depende...</td>\n",
       "      <td>Embryo lethality of cpk10 cpk30 double mutant ...</td>\n",
       "      <td>GO:0005515|GO:0005886|PO:0000013|PO:0000037|PO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4352</th>\n",
       "      <td>4352</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G74910|KJC1|KONJAC 1|F25A4.12|F25A4_12|AT1G...</td>\n",
       "      <td>Reduced levels of GDP-Man. Severe dwarf, small...</td>\n",
       "      <td>GO:0005829|GO:0005777|GO:0046686|PO:0000013|PO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4353</th>\n",
       "      <td>4353</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G75080|BZR1|BRASSINAZOLE-RESISTANT 1|F9E10....</td>\n",
       "      <td>Insensitive to brassinazole (BRZ), an inhibito...</td>\n",
       "      <td>GO:0045892|GO:0048481|GO:0003700|GO:0005515|GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4354</th>\n",
       "      <td>4354</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G75520|SRS5|SHI-related sequence 5|F1B16.17</td>\n",
       "      <td>18-25% of flowers have homeotic conversion pet...</td>\n",
       "      <td>GO:0048467|PO:0000037|PO:0009009|PO:0009010|PO...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id species                                         gene_names                                        description                                           term_ids\n",
       "1        1     ath  At3g49600|UBP26|AT3G49600|SUP32|ATUBP26|ubiqui...  50% defective seeds. Low penetrance of endospe...  GO:0005730|GO:0048316|PO:0000013|PO:0000037|PO...\n",
       "4346  4346     ath  AT1G74380|XXT5|xyloglucan xylosyltransferase 5...  Abnormal roothairs. Reduction in xyloglucan le...  GO:0005794|GO:0048767|GO:0005515|GO:0000139|GO...\n",
       "4347  4347     ath            AT1G74450|AT1G74450.1|F1M20.13|F1M20_13                              No visible phenotype.  GO:0003674|GO:0008150|PO:0000013|PO:0000037|PO...\n",
       "4348  4348     ath  AT1G74560|AT2G03440|NRP1|NAP1-related protein ...  mutants did not show any phenotype under in vi...  GO:0005634|GO:0005829|GO:0046686|GO:0003682|GO...\n",
       "4349  4349     ath  AT1G74660|MIF1|mini zinc finger 1|F1M20.34|F1M...  Constitutive overexpression of MIF1 caused dra...  GO:0048509|GO:0045892|GO:0009640|GO:0003677|GO...\n",
       "4350  4350     ath                   AT1G74730|RIQ2|F25A4.30|F25A4_30  Reduced NPQ, affected organization of light-ha...  GO:0009535|GO:0009534|GO:0003674|GO:0009507|GO...\n",
       "4351  4351     ath  AT1G74740|CPK30|CDPK1A|ATCPK30|calcium-depende...  Embryo lethality of cpk10 cpk30 double mutant ...  GO:0005515|GO:0005886|PO:0000013|PO:0000037|PO...\n",
       "4352  4352     ath  AT1G74910|KJC1|KONJAC 1|F25A4.12|F25A4_12|AT1G...  Reduced levels of GDP-Man. Severe dwarf, small...  GO:0005829|GO:0005777|GO:0046686|PO:0000013|PO...\n",
       "4353  4353     ath  AT1G75080|BZR1|BRASSINAZOLE-RESISTANT 1|F9E10....  Insensitive to brassinazole (BRZ), an inhibito...  GO:0045892|GO:0048481|GO:0003700|GO:0005515|GO...\n",
       "4354  4354     ath     AT1G75520|SRS5|SHI-related sequence 5|F1B16.17  18-25% of flowers have homeotic conversion pet...  GO:0048467|PO:0000037|PO:0009009|PO:0009010|PO..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_filename = training_dataset_filename if NOTEBOOK_TAGS[\"training\"] else testing_dataset_filename\n",
    "dataset = load_from_pickle(dataset_filename)\n",
    "dataset.describe()\n",
    "dataset.filter_by_species(\"ath\",\"zma\",\"sly\")\n",
    "dataset.filter_has_description()\n",
    "dataset.describe()\n",
    "if not NOTEBOOK_TAGS[\"training\"]: dataset.filter_has_annotation()\n",
    "if not NOTEBOOK_TAGS[\"training\"]: dataset.filter_has_annotation(\"GO\")\n",
    "if not NOTEBOOK_TAGS[\"training\"]: dataset.filter_has_annotation(\"PO\")\n",
    "dataset.describe()\n",
    "dataset.to_pandas().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"read_other_data\"></a>\n",
    "### Reading in the dataset of groupings from KEGG, Plant Metabolic Network, and other hierarchies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>group_id</th>\n",
       "      <th>gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>FSM</td>\n",
       "      <td>at1g01030|nga3|top1|ngatha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>EMB|FSM|OVP|SRF</td>\n",
       "      <td>at1g01040|sus1|dcl1|sin1|caf|abnormal suspensor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>CDR|LIT</td>\n",
       "      <td>at1g01060|lhy|late elongated hypocotyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>IST|WAT</td>\n",
       "      <td>at1g01120|kcs1|3-ketoacyl-coa synthase defective</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>OVP|SRF</td>\n",
       "      <td>at1g01280|cyp703a2|cytochrome p450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>EMB</td>\n",
       "      <td>at1g01370|cenh3|centromere-specific histone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>CHS</td>\n",
       "      <td>at1g01460|pipk11|phosphatidylinositol phosphat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>NLS|GRS|IST</td>\n",
       "      <td>at1g01480|acs2|aminocyclopropane carboxylate s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>LEF|FSM</td>\n",
       "      <td>at1g01510|an|angustifolia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>SRL|ROT|LEF|MSL|STT|RTH|TCM|TMP</td>\n",
       "      <td>at1g01550|bps1|bypass</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species                         group_id                                         gene_names\n",
       "0     ath                              FSM                         at1g01030|nga3|top1|ngatha\n",
       "1     ath                  EMB|FSM|OVP|SRF    at1g01040|sus1|dcl1|sin1|caf|abnormal suspensor\n",
       "2     ath                          CDR|LIT             at1g01060|lhy|late elongated hypocotyl\n",
       "3     ath                          IST|WAT   at1g01120|kcs1|3-ketoacyl-coa synthase defective\n",
       "4     ath                          OVP|SRF                 at1g01280|cyp703a2|cytochrome p450\n",
       "5     ath                              EMB        at1g01370|cenh3|centromere-specific histone\n",
       "6     ath                              CHS  at1g01460|pipk11|phosphatidylinositol phosphat...\n",
       "7     ath                      NLS|GRS|IST  at1g01480|acs2|aminocyclopropane carboxylate s...\n",
       "8     ath                          LEF|FSM                          at1g01510|an|angustifolia\n",
       "9     ath  SRL|ROT|LEF|MSL|STT|RTH|TCM|TMP                              at1g01550|bps1|bypass"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupings_filename = \"\"\n",
    "groupings_filename = kegg_pathways_filename if NOTEBOOK_TAGS[\"kegg\"] else groupings_filename\n",
    "groupings_filename = pmn_pathways_filename if NOTEBOOK_TAGS[\"pmn\"] else groupings_filename\n",
    "groupings_filename = lloyd_subsets_filename if NOTEBOOK_TAGS[\"subsets\"] else groupings_filename\n",
    "groupings_filename = lloyd_classes_filename if NOTEBOOK_TAGS[\"classes\"] else groupings_filename\n",
    "groups = load_from_pickle(groupings_filename)\n",
    "id_to_group_ids = groups.get_id_to_group_ids_dict(dataset.get_gene_dictionary())\n",
    "group_id_to_ids = groups.get_group_id_to_ids_dict(dataset.get_gene_dictionary())\n",
    "group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "groups.to_csv(os.path.join(OUTPUT_DIR,\"part_1_groupings.csv\"))\n",
    "groups.to_pandas().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"relating\"></a>\n",
    "### Relating the dataset of genes to the dataset of groupings or categories\n",
    "This section generates tables that indicate how the genes present in the dataset were mapped to the defined pathways or groups. This includes a summary table that indicates how many genes by species were succcessfully mapped to atleast one pathway or group, as well as a more detailed table describing how many genes from each species were mapped to each particular pathway or group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate a table describing how many of the genes input from each species map to atleast one group.\n",
    "summary = defaultdict(dict)\n",
    "species_dict = dataset.get_species_dictionary()\n",
    "for species in dataset.get_species():\n",
    "    summary[species][\"input\"] = len([x for x in dataset.get_ids() if species_dict[x]==species])\n",
    "    summary[species][\"mapped\"] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "table = pd.DataFrame(summary).transpose()\n",
    "table.loc[\"total\"]= table.sum()\n",
    "table[\"fraction\"] = table.apply(lambda row: \"{:0.4f}\".format(row[\"mapped\"]/row[\"input\"]), axis=1)\n",
    "table = table.reset_index(inplace=False)\n",
    "table = table.rename({\"index\":\"species\"}, axis=\"columns\")\n",
    "table.to_csv(os.path.join(OUTPUT_DIR,\"part_1_mappings_summary.csv\"), index=False)\n",
    "\n",
    "# Generate a table describing how many genes from each species map to which particular group.\n",
    "summary = defaultdict(dict)\n",
    "for group_id,ids in group_id_to_ids.items():\n",
    "    summary[group_id].update({species:len([x for x in ids if species_dict[x]==species]) for species in dataset.get_species()})\n",
    "    summary[group_id][\"total\"] = len([x for x in ids])\n",
    "table = pd.DataFrame(summary).transpose()\n",
    "table = table.sort_values(by=\"total\", ascending=False)\n",
    "table = table.reset_index(inplace=False)\n",
    "table = table.rename({\"index\":\"pathway_id\"}, axis=\"columns\")\n",
    "table[\"pathway_name\"] = table[\"pathway_id\"].map(groups.get_long_name)\n",
    "table.loc[\"total\"] = table.sum()\n",
    "table.loc[\"total\",\"pathway_id\"] = \"total\"\n",
    "table.loc[\"total\",\"pathway_name\"] = \"total\"\n",
    "table = table[table.columns.tolist()[-1:] + table.columns.tolist()[:-1]]\n",
    "table.to_csv(os.path.join(OUTPUT_DIR,\"part_1_mappings_by_group.csv\"), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"filtering\"></a>\n",
    "### Option 1: Filtering the dataset based on presence in the curated Oellrich, Walls et al. (2015) dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>1899</td>\n",
       "      <td>1692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>total</td>\n",
       "      <td>1899</td>\n",
       "      <td>1692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath       1899                 1692\n",
       "1   total       1899                 1692"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter the dataset based on whether or not the genes were in the curated dataset.\n",
    "# This is similar to filtering based on protein interaction data because the dataset is a list of edge values.\n",
    "pppn_edgelist_path = \"../data/supplemental_files_oellrich_walls/13007_2015_53_MOESM9_ESM.txt\"\n",
    "pppn_edgelist = Edges(dataset.get_name_to_id_dictionary(), pppn_edgelist_path)\n",
    "if not NOTEBOOK_TAGS[\"training\"]: dataset.filter_with_ids(pppn_edgelist.ids)\n",
    "dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 2: Filtering the dataset based on protein-protein interactions\n",
    "This is done to only include genes (and the corresponding phenotype descriptions and annotations) which are useful for the current analysis. In this case we want to only retain genes that are mentioned atleast one time in the STRING database for a given species. If a gene is not mentioned at all in STRING, there is no information available for whether or not it interacts with any other proteins in the dataset so choose to not include it in the analysis. Only genes that have atleast one true positive are included because these are the only ones for which the missing information (negatives) is meaningful. This should be run instead of the subsequent cell, or the other way around, based on whether or not protein-protein interactions is the prediction goal for the current analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = dataset.get_gene_dictionary()\n",
    "string_data = String(genes, naming_file, *interaction_files)\n",
    "dataset.filter_with_ids(string_data.ids)\n",
    "dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 3: Filtering the dataset based on membership in pathways or phenotype category\n",
    "This is done to only include genes (and the corresponding phenotype descriptions and annotations) which are useful for the current analysis. In this case we want to only retain genes that are mapped to atleast one pathway in whatever the source of pathway membership we are using is (KEGG, Plant Metabolic Network, etc). This is because for these genes, it will be impossible to correctly predict their pathway membership, and we have no evidence that they belong or do not belong in certain pathways so they can not be identified as being true or false negatives in any case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Filter based on succcessful mappings to groups or pathways.\n",
    "dataset.filter_with_ids(group_mapped_ids)\n",
    "dataset.describe()\n",
    "# Get the mappings in each direction again now that the dataset has been subset.\n",
    "id_to_group_ids = groups.get_id_to_group_ids_dict(dataset.get_gene_dictionary())\n",
    "group_id_to_ids = groups.get_group_id_to_ids_dict(dataset.get_gene_dictionary())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_2\"></a>\n",
    "# Part 2. NLP Models\n",
    "\n",
    "\n",
    "<a id=\"word2vec_doc2vec\"></a>\n",
    "### Word2Vec and Doc2Vec\n",
    "Word2Vec is a word embedding technique using a neural network trained on a so-called *false task*, namely either predicting a missing word from within a sequence of context words drawn from a sentence or phrase, or predicting which contexts words surround some given input word drawn from a sentence or phrase. Each of these tasks are supervised (the correct answer is fixed and known), but can be generated from unlabelled text data such as a collection of books or wikipedia articles, meaning that even though the task itself is supervised the training data can be generated automatically, enabling the creation of enormous training sets. The internal representation for particular words learned during the training process contain semantically informative features related to that given word, and can therefore be used as embeddings used downstream for tasks such as finding similarity between words or as input into additional models. Doc2Vec is an extension of this technique that determines vector embeddings for entire documents (strings containing multiple words, could be sentences, paragraphs, or documents).\n",
    "\n",
    "\n",
    "<a id=\"bert_biobert\"></a>\n",
    "### BERT and BioBERT\n",
    "BERT ('Bidirectional Encoder Representations from Transformers') is another neueral network-based model trained on two different false tasks, namely predicting the subsequent sentence given some input sentence, or predicting the identity of a set of words masked from an input sentence. Like Word2Vec, this architecture can be used to generate vector embeddings for a particular input word by extracting values from a subset of the encoder layers that correspond to that input word. Practically, a major difference is that because the input word is input in the context of its surrounding sentence, the embedding reflects the meaning of a particular word in a particular context (such as the difference in the meaning of *root* in the phrases *plant root* and *root of the problem*. BioBERT refers to a set of BERT models which have been finetuned on the PubMed and PMC corpora. See the list of relevant links for the publications and pages associated with these models.\n",
    "\n",
    "<a id=\"load_models\"></a>\n",
    "### Loading trained and saved models\n",
    "Versions of the architectures discussed above which have been saved as trained models are loaded here. Some of these models are loaded as pretrained models from the work of other groups, and some were trained on data specific to this notebook and loaded here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Files and models related to the machine learning text embedding methods used here.\n",
    "doc2vec_wiki_model = gensim.models.Doc2Vec.load(doc2vec_wikipedia_filename)\n",
    "doc2vec_pubmed_model = gensim.models.Doc2Vec.load(doc2vec_pubmed_filename)\n",
    "word2vec_model = gensim.models.Word2Vec.load(word2vec_model_filename)\n",
    "bert_tokenizer_base = BertTokenizer.from_pretrained('bert-base-uncased')\n",
    "bert_tokenizer_pmc = BertTokenizer.from_pretrained(biobert_pmc_path)\n",
    "bert_tokenizer_pubmed = BertTokenizer.from_pretrained(biobert_pubmed_path)\n",
    "bert_tokenizer_pubmed_pmc = BertTokenizer.from_pretrained(biobert_pubmed_pmc_path)\n",
    "bert_model_base = BertModel.from_pretrained('bert-base-uncased')\n",
    "bert_model_pmc = BertModel.from_pretrained(biobert_pmc_path)\n",
    "bert_model_pubmed = BertModel.from_pretrained(biobert_pubmed_path)\n",
    "bert_model_pubmed_pmc = BertModel.from_pretrained(biobert_pubmed_pmc_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_3\"></a>\n",
    "# Part 3. NLP Choices\n",
    "\n",
    "<a id=\"preprocessing\"></a>\n",
    "### Preprocessing text descriptions\n",
    "The preprocessing methods applied to the phenotype descriptions are a choice which impacts the subsequent vectorization and similarity methods which construct the pairwise distance matrix from each of these descriptions. The preprocessing methods that make sense are also highly dependent on the vectorization method or embedding method that is to be applied. For example, stemming (which is part of the full proprocessing done below using the Gensim preprocessing function) is useful for the n-grams and bag-of-words methods but not for the document embeddings methods which need each token to be in the vocabulary that was constructed and used when the model was trained. For this reason, embedding methods with pretrained models where the vocabulary is fixed should have a lighter degree of preprocessing not involving stemming or lemmatization but should involve things like removal of non-alphanumerics and normalizing case. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Obtain a mapping between IDs and the raw text descriptions associated with that ID from the dataset.\n",
    "descriptions = dataset.get_description_dictionary()\n",
    "\n",
    "# Preprocessing of the text descriptions. Different methods are necessary for different approaches.\n",
    "descriptions_full_preprocessing = {i:\" \".join(preprocess_string(d)) for i,d in descriptions.items()}\n",
    "descriptions_simple_preprocessing = {i:\" \".join(simple_preprocess(d)) for i,d in descriptions.items()}\n",
    "descriptions_no_stopwords = {i:remove_stopwords(d) for i,d in descriptions.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"pos_tagging\"></a>\n",
    "### POS tagging the phenotype descriptions for nouns and adjectives\n",
    "Note that preprocessing of the descriptions should be done after part-of-speech tagging, because tokens that are removed during preprocessing before n-gram analysis contain information that the parser needs to accurately call parts-of-speech. This step should be done on the raw descriptions and then the resulting bags of words can be subset using additional preprocesssing steps before input in one of the vectorization methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "get_pos_tokens = lambda text,pos: \" \".join([t[0] for t in nltk.pos_tag(word_tokenize(text)) if t[1].lower()==pos.lower()])\n",
    "descriptions_noun_only =  {i:get_pos_tokens(d,\"NN\") for i,d in descriptions.items()}\n",
    "descriptions_noun_only_full_preprocessing = {i:\" \".join(preprocess_string(d)) for i,d in descriptions_noun_only.items()}\n",
    "descriptions_noun_only_simple_preprocessing = {i:\" \".join(simple_preprocess(d)) for i,d in descriptions_noun_only.items()}\n",
    "descriptions_adj_only =  {i:get_pos_tokens(d,\"JJ\") for i,d in descriptions.items()}\n",
    "descriptions_adj_only_full_preprocessing = {i:\" \".join(preprocess_string(d)) for i,d in descriptions_adj_only.items()}\n",
    "descriptions_adj_only_simple_preprocessing = {i:\" \".join(simple_preprocess(d)) for i,d in descriptions_adj_only.items()}\n",
    "descriptions_noun_adj = {i:\"{} {}\".format(descriptions_noun_only[i],descriptions_adj_only[i]) for i in descriptions.keys()}\n",
    "descriptions_noun_adj_full_preprocessing = {i:\"{} {}\".format(descriptions_noun_only_full_preprocessing[i],descriptions_adj_only_full_preprocessing[i]) for i in descriptions.keys()}\n",
    "descriptions_noun_adj_simple_preprocessing = {i:\"{} {}\".format(descriptions_noun_only_simple_preprocessing[i],descriptions_adj_only_simple_preprocessing[i]) for i in descriptions.keys()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"vocab\"></a>\n",
    "### Reducing the vocabulary size using a word distance matrix\n",
    "These approaches for reducing the vocabulary size of the dataset work by replacing multiple words that occur throughout the dataset of descriptions with an identical word that is representative of this larger group of words. The total number of unique words across all descriptions is therefore reduced, and when observing n-gram overlaps between vector representations of these descriptions, overlaps will now occur between descriptions that included different but similar words. These methods work by actually generating versions of these descriptions that have the word replacements present. The returned objects for these methods are the revised description dictionary, a dictionary mapping tokens in the full vocabulary to tokens in the reduced vocabulary, and a dictionary mapping tokens in the reduced vocabulary to a list of tokens in the full vocabulary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Reducing the size of the vocabulary for descriptions treated with simple preprocessing.\n",
    "tokens = list(set([w for w in flatten(d.split() for d in descriptions_simple_preprocessing.values())]))\n",
    "tokens_dict = {i:w for i,w in enumerate(tokens)}\n",
    "graph = pw.pairwise_square_word2vec(word2vec_model, tokens_dict, \"cosine\")\n",
    "\n",
    "# Make sure that the tokens list is in the same order as the indices representing each word in the distance matrix.\n",
    "# This is only trivial here because the IDs used are ordered integers 0 to n, but this might not always be the case.\n",
    "distance_matrix = graph.array\n",
    "tokens = [tokens_dict[graph.index_to_id[index]] for index in np.arange(distance_matrix.shape[0])]\n",
    "n = 3\n",
    "threshold = 0.2\n",
    "descriptions_linares_pontes, reduce_lp, unreduce_lp = reduce_vocabulary_linares_pontes(descriptions_simple_preprocessing, tokens, distance_matrix, n)\n",
    "descriptions_connected_components, reduce_cc, unreduce_cc = reduce_vocabulary_connected_components(descriptions_simple_preprocessing, tokens, distance_matrix, threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reducing vocabulary size based on identifying important words\n",
    "These approcahes for reducing the vocabulary size of the dataset work by identifying which words in the descriptions are likely to be the most important for identifying differences between the phenotypes and meaning of the descriptions. One approach is to determine which words occur at a higher rate in text of interest such as articles about plant phenotypes as compared to their rates in more general texts such as a corpus of news articles. These approaches do not create modified versions of the descriptions but rather provide vocabulary objects that can be passed to the sklearn vectorizer or constructors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Constructing a vocabulary by looking at what words are overrepresented in domain specific text.\n",
    "background_corpus = open(background_corpus_filename,\"r\").read()\n",
    "phenotypes_corpus = open(phenotypes_corpus_filename,\"r\").read()\n",
    "tokens = get_overrepresented_tokens(phenotypes_corpus, background_corpus, max_features=5000)\n",
    "vocabulary_from_text = get_vocabulary_from_tokens(tokens)\n",
    "\n",
    "# Constructing a vocabulary by assuming all words present in a given ontology are important.\n",
    "ontology = Ontology(ontology_filename)\n",
    "vocabulary_from_ontology = get_vocabulary_from_tokens(ontology.get_tokens())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"annotation\"></a>\n",
    "### Annotating descriptions with ontology terms\n",
    "This section generates dictionaries that map gene IDs from the dataset to lists of strings, where those strings are ontology term IDs. How the term IDs are found for each gene entry with its corresponding phenotype description depends on the cell below. Firstly, the terms are found by using the NOBLE Coder annotation tool through these wrapper functions to identify the terms by looking for instances of the term's label or synonyms in the actual text of the phenotype descriptions. Secondly, the next cell just draws the terms directly from the dataset itself. In this case, these are high-confidence annotations done by curators for a comparison against what can be accomplished through computational analysis of the text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Run the ontology term annotators over the raw input text descriptions. NOBLE-Coder handles simple issues like case\n",
    "# normalization so preprocessed descriptions are not used for this step.\n",
    "ontology = Ontology(ontology_filename)\n",
    "annotations_noblecoder_precise = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"mo\", precise=1)\n",
    "annotations_noblecoder_partial = annotate_using_noble_coder(descriptions, noblecoder_jarfile_path, \"mo\", precise=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get the ID to term list annotation dictionaries for each ontology in the dataset.\n",
    "annotations = dataset.get_annotations_dictionary()\n",
    "go_annotations = {k:[term for term in v if term[0:2]==\"GO\"] for k,v in annotations.items()}\n",
    "po_annotations = {k:[term for term in v if term[0:2]==\"PO\"] for k,v in annotations.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"phenes\"></a>\n",
    "### Splitting the descriptions into individual phenes\n",
    "As a preprocessing step, split into a new set of descriptions that's larger. Note that phenotypes are split into phenes, and the phenes that are identical are retained as separate entries in the dataset. This makes the distance matrix calculation more needlessly expensive, because vectors need to be found for the same string more than once, but it simplifies converting the edgelist back to having IDs that reference the genes (full phenotypes) instead of the smaller phenes. If anything, that problem should be addressed in the pairwise functions, not here. (The package should handle it, not when creating input data for those methods)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary of phene descriptions and a dictionary to convert back to the phenotype/gene IDs.\n",
    "phenes = {}\n",
    "phene_id_to_id = {}\n",
    "phene_id = 0\n",
    "for i,phene_list in {i:sent_tokenize(d) for i,d in descriptions.items()}.items():\n",
    "    for phene in phene_list:\n",
    "        phenes[phene_id] = phene\n",
    "        phene_id_to_id[phene_id] = i\n",
    "        phene_id = phene_id+1\n",
    "        \n",
    "# Repeating the reprocessing options for the individual phenes instead of the full phenotype descriptions.\n",
    "phenes_full_preprocessing = {i:\" \".join(preprocess_string(d)) for i,d in phenes.items()}\n",
    "phenes_simple_preprocessing = {i:\" \".join(simple_preprocess(d)) for i,d in phenes.items()}\n",
    "phenes_no_stopwords = {i:remove_stopwords(d) for i,d in phenes.items()}\n",
    "get_pos_tokens = lambda text,pos: \" \".join([t[0] for t in nltk.pos_tag(word_tokenize(text)) if t[1].lower()==pos.lower()])\n",
    "phenes_noun_only =  {i:get_pos_tokens(d,\"NN\") for i,d in phenes.items()}\n",
    "phenes_noun_only_full_preprocessing = {i:\" \".join(preprocess_string(d)) for i,d in phenes_noun_only.items()}\n",
    "phenes_noun_only_simple_preprocessing = {i:\" \".join(simple_preprocess(d)) for i,d in phenes_noun_only.items()}\n",
    "phenes_adj_only =  {i:get_pos_tokens(d,\"JJ\") for i,d in phenes.items()}\n",
    "phenes_adj_only_full_preprocessing = {i:\" \".join(preprocess_string(d)) for i,d in phenes_adj_only.items()}\n",
    "phenes_adj_only_simple_preprocessing = {i:\" \".join(simple_preprocess(d)) for i,d in phenes_adj_only.items()}\n",
    "phenes_noun_adj = {i:\"{} {}\".format(phenes_noun_only[i],phenes_adj_only[i]) for i in phenes.keys()}\n",
    "phenes_noun_adj_full_preprocessing = {i:\"{} {}\".format(phenes_noun_only_full_preprocessing[i],phenes_adj_only_full_preprocessing[i]) for i in phenes.keys()}\n",
    "phenes_noun_adj_simple_preprocessing = {i:\"{} {}\".format(phenes_noun_only_simple_preprocessing[i],phenes_adj_only_simple_preprocessing[i]) for i in phenes.keys()}\n",
    "\n",
    "# Repeating the vocbulary reduction step using the individual phenes instead of full phenotype descriptions.\n",
    "tokens = list(set([w for w in flatten(d.split() for d in phenes_simple_preprocessing.values())]))\n",
    "tokens_dict = {i:w for i,w in enumerate(tokens)}\n",
    "graph = pw.pairwise_square_word2vec(word2vec_model, tokens_dict, \"cosine\")\n",
    "distance_matrix = graph.array\n",
    "tokens = [tokens_dict[graph.index_to_id[index]] for index in np.arange(distance_matrix.shape[0])]\n",
    "n = 3\n",
    "threshold = 0.2\n",
    "phenes_linares_pontes, reduce_lp, unreduce_lp = reduce_vocabulary_linares_pontes(phenes_simple_preprocessing, tokens, distance_matrix, n)\n",
    "phenes_connected_components, reduce_cc, unreduce_cc = reduce_vocabulary_connected_components(phenes_simple_preprocessing, tokens, distance_matrix, threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_4\"></a>\n",
    "# Part 4. Generating vector representations and pairwise distances matrices\n",
    "This section uses the text descriptions, preprocessed text descriptions, or ontology term annotations created or read in the previous sections to generate a vector representation for each gene and build a pairwise distance matrix for the whole dataset. Each method specified is a unique combination of a method of vectorization (bag-of-words, n-grams, document embedding model, etc) and distance metric (Euclidean, Jaccard, cosine, etc) applied to those vectors in constructing the pairwise matrix. The method of vectorization here is equivalent to feature selection, so the task is to figure out which type of vectors will encode features that are useful (n-grams, full words, only words from a certain vocabulary, etc).\n",
    "\n",
    "<a id=\"methods\"></a>\n",
    "### Specifying a list of NLP methods to use\n",
    "Something here if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "training_methods = [\n",
    "\n",
    "\n",
    "    # Full phenotype descriptions\n",
    "\n",
    "\n",
    "    # Methods that use neural networks to generate embeddings.\n",
    "    Method(\"Doc2Vec\", \"Wikipedia,Size=300\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_wiki_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\"}, spatial.distance.cosine),\n",
    "    Method(\"Doc2Vec\", \"PubMed,Size=100\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_pubmed_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Mean\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Max\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Mean,Nouns,Adjectives\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions_noun_adj_simple_preprocessing, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Max,Nouns,Adjectives\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions_noun_adj_simple_preprocessing, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Mean,Linares Pontes\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Max,Linares Pontes\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine),\n",
    "\n",
    "    #Method(\"BERT\", \"Base:Layers=2,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":2}, spatial.distance.cosine),\n",
    "    #Method(\"BERT\", \"Base:Layers=3,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":3}, spatial.distance.cosine),\n",
    "    #Method(\"BERT\", \"Base:Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine),\n",
    "    #Method(\"BERT\", \"Base:Layers=2,Summed\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"sum\", \"layers\":2}, spatial.distance.cosine),\n",
    "    #Method(\"BERT\", \" Base:Layers=3,Summed\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"sum\", \"layers\":3}, spatial.distance.cosine),\n",
    "    #Method(\"BERT\", \"Base:Layers=4,Summed\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"sum\", \"layers\":4}, spatial.distance.cosine),\n",
    "    #Method(\"BioBERT\", \"PMC,Layers=2,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pmc, \"tokenizer\":bert_tokenizer_pmc, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":2}, spatial.distance.cosine),\n",
    "    #Method(\"BioBERT\", \"PMC,Layers=3,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pmc, \"tokenizer\":bert_tokenizer_pmc, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":3}, spatial.distance.cosine),\n",
    "    #Method(\"BioBERT\", \"PMC,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pmc, \"tokenizer\":bert_tokenizer_pmc, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine),\n",
    "    #Method(\"BioBERT\", \"PubMed,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pubmed, \"tokenizer\":bert_tokenizer_pubmed, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine),\n",
    "    #Method(\"BioBERT\", \"PubMed,PMC,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pubmed_pmc, \"tokenizer\":bert_tokenizer_pubmed_pmc, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine),\n",
    "\n",
    "\n",
    "    # Methods that use topic models to generate embeddings.\n",
    "    Method(\"Topic Models\", \"LDA,Simple,Topics=20\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":20, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Simple,Topics=50\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":50, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Simple,Topics=100\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":100, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Simple,Topics=200\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":200, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Full,Topics=20\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":20, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Full,Topics=50\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":50, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Full,Topics=100\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":100, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"LDA,Full,Topics=200\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":200, \"algorithm\":\"lda\"}, spatial.distance.cosine),\n",
    "\n",
    "    Method(\"Topic Models\", \"NMF,Simple,Topics=20\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":20, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Simple,Topics=50\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":50, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Simple,Topics=100\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":100, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Simple,Topics=200\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"num_topics\":200, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=20\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":20, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=50\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":50, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=100\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":100, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "    Method(\"Topic Models\", \"NMF,Full,Topics=200\", pw.pairwise_square_topic_model, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"num_topics\":200, \"algorithm\":\"nmf\"}, spatial.distance.cosine),\n",
    "\n",
    "\n",
    "    # Methods that use variations on the n-grams approach with full preprocessing (includes stemming).\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,2-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,2-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,2-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,2-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Words,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use variations on the n-grams approach with simple preprocessing (no stemming).\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,2-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,2-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,2-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,2-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_simple_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use variations on the n-grams approach selecting for specific parts-of-speech (includes stemming).\n",
    "    Method(\"N-Grams\", \"Full,Nouns,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_only_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Adjectives,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_adj_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Adjectives,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_adj_only_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Full,Adjectives,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_adj_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Adjectives,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_adj_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,Adjectives,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_adj_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,Adjectives,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_adj_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,Adjectives,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_adj_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Full,Nouns,Adjectives,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_noun_adj_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use variations on the n-grams approach with a reduced vocabulary size and simple preprocessing (no stemming).\n",
    "    Method(\"N-Grams\", \"Simple,Words,Linares Pontes,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,Linares Pontes,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"N-Grams\", \"Simple,Words,Linares Pontes,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,Linares Pontes,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "\n",
    "\n",
    "    # Methods that use terms inferred from automated annotation of the text.\n",
    "    Method(\"NOBLE Coder\", \"Precise\", pw.pairwise_square_annotations, {\"ids_to_annotations\":annotations_noblecoder_precise, \"ontology\":ontology, \"binary\":True, \"metric\":\"jaccard\", \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"NOBLE Coder\", \"Partial\", pw.pairwise_square_annotations, {\"ids_to_annotations\":annotations_noblecoder_partial, \"ontology\":ontology, \"binary\":True, \"metric\":\"jaccard\", \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"NOBLE Coder\", \"Precise,TFIDF\", pw.pairwise_square_annotations, {\"ids_to_annotations\":annotations_noblecoder_precise, \"ontology\":ontology, \"binary\":True, \"metric\":\"cosine\", \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"NOBLE Coder\", \"Partial,TFIDF\", pw.pairwise_square_annotations, {\"ids_to_annotations\":annotations_noblecoder_partial, \"ontology\":ontology, \"binary\":True, \"metric\":\"cosine\", \"tfidf\":True}, spatial.distance.cosine),\n",
    "\n",
    "    # Methods that use terms assigned by humans that are present in the dataset.\n",
    "    Method(\"GO\", \"None\", pw.pairwise_square_annotations, {\"ids_to_annotations\":go_annotations, \"ontology\":ontology, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"PO\", \"None\", pw.pairwise_square_annotations, {\"ids_to_annotations\":po_annotations, \"ontology\":ontology, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"tfidf\":False}, spatial.distance.jaccard),\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # Individual phenes from those larger phenotypes.\n",
    "\n",
    "\n",
    "    # Approaches were the phenotype descriptions were split into individual phenes first (computationally expensive).\n",
    "    Method(\"Doc2Vec (Phenes)\", \"Wikipedia,Size=300\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_wiki_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Doc2Vec (Phenes)\", \"PubMed,Size=100\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_pubmed_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Mean\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Max\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Mean,Nouns,Adjectives\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes_noun_adj_simple_preprocessing, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Max,Nouns,Adjectives\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes_noun_adj_simple_preprocessing, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Mean,Linares Pontes\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes_linares_pontes, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Max,Linares Pontes\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes_linares_pontes, \"metric\":\"cosine\", \"method\":\"max\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "\n",
    "    #Method(\"BERT (Phenes)\", \"Base:Layers=2,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":2}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BERT (Phenes)\", \"Base:Layers=3,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":3}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BERT (Phenes)\", \"Base:Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BERT (Phenes)\", \"Base:Layers=2,Summed\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"sum\", \"layers\":2}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BERT (Phenes)\", \"Base:Layers=3,Summed\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"sum\", \"layers\":3}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BERT (Phenes)\", \"Base:Layers=4,Summed\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"sum\", \"layers\":4}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BioBERT (Phenes)\", \"PMC,Layers=2,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pmc, \"tokenizer\":bert_tokenizer_pmc, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":2}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BioBERT (Phenes)\", \"PMC,Layers=3,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pmc, \"tokenizer\":bert_tokenizer_pmc, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":3}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BioBERT (Phenes)\", \"PMC,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pmc, \"tokenizer\":bert_tokenizer_pmc, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BioBERT (Phenes)\", \"PubMed,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pubmed, \"tokenizer\":bert_tokenizer_pubmed, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    #Method(\"BioBERT (Phenes)\", \"PubMed,PMC,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pubmed_pmc, \"tokenizer\":bert_tokenizer_pubmed_pmc, \"ids_to_texts\":phenes, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "\n",
    "\n",
    "\n",
    "    # Methods that use variations on the N-Grams (Phenes) approach with full preprocessing (includes stemming).\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,2-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,2-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,2-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,2-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Words,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "\n",
    "    # Methods that use variations on the N-Grams (Phenes) approach with simple preprocessing (no stemming).\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,2-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,2-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,2-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,2),\"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,2-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,2), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "\n",
    "    # Methods that use variations on the N-Grams (Phenes) approach selecting for specific parts-of-speech (includes stemming).\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_only_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Adjectives,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_adj_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Adjectives,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_adj_only_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Adjectives,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_adj_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Adjectives,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_adj_only_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,Adjectives,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_adj_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,Adjectives,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_adj_full_preprocessing, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,Adjectives,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_adj_full_preprocessing, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Full,Nouns,Adjectives,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_noun_adj_full_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "\n",
    "    # Methods that use variations on the N-Grams (Phenes) approach with a reduced vocabulary size and simple preprocessing (no stemming).\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,Linares Pontes,1-grams\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_linares_pontes, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,Linares Pontes,1-grams,Binary\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_linares_pontes, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":False}, spatial.distance.jaccard, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,Linares Pontes,1-grams,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_linares_pontes, \"metric\":\"cosine\", \"binary\":False, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,Linares Pontes,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_linares_pontes, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "testing_methods = [\n",
    "    # Curation-based methods\n",
    "    Method(\"GO\", \"None\", pw.pairwise_square_annotations, {\"ids_to_annotations\":go_annotations, \"ontology\":ontology, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    Method(\"PO\", \"None\", pw.pairwise_square_annotations, {\"ids_to_annotations\":po_annotations, \"ontology\":ontology, \"metric\":\"jaccard\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"tfidf\":False}, spatial.distance.jaccard),\n",
    "    # Phenotype descriptions\n",
    "    Method(\"Doc2Vec\", \"Wikipedia,Size=300\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_wiki_model, \"ids_to_texts\":descriptions, \"metric\":\"cosine\"}, spatial.distance.cosine),\n",
    "    Method(\"Word2Vec\", \"Wikipedia,Size=300,Mean,Nouns,Adjectives\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":descriptions_noun_adj_simple_preprocessing, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine),\n",
    "    Method(\"BERT\", \"Base:Layers=3,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_base, \"tokenizer\":bert_tokenizer_base, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":3}, spatial.distance.cosine),\n",
    "    Method(\"BioBERT\", \"PubMed,PMC,Layers=4,Concatenated\", pw.pairwise_square_bert, {\"model\":bert_model_pubmed_pmc, \"tokenizer\":bert_tokenizer_pubmed_pmc, \"ids_to_texts\":descriptions, \"metric\":\"cosine\", \"method\":\"concat\", \"layers\":4}, spatial.distance.cosine),\n",
    "    Method(\"N-Grams\", \"Simple,Words,Linares Pontes,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":descriptions_linares_pontes, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine),\n",
    "    Method(\"NOBLE Coder\", \"Partial,TFIDF\", pw.pairwise_square_annotations, {\"ids_to_annotations\":annotations_noblecoder_partial, \"ontology\":ontology, \"binary\":True, \"metric\":\"cosine\", \"tfidf\":True}, spatial.distance.cosine),\n",
    "    # Phene descriptions\n",
    "    Method(\"Doc2Vec (Phenes)\", \"Wikipedia,Size=300\", pw.pairwise_square_doc2vec, {\"model\":doc2vec_wiki_model, \"ids_to_texts\":phenes, \"metric\":\"cosine\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"Word2Vec (Phenes)\", \"Wikipedia,Size=300,Mean,Nouns,Adjectives\", pw.pairwise_square_word2vec, {\"model\":word2vec_model, \"ids_to_texts\":phenes_noun_adj_simple_preprocessing, \"metric\":\"cosine\", \"method\":\"mean\"}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "    Method(\"N-Grams (Phenes)\", \"Simple,Words,1-grams,Binary,TFIDF\", pw.pairwise_square_ngrams, {\"ids_to_texts\":phenes_simple_preprocessing, \"metric\":\"cosine\", \"binary\":True, \"analyzer\":\"word\", \"ngram_range\":(1,1), \"max_features\":10000, \"tfidf\":True}, spatial.distance.cosine, tag=\"phenes\"),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "methods = training_methods if NOTEBOOK_TAGS[\"training\"] else testing_methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"running\"></a>\n",
    "### Running all of the methods to generate distance matrices\n",
    "Something here if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO:None                                                                00:00:03\n",
      "PO:None                                                                00:00:01\n",
      "Doc2Vec:Wikipedia,Size=300                                             00:00:07\n",
      "Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives                      00:00:01\n",
      "BERT:Base:Layers=3,Concatenated                                        00:14:12\n",
      "BioBERT:PubMed,PMC,Layers=4,Concatenated                               00:14:18\n",
      "N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF               00:00:06\n",
      "NOBLE Coder:Partial,TFIDF                                              00:00:03\n",
      "Doc2Vec (Phenes):Wikipedia,Size=300                                    00:01:22\n",
      "Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives             00:01:39\n",
      "N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF                     00:11:07\n"
     ]
    }
   ],
   "source": [
    "# Generate all the pairwise distance matrices (not in parallel).\n",
    "graphs = {}\n",
    "names = []\n",
    "durations = []\n",
    "for method in methods:\n",
    "    graph,duration = function_wrapper_with_duration(function=method.function, args=method.kwargs)\n",
    "    graphs[method.name_with_hyperparameters] = graph\n",
    "    names.append(method.name_with_hyperparameters)\n",
    "    durations.append(to_hms(duration))\n",
    "    print(\"{:70} {}\".format(method.name_with_hyperparameters,to_hms(duration)))\n",
    "durations_df = pd.DataFrame({\"method\":names,\"duration\":durations})\n",
    "durations_df.to_csv(os.path.join(OUTPUT_DIR,\"part_4_durations.csv\"), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"merging\"></a>\n",
    "### Merging all of the distance matrices into a single dataframe specifying edges\n",
    "This section also handles replacing IDs from the individual methods that are references individual phenes that are part of a larger phenotype, and replacing those IDs with IDs referencing the full phenotypes (one-to-one relationship between phenotypes and genes). In this case, the minimum distance found between any two phenes from those two phenotypes represents the distance between that pair of phenotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>BERT:Base:Layers=3,Concatenated</th>\n",
       "      <th>BioBERT:PubMed,PMC,Layers=4,Concatenated</th>\n",
       "      <th>N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>Doc2Vec (Phenes):Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2497</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.492080</td>\n",
       "      <td>0.273918</td>\n",
       "      <td>0.117563</td>\n",
       "      <td>0.037339</td>\n",
       "      <td>0.945848</td>\n",
       "      <td>0.803728</td>\n",
       "      <td>0.204753</td>\n",
       "      <td>0.278276</td>\n",
       "      <td>0.796365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2498</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.427673</td>\n",
       "      <td>0.512304</td>\n",
       "      <td>0.461870</td>\n",
       "      <td>0.228432</td>\n",
       "      <td>0.069785</td>\n",
       "      <td>0.994671</td>\n",
       "      <td>0.915870</td>\n",
       "      <td>0.353274</td>\n",
       "      <td>0.435056</td>\n",
       "      <td>0.961496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2499</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.564400</td>\n",
       "      <td>0.525348</td>\n",
       "      <td>0.213786</td>\n",
       "      <td>0.079023</td>\n",
       "      <td>0.994793</td>\n",
       "      <td>0.935954</td>\n",
       "      <td>0.369662</td>\n",
       "      <td>0.514471</td>\n",
       "      <td>0.959023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2500</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.036232</td>\n",
       "      <td>0.532318</td>\n",
       "      <td>0.537138</td>\n",
       "      <td>0.301149</td>\n",
       "      <td>0.119390</td>\n",
       "      <td>0.991823</td>\n",
       "      <td>0.935954</td>\n",
       "      <td>0.371638</td>\n",
       "      <td>0.551040</td>\n",
       "      <td>0.959023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>2501</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.056738</td>\n",
       "      <td>0.590537</td>\n",
       "      <td>0.334127</td>\n",
       "      <td>0.188197</td>\n",
       "      <td>0.058344</td>\n",
       "      <td>0.986016</td>\n",
       "      <td>0.941871</td>\n",
       "      <td>0.375241</td>\n",
       "      <td>0.331523</td>\n",
       "      <td>0.946937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>2502</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.499923</td>\n",
       "      <td>0.266366</td>\n",
       "      <td>0.103811</td>\n",
       "      <td>0.033192</td>\n",
       "      <td>0.953700</td>\n",
       "      <td>0.815053</td>\n",
       "      <td>0.213289</td>\n",
       "      <td>0.278276</td>\n",
       "      <td>0.873528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>2503</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.092199</td>\n",
       "      <td>0.555250</td>\n",
       "      <td>0.524732</td>\n",
       "      <td>0.186933</td>\n",
       "      <td>0.072269</td>\n",
       "      <td>0.995320</td>\n",
       "      <td>0.940648</td>\n",
       "      <td>0.401548</td>\n",
       "      <td>0.540968</td>\n",
       "      <td>0.972136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>2504</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.043165</td>\n",
       "      <td>0.497086</td>\n",
       "      <td>0.440370</td>\n",
       "      <td>0.184870</td>\n",
       "      <td>0.054795</td>\n",
       "      <td>0.978726</td>\n",
       "      <td>0.886473</td>\n",
       "      <td>0.305292</td>\n",
       "      <td>0.478938</td>\n",
       "      <td>0.970834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>2505</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.131387</td>\n",
       "      <td>0.380056</td>\n",
       "      <td>0.169344</td>\n",
       "      <td>0.049516</td>\n",
       "      <td>0.022890</td>\n",
       "      <td>0.847976</td>\n",
       "      <td>0.620993</td>\n",
       "      <td>0.261939</td>\n",
       "      <td>0.260708</td>\n",
       "      <td>0.857565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>2506</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.043165</td>\n",
       "      <td>0.514861</td>\n",
       "      <td>0.474861</td>\n",
       "      <td>0.272169</td>\n",
       "      <td>0.110464</td>\n",
       "      <td>0.987815</td>\n",
       "      <td>0.935954</td>\n",
       "      <td>0.332465</td>\n",
       "      <td>0.481888</td>\n",
       "      <td>0.964217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>2507</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.179856</td>\n",
       "      <td>0.436388</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.056229</td>\n",
       "      <td>0.024944</td>\n",
       "      <td>0.861806</td>\n",
       "      <td>0.654354</td>\n",
       "      <td>0.292189</td>\n",
       "      <td>0.260708</td>\n",
       "      <td>0.872681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>2508</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.043165</td>\n",
       "      <td>0.616098</td>\n",
       "      <td>0.323756</td>\n",
       "      <td>0.114476</td>\n",
       "      <td>0.040706</td>\n",
       "      <td>0.983865</td>\n",
       "      <td>0.869955</td>\n",
       "      <td>0.353538</td>\n",
       "      <td>0.364396</td>\n",
       "      <td>0.962118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>2510</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.035211</td>\n",
       "      <td>0.495707</td>\n",
       "      <td>0.297705</td>\n",
       "      <td>0.083310</td>\n",
       "      <td>0.033765</td>\n",
       "      <td>0.906948</td>\n",
       "      <td>0.768903</td>\n",
       "      <td>0.237985</td>\n",
       "      <td>0.374776</td>\n",
       "      <td>0.882060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>2512</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.330986</td>\n",
       "      <td>0.507190</td>\n",
       "      <td>0.252597</td>\n",
       "      <td>0.097275</td>\n",
       "      <td>0.035667</td>\n",
       "      <td>0.921562</td>\n",
       "      <td>0.714306</td>\n",
       "      <td>0.251288</td>\n",
       "      <td>0.200374</td>\n",
       "      <td>0.872402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>2513</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.036232</td>\n",
       "      <td>0.389739</td>\n",
       "      <td>0.330207</td>\n",
       "      <td>0.156228</td>\n",
       "      <td>0.044084</td>\n",
       "      <td>0.928149</td>\n",
       "      <td>0.838941</td>\n",
       "      <td>0.259569</td>\n",
       "      <td>0.340708</td>\n",
       "      <td>0.851877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>2514</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.388489</td>\n",
       "      <td>0.460988</td>\n",
       "      <td>0.296110</td>\n",
       "      <td>0.106274</td>\n",
       "      <td>0.029168</td>\n",
       "      <td>0.964941</td>\n",
       "      <td>0.693033</td>\n",
       "      <td>0.250454</td>\n",
       "      <td>0.374776</td>\n",
       "      <td>0.910322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>2516</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.532847</td>\n",
       "      <td>0.318500</td>\n",
       "      <td>0.176675</td>\n",
       "      <td>0.056443</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.867980</td>\n",
       "      <td>0.494596</td>\n",
       "      <td>0.222775</td>\n",
       "      <td>0.227615</td>\n",
       "      <td>0.752867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>2517</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.226667</td>\n",
       "      <td>0.429743</td>\n",
       "      <td>0.249818</td>\n",
       "      <td>0.072738</td>\n",
       "      <td>0.029424</td>\n",
       "      <td>0.958012</td>\n",
       "      <td>0.563274</td>\n",
       "      <td>0.289452</td>\n",
       "      <td>0.278276</td>\n",
       "      <td>0.949201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>2518</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.506849</td>\n",
       "      <td>0.378157</td>\n",
       "      <td>0.156303</td>\n",
       "      <td>0.051445</td>\n",
       "      <td>0.022125</td>\n",
       "      <td>0.917711</td>\n",
       "      <td>0.545262</td>\n",
       "      <td>0.221726</td>\n",
       "      <td>0.207076</td>\n",
       "      <td>0.842986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>2545</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.283019</td>\n",
       "      <td>0.534610</td>\n",
       "      <td>0.258736</td>\n",
       "      <td>0.069371</td>\n",
       "      <td>0.027680</td>\n",
       "      <td>0.931660</td>\n",
       "      <td>0.765625</td>\n",
       "      <td>0.172289</td>\n",
       "      <td>0.307097</td>\n",
       "      <td>0.850262</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to   GO:None   PO:None  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives  BERT:Base:Layers=3,Concatenated  BioBERT:PubMed,PMC,Layers=4,Concatenated  N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF  NOBLE Coder:Partial,TFIDF  Doc2Vec (Phenes):Wikipedia,Size=300  Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives  N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF\n",
       "1      1  2497  0.928571  0.029197                    0.492080                                           0.273918                         0.117563                                  0.037339                                           0.945848                          0.803728                             0.204753                                           0.278276                                                    0.796365 \n",
       "2      1  2498  0.857143  0.427673                    0.512304                                           0.461870                         0.228432                                  0.069785                                           0.994671                          0.915870                             0.353274                                           0.435056                                                    0.961496 \n",
       "3      1  2499  0.875000  0.029197                    0.564400                                           0.525348                         0.213786                                  0.079023                                           0.994793                          0.935954                             0.369662                                           0.514471                                                    0.959023 \n",
       "4      1  2500  0.833333  0.036232                    0.532318                                           0.537138                         0.301149                                  0.119390                                           0.991823                          0.935954                             0.371638                                           0.551040                                                    0.959023 \n",
       "5      1  2501  0.888889  0.056738                    0.590537                                           0.334127                         0.188197                                  0.058344                                           0.986016                          0.941871                             0.375241                                           0.331523                                                    0.946937 \n",
       "6      1  2502  0.916667  0.029197                    0.499923                                           0.266366                         0.103811                                  0.033192                                           0.953700                          0.815053                             0.213289                                           0.278276                                                    0.873528 \n",
       "7      1  2503  0.750000  0.092199                    0.555250                                           0.524732                         0.186933                                  0.072269                                           0.995320                          0.940648                             0.401548                                           0.540968                                                    0.972136 \n",
       "8      1  2504  0.833333  0.043165                    0.497086                                           0.440370                         0.184870                                  0.054795                                           0.978726                          0.886473                             0.305292                                           0.478938                                                    0.970834 \n",
       "9      1  2505  0.923077  0.131387                    0.380056                                           0.169344                         0.049516                                  0.022890                                           0.847976                          0.620993                             0.261939                                           0.260708                                                    0.857565 \n",
       "10     1  2506  0.833333  0.043165                    0.514861                                           0.474861                         0.272169                                  0.110464                                           0.987815                          0.935954                             0.332465                                           0.481888                                                    0.964217 \n",
       "11     1  2507  0.909091  0.179856                    0.436388                                           0.173617                         0.056229                                  0.024944                                           0.861806                          0.654354                             0.292189                                           0.260708                                                    0.872681 \n",
       "12     1  2508  0.916667  0.043165                    0.616098                                           0.323756                         0.114476                                  0.040706                                           0.983865                          0.869955                             0.353538                                           0.364396                                                    0.962118 \n",
       "13     1  2510  0.928571  0.035211                    0.495707                                           0.297705                         0.083310                                  0.033765                                           0.906948                          0.768903                             0.237985                                           0.374776                                                    0.882060 \n",
       "14     1  2512  0.857143  0.330986                    0.507190                                           0.252597                         0.097275                                  0.035667                                           0.921562                          0.714306                             0.251288                                           0.200374                                                    0.872402 \n",
       "15     1  2513  0.833333  0.036232                    0.389739                                           0.330207                         0.156228                                  0.044084                                           0.928149                          0.838941                             0.259569                                           0.340708                                                    0.851877 \n",
       "16     1  2514  0.875000  0.388489                    0.460988                                           0.296110                         0.106274                                  0.029168                                           0.964941                          0.693033                             0.250454                                           0.374776                                                    0.910322 \n",
       "17     1  2516  0.875000  0.532847                    0.318500                                           0.176675                         0.056443                                  0.022727                                           0.867980                          0.494596                             0.222775                                           0.227615                                                    0.752867 \n",
       "18     1  2517  0.833333  0.226667                    0.429743                                           0.249818                         0.072738                                  0.029424                                           0.958012                          0.563274                             0.289452                                           0.278276                                                    0.949201 \n",
       "19     1  2518  0.800000  0.506849                    0.378157                                           0.156303                         0.051445                                  0.022125                                           0.917711                          0.545262                             0.221726                                           0.207076                                                    0.842986 \n",
       "20     1  2545  0.875000  0.283019                    0.534610                                           0.258736                         0.069371                                  0.027680                                           0.931660                          0.765625                             0.172289                                           0.307097                                                    0.850262 "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging all the edgelists together.\n",
    "metric_dict = {method.name_with_hyperparameters:method.metric for method in methods}\n",
    "tags_dict = {method.name_with_hyperparameters:method.tag for method in methods}\n",
    "names = list(graphs.keys())\n",
    "edgelists = {k:v.edgelist for k,v in graphs.items()}\n",
    "\n",
    "# Modify the edgelists for the methods that were using a phene split.\n",
    "for name,edgelist in edgelists.items():\n",
    "    # Converting phene IDs back to phenotype (gene) IDs where applicable.\n",
    "    if \"phene\" in tags_dict[name]:\n",
    "        edgelist[\"from\"] = edgelist[\"from\"].map(lambda x: phene_id_to_id[x])\n",
    "        edgelist[\"to\"] = edgelist[\"to\"].map(lambda x: phene_id_to_id[x])\n",
    "        edgelist = edgelist.groupby([\"from\",\"to\"], as_index=False).min()\n",
    "    # Making sure the edges are listed with the nodes sorted consistently.\n",
    "    cond = edgelist[\"from\"] > edgelist[\"to\"]\n",
    "    edgelist.loc[cond, ['from', 'to']] = edgelist.loc[cond, ['to', 'from']].values\n",
    "    edgelists[name] = edgelist\n",
    "\n",
    "# Do the merge step and remove self edges from the full dataframe.\n",
    "df = merge_edgelists(edgelists, default_value=1.000)\n",
    "df = remove_self_loops(df)\n",
    "df[\"from\"] = df[\"from\"].astype(\"int64\")\n",
    "df[\"to\"] = df[\"to\"].astype(\"int64\")\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ensemble\"></a>\n",
    "### Combining multiple distances measurements into summarizing distance values\n",
    "The purpose of this section is to iteratively train models on subsections of the dataset using simple regression or machine learning approaches to predict a value from zero to one indicating indicating how likely is it that two genes share atleast one of the specified groups in common. The information input to these models is the distance scores provided by each method in some set of all the methods used in this notebook. The purpose is to see whether or not a function of these similarity scores specifically trained to the task of predicting common groupings is better able to used the distance metric information to report a score for this task."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>BERT:Base:Layers=3,Concatenated</th>\n",
       "      <th>BioBERT:PubMed,PMC,Layers=4,Concatenated</th>\n",
       "      <th>N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>Doc2Vec (Phenes):Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2497</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.492080</td>\n",
       "      <td>0.273918</td>\n",
       "      <td>0.117563</td>\n",
       "      <td>0.037339</td>\n",
       "      <td>0.945848</td>\n",
       "      <td>0.803728</td>\n",
       "      <td>0.204753</td>\n",
       "      <td>0.278276</td>\n",
       "      <td>0.796365</td>\n",
       "      <td>0.353725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2498</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.427673</td>\n",
       "      <td>0.512304</td>\n",
       "      <td>0.461870</td>\n",
       "      <td>0.228432</td>\n",
       "      <td>0.069785</td>\n",
       "      <td>0.994671</td>\n",
       "      <td>0.915870</td>\n",
       "      <td>0.353274</td>\n",
       "      <td>0.435056</td>\n",
       "      <td>0.961496</td>\n",
       "      <td>0.735796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2499</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.564400</td>\n",
       "      <td>0.525348</td>\n",
       "      <td>0.213786</td>\n",
       "      <td>0.079023</td>\n",
       "      <td>0.994793</td>\n",
       "      <td>0.935954</td>\n",
       "      <td>0.369662</td>\n",
       "      <td>0.514471</td>\n",
       "      <td>0.959023</td>\n",
       "      <td>0.793990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2500</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.036232</td>\n",
       "      <td>0.532318</td>\n",
       "      <td>0.537138</td>\n",
       "      <td>0.301149</td>\n",
       "      <td>0.119390</td>\n",
       "      <td>0.991823</td>\n",
       "      <td>0.935954</td>\n",
       "      <td>0.371638</td>\n",
       "      <td>0.551040</td>\n",
       "      <td>0.959023</td>\n",
       "      <td>0.837148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>2501</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.056738</td>\n",
       "      <td>0.590537</td>\n",
       "      <td>0.334127</td>\n",
       "      <td>0.188197</td>\n",
       "      <td>0.058344</td>\n",
       "      <td>0.986016</td>\n",
       "      <td>0.941871</td>\n",
       "      <td>0.375241</td>\n",
       "      <td>0.331523</td>\n",
       "      <td>0.946937</td>\n",
       "      <td>0.659988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>2502</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.499923</td>\n",
       "      <td>0.266366</td>\n",
       "      <td>0.103811</td>\n",
       "      <td>0.033192</td>\n",
       "      <td>0.953700</td>\n",
       "      <td>0.815053</td>\n",
       "      <td>0.213289</td>\n",
       "      <td>0.278276</td>\n",
       "      <td>0.873528</td>\n",
       "      <td>0.367527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>2503</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.092199</td>\n",
       "      <td>0.555250</td>\n",
       "      <td>0.524732</td>\n",
       "      <td>0.186933</td>\n",
       "      <td>0.072269</td>\n",
       "      <td>0.995320</td>\n",
       "      <td>0.940648</td>\n",
       "      <td>0.401548</td>\n",
       "      <td>0.540968</td>\n",
       "      <td>0.972136</td>\n",
       "      <td>0.794708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>2504</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.043165</td>\n",
       "      <td>0.497086</td>\n",
       "      <td>0.440370</td>\n",
       "      <td>0.184870</td>\n",
       "      <td>0.054795</td>\n",
       "      <td>0.978726</td>\n",
       "      <td>0.886473</td>\n",
       "      <td>0.305292</td>\n",
       "      <td>0.478938</td>\n",
       "      <td>0.970834</td>\n",
       "      <td>0.646634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>2505</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.131387</td>\n",
       "      <td>0.380056</td>\n",
       "      <td>0.169344</td>\n",
       "      <td>0.049516</td>\n",
       "      <td>0.022890</td>\n",
       "      <td>0.847976</td>\n",
       "      <td>0.620993</td>\n",
       "      <td>0.261939</td>\n",
       "      <td>0.260708</td>\n",
       "      <td>0.857565</td>\n",
       "      <td>0.160443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>2506</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.043165</td>\n",
       "      <td>0.514861</td>\n",
       "      <td>0.474861</td>\n",
       "      <td>0.272169</td>\n",
       "      <td>0.110464</td>\n",
       "      <td>0.987815</td>\n",
       "      <td>0.935954</td>\n",
       "      <td>0.332465</td>\n",
       "      <td>0.481888</td>\n",
       "      <td>0.964217</td>\n",
       "      <td>0.779092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>2507</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.179856</td>\n",
       "      <td>0.436388</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.056229</td>\n",
       "      <td>0.024944</td>\n",
       "      <td>0.861806</td>\n",
       "      <td>0.654354</td>\n",
       "      <td>0.292189</td>\n",
       "      <td>0.260708</td>\n",
       "      <td>0.872681</td>\n",
       "      <td>0.219638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>2508</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.043165</td>\n",
       "      <td>0.616098</td>\n",
       "      <td>0.323756</td>\n",
       "      <td>0.114476</td>\n",
       "      <td>0.040706</td>\n",
       "      <td>0.983865</td>\n",
       "      <td>0.869955</td>\n",
       "      <td>0.353538</td>\n",
       "      <td>0.364396</td>\n",
       "      <td>0.962118</td>\n",
       "      <td>0.579428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>2510</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.035211</td>\n",
       "      <td>0.495707</td>\n",
       "      <td>0.297705</td>\n",
       "      <td>0.083310</td>\n",
       "      <td>0.033765</td>\n",
       "      <td>0.906948</td>\n",
       "      <td>0.768903</td>\n",
       "      <td>0.237985</td>\n",
       "      <td>0.374776</td>\n",
       "      <td>0.882060</td>\n",
       "      <td>0.355406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>2512</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.330986</td>\n",
       "      <td>0.507190</td>\n",
       "      <td>0.252597</td>\n",
       "      <td>0.097275</td>\n",
       "      <td>0.035667</td>\n",
       "      <td>0.921562</td>\n",
       "      <td>0.714306</td>\n",
       "      <td>0.251288</td>\n",
       "      <td>0.200374</td>\n",
       "      <td>0.872402</td>\n",
       "      <td>0.307721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>2513</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.036232</td>\n",
       "      <td>0.389739</td>\n",
       "      <td>0.330207</td>\n",
       "      <td>0.156228</td>\n",
       "      <td>0.044084</td>\n",
       "      <td>0.928149</td>\n",
       "      <td>0.838941</td>\n",
       "      <td>0.259569</td>\n",
       "      <td>0.340708</td>\n",
       "      <td>0.851877</td>\n",
       "      <td>0.389242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>2514</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.388489</td>\n",
       "      <td>0.460988</td>\n",
       "      <td>0.296110</td>\n",
       "      <td>0.106274</td>\n",
       "      <td>0.029168</td>\n",
       "      <td>0.964941</td>\n",
       "      <td>0.693033</td>\n",
       "      <td>0.250454</td>\n",
       "      <td>0.374776</td>\n",
       "      <td>0.910322</td>\n",
       "      <td>0.364662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>2516</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.532847</td>\n",
       "      <td>0.318500</td>\n",
       "      <td>0.176675</td>\n",
       "      <td>0.056443</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.867980</td>\n",
       "      <td>0.494596</td>\n",
       "      <td>0.222775</td>\n",
       "      <td>0.227615</td>\n",
       "      <td>0.752867</td>\n",
       "      <td>0.101684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>2517</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.226667</td>\n",
       "      <td>0.429743</td>\n",
       "      <td>0.249818</td>\n",
       "      <td>0.072738</td>\n",
       "      <td>0.029424</td>\n",
       "      <td>0.958012</td>\n",
       "      <td>0.563274</td>\n",
       "      <td>0.289452</td>\n",
       "      <td>0.278276</td>\n",
       "      <td>0.949201</td>\n",
       "      <td>0.300515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>2518</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.506849</td>\n",
       "      <td>0.378157</td>\n",
       "      <td>0.156303</td>\n",
       "      <td>0.051445</td>\n",
       "      <td>0.022125</td>\n",
       "      <td>0.917711</td>\n",
       "      <td>0.545262</td>\n",
       "      <td>0.221726</td>\n",
       "      <td>0.207076</td>\n",
       "      <td>0.842986</td>\n",
       "      <td>0.143963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>2545</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.283019</td>\n",
       "      <td>0.534610</td>\n",
       "      <td>0.258736</td>\n",
       "      <td>0.069371</td>\n",
       "      <td>0.027680</td>\n",
       "      <td>0.931660</td>\n",
       "      <td>0.765625</td>\n",
       "      <td>0.172289</td>\n",
       "      <td>0.307097</td>\n",
       "      <td>0.850262</td>\n",
       "      <td>0.316533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    from    to   GO:None   PO:None  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives  BERT:Base:Layers=3,Concatenated  BioBERT:PubMed,PMC,Layers=4,Concatenated  N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF  NOBLE Coder:Partial,TFIDF  Doc2Vec (Phenes):Wikipedia,Size=300  Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives  N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF      Mean\n",
       "1      1  2497  0.928571  0.029197                    0.492080                                           0.273918                         0.117563                                  0.037339                                           0.945848                          0.803728                             0.204753                                           0.278276                                                    0.796365   0.353725\n",
       "2      1  2498  0.857143  0.427673                    0.512304                                           0.461870                         0.228432                                  0.069785                                           0.994671                          0.915870                             0.353274                                           0.435056                                                    0.961496   0.735796\n",
       "3      1  2499  0.875000  0.029197                    0.564400                                           0.525348                         0.213786                                  0.079023                                           0.994793                          0.935954                             0.369662                                           0.514471                                                    0.959023   0.793990\n",
       "4      1  2500  0.833333  0.036232                    0.532318                                           0.537138                         0.301149                                  0.119390                                           0.991823                          0.935954                             0.371638                                           0.551040                                                    0.959023   0.837148\n",
       "5      1  2501  0.888889  0.056738                    0.590537                                           0.334127                         0.188197                                  0.058344                                           0.986016                          0.941871                             0.375241                                           0.331523                                                    0.946937   0.659988\n",
       "6      1  2502  0.916667  0.029197                    0.499923                                           0.266366                         0.103811                                  0.033192                                           0.953700                          0.815053                             0.213289                                           0.278276                                                    0.873528   0.367527\n",
       "7      1  2503  0.750000  0.092199                    0.555250                                           0.524732                         0.186933                                  0.072269                                           0.995320                          0.940648                             0.401548                                           0.540968                                                    0.972136   0.794708\n",
       "8      1  2504  0.833333  0.043165                    0.497086                                           0.440370                         0.184870                                  0.054795                                           0.978726                          0.886473                             0.305292                                           0.478938                                                    0.970834   0.646634\n",
       "9      1  2505  0.923077  0.131387                    0.380056                                           0.169344                         0.049516                                  0.022890                                           0.847976                          0.620993                             0.261939                                           0.260708                                                    0.857565   0.160443\n",
       "10     1  2506  0.833333  0.043165                    0.514861                                           0.474861                         0.272169                                  0.110464                                           0.987815                          0.935954                             0.332465                                           0.481888                                                    0.964217   0.779092\n",
       "11     1  2507  0.909091  0.179856                    0.436388                                           0.173617                         0.056229                                  0.024944                                           0.861806                          0.654354                             0.292189                                           0.260708                                                    0.872681   0.219638\n",
       "12     1  2508  0.916667  0.043165                    0.616098                                           0.323756                         0.114476                                  0.040706                                           0.983865                          0.869955                             0.353538                                           0.364396                                                    0.962118   0.579428\n",
       "13     1  2510  0.928571  0.035211                    0.495707                                           0.297705                         0.083310                                  0.033765                                           0.906948                          0.768903                             0.237985                                           0.374776                                                    0.882060   0.355406\n",
       "14     1  2512  0.857143  0.330986                    0.507190                                           0.252597                         0.097275                                  0.035667                                           0.921562                          0.714306                             0.251288                                           0.200374                                                    0.872402   0.307721\n",
       "15     1  2513  0.833333  0.036232                    0.389739                                           0.330207                         0.156228                                  0.044084                                           0.928149                          0.838941                             0.259569                                           0.340708                                                    0.851877   0.389242\n",
       "16     1  2514  0.875000  0.388489                    0.460988                                           0.296110                         0.106274                                  0.029168                                           0.964941                          0.693033                             0.250454                                           0.374776                                                    0.910322   0.364662\n",
       "17     1  2516  0.875000  0.532847                    0.318500                                           0.176675                         0.056443                                  0.022727                                           0.867980                          0.494596                             0.222775                                           0.227615                                                    0.752867   0.101684\n",
       "18     1  2517  0.833333  0.226667                    0.429743                                           0.249818                         0.072738                                  0.029424                                           0.958012                          0.563274                             0.289452                                           0.278276                                                    0.949201   0.300515\n",
       "19     1  2518  0.800000  0.506849                    0.378157                                           0.156303                         0.051445                                  0.022125                                           0.917711                          0.545262                             0.221726                                           0.207076                                                    0.842986   0.143963\n",
       "20     1  2545  0.875000  0.283019                    0.534610                                           0.258736                         0.069371                                  0.027680                                           0.931660                          0.765625                             0.172289                                           0.307097                                                    0.850262   0.316533"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the average distance percentile as a means of combining multiple scores.\n",
    "name = \"Mean\"\n",
    "names_to_use_for_mean = [name for name in names if not name in [\"GO:None\",\"PO:None\"]]\n",
    "df[name] = df[names_to_use_for_mean].rank(pct=True).mean(axis=1)\n",
    "names.append(name)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO:None\n",
      "PO:None\n",
      "Doc2Vec:Wikipedia,Size=300\n",
      "Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives\n",
      "BERT:Base:Layers=3,Concatenated\n",
      "BioBERT:PubMed,PMC,Layers=4,Concatenated\n",
      "N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF\n",
      "NOBLE Coder:Partial,TFIDF\n",
      "Doc2Vec (Phenes):Wikipedia,Size=300\n",
      "Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives\n",
      "N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF\n",
      "Mean\n"
     ]
    }
   ],
   "source": [
    "# Normalizing all of the array representations of the graphs so they can be combined. Then this version of the arrays\n",
    "# should be used by any other cells that need all of the arrays, rather than the arrays accessed from the graph\n",
    "# objects. This is necessary for this analysis because some of the graph objects refer to phene datasets not\n",
    "# phenotype datasets.\n",
    "name_to_array = {}\n",
    "ids = list(descriptions.keys())\n",
    "n = len(descriptions)\n",
    "id_to_array_index = {i:idx for idx,i in enumerate(ids)}\n",
    "array_index_to_id = {idx:i for i,idx in id_to_array_index.items()}\n",
    "for name in names:\n",
    "    print(name)\n",
    "    idx = list(df.columns).index(name)+1\n",
    "    arr = np.ones((n, n))\n",
    "    for row in df.itertuples():\n",
    "        arr[id_to_array_index[row[1]]][id_to_array_index[row[2]]] = row[idx]\n",
    "        arr[id_to_array_index[row[2]]][id_to_array_index[row[1]]] = row[idx]\n",
    "    np.fill_diagonal(arr, 0.000) \n",
    "    name_to_array[name] = arr    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_5\"></a>\n",
    "# Part 5. Cluster Analysis\n",
    "The purpose of this section is to look at different ways that the embeddings obtained for the dataset of phenotype descriptions can be used to cluster or organize the genes to which those phenotypes are mapped into subgroups or representations. These approaches include generating topic models from the data, and doing agglomerative clustering to find clusters to which each gene belongs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"topic_modeling\"></a>\n",
    "### Approach 1: Topic modeling based on n-grams with a reduced vocabulary\n",
    "Topic modelling learns a set of word probability distributions from the dataset of text descriptions, which represent distinct topics which are present in the dataset. Each text description can then be represented as a discrete probability distribution over the learned topics based on the probability that a given piece of text belongs to each particular topics. This is a form of data reduction because a high dimensionsal bag-of-words can be represented as a vector of *k* probabilities where *k* is the number of topics. The main advantages of topic modelling over clustering is that topic modelling provides soft classifications that can be additionally interpreted, rather than hard classifications into a single cluster. Topic models are also explainable, because the word probability distributions for that topic can be used to determine which words are most representative of any given topic. One problem with topic modelling is that is uses the n-grams embeddings to semantic similarity between different words is not accounted for. To help alleviate this, this section uses implementations of some existing algorithms to compress the vocabulary as a preprocessing step based on word distance matrices generated using word embeddings.\n",
    "\n",
    "Topic models define topics present in a dataset of texts as word or n-gram probability distributions. These models represent each instance of text then as being composed of or generated as as mixture of these topics. The vector for each text that indicates which fraction of that text is generated by a each topic is of length *n* where *n* is the number of topics, and can be used as a reduced dimensionality of the text, with a much smaller vector length than the n-grams embedding itself. Therefore we can build a topic model of the data with 100 topics for example in order to then represent each description in the dataset as a a vector of length 100. This section constructs topic models from the n-gram representations of the dataset and selects different values for the number of topics in order to find a value that works well during the grid search over the training dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>topic</th>\n",
       "      <th>GAM</th>\n",
       "      <th>GEM</th>\n",
       "      <th>EMG</th>\n",
       "      <th>MGD</th>\n",
       "      <th>EMB</th>\n",
       "      <th>MSD</th>\n",
       "      <th>NHM</th>\n",
       "      <th>SRL</th>\n",
       "      <th>GER</th>\n",
       "      <th>NLS</th>\n",
       "      <th>PIG</th>\n",
       "      <th>GRS</th>\n",
       "      <th>ROT</th>\n",
       "      <th>LEF</th>\n",
       "      <th>IST</th>\n",
       "      <th>ARC</th>\n",
       "      <th>MSL</th>\n",
       "      <th>FSM</th>\n",
       "      <th>OVP</th>\n",
       "      <th>SRF</th>\n",
       "      <th>SSC</th>\n",
       "      <th>FLT</th>\n",
       "      <th>SEN</th>\n",
       "      <th>CDR</th>\n",
       "      <th>MTM</th>\n",
       "      <th>STT</th>\n",
       "      <th>RTH</th>\n",
       "      <th>TCM</th>\n",
       "      <th>CUL</th>\n",
       "      <th>PRA</th>\n",
       "      <th>CPR</th>\n",
       "      <th>WAT</th>\n",
       "      <th>TMP</th>\n",
       "      <th>LIT</th>\n",
       "      <th>MEC</th>\n",
       "      <th>MPH</th>\n",
       "      <th>NUT</th>\n",
       "      <th>HRM</th>\n",
       "      <th>CHS</th>\n",
       "      <th>MCH</th>\n",
       "      <th>PTH</th>\n",
       "      <th>OBI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.233855</td>\n",
       "      <td>0.216993</td>\n",
       "      <td>0.124526</td>\n",
       "      <td>0.139415</td>\n",
       "      <td>0.010228</td>\n",
       "      <td>0.055123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005404</td>\n",
       "      <td>0.011408</td>\n",
       "      <td>0.005775</td>\n",
       "      <td>0.002957</td>\n",
       "      <td>0.007177</td>\n",
       "      <td>0.010170</td>\n",
       "      <td>0.007156</td>\n",
       "      <td>0.005666</td>\n",
       "      <td>0.006718</td>\n",
       "      <td>0.006163</td>\n",
       "      <td>0.007406</td>\n",
       "      <td>0.005776</td>\n",
       "      <td>0.014127</td>\n",
       "      <td>0.001352</td>\n",
       "      <td>0.002740</td>\n",
       "      <td>0.003543</td>\n",
       "      <td>0.000427</td>\n",
       "      <td>9.202232e-03</td>\n",
       "      <td>0.001941</td>\n",
       "      <td>0.009646</td>\n",
       "      <td>0.007024</td>\n",
       "      <td>0.003604</td>\n",
       "      <td>0.000939</td>\n",
       "      <td>0.004591</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>0.002933</td>\n",
       "      <td>0.002393</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.000972</td>\n",
       "      <td>0.001199</td>\n",
       "      <td>0.001053</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.004134</td>\n",
       "      <td>0.009401</td>\n",
       "      <td>0.018687</td>\n",
       "      <td>0.001565</td>\n",
       "      <td>0.040489</td>\n",
       "      <td>0.011614</td>\n",
       "      <td>0.000562</td>\n",
       "      <td>0.002660</td>\n",
       "      <td>0.000542</td>\n",
       "      <td>0.000554</td>\n",
       "      <td>0.002779</td>\n",
       "      <td>0.002613</td>\n",
       "      <td>0.001889</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>0.000612</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.006118</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>0.001652</td>\n",
       "      <td>0.000881</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.002562</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000731</td>\n",
       "      <td>0.000427</td>\n",
       "      <td>0.004544</td>\n",
       "      <td>0.001261</td>\n",
       "      <td>0.000394</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.001825</td>\n",
       "      <td>0.000219</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000653</td>\n",
       "      <td>0.000636</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>0.008770</td>\n",
       "      <td>0.013485</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.009631</td>\n",
       "      <td>0.036758</td>\n",
       "      <td>0.027312</td>\n",
       "      <td>0.007177</td>\n",
       "      <td>0.013111</td>\n",
       "      <td>0.009136</td>\n",
       "      <td>0.004887</td>\n",
       "      <td>0.008099</td>\n",
       "      <td>0.004785</td>\n",
       "      <td>0.004810</td>\n",
       "      <td>0.005188</td>\n",
       "      <td>0.005258</td>\n",
       "      <td>0.004650</td>\n",
       "      <td>0.005122</td>\n",
       "      <td>0.008899</td>\n",
       "      <td>0.008305</td>\n",
       "      <td>0.007519</td>\n",
       "      <td>0.002659</td>\n",
       "      <td>0.003938</td>\n",
       "      <td>0.004301</td>\n",
       "      <td>0.000713</td>\n",
       "      <td>6.296012e-04</td>\n",
       "      <td>0.004960</td>\n",
       "      <td>0.005831</td>\n",
       "      <td>0.007089</td>\n",
       "      <td>0.003573</td>\n",
       "      <td>0.004343</td>\n",
       "      <td>0.001303</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>0.003178</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>0.001055</td>\n",
       "      <td>0.004622</td>\n",
       "      <td>0.004412</td>\n",
       "      <td>0.003675</td>\n",
       "      <td>0.005247</td>\n",
       "      <td>0.002479</td>\n",
       "      <td>0.007914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>0.002796</td>\n",
       "      <td>0.008644</td>\n",
       "      <td>0.011394</td>\n",
       "      <td>0.001315</td>\n",
       "      <td>0.038993</td>\n",
       "      <td>0.006750</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.002374</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.001491</td>\n",
       "      <td>0.000857</td>\n",
       "      <td>0.001067</td>\n",
       "      <td>0.000863</td>\n",
       "      <td>0.001919</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000485</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.000811</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>4.641771e-06</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000724</td>\n",
       "      <td>0.002106</td>\n",
       "      <td>0.005886</td>\n",
       "      <td>0.000637</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.001006</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001208</td>\n",
       "      <td>0.001016</td>\n",
       "      <td>0.000919</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.000980</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>0.017270</td>\n",
       "      <td>0.049438</td>\n",
       "      <td>0.062597</td>\n",
       "      <td>0.009543</td>\n",
       "      <td>0.096471</td>\n",
       "      <td>0.043610</td>\n",
       "      <td>0.005191</td>\n",
       "      <td>0.027866</td>\n",
       "      <td>0.013209</td>\n",
       "      <td>0.017232</td>\n",
       "      <td>0.017679</td>\n",
       "      <td>0.008049</td>\n",
       "      <td>0.009300</td>\n",
       "      <td>0.010925</td>\n",
       "      <td>0.006650</td>\n",
       "      <td>0.006165</td>\n",
       "      <td>0.008678</td>\n",
       "      <td>0.008245</td>\n",
       "      <td>0.003203</td>\n",
       "      <td>0.006532</td>\n",
       "      <td>0.007222</td>\n",
       "      <td>0.006100</td>\n",
       "      <td>0.008059</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002005</td>\n",
       "      <td>0.007473</td>\n",
       "      <td>0.017546</td>\n",
       "      <td>0.007957</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.001259</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.004076</td>\n",
       "      <td>0.006525</td>\n",
       "      <td>0.022467</td>\n",
       "      <td>0.005462</td>\n",
       "      <td>0.003984</td>\n",
       "      <td>0.004436</td>\n",
       "      <td>0.003182</td>\n",
       "      <td>0.002036</td>\n",
       "      <td>0.001968</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>0.015601</td>\n",
       "      <td>0.029468</td>\n",
       "      <td>0.053033</td>\n",
       "      <td>0.011858</td>\n",
       "      <td>0.016819</td>\n",
       "      <td>0.125936</td>\n",
       "      <td>0.063508</td>\n",
       "      <td>0.006490</td>\n",
       "      <td>0.004538</td>\n",
       "      <td>0.005119</td>\n",
       "      <td>0.003709</td>\n",
       "      <td>0.004552</td>\n",
       "      <td>0.003197</td>\n",
       "      <td>0.004199</td>\n",
       "      <td>0.006508</td>\n",
       "      <td>0.006778</td>\n",
       "      <td>0.007274</td>\n",
       "      <td>0.009653</td>\n",
       "      <td>0.025042</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.010476</td>\n",
       "      <td>0.002064</td>\n",
       "      <td>0.004736</td>\n",
       "      <td>0.002294</td>\n",
       "      <td>4.709899e-04</td>\n",
       "      <td>0.003747</td>\n",
       "      <td>0.002080</td>\n",
       "      <td>0.007808</td>\n",
       "      <td>0.051914</td>\n",
       "      <td>0.003169</td>\n",
       "      <td>0.002024</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.003863</td>\n",
       "      <td>0.002706</td>\n",
       "      <td>0.000241</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>0.003014</td>\n",
       "      <td>0.002476</td>\n",
       "      <td>0.001413</td>\n",
       "      <td>0.001946</td>\n",
       "      <td>0.001280</td>\n",
       "      <td>0.001921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004831</td>\n",
       "      <td>0.001237</td>\n",
       "      <td>0.007238</td>\n",
       "      <td>0.014955</td>\n",
       "      <td>0.006299</td>\n",
       "      <td>0.005428</td>\n",
       "      <td>0.042809</td>\n",
       "      <td>0.013324</td>\n",
       "      <td>0.009774</td>\n",
       "      <td>0.018999</td>\n",
       "      <td>0.016906</td>\n",
       "      <td>0.016497</td>\n",
       "      <td>0.015769</td>\n",
       "      <td>0.015358</td>\n",
       "      <td>0.015201</td>\n",
       "      <td>0.019890</td>\n",
       "      <td>0.018986</td>\n",
       "      <td>0.015196</td>\n",
       "      <td>0.012471</td>\n",
       "      <td>0.014936</td>\n",
       "      <td>0.010181</td>\n",
       "      <td>0.018489</td>\n",
       "      <td>0.012970</td>\n",
       "      <td>0.015334</td>\n",
       "      <td>1.264738e-02</td>\n",
       "      <td>0.015273</td>\n",
       "      <td>0.011626</td>\n",
       "      <td>0.014952</td>\n",
       "      <td>0.016100</td>\n",
       "      <td>0.011959</td>\n",
       "      <td>0.008958</td>\n",
       "      <td>0.021323</td>\n",
       "      <td>0.013904</td>\n",
       "      <td>0.017438</td>\n",
       "      <td>0.026334</td>\n",
       "      <td>0.017067</td>\n",
       "      <td>0.012287</td>\n",
       "      <td>0.013811</td>\n",
       "      <td>0.015289</td>\n",
       "      <td>0.017314</td>\n",
       "      <td>0.007265</td>\n",
       "      <td>0.013958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>38</td>\n",
       "      <td>0.002984</td>\n",
       "      <td>0.015327</td>\n",
       "      <td>0.007291</td>\n",
       "      <td>0.005791</td>\n",
       "      <td>0.013113</td>\n",
       "      <td>0.015017</td>\n",
       "      <td>0.028256</td>\n",
       "      <td>0.104313</td>\n",
       "      <td>0.017214</td>\n",
       "      <td>0.016947</td>\n",
       "      <td>0.045372</td>\n",
       "      <td>0.008737</td>\n",
       "      <td>0.008832</td>\n",
       "      <td>0.008744</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.007902</td>\n",
       "      <td>0.007329</td>\n",
       "      <td>0.005381</td>\n",
       "      <td>0.000881</td>\n",
       "      <td>0.004174</td>\n",
       "      <td>0.001654</td>\n",
       "      <td>0.006909</td>\n",
       "      <td>0.009680</td>\n",
       "      <td>0.011118</td>\n",
       "      <td>2.458083e-03</td>\n",
       "      <td>0.005156</td>\n",
       "      <td>0.004924</td>\n",
       "      <td>0.010508</td>\n",
       "      <td>0.007123</td>\n",
       "      <td>0.003579</td>\n",
       "      <td>0.001895</td>\n",
       "      <td>0.005913</td>\n",
       "      <td>0.012561</td>\n",
       "      <td>0.011679</td>\n",
       "      <td>0.024150</td>\n",
       "      <td>0.009111</td>\n",
       "      <td>0.013006</td>\n",
       "      <td>0.010513</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>0.009365</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>0.010042</td>\n",
       "      <td>0.026018</td>\n",
       "      <td>0.034661</td>\n",
       "      <td>0.000700</td>\n",
       "      <td>0.001941</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.039996</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>0.000864</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.000918</td>\n",
       "      <td>0.000709</td>\n",
       "      <td>0.000973</td>\n",
       "      <td>0.000725</td>\n",
       "      <td>0.001127</td>\n",
       "      <td>0.000494</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.001904</td>\n",
       "      <td>0.002540</td>\n",
       "      <td>0.010631</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000897</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.001355</td>\n",
       "      <td>0.000637</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.000914</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>0.000693</td>\n",
       "      <td>0.000950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.000612</td>\n",
       "      <td>0.000516</td>\n",
       "      <td>0.000776</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>0.005503</td>\n",
       "      <td>0.004971</td>\n",
       "      <td>0.003548</td>\n",
       "      <td>0.005822</td>\n",
       "      <td>0.004744</td>\n",
       "      <td>0.004161</td>\n",
       "      <td>0.002542</td>\n",
       "      <td>0.007354</td>\n",
       "      <td>0.093655</td>\n",
       "      <td>0.010580</td>\n",
       "      <td>0.008660</td>\n",
       "      <td>0.011068</td>\n",
       "      <td>0.010090</td>\n",
       "      <td>0.012253</td>\n",
       "      <td>0.009013</td>\n",
       "      <td>0.009689</td>\n",
       "      <td>0.009877</td>\n",
       "      <td>0.006306</td>\n",
       "      <td>0.009561</td>\n",
       "      <td>0.008840</td>\n",
       "      <td>0.018755</td>\n",
       "      <td>0.016242</td>\n",
       "      <td>0.015312</td>\n",
       "      <td>0.005352</td>\n",
       "      <td>8.766048e-02</td>\n",
       "      <td>0.006582</td>\n",
       "      <td>0.000965</td>\n",
       "      <td>0.006136</td>\n",
       "      <td>0.005109</td>\n",
       "      <td>0.005031</td>\n",
       "      <td>0.003614</td>\n",
       "      <td>0.013814</td>\n",
       "      <td>0.026101</td>\n",
       "      <td>0.008464</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.049021</td>\n",
       "      <td>0.003964</td>\n",
       "      <td>0.021643</td>\n",
       "      <td>0.017115</td>\n",
       "      <td>0.004798</td>\n",
       "      <td>0.003598</td>\n",
       "      <td>0.005498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>32</td>\n",
       "      <td>0.002357</td>\n",
       "      <td>0.000661</td>\n",
       "      <td>0.002660</td>\n",
       "      <td>0.005078</td>\n",
       "      <td>0.004763</td>\n",
       "      <td>0.000666</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.010805</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>0.046585</td>\n",
       "      <td>0.013480</td>\n",
       "      <td>0.011606</td>\n",
       "      <td>0.016561</td>\n",
       "      <td>0.008580</td>\n",
       "      <td>0.017896</td>\n",
       "      <td>0.015488</td>\n",
       "      <td>0.010560</td>\n",
       "      <td>0.008081</td>\n",
       "      <td>0.001019</td>\n",
       "      <td>0.010691</td>\n",
       "      <td>0.006602</td>\n",
       "      <td>0.015584</td>\n",
       "      <td>0.014367</td>\n",
       "      <td>0.033849</td>\n",
       "      <td>4.266287e-03</td>\n",
       "      <td>0.009154</td>\n",
       "      <td>0.014316</td>\n",
       "      <td>0.014305</td>\n",
       "      <td>0.009372</td>\n",
       "      <td>0.001682</td>\n",
       "      <td>0.001565</td>\n",
       "      <td>0.002782</td>\n",
       "      <td>0.005907</td>\n",
       "      <td>0.044818</td>\n",
       "      <td>0.018351</td>\n",
       "      <td>0.004746</td>\n",
       "      <td>0.008106</td>\n",
       "      <td>0.017760</td>\n",
       "      <td>0.008829</td>\n",
       "      <td>0.008785</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005149</td>\n",
       "      <td>0.001335</td>\n",
       "      <td>0.008332</td>\n",
       "      <td>0.001858</td>\n",
       "      <td>0.017344</td>\n",
       "      <td>0.022282</td>\n",
       "      <td>0.015581</td>\n",
       "      <td>0.007646</td>\n",
       "      <td>0.061410</td>\n",
       "      <td>0.023523</td>\n",
       "      <td>0.005288</td>\n",
       "      <td>0.011936</td>\n",
       "      <td>0.011803</td>\n",
       "      <td>0.007843</td>\n",
       "      <td>0.023999</td>\n",
       "      <td>0.007527</td>\n",
       "      <td>0.002020</td>\n",
       "      <td>0.008303</td>\n",
       "      <td>0.014339</td>\n",
       "      <td>0.010217</td>\n",
       "      <td>0.008608</td>\n",
       "      <td>0.004586</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.006978</td>\n",
       "      <td>0.003033</td>\n",
       "      <td>0.004380</td>\n",
       "      <td>0.009826</td>\n",
       "      <td>0.003670</td>\n",
       "      <td>0.003970</td>\n",
       "      <td>0.005145</td>\n",
       "      <td>0.005947</td>\n",
       "      <td>0.008177</td>\n",
       "      <td>0.033739</td>\n",
       "      <td>0.006226</td>\n",
       "      <td>0.006631</td>\n",
       "      <td>0.003026</td>\n",
       "      <td>0.006739</td>\n",
       "      <td>0.004216</td>\n",
       "      <td>0.004712</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>27</td>\n",
       "      <td>0.002619</td>\n",
       "      <td>0.003727</td>\n",
       "      <td>0.001753</td>\n",
       "      <td>0.005362</td>\n",
       "      <td>0.003686</td>\n",
       "      <td>0.002151</td>\n",
       "      <td>0.030106</td>\n",
       "      <td>0.004513</td>\n",
       "      <td>0.017063</td>\n",
       "      <td>0.011274</td>\n",
       "      <td>0.015639</td>\n",
       "      <td>0.051975</td>\n",
       "      <td>0.012616</td>\n",
       "      <td>0.021316</td>\n",
       "      <td>0.017783</td>\n",
       "      <td>0.023027</td>\n",
       "      <td>0.023274</td>\n",
       "      <td>0.014266</td>\n",
       "      <td>0.003262</td>\n",
       "      <td>0.021104</td>\n",
       "      <td>0.005124</td>\n",
       "      <td>0.015924</td>\n",
       "      <td>0.023766</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.004374</td>\n",
       "      <td>0.015744</td>\n",
       "      <td>0.015697</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.001797</td>\n",
       "      <td>0.005745</td>\n",
       "      <td>0.006434</td>\n",
       "      <td>0.008450</td>\n",
       "      <td>0.011377</td>\n",
       "      <td>0.012686</td>\n",
       "      <td>0.003449</td>\n",
       "      <td>0.007249</td>\n",
       "      <td>0.005649</td>\n",
       "      <td>0.002325</td>\n",
       "      <td>0.010392</td>\n",
       "      <td>0.000018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>26</td>\n",
       "      <td>0.002685</td>\n",
       "      <td>0.004713</td>\n",
       "      <td>0.001862</td>\n",
       "      <td>0.014777</td>\n",
       "      <td>0.003262</td>\n",
       "      <td>0.006229</td>\n",
       "      <td>0.011269</td>\n",
       "      <td>0.009096</td>\n",
       "      <td>0.019509</td>\n",
       "      <td>0.015004</td>\n",
       "      <td>0.010441</td>\n",
       "      <td>0.025146</td>\n",
       "      <td>0.015887</td>\n",
       "      <td>0.012120</td>\n",
       "      <td>0.009642</td>\n",
       "      <td>0.012163</td>\n",
       "      <td>0.011361</td>\n",
       "      <td>0.007889</td>\n",
       "      <td>0.001690</td>\n",
       "      <td>0.010427</td>\n",
       "      <td>0.011241</td>\n",
       "      <td>0.012197</td>\n",
       "      <td>0.008314</td>\n",
       "      <td>0.003635</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.003645</td>\n",
       "      <td>0.007680</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>0.009233</td>\n",
       "      <td>0.004706</td>\n",
       "      <td>0.003533</td>\n",
       "      <td>0.021326</td>\n",
       "      <td>0.011030</td>\n",
       "      <td>0.008461</td>\n",
       "      <td>0.008776</td>\n",
       "      <td>0.018320</td>\n",
       "      <td>0.018552</td>\n",
       "      <td>0.012593</td>\n",
       "      <td>0.010363</td>\n",
       "      <td>0.012605</td>\n",
       "      <td>0.004404</td>\n",
       "      <td>0.006856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>37</td>\n",
       "      <td>0.006129</td>\n",
       "      <td>0.001525</td>\n",
       "      <td>0.003586</td>\n",
       "      <td>0.014037</td>\n",
       "      <td>0.004350</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.007254</td>\n",
       "      <td>0.017148</td>\n",
       "      <td>0.030966</td>\n",
       "      <td>0.010404</td>\n",
       "      <td>0.018616</td>\n",
       "      <td>0.043553</td>\n",
       "      <td>0.025761</td>\n",
       "      <td>0.035113</td>\n",
       "      <td>0.026167</td>\n",
       "      <td>0.018491</td>\n",
       "      <td>0.020786</td>\n",
       "      <td>0.003878</td>\n",
       "      <td>0.022102</td>\n",
       "      <td>0.019332</td>\n",
       "      <td>0.033067</td>\n",
       "      <td>0.014019</td>\n",
       "      <td>0.036732</td>\n",
       "      <td>2.977250e-02</td>\n",
       "      <td>0.010993</td>\n",
       "      <td>0.028490</td>\n",
       "      <td>0.022923</td>\n",
       "      <td>0.007094</td>\n",
       "      <td>0.004071</td>\n",
       "      <td>0.000883</td>\n",
       "      <td>0.005883</td>\n",
       "      <td>0.004062</td>\n",
       "      <td>0.043550</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>0.012270</td>\n",
       "      <td>0.008078</td>\n",
       "      <td>0.011307</td>\n",
       "      <td>0.007066</td>\n",
       "      <td>0.003028</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>0.002895</td>\n",
       "      <td>0.005139</td>\n",
       "      <td>0.003636</td>\n",
       "      <td>0.018081</td>\n",
       "      <td>0.005380</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011104</td>\n",
       "      <td>0.013222</td>\n",
       "      <td>0.027899</td>\n",
       "      <td>0.007674</td>\n",
       "      <td>0.016596</td>\n",
       "      <td>0.087128</td>\n",
       "      <td>0.018195</td>\n",
       "      <td>0.019698</td>\n",
       "      <td>0.019645</td>\n",
       "      <td>0.011624</td>\n",
       "      <td>0.010983</td>\n",
       "      <td>0.008680</td>\n",
       "      <td>0.014280</td>\n",
       "      <td>0.003941</td>\n",
       "      <td>0.012801</td>\n",
       "      <td>0.012177</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>8.281440e-03</td>\n",
       "      <td>0.011035</td>\n",
       "      <td>0.026046</td>\n",
       "      <td>0.030679</td>\n",
       "      <td>0.005444</td>\n",
       "      <td>0.009548</td>\n",
       "      <td>0.009983</td>\n",
       "      <td>0.014836</td>\n",
       "      <td>0.008812</td>\n",
       "      <td>0.015839</td>\n",
       "      <td>0.052187</td>\n",
       "      <td>0.020165</td>\n",
       "      <td>0.018246</td>\n",
       "      <td>0.037686</td>\n",
       "      <td>0.017439</td>\n",
       "      <td>0.035924</td>\n",
       "      <td>0.004585</td>\n",
       "      <td>0.026493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>0.003004</td>\n",
       "      <td>0.002332</td>\n",
       "      <td>0.006356</td>\n",
       "      <td>0.007666</td>\n",
       "      <td>0.005608</td>\n",
       "      <td>0.002332</td>\n",
       "      <td>0.023664</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>0.011981</td>\n",
       "      <td>0.025315</td>\n",
       "      <td>0.021178</td>\n",
       "      <td>0.027493</td>\n",
       "      <td>0.020268</td>\n",
       "      <td>0.050045</td>\n",
       "      <td>0.025557</td>\n",
       "      <td>0.028673</td>\n",
       "      <td>0.027159</td>\n",
       "      <td>0.026575</td>\n",
       "      <td>0.005439</td>\n",
       "      <td>0.019604</td>\n",
       "      <td>0.020494</td>\n",
       "      <td>0.023567</td>\n",
       "      <td>0.028631</td>\n",
       "      <td>0.003448</td>\n",
       "      <td>3.724145e-02</td>\n",
       "      <td>0.019643</td>\n",
       "      <td>0.007581</td>\n",
       "      <td>0.023601</td>\n",
       "      <td>0.006305</td>\n",
       "      <td>0.008181</td>\n",
       "      <td>0.007223</td>\n",
       "      <td>0.008001</td>\n",
       "      <td>0.009805</td>\n",
       "      <td>0.012355</td>\n",
       "      <td>0.029097</td>\n",
       "      <td>0.026537</td>\n",
       "      <td>0.012839</td>\n",
       "      <td>0.008255</td>\n",
       "      <td>0.004668</td>\n",
       "      <td>0.009286</td>\n",
       "      <td>0.009435</td>\n",
       "      <td>0.000704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000870</td>\n",
       "      <td>0.000264</td>\n",
       "      <td>0.003455</td>\n",
       "      <td>0.004486</td>\n",
       "      <td>0.003805</td>\n",
       "      <td>0.000286</td>\n",
       "      <td>0.001895</td>\n",
       "      <td>0.004295</td>\n",
       "      <td>0.019218</td>\n",
       "      <td>0.018581</td>\n",
       "      <td>0.010083</td>\n",
       "      <td>0.014595</td>\n",
       "      <td>0.014525</td>\n",
       "      <td>0.019771</td>\n",
       "      <td>0.078084</td>\n",
       "      <td>0.044927</td>\n",
       "      <td>0.014128</td>\n",
       "      <td>0.025925</td>\n",
       "      <td>0.011205</td>\n",
       "      <td>0.014240</td>\n",
       "      <td>0.009558</td>\n",
       "      <td>0.009776</td>\n",
       "      <td>0.013238</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.011245</td>\n",
       "      <td>0.003679</td>\n",
       "      <td>0.017921</td>\n",
       "      <td>0.005703</td>\n",
       "      <td>0.001859</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.014647</td>\n",
       "      <td>0.003289</td>\n",
       "      <td>0.005922</td>\n",
       "      <td>0.017998</td>\n",
       "      <td>0.001351</td>\n",
       "      <td>0.004002</td>\n",
       "      <td>0.004212</td>\n",
       "      <td>0.003250</td>\n",
       "      <td>0.002405</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>0.004228</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.002365</td>\n",
       "      <td>0.007646</td>\n",
       "      <td>0.003725</td>\n",
       "      <td>0.003119</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005732</td>\n",
       "      <td>0.009388</td>\n",
       "      <td>0.018852</td>\n",
       "      <td>0.005405</td>\n",
       "      <td>0.014532</td>\n",
       "      <td>0.009130</td>\n",
       "      <td>0.025307</td>\n",
       "      <td>0.031224</td>\n",
       "      <td>0.045235</td>\n",
       "      <td>0.020942</td>\n",
       "      <td>0.077240</td>\n",
       "      <td>0.036374</td>\n",
       "      <td>0.023349</td>\n",
       "      <td>0.010947</td>\n",
       "      <td>0.020304</td>\n",
       "      <td>0.021025</td>\n",
       "      <td>0.000970</td>\n",
       "      <td>1.412302e-02</td>\n",
       "      <td>0.009610</td>\n",
       "      <td>0.004241</td>\n",
       "      <td>0.025455</td>\n",
       "      <td>0.006155</td>\n",
       "      <td>0.001063</td>\n",
       "      <td>0.001538</td>\n",
       "      <td>0.001367</td>\n",
       "      <td>0.003342</td>\n",
       "      <td>0.006320</td>\n",
       "      <td>0.005240</td>\n",
       "      <td>0.000492</td>\n",
       "      <td>0.003393</td>\n",
       "      <td>0.004874</td>\n",
       "      <td>0.002646</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>0.001090</td>\n",
       "      <td>0.015124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0.031163</td>\n",
       "      <td>0.037378</td>\n",
       "      <td>0.015377</td>\n",
       "      <td>0.057391</td>\n",
       "      <td>0.002099</td>\n",
       "      <td>0.031820</td>\n",
       "      <td>0.002747</td>\n",
       "      <td>0.001319</td>\n",
       "      <td>0.001792</td>\n",
       "      <td>0.005340</td>\n",
       "      <td>0.001910</td>\n",
       "      <td>0.003595</td>\n",
       "      <td>0.002405</td>\n",
       "      <td>0.004156</td>\n",
       "      <td>0.008519</td>\n",
       "      <td>0.003893</td>\n",
       "      <td>0.002347</td>\n",
       "      <td>0.005710</td>\n",
       "      <td>0.074739</td>\n",
       "      <td>0.020427</td>\n",
       "      <td>0.007097</td>\n",
       "      <td>0.003376</td>\n",
       "      <td>0.015303</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>6.356093e-04</td>\n",
       "      <td>0.000479</td>\n",
       "      <td>0.008951</td>\n",
       "      <td>0.003442</td>\n",
       "      <td>0.003413</td>\n",
       "      <td>0.002122</td>\n",
       "      <td>0.008090</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>0.005965</td>\n",
       "      <td>0.001050</td>\n",
       "      <td>0.012717</td>\n",
       "      <td>0.000493</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>0.002633</td>\n",
       "      <td>0.001827</td>\n",
       "      <td>0.001898</td>\n",
       "      <td>0.000909</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>30</td>\n",
       "      <td>0.012632</td>\n",
       "      <td>0.004357</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>0.012891</td>\n",
       "      <td>0.002499</td>\n",
       "      <td>0.016007</td>\n",
       "      <td>0.002550</td>\n",
       "      <td>0.005409</td>\n",
       "      <td>0.005386</td>\n",
       "      <td>0.005901</td>\n",
       "      <td>0.004666</td>\n",
       "      <td>0.011086</td>\n",
       "      <td>0.004717</td>\n",
       "      <td>0.007230</td>\n",
       "      <td>0.009130</td>\n",
       "      <td>0.009982</td>\n",
       "      <td>0.006807</td>\n",
       "      <td>0.019259</td>\n",
       "      <td>0.034602</td>\n",
       "      <td>0.052767</td>\n",
       "      <td>0.005686</td>\n",
       "      <td>0.004455</td>\n",
       "      <td>0.013537</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.422100e-03</td>\n",
       "      <td>0.006395</td>\n",
       "      <td>0.002160</td>\n",
       "      <td>0.008303</td>\n",
       "      <td>0.008397</td>\n",
       "      <td>0.001968</td>\n",
       "      <td>0.001177</td>\n",
       "      <td>0.003123</td>\n",
       "      <td>0.000985</td>\n",
       "      <td>0.003179</td>\n",
       "      <td>0.008666</td>\n",
       "      <td>0.000887</td>\n",
       "      <td>0.003155</td>\n",
       "      <td>0.001525</td>\n",
       "      <td>0.005234</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>0.002174</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003209</td>\n",
       "      <td>0.010795</td>\n",
       "      <td>0.005465</td>\n",
       "      <td>0.042080</td>\n",
       "      <td>0.006191</td>\n",
       "      <td>0.014428</td>\n",
       "      <td>0.004383</td>\n",
       "      <td>0.005384</td>\n",
       "      <td>0.022716</td>\n",
       "      <td>0.020312</td>\n",
       "      <td>0.014833</td>\n",
       "      <td>0.023441</td>\n",
       "      <td>0.022165</td>\n",
       "      <td>0.021830</td>\n",
       "      <td>0.030705</td>\n",
       "      <td>0.034050</td>\n",
       "      <td>0.015475</td>\n",
       "      <td>0.024663</td>\n",
       "      <td>0.024133</td>\n",
       "      <td>0.071650</td>\n",
       "      <td>0.023465</td>\n",
       "      <td>0.021892</td>\n",
       "      <td>0.036927</td>\n",
       "      <td>0.004456</td>\n",
       "      <td>2.484060e-02</td>\n",
       "      <td>0.005602</td>\n",
       "      <td>0.009466</td>\n",
       "      <td>0.024718</td>\n",
       "      <td>0.010908</td>\n",
       "      <td>0.005612</td>\n",
       "      <td>0.015441</td>\n",
       "      <td>0.009490</td>\n",
       "      <td>0.012971</td>\n",
       "      <td>0.013677</td>\n",
       "      <td>0.016407</td>\n",
       "      <td>0.005222</td>\n",
       "      <td>0.012124</td>\n",
       "      <td>0.010440</td>\n",
       "      <td>0.009188</td>\n",
       "      <td>0.007100</td>\n",
       "      <td>0.003506</td>\n",
       "      <td>0.014644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>0.003391</td>\n",
       "      <td>0.007254</td>\n",
       "      <td>0.004649</td>\n",
       "      <td>0.009401</td>\n",
       "      <td>0.044719</td>\n",
       "      <td>0.001333</td>\n",
       "      <td>0.008400</td>\n",
       "      <td>0.020318</td>\n",
       "      <td>0.003554</td>\n",
       "      <td>0.015157</td>\n",
       "      <td>0.004750</td>\n",
       "      <td>0.003257</td>\n",
       "      <td>0.004925</td>\n",
       "      <td>0.003766</td>\n",
       "      <td>0.003597</td>\n",
       "      <td>0.005636</td>\n",
       "      <td>0.004945</td>\n",
       "      <td>0.005552</td>\n",
       "      <td>0.006500</td>\n",
       "      <td>0.075389</td>\n",
       "      <td>0.005409</td>\n",
       "      <td>0.009992</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>2.931361e-03</td>\n",
       "      <td>0.006633</td>\n",
       "      <td>0.005253</td>\n",
       "      <td>0.004226</td>\n",
       "      <td>0.003336</td>\n",
       "      <td>0.009447</td>\n",
       "      <td>0.001423</td>\n",
       "      <td>0.006400</td>\n",
       "      <td>0.006832</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.003963</td>\n",
       "      <td>0.012614</td>\n",
       "      <td>0.004590</td>\n",
       "      <td>0.006466</td>\n",
       "      <td>0.006169</td>\n",
       "      <td>0.003606</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>7</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006743</td>\n",
       "      <td>0.006461</td>\n",
       "      <td>0.004717</td>\n",
       "      <td>0.002232</td>\n",
       "      <td>0.000464</td>\n",
       "      <td>0.003508</td>\n",
       "      <td>0.015432</td>\n",
       "      <td>0.012939</td>\n",
       "      <td>0.009898</td>\n",
       "      <td>0.016662</td>\n",
       "      <td>0.011520</td>\n",
       "      <td>0.020080</td>\n",
       "      <td>0.014002</td>\n",
       "      <td>0.016670</td>\n",
       "      <td>0.009840</td>\n",
       "      <td>0.017272</td>\n",
       "      <td>0.003581</td>\n",
       "      <td>0.015964</td>\n",
       "      <td>0.012662</td>\n",
       "      <td>0.098446</td>\n",
       "      <td>0.031923</td>\n",
       "      <td>0.028133</td>\n",
       "      <td>4.063189e-02</td>\n",
       "      <td>0.005223</td>\n",
       "      <td>0.006173</td>\n",
       "      <td>0.009685</td>\n",
       "      <td>0.006139</td>\n",
       "      <td>0.002325</td>\n",
       "      <td>0.003318</td>\n",
       "      <td>0.012070</td>\n",
       "      <td>0.006914</td>\n",
       "      <td>0.018279</td>\n",
       "      <td>0.012279</td>\n",
       "      <td>0.008313</td>\n",
       "      <td>0.004683</td>\n",
       "      <td>0.005494</td>\n",
       "      <td>0.006619</td>\n",
       "      <td>0.004051</td>\n",
       "      <td>0.002248</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000231</td>\n",
       "      <td>0.001776</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.000611</td>\n",
       "      <td>0.002743</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000454</td>\n",
       "      <td>0.011499</td>\n",
       "      <td>0.009972</td>\n",
       "      <td>0.013202</td>\n",
       "      <td>0.019861</td>\n",
       "      <td>0.005993</td>\n",
       "      <td>0.005954</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>0.006986</td>\n",
       "      <td>0.005189</td>\n",
       "      <td>0.012679</td>\n",
       "      <td>0.003829</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>0.003603</td>\n",
       "      <td>0.024249</td>\n",
       "      <td>0.012850</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>0.128229</td>\n",
       "      <td>2.151660e-03</td>\n",
       "      <td>0.005480</td>\n",
       "      <td>0.007371</td>\n",
       "      <td>0.001715</td>\n",
       "      <td>0.007800</td>\n",
       "      <td>0.005614</td>\n",
       "      <td>0.021080</td>\n",
       "      <td>0.006178</td>\n",
       "      <td>0.006874</td>\n",
       "      <td>0.075474</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.004063</td>\n",
       "      <td>0.004266</td>\n",
       "      <td>0.002911</td>\n",
       "      <td>0.006214</td>\n",
       "      <td>0.003424</td>\n",
       "      <td>0.001644</td>\n",
       "      <td>0.000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000374</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>0.001483</td>\n",
       "      <td>0.001946</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.000198</td>\n",
       "      <td>0.000964</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.008333</td>\n",
       "      <td>0.009214</td>\n",
       "      <td>0.003892</td>\n",
       "      <td>0.006347</td>\n",
       "      <td>0.009242</td>\n",
       "      <td>0.006363</td>\n",
       "      <td>0.005526</td>\n",
       "      <td>0.009229</td>\n",
       "      <td>0.012236</td>\n",
       "      <td>0.006906</td>\n",
       "      <td>0.002169</td>\n",
       "      <td>0.005820</td>\n",
       "      <td>0.002597</td>\n",
       "      <td>0.006962</td>\n",
       "      <td>0.005227</td>\n",
       "      <td>0.017293</td>\n",
       "      <td>8.650326e-02</td>\n",
       "      <td>0.008001</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>0.007510</td>\n",
       "      <td>0.005317</td>\n",
       "      <td>0.007727</td>\n",
       "      <td>0.013663</td>\n",
       "      <td>0.008322</td>\n",
       "      <td>0.008354</td>\n",
       "      <td>0.012819</td>\n",
       "      <td>0.016396</td>\n",
       "      <td>0.005037</td>\n",
       "      <td>0.025522</td>\n",
       "      <td>0.027315</td>\n",
       "      <td>0.011950</td>\n",
       "      <td>0.017137</td>\n",
       "      <td>0.021110</td>\n",
       "      <td>0.007271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>13</td>\n",
       "      <td>0.001251</td>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.001518</td>\n",
       "      <td>0.002308</td>\n",
       "      <td>0.001111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006238</td>\n",
       "      <td>0.002576</td>\n",
       "      <td>0.001849</td>\n",
       "      <td>0.007898</td>\n",
       "      <td>0.004352</td>\n",
       "      <td>0.005649</td>\n",
       "      <td>0.005759</td>\n",
       "      <td>0.008449</td>\n",
       "      <td>0.004576</td>\n",
       "      <td>0.007966</td>\n",
       "      <td>0.003033</td>\n",
       "      <td>0.006768</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.025202</td>\n",
       "      <td>0.004593</td>\n",
       "      <td>0.003151</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.732154e-03</td>\n",
       "      <td>0.092212</td>\n",
       "      <td>0.019396</td>\n",
       "      <td>0.011659</td>\n",
       "      <td>0.006270</td>\n",
       "      <td>0.001193</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.017985</td>\n",
       "      <td>0.001241</td>\n",
       "      <td>0.006759</td>\n",
       "      <td>0.028372</td>\n",
       "      <td>0.008770</td>\n",
       "      <td>0.008437</td>\n",
       "      <td>0.009677</td>\n",
       "      <td>0.004669</td>\n",
       "      <td>0.002782</td>\n",
       "      <td>0.001134</td>\n",
       "      <td>0.000249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>0.002154</td>\n",
       "      <td>0.000250</td>\n",
       "      <td>0.001455</td>\n",
       "      <td>0.010430</td>\n",
       "      <td>0.000735</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001621</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.003324</td>\n",
       "      <td>0.001055</td>\n",
       "      <td>0.002660</td>\n",
       "      <td>0.004962</td>\n",
       "      <td>0.002605</td>\n",
       "      <td>0.001867</td>\n",
       "      <td>0.000880</td>\n",
       "      <td>0.001060</td>\n",
       "      <td>0.001139</td>\n",
       "      <td>0.004029</td>\n",
       "      <td>0.002540</td>\n",
       "      <td>0.001874</td>\n",
       "      <td>0.002350</td>\n",
       "      <td>0.004662</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>9.889729e-06</td>\n",
       "      <td>0.018552</td>\n",
       "      <td>0.115779</td>\n",
       "      <td>0.008614</td>\n",
       "      <td>0.002085</td>\n",
       "      <td>0.000764</td>\n",
       "      <td>0.000845</td>\n",
       "      <td>0.000683</td>\n",
       "      <td>0.001197</td>\n",
       "      <td>0.002102</td>\n",
       "      <td>0.002466</td>\n",
       "      <td>0.001877</td>\n",
       "      <td>0.004213</td>\n",
       "      <td>0.003791</td>\n",
       "      <td>0.001714</td>\n",
       "      <td>0.001404</td>\n",
       "      <td>0.001908</td>\n",
       "      <td>0.001590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>0.008786</td>\n",
       "      <td>0.007900</td>\n",
       "      <td>0.006913</td>\n",
       "      <td>0.015062</td>\n",
       "      <td>0.009527</td>\n",
       "      <td>0.016152</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.008910</td>\n",
       "      <td>0.007131</td>\n",
       "      <td>0.017226</td>\n",
       "      <td>0.005772</td>\n",
       "      <td>0.010860</td>\n",
       "      <td>0.018627</td>\n",
       "      <td>0.014757</td>\n",
       "      <td>0.013562</td>\n",
       "      <td>0.013564</td>\n",
       "      <td>0.010113</td>\n",
       "      <td>0.014785</td>\n",
       "      <td>0.026514</td>\n",
       "      <td>0.016684</td>\n",
       "      <td>0.008537</td>\n",
       "      <td>0.005715</td>\n",
       "      <td>0.012519</td>\n",
       "      <td>0.002086</td>\n",
       "      <td>4.485012e-07</td>\n",
       "      <td>0.025996</td>\n",
       "      <td>0.031062</td>\n",
       "      <td>0.030288</td>\n",
       "      <td>0.024631</td>\n",
       "      <td>0.012115</td>\n",
       "      <td>0.002161</td>\n",
       "      <td>0.004566</td>\n",
       "      <td>0.003142</td>\n",
       "      <td>0.005541</td>\n",
       "      <td>0.019499</td>\n",
       "      <td>0.002029</td>\n",
       "      <td>0.001721</td>\n",
       "      <td>0.005731</td>\n",
       "      <td>0.006044</td>\n",
       "      <td>0.002374</td>\n",
       "      <td>0.007390</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>16</td>\n",
       "      <td>0.003888</td>\n",
       "      <td>0.001823</td>\n",
       "      <td>0.005976</td>\n",
       "      <td>0.010775</td>\n",
       "      <td>0.005325</td>\n",
       "      <td>0.005898</td>\n",
       "      <td>0.006465</td>\n",
       "      <td>0.005657</td>\n",
       "      <td>0.006330</td>\n",
       "      <td>0.016064</td>\n",
       "      <td>0.005181</td>\n",
       "      <td>0.013653</td>\n",
       "      <td>0.013613</td>\n",
       "      <td>0.020260</td>\n",
       "      <td>0.010452</td>\n",
       "      <td>0.019251</td>\n",
       "      <td>0.007429</td>\n",
       "      <td>0.020835</td>\n",
       "      <td>0.014591</td>\n",
       "      <td>0.012898</td>\n",
       "      <td>0.009953</td>\n",
       "      <td>0.012291</td>\n",
       "      <td>0.008709</td>\n",
       "      <td>0.011058</td>\n",
       "      <td>1.004192e-02</td>\n",
       "      <td>0.022119</td>\n",
       "      <td>0.010812</td>\n",
       "      <td>0.027924</td>\n",
       "      <td>0.017788</td>\n",
       "      <td>0.009603</td>\n",
       "      <td>0.005041</td>\n",
       "      <td>0.004261</td>\n",
       "      <td>0.006692</td>\n",
       "      <td>0.004856</td>\n",
       "      <td>0.006978</td>\n",
       "      <td>0.010155</td>\n",
       "      <td>0.006516</td>\n",
       "      <td>0.006560</td>\n",
       "      <td>0.003896</td>\n",
       "      <td>0.004245</td>\n",
       "      <td>0.002115</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "      <td>0.005803</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.000468</td>\n",
       "      <td>0.007171</td>\n",
       "      <td>0.005657</td>\n",
       "      <td>0.001195</td>\n",
       "      <td>0.007195</td>\n",
       "      <td>0.014805</td>\n",
       "      <td>0.008077</td>\n",
       "      <td>0.007729</td>\n",
       "      <td>0.020889</td>\n",
       "      <td>0.008545</td>\n",
       "      <td>0.003662</td>\n",
       "      <td>0.005611</td>\n",
       "      <td>0.004953</td>\n",
       "      <td>0.002194</td>\n",
       "      <td>0.008638</td>\n",
       "      <td>0.003270</td>\n",
       "      <td>0.001350</td>\n",
       "      <td>0.003437</td>\n",
       "      <td>0.001422</td>\n",
       "      <td>0.006077</td>\n",
       "      <td>0.001804</td>\n",
       "      <td>0.005117</td>\n",
       "      <td>1.227654e-03</td>\n",
       "      <td>0.002059</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.003056</td>\n",
       "      <td>0.071982</td>\n",
       "      <td>0.003140</td>\n",
       "      <td>0.005073</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.003954</td>\n",
       "      <td>0.008951</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>0.001171</td>\n",
       "      <td>0.003224</td>\n",
       "      <td>0.003054</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.001310</td>\n",
       "      <td>0.002697</td>\n",
       "      <td>0.000572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>22</td>\n",
       "      <td>0.001536</td>\n",
       "      <td>0.003794</td>\n",
       "      <td>0.001343</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>0.004313</td>\n",
       "      <td>0.001723</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.003443</td>\n",
       "      <td>0.003253</td>\n",
       "      <td>0.002264</td>\n",
       "      <td>0.004509</td>\n",
       "      <td>0.010360</td>\n",
       "      <td>0.002404</td>\n",
       "      <td>0.003702</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.008882</td>\n",
       "      <td>0.001693</td>\n",
       "      <td>0.002735</td>\n",
       "      <td>0.002655</td>\n",
       "      <td>0.004645</td>\n",
       "      <td>0.003636</td>\n",
       "      <td>0.003620</td>\n",
       "      <td>0.006590</td>\n",
       "      <td>1.562006e-03</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.005072</td>\n",
       "      <td>0.003434</td>\n",
       "      <td>0.002956</td>\n",
       "      <td>0.037968</td>\n",
       "      <td>0.004111</td>\n",
       "      <td>0.002431</td>\n",
       "      <td>0.003269</td>\n",
       "      <td>0.005270</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000320</td>\n",
       "      <td>0.009420</td>\n",
       "      <td>0.002391</td>\n",
       "      <td>0.003596</td>\n",
       "      <td>0.001316</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>0.000649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>15</td>\n",
       "      <td>0.001503</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>0.001757</td>\n",
       "      <td>0.002907</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.004518</td>\n",
       "      <td>0.009612</td>\n",
       "      <td>0.003877</td>\n",
       "      <td>0.005575</td>\n",
       "      <td>0.008098</td>\n",
       "      <td>0.005010</td>\n",
       "      <td>0.005064</td>\n",
       "      <td>0.005359</td>\n",
       "      <td>0.004123</td>\n",
       "      <td>0.010781</td>\n",
       "      <td>0.002354</td>\n",
       "      <td>0.005695</td>\n",
       "      <td>0.004617</td>\n",
       "      <td>0.005700</td>\n",
       "      <td>0.005871</td>\n",
       "      <td>0.004707</td>\n",
       "      <td>0.003462</td>\n",
       "      <td>9.445052e-05</td>\n",
       "      <td>0.004137</td>\n",
       "      <td>0.003621</td>\n",
       "      <td>0.003801</td>\n",
       "      <td>0.006968</td>\n",
       "      <td>0.046871</td>\n",
       "      <td>0.011266</td>\n",
       "      <td>0.010440</td>\n",
       "      <td>0.016029</td>\n",
       "      <td>0.006311</td>\n",
       "      <td>0.001971</td>\n",
       "      <td>0.001527</td>\n",
       "      <td>0.018571</td>\n",
       "      <td>0.004326</td>\n",
       "      <td>0.007603</td>\n",
       "      <td>0.007039</td>\n",
       "      <td>0.005917</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>24</td>\n",
       "      <td>0.002479</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.002817</td>\n",
       "      <td>0.005953</td>\n",
       "      <td>0.002316</td>\n",
       "      <td>0.004175</td>\n",
       "      <td>0.011399</td>\n",
       "      <td>0.003759</td>\n",
       "      <td>0.011954</td>\n",
       "      <td>0.016293</td>\n",
       "      <td>0.007691</td>\n",
       "      <td>0.016884</td>\n",
       "      <td>0.017026</td>\n",
       "      <td>0.017779</td>\n",
       "      <td>0.024130</td>\n",
       "      <td>0.031312</td>\n",
       "      <td>0.012454</td>\n",
       "      <td>0.018282</td>\n",
       "      <td>0.010040</td>\n",
       "      <td>0.011706</td>\n",
       "      <td>0.010083</td>\n",
       "      <td>0.021255</td>\n",
       "      <td>0.016545</td>\n",
       "      <td>0.006368</td>\n",
       "      <td>1.136601e-02</td>\n",
       "      <td>0.026263</td>\n",
       "      <td>0.018037</td>\n",
       "      <td>0.015889</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.009681</td>\n",
       "      <td>0.031673</td>\n",
       "      <td>0.012813</td>\n",
       "      <td>0.009514</td>\n",
       "      <td>0.009578</td>\n",
       "      <td>0.007019</td>\n",
       "      <td>0.004488</td>\n",
       "      <td>0.007321</td>\n",
       "      <td>0.013815</td>\n",
       "      <td>0.010892</td>\n",
       "      <td>0.003244</td>\n",
       "      <td>0.007871</td>\n",
       "      <td>0.020603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>39</td>\n",
       "      <td>0.004907</td>\n",
       "      <td>0.007882</td>\n",
       "      <td>0.005158</td>\n",
       "      <td>0.004248</td>\n",
       "      <td>0.004013</td>\n",
       "      <td>0.005876</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004438</td>\n",
       "      <td>0.010565</td>\n",
       "      <td>0.004637</td>\n",
       "      <td>0.005175</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>0.005557</td>\n",
       "      <td>0.006583</td>\n",
       "      <td>0.003998</td>\n",
       "      <td>0.007094</td>\n",
       "      <td>0.005253</td>\n",
       "      <td>0.006637</td>\n",
       "      <td>0.008432</td>\n",
       "      <td>0.014632</td>\n",
       "      <td>0.002685</td>\n",
       "      <td>0.012521</td>\n",
       "      <td>0.009855</td>\n",
       "      <td>0.009739</td>\n",
       "      <td>9.181121e-03</td>\n",
       "      <td>0.007412</td>\n",
       "      <td>0.005050</td>\n",
       "      <td>0.007820</td>\n",
       "      <td>0.014988</td>\n",
       "      <td>0.009338</td>\n",
       "      <td>0.104472</td>\n",
       "      <td>0.007809</td>\n",
       "      <td>0.005933</td>\n",
       "      <td>0.006879</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007539</td>\n",
       "      <td>0.009420</td>\n",
       "      <td>0.009348</td>\n",
       "      <td>0.017438</td>\n",
       "      <td>0.004732</td>\n",
       "      <td>0.006888</td>\n",
       "      <td>0.040894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0.002188</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.001863</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000616</td>\n",
       "      <td>0.001649</td>\n",
       "      <td>0.013398</td>\n",
       "      <td>0.006123</td>\n",
       "      <td>0.004466</td>\n",
       "      <td>0.007914</td>\n",
       "      <td>0.010261</td>\n",
       "      <td>0.003761</td>\n",
       "      <td>0.004412</td>\n",
       "      <td>0.002293</td>\n",
       "      <td>0.008816</td>\n",
       "      <td>0.001530</td>\n",
       "      <td>0.003051</td>\n",
       "      <td>0.003778</td>\n",
       "      <td>0.001645</td>\n",
       "      <td>0.010915</td>\n",
       "      <td>0.009478</td>\n",
       "      <td>0.005182</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001375</td>\n",
       "      <td>0.004451</td>\n",
       "      <td>0.005785</td>\n",
       "      <td>0.007838</td>\n",
       "      <td>0.003545</td>\n",
       "      <td>0.008996</td>\n",
       "      <td>0.077560</td>\n",
       "      <td>0.023216</td>\n",
       "      <td>0.006355</td>\n",
       "      <td>0.001691</td>\n",
       "      <td>0.031084</td>\n",
       "      <td>0.034143</td>\n",
       "      <td>0.033020</td>\n",
       "      <td>0.076647</td>\n",
       "      <td>0.044992</td>\n",
       "      <td>0.007364</td>\n",
       "      <td>0.005049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>19</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>0.004301</td>\n",
       "      <td>0.002100</td>\n",
       "      <td>0.000962</td>\n",
       "      <td>0.018583</td>\n",
       "      <td>0.003524</td>\n",
       "      <td>0.010751</td>\n",
       "      <td>0.004686</td>\n",
       "      <td>0.002164</td>\n",
       "      <td>0.005184</td>\n",
       "      <td>0.004450</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>0.010446</td>\n",
       "      <td>0.003300</td>\n",
       "      <td>0.010849</td>\n",
       "      <td>0.003983</td>\n",
       "      <td>0.001335</td>\n",
       "      <td>0.002799</td>\n",
       "      <td>0.005940</td>\n",
       "      <td>0.007988</td>\n",
       "      <td>0.007054</td>\n",
       "      <td>0.008968</td>\n",
       "      <td>2.450986e-03</td>\n",
       "      <td>0.004465</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>0.005114</td>\n",
       "      <td>0.005431</td>\n",
       "      <td>0.008107</td>\n",
       "      <td>0.002413</td>\n",
       "      <td>0.041366</td>\n",
       "      <td>0.125182</td>\n",
       "      <td>0.005650</td>\n",
       "      <td>0.004090</td>\n",
       "      <td>0.015171</td>\n",
       "      <td>0.004466</td>\n",
       "      <td>0.005931</td>\n",
       "      <td>0.009950</td>\n",
       "      <td>0.005255</td>\n",
       "      <td>0.003352</td>\n",
       "      <td>0.000094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>41</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.000375</td>\n",
       "      <td>0.000705</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000871</td>\n",
       "      <td>0.000236</td>\n",
       "      <td>0.031901</td>\n",
       "      <td>0.006057</td>\n",
       "      <td>0.002157</td>\n",
       "      <td>0.004989</td>\n",
       "      <td>0.005433</td>\n",
       "      <td>0.007036</td>\n",
       "      <td>0.006706</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003647</td>\n",
       "      <td>0.002968</td>\n",
       "      <td>0.003604</td>\n",
       "      <td>0.002279</td>\n",
       "      <td>0.002219</td>\n",
       "      <td>0.003166</td>\n",
       "      <td>0.001100</td>\n",
       "      <td>0.003948</td>\n",
       "      <td>0.021010</td>\n",
       "      <td>0.005061</td>\n",
       "      <td>1.084785e-03</td>\n",
       "      <td>0.005550</td>\n",
       "      <td>0.000973</td>\n",
       "      <td>0.003425</td>\n",
       "      <td>0.004322</td>\n",
       "      <td>0.007975</td>\n",
       "      <td>0.027454</td>\n",
       "      <td>0.003874</td>\n",
       "      <td>0.007558</td>\n",
       "      <td>0.051611</td>\n",
       "      <td>0.005238</td>\n",
       "      <td>0.078041</td>\n",
       "      <td>0.006573</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>0.013947</td>\n",
       "      <td>0.007533</td>\n",
       "      <td>0.004067</td>\n",
       "      <td>0.000009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "      <td>0.001208</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.011101</td>\n",
       "      <td>0.004073</td>\n",
       "      <td>0.002759</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.030902</td>\n",
       "      <td>0.016638</td>\n",
       "      <td>0.007384</td>\n",
       "      <td>0.012506</td>\n",
       "      <td>0.007956</td>\n",
       "      <td>0.010088</td>\n",
       "      <td>0.005665</td>\n",
       "      <td>0.004280</td>\n",
       "      <td>0.002158</td>\n",
       "      <td>0.003066</td>\n",
       "      <td>0.002602</td>\n",
       "      <td>0.003449</td>\n",
       "      <td>0.003676</td>\n",
       "      <td>0.002518</td>\n",
       "      <td>0.001798</td>\n",
       "      <td>0.007210</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>4.567760e-03</td>\n",
       "      <td>0.001255</td>\n",
       "      <td>0.001856</td>\n",
       "      <td>0.004561</td>\n",
       "      <td>0.003381</td>\n",
       "      <td>0.014285</td>\n",
       "      <td>0.002437</td>\n",
       "      <td>0.003139</td>\n",
       "      <td>0.005483</td>\n",
       "      <td>0.002755</td>\n",
       "      <td>0.012982</td>\n",
       "      <td>0.012367</td>\n",
       "      <td>0.039782</td>\n",
       "      <td>0.012846</td>\n",
       "      <td>0.006647</td>\n",
       "      <td>0.009521</td>\n",
       "      <td>0.002123</td>\n",
       "      <td>0.000206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000489</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.002095</td>\n",
       "      <td>0.002569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000682</td>\n",
       "      <td>0.002841</td>\n",
       "      <td>0.015941</td>\n",
       "      <td>0.004382</td>\n",
       "      <td>0.002594</td>\n",
       "      <td>0.004601</td>\n",
       "      <td>0.009447</td>\n",
       "      <td>0.004057</td>\n",
       "      <td>0.003658</td>\n",
       "      <td>0.004648</td>\n",
       "      <td>0.014961</td>\n",
       "      <td>0.001967</td>\n",
       "      <td>0.000520</td>\n",
       "      <td>0.002553</td>\n",
       "      <td>0.001609</td>\n",
       "      <td>0.005486</td>\n",
       "      <td>0.007524</td>\n",
       "      <td>0.004027</td>\n",
       "      <td>2.309467e-04</td>\n",
       "      <td>0.009231</td>\n",
       "      <td>0.008048</td>\n",
       "      <td>0.005033</td>\n",
       "      <td>0.000654</td>\n",
       "      <td>0.005222</td>\n",
       "      <td>0.002791</td>\n",
       "      <td>0.038283</td>\n",
       "      <td>0.020127</td>\n",
       "      <td>0.005593</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.028727</td>\n",
       "      <td>0.016776</td>\n",
       "      <td>0.089894</td>\n",
       "      <td>0.033938</td>\n",
       "      <td>0.011498</td>\n",
       "      <td>0.005139</td>\n",
       "      <td>0.000121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>0.002676</td>\n",
       "      <td>0.002221</td>\n",
       "      <td>0.002050</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002193</td>\n",
       "      <td>0.001145</td>\n",
       "      <td>0.004202</td>\n",
       "      <td>0.004522</td>\n",
       "      <td>0.004020</td>\n",
       "      <td>0.005658</td>\n",
       "      <td>0.002883</td>\n",
       "      <td>0.007014</td>\n",
       "      <td>0.004985</td>\n",
       "      <td>0.006217</td>\n",
       "      <td>0.005476</td>\n",
       "      <td>0.004040</td>\n",
       "      <td>0.013556</td>\n",
       "      <td>0.003878</td>\n",
       "      <td>0.002625</td>\n",
       "      <td>0.003382</td>\n",
       "      <td>0.002333</td>\n",
       "      <td>0.005521</td>\n",
       "      <td>0.005030</td>\n",
       "      <td>0.001764</td>\n",
       "      <td>2.868712e-04</td>\n",
       "      <td>0.003624</td>\n",
       "      <td>0.004730</td>\n",
       "      <td>0.002416</td>\n",
       "      <td>0.001750</td>\n",
       "      <td>0.009800</td>\n",
       "      <td>0.003024</td>\n",
       "      <td>0.014539</td>\n",
       "      <td>0.012149</td>\n",
       "      <td>0.002792</td>\n",
       "      <td>0.008759</td>\n",
       "      <td>0.004254</td>\n",
       "      <td>0.006191</td>\n",
       "      <td>0.017340</td>\n",
       "      <td>0.016219</td>\n",
       "      <td>0.025542</td>\n",
       "      <td>0.091743</td>\n",
       "      <td>0.004946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001015</td>\n",
       "      <td>0.000772</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0.001632</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.004861</td>\n",
       "      <td>0.004180</td>\n",
       "      <td>0.001928</td>\n",
       "      <td>0.002187</td>\n",
       "      <td>0.001789</td>\n",
       "      <td>0.005358</td>\n",
       "      <td>0.003258</td>\n",
       "      <td>0.005490</td>\n",
       "      <td>0.002427</td>\n",
       "      <td>0.001154</td>\n",
       "      <td>0.019303</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>0.000656</td>\n",
       "      <td>0.002425</td>\n",
       "      <td>0.001070</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>0.001973</td>\n",
       "      <td>0.000193</td>\n",
       "      <td>2.327494e-03</td>\n",
       "      <td>0.001713</td>\n",
       "      <td>0.004906</td>\n",
       "      <td>0.001570</td>\n",
       "      <td>0.003260</td>\n",
       "      <td>0.008483</td>\n",
       "      <td>0.001294</td>\n",
       "      <td>0.012950</td>\n",
       "      <td>0.006502</td>\n",
       "      <td>0.003452</td>\n",
       "      <td>0.003198</td>\n",
       "      <td>0.000502</td>\n",
       "      <td>0.003243</td>\n",
       "      <td>0.009363</td>\n",
       "      <td>0.016670</td>\n",
       "      <td>0.023373</td>\n",
       "      <td>0.107448</td>\n",
       "      <td>0.119257</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    order  topic       GAM       GEM       EMG       MGD       EMB       MSD       NHM       SRL       GER       NLS       PIG       GRS       ROT       LEF       IST       ARC       MSL       FSM       OVP       SRF       SSC       FLT       SEN       CDR           MTM       STT       RTH       TCM       CUL       PRA       CPR       WAT       TMP       LIT       MEC       MPH       NUT       HRM       CHS       MCH       PTH       OBI\n",
       "0       0      2  0.233855  0.216993  0.124526  0.139415  0.010228  0.055123  0.000000  0.005404  0.011408  0.005775  0.002957  0.007177  0.010170  0.007156  0.005666  0.006718  0.006163  0.007406  0.005776  0.014127  0.001352  0.002740  0.003543  0.000427  9.202232e-03  0.001941  0.009646  0.007024  0.003604  0.000939  0.004591  0.000369  0.002933  0.002393  0.000306  0.000000  0.001238  0.001822  0.000972  0.001199  0.001053  0.000000\n",
       "1       1      0  0.004134  0.009401  0.018687  0.001565  0.040489  0.011614  0.000562  0.002660  0.000542  0.000554  0.002779  0.002613  0.001889  0.001963  0.000612  0.000209  0.006118  0.000795  0.001652  0.000881  0.000398  0.002562  0.000172  0.000131  0.000000e+00  0.000017  0.000731  0.000427  0.004544  0.001261  0.000394  0.000061  0.001825  0.000219  0.000005  0.000010  0.000653  0.000636  0.000461  0.000477  0.000222  0.000000\n",
       "2       2     31  0.008770  0.013485  0.024390  0.009631  0.036758  0.027312  0.007177  0.013111  0.009136  0.004887  0.008099  0.004785  0.004810  0.005188  0.005258  0.004650  0.005122  0.008899  0.008305  0.007519  0.002659  0.003938  0.004301  0.000713  6.296012e-04  0.004960  0.005831  0.007089  0.003573  0.004343  0.001303  0.001765  0.006293  0.003178  0.001670  0.001055  0.004622  0.004412  0.003675  0.005247  0.002479  0.007914\n",
       "3       3     11  0.002796  0.008644  0.011394  0.001315  0.038993  0.006750  0.000255  0.002374  0.000437  0.000630  0.001491  0.000857  0.001067  0.000863  0.001919  0.000052  0.000485  0.000896  0.000302  0.000811  0.000327  0.001162  0.000051  0.000109  4.641771e-06  0.000009  0.000724  0.002106  0.005886  0.000637  0.000458  0.000054  0.001006  0.000215  0.000000  0.000000  0.001208  0.001016  0.000919  0.000233  0.000980  0.000010\n",
       "4       4      9  0.017270  0.049438  0.062597  0.009543  0.096471  0.043610  0.005191  0.027866  0.013209  0.017232  0.017679  0.008049  0.009300  0.010925  0.006650  0.006165  0.008678  0.008245  0.003203  0.006532  0.007222  0.006100  0.008059  0.001102  0.000000e+00  0.002005  0.007473  0.017546  0.007957  0.001027  0.001259  0.000119  0.004076  0.006525  0.022467  0.005462  0.003984  0.004436  0.003182  0.002036  0.001968  0.000000\n",
       "5       5     36  0.015601  0.029468  0.053033  0.011858  0.016819  0.125936  0.063508  0.006490  0.004538  0.005119  0.003709  0.004552  0.003197  0.004199  0.006508  0.006778  0.007274  0.009653  0.025042  0.006045  0.010476  0.002064  0.004736  0.002294  4.709899e-04  0.003747  0.002080  0.007808  0.051914  0.003169  0.002024  0.000201  0.003863  0.002706  0.000241  0.001119  0.003014  0.002476  0.001413  0.001946  0.001280  0.001921\n",
       "6       6      1  0.004831  0.001237  0.007238  0.014955  0.006299  0.005428  0.042809  0.013324  0.009774  0.018999  0.016906  0.016497  0.015769  0.015358  0.015201  0.019890  0.018986  0.015196  0.012471  0.014936  0.010181  0.018489  0.012970  0.015334  1.264738e-02  0.015273  0.011626  0.014952  0.016100  0.011959  0.008958  0.021323  0.013904  0.017438  0.026334  0.017067  0.012287  0.013811  0.015289  0.017314  0.007265  0.013958\n",
       "7       7     38  0.002984  0.015327  0.007291  0.005791  0.013113  0.015017  0.028256  0.104313  0.017214  0.016947  0.045372  0.008737  0.008832  0.008744  0.006065  0.007902  0.007329  0.005381  0.000881  0.004174  0.001654  0.006909  0.009680  0.011118  2.458083e-03  0.005156  0.004924  0.010508  0.007123  0.003579  0.001895  0.005913  0.012561  0.011679  0.024150  0.009111  0.013006  0.010513  0.006421  0.009365  0.003082  0.000000\n",
       "8       8      4  0.010042  0.026018  0.034661  0.000700  0.001941  0.000123  0.000677  0.039996  0.000271  0.000864  0.033981  0.000918  0.000709  0.000973  0.000725  0.001127  0.000494  0.001155  0.001904  0.002540  0.010631  0.000825  0.001209  0.000175  0.000000e+00  0.000897  0.001000  0.001355  0.000637  0.000544  0.000914  0.000480  0.000693  0.000950  0.000000  0.000461  0.000612  0.000516  0.000776  0.000174  0.000397  0.000000\n",
       "9       9     17  0.005503  0.004971  0.003548  0.005822  0.004744  0.004161  0.002542  0.007354  0.093655  0.010580  0.008660  0.011068  0.010090  0.012253  0.009013  0.009689  0.009877  0.006306  0.009561  0.008840  0.018755  0.016242  0.015312  0.005352  8.766048e-02  0.006582  0.000965  0.006136  0.005109  0.005031  0.003614  0.013814  0.026101  0.008464  0.003082  0.049021  0.003964  0.021643  0.017115  0.004798  0.003598  0.005498\n",
       "10     10     32  0.002357  0.000661  0.002660  0.005078  0.004763  0.000666  0.000640  0.010805  0.006291  0.046585  0.013480  0.011606  0.016561  0.008580  0.017896  0.015488  0.010560  0.008081  0.001019  0.010691  0.006602  0.015584  0.014367  0.033849  4.266287e-03  0.009154  0.014316  0.014305  0.009372  0.001682  0.001565  0.002782  0.005907  0.044818  0.018351  0.004746  0.008106  0.017760  0.008829  0.008785  0.001631  0.000000\n",
       "11     11     10  0.000729  0.000033  0.005149  0.001335  0.008332  0.001858  0.017344  0.022282  0.015581  0.007646  0.061410  0.023523  0.005288  0.011936  0.011803  0.007843  0.023999  0.007527  0.002020  0.008303  0.014339  0.010217  0.008608  0.004586  0.000000e+00  0.006978  0.003033  0.004380  0.009826  0.003670  0.003970  0.005145  0.005947  0.008177  0.033739  0.006226  0.006631  0.003026  0.006739  0.004216  0.004712  0.000000\n",
       "12     12     27  0.002619  0.003727  0.001753  0.005362  0.003686  0.002151  0.030106  0.004513  0.017063  0.011274  0.015639  0.051975  0.012616  0.021316  0.017783  0.023027  0.023274  0.014266  0.003262  0.021104  0.005124  0.015924  0.023766  0.000124  0.000000e+00  0.006496  0.004374  0.015744  0.015697  0.001125  0.001797  0.005745  0.006434  0.008450  0.011377  0.012686  0.003449  0.007249  0.005649  0.002325  0.010392  0.000018\n",
       "13     13     26  0.002685  0.004713  0.001862  0.014777  0.003262  0.006229  0.011269  0.009096  0.019509  0.015004  0.010441  0.025146  0.015887  0.012120  0.009642  0.012163  0.011361  0.007889  0.001690  0.010427  0.011241  0.012197  0.008314  0.003635  0.000000e+00  0.003645  0.007680  0.012778  0.009233  0.004706  0.003533  0.021326  0.011030  0.008461  0.008776  0.018320  0.018552  0.012593  0.010363  0.012605  0.004404  0.006856\n",
       "14     14     37  0.006129  0.001525  0.003586  0.014037  0.004350  0.005530  0.000607  0.007254  0.017148  0.030966  0.010404  0.018616  0.043553  0.025761  0.035113  0.026167  0.018491  0.020786  0.003878  0.022102  0.019332  0.033067  0.014019  0.036732  2.977250e-02  0.010993  0.028490  0.022923  0.007094  0.004071  0.000883  0.005883  0.004062  0.043550  0.008481  0.012270  0.008078  0.011307  0.007066  0.003028  0.001309  0.000000\n",
       "15     15      5  0.002895  0.005139  0.003636  0.018081  0.005380  0.000349  0.000000  0.011104  0.013222  0.027899  0.007674  0.016596  0.087128  0.018195  0.019698  0.019645  0.011624  0.010983  0.008680  0.014280  0.003941  0.012801  0.012177  0.000461  8.281440e-03  0.011035  0.026046  0.030679  0.005444  0.009548  0.009983  0.014836  0.008812  0.015839  0.052187  0.020165  0.018246  0.037686  0.017439  0.035924  0.004585  0.026493\n",
       "16     16     18  0.003004  0.002332  0.006356  0.007666  0.005608  0.002332  0.023664  0.012108  0.011981  0.025315  0.021178  0.027493  0.020268  0.050045  0.025557  0.028673  0.027159  0.026575  0.005439  0.019604  0.020494  0.023567  0.028631  0.003448  3.724145e-02  0.019643  0.007581  0.023601  0.006305  0.008181  0.007223  0.008001  0.009805  0.012355  0.029097  0.026537  0.012839  0.008255  0.004668  0.009286  0.009435  0.000704\n",
       "17     17     21  0.000870  0.000264  0.003455  0.004486  0.003805  0.000286  0.001895  0.004295  0.019218  0.018581  0.010083  0.014595  0.014525  0.019771  0.078084  0.044927  0.014128  0.025925  0.011205  0.014240  0.009558  0.009776  0.013238  0.000248  0.000000e+00  0.011245  0.003679  0.017921  0.005703  0.001859  0.001155  0.014647  0.003289  0.005922  0.017998  0.001351  0.004002  0.004212  0.003250  0.002405  0.000984  0.000000\n",
       "18     18     20  0.004228  0.000099  0.002365  0.007646  0.003725  0.003119  0.000000  0.005732  0.009388  0.018852  0.005405  0.014532  0.009130  0.025307  0.031224  0.045235  0.020942  0.077240  0.036374  0.023349  0.010947  0.020304  0.021025  0.000970  1.412302e-02  0.009610  0.004241  0.025455  0.006155  0.001063  0.001538  0.001367  0.003342  0.006320  0.005240  0.000492  0.003393  0.004874  0.002646  0.000507  0.001090  0.015124\n",
       "19     19     14  0.031163  0.037378  0.015377  0.057391  0.002099  0.031820  0.002747  0.001319  0.001792  0.005340  0.001910  0.003595  0.002405  0.004156  0.008519  0.003893  0.002347  0.005710  0.074739  0.020427  0.007097  0.003376  0.015303  0.000441  6.356093e-04  0.000479  0.008951  0.003442  0.003413  0.002122  0.008090  0.001182  0.005965  0.001050  0.012717  0.000493  0.007632  0.002633  0.001827  0.001898  0.000909  0.000000\n",
       "20     20     30  0.012632  0.004357  0.002999  0.012891  0.002499  0.016007  0.002550  0.005409  0.005386  0.005901  0.004666  0.011086  0.004717  0.007230  0.009130  0.009982  0.006807  0.019259  0.034602  0.052767  0.005686  0.004455  0.013537  0.000000  2.422100e-03  0.006395  0.002160  0.008303  0.008397  0.001968  0.001177  0.003123  0.000985  0.003179  0.008666  0.000887  0.003155  0.001525  0.005234  0.001511  0.002174  0.000000\n",
       "21     21     29  0.003209  0.010795  0.005465  0.042080  0.006191  0.014428  0.004383  0.005384  0.022716  0.020312  0.014833  0.023441  0.022165  0.021830  0.030705  0.034050  0.015475  0.024663  0.024133  0.071650  0.023465  0.021892  0.036927  0.004456  2.484060e-02  0.005602  0.009466  0.024718  0.010908  0.005612  0.015441  0.009490  0.012971  0.013677  0.016407  0.005222  0.012124  0.010440  0.009188  0.007100  0.003506  0.014644\n",
       "22     22     12  0.002175  0.003391  0.007254  0.004649  0.009401  0.044719  0.001333  0.008400  0.020318  0.003554  0.015157  0.004750  0.003257  0.004925  0.003766  0.003597  0.005636  0.004945  0.005552  0.006500  0.075389  0.005409  0.009992  0.000096  2.931361e-03  0.006633  0.005253  0.004226  0.003336  0.009447  0.001423  0.006400  0.006832  0.004115  0.003963  0.012614  0.004590  0.006466  0.006169  0.003606  0.001552  0.000000\n",
       "23     23      7  0.003922  0.000000  0.006743  0.006461  0.004717  0.002232  0.000464  0.003508  0.015432  0.012939  0.009898  0.016662  0.011520  0.020080  0.014002  0.016670  0.009840  0.017272  0.003581  0.015964  0.012662  0.098446  0.031923  0.028133  4.063189e-02  0.005223  0.006173  0.009685  0.006139  0.002325  0.003318  0.012070  0.006914  0.018279  0.012279  0.008313  0.004683  0.005494  0.006619  0.004051  0.002248  0.000000\n",
       "24     24      8  0.000231  0.001776  0.000931  0.000611  0.002743  0.000016  0.000454  0.011499  0.009972  0.013202  0.019861  0.005993  0.005954  0.005661  0.006986  0.005189  0.012679  0.003829  0.000990  0.003603  0.024249  0.012850  0.003350  0.128229  2.151660e-03  0.005480  0.007371  0.001715  0.007800  0.005614  0.021080  0.006178  0.006874  0.075474  0.000017  0.004063  0.004266  0.002911  0.006214  0.003424  0.001644  0.000023\n",
       "25     25     35  0.000374  0.000301  0.001483  0.001946  0.001590  0.000198  0.000964  0.002204  0.008333  0.009214  0.003892  0.006347  0.009242  0.006363  0.005526  0.009229  0.012236  0.006906  0.002169  0.005820  0.002597  0.006962  0.005227  0.017293  8.650326e-02  0.008001  0.008403  0.007510  0.005317  0.007727  0.013663  0.008322  0.008354  0.012819  0.016396  0.005037  0.025522  0.027315  0.011950  0.017137  0.021110  0.007271\n",
       "26     26     13  0.001251  0.000304  0.001518  0.002308  0.001111  0.000000  0.006238  0.002576  0.001849  0.007898  0.004352  0.005649  0.005759  0.008449  0.004576  0.007966  0.003033  0.006768  0.002446  0.004854  0.025202  0.004593  0.003151  0.000000  2.732154e-03  0.092212  0.019396  0.011659  0.006270  0.001193  0.001597  0.017985  0.001241  0.006759  0.028372  0.008770  0.008437  0.009677  0.004669  0.002782  0.001134  0.000249\n",
       "27     27     23  0.002154  0.000250  0.001455  0.010430  0.000735  0.000784  0.000000  0.001621  0.000342  0.003324  0.001055  0.002660  0.004962  0.002605  0.001867  0.000880  0.001060  0.001139  0.004029  0.002540  0.001874  0.002350  0.004662  0.000243  9.889729e-06  0.018552  0.115779  0.008614  0.002085  0.000764  0.000845  0.000683  0.001197  0.002102  0.002466  0.001877  0.004213  0.003791  0.001714  0.001404  0.001908  0.001590\n",
       "28     28     25  0.008786  0.007900  0.006913  0.015062  0.009527  0.016152  0.000175  0.008910  0.007131  0.017226  0.005772  0.010860  0.018627  0.014757  0.013562  0.013564  0.010113  0.014785  0.026514  0.016684  0.008537  0.005715  0.012519  0.002086  4.485012e-07  0.025996  0.031062  0.030288  0.024631  0.012115  0.002161  0.004566  0.003142  0.005541  0.019499  0.002029  0.001721  0.005731  0.006044  0.002374  0.007390  0.000000\n",
       "29     29     16  0.003888  0.001823  0.005976  0.010775  0.005325  0.005898  0.006465  0.005657  0.006330  0.016064  0.005181  0.013653  0.013613  0.020260  0.010452  0.019251  0.007429  0.020835  0.014591  0.012898  0.009953  0.012291  0.008709  0.011058  1.004192e-02  0.022119  0.010812  0.027924  0.017788  0.009603  0.005041  0.004261  0.006692  0.004856  0.006978  0.010155  0.006516  0.006560  0.003896  0.004245  0.002115  0.000000\n",
       "30     30     34  0.005803  0.000145  0.000468  0.007171  0.005657  0.001195  0.007195  0.014805  0.008077  0.007729  0.020889  0.008545  0.003662  0.005611  0.004953  0.002194  0.008638  0.003270  0.001350  0.003437  0.001422  0.006077  0.001804  0.005117  1.227654e-03  0.002059  0.001343  0.003056  0.071982  0.003140  0.005073  0.001340  0.003954  0.008951  0.000301  0.001171  0.003224  0.003054  0.002049  0.001310  0.002697  0.000572\n",
       "31     31     22  0.001536  0.003794  0.001343  0.001894  0.004313  0.001723  0.000042  0.003443  0.003253  0.002264  0.004509  0.010360  0.002404  0.003702  0.001485  0.001174  0.008882  0.001693  0.002735  0.002655  0.004645  0.003636  0.003620  0.006590  1.562006e-03  0.000629  0.005072  0.003434  0.002956  0.037968  0.004111  0.002431  0.003269  0.005270  0.000000  0.000320  0.009420  0.002391  0.003596  0.001316  0.001406  0.000649\n",
       "32     32     15  0.001503  0.000000  0.000473  0.001757  0.002907  0.000297  0.000318  0.004518  0.009612  0.003877  0.005575  0.008098  0.005010  0.005064  0.005359  0.004123  0.010781  0.002354  0.005695  0.004617  0.005700  0.005871  0.004707  0.003462  9.445052e-05  0.004137  0.003621  0.003801  0.006968  0.046871  0.011266  0.010440  0.016029  0.006311  0.001971  0.001527  0.018571  0.004326  0.007603  0.007039  0.005917  0.000000\n",
       "33     33     24  0.002479  0.000061  0.002817  0.005953  0.002316  0.004175  0.011399  0.003759  0.011954  0.016293  0.007691  0.016884  0.017026  0.017779  0.024130  0.031312  0.012454  0.018282  0.010040  0.011706  0.010083  0.021255  0.016545  0.006368  1.136601e-02  0.026263  0.018037  0.015889  0.008593  0.009681  0.031673  0.012813  0.009514  0.009578  0.007019  0.004488  0.007321  0.013815  0.010892  0.003244  0.007871  0.020603\n",
       "34     34     39  0.004907  0.007882  0.005158  0.004248  0.004013  0.005876  0.000000  0.004438  0.010565  0.004637  0.005175  0.007299  0.005557  0.006583  0.003998  0.007094  0.005253  0.006637  0.008432  0.014632  0.002685  0.012521  0.009855  0.009739  9.181121e-03  0.007412  0.005050  0.007820  0.014988  0.009338  0.104472  0.007809  0.005933  0.006879  0.000000  0.007539  0.009420  0.009348  0.017438  0.004732  0.006888  0.040894\n",
       "35     35     28  0.000061  0.000382  0.002188  0.001080  0.001863  0.000066  0.000616  0.001649  0.013398  0.006123  0.004466  0.007914  0.010261  0.003761  0.004412  0.002293  0.008816  0.001530  0.003051  0.003778  0.001645  0.010915  0.009478  0.005182  0.000000e+00  0.001375  0.004451  0.005785  0.007838  0.003545  0.008996  0.077560  0.023216  0.006355  0.001691  0.031084  0.034143  0.033020  0.076647  0.044992  0.007364  0.005049\n",
       "36     36     19  0.001084  0.000103  0.001992  0.004301  0.002100  0.000962  0.018583  0.003524  0.010751  0.004686  0.002164  0.005184  0.004450  0.004098  0.010446  0.003300  0.010849  0.003983  0.001335  0.002799  0.005940  0.007988  0.007054  0.008968  2.450986e-03  0.004465  0.008632  0.005114  0.005431  0.008107  0.002413  0.041366  0.125182  0.005650  0.004090  0.015171  0.004466  0.005931  0.009950  0.005255  0.003352  0.000094\n",
       "37     37     41  0.000163  0.000375  0.000705  0.001369  0.000871  0.000236  0.031901  0.006057  0.002157  0.004989  0.005433  0.007036  0.006706  0.003879  0.003647  0.002968  0.003604  0.002279  0.002219  0.003166  0.001100  0.003948  0.021010  0.005061  1.084785e-03  0.005550  0.000973  0.003425  0.004322  0.007975  0.027454  0.003874  0.007558  0.051611  0.005238  0.078041  0.006573  0.006410  0.013947  0.007533  0.004067  0.000009\n",
       "38     38     33  0.001208  0.000090  0.000124  0.011101  0.004073  0.002759  0.000144  0.030902  0.016638  0.007384  0.012506  0.007956  0.010088  0.005665  0.004280  0.002158  0.003066  0.002602  0.003449  0.003676  0.002518  0.001798  0.007210  0.000272  4.567760e-03  0.001255  0.001856  0.004561  0.003381  0.014285  0.002437  0.003139  0.005483  0.002755  0.012982  0.012367  0.039782  0.012846  0.006647  0.009521  0.002123  0.000206\n",
       "39     39      3  0.000088  0.000489  0.000273  0.002095  0.002569  0.000000  0.000682  0.002841  0.015941  0.004382  0.002594  0.004601  0.009447  0.004057  0.003658  0.004648  0.014961  0.001967  0.000520  0.002553  0.001609  0.005486  0.007524  0.004027  2.309467e-04  0.009231  0.008048  0.005033  0.000654  0.005222  0.002791  0.038283  0.020127  0.005593  0.002728  0.028727  0.016776  0.089894  0.033938  0.011498  0.005139  0.000121\n",
       "40     40     40  0.002676  0.002221  0.002050  0.001377  0.002193  0.001145  0.004202  0.004522  0.004020  0.005658  0.002883  0.007014  0.004985  0.006217  0.005476  0.004040  0.013556  0.003878  0.002625  0.003382  0.002333  0.005521  0.005030  0.001764  2.868712e-04  0.003624  0.004730  0.002416  0.001750  0.009800  0.003024  0.014539  0.012149  0.002792  0.008759  0.004254  0.006191  0.017340  0.016219  0.025542  0.091743  0.004946\n",
       "41     41      6  0.001015  0.000772  0.000536  0.000723  0.001632  0.002502  0.004861  0.004180  0.001928  0.002187  0.001789  0.005358  0.003258  0.005490  0.002427  0.001154  0.019303  0.001182  0.000656  0.002425  0.001070  0.003465  0.001973  0.000193  2.327494e-03  0.001713  0.004906  0.001570  0.003260  0.008483  0.001294  0.012950  0.006502  0.003452  0.003198  0.000502  0.003243  0.009363  0.016670  0.023373  0.107448  0.119257"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get a list of texts to create a topic model from, from one of the processed description dictionaries above. \n",
    "texts = [description for i,description in descriptions_linares_pontes.items()]\n",
    "\n",
    "# Creating and fitting the topic model, either NFM or LDA.\n",
    "number_of_topics = 42\n",
    "seed = 0\n",
    "vectorizer = TfidfVectorizer(max_features=10000, stop_words=\"english\", max_df=0.95, min_df=2, lowercase=False)\n",
    "features = vectorizer.fit_transform(texts)\n",
    "cls = NMF(n_components=number_of_topics, random_state=seed)\n",
    "cls.fit(features)\n",
    "\n",
    "# Function for retrieving the topic vectors for a list of text descriptions.\n",
    "def get_topic_embeddings(texts, model, vectorizer):\n",
    "    ngrams_vectors = vectorizer.transform(texts).toarray()\n",
    "    topic_vectors = model.transform(ngrams_vectors)\n",
    "    return(topic_vectors)\n",
    "    \n",
    "# Create the dataframe containing the average score assigned to each topic for the genes from each subset.\n",
    "group_to_topic_vector = {}\n",
    "for group_id,ids in group_id_to_ids.items():\n",
    "    texts = [descriptions_linares_pontes[i] for i in ids]\n",
    "    topic_vectors = get_topic_embeddings(texts, cls, vectorizer)\n",
    "    mean_topic_vector = np.mean(topic_vectors, axis=0)\n",
    "    group_to_topic_vector[group_id] = mean_topic_vector\n",
    "    \n",
    "tm_df = pd.DataFrame(group_to_topic_vector)\n",
    "\n",
    "# Changing the order of the Lloyd, Meinke phenotype subsets to match other figures for consistency.\n",
    "if NOTEBOOK_TAGS[\"subsets\"]:\n",
    "    filename = \"../data/group_related_files/lloyd/lloyd_function_hierarchy_irb_cleaned.csv\"\n",
    "    lmtm_df = pd.read_csv(filename)    \n",
    "    columns_in_order = [col for col in lmtm_df[\"Subset Symbol\"].values if col in tm_df.columns]\n",
    "    tm_df = tm_df[columns_in_order]\n",
    "    \n",
    "# Reordering so consistency with the curated subsets can be checked by looking at the diagonal.\n",
    "tm_df[\"idxmax\"] = tm_df.idxmax(axis = 1)\n",
    "tm_df[\"idxmax\"] = tm_df[\"idxmax\"].apply(lambda x: tm_df.columns.get_loc(x))\n",
    "tm_df = tm_df.sort_values(by=\"idxmax\")\n",
    "tm_df.drop(columns=[\"idxmax\"], inplace=True)\n",
    "tm_df = tm_df.reset_index(drop=False).rename({\"index\":\"topic\"},axis=1).reset_index(drop=False).rename({\"index\":\"order\"},axis=1)\n",
    "tm_df.to_csv(os.path.join(OUTPUT_DIR,\"part_5_topic_modeling.csv\"), index=False)\n",
    "tm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0: globular embryos embryogenesis embryo fertilization germ placenta embryogenic organogenesis placentae zygotic embryonic \n",
      "1: mutated albino mutants shapeshifter mutant wild \n",
      "2: meristem suspensors gametophytes gametophyte male female \n",
      "3: chl ala aba iba sensitivity insensitive hypersensitive sensitivities sensitive \n",
      "4: pigment pigments tapetum pigmented inferred \n",
      "5: root roots laterals laterally lateral \n",
      "6: antibiotics bacteria yeast avirulent toxin sulfoximine pathogens nonpathogenic microbial coli agrobacterium bacterial virulence infection untreated flu infected coinfection syncytial aseptic disease hiv infections cmv viral \n",
      "7: rotundifolia flowering decondense succulent mid earliest early late beginning \n",
      "8: blue red yellow illumination light \n",
      "9: embryos embryogenesis embryo fertilization germ placenta embryogenic organogenesis placentae zygotic embryonic cotyledon cotyledons \n",
      "10: fore pale wavy green greener \n",
      "11: gravicurvature preglobular wilty phytogylcogens embryos embryogenesis embryo fertilization germ placenta embryogenic organogenesis placentae zygotic embryonic \n",
      "12: seed seedings seeds white brown speck \n",
      "13: trichome stomatal sporopollenin mucilage trichomes branching branches branched \n",
      "14: pollen pollination pollinated pollens tubular tube tubes \n",
      "15: level tiers higher levels paraclade minimal low \n",
      "16: malformed aberrant abnormalities abnormality abnormal normal abnormally morphology morphological generative morphologically morphologies \n",
      "17: germinates vegetative germinating desiccation germinate germination dormancy vernalization photoperiods delayed delay delays halted \n",
      "18: lacunae leaves rosettes rosette \n",
      "19: humidity diurnal heat ambient temperatures temperature freezing paraclade minimal low \n",
      "20: flowered flower blooming bloom floral flowers petals sepals \n",
      "21: unbranched bracts inflorescences pedicels bract inflorescence axils raceme stems stem habit borne \n",
      "22: pectin starch amylopectin dextrin maltodextrin sugar level tiers higher levels \n",
      "23: hairy fleshy hair hairs hairiness root roots \n",
      "24: increased decreased tenfold increasing increase doubling number numbers \n",
      "25: cell cells cellular inside wall walls palisade \n",
      "26: proliferating growth regrowth slow slowing pace slowed slower steady \n",
      "27: giant dwarfs elf dwarf semi \n",
      "28: sensitivity insensitive hypersensitive sensitivities sensitive strain stressed stress stresses \n",
      "29: diminished reduced halved fertility fecundity infertility \n",
      "30: sterility sterile male female \n",
      "31: embryos embryogenesis embryo fertilization germ placenta embryogenic organogenesis placentae zygotic embryonic arrest arrests semidwarfism arresting menbrane arrested \n",
      "32: radicle hypocotyls hypocotyl darkened darkness darkening dark darker black darkly faint dull bright \n",
      "33: xylose sugars invertase fructose sucrose autotrophy glucose maltose sorbitol exogenously exogenous \n",
      "34: amyloplasts flagella plastids chloroplasts chloroplast thylakoids sinapoylglucose radical photooxidative movements progressive rastafari movement \n",
      "35: response youg respond responding responses rearranged modified alterations original altered alteration alter \n",
      "36: locule ovary ovule nucleus nuclei \n",
      "37: long short weeks hours months day month days \n",
      "38: seedlings plantlets germinated planting bonsai seedling accessions lethality lethal \n",
      "39: unmethylated demethylation acetylation hypomethylation methylated cpg methylation trimethylation dna strand \n",
      "40: resistant susceptible clubroot crinkling rot blight mildew powdery botrytis rust downy \n",
      "41: actinic fluorescent uv sunscreens ultraviolet ibr illumination light \n"
     ]
    }
   ],
   "source": [
    "# Describing what the most representative tokens for each topic in the model are.\n",
    "num_top_words = 2\n",
    "feature_names = vectorizer.get_feature_names()\n",
    "for i,topic_vec in enumerate(cls.components_):\n",
    "    print(i,end=\": \")\n",
    "    for fid in topic_vec.argsort()[-1:-num_top_words-1:-1]:\n",
    "        word = feature_names[fid]\n",
    "        word = \" \".join(unreduce_lp[word])\n",
    "        print(word, end=\" \")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"clustering\"></a>\n",
    "### Approach 2: Agglomerative clustering and comparison to predefined groups\n",
    "This clustering approach uses agglomerative clustering to cluster the genes into a fixed number of clusters based off the distances between their embedding representations using all of the above methods. Clustering into a fixed number of clusters allows for clustering into a similar number of groups as a present in some existing grouping of the data, such as phenotype categories or biochemical pathways, and then determining if the clusters obtained are at all similar to the groupings that already exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Generate the numpy array where values are mean distance percentiles between all the methods.\n",
    "mean_pct_array = name_to_array[\"Mean\"]\n",
    "to_id = array_index_to_id\n",
    "\n",
    "# Do agglomerative clustering based on that distance matrix.\n",
    "number_of_clusters = 42\n",
    "ac = AgglomerativeClustering(n_clusters=number_of_clusters, linkage=\"complete\", affinity=\"precomputed\")\n",
    "clustering = ac.fit(mean_pct_array)\n",
    "id_to_cluster = {}\n",
    "cluster_to_ids = defaultdict(list)\n",
    "for idx,c in enumerate(clustering.labels_):\n",
    "    id_to_cluster[to_id[idx]] = c\n",
    "    cluster_to_ids[c].append(to_id[idx])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>cluster</th>\n",
       "      <th>MSD</th>\n",
       "      <th>SRF</th>\n",
       "      <th>SSC</th>\n",
       "      <th>PTH</th>\n",
       "      <th>LEF</th>\n",
       "      <th>IST</th>\n",
       "      <th>NLS</th>\n",
       "      <th>HRM</th>\n",
       "      <th>TMP</th>\n",
       "      <th>LIT</th>\n",
       "      <th>OVP</th>\n",
       "      <th>ROT</th>\n",
       "      <th>GRS</th>\n",
       "      <th>FSM</th>\n",
       "      <th>SEN</th>\n",
       "      <th>CHS</th>\n",
       "      <th>MCH</th>\n",
       "      <th>PRA</th>\n",
       "      <th>WAT</th>\n",
       "      <th>ARC</th>\n",
       "      <th>STT</th>\n",
       "      <th>MSL</th>\n",
       "      <th>PIG</th>\n",
       "      <th>RTH</th>\n",
       "      <th>NUT</th>\n",
       "      <th>MPH</th>\n",
       "      <th>GER</th>\n",
       "      <th>FLT</th>\n",
       "      <th>TCM</th>\n",
       "      <th>CUL</th>\n",
       "      <th>SRL</th>\n",
       "      <th>CPR</th>\n",
       "      <th>MEC</th>\n",
       "      <th>MGD</th>\n",
       "      <th>EMG</th>\n",
       "      <th>EMB</th>\n",
       "      <th>GEM</th>\n",
       "      <th>MTM</th>\n",
       "      <th>GAM</th>\n",
       "      <th>OBI</th>\n",
       "      <th>CDR</th>\n",
       "      <th>NHM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.028736</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.022059</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.013605</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.152074</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.138462</td>\n",
       "      <td>0.248227</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.078947</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.354478</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.022059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160920</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.022989</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.016575</td>\n",
       "      <td>0.016461</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011905</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.004608</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.046154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028736</td>\n",
       "      <td>0.257143</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>0.022321</td>\n",
       "      <td>0.022989</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011050</td>\n",
       "      <td>0.041152</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.216590</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.094340</td>\n",
       "      <td>0.014925</td>\n",
       "      <td>0.011111</td>\n",
       "      <td>0.015385</td>\n",
       "      <td>0.099291</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>0.032520</td>\n",
       "      <td>0.053571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.028807</td>\n",
       "      <td>0.007353</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013825</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014925</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007092</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.520325</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034014</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.039216</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007092</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003731</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.075893</td>\n",
       "      <td>0.298851</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.046083</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>0.044444</td>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026119</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.013393</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.195312</td>\n",
       "      <td>0.135593</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016575</td>\n",
       "      <td>0.016461</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102041</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>0.009804</td>\n",
       "      <td>0.085714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.022388</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.061538</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>33</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043860</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068182</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004608</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.057471</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011905</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009217</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.206349</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>29</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003731</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022099</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.007353</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088435</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.039216</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054688</td>\n",
       "      <td>0.135593</td>\n",
       "      <td>0.068182</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011050</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.093023</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>39</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.039216</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.008230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.137255</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>40</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004608</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003731</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.011905</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009217</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.470588</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.009217</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022388</td>\n",
       "      <td>0.077778</td>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.016461</td>\n",
       "      <td>0.007353</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.191176</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023041</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.046512</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.011111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003731</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>37</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.055300</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007092</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.038674</td>\n",
       "      <td>0.008230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037879</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.132597</td>\n",
       "      <td>0.008230</td>\n",
       "      <td>0.022059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009804</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.044118</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.174419</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011111</td>\n",
       "      <td>0.015385</td>\n",
       "      <td>0.007092</td>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>34</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.120690</td>\n",
       "      <td>0.171429</td>\n",
       "      <td>0.162602</td>\n",
       "      <td>0.151786</td>\n",
       "      <td>0.195402</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.335938</td>\n",
       "      <td>0.372881</td>\n",
       "      <td>0.174242</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.353591</td>\n",
       "      <td>0.246914</td>\n",
       "      <td>0.088235</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.326531</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.314286</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.275862</td>\n",
       "      <td>0.156682</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.290698</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>0.150943</td>\n",
       "      <td>0.171642</td>\n",
       "      <td>0.255556</td>\n",
       "      <td>0.276923</td>\n",
       "      <td>0.092199</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.033582</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.022321</td>\n",
       "      <td>0.045977</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.027624</td>\n",
       "      <td>0.016461</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009804</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027650</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.188679</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>31</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.113208</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007092</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.049107</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037879</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027624</td>\n",
       "      <td>0.041152</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004608</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>0.365672</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014925</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>23</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.023438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.059908</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015385</td>\n",
       "      <td>0.269504</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>36</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.068182</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016461</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.009217</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061538</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.046512</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029851</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.239130</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>27</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.034884</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>8</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033149</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.023041</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>0.014925</td>\n",
       "      <td>0.011111</td>\n",
       "      <td>0.015385</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.543478</td>\n",
       "      <td>0.473881</td>\n",
       "      <td>0.541667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>41</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.113821</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011905</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.004608</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007092</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>24</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.287356</td>\n",
       "      <td>0.171429</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>0.392857</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.394737</td>\n",
       "      <td>0.085938</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378788</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.182320</td>\n",
       "      <td>0.230453</td>\n",
       "      <td>0.551471</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.027211</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009804</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.452381</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>0.138249</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.034884</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.094340</td>\n",
       "      <td>0.201493</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.138462</td>\n",
       "      <td>0.099291</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.131579</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.033582</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016575</td>\n",
       "      <td>0.020576</td>\n",
       "      <td>0.044118</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022388</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011905</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.009217</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003731</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    order  cluster       MSD       SRF       SSC       PTH       LEF       IST       NLS       HRM       TMP       LIT       OVP       ROT       GRS       FSM       SEN       CHS       MCH       PRA       WAT       ARC       STT       MSL       PIG   RTH       NUT       MPH       GER       FLT       TCM       CUL       SRL       CPR    MEC       MGD       EMG       EMB       GEM       MTM       GAM  OBI       CDR       NHM\n",
       "0       0       14  0.666667  0.028736  0.028571  0.008130  0.008929  0.000000  0.026316  0.015625  0.033898  0.007576  0.035714  0.005525  0.012346  0.022059  0.030303  0.013605  0.095238  0.019608  0.000000  0.000000  0.000000  0.017241  0.152074  0.00  0.023256  0.000000  0.018868  0.007463  0.033333  0.138462  0.248227  0.000000  0.000  0.078947  0.260870  0.354478  0.125000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "1       1       38  0.000000  0.068966  0.000000  0.000000  0.004464  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.004115  0.022059  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "2       2        9  0.000000  0.160920  0.057143  0.000000  0.017857  0.022989  0.008772  0.000000  0.000000  0.000000  0.071429  0.016575  0.016461  0.014706  0.090909  0.000000  0.000000  0.000000  0.000000  0.011905  0.000000  0.017241  0.004608  0.00  0.000000  0.000000  0.018868  0.007463  0.022222  0.046154  0.000000  0.021739  0.000  0.026316  0.000000  0.007463  0.000000  0.000000  0.000000  0.0  0.000000  0.083333\n",
       "3       3       25  0.000000  0.028736  0.257143  0.016260  0.022321  0.022989  0.017544  0.031250  0.033898  0.022727  0.000000  0.011050  0.041152  0.014706  0.000000  0.020408  0.000000  0.000000  0.000000  0.023810  0.014706  0.068966  0.216590  0.02  0.011628  0.000000  0.094340  0.014925  0.011111  0.015385  0.099291  0.021739  0.125  0.000000  0.000000  0.007463  0.000000  0.000000  0.000000  0.0  0.000000  0.083333\n",
       "4       4        5  0.000000  0.000000  0.114286  0.032520  0.053571  0.000000  0.008772  0.000000  0.016949  0.015152  0.000000  0.005525  0.028807  0.007353  0.000000  0.006803  0.047619  0.000000  0.000000  0.000000  0.000000  0.000000  0.013825  0.00  0.011628  0.000000  0.000000  0.014925  0.000000  0.000000  0.007092  0.021739  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "5       5        2  0.000000  0.005747  0.000000  0.520325  0.004464  0.000000  0.000000  0.023438  0.000000  0.007576  0.000000  0.000000  0.008230  0.000000  0.000000  0.034014  0.047619  0.039216  0.000000  0.000000  0.000000  0.068966  0.000000  0.02  0.000000  0.000000  0.000000  0.007463  0.000000  0.000000  0.007092  0.000000  0.000  0.000000  0.000000  0.003731  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "6       6       35  0.000000  0.000000  0.000000  0.000000  0.017857  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.004115  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "7       7        0  0.000000  0.040230  0.000000  0.000000  0.075893  0.298851  0.026316  0.015625  0.000000  0.015152  0.035714  0.066298  0.049383  0.117647  0.000000  0.000000  0.000000  0.000000  0.057143  0.095238  0.058824  0.051724  0.046083  0.00  0.000000  0.083333  0.037736  0.044776  0.044444  0.030769  0.021277  0.000000  0.125  0.000000  0.000000  0.026119  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "8       8       16  0.000000  0.000000  0.000000  0.008130  0.013393  0.011494  0.017544  0.195312  0.135593  0.000000  0.000000  0.016575  0.016461  0.000000  0.000000  0.102041  0.190476  0.009804  0.085714  0.000000  0.000000  0.000000  0.018433  0.00  0.139535  0.083333  0.018868  0.022388  0.022222  0.061538  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "9       9       33  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.043860  0.000000  0.000000  0.068182  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.004608  0.00  0.000000  0.000000  0.000000  0.007463  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "10     10        7  0.166667  0.057471  0.000000  0.000000  0.008929  0.011494  0.017544  0.007812  0.033898  0.000000  0.357143  0.005525  0.012346  0.014706  0.060606  0.006803  0.000000  0.000000  0.000000  0.011905  0.014706  0.000000  0.009217  0.06  0.023256  0.000000  0.000000  0.007463  0.000000  0.000000  0.000000  0.021739  0.000  0.315789  0.043478  0.000000  0.333333  0.000000  0.206349  0.0  0.000000  0.000000\n",
       "11     11        1  0.000000  0.005747  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.012346  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "12     12       29  0.000000  0.011494  0.000000  0.000000  0.004464  0.000000  0.000000  0.000000  0.000000  0.007576  0.000000  0.000000  0.012346  0.000000  0.181818  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.017241  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.003731  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "13     13       18  0.000000  0.011494  0.000000  0.000000  0.008929  0.000000  0.000000  0.062500  0.016949  0.007576  0.000000  0.022099  0.012346  0.007353  0.000000  0.088435  0.000000  0.039216  0.000000  0.000000  0.000000  0.051724  0.000000  0.00  0.023256  0.083333  0.075472  0.007463  0.000000  0.030769  0.000000  0.065217  0.000  0.000000  0.000000  0.007463  0.000000  0.000000  0.000000  0.0  0.076923  0.000000\n",
       "14     14        6  0.000000  0.000000  0.000000  0.048780  0.008929  0.000000  0.000000  0.054688  0.135593  0.068182  0.000000  0.011050  0.004115  0.000000  0.000000  0.238095  0.285714  0.029412  0.142857  0.000000  0.000000  0.017241  0.000000  0.00  0.093023  0.000000  0.018868  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "15     15       39  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.039216  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "16     16       32  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.004115  0.000000  0.030303  0.000000  0.000000  0.107843  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "17     17       13  0.000000  0.000000  0.000000  0.000000  0.004464  0.011494  0.000000  0.000000  0.033898  0.007576  0.000000  0.005525  0.008230  0.000000  0.000000  0.006803  0.000000  0.137255  0.028571  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.018868  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.007463  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "18     18       40  0.000000  0.040230  0.000000  0.000000  0.004464  0.000000  0.000000  0.000000  0.000000  0.000000  0.035714  0.005525  0.000000  0.000000  0.000000  0.000000  0.000000  0.068627  0.000000  0.000000  0.000000  0.000000  0.004608  0.02  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.043478  0.000  0.000000  0.000000  0.003731  0.000000  0.000000  0.015873  0.0  0.000000  0.000000\n",
       "19     19       15  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.015625  0.016949  0.007576  0.000000  0.000000  0.037037  0.000000  0.000000  0.013605  0.000000  0.000000  0.057143  0.011905  0.000000  0.000000  0.009217  0.00  0.000000  0.000000  0.018868  0.007463  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "20     20       17  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.022059  0.000000  0.000000  0.000000  0.000000  0.028571  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "21     21       19  0.000000  0.017241  0.142857  0.008130  0.017857  0.011494  0.026316  0.007812  0.000000  0.022727  0.000000  0.000000  0.012346  0.014706  0.000000  0.006803  0.000000  0.029412  0.028571  0.000000  0.470588  0.017241  0.009217  0.06  0.000000  0.000000  0.000000  0.022388  0.077778  0.030769  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "22     22       30  0.000000  0.017241  0.000000  0.000000  0.008929  0.000000  0.008772  0.062500  0.000000  0.007576  0.000000  0.005525  0.016461  0.007353  0.000000  0.020408  0.000000  0.000000  0.142857  0.023810  0.191176  0.000000  0.023041  0.00  0.046512  0.083333  0.000000  0.007463  0.011111  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.003731  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "23     23       37  0.000000  0.000000  0.000000  0.000000  0.004464  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.024691  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.017241  0.055300  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.007092  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "24     24       20  0.000000  0.000000  0.000000  0.000000  0.008929  0.000000  0.026316  0.007812  0.000000  0.015152  0.000000  0.038674  0.008230  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.12  0.000000  0.000000  0.000000  0.000000  0.022222  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "25     25       11  0.000000  0.011494  0.000000  0.008130  0.008929  0.000000  0.017544  0.015625  0.000000  0.037879  0.000000  0.132597  0.008230  0.022059  0.000000  0.000000  0.000000  0.009804  0.028571  0.023810  0.044118  0.000000  0.000000  0.38  0.011628  0.083333  0.000000  0.000000  0.033333  0.000000  0.000000  0.021739  0.250  0.000000  0.000000  0.000000  0.000000  0.142857  0.000000  0.0  0.000000  0.000000\n",
       "26     26       21  0.000000  0.011494  0.000000  0.000000  0.004464  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.008230  0.000000  0.000000  0.006803  0.000000  0.029412  0.000000  0.000000  0.000000  0.017241  0.000000  0.00  0.174419  0.000000  0.000000  0.000000  0.011111  0.015385  0.007092  0.065217  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "27     27       34  0.000000  0.120690  0.171429  0.162602  0.151786  0.195402  0.263158  0.335938  0.372881  0.174242  0.107143  0.353591  0.246914  0.088235  0.272727  0.326531  0.285714  0.333333  0.314286  0.214286  0.058824  0.275862  0.156682  0.22  0.290698  0.416667  0.150943  0.171642  0.255556  0.276923  0.092199  0.043478  0.250  0.105263  0.065217  0.033582  0.000000  0.142857  0.000000  0.0  0.000000  0.000000\n",
       "28     28       12  0.000000  0.005747  0.028571  0.008130  0.022321  0.045977  0.035088  0.000000  0.084746  0.007576  0.035714  0.027624  0.016461  0.014706  0.000000  0.000000  0.000000  0.009804  0.028571  0.035714  0.000000  0.000000  0.027650  0.00  0.011628  0.000000  0.188679  0.007463  0.000000  0.030769  0.085106  0.000000  0.000  0.026316  0.000000  0.007463  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "29     29       31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.007812  0.033898  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.013605  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.083333  0.113208  0.000000  0.000000  0.000000  0.007092  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "30     30       22  0.000000  0.028736  0.000000  0.000000  0.049107  0.011494  0.008772  0.000000  0.000000  0.037879  0.000000  0.027624  0.041152  0.014706  0.030303  0.000000  0.000000  0.000000  0.028571  0.023810  0.000000  0.000000  0.004608  0.00  0.011628  0.000000  0.037736  0.365672  0.000000  0.000000  0.000000  0.000000  0.000  0.000000  0.000000  0.014925  0.000000  0.285714  0.000000  0.0  0.000000  0.000000\n",
       "31     31       23  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.009804  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.022222  0.030769  0.000000  0.000000  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "32     32       26  0.000000  0.005747  0.000000  0.000000  0.000000  0.000000  0.008772  0.023438  0.000000  0.000000  0.000000  0.005525  0.004115  0.000000  0.030303  0.006803  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.059908  0.00  0.000000  0.000000  0.056604  0.000000  0.000000  0.015385  0.269504  0.000000  0.000  0.000000  0.000000  0.007463  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "33     33       36  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.015152  0.000000  0.000000  0.000000  0.000000  0.000000  0.006803  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.015385  0.000000  0.043478  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "34     34        4  0.000000  0.005747  0.000000  0.008130  0.004464  0.000000  0.008772  0.000000  0.016949  0.068182  0.000000  0.000000  0.016461  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.017241  0.009217  0.00  0.000000  0.000000  0.018868  0.000000  0.000000  0.061538  0.000000  0.260870  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.083333\n",
       "35     35        3  0.000000  0.000000  0.000000  0.008130  0.000000  0.000000  0.000000  0.015625  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.030303  0.020408  0.000000  0.029412  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.046512  0.000000  0.000000  0.029851  0.000000  0.015385  0.000000  0.239130  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.000000\n",
       "36     36       27  0.000000  0.000000  0.000000  0.024390  0.004464  0.000000  0.000000  0.007812  0.016949  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.006803  0.047619  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00  0.034884  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.043478  0.000  0.000000  0.000000  0.000000  0.000000  0.285714  0.000000  0.0  0.000000  0.000000\n",
       "37     37        8  0.166667  0.017241  0.000000  0.000000  0.017857  0.000000  0.008772  0.000000  0.000000  0.000000  0.000000  0.033149  0.037037  0.000000  0.000000  0.006803  0.000000  0.000000  0.000000  0.000000  0.000000  0.051724  0.023041  0.02  0.000000  0.000000  0.018868  0.014925  0.011111  0.015385  0.021277  0.000000  0.000  0.315789  0.543478  0.473881  0.541667  0.000000  0.666667  0.0  0.000000  0.000000\n",
       "38     38       41  0.000000  0.000000  0.000000  0.113821  0.004464  0.000000  0.008772  0.000000  0.000000  0.000000  0.000000  0.005525  0.012346  0.000000  0.030303  0.020408  0.000000  0.029412  0.000000  0.011905  0.000000  0.017241  0.004608  0.00  0.000000  0.000000  0.000000  0.007463  0.000000  0.000000  0.007092  0.021739  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  1.0  0.000000  0.000000\n",
       "39     39       24  0.000000  0.287356  0.171429  0.016260  0.392857  0.344828  0.394737  0.085938  0.000000  0.378788  0.321429  0.182320  0.230453  0.551471  0.212121  0.027211  0.000000  0.009804  0.028571  0.452381  0.117647  0.241379  0.138249  0.08  0.034884  0.083333  0.094340  0.201493  0.366667  0.138462  0.099291  0.043478  0.250  0.131579  0.086957  0.033582  0.000000  0.142857  0.111111  0.0  0.769231  0.083333\n",
       "40     40       10  0.000000  0.011494  0.000000  0.000000  0.035714  0.011494  0.008772  0.007812  0.000000  0.000000  0.000000  0.016575  0.020576  0.044118  0.000000  0.000000  0.000000  0.019608  0.000000  0.047619  0.029412  0.000000  0.000000  0.00  0.011628  0.000000  0.000000  0.022388  0.055556  0.030769  0.000000  0.021739  0.000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  0.153846  0.000000\n",
       "41     41       28  0.000000  0.000000  0.028571  0.008130  0.004464  0.000000  0.008772  0.000000  0.016949  0.000000  0.000000  0.000000  0.012346  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.011905  0.000000  0.034483  0.009217  0.00  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.021277  0.000000  0.000  0.000000  0.000000  0.003731  0.000000  0.000000  0.000000  0.0  0.000000  0.666667"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the dataframe containing the average score assigned to each topic for the genes from each subset.\n",
    "group_to_cluster_vector = {}\n",
    "for group_id,ids in group_id_to_ids.items():\n",
    "    \n",
    "    mean_cluster_vector = np.zeros(number_of_clusters)\n",
    "    for i in ids:\n",
    "        cluster = id_to_cluster[i]\n",
    "        mean_cluster_vector[cluster] = mean_cluster_vector[cluster]+1\n",
    "    mean_cluster_vector = mean_cluster_vector/mean_cluster_vector.sum(axis=0,keepdims=1)\n",
    "    group_to_cluster_vector[group_id] = mean_cluster_vector\n",
    "    \n",
    "ac_df = pd.DataFrame(group_to_cluster_vector)\n",
    "\n",
    "# Changing the order of the Lloyd, Meinke phenotype subsets to match other figures for consistency.\n",
    "if NOTEBOOK_TAGS[\"subsets\"]:\n",
    "    filename = \"../data/group_related_files/lloyd/lloyd_function_hierarchy_irb_cleaned.csv\"\n",
    "    lmtm_df = pd.read_csv(filename)    \n",
    "    columns_in_order = [col for col in lmtm_df[\"Subset Symbol\"].values if col in tm_df.columns]\n",
    "    tm_df = tm_df[columns_in_order]\n",
    "\n",
    "# Reordering so consistency with the curated subsets can be checked by looking at the diagonal.\n",
    "ac_df[\"idxmax\"] = ac_df.idxmax(axis = 1)\n",
    "ac_df[\"idxmax\"] = ac_df[\"idxmax\"].apply(lambda x: ac_df.columns.get_loc(x))\n",
    "ac_df = ac_df.sort_values(by=\"idxmax\")\n",
    "ac_df.drop(columns=[\"idxmax\"], inplace=True)\n",
    "ac_df = ac_df.reset_index(drop=False).rename({\"index\":\"cluster\"},axis=1).reset_index(drop=False).rename({\"index\":\"order\"},axis=1)\n",
    "ac_df.to_csv(os.path.join(OUTPUT_DIR,\"part_5_agglomerative_clustering.csv\"), index=False)\n",
    "ac_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"phenologs\"></a>\n",
    "### Approach 3: Looking for phenolog relationships between clusters and OMIM disease phenotypes\n",
    "This section produces a table of values that provides a score for the a particular pair of a cluster found for this dataset of plant genes and a disease phenotype. Currently the value indicates the fraction of the plant genes in that cluster that have orthologs associated with that disease phenotype. This should be replaced or supplemented with a p-value for evaluating the significance of this value given the distribution of genes and their mappings to all of the disease phenotypes. All the rows from the input dataframe containing the PantherDB and OMIM information where the ID from this dataset is not known or the mapping to a phenotype was unsuccessful are removed at this step, fix this if the metric for evaluating cluster to phenotype phenolog mappings need this information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_identifier</th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "      <th>gene_mim_number</th>\n",
       "      <th>phenotype_mim_number</th>\n",
       "      <th>phenotype_mim_name</th>\n",
       "      <th>id</th>\n",
       "      <th>compressed_phenotype_mim_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G01950</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 1</td>\n",
       "      <td>2558</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G01950</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 3B</td>\n",
       "      <td>2558</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02140</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 1</td>\n",
       "      <td>1468</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02140</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 3B</td>\n",
       "      <td>1468</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G02120</td>\n",
       "      <td>KIF21A</td>\n",
       "      <td>608283.0</td>\n",
       "      <td>135700</td>\n",
       "      <td>Fibrosis of extraocular muscles, congenital, 1</td>\n",
       "      <td>232</td>\n",
       "      <td>Fibrosis of extraocular muscles</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    species gene_identifier human_ortholog_gene_symbol  gene_mim_number  phenotype_mim_number                               phenotype_mim_name    id    compressed_phenotype_mim_name\n",
       "77    ARATH       AT1G01950                     KIF21A         608283.0                135700   Fibrosis of extraocular muscles, congenital, 1  2558  Fibrosis of extraocular muscles\n",
       "78    ARATH       AT1G01950                     KIF21A         608283.0                135700  Fibrosis of extraocular muscles, congenital, 3B  2558  Fibrosis of extraocular muscles\n",
       "113   ARATH       AT1G02140                     KIF21A         608283.0                135700   Fibrosis of extraocular muscles, congenital, 1  1468  Fibrosis of extraocular muscles\n",
       "114   ARATH       AT1G02140                     KIF21A         608283.0                135700  Fibrosis of extraocular muscles, congenital, 3B  1468  Fibrosis of extraocular muscles\n",
       "167   ARATH       AT1G02120                     KIF21A         608283.0                135700   Fibrosis of extraocular muscles, congenital, 1   232  Fibrosis of extraocular muscles"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the dataframe mapping plant genes --> human orthologs --> disease phenotypes.\n",
    "omim_df = pd.read_csv(panther_to_omim_filename)\n",
    "# Add a column that indicates which ID in the dataset those plant genes refer to, for mapping to phenotypes.\n",
    "name_to_id = dataset.get_name_to_id_dictionary()\n",
    "omim_df[\"id\"] = omim_df[\"gene_identifier\"].map(lambda x: name_to_id.get(x,None))\n",
    "omim_df = omim_df.dropna(subset=[\"id\",\"phenotype_mim_name\"], inplace=False)\n",
    "omim_df[\"phenotype_mim_name\"] = omim_df[\"phenotype_mim_name\"].astype(str)\n",
    "omim_df[\"compressed_phenotype_mim_name\"] = omim_df[\"phenotype_mim_name\"].map(lambda x: x.split(\",\")[0])\n",
    "omim_df[\"id\"] = omim_df[\"id\"].astype(\"int64\")\n",
    "omim_df[\"phenotype_mim_number\"] = omim_df[\"phenotype_mim_number\"].astype(\"int64\")\n",
    "# Generate mappings between the IDs in this dataset and disease phenotypes or orthologous genes.\n",
    "id_to_mim_phenotype_names = defaultdict(list)\n",
    "for i,p in zip(omim_df[\"id\"].values,omim_df[\"compressed_phenotype_mim_name\"].values):\n",
    "    id_to_mim_phenotype_names[i].append(p)\n",
    "id_to_human_gene_symbols = defaultdict(list)\n",
    "for i,s in zip(omim_df[\"id\"].values,omim_df[\"human_ortholog_gene_symbol\"].values):\n",
    "    id_to_human_gene_symbols[i].append(s)\n",
    "omim_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "compressed_phenotype_mim_name\n",
      "3-Methylcrotonyl-CoA carboxylase 1 deficiency     3\n",
      "3-Methylcrotonyl-CoA carboxylase 2 deficiency     3\n",
      "3-hydroxyisobutryl-CoA hydrolase deficiency       4\n",
      "3-methylglutaconic aciduria                       7\n",
      "5-fluorouracil toxicity                           3\n",
      "                                                 ..\n",
      "{Rhabdoid tumor predisposition syndrome 2}        9\n",
      "{Rheumatoid arthritis                             5\n",
      "{Schizophrenia                                    3\n",
      "{Yao syndrome}                                    4\n",
      "{von Hippel-Lindau syndrome                      14\n",
      "Length: 624, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# How many genes in our dataset map to orthologs that map to the same OMIM phenotype?\n",
    "print(omim_df.groupby(\"compressed_phenotype_mim_name\").size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>14</th>\n",
       "      <th>34</th>\n",
       "      <th>2</th>\n",
       "      <th>41</th>\n",
       "      <th>24</th>\n",
       "      <th>12</th>\n",
       "      <th>7</th>\n",
       "      <th>19</th>\n",
       "      <th>0</th>\n",
       "      <th>18</th>\n",
       "      <th>6</th>\n",
       "      <th>37</th>\n",
       "      <th>25</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>21</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>16</th>\n",
       "      <th>5</th>\n",
       "      <th>22</th>\n",
       "      <th>15</th>\n",
       "      <th>28</th>\n",
       "      <th>10</th>\n",
       "      <th>33</th>\n",
       "      <th>9</th>\n",
       "      <th>30</th>\n",
       "      <th>8</th>\n",
       "      <th>26</th>\n",
       "      <th>32</th>\n",
       "      <th>29</th>\n",
       "      <th>39</th>\n",
       "      <th>40</th>\n",
       "      <th>17</th>\n",
       "      <th>27</th>\n",
       "      <th>23</th>\n",
       "      <th>35</th>\n",
       "      <th>38</th>\n",
       "      <th>1</th>\n",
       "      <th>20</th>\n",
       "      <th>31</th>\n",
       "      <th>36</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Fibrosis of extraocular muscles</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thiamine metabolism dysfunction syndrome 5 (episodic encephalopathy type)</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Weaver syndrome</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Heimler syndrome 2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peroxisome biogenesis disorder 4A (Zellweger)</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    14  34  2   41  24  12  7   19  0   18  6   37  25  11  13  21  3   4   16  5   22  15  28  10  33  9   30  8   26  32  29  39  40  17  27  23  35  38  1   20  31  36\n",
       "Fibrosis of extraocular muscles                      0   1   0   1   1   0   1   0   1   0   0   0   1   1   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Thiamine metabolism dysfunction syndrome 5 (epi...   1   1   0   0   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Weaver syndrome                                      1   1   0   0   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Heimler syndrome 2                                   1   1   0   0   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Peroxisome biogenesis disorder 4A (Zellweger)        1   1   0   0   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenolog_x_dict = defaultdict(dict)\n",
    "phenolog_p_dict = defaultdict(dict)\n",
    "candidate_genes_dict = defaultdict(dict)\n",
    "phenotypes = pd.unique(omim_df[\"compressed_phenotype_mim_name\"].values)\n",
    "clusters = list(cluster_to_ids.keys())\n",
    "for cluster,phenotype in itertools.product(clusters,phenotypes):\n",
    "    \n",
    "    # What are the candidate genes predicted if this phenolog pairing is real?\n",
    "    ids = cluster_to_ids[cluster]\n",
    "    candidate_genes_dict[cluster][phenotype] = list(set(flatten([id_to_human_gene_symbols[i] for i in ids if phenotype not in id_to_mim_phenotype_names.get(i,[])])))\n",
    "\n",
    "    # What is the p-value for this phenolog pairing?\n",
    "    # The size of the population (genes in the dataset).\n",
    "    M = len(id_to_cluster.keys())\n",
    "    # The number of elements we draw without replacement (genes in the cluster).\n",
    "    N = len(cluster_to_ids[cluster])     \n",
    "    # The number of available successes in the population (genes that map to orthologs that map to this phenotype).\n",
    "    n = len([i for i in id_to_cluster.keys() if phenotype in id_to_mim_phenotype_names.get(i,[])])\n",
    "    # The number of successes drawn (genes in this cluster that map to orthologs that map to this phenotype).\n",
    "    x = list(set(flatten([id_to_mim_phenotype_names.get(i,[]) for i in ids]))).count(phenotype)\n",
    "    prob = 1-hypergeom.cdf(x-1, M, n, N) # Equivalent to prob = 1-sum([hypergeom.pmf(x_i, M, n, N) for x_i in range(0,x)])\n",
    "    phenolog_x_dict[cluster][phenotype] = x\n",
    "    phenolog_p_dict[cluster][phenotype] = prob\n",
    "    \n",
    "\n",
    "# Convert the dictionary to a table of values with cluster and phenotype as the rows and columns.\n",
    "phenolog_matrix = pd.DataFrame(phenolog_x_dict)        \n",
    "phenolog_matrix.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>omim_phenotype_name</th>\n",
       "      <th>cluster</th>\n",
       "      <th>size</th>\n",
       "      <th>x</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_adjusted</th>\n",
       "      <th>candidate_gene_symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20117</th>\n",
       "      <td>{Deafness</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008406</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MAPKAPK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22074</th>\n",
       "      <td>Smith-McCort dysplasia 2</td>\n",
       "      <td>35</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010499</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1290</th>\n",
       "      <td>Glycogen storage disease of heart</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>0.012112</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FAN1|SNIP1|SET|TPI1|EED|DNMT3B|MRE11|DHX38|DDX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1291</th>\n",
       "      <td>Wolff-Parkinson-White syndrome</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>0.012112</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FAN1|SNIP1|SET|TPI1|EED|DNMT3B|MRE11|DHX38|DDX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14837</th>\n",
       "      <td>Pontocerebellar hypoplasia type 2B</td>\n",
       "      <td>33</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014694</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GLDC|SEC23B|SCO1|SEC23A|WARS2|STX11|SIL1|XPR1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14838</th>\n",
       "      <td>Neurodevelopmental disorder with dysmorphic fa...</td>\n",
       "      <td>33</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.014694</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GLDC|SEC23B|SCO1|SEC23A|WARS2|STX11|SIL1|XPR1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>?Aniridia 2</td>\n",
       "      <td>12</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>0.015271</td>\n",
       "      <td>1.0</td>\n",
       "      <td>IGF2BP2|IQSEC2|TFRC|ACTG1|HSPD1|DNMT3B|PRPF31|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3035</th>\n",
       "      <td>Retinitis pigmentosa 11</td>\n",
       "      <td>12</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>0.015271</td>\n",
       "      <td>1.0</td>\n",
       "      <td>IGF2BP2|IQSEC2|TFRC|ACTG1|HSPD1|DNMT3B|ACTB|LC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6693</th>\n",
       "      <td>Frontometaphyseal dysplasia 1</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6694</th>\n",
       "      <td>Heterotopia</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6695</th>\n",
       "      <td>Intestinal pseudoobstruction</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6697</th>\n",
       "      <td>Otopalatodigital syndrome</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6700</th>\n",
       "      <td>Boomerang dysplasia</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6699</th>\n",
       "      <td>Atelosteogenesis</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6692</th>\n",
       "      <td>Congenital short bowel syndrome</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6701</th>\n",
       "      <td>Larsen syndrome</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6702</th>\n",
       "      <td>Spondylocarpotarsal synostosis syndrome</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6698</th>\n",
       "      <td>Terminal osseous dysplasia</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6691</th>\n",
       "      <td>Cardiac valvular dysplasia</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6696</th>\n",
       "      <td>Melnick-Needles syndrome</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21096</th>\n",
       "      <td>Antley-Bixler syndrome with genital anomalies ...</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6689</th>\n",
       "      <td>?Cataracts</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21097</th>\n",
       "      <td>Disordered steroidogenesis due to cytochrome P...</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21094</th>\n",
       "      <td>{Malaria</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21095</th>\n",
       "      <td>{Alzheimer disease</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6690</th>\n",
       "      <td>?FG syndrome 2</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17243</th>\n",
       "      <td>Glomuvenous malformations</td>\n",
       "      <td>26</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCAPD3|MSH6|UPB1|IQSEC2|TFRC|PEX14|SNIP1|SUMO1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22075</th>\n",
       "      <td>Oocyte maturation defect 4</td>\n",
       "      <td>35</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020914</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20118</th>\n",
       "      <td>?Spinocerebellar ataxia</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020914</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MAPKAPK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8328</th>\n",
       "      <td>{Encephalopathy</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020958</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RSPH1|DNAJC6|RAD51C|DGAT1|GNB5|NMNAT1|TUBG1|PE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     omim_phenotype_name cluster  size  x   p_value  p_adjusted                             candidate_gene_symbols\n",
       "20117                                          {Deafness      17     4  1  0.008406         1.0                                           MAPKAPK3\n",
       "22074                           Smith-McCort dysplasia 2      35     4  1  0.010499         1.0                                                   \n",
       "1290                   Glycogen storage disease of heart      41    23  1  0.012112         1.0  FAN1|SNIP1|SET|TPI1|EED|DNMT3B|MRE11|DHX38|DDX...\n",
       "1291                      Wolff-Parkinson-White syndrome      41    23  1  0.012112         1.0  FAN1|SNIP1|SET|TPI1|EED|DNMT3B|MRE11|DHX38|DDX...\n",
       "14837                 Pontocerebellar hypoplasia type 2B      33    14  1  0.014694         1.0  GLDC|SEC23B|SCO1|SEC23A|WARS2|STX11|SIL1|XPR1|...\n",
       "14838  Neurodevelopmental disorder with dysmorphic fa...      33    14  1  0.014694         1.0  GLDC|SEC23B|SCO1|SEC23A|WARS2|STX11|SIL1|XPR1|...\n",
       "2701                                         ?Aniridia 2      12    29  1  0.015271         1.0  IGF2BP2|IQSEC2|TFRC|ACTG1|HSPD1|DNMT3B|PRPF31|...\n",
       "3035                             Retinitis pigmentosa 11      12    29  1  0.015271         1.0  IGF2BP2|IQSEC2|TFRC|ACTG1|HSPD1|DNMT3B|ACTB|LC...\n",
       "6693                       Frontometaphyseal dysplasia 1      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6694                                         Heterotopia      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6695                        Intestinal pseudoobstruction      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6697                           Otopalatodigital syndrome      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6700                                 Boomerang dysplasia      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6699                                    Atelosteogenesis      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6692                     Congenital short bowel syndrome      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6701                                     Larsen syndrome      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6702             Spondylocarpotarsal synostosis syndrome      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6698                          Terminal osseous dysplasia      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6691                          Cardiac valvular dysplasia      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "6696                            Melnick-Needles syndrome      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "21096  Antley-Bixler syndrome with genital anomalies ...      27    12  1  0.018848         1.0  MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...\n",
       "6689                                          ?Cataracts      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "21097  Disordered steroidogenesis due to cytochrome P...      27    12  1  0.018848         1.0  MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...\n",
       "21094                                           {Malaria      27    12  1  0.018848         1.0  MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...\n",
       "21095                                 {Alzheimer disease      27    12  1  0.018848         1.0  MEF2C|SLC25A15|OAT|RBBP8|MEF2A|CCND2|BAG3|CCND...\n",
       "6690                                      ?FG syndrome 2      37    12  1  0.018848         1.0  CYP26B1|KDM5B|CYP26C1|IGF2BP2|SLC11A2|SLC11A1|...\n",
       "17243                          Glomuvenous malformations      26    38  1  0.020011         1.0  NCAPD3|MSH6|UPB1|IQSEC2|TFRC|PEX14|SNIP1|SUMO1...\n",
       "22075                         Oocyte maturation defect 4      35     4  1  0.020914         1.0                                                   \n",
       "20118                            ?Spinocerebellar ataxia      17     4  1  0.020914         1.0                                           MAPKAPK3\n",
       "8328                                     {Encephalopathy      13    20  1  0.020958         1.0  RSPH1|DNAJC6|RAD51C|DGAT1|GNB5|NMNAT1|TUBG1|PE..."
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Produce a melted version of the phenolog matrix sorted by value and including predicted candidate genes.\n",
    "phenolog_matrix_reset = phenolog_matrix.reset_index(drop=False).rename({\"index\":\"omim_phenotype_name\"}, axis=\"columns\")\n",
    "phenolog_df = pd.melt(phenolog_matrix_reset, id_vars=[\"omim_phenotype_name\"], value_vars=phenolog_matrix.columns[1:], var_name=\"cluster\", value_name=\"x\")\n",
    "# What other information should be present in this melted phenologs matrix?\n",
    "phenolog_df[\"size\"] = phenolog_df[\"cluster\"].map(lambda x: len(cluster_to_ids[x]))\n",
    "phenolog_df[\"candidate_gene_symbols\"] = np.vectorize(lambda x,y: concatenate_with_bar_delim(*candidate_genes_dict[x][y]))(phenolog_df[\"cluster\"], phenolog_df[\"omim_phenotype_name\"])\n",
    "phenolog_df[\"p_value\"] = np.vectorize(lambda x,y: phenolog_p_dict[x][y])(phenolog_df[\"cluster\"], phenolog_df[\"omim_phenotype_name\"])\n",
    "phenolog_df[\"p_adjusted\"] = multipletests(phenolog_df[\"p_value\"].values, method='bonferroni')[1]\n",
    "phenolog_df.sort_values(by=[\"p_value\"], inplace=True, ascending=True)\n",
    "phenolog_df = phenolog_df[[\"omim_phenotype_name\", \"cluster\", \"size\", \"x\", \"p_value\", \"p_adjusted\", \"candidate_gene_symbols\"]]\n",
    "phenolog_df.to_csv(os.path.join(OUTPUT_DIR,\"part_5_phenologs.csv\"), index=False)\n",
    "phenolog_df.head(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Approach 4: Agglomerative clustering and sillhouette scores for each NLP method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>BERT:Base:Layers=3,Concatenated</th>\n",
       "      <th>BioBERT:PubMed,PMC,Layers=4,Concatenated</th>\n",
       "      <th>N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>Doc2Vec (Phenes):Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>0.040338</td>\n",
       "      <td>0.552907</td>\n",
       "      <td>0.055877</td>\n",
       "      <td>0.044010</td>\n",
       "      <td>0.061637</td>\n",
       "      <td>0.089547</td>\n",
       "      <td>0.003484</td>\n",
       "      <td>0.059101</td>\n",
       "      <td>0.243716</td>\n",
       "      <td>0.097867</td>\n",
       "      <td>0.081677</td>\n",
       "      <td>0.094990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>0.042736</td>\n",
       "      <td>0.549014</td>\n",
       "      <td>0.057230</td>\n",
       "      <td>0.025881</td>\n",
       "      <td>0.072647</td>\n",
       "      <td>0.100706</td>\n",
       "      <td>0.014961</td>\n",
       "      <td>0.075144</td>\n",
       "      <td>0.251999</td>\n",
       "      <td>0.115777</td>\n",
       "      <td>0.088976</td>\n",
       "      <td>0.092947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>0.037298</td>\n",
       "      <td>0.558779</td>\n",
       "      <td>0.059331</td>\n",
       "      <td>0.090848</td>\n",
       "      <td>0.073382</td>\n",
       "      <td>0.092769</td>\n",
       "      <td>0.016356</td>\n",
       "      <td>0.083506</td>\n",
       "      <td>0.258298</td>\n",
       "      <td>0.093764</td>\n",
       "      <td>0.092112</td>\n",
       "      <td>0.085023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32</td>\n",
       "      <td>0.038338</td>\n",
       "      <td>0.560432</td>\n",
       "      <td>0.059144</td>\n",
       "      <td>0.095460</td>\n",
       "      <td>0.081567</td>\n",
       "      <td>0.092226</td>\n",
       "      <td>0.021565</td>\n",
       "      <td>0.095497</td>\n",
       "      <td>0.264355</td>\n",
       "      <td>0.105933</td>\n",
       "      <td>0.103885</td>\n",
       "      <td>0.097224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36</td>\n",
       "      <td>0.039492</td>\n",
       "      <td>0.548790</td>\n",
       "      <td>0.059690</td>\n",
       "      <td>0.098910</td>\n",
       "      <td>0.084490</td>\n",
       "      <td>0.092401</td>\n",
       "      <td>0.022260</td>\n",
       "      <td>0.100233</td>\n",
       "      <td>0.278200</td>\n",
       "      <td>0.093484</td>\n",
       "      <td>0.114564</td>\n",
       "      <td>0.097568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40</td>\n",
       "      <td>0.038411</td>\n",
       "      <td>0.553280</td>\n",
       "      <td>0.061743</td>\n",
       "      <td>0.102653</td>\n",
       "      <td>0.088682</td>\n",
       "      <td>0.098438</td>\n",
       "      <td>0.026081</td>\n",
       "      <td>0.112990</td>\n",
       "      <td>0.277152</td>\n",
       "      <td>0.095957</td>\n",
       "      <td>0.119829</td>\n",
       "      <td>0.092690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>44</td>\n",
       "      <td>0.039335</td>\n",
       "      <td>0.554208</td>\n",
       "      <td>0.063317</td>\n",
       "      <td>0.106875</td>\n",
       "      <td>0.088759</td>\n",
       "      <td>0.100177</td>\n",
       "      <td>0.028844</td>\n",
       "      <td>0.116996</td>\n",
       "      <td>0.280180</td>\n",
       "      <td>0.110355</td>\n",
       "      <td>0.131681</td>\n",
       "      <td>0.095001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>48</td>\n",
       "      <td>0.039680</td>\n",
       "      <td>0.556447</td>\n",
       "      <td>0.064759</td>\n",
       "      <td>0.112910</td>\n",
       "      <td>0.088818</td>\n",
       "      <td>0.096962</td>\n",
       "      <td>0.033479</td>\n",
       "      <td>0.123768</td>\n",
       "      <td>0.284644</td>\n",
       "      <td>0.131322</td>\n",
       "      <td>0.142153</td>\n",
       "      <td>0.101552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>52</td>\n",
       "      <td>0.041080</td>\n",
       "      <td>0.560009</td>\n",
       "      <td>0.067402</td>\n",
       "      <td>0.115100</td>\n",
       "      <td>0.087781</td>\n",
       "      <td>0.094912</td>\n",
       "      <td>0.035242</td>\n",
       "      <td>0.130476</td>\n",
       "      <td>0.287420</td>\n",
       "      <td>0.136194</td>\n",
       "      <td>0.135860</td>\n",
       "      <td>0.091996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>56</td>\n",
       "      <td>0.042142</td>\n",
       "      <td>0.559372</td>\n",
       "      <td>0.067702</td>\n",
       "      <td>0.121932</td>\n",
       "      <td>0.089189</td>\n",
       "      <td>0.072494</td>\n",
       "      <td>0.041302</td>\n",
       "      <td>0.129662</td>\n",
       "      <td>0.292647</td>\n",
       "      <td>0.141877</td>\n",
       "      <td>0.141239</td>\n",
       "      <td>0.099073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    n   GO:None   PO:None  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives  BERT:Base:Layers=3,Concatenated  BioBERT:PubMed,PMC,Layers=4,Concatenated  N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF  NOBLE Coder:Partial,TFIDF  Doc2Vec (Phenes):Wikipedia,Size=300  Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives  N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF      Mean\n",
       "0  20  0.040338  0.552907                    0.055877                                           0.044010                         0.061637                                  0.089547                                           0.003484                          0.059101                             0.243716                                           0.097867                                                    0.081677   0.094990\n",
       "1  24  0.042736  0.549014                    0.057230                                           0.025881                         0.072647                                  0.100706                                           0.014961                          0.075144                             0.251999                                           0.115777                                                    0.088976   0.092947\n",
       "2  28  0.037298  0.558779                    0.059331                                           0.090848                         0.073382                                  0.092769                                           0.016356                          0.083506                             0.258298                                           0.093764                                                    0.092112   0.085023\n",
       "3  32  0.038338  0.560432                    0.059144                                           0.095460                         0.081567                                  0.092226                                           0.021565                          0.095497                             0.264355                                           0.105933                                                    0.103885   0.097224\n",
       "4  36  0.039492  0.548790                    0.059690                                           0.098910                         0.084490                                  0.092401                                           0.022260                          0.100233                             0.278200                                           0.093484                                                    0.114564   0.097568\n",
       "5  40  0.038411  0.553280                    0.061743                                           0.102653                         0.088682                                  0.098438                                           0.026081                          0.112990                             0.277152                                           0.095957                                                    0.119829   0.092690\n",
       "6  44  0.039335  0.554208                    0.063317                                           0.106875                         0.088759                                  0.100177                                           0.028844                          0.116996                             0.280180                                           0.110355                                                    0.131681   0.095001\n",
       "7  48  0.039680  0.556447                    0.064759                                           0.112910                         0.088818                                  0.096962                                           0.033479                          0.123768                             0.284644                                           0.131322                                                    0.142153   0.101552\n",
       "8  52  0.041080  0.560009                    0.067402                                           0.115100                         0.087781                                  0.094912                                           0.035242                          0.130476                             0.287420                                           0.136194                                                    0.135860   0.091996\n",
       "9  56  0.042142  0.559372                    0.067702                                           0.121932                         0.089189                                  0.072494                                           0.041302                          0.129662                             0.292647                                           0.141877                                                    0.141239   0.099073"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics.cluster import silhouette_score\n",
    "# Note that homogeneity scores don't fit for evaluating how close the clustering is to pathway membership, etc.\n",
    "# This is because genes can be assigned to more than one pathway, metric would have to be changed to account for this.\n",
    "# So all this section does is determines which values of n_clusters provide good clustering results for each matrix.\n",
    "n_clusters_silhouette_scores = defaultdict(dict)\n",
    "min_n_clusters = 20\n",
    "max_n_clusters = 80\n",
    "step_size = 4\n",
    "number_of_clusters = np.arange(min_n_clusters, max_n_clusters, step_size)\n",
    "for n in number_of_clusters:\n",
    "    for name in names:\n",
    "        distance_matrix = name_to_array[name]\n",
    "        #to_id = array_index_to_id\n",
    "        ac = AgglomerativeClustering(n_clusters=n, linkage=\"complete\", affinity=\"precomputed\")\n",
    "        clustering = ac.fit(distance_matrix)\n",
    "        sil_score = silhouette_score(distance_matrix, clustering.labels_, metric=\"precomputed\")\n",
    "        n_clusters_silhouette_scores[name][n] = sil_score\n",
    "sil_df = pd.DataFrame(n_clusters_silhouette_scores).reset_index(drop=False).rename({\"index\":\"n\"},axis=\"columns\")\n",
    "sil_df.to_csv(os.path.join(OUTPUT_DIR,\"part_5_silhouette_scores_by_n.csv\"), index=False)\n",
    "sil_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"part_6\"></a>\n",
    "# Part 6. Supervised Tasks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"merging\"></a>\n",
    "### Option 1: Merging in the previously curated similarity values from the Oellrich, Walls et al. (2015) dataset\n",
    "This section reads in a file that contains the previously calculated distance values from the Oellrich, Walls et al. (2015) dataset, and merges it with the values which are obtained here for all of the applicable natural language processing or machine learning methods used, so that the graphs which are specified by these sets of distances values can be evaluated side by side in the subsequent sections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if not NOTEBOOK_TAGS[\"training\"]:\n",
    "    # Add a column that indicates the distance estimated using curated EQ statements.\n",
    "    df = df.merge(right=pppn_edgelist.df, how=\"left\", on=[\"from\",\"to\"])\n",
    "    df.fillna(value=0.000,inplace=True)\n",
    "    df.rename(columns={\"value\":\"EQs\"}, inplace=True)\n",
    "    df[\"EQs\"] = 1-df[\"EQs\"]\n",
    "    names.append(\"EQs\")\n",
    "    df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 2: Merging with information about shared biochemical pathways or groups.\n",
    "The relevant information for each edge includes questions like whether or not the two genes that edge connects share a group or biochemical pathway in common, or if those genes are from the same species. This information can then later be used as the target values for predictive models, or for filtering the graphs represented by these edge lists. Either the grouping information or the protein-protein interaction information should be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({0: 1585301, 1: 221352})\n",
      "Counter({1: 1806653})\n"
     ]
    }
   ],
   "source": [
    "# Column indicating whether or not the two genes share this features (e.g., pathway in common, same group).\n",
    "df[\"shared\"] = df[[\"from\",\"to\"]].apply(lambda x: len(set(id_to_group_ids[x[\"from\"]]).intersection(set(id_to_group_ids[x[\"to\"]])))>0, axis=1)*1\n",
    "# Column indicating whether the two genes are from the same species.\n",
    "species_dict = dataset.get_species_dictionary()\n",
    "df[\"same\"] = df[[\"from\",\"to\"]].apply(lambda x: species_dict[x[\"from\"]]==species_dict[x[\"to\"]],axis=1)*1\n",
    "print(Counter(df[\"shared\"].values))\n",
    "print(Counter(df[\"same\"].values))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 3: Merging with information about protein-protein interactions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Merging information from the protein-protein interaction database with this dataset.\n",
    "df = df.merge(right=string_data.df, how=\"left\", on=[\"from\",\"to\"])\n",
    "df.fillna(value=0,inplace=True)\n",
    "df[\"shared\"] = (df[\"combined_score\"] != 0.00)*1\n",
    "df.tail(12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ks\"></a>\n",
    "### Do the edges joining genes that share a group, pathway, or interaction come from a different distribution?\n",
    "The purpose of this section is to visualize kernel estimates for the distributions of distance or similarity scores generated by each of the methods tested for measuring semantic similarity or generating vector representations of the phenotype descriptions. Ideally, better methods should show better separation betwene the distributions for distance values between two genes involved in a common specified group or two genes that are not. Additionally, a statistical test is used to check whether these two distributions are significantly different from each other or not, although this is a less informative measure than the other tests used in subsequent sections, because it does not address how useful these differences in the distributions actually are for making predictions about group membership."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Use Kolmogorov-Smirnov test to see if edges between genes that share a group come from a distinct distribution.\n",
    "ppi_pos_dict = {name:(df[df[\"shared\"] > 0.00][name].values) for name in names}\n",
    "ppi_neg_dict = {name:(df[df[\"shared\"] == 0.00][name].values) for name in names}\n",
    "for name in names:\n",
    "    stat,p = ks_2samp(ppi_pos_dict[name],ppi_neg_dict[name])\n",
    "    pos_mean = np.average(ppi_pos_dict[name])\n",
    "    neg_mean = np.average(ppi_neg_dict[name])\n",
    "    pos_n = len(ppi_pos_dict[name])\n",
    "    neg_n = len(ppi_neg_dict[name])\n",
    "    TABLE[name].update({\"mean_1\":pos_mean, \"mean_0\":neg_mean, \"n_1\":pos_n, \"n_0\":neg_n})\n",
    "    TABLE[name].update({\"ks\":stat, \"ks_pval\":p})\n",
    "    \n",
    "    \n",
    "# Show the kernel estimates for each distribution of weights for each method.\n",
    "#num_plots, plots_per_row, row_width, row_height = (len(names), 4, 14, 3)\n",
    "#fig,axs = plt.subplots(math.ceil(num_plots/plots_per_row), plots_per_row, squeeze=False)\n",
    "#for name,ax in zip(names,axs.flatten()):\n",
    "#    ax.set_title(name)\n",
    "#    ax.set_xlabel(\"value\")\n",
    "#    ax.set_ylabel(\"density\")\n",
    "#    sns.kdeplot(ppi_pos_dict[name], color=\"black\", shade=False, alpha=1.0, ax=ax)\n",
    "#    sns.kdeplot(ppi_neg_dict[name], color=\"black\", shade=True, alpha=0.1, ax=ax) \n",
    "#fig.set_size_inches(row_width, row_height*math.ceil(num_plots/plots_per_row))\n",
    "#fig.tight_layout()\n",
    "#fig.savefig(os.path.join(OUTPUT_DIR,\"part_6_kernel_density.png\"),dpi=400)\n",
    "#plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"within\"></a>\n",
    "### Looking at within-group or within-pathway distances in each graph\n",
    "The purpose of this section is to determine which methods generated graphs which tightly group genes which share common pathways or group membership with one another. In order to compare across different methods where the distance value distributions are different, the mean distance values for each group for each method are convereted to percentile scores. Lower percentile scores indicate that the average distance value between any two genes that belong to that group is lower than most of the distance values in the entire distribution for that method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group_id</th>\n",
       "      <th>full_name</th>\n",
       "      <th>n</th>\n",
       "      <th>mean_percentile</th>\n",
       "      <th>mean_rank</th>\n",
       "      <th>GO:None</th>\n",
       "      <th>PO:None</th>\n",
       "      <th>Doc2Vec:Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>BERT:Base:Layers=3,Concatenated</th>\n",
       "      <th>BioBERT:PubMed,PMC,Layers=4,Concatenated</th>\n",
       "      <th>N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF</th>\n",
       "      <th>NOBLE Coder:Partial,TFIDF</th>\n",
       "      <th>Doc2Vec (Phenes):Wikipedia,Size=300</th>\n",
       "      <th>Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives</th>\n",
       "      <th>N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF</th>\n",
       "      <th>Mean</th>\n",
       "      <th>EQs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OBI</td>\n",
       "      <td>Other biological interactors</td>\n",
       "      <td>4</td>\n",
       "      <td>10.405391</td>\n",
       "      <td>8.230769</td>\n",
       "      <td>4.6266</td>\n",
       "      <td>18.2828</td>\n",
       "      <td>1.2582</td>\n",
       "      <td>13.1061</td>\n",
       "      <td>26.2137</td>\n",
       "      <td>39.4603</td>\n",
       "      <td>0.9049</td>\n",
       "      <td>4.7270</td>\n",
       "      <td>7.5859</td>\n",
       "      <td>8.3546</td>\n",
       "      <td>4.9824</td>\n",
       "      <td>3.6392</td>\n",
       "      <td>4.3030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GEM</td>\n",
       "      <td>Gametophyte, embryo defective (2-10% mutant se...</td>\n",
       "      <td>24</td>\n",
       "      <td>10.414501</td>\n",
       "      <td>5.192308</td>\n",
       "      <td>23.8242</td>\n",
       "      <td>64.4687</td>\n",
       "      <td>1.4460</td>\n",
       "      <td>2.5669</td>\n",
       "      <td>19.1636</td>\n",
       "      <td>13.6088</td>\n",
       "      <td>0.9671</td>\n",
       "      <td>3.7343</td>\n",
       "      <td>1.8435</td>\n",
       "      <td>1.6708</td>\n",
       "      <td>3.0778</td>\n",
       "      <td>1.2070</td>\n",
       "      <td>3.0320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDR</td>\n",
       "      <td>Circadian rhythms</td>\n",
       "      <td>13</td>\n",
       "      <td>11.958714</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>15.3052</td>\n",
       "      <td>57.2714</td>\n",
       "      <td>13.5902</td>\n",
       "      <td>7.8281</td>\n",
       "      <td>15.0445</td>\n",
       "      <td>24.5121</td>\n",
       "      <td>1.6572</td>\n",
       "      <td>3.1262</td>\n",
       "      <td>6.9848</td>\n",
       "      <td>5.0485</td>\n",
       "      <td>3.5098</td>\n",
       "      <td>2.4337</td>\n",
       "      <td>5.1103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GAM</td>\n",
       "      <td>Gametophyte defective (&lt;2% mutant seeds)</td>\n",
       "      <td>63</td>\n",
       "      <td>12.394160</td>\n",
       "      <td>6.653846</td>\n",
       "      <td>23.8210</td>\n",
       "      <td>60.2561</td>\n",
       "      <td>2.5926</td>\n",
       "      <td>3.1106</td>\n",
       "      <td>26.7058</td>\n",
       "      <td>23.8934</td>\n",
       "      <td>0.9519</td>\n",
       "      <td>7.2648</td>\n",
       "      <td>3.8151</td>\n",
       "      <td>4.5914</td>\n",
       "      <td>3.2140</td>\n",
       "      <td>3.4049</td>\n",
       "      <td>3.2428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EMG</td>\n",
       "      <td>Embryo, gametophyte defective (&gt;10% mutant seeds)</td>\n",
       "      <td>46</td>\n",
       "      <td>13.394251</td>\n",
       "      <td>7.692308</td>\n",
       "      <td>15.5231</td>\n",
       "      <td>52.3486</td>\n",
       "      <td>2.4483</td>\n",
       "      <td>28.1632</td>\n",
       "      <td>27.2377</td>\n",
       "      <td>23.3622</td>\n",
       "      <td>1.3685</td>\n",
       "      <td>3.9908</td>\n",
       "      <td>6.4616</td>\n",
       "      <td>5.4030</td>\n",
       "      <td>3.5091</td>\n",
       "      <td>6.2858</td>\n",
       "      <td>3.7253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  group_id                                          full_name   n  mean_percentile  mean_rank  GO:None  PO:None  Doc2Vec:Wikipedia,Size=300  Word2Vec:Wikipedia,Size=300,Mean,Nouns,Adjectives  BERT:Base:Layers=3,Concatenated  BioBERT:PubMed,PMC,Layers=4,Concatenated  N-Grams:Simple,Words,Linares Pontes,1-grams,Binary,TFIDF  NOBLE Coder:Partial,TFIDF  Doc2Vec (Phenes):Wikipedia,Size=300  Word2Vec (Phenes):Wikipedia,Size=300,Mean,Nouns,Adjectives  N-Grams (Phenes):Simple,Words,1-grams,Binary,TFIDF    Mean     EQs\n",
       "0      OBI                       Other biological interactors   4        10.405391   8.230769   4.6266  18.2828                      1.2582                                            13.1061                          26.2137                                   39.4603                                             0.9049                            4.7270                               7.5859                                             8.3546                                                      4.9824   3.6392  4.3030\n",
       "1      GEM  Gametophyte, embryo defective (2-10% mutant se...  24        10.414501   5.192308  23.8242  64.4687                      1.4460                                             2.5669                          19.1636                                   13.6088                                             0.9671                            3.7343                               1.8435                                             1.6708                                                      3.0778   1.2070  3.0320\n",
       "2      CDR                                  Circadian rhythms  13        11.958714   6.000000  15.3052  57.2714                     13.5902                                             7.8281                          15.0445                                   24.5121                                             1.6572                            3.1262                               6.9848                                             5.0485                                                      3.5098   2.4337  5.1103\n",
       "3      GAM           Gametophyte defective (<2% mutant seeds)  63        12.394160   6.653846  23.8210  60.2561                      2.5926                                             3.1106                          26.7058                                   23.8934                                             0.9519                            7.2648                               3.8151                                             4.5914                                                      3.2140   3.4049  3.2428\n",
       "4      EMG  Embryo, gametophyte defective (>10% mutant seeds)  46        13.394251   7.692308  15.5231  52.3486                      2.4483                                            28.1632                          27.2377                                   23.3622                                             1.3685                            3.9908                               6.4616                                             5.4030                                                      3.5091   6.2858  3.7253"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all the average within-pathway phenotype distance values for each method for each particular pathway.\n",
    "group_id_to_ids = groups.get_group_id_to_ids_dict(dataset.get_gene_dictionary())\n",
    "group_ids = list(group_id_to_ids.keys())\n",
    "graph = IndexedGraph(df)\n",
    "within_weights_dict = defaultdict(lambda: defaultdict(list))\n",
    "within_percentiles_dict = defaultdict(lambda: defaultdict(list))\n",
    "all_weights_dict = {}\n",
    "for name in names:\n",
    "    all_weights_dict[name] = df[name].values\n",
    "    for group in group_ids:\n",
    "        within_ids = group_id_to_ids[group]\n",
    "        within_pairs = [(i,j) for i,j in itertools.permutations(within_ids,2)]\n",
    "        mean_weight = np.mean((graph.get_values(within_pairs, kind=name)))\n",
    "        within_weights_dict[name][group] = mean_weight\n",
    "        within_percentiles_dict[name][group] = stats.percentileofscore(df[name].values, mean_weight, kind=\"rank\")\n",
    "\n",
    "# Generating a dataframe of percentiles of the mean in-group distance scores.\n",
    "within_dist_data = pd.DataFrame(within_percentiles_dict)\n",
    "within_dist_data = within_dist_data.dropna(axis=0, inplace=False)\n",
    "within_dist_data = within_dist_data.round(4)\n",
    "\n",
    "# Adding relevant information to this dataframe and saving.\n",
    "within_dist_data[\"mean_rank\"] = within_dist_data.rank().mean(axis=1)\n",
    "within_dist_data[\"mean_percentile\"] = within_dist_data.mean(axis=1)\n",
    "within_dist_data.sort_values(by=\"mean_percentile\", inplace=True)\n",
    "within_dist_data.reset_index(inplace=True)\n",
    "within_dist_data[\"group_id\"] = within_dist_data[\"index\"]\n",
    "within_dist_data[\"full_name\"] = within_dist_data[\"group_id\"].apply(lambda x: groups.get_long_name(x))\n",
    "within_dist_data[\"n\"] = within_dist_data[\"group_id\"].apply(lambda x: len(group_id_to_ids[x]))\n",
    "within_dist_data = within_dist_data[flatten([\"group_id\",\"full_name\",\"n\",\"mean_percentile\",\"mean_rank\",names])]\n",
    "within_dist_data.to_csv(os.path.join(OUTPUT_DIR,\"part_6_within_distances.csv\"), index=False)\n",
    "within_dist_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"auc\"></a>\n",
    "### Predicting whether two genes belong to the same group, pathway, or share an interaction\n",
    "The purpose of this section is to see if whether or not two genes share atleast one common pathway can be predicted from the distance scores assigned using analysis of text similarity. The evaluation of predictability is done by reporting a precision and recall curve for each method, as well as remembering the area under the curve, and ratio between the area under the curve and the baseline (expected area when guessing randomly) for each method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bootstrap(fraction, num_iterations, y_true, y_prob):\n",
    "    # Run the desired number of bootstrap iterations over the full population of predictions and return st devs.\n",
    "    scores = pd.DataFrame([bootstrap_iteration(fraction, y_true, y_prob) for i in range(num_iterations)])\n",
    "    standard_deviations = {\n",
    "        \"f_1_max_std\": np.std(scores[\"f_1_max\"].values),\n",
    "        \"f_2_max_std\": np.std(scores[\"f_2_max\"].values),\n",
    "        \"f_point5_max_std\": np.std(scores[\"f_point5_max\"].values)}\n",
    "    return(standard_deviations)\n",
    "\n",
    "\n",
    "def bootstrap_iteration(fraction, y_true, y_prob):\n",
    "    assert len(y_true) == len(y_prob)\n",
    "    # Subset the total population of predictions using the provided fraction.\n",
    "    num_predictions = len(y_true)\n",
    "    bootstrapping_fraction = fraction\n",
    "    num_to_retain = int(np.ceil(num_predictions*bootstrapping_fraction))\n",
    "    idx = np.random.choice(np.arange(num_predictions), num_to_retain, replace=False)\n",
    "    y_true_sample = y_true[idx]\n",
    "    y_prob_sample = y_prob[idx]\n",
    "    \n",
    "    # Calculate any desired metrics using just that subset.\n",
    "    n_pos, n_neg = Counter(y_true_sample)[1], Counter(y_true_sample)[0]\n",
    "    precision, recall, thresholds = precision_recall_curve(y_true_sample, y_prob_sample)\n",
    "    baseline = Counter(y_true_sample)[1]/len(y_true_sample) \n",
    "    area = auc(recall, precision)\n",
    "    auc_to_baseline_auc_ratio = area/baseline\n",
    "    \n",
    "    # Find the maximum F score for different values of .  \n",
    "    f_beta = lambda pr,re,beta: [((1+beta**2)*p*r)/((((beta**2)*p)+r)) for p,r in zip(pr,re)]\n",
    "    f_1_scores = f_beta(precision,recall,beta=1)\n",
    "    f_2_scores = f_beta(precision,recall,beta=2)\n",
    "    f_point5_scores = f_beta(precision,recall,beta=0.5)\n",
    "    \n",
    "    # Create a dictionary of those metric values to return.\n",
    "    scores={\"f_1_max\":np.nanmax(f_1_scores),\"f_2_max\":np.nanmax(f_2_scores),\"f_point5_max\":np.nanmax(f_point5_scores)}\n",
    "    return(scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "y_true_dict = {name:df[\"shared\"].values for name in names}       #just added .values here...\n",
    "y_prob_dict = {name:(1 - df[name].values) for name in names}\n",
    "num_plots, plots_per_row, row_width, row_height = (len(names), 4, 14, 3)\n",
    "fig,axs = plt.subplots(math.ceil(num_plots/plots_per_row), plots_per_row, squeeze=False)\n",
    "for name,ax in zip(names, axs.flatten()):\n",
    "    \n",
    "    # Obtaining the values and metrics.\n",
    "    y_true, y_prob = y_true_dict[name], y_prob_dict[name]\n",
    "    n_pos, n_neg = Counter(y_true)[1], Counter(y_true)[0]\n",
    "    precision, recall, thresholds = precision_recall_curve(y_true, y_prob)\n",
    "    baseline_auc = Counter(y_true)[1]/len(y_true) \n",
    "    area = auc(recall, precision)\n",
    "    auc_to_baseline_auc_ratio = area/baseline_auc\n",
    "    # The baseline F1 max has a precision of the ratio of positives to all samples and a recall of 1.\n",
    "    # This is because a random classifier achieves that precision at all recall values, so recall is maximized to\n",
    "    # find the maximum F1 value that can be expected due to random chance.\n",
    "    baseline_f1_max = (2*baseline_auc*1)/(baseline_auc+1)\n",
    "    TABLE[name].update({\"auc\":area,\"ratio\":auc_to_baseline_auc_ratio, \"baseline\":baseline_f1_max, })\n",
    "    \n",
    "    # Find the maximum F score for different values of .  \n",
    "    f_beta = lambda pr,re,beta: [((1+beta**2)*p*r)/((((beta**2)*p)+r)) for p,r in zip(pr,re)]\n",
    "    f_1_scores = f_beta(precision,recall,beta=1)\n",
    "    f_2_scores = f_beta(precision,recall,beta=2)\n",
    "    f_point5_scores = f_beta(precision,recall,beta=0.5)\n",
    "    f_1_max, f_1_std = np.nanmax(f_1_scores), np.std(f_1_scores)\n",
    "    f_2_max, f_2_std = np.nanmax(f_2_scores), np.std(f_2_scores)\n",
    "    f_point5_max, f_point5_std = np.nanmax(f_point5_scores), np.std(f_point5_scores)\n",
    "    \n",
    "    # Find the standard deviation of each metric when subsampling the dataset of predictions for each method.\n",
    "    bootstrap_fraction = 0.5\n",
    "    bootstrap_iterations = 10\n",
    "    bootstrapped_std_dict = bootstrap(bootstrap_fraction, bootstrap_iterations, y_true, y_prob)\n",
    "    \n",
    "    TABLE[name].update({\"f1_max\":f_1_max, \"f5_max\":f_point5_max, \"f2_max\":f_2_max})\n",
    "    #TABLE[name].update({\"f1_std\":f_1_std, \"f5_std\":f_point5_std, \"f2_std\":f_2_std})\n",
    "    TABLE[name].update({\"f1_std\":bootstrapped_std_dict[\"f_1_max_std\"], \n",
    "                        \"f5_std\":bootstrapped_std_dict[\"f_point5_max_std\"], \n",
    "                        \"f2_std\":bootstrapped_std_dict[\"f_2_max_std\"]}) \n",
    "    \n",
    "    # Producing the precision recall curve.\n",
    "    step_kwargs = ({'step': 'post'} if 'step' in signature(plt.fill_between).parameters else {})\n",
    "    ax.step(recall, precision, color='black', alpha=0.2, where='post')\n",
    "    ax.fill_between(recall, precision, alpha=0.7, color='black', **step_kwargs)\n",
    "    ax.axhline(baseline_auc, linestyle=\"--\", color=\"lightgray\")\n",
    "    ax.set_xlabel('Recall')\n",
    "    ax.set_ylabel('Precision')\n",
    "    ax.set_ylim([0.0, 1.05])\n",
    "    ax.set_xlim([0.0, 1.0])\n",
    "    ax.set_title(\"PR {0} (Baseline={1:0.3f})\".format(name, baseline_auc))\n",
    "    \n",
    "fig.set_size_inches(row_width, row_height*math.ceil(num_plots/plots_per_row))\n",
    "fig.tight_layout()\n",
    "fig.savefig(os.path.join(OUTPUT_DIR,\"part_6_prcurve_shared.png\"),dpi=400)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"y\"></a>\n",
    "### Are genes in the same group or pathway ranked higher with respect to individual nodes?\n",
    "This is a way of statistically seeing if for some value k, the graph ranks more edges from some particular gene to any other gene that it has a true protein-protein interaction with higher or equal to rank k, than we would expect due to random chance. This way of looking at the problem helps to be less ambiguous than the previous methods, because it gets at the core of how this would actually be used. In other words, we don't really care how much true information we're missing as long as we're still able to pick up some new useful information by building these networks, so even though we could be missing a lot, what's going on at the very top of the results? These results should be comparable to very strictly thresholding the network and saying that the remaining edges are our guesses at interactions. This is comparable to just looking at the far left-hand side of the precision recall curves, but just quantifies it slightly differently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# When the edgelist is generated above, only the lower triangle of the pairwise matrix is retained for edges in the \n",
    "# graph. This means that in terms of the indices of each node, only the (i,j) node is listed in the edge list where\n",
    "# i is less than j. This makes sense because the graph that's specified is assumed to already be undirected. However\n",
    "# in order to be able to easily subset the edgelist by a single column to obtain rows that correspond to all edges\n",
    "# connected to a particular node, this method will double the number of rows to include both (i,j) and (j,i) edges.\n",
    "df = make_undirected(df)\n",
    "\n",
    "# What's the number of functional partners ranked k or higher in terms of phenotypic description similarity for \n",
    "# each gene? Also figure out the maximum possible number of functional partners that could be theoretically\n",
    "# recovered in this dataset if recovered means being ranked as k or higher here.\n",
    "k = 10      # The threshold of interest for gene ranks.\n",
    "n = 100     # Number of Monte Carlo simulation iterations to complete.\n",
    "df[list(names)] = df.groupby(\"from\")[list(names)].rank()\n",
    "ys = df[df[\"shared\"]==1][list(names)].apply(lambda s: len([x for x in s if x<=k]))\n",
    "ymax = sum(df.groupby(\"from\")[\"shared\"].apply(lambda s: min(len([x for x in s if x==1]),k)))\n",
    "\n",
    "# Monte Carlo simulation to see what the probability is of achieving each y-value by just randomly pulling k \n",
    "# edges for each gene rather than taking the top k ones that the similarity methods specifies when ranking.\n",
    "ysims = [sum(df.groupby(\"from\")[\"shared\"].apply(lambda s: len([x for x in s.sample(k) if x>0.00]))) for i in range(n)]\n",
    "for name in names:\n",
    "    pvalue = len([ysim for ysim in ysims if ysim>=ys[name]])/float(n)\n",
    "    TABLE[name].update({\"y\":ys[name], \"y_max\":ymax, \"y_ratio\":ys[name]/ymax, \"y_pval\":pvalue})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"mean\"></a>\n",
    "### Predicting biochemical pathway or group membership based on mean vectors\n",
    "This section looks at how well the biochemical pathways that a particular gene is a member of can be predicted based on the similarity between the vector representation of the phenotype descriptions for that gene and the average vector for all the vector representations of phenotypes asociated with genes that belong to that particular pathway. In calculating the average vector for a given biochemical pathway, the vector corresponding to the gene that is currently being classified is not accounted for, to avoid overestimating the performance by including information about the ground truth during classification. This leads to missing information in the case of biochemical pathways that have only one member. This can be accounted for by only limiting the overall dataset to only include genes that belong to pathways that have atleast two genes mapped to them, and only including those pathways, or by removing the missing values before calculating the performance metrics below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get the list of methods to look at, and a mapping between each method and the correct similarity metric to apply.\n",
    "# vector_dicts = {k:v.vector_dictionary for k,v in graphs.items()}\n",
    "# names = list(vector_dicts.keys())\n",
    "# group_id_to_ids = groups.get_group_id_to_ids_dict(dataset.get_gene_dictionary())\n",
    "# valid_group_ids = [group for group,id_list in group_id_to_ids.items() if len(id_list)>1]\n",
    "# valid_ids = [i for i in dataset.get_ids() if len(set(valid_group_ids).intersection(set(id_to_group_ids[i])))>0]\n",
    "# pred_dict = defaultdict(lambda: defaultdict(dict))\n",
    "# true_dict = defaultdict(lambda: defaultdict(dict))\n",
    "# for name in names:\n",
    "#     for group in valid_group_ids:\n",
    "#         ids = group_id_to_ids[group]\n",
    "#         for identifier in valid_ids:\n",
    "#             # What's the mean vector of this group, without this particular one that we're trying to classify.\n",
    "#             vectors = np.array([vector_dicts[name][some_id] for some_id in ids if not some_id==identifier])\n",
    "#             mean_vector = vectors.mean(axis=0)\n",
    "#             this_vector = vector_dicts[name][identifier]\n",
    "#             pred_dict[name][identifier][group] = 1-metric_dict[name](mean_vector, this_vector)\n",
    "#             true_dict[name][identifier][group] = (identifier in group_id_to_ids[group])*1                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# num_plots, plots_per_row, row_width, row_height = (len(names), 4, 14, 3)\n",
    "# fig,axs = plt.subplots(math.ceil(num_plots/plots_per_row), plots_per_row, squeeze=False)\n",
    "# for name,ax in zip(names, axs.flatten()):\n",
    "#     \n",
    "#     # Obtaining the values and metrics.\n",
    "#     y_true = pd.DataFrame(true_dict[name]).as_matrix().flatten()\n",
    "#     y_prob = pd.DataFrame(pred_dict[name]).as_matrix().flatten()\n",
    "#     n_pos, n_neg = Counter(y_true)[1], Counter(y_true)[0]\n",
    "#     precision, recall, thresholds = precision_recall_curve(y_true, y_prob)\n",
    "#     baseline = Counter(y_true)[1]/len(y_true) \n",
    "#     area = auc(recall, precision)\n",
    "#     auc_to_baseline_auc_ratio = area/baseline\n",
    "#     TABLE[name].update({\"mean_auc\":area, \"mean_baseline\":baseline, \"mean_ratio\":auc_to_baseline_auc_ratio})\n",
    "# \n",
    "#     # Producing the precision recall curve.\n",
    "#     step_kwargs = ({'step': 'post'} if 'step' in signature(plt.fill_between).parameters else {})\n",
    "#     ax.step(recall, precision, color='black', alpha=0.2, where='post')\n",
    "#     ax.fill_between(recall, precision, alpha=0.7, color='black', **step_kwargs)\n",
    "#     ax.axhline(baseline, linestyle=\"--\", color=\"lightgray\")\n",
    "#     ax.set_xlabel('Recall')\n",
    "#     ax.set_ylabel('Precision')\n",
    "#     ax.set_ylim([0.0, 1.05])\n",
    "#     ax.set_xlim([0.0, 1.0])\n",
    "#     ax.set_title(\"PR {0} (Baseline={1:0.3f})\".format(name[:10], baseline))\n",
    "#     \n",
    "# fig.set_size_inches(row_width, row_height*math.ceil(num_plots/plots_per_row))\n",
    "# fig.tight_layout()\n",
    "# fig.savefig(os.path.join(OUTPUT_DIR,\"part_6_prcurve_mean_classifier.png\"),dpi=400)\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicting biochemical pathway membership based on mean similarity values\n",
    "This section looks at how well the biochemical pathways that a particular gene is a member of can be predicted based on the average similarity between the vector representationt of the phenotype descriptions for that gene and each of the vector representations for other phenotypes associated with genes that belong to that particular pathway. In calculating the average similarity to other genes from a given biochemical pathway, the gene that is currently being classified is not accounted for, to avoid overestimating the performance by including information about the ground truth during classification. This leads to missing information in the case of biochemical pathways that have only one member. This can be accounted for by only limiting the overall dataset to only include genes that belong to pathways that have atleast two genes mapped to them, and only including those pathways, or by removing the missing values before calculating the performance metrics below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicting biochemical pathway or group membership with KNN classifier\n",
    "This section looks at how well the group(s) or biochemical pathway(s) that a particular gene belongs to can be predicted based on a KNN classifier generated using every other gene. For this section, only the groups or pathways which contain more than one gene, and the genes mapped to those groups or pathways, are of interest. This is because for other genes, if we consider them then it will be true that that gene belongs to that group in the target vector, but the KNN classifier could never predict this because when that gene is held out, nothing could provide a vote for that group, because there are zero genes available to be members of the K nearest neighbors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"output\"></a>\n",
    "### Summarizing the results for this notebook\n",
    "Write a large table of results to an output file. Columns are generally metrics and rows are generally methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Method</th>\n",
       "      <th>Hyperparameters</th>\n",
       "      <th>Group</th>\n",
       "      <th>Order</th>\n",
       "      <th>Topic</th>\n",
       "      <th>Data</th>\n",
       "      <th>mean_1</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>n_1</th>\n",
       "      <th>n_0</th>\n",
       "      <th>ks</th>\n",
       "      <th>ks_pval</th>\n",
       "      <th>auc</th>\n",
       "      <th>ratio</th>\n",
       "      <th>baseline</th>\n",
       "      <th>f1_max</th>\n",
       "      <th>f5_max</th>\n",
       "      <th>f2_max</th>\n",
       "      <th>f1_std</th>\n",
       "      <th>f5_std</th>\n",
       "      <th>f2_std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO</td>\n",
       "      <td>None</td>\n",
       "      <td>NLP</td>\n",
       "      <td>0</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.889960</td>\n",
       "      <td>0.894548</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.031194</td>\n",
       "      <td>1.382797e-164</td>\n",
       "      <td>0.132613</td>\n",
       "      <td>1.082373</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.149586</td>\n",
       "      <td>0.411120</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>0.000321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PO</td>\n",
       "      <td>None</td>\n",
       "      <td>NLP</td>\n",
       "      <td>1</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.237214</td>\n",
       "      <td>0.283063</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.095043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.158489</td>\n",
       "      <td>1.293568</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.230242</td>\n",
       "      <td>0.168826</td>\n",
       "      <td>0.411420</td>\n",
       "      <td>0.000506</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0.000497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Doc2Vec</td>\n",
       "      <td>Wikipedia,Size=300</td>\n",
       "      <td>NLP</td>\n",
       "      <td>2</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.386482</td>\n",
       "      <td>0.445125</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.243296</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.279419</td>\n",
       "      <td>2.280588</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.302845</td>\n",
       "      <td>0.318254</td>\n",
       "      <td>0.432778</td>\n",
       "      <td>0.000552</td>\n",
       "      <td>0.000983</td>\n",
       "      <td>0.000579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Word2Vec</td>\n",
       "      <td>Wikipedia,Size=300,Mean,Nouns,Adjectives</td>\n",
       "      <td>NLP</td>\n",
       "      <td>3</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.288637</td>\n",
       "      <td>0.391489</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.277843</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.265836</td>\n",
       "      <td>2.169724</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.316445</td>\n",
       "      <td>0.304437</td>\n",
       "      <td>0.453653</td>\n",
       "      <td>0.000466</td>\n",
       "      <td>0.000992</td>\n",
       "      <td>0.000642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BERT</td>\n",
       "      <td>Base</td>\n",
       "      <td>NLP</td>\n",
       "      <td>4</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.139482</td>\n",
       "      <td>0.182477</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.239832</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.210654</td>\n",
       "      <td>1.719338</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.284718</td>\n",
       "      <td>0.237876</td>\n",
       "      <td>0.446076</td>\n",
       "      <td>0.000453</td>\n",
       "      <td>0.000850</td>\n",
       "      <td>0.000628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BioBERT</td>\n",
       "      <td>PubMed,PMC,Layers=4,Concatenated</td>\n",
       "      <td>NLP</td>\n",
       "      <td>5</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.043455</td>\n",
       "      <td>0.059269</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.257316</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.226192</td>\n",
       "      <td>1.846161</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.291718</td>\n",
       "      <td>0.258599</td>\n",
       "      <td>0.455249</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000814</td>\n",
       "      <td>0.000694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>N-Grams</td>\n",
       "      <td>Simple,Words,Linares Pontes,1-grams,Binary,TFIDF</td>\n",
       "      <td>NLP</td>\n",
       "      <td>6</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.867449</td>\n",
       "      <td>0.954430</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.454706</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.389964</td>\n",
       "      <td>3.182851</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.414292</td>\n",
       "      <td>0.408670</td>\n",
       "      <td>0.549205</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.000906</td>\n",
       "      <td>0.000431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NOBLE Coder</td>\n",
       "      <td>Partial,TFIDF</td>\n",
       "      <td>NLP</td>\n",
       "      <td>7</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.647827</td>\n",
       "      <td>0.770909</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.332715</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.300828</td>\n",
       "      <td>2.455329</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.350433</td>\n",
       "      <td>0.340740</td>\n",
       "      <td>0.476043</td>\n",
       "      <td>0.001048</td>\n",
       "      <td>0.001046</td>\n",
       "      <td>0.000823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Doc2Vec (Phenes)</td>\n",
       "      <td>Wikipedia,Size=300</td>\n",
       "      <td>NLP</td>\n",
       "      <td>8</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.184149</td>\n",
       "      <td>0.309738</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.519054</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.463947</td>\n",
       "      <td>3.786689</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.499558</td>\n",
       "      <td>0.552394</td>\n",
       "      <td>0.585175</td>\n",
       "      <td>0.000888</td>\n",
       "      <td>0.001312</td>\n",
       "      <td>0.000610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Word2Vec (Phenes)</td>\n",
       "      <td>Wikipedia,Size=300,Mean,Nouns,Adjectives</td>\n",
       "      <td>NLP</td>\n",
       "      <td>9</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.238348</td>\n",
       "      <td>0.408866</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.372733</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.409688</td>\n",
       "      <td>3.343833</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.418810</td>\n",
       "      <td>0.499060</td>\n",
       "      <td>0.497894</td>\n",
       "      <td>0.001198</td>\n",
       "      <td>0.001026</td>\n",
       "      <td>0.000610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>N-Grams (Phenes)</td>\n",
       "      <td>Simple,Words,1-grams,Binary,TFIDF</td>\n",
       "      <td>NLP</td>\n",
       "      <td>10</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.521508</td>\n",
       "      <td>0.892898</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.581513</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.578178</td>\n",
       "      <td>4.719035</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.571485</td>\n",
       "      <td>0.614575</td>\n",
       "      <td>0.629596</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.000995</td>\n",
       "      <td>0.001246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Mean</td>\n",
       "      <td>None</td>\n",
       "      <td>NLP</td>\n",
       "      <td>11</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.288270</td>\n",
       "      <td>0.528838</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.465997</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.405470</td>\n",
       "      <td>3.309410</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.427383</td>\n",
       "      <td>0.430977</td>\n",
       "      <td>0.553572</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>0.000755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EQs</td>\n",
       "      <td>None</td>\n",
       "      <td>NLP</td>\n",
       "      <td>12</td>\n",
       "      <td>Biochemical Pathways</td>\n",
       "      <td>Filtered</td>\n",
       "      <td>0.786415</td>\n",
       "      <td>0.976061</td>\n",
       "      <td>221352.0</td>\n",
       "      <td>1585301.0</td>\n",
       "      <td>0.454901</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.454573</td>\n",
       "      <td>3.710176</td>\n",
       "      <td>0.218295</td>\n",
       "      <td>0.479610</td>\n",
       "      <td>0.448394</td>\n",
       "      <td>0.528150</td>\n",
       "      <td>0.000878</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>0.000784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Method                                   Hyperparameters Group  Order                 Topic      Data    mean_1    mean_0       n_1        n_0        ks        ks_pval       auc     ratio  baseline    f1_max    f5_max    f2_max    f1_std    f5_std    f2_std\n",
       "0                  GO                                              None   NLP      0  Biochemical Pathways  Filtered  0.889960  0.894548  221352.0  1585301.0  0.031194  1.382797e-164  0.132613  1.082373  0.218295  0.218295  0.149586  0.411120  0.000227  0.000593  0.000321\n",
       "1                  PO                                              None   NLP      1  Biochemical Pathways  Filtered  0.237214  0.283063  221352.0  1585301.0  0.095043   0.000000e+00  0.158489  1.293568  0.218295  0.230242  0.168826  0.411420  0.000506  0.000430  0.000497\n",
       "2             Doc2Vec                                Wikipedia,Size=300   NLP      2  Biochemical Pathways  Filtered  0.386482  0.445125  221352.0  1585301.0  0.243296   0.000000e+00  0.279419  2.280588  0.218295  0.302845  0.318254  0.432778  0.000552  0.000983  0.000579\n",
       "3            Word2Vec          Wikipedia,Size=300,Mean,Nouns,Adjectives   NLP      3  Biochemical Pathways  Filtered  0.288637  0.391489  221352.0  1585301.0  0.277843   0.000000e+00  0.265836  2.169724  0.218295  0.316445  0.304437  0.453653  0.000466  0.000992  0.000642\n",
       "4                BERT                                              Base   NLP      4  Biochemical Pathways  Filtered  0.139482  0.182477  221352.0  1585301.0  0.239832   0.000000e+00  0.210654  1.719338  0.218295  0.284718  0.237876  0.446076  0.000453  0.000850  0.000628\n",
       "5             BioBERT                  PubMed,PMC,Layers=4,Concatenated   NLP      5  Biochemical Pathways  Filtered  0.043455  0.059269  221352.0  1585301.0  0.257316   0.000000e+00  0.226192  1.846161  0.218295  0.291718  0.258599  0.455249  0.000594  0.000814  0.000694\n",
       "6             N-Grams  Simple,Words,Linares Pontes,1-grams,Binary,TFIDF   NLP      6  Biochemical Pathways  Filtered  0.867449  0.954430  221352.0  1585301.0  0.454706   0.000000e+00  0.389964  3.182851  0.218295  0.414292  0.408670  0.549205  0.000873  0.000906  0.000431\n",
       "7         NOBLE Coder                                     Partial,TFIDF   NLP      7  Biochemical Pathways  Filtered  0.647827  0.770909  221352.0  1585301.0  0.332715   0.000000e+00  0.300828  2.455329  0.218295  0.350433  0.340740  0.476043  0.001048  0.001046  0.000823\n",
       "8    Doc2Vec (Phenes)                                Wikipedia,Size=300   NLP      8  Biochemical Pathways  Filtered  0.184149  0.309738  221352.0  1585301.0  0.519054   0.000000e+00  0.463947  3.786689  0.218295  0.499558  0.552394  0.585175  0.000888  0.001312  0.000610\n",
       "9   Word2Vec (Phenes)          Wikipedia,Size=300,Mean,Nouns,Adjectives   NLP      9  Biochemical Pathways  Filtered  0.238348  0.408866  221352.0  1585301.0  0.372733   0.000000e+00  0.409688  3.343833  0.218295  0.418810  0.499060  0.497894  0.001198  0.001026  0.000610\n",
       "10   N-Grams (Phenes)                 Simple,Words,1-grams,Binary,TFIDF   NLP     10  Biochemical Pathways  Filtered  0.521508  0.892898  221352.0  1585301.0  0.581513   0.000000e+00  0.578178  4.719035  0.218295  0.571485  0.614575  0.629596  0.000908  0.000995  0.001246\n",
       "11               Mean                                              None   NLP     11  Biochemical Pathways  Filtered  0.288270  0.528838  221352.0  1585301.0  0.465997   0.000000e+00  0.405470  3.309410  0.218295  0.427383  0.430977  0.553572  0.000748  0.001348  0.000755\n",
       "12                EQs                                              None   NLP     12  Biochemical Pathways  Filtered  0.786415  0.976061  221352.0  1585301.0  0.454901   0.000000e+00  0.454573  3.710176  0.218295  0.479610  0.448394  0.528150  0.000878  0.000935  0.000784"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = pd.DataFrame(TABLE).transpose()\n",
    "columns = flatten([\"Hyperparameters\",\"Group\",\"Order\",\"Topic\",\"Data\",results.columns])\n",
    "results[\"Hyperparameters\"] = \"\"\n",
    "results[\"Group\"] = \"NLP\"\n",
    "results[\"Order\"] = np.arange(results.shape[0])\n",
    "results[\"Topic\"] = TOPIC\n",
    "results[\"Data\"] = DATA\n",
    "results = results[columns]\n",
    "results.reset_index(inplace=True)\n",
    "results = results.rename({\"index\":\"Method\"}, axis=\"columns\")\n",
    "hyperparam_sep = \":\"\n",
    "results[\"Hyperparameters\"] = results[\"Method\"].map(lambda x: x.split(hyperparam_sep)[1] if hyperparam_sep in x else \"None\")\n",
    "results[\"Method\"] = results[\"Method\"].map(lambda x: x.split(hyperparam_sep)[0])\n",
    "results.to_csv(os.path.join(OUTPUT_DIR,\"part_6_full_table.csv\"), index=False)\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
