{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at the Dataset\n",
    "The purpose of this notebook is to look closer at the dataset of genes, natural language descriptions, and ontology term annotations that are used in this work. As included in the preprocessing notebooks, these data are drawn from files from either publications supplements like Oellrich, Walls et al. (2015) or model species databases such as TAIR, MaizeGDB, and SGN. The datasets are already loaded and merged using classes available through the oats package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warming up PyWSD (takes ~10 secs)... took 4.339249134063721 secs.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.probability import FreqDist\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.annotation.annotation import term_enrichment\n",
    "from oats.nlp.vocabulary import token_enrichment\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>628658</td>\n",
       "      <td>9110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>156</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>342</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>772</td>\n",
       "      <td>389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>786</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>6526</td>\n",
       "      <td>998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>637240</td>\n",
       "      <td>11015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath     628658                 9110\n",
       "1     gmx        156                   49\n",
       "2     mtr        342                  155\n",
       "3     osa        772                  389\n",
       "4     sly        786                  314\n",
       "5     zma       6526                  998\n",
       "6   total     637240                11015"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = load_from_pickle(\"../data/pickles/gene_phenotype_dataset_all_text_and_annotations_unmerged.pickle\")\n",
    "data.to_pandas().head()\n",
    "data.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath       5851                 3527\n",
       "1     gmx         30                   24\n",
       "2     mtr         37                   36\n",
       "3     osa         92                   85\n",
       "4     sly         70                   70\n",
       "5     zma       1405                  811\n",
       "6   total       7485                 4553"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = load_from_pickle(\"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\")\n",
    "data.filter_has_description()\n",
    "data.to_pandas().head()\n",
    "data.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's there for each species?\n",
    "The previously loaded dataset contains all of the genes that across six plant species that have natural language description data for phenotype(s) related to that gene. Each gene can have multiple descriptions annotated to it, which were combined or concatenated when the datasets from multiple sources were merged in creating the pickled datasets. Arabidopsis has the highest number of genes that satisfy this criteria, followed by maize, and then followed by the other four species which have a relatively low number of genes that satisfy this criteria, atleast given the sources used for this work. Note that the number of unique descriptions is lower than the number of genes in call cases, because multiple genes can have the same phenotype description associated with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wnl = WordNetLemmatizer()\n",
    "lemmatize_doc = lambda d: [wnl.lemmatize(x) for x in simple_preprocess(d)]\n",
    "\n",
    "dists = defaultdict(list)\n",
    "\n",
    "sent_lists = {}\n",
    "token_lists = {}\n",
    "stems_lists = {}\n",
    "lemma_lists = {}\n",
    "\n",
    "\n",
    "# For each individual species.\n",
    "for species in data.get_species():\n",
    "    df = data.to_pandas()\n",
    "    subset = df[df[\"species\"]==species]\n",
    "    sentences = [sent_tokenize(d) for d in subset[\"description\"].values]\n",
    "    descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"description\"].values]\n",
    "    descriptions_stemmed = [preprocess_string(d) for d in subset[\"description\"].values]\n",
    "    descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"description\"].values]\n",
    "    sent_lists[species] = flatten(sentences)\n",
    "    token_lists[species] = flatten(descriptions_not_stemmed)\n",
    "    stems_lists[species] = flatten(descriptions_stemmed)    \n",
    "    lemma_lists[species] = flatten(descriptions_lemmatized)\n",
    "    \n",
    "    # What about the distributions of words per gene and sentences per gene?\n",
    "    dists[\"species\"].extend([species]*subset.shape[0])\n",
    "    dists[\"num_words\"].extend([len(word_tokenize(x)) for x in subset[\"description\"].values])\n",
    "    dists[\"num_sents\"].extend([len(sent_tokenize(x)) for x in subset[\"description\"].values])\n",
    "    \n",
    "# For the entire dataset including all of the species.\n",
    "df = data.to_pandas()\n",
    "subset = df\n",
    "sentences = [sent_tokenize(d) for d in subset[\"description\"].values]\n",
    "descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"description\"].values]\n",
    "descriptions_stemmed = [preprocess_string(d) for d in subset[\"description\"].values]\n",
    "descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"description\"].values]\n",
    "sent_lists[\"total\"] = flatten(sentences)\n",
    "token_lists[\"total\"] = flatten(descriptions_not_stemmed)\n",
    "stems_lists[\"total\"] = flatten(descriptions_stemmed)    \n",
    "lemma_lists[\"total\"] = flatten(descriptions_lemmatized)\n",
    "\n",
    "# What about lemmas that are uniquely used for a particular species?\n",
    "lemma_sets_unique_to_species = {}\n",
    "for species in data.get_species():\n",
    "    other_species = [s for s in data.get_species() if s != species]\n",
    "    lemmas_used_in_other_species = set(flatten([lemma_lists[s] for s in other_species]))\n",
    "    unique_lemmas = set(lemma_lists[species]).difference(lemmas_used_in_other_species)\n",
    "    lemma_sets_unique_to_species[species] = unique_lemmas\n",
    "lemma_sets_unique_to_species[\"total\"] = flatten([list(s) for s in lemma_sets_unique_to_species.values()])\n",
    "\n",
    "    \n",
    "# Create a dataframe to contain the summarizing information about this dataset, and sort it by number of genes.\n",
    "df = data.describe() \n",
    "condition = (df.species==\"total\")\n",
    "excluded = df[condition]\n",
    "included = df[~condition]\n",
    "df_sorted = included.sort_values(by=\"num_genes\", ascending=False)\n",
    "df = pd.concat([df_sorted,excluded])\n",
    "\n",
    "# Add columns summarizing information about the text descriptions in the dataset.\n",
    "df[\"total_sents\"] = df[\"species\"].map(lambda x: len(sent_lists[x]))\n",
    "df[\"total_words\"] = df[\"species\"].map(lambda x: len(token_lists[x]))\n",
    "df[\"unique_words\"] = df[\"species\"].map(lambda x: len(set(token_lists[x])))\n",
    "df[\"unique_stems\"] = df[\"species\"].map(lambda x: len(set(stems_lists[x])))\n",
    "df[\"total_lemmas\"] = df[\"species\"].map(lambda x: len(lemma_lists[x]))\n",
    "df[\"unique_lemmas\"] = df[\"species\"].map(lambda x: len(set(lemma_lists[x])))\n",
    "df[\"unique_lemmas_to_species\"] = df[\"species\"].map(lambda x: len(lemma_sets_unique_to_species[x]))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_words</th>\n",
       "      <th>num_sents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>157</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>22</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>182</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>104</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>127</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>343</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>49</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>169</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>212</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>31</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species  num_words  num_sents\n",
       "0      ath          8          2\n",
       "1      ath         17          4\n",
       "2      ath         16          2\n",
       "3      ath         30          3\n",
       "4      ath        157         11\n",
       "5      ath         22          4\n",
       "6      ath        182         33\n",
       "7      ath        104          7\n",
       "8      ath        127         13\n",
       "9      ath        343         30\n",
       "10     ath         49          6\n",
       "11     ath         40          3\n",
       "12     ath          3          1\n",
       "13     ath          3          1\n",
       "14     ath        169         15\n",
       "15     ath          5          1\n",
       "16     ath        212         16\n",
       "17     ath          5          1\n",
       "18     ath         23          3\n",
       "19     ath         31          7"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_distributions = pd.DataFrame(dists)\n",
    "text_distributions.to_csv(\"../data/scratch/word_sent_distributions.csv\", index=False)\n",
    "text_distributions.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the ontology term annotations for each species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many of the genes in this dataset for each species are mapped to atleast one term from a given ontology?\n",
    "num_mapped_go = {}\n",
    "num_mapped_po = {}\n",
    "for species in data.get_species():\n",
    "    d = data.to_pandas()\n",
    "    subset = d[d[\"species\"]==species]    \n",
    "    num_mapped_po[species] = len([t for t in subset[\"term_ids\"].values if \"PO\" in t])\n",
    "    num_mapped_go[species] = len([t for t in subset[\"term_ids\"].values if \"GO\" in t])\n",
    "num_mapped_go[\"total\"] = sum(list(num_mapped_go.values()))   \n",
    "num_mapped_po[\"total\"] = sum(list(num_mapped_po.values()))\n",
    "df[\"go\"] = df[\"species\"].map(lambda x: num_mapped_go[x])\n",
    "df[\"po\"] = df[\"species\"].map(lambda x: num_mapped_po[x])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the biologically relevant groups like biochemical pathways and phenotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "      <td>1345</td>\n",
       "      <td>807</td>\n",
       "      <td>2355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "      <td>1519</td>\n",
       "      <td>947</td>\n",
       "      <td>2355</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003  1345       807   2355\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111   156       133      0\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92     0         3      0\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65    17         2      0\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32     0         2      0\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27     1         0      0\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330  1519       947   2355"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the groupings that we're interested in mapping to?\n",
    "kegg_pathways_filename = \"../data/pickles/groupings_from_kegg_pathways.pickle\" \n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "lloyd_subsets_filename = \"../data/pickles/groupings_from_lloyd_subsets.pickle\"                     \n",
    "groupings_dict = {\"kegg\":kegg_pathways_filename,\"plantcyc\":pmn_pathways_filename,\"lloyd\":lloyd_subsets_filename}\n",
    "\n",
    "for name,filename in groupings_dict.items():\n",
    "    groups = load_from_pickle(filename)\n",
    "    id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "    group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "    species_dict = data.get_species_dictionary()\n",
    "    num_mapped = {}\n",
    "    for species in data.get_species():\n",
    "        num_mapped[species] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "    num_mapped[\"total\"] = sum(list(num_mapped.values()))    \n",
    "    df[name] = df[\"species\"].map(lambda x: num_mapped[x])  \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the other biologically relevant information like orthologous genes and protein interactions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd</th>\n",
       "      <th>panther</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "      <td>1345</td>\n",
       "      <td>807</td>\n",
       "      <td>2355</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "      <td>1519</td>\n",
       "      <td>947</td>\n",
       "      <td>2355</td>\n",
       "      <td>884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd  panther\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003  1345       807   2355      344\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111   156       133      0      443\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92     0         3      0       86\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65    17         2      0       11\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32     0         2      0        0\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27     1         0      0        0\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330  1519       947   2355      884"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PantherDB for plant orthologs.\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\"\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in ortholog_edgelist.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"panther\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd</th>\n",
       "      <th>panther</th>\n",
       "      <th>stringdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "      <td>1345</td>\n",
       "      <td>807</td>\n",
       "      <td>2355</td>\n",
       "      <td>344</td>\n",
       "      <td>3898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "      <td>1519</td>\n",
       "      <td>947</td>\n",
       "      <td>2355</td>\n",
       "      <td>884</td>\n",
       "      <td>4134</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd  panther  stringdb\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003  1345       807   2355      344      3898\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111   156       133      0      443       168\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92     0         3      0       86        45\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65    17         2      0       11         8\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32     0         2      0        0        13\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27     1         0      0        0         2\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330  1519       947   2355      884      4134"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# STRING DB for protein-protein interactions.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in string_data.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"stringdb\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the vocabularies used for different species compare?\n",
    "One of the things we are interested in is discovering or recovering phenotype similarity between different species in order to identify phenologs (phenotypes between species that share some underlying genetic cause). For this reason, we are interested in how the vocabularies used to describe phenotypes between different species vary, because this will impact how feasible it is to use a dataset like this to identify phenologs. Because the Arabidopsis and maize datasets are the largest in this case, we will compare the vocabularies used in describing the phenotypes associated with the genes from these species in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chang</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>-0.001999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seedligs</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>vice</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8</td>\n",
       "      <td>0.015991</td>\n",
       "      <td>-0.015991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>processed</td>\n",
       "      <td>VB</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glutamine</td>\n",
       "      <td>NN</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003408</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>allantoin</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>splitting</td>\n",
       "      <td>NN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001136</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mechanism</td>\n",
       "      <td>NN</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>groove</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>swize</td>\n",
       "      <td>VB</td>\n",
       "      <td>6</td>\n",
       "      <td>0.002272</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002272</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "0      chang             NN         0  0.000000         1  0.001999 -0.001999\n",
       "1   seedligs             NN         2  0.000757         0  0.000000  0.000757\n",
       "2       vice             NN         0  0.000000         8  0.015991 -0.015991\n",
       "3  processed             VB         2  0.000757         0  0.000000  0.000757\n",
       "4  glutamine             NN         9  0.003408         0  0.000000  0.003408\n",
       "5  allantoin             NN         2  0.000757         0  0.000000  0.000757\n",
       "6  splitting             NN         3  0.001136         0  0.000000  0.001136\n",
       "7  mechanism             NN         4  0.001515         0  0.000000  0.001515\n",
       "8     groove             NN         1  0.000379         0  0.000000  0.000379\n",
       "9      swize             VB         6  0.002272         0  0.000000  0.002272"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using lemmas as the vocabulary components.\n",
    "vocabs = {s:set(lemma_list) for s,lemma_list in lemma_lists.items()}\n",
    "fdist_zma = FreqDist(lemma_lists[\"zma\"])\n",
    "fdist_ath = FreqDist(lemma_lists[\"ath\"])\n",
    "\n",
    "# Using word stems as the vocabulary components.\n",
    "#vocabs = {s:set(stems_list) for s,stems_list in stems_lists.items()}\n",
    "#fdist_zma = FreqDist(stems_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(stems_lists[\"ath\"])\n",
    "\n",
    "# Using tokens (full words) as the vocabulary components.\n",
    "#vocabs = {s:set(token_list) for s,token_list in token_lists.items()}\n",
    "#fdist_zma = FreqDist(token_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(token_lists[\"ath\"])\n",
    "\n",
    "union_vocab = vocabs[\"zma\"].union(vocabs[\"ath\"])\n",
    "table = pd.DataFrame({\"token\":list(union_vocab)})\n",
    "stops = set(stopwords.words('english'))\n",
    "table = table[~table.token.isin(stops)]\n",
    "table[\"part_of_speech\"] = table[\"token\"].map(lambda x: nltk.pos_tag([x])[0][1][:2])\n",
    "table[\"ath_freq\"] = table[\"token\"].map(lambda x: fdist_ath[x])\n",
    "table[\"ath_rate\"] = table[\"ath_freq\"]*100/len(token_lists[\"ath\"])\n",
    "table[\"zma_freq\"] = table[\"token\"].map(lambda x: fdist_zma[x])\n",
    "table[\"zma_rate\"] = table[\"zma_freq\"]*100/len(token_lists[\"zma\"])\n",
    "table[\"diff\"] = table[\"ath_rate\"]-table[\"zma_rate\"]\n",
    "table.to_csv(\"../data/scratch/token_frequencies.csv\")\n",
    "table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>embryo</td>\n",
       "      <td>NN</td>\n",
       "      <td>4987</td>\n",
       "      <td>1.888307</td>\n",
       "      <td>151</td>\n",
       "      <td>0.301825</td>\n",
       "      <td>1.586482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4528</th>\n",
       "      <td>mutant</td>\n",
       "      <td>NN</td>\n",
       "      <td>4778</td>\n",
       "      <td>1.809170</td>\n",
       "      <td>256</td>\n",
       "      <td>0.511703</td>\n",
       "      <td>1.297467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5189</th>\n",
       "      <td>phenotype</td>\n",
       "      <td>NN</td>\n",
       "      <td>3583</td>\n",
       "      <td>1.356688</td>\n",
       "      <td>78</td>\n",
       "      <td>0.155910</td>\n",
       "      <td>1.200779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6603</th>\n",
       "      <td>root</td>\n",
       "      <td>NN</td>\n",
       "      <td>2942</td>\n",
       "      <td>1.113976</td>\n",
       "      <td>72</td>\n",
       "      <td>0.143917</td>\n",
       "      <td>0.970060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5992</th>\n",
       "      <td>type</td>\n",
       "      <td>NN</td>\n",
       "      <td>2540</td>\n",
       "      <td>0.961761</td>\n",
       "      <td>15</td>\n",
       "      <td>0.029983</td>\n",
       "      <td>0.931778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5794</th>\n",
       "      <td>wild</td>\n",
       "      <td>NN</td>\n",
       "      <td>2456</td>\n",
       "      <td>0.929954</td>\n",
       "      <td>7</td>\n",
       "      <td>0.013992</td>\n",
       "      <td>0.915962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6818</th>\n",
       "      <td>defective</td>\n",
       "      <td>JJ</td>\n",
       "      <td>3332</td>\n",
       "      <td>1.261648</td>\n",
       "      <td>285</td>\n",
       "      <td>0.569670</td>\n",
       "      <td>0.691979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>reduced</td>\n",
       "      <td>VB</td>\n",
       "      <td>2857</td>\n",
       "      <td>1.081791</td>\n",
       "      <td>216</td>\n",
       "      <td>0.431750</td>\n",
       "      <td>0.650042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5697</th>\n",
       "      <td>stage</td>\n",
       "      <td>NN</td>\n",
       "      <td>1814</td>\n",
       "      <td>0.686864</td>\n",
       "      <td>63</td>\n",
       "      <td>0.125927</td>\n",
       "      <td>0.560937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3596</th>\n",
       "      <td>cotyledon</td>\n",
       "      <td>NN</td>\n",
       "      <td>1439</td>\n",
       "      <td>0.544871</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.544871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2549</th>\n",
       "      <td>growth</td>\n",
       "      <td>NN</td>\n",
       "      <td>1829</td>\n",
       "      <td>0.692543</td>\n",
       "      <td>79</td>\n",
       "      <td>0.157908</td>\n",
       "      <td>0.534635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3251</th>\n",
       "      <td>seed</td>\n",
       "      <td>NN</td>\n",
       "      <td>1890</td>\n",
       "      <td>0.715641</td>\n",
       "      <td>113</td>\n",
       "      <td>0.225869</td>\n",
       "      <td>0.489772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2504</th>\n",
       "      <td>increased</td>\n",
       "      <td>VB</td>\n",
       "      <td>1315</td>\n",
       "      <td>0.497919</td>\n",
       "      <td>11</td>\n",
       "      <td>0.021987</td>\n",
       "      <td>0.475932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6933</th>\n",
       "      <td>terminal</td>\n",
       "      <td>NN</td>\n",
       "      <td>1268</td>\n",
       "      <td>0.480123</td>\n",
       "      <td>4</td>\n",
       "      <td>0.007995</td>\n",
       "      <td>0.472128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2644</th>\n",
       "      <td>level</td>\n",
       "      <td>NN</td>\n",
       "      <td>1336</td>\n",
       "      <td>0.505871</td>\n",
       "      <td>22</td>\n",
       "      <td>0.043974</td>\n",
       "      <td>0.461896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5707</th>\n",
       "      <td>cell</td>\n",
       "      <td>NN</td>\n",
       "      <td>1646</td>\n",
       "      <td>0.623251</td>\n",
       "      <td>112</td>\n",
       "      <td>0.223870</td>\n",
       "      <td>0.399381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1220</th>\n",
       "      <td>rosette</td>\n",
       "      <td>NN</td>\n",
       "      <td>966</td>\n",
       "      <td>0.365772</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>0.363773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4672</th>\n",
       "      <td>double</td>\n",
       "      <td>RB</td>\n",
       "      <td>956</td>\n",
       "      <td>0.361985</td>\n",
       "      <td>3</td>\n",
       "      <td>0.005997</td>\n",
       "      <td>0.355989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2647</th>\n",
       "      <td>globular</td>\n",
       "      <td>NN</td>\n",
       "      <td>905</td>\n",
       "      <td>0.342675</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.342675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1384</th>\n",
       "      <td>abnormal</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1048</td>\n",
       "      <td>0.396821</td>\n",
       "      <td>28</td>\n",
       "      <td>0.055968</td>\n",
       "      <td>0.340853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3061</th>\n",
       "      <td>wa</td>\n",
       "      <td>NN</td>\n",
       "      <td>886</td>\n",
       "      <td>0.335480</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>0.333481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3400</th>\n",
       "      <td>flower</td>\n",
       "      <td>NN</td>\n",
       "      <td>940</td>\n",
       "      <td>0.355927</td>\n",
       "      <td>14</td>\n",
       "      <td>0.027984</td>\n",
       "      <td>0.327943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5893</th>\n",
       "      <td>flowering</td>\n",
       "      <td>VB</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.390005</td>\n",
       "      <td>32</td>\n",
       "      <td>0.063963</td>\n",
       "      <td>0.326042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4006</th>\n",
       "      <td>short</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1504</td>\n",
       "      <td>0.569483</td>\n",
       "      <td>123</td>\n",
       "      <td>0.245857</td>\n",
       "      <td>0.323626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3358</th>\n",
       "      <td>compared</td>\n",
       "      <td>VB</td>\n",
       "      <td>882</td>\n",
       "      <td>0.333966</td>\n",
       "      <td>9</td>\n",
       "      <td>0.017990</td>\n",
       "      <td>0.315976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3045</th>\n",
       "      <td>silique</td>\n",
       "      <td>NN</td>\n",
       "      <td>798</td>\n",
       "      <td>0.302159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.302159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3633</th>\n",
       "      <td>sensitive</td>\n",
       "      <td>JJ</td>\n",
       "      <td>826</td>\n",
       "      <td>0.312762</td>\n",
       "      <td>6</td>\n",
       "      <td>0.011993</td>\n",
       "      <td>0.300768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4334</th>\n",
       "      <td>visible</td>\n",
       "      <td>JJ</td>\n",
       "      <td>838</td>\n",
       "      <td>0.317305</td>\n",
       "      <td>13</td>\n",
       "      <td>0.025985</td>\n",
       "      <td>0.291320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4689</th>\n",
       "      <td>development</td>\n",
       "      <td>NN</td>\n",
       "      <td>862</td>\n",
       "      <td>0.326393</td>\n",
       "      <td>19</td>\n",
       "      <td>0.037978</td>\n",
       "      <td>0.288415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6242</th>\n",
       "      <td>hypocotyl</td>\n",
       "      <td>NN</td>\n",
       "      <td>752</td>\n",
       "      <td>0.284742</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.284742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "310        embryo             NN      4987  1.888307       151  0.301825  1.586482\n",
       "4528       mutant             NN      4778  1.809170       256  0.511703  1.297467\n",
       "5189    phenotype             NN      3583  1.356688        78  0.155910  1.200779\n",
       "6603         root             NN      2942  1.113976        72  0.143917  0.970060\n",
       "5992         type             NN      2540  0.961761        15  0.029983  0.931778\n",
       "5794         wild             NN      2456  0.929954         7  0.013992  0.915962\n",
       "6818    defective             JJ      3332  1.261648       285  0.569670  0.691979\n",
       "1664      reduced             VB      2857  1.081791       216  0.431750  0.650042\n",
       "5697        stage             NN      1814  0.686864        63  0.125927  0.560937\n",
       "3596    cotyledon             NN      1439  0.544871         0  0.000000  0.544871\n",
       "2549       growth             NN      1829  0.692543        79  0.157908  0.534635\n",
       "3251         seed             NN      1890  0.715641       113  0.225869  0.489772\n",
       "2504    increased             VB      1315  0.497919        11  0.021987  0.475932\n",
       "6933     terminal             NN      1268  0.480123         4  0.007995  0.472128\n",
       "2644        level             NN      1336  0.505871        22  0.043974  0.461896\n",
       "5707         cell             NN      1646  0.623251       112  0.223870  0.399381\n",
       "1220      rosette             NN       966  0.365772         1  0.001999  0.363773\n",
       "4672       double             RB       956  0.361985         3  0.005997  0.355989\n",
       "2647     globular             NN       905  0.342675         0  0.000000  0.342675\n",
       "1384     abnormal             JJ      1048  0.396821        28  0.055968  0.340853\n",
       "3061           wa             NN       886  0.335480         1  0.001999  0.333481\n",
       "3400       flower             NN       940  0.355927        14  0.027984  0.327943\n",
       "5893    flowering             VB      1030  0.390005        32  0.063963  0.326042\n",
       "4006        short             JJ      1504  0.569483       123  0.245857  0.323626\n",
       "3358     compared             VB       882  0.333966         9  0.017990  0.315976\n",
       "3045      silique             NN       798  0.302159         0  0.000000  0.302159\n",
       "3633    sensitive             JJ       826  0.312762         6  0.011993  0.300768\n",
       "4334      visible             JJ       838  0.317305        13  0.025985  0.291320\n",
       "4689  development             NN       862  0.326393        19  0.037978  0.288415\n",
       "6242    hypocotyl             NN       752  0.284742         0  0.000000  0.284742"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for Arabidopsis than maize descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=False, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4794</th>\n",
       "      <td>endosperm</td>\n",
       "      <td>NN</td>\n",
       "      <td>124</td>\n",
       "      <td>0.046952</td>\n",
       "      <td>1078</td>\n",
       "      <td>2.154750</td>\n",
       "      <td>-2.107798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>seedling</td>\n",
       "      <td>VB</td>\n",
       "      <td>1560</td>\n",
       "      <td>0.590688</td>\n",
       "      <td>1318</td>\n",
       "      <td>2.634472</td>\n",
       "      <td>-2.043784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>kernel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>766</td>\n",
       "      <td>1.531112</td>\n",
       "      <td>-1.531112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>yellow</td>\n",
       "      <td>NN</td>\n",
       "      <td>304</td>\n",
       "      <td>0.115108</td>\n",
       "      <td>775</td>\n",
       "      <td>1.549102</td>\n",
       "      <td>-1.433993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4670</th>\n",
       "      <td>leaf</td>\n",
       "      <td>NN</td>\n",
       "      <td>3876</td>\n",
       "      <td>1.467631</td>\n",
       "      <td>1445</td>\n",
       "      <td>2.888325</td>\n",
       "      <td>-1.420693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4318</th>\n",
       "      <td>green</td>\n",
       "      <td>JJ</td>\n",
       "      <td>883</td>\n",
       "      <td>0.334344</td>\n",
       "      <td>788</td>\n",
       "      <td>1.575086</td>\n",
       "      <td>-1.240742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5367</th>\n",
       "      <td>white</td>\n",
       "      <td>JJ</td>\n",
       "      <td>375</td>\n",
       "      <td>0.141992</td>\n",
       "      <td>642</td>\n",
       "      <td>1.283256</td>\n",
       "      <td>-1.141264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5355</th>\n",
       "      <td>albino</td>\n",
       "      <td>NN</td>\n",
       "      <td>222</td>\n",
       "      <td>0.084059</td>\n",
       "      <td>396</td>\n",
       "      <td>0.791541</td>\n",
       "      <td>-0.707482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2337</th>\n",
       "      <td>usually</td>\n",
       "      <td>RB</td>\n",
       "      <td>46</td>\n",
       "      <td>0.017418</td>\n",
       "      <td>353</td>\n",
       "      <td>0.705591</td>\n",
       "      <td>-0.688173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>color</td>\n",
       "      <td>NN</td>\n",
       "      <td>43</td>\n",
       "      <td>0.016282</td>\n",
       "      <td>330</td>\n",
       "      <td>0.659617</td>\n",
       "      <td>-0.643336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5620</th>\n",
       "      <td>tassel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>310</td>\n",
       "      <td>0.619641</td>\n",
       "      <td>-0.619641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1670</th>\n",
       "      <td>ear</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>284</td>\n",
       "      <td>0.567671</td>\n",
       "      <td>-0.567671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4221</th>\n",
       "      <td>often</td>\n",
       "      <td>RB</td>\n",
       "      <td>149</td>\n",
       "      <td>0.056418</td>\n",
       "      <td>278</td>\n",
       "      <td>0.555678</td>\n",
       "      <td>-0.499259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1613</th>\n",
       "      <td>aleurone</td>\n",
       "      <td>NN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001136</td>\n",
       "      <td>248</td>\n",
       "      <td>0.495712</td>\n",
       "      <td>-0.494577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2103</th>\n",
       "      <td>opaque</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>244</td>\n",
       "      <td>0.487717</td>\n",
       "      <td>-0.487338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3925</th>\n",
       "      <td>class</td>\n",
       "      <td>NN</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003029</td>\n",
       "      <td>240</td>\n",
       "      <td>0.479722</td>\n",
       "      <td>-0.476693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6212</th>\n",
       "      <td>tissue</td>\n",
       "      <td>NN</td>\n",
       "      <td>236</td>\n",
       "      <td>0.089360</td>\n",
       "      <td>259</td>\n",
       "      <td>0.517700</td>\n",
       "      <td>-0.428339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2903</th>\n",
       "      <td>small</td>\n",
       "      <td>JJ</td>\n",
       "      <td>669</td>\n",
       "      <td>0.253314</td>\n",
       "      <td>323</td>\n",
       "      <td>0.645626</td>\n",
       "      <td>-0.392311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4005</th>\n",
       "      <td>pale</td>\n",
       "      <td>NN</td>\n",
       "      <td>651</td>\n",
       "      <td>0.246498</td>\n",
       "      <td>318</td>\n",
       "      <td>0.635631</td>\n",
       "      <td>-0.389133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1644</th>\n",
       "      <td>vp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190</td>\n",
       "      <td>0.379780</td>\n",
       "      <td>-0.379780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2420</th>\n",
       "      <td>palegreen</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>171</td>\n",
       "      <td>0.341802</td>\n",
       "      <td>-0.341423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055</th>\n",
       "      <td>general</td>\n",
       "      <td>JJ</td>\n",
       "      <td>13</td>\n",
       "      <td>0.004922</td>\n",
       "      <td>171</td>\n",
       "      <td>0.341802</td>\n",
       "      <td>-0.336879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1470</th>\n",
       "      <td>term</td>\n",
       "      <td>NN</td>\n",
       "      <td>22</td>\n",
       "      <td>0.008330</td>\n",
       "      <td>172</td>\n",
       "      <td>0.343801</td>\n",
       "      <td>-0.335470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4491</th>\n",
       "      <td>surface</td>\n",
       "      <td>NN</td>\n",
       "      <td>138</td>\n",
       "      <td>0.052253</td>\n",
       "      <td>190</td>\n",
       "      <td>0.379780</td>\n",
       "      <td>-0.327527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6109</th>\n",
       "      <td>appearance</td>\n",
       "      <td>NN</td>\n",
       "      <td>131</td>\n",
       "      <td>0.049603</td>\n",
       "      <td>188</td>\n",
       "      <td>0.375782</td>\n",
       "      <td>-0.326179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7006</th>\n",
       "      <td>texture</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>163</td>\n",
       "      <td>0.325811</td>\n",
       "      <td>-0.325054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2193</th>\n",
       "      <td>otherwise</td>\n",
       "      <td>RB</td>\n",
       "      <td>10</td>\n",
       "      <td>0.003786</td>\n",
       "      <td>158</td>\n",
       "      <td>0.315817</td>\n",
       "      <td>-0.312030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7061</th>\n",
       "      <td>anther</td>\n",
       "      <td>NN</td>\n",
       "      <td>139</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>167</td>\n",
       "      <td>0.333806</td>\n",
       "      <td>-0.281175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3406</th>\n",
       "      <td>pericarp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>140</td>\n",
       "      <td>0.279838</td>\n",
       "      <td>-0.279838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3510</th>\n",
       "      <td>sometimes</td>\n",
       "      <td>RB</td>\n",
       "      <td>44</td>\n",
       "      <td>0.016660</td>\n",
       "      <td>148</td>\n",
       "      <td>0.295828</td>\n",
       "      <td>-0.279168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "4794   endosperm             NN       124  0.046952      1078  2.154750 -2.107798\n",
       "379     seedling             VB      1560  0.590688      1318  2.634472 -2.043784\n",
       "667       kernel             NN         0  0.000000       766  1.531112 -1.531112\n",
       "238       yellow             NN       304  0.115108       775  1.549102 -1.433993\n",
       "4670        leaf             NN      3876  1.467631      1445  2.888325 -1.420693\n",
       "4318       green             JJ       883  0.334344       788  1.575086 -1.240742\n",
       "5367       white             JJ       375  0.141992       642  1.283256 -1.141264\n",
       "5355      albino             NN       222  0.084059       396  0.791541 -0.707482\n",
       "2337     usually             RB        46  0.017418       353  0.705591 -0.688173\n",
       "80         color             NN        43  0.016282       330  0.659617 -0.643336\n",
       "5620      tassel             NN         0  0.000000       310  0.619641 -0.619641\n",
       "1670         ear             NN         0  0.000000       284  0.567671 -0.567671\n",
       "4221       often             RB       149  0.056418       278  0.555678 -0.499259\n",
       "1613    aleurone             NN         3  0.001136       248  0.495712 -0.494577\n",
       "2103      opaque             NN         1  0.000379       244  0.487717 -0.487338\n",
       "3925       class             NN         8  0.003029       240  0.479722 -0.476693\n",
       "6212      tissue             NN       236  0.089360       259  0.517700 -0.428339\n",
       "2903       small             JJ       669  0.253314       323  0.645626 -0.392311\n",
       "4005        pale             NN       651  0.246498       318  0.635631 -0.389133\n",
       "1644          vp             NN         0  0.000000       190  0.379780 -0.379780\n",
       "2420   palegreen             NN         1  0.000379       171  0.341802 -0.341423\n",
       "2055     general             JJ        13  0.004922       171  0.341802 -0.336879\n",
       "1470        term             NN        22  0.008330       172  0.343801 -0.335470\n",
       "4491     surface             NN       138  0.052253       190  0.379780 -0.327527\n",
       "6109  appearance             NN       131  0.049603       188  0.375782 -0.326179\n",
       "7006     texture             NN         2  0.000757       163  0.325811 -0.325054\n",
       "2193   otherwise             RB        10  0.003786       158  0.315817 -0.312030\n",
       "7061      anther             NN       139  0.052632       167  0.333806 -0.281175\n",
       "3406    pericarp             NN         0  0.000000       140  0.279838 -0.279838\n",
       "3510   sometimes             RB        44  0.016660       148  0.295828 -0.279168"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for maize than Arabidopsis descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=True, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MD</td>\n",
       "      <td>0.059431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD</td>\n",
       "      <td>0.027896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IN</td>\n",
       "      <td>0.026204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JJ</td>\n",
       "      <td>0.023686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DT</td>\n",
       "      <td>0.019075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RB</td>\n",
       "      <td>0.016302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NN</td>\n",
       "      <td>0.014463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VB</td>\n",
       "      <td>0.012130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC</td>\n",
       "      <td>0.007617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WP</td>\n",
       "      <td>0.001620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  part_of_speech  abs_diff\n",
       "0             MD  0.059431\n",
       "1             CD  0.027896\n",
       "2             IN  0.026204\n",
       "3             JJ  0.023686\n",
       "4             DT  0.019075\n",
       "5             RB  0.016302\n",
       "6             NN  0.014463\n",
       "7             VB  0.012130\n",
       "8             CC  0.007617\n",
       "9             WP  0.001620"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is the mean absolute value of the rate differences different between the different parts of speech?\n",
    "table[\"abs_diff\"] = abs(table[\"diff\"])\n",
    "pos_table = table.groupby(\"part_of_speech\").mean()\n",
    "pos_table.sort_values(by=\"abs_diff\", inplace=True, ascending=False)\n",
    "pos_table = pos_table[[\"abs_diff\"]]\n",
    "pos_table.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Working on the Venn Diagram for this part, unused currently.\n",
    "#print(table.shape)\n",
    "#zma_only = table[table[\"ath_rate\"]==0]\n",
    "#ath_only = table[table[\"zma_rate\"]==0]\n",
    "#print(zma_only.shape)\n",
    "#print(ath_only.shape)\n",
    "#print(ath_only.shape[0]+zma_only.shape[0])\n",
    "#ath_only.head(10)\n",
    "# We need to create a mapping between stems and the words that were present for them.\n",
    "# This is because what we want is the stems that are exclusive to a species.\n",
    "# but then the words that are actually there for those stems, so that we can count their parts of speech."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at Term and Word Enrichment for Groups of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3815"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the dataset of phenotype descriptions and ontology annotations.\n",
    "data = load_from_pickle(\"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\")\n",
    "data.filter_has_description()\n",
    "data.filter_has_annotation(\"GO\")\n",
    "data.filter_has_annotation(\"PO\")\n",
    "d = data.get_description_dictionary()\n",
    "texts = {i:\" \".join(simple_preprocess(t)) for i,t in d.items()}\n",
    "len(texts)                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create ontology objects for all the biological ontologies being used.\n",
    "go_filename = \"../ontologies/go.obo\"                                                                \n",
    "po_filename = \"../ontologies/po.obo\"                                                                 \n",
    "pato_filename = \"../ontologies/pato.obo\"   \n",
    "pato = Ontology(pato_filename)\n",
    "po = Ontology(po_filename)\n",
    "go = Ontology(go_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "curated_go_annotations = data.get_annotations_dictionary(\"GO\")\n",
    "curated_po_annotations = data.get_annotations_dictionary(\"PO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>num_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sporopollenin precursors biosynthesis</td>\n",
       "      <td>PWY-6733</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gluconeogenesis III</td>\n",
       "      <td>PWY66-399</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gluconeogenesis I</td>\n",
       "      <td>GLUCONEO-PWY</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>suberin monomers biosynthesis</td>\n",
       "      <td>PWY-1121</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phosphatidylcholine acyl editing</td>\n",
       "      <td>PWY-6803</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>indole-3-acetate biosynthesis II</td>\n",
       "      <td>PWY-581</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>glycolysis I (from glucose 6-phosphate)</td>\n",
       "      <td>GLYCOLYSIS</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>glycolysis II (from fructose 6-phosphate)</td>\n",
       "      <td>PWY-5484</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>palmitate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5971</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3-phosphoinositide biosynthesis</td>\n",
       "      <td>PWY-6352</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>phenylpropanoid biosynthesis</td>\n",
       "      <td>PWY-361</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>L-leucine biosynthesis</td>\n",
       "      <td>LEUSYN-PWY</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>starch biosynthesis</td>\n",
       "      <td>PWY-622</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>glycolysis IV (plant cytosol)</td>\n",
       "      <td>PWY-1042</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>stearate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5989</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       pathway_name    pathway_id  num_genes\n",
       "0             sporopollenin precursors biosynthesis      PWY-6733         21\n",
       "1                               gluconeogenesis III     PWY66-399         20\n",
       "2                                 gluconeogenesis I  GLUCONEO-PWY         20\n",
       "3                     suberin monomers biosynthesis      PWY-1121         20\n",
       "4                  phosphatidylcholine acyl editing      PWY-6803         19\n",
       "5                  indole-3-acetate biosynthesis II       PWY-581         18\n",
       "6           glycolysis I (from glucose 6-phosphate)    GLYCOLYSIS         16\n",
       "7         glycolysis II (from fructose 6-phosphate)      PWY-5484         15\n",
       "8   palmitate biosynthesis II (bacteria and plants)      PWY-5971         15\n",
       "9                   3-phosphoinositide biosynthesis      PWY-6352         15\n",
       "10                     phenylpropanoid biosynthesis       PWY-361         15\n",
       "11                           L-leucine biosynthesis    LEUSYN-PWY         14\n",
       "12                              starch biosynthesis       PWY-622         14\n",
       "13                    glycolysis IV (plant cytosol)      PWY-1042         14\n",
       "14   stearate biosynthesis II (bacteria and plants)      PWY-5989         14"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the mappings from this dataset to PlantCyc information.\n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "groups = load_from_pickle(pmn_pathways_filename)\n",
    "id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "\n",
    "# Look at which pathways are best represented in this dataset.\n",
    "pathways_sorted = sorted(group_id_to_ids.items(), key=lambda item: len(item[1]), reverse=True)\n",
    "pathways_sorted_lengths = [(i,len(l)) for (i,l) in pathways_sorted]\n",
    "pathways_df = pd.DataFrame(pathways_sorted_lengths, columns=[\"pathway_id\",\"num_genes\"])\n",
    "pathways_df[\"pathway_name\"] = pathways_df[\"pathway_id\"].map(lambda x: groups.get_long_name(x))\n",
    "pathways_df = pathways_df[[\"pathway_name\",\"pathway_id\",\"num_genes\"]]\n",
    "pathways_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For some example pathway to use.\n",
    "pathway_id = \"PWY-6733\"\n",
    "gene_ids_in_this_pathway = group_id_to_ids[pathway_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PO:0020008</td>\n",
       "      <td>exine</td>\n",
       "      <td>9</td>\n",
       "      <td>3806</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>5.718655e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PO:0025313</td>\n",
       "      <td>tapetum</td>\n",
       "      <td>33</td>\n",
       "      <td>3782</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>3.593559e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PO:0025059</td>\n",
       "      <td>portion of ground tissue</td>\n",
       "      <td>173</td>\n",
       "      <td>3642</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>1.797079e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PO:0009071</td>\n",
       "      <td>anther wall tapetum</td>\n",
       "      <td>29</td>\n",
       "      <td>3786</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>2.452198e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PO:0025314</td>\n",
       "      <td>microsporangium tapetum</td>\n",
       "      <td>29</td>\n",
       "      <td>3786</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>2.452198e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PO:0005059</td>\n",
       "      <td>root endodermis</td>\n",
       "      <td>18</td>\n",
       "      <td>3797</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>1.765972e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PO:0000252</td>\n",
       "      <td>endodermis</td>\n",
       "      <td>25</td>\n",
       "      <td>3790</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>4.243818e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PO:0009088</td>\n",
       "      <td>seed coat</td>\n",
       "      <td>37</td>\n",
       "      <td>3778</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>1.226875e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PO:0009066</td>\n",
       "      <td>anther</td>\n",
       "      <td>136</td>\n",
       "      <td>3679</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>6.267905e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PO:0009002</td>\n",
       "      <td>plant cell</td>\n",
       "      <td>3381</td>\n",
       "      <td>434</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>6.969615e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PO:0028003</td>\n",
       "      <td>gametophyte development stage</td>\n",
       "      <td>857</td>\n",
       "      <td>2958</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>7.322126e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PO:0009031</td>\n",
       "      <td>sepal</td>\n",
       "      <td>2910</td>\n",
       "      <td>905</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>7.349703e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PO:0025602</td>\n",
       "      <td>microgametophyte development stage</td>\n",
       "      <td>849</td>\n",
       "      <td>2966</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>7.415549e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PO:0001007</td>\n",
       "      <td>pollen development stage</td>\n",
       "      <td>849</td>\n",
       "      <td>2966</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>7.415549e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PO:0025606</td>\n",
       "      <td>native plant cell</td>\n",
       "      <td>3375</td>\n",
       "      <td>440</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>7.508613e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PO:0025395</td>\n",
       "      <td>floral organ</td>\n",
       "      <td>3018</td>\n",
       "      <td>797</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1.271618e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PO:0001017</td>\n",
       "      <td>M germinated pollen stage</td>\n",
       "      <td>808</td>\n",
       "      <td>3007</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>1.274945e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PO:0025202</td>\n",
       "      <td>microsporangium</td>\n",
       "      <td>3</td>\n",
       "      <td>3812</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2.172708e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PO:0025277</td>\n",
       "      <td>pollen sac</td>\n",
       "      <td>3</td>\n",
       "      <td>3812</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2.172708e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PO:0025338</td>\n",
       "      <td>collective plant organ structure development s...</td>\n",
       "      <td>3095</td>\n",
       "      <td>720</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>2.187300e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PO:0030114</td>\n",
       "      <td>fruit pedicel</td>\n",
       "      <td>4</td>\n",
       "      <td>3811</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2.708817e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PO:0004536</td>\n",
       "      <td>infructescence fruit pedicel</td>\n",
       "      <td>4</td>\n",
       "      <td>3811</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2.708817e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PO:0004013</td>\n",
       "      <td>epidermal cell</td>\n",
       "      <td>3249</td>\n",
       "      <td>566</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>2.778058e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PO:0009073</td>\n",
       "      <td>stigma</td>\n",
       "      <td>48</td>\n",
       "      <td>3767</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2.986253e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PO:0025165</td>\n",
       "      <td>shoot epidermal cell</td>\n",
       "      <td>3224</td>\n",
       "      <td>591</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>3.434285e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>guard cell</td>\n",
       "      <td>3217</td>\n",
       "      <td>598</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>3.635925e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PO:0009067</td>\n",
       "      <td>filament</td>\n",
       "      <td>54</td>\n",
       "      <td>3761</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>3.681205e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PO:0001016</td>\n",
       "      <td>L mature pollen stage</td>\n",
       "      <td>662</td>\n",
       "      <td>3153</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>3.725107e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PO:0009007</td>\n",
       "      <td>portion of plant tissue</td>\n",
       "      <td>2900</td>\n",
       "      <td>915</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>3.910610e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PO:0009072</td>\n",
       "      <td>plant ovary</td>\n",
       "      <td>7</td>\n",
       "      <td>3808</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>4.300410e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       term_id                                         term_label  genes_with  genes_without  group_genes_with  group_genes_without       p_value\n",
       "0   PO:0020008                                              exine           9           3806                 5                   16  5.718655e-09\n",
       "1   PO:0025313                                            tapetum          33           3782                 6                   15  3.593559e-08\n",
       "2   PO:0025059                           portion of ground tissue         173           3642                 9                   12  1.797079e-07\n",
       "3   PO:0009071                                anther wall tapetum          29           3786                 4                   17  2.452198e-05\n",
       "4   PO:0025314                            microsporangium tapetum          29           3786                 4                   17  2.452198e-05\n",
       "5   PO:0005059                                    root endodermis          18           3797                 3                   18  1.765972e-04\n",
       "6   PO:0000252                                         endodermis          25           3790                 3                   18  4.243818e-04\n",
       "7   PO:0009088                                          seed coat          37           3778                 3                   18  1.226875e-03\n",
       "8   PO:0009066                                             anther         136           3679                 4                   17  6.267905e-03\n",
       "9   PO:0009002                                         plant cell        3381            434                14                    7  6.969615e-03\n",
       "10  PO:0028003                      gametophyte development stage         857           2958                 0                   21  7.322126e-03\n",
       "11  PO:0009031                                              sepal        2910            905                21                    0  7.349703e-03\n",
       "12  PO:0025602                 microgametophyte development stage         849           2966                 0                   21  7.415549e-03\n",
       "13  PO:0001007                           pollen development stage         849           2966                 0                   21  7.415549e-03\n",
       "14  PO:0025606                                  native plant cell        3375            440                14                    7  7.508613e-03\n",
       "15  PO:0025395                                       floral organ        3018            797                21                    0  1.271618e-02\n",
       "16  PO:0001017                          M germinated pollen stage         808           3007                 0                   21  1.274945e-02\n",
       "17  PO:0025202                                    microsporangium           3           3812                 1                   20  2.172708e-02\n",
       "18  PO:0025277                                         pollen sac           3           3812                 1                   20  2.172708e-02\n",
       "19  PO:0025338  collective plant organ structure development s...        3095            720                21                    0  2.187300e-02\n",
       "20  PO:0030114                                      fruit pedicel           4           3811                 1                   20  2.708817e-02\n",
       "21  PO:0004536                       infructescence fruit pedicel           4           3811                 1                   20  2.708817e-02\n",
       "22  PO:0004013                                     epidermal cell        3249            566                14                    7  2.778058e-02\n",
       "23  PO:0009073                                             stigma          48           3767                 2                   19  2.986253e-02\n",
       "24  PO:0025165                               shoot epidermal cell        3224            591                14                    7  3.434285e-02\n",
       "25  PO:0000293                                         guard cell        3217            598                14                    7  3.635925e-02\n",
       "26  PO:0009067                                           filament          54           3761                 2                   19  3.681205e-02\n",
       "27  PO:0001016                              L mature pollen stage         662           3153                 0                   21  3.725107e-02\n",
       "28  PO:0009007                            portion of plant tissue        2900            915                20                    1  3.910610e-02\n",
       "29  PO:0009072                                        plant ovary           7           3808                 1                   20  4.300410e-02"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_enrichment(curated_po_annotations, gene_ids_in_this_pathway, po).head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>layer</td>\n",
       "      <td>65</td>\n",
       "      <td>3750</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>5.499473e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exine</td>\n",
       "      <td>19</td>\n",
       "      <td>3796</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>2.045050e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>suberin</td>\n",
       "      <td>8</td>\n",
       "      <td>3807</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>2.269517e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lacs</td>\n",
       "      <td>2</td>\n",
       "      <td>3813</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.701698e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cutin</td>\n",
       "      <td>4</td>\n",
       "      <td>3811</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>4.226197e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>wax</td>\n",
       "      <td>32</td>\n",
       "      <td>3783</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>8.271913e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>permeability</td>\n",
       "      <td>7</td>\n",
       "      <td>3808</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.004279e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>flavonoids</td>\n",
       "      <td>8</td>\n",
       "      <td>3807</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.251207e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>flavonoid</td>\n",
       "      <td>9</td>\n",
       "      <td>3806</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.524210e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pollen</td>\n",
       "      <td>251</td>\n",
       "      <td>3564</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>1.964829e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>glossy</td>\n",
       "      <td>11</td>\n",
       "      <td>3804</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2.147377e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>composition</td>\n",
       "      <td>46</td>\n",
       "      <td>3769</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>2.216439e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>no</td>\n",
       "      <td>981</td>\n",
       "      <td>2834</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>2.269727e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cuticle</td>\n",
       "      <td>12</td>\n",
       "      <td>3803</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2.497017e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>visible</td>\n",
       "      <td>611</td>\n",
       "      <td>3204</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>3.333864e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>synergistic</td>\n",
       "      <td>15</td>\n",
       "      <td>3800</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>3.695047e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>precursors</td>\n",
       "      <td>16</td>\n",
       "      <td>3799</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>4.143240e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>drastically</td>\n",
       "      <td>16</td>\n",
       "      <td>3799</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>4.143240e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>abnormal</td>\n",
       "      <td>511</td>\n",
       "      <td>3304</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>4.291804e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>coat</td>\n",
       "      <td>60</td>\n",
       "      <td>3755</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>4.547485e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>epicuticular</td>\n",
       "      <td>17</td>\n",
       "      <td>3798</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>4.615436e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>humidity</td>\n",
       "      <td>19</td>\n",
       "      <td>3796</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>5.630835e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>characteristic</td>\n",
       "      <td>20</td>\n",
       "      <td>3795</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>6.173542e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>been</td>\n",
       "      <td>20</td>\n",
       "      <td>3795</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>6.173542e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>contents</td>\n",
       "      <td>23</td>\n",
       "      <td>3792</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>7.938758e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>substantially</td>\n",
       "      <td>24</td>\n",
       "      <td>3791</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>8.572055e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>bright</td>\n",
       "      <td>27</td>\n",
       "      <td>3788</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.060329e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>alongside</td>\n",
       "      <td>1</td>\n",
       "      <td>3814</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1.092036e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>yielded</td>\n",
       "      <td>1</td>\n",
       "      <td>3814</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1.092036e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>fluoroacetate</td>\n",
       "      <td>1</td>\n",
       "      <td>3814</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1.092036e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             token  genes_with  genes_without  group_genes_with  group_genes_without       p_value\n",
       "0            layer          65           3750                 9                   12  5.499473e-11\n",
       "1            exine          19           3796                 6                   15  2.045050e-09\n",
       "2          suberin           8           3807                 3                   18  2.269517e-05\n",
       "3             lacs           2           3813                 2                   19  1.701698e-04\n",
       "4            cutin           4           3811                 2                   19  4.226197e-04\n",
       "5              wax          32           3783                 3                   18  8.271913e-04\n",
       "6     permeability           7           3808                 2                   19  1.004279e-03\n",
       "7       flavonoids           8           3807                 2                   19  1.251207e-03\n",
       "8        flavonoid           9           3806                 2                   19  1.524210e-03\n",
       "9           pollen         251           3564                 6                   15  1.964829e-03\n",
       "10          glossy          11           3804                 2                   19  2.147377e-03\n",
       "11     composition          46           3769                 3                   18  2.216439e-03\n",
       "12              no         981           2834                12                    9  2.269727e-03\n",
       "13         cuticle          12           3803                 2                   19  2.497017e-03\n",
       "14         visible         611           3204                 9                   12  3.333864e-03\n",
       "15     synergistic          15           3800                 2                   19  3.695047e-03\n",
       "16      precursors          16           3799                 2                   19  4.143240e-03\n",
       "17     drastically          16           3799                 2                   19  4.143240e-03\n",
       "18        abnormal         511           3304                 8                   13  4.291804e-03\n",
       "19            coat          60           3755                 3                   18  4.547485e-03\n",
       "20    epicuticular          17           3798                 2                   19  4.615436e-03\n",
       "21        humidity          19           3796                 2                   19  5.630835e-03\n",
       "22  characteristic          20           3795                 2                   19  6.173542e-03\n",
       "23            been          20           3795                 2                   19  6.173542e-03\n",
       "24        contents          23           3792                 2                   19  7.938758e-03\n",
       "25   substantially          24           3791                 2                   19  8.572055e-03\n",
       "26          bright          27           3788                 2                   19  1.060329e-02\n",
       "27       alongside           1           3814                 1                   20  1.092036e-02\n",
       "28         yielded           1           3814                 1                   20  1.092036e-02\n",
       "29   fluoroacetate           1           3814                 1                   20  1.092036e-02"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_enrichment(texts, gene_ids_in_this_pathway).head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
