{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at the Dataset\n",
    "The purpose of this notebook is to look closer at the dataset of genes, natural language descriptions, and ontology term annotations that are used in this work. As included in the preprocessing notebooks, these data are drawn from files from either publications supplements like Oellrich, Walls et al. (2015) or model species databases such as TAIR, MaizeGDB, and SGN. The datasets are already loaded and merged using classes available through the oats package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warming up PyWSD (takes ~10 secs)... took 5.291064262390137 secs.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.probability import FreqDist\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, flatten, to_hms\n",
    "from oats.utils.utils import function_wrapper_with_duration, remove_duplicates_retain_order\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.annotation.annotation import annotate_using_noble_coder, term_enrichment\n",
    "from oats.distances import pairwise as pw\n",
    "from oats.nlp.vocabulary import get_overrepresented_tokens, get_vocab_from_tokens\n",
    "from oats.nlp.vocabulary import reduce_vocab_connected_components, reduce_vocab_linares_pontes, token_enrichment\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paths to the files that are used for this notebook.\n",
    "plant_dataset_path = \"../../plant-data/genes_texts_annots.csv\"\n",
    "\n",
    "# Paths to files with mappings to groups.\n",
    "kegg_pathways_path = \"../../plant-data/reshaped_data/kegg_pathways.csv\" \n",
    "plantcyc_pathways_path = \"../../plant-data/reshaped_data/plantcyc_pathways.csv\" \n",
    "lloyd_meinke_subsets_path = \"../../plant-data/reshaped_data/lloyd_meinke_subsets.csv\" \n",
    "lloyd_meinke_classes_path = \"../../plant-data/reshaped_data/lloyd_meinke_classes.csv\" \n",
    "\n",
    "# Paths to files that map group identifers to longer group names.\n",
    "kegg_pathways_names_path = \"../../plant-data/reshaped_data/kegg_pathways_name_map.csv\"\n",
    "plantcyc_pathways_names_path = \"../../plant-data/reshaped_data/plantcyc_pathways_name_map.csv\"\n",
    "lloyd_meinke_subsets_names_path = \"../../plant-data/reshaped_data/lloyd_meinke_subsets_name_map.csv\"\n",
    "lloyd_meinke_classes_names_path = \"../../plant-data/reshaped_data/lloyd_meinke_classes_name_map.csv\"\n",
    "\n",
    "# Path to file with plant ortholog mappings.\n",
    "ortholog_file_path = \"../../plant-data/databases/panther/PlantGenomeOrthologs_IRB_Modified.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create and name an output directory according to when the notebooks was run.\n",
    "OUTPUT_NAME = \"composition\"\n",
    "OUTPUT_DIR = os.path.join(\"../outputs\",\"{}_{}_{}\".format(OUTPUT_NAME,datetime.datetime.now().strftime('%m_%d_%Y_h%Hm%Ms%S'),random.randrange(1000,9999)))\n",
    "os.mkdir(OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions\n",
       "0     ath                     5066                 2975\n",
       "1     gmx                       30                   23\n",
       "2     mtr                       37                   36\n",
       "3     osa                       92                   85\n",
       "4     sly                       70                   70\n",
       "5     zma                     1405                  810\n",
       "6   total                     6700                 3999"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading in and describing the dataset of plant genes.\n",
    "plant_dataset = Dataset(plant_dataset_path)\n",
    "plant_dataset.filter_has_description()\n",
    "plant_dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's there for each species?\n",
    "The previously loaded dataset contains all of the genes that across six plant species that have natural language description data for phenotype(s) related to that gene. Each gene can have multiple descriptions annotated to it, which were combined or concatenated when the datasets from multiple sources were merged in creating the pickled datasets. Arabidopsis has the highest number of genes that satisfy this criteria, followed by maize, and then followed by the other four species which have a relatively low number of genes that satisfy this criteria, atleast given the sources used for this work. Note that the number of unique descriptions is lower than the number of genes in call cases, because multiple genes can have the same phenotype description associated with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = plant_dataset\n",
    "\n",
    "wnl = WordNetLemmatizer()\n",
    "lemmatize_doc = lambda d: [wnl.lemmatize(x) for x in simple_preprocess(d)]\n",
    "\n",
    "dists = defaultdict(list)\n",
    "\n",
    "sent_lists = {}\n",
    "token_lists = {}\n",
    "stems_lists = {}\n",
    "lemma_lists = {}\n",
    "\n",
    "\n",
    "# For each individual species.\n",
    "for species in data.get_species():\n",
    "    df = data.to_pandas()\n",
    "    subset = df[df[\"species\"]==species]\n",
    "    sentences = [sent_tokenize(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_stemmed = [preprocess_string(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"descriptions\"].values]\n",
    "    sent_lists[species] = flatten(sentences)\n",
    "    token_lists[species] = flatten(descriptions_not_stemmed)\n",
    "    stems_lists[species] = flatten(descriptions_stemmed)    \n",
    "    lemma_lists[species] = flatten(descriptions_lemmatized)\n",
    "    \n",
    "    # What about the distributions of words per gene and sentences per gene?\n",
    "    dists[\"species\"].extend([species]*subset.shape[0])\n",
    "    dists[\"num_words\"].extend([len(word_tokenize(x)) for x in subset[\"descriptions\"].values])\n",
    "    dists[\"num_sents\"].extend([len(sent_tokenize(x)) for x in subset[\"descriptions\"].values])\n",
    "    \n",
    "# For the entire dataset including all of the species.\n",
    "df = data.to_pandas()\n",
    "subset = df\n",
    "sentences = [sent_tokenize(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_stemmed = [preprocess_string(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"descriptions\"].values]\n",
    "sent_lists[\"total\"] = flatten(sentences)\n",
    "token_lists[\"total\"] = flatten(descriptions_not_stemmed)\n",
    "stems_lists[\"total\"] = flatten(descriptions_stemmed)    \n",
    "lemma_lists[\"total\"] = flatten(descriptions_lemmatized)\n",
    "\n",
    "# What about lemmas that are uniquely used for a particular species?\n",
    "lemma_sets_unique_to_species = {}\n",
    "for species in data.get_species():\n",
    "    other_species = [s for s in data.get_species() if s != species]\n",
    "    lemmas_used_in_other_species = set(flatten([lemma_lists[s] for s in other_species]))\n",
    "    unique_lemmas = set(lemma_lists[species]).difference(lemmas_used_in_other_species)\n",
    "    lemma_sets_unique_to_species[species] = unique_lemmas\n",
    "lemma_sets_unique_to_species[\"total\"] = flatten([list(s) for s in lemma_sets_unique_to_species.values()])\n",
    "\n",
    "    \n",
    "# Create a dataframe to contain the summarizing information about this dataset, and sort it by number of genes.\n",
    "# Unique gene identifiers is just the total number of genes, this column name should be changed in the class...\n",
    "df = data.describe() \n",
    "condition = (df.species==\"total\")\n",
    "excluded = df[condition]\n",
    "included = df[~condition]\n",
    "df_sorted = included.sort_values(by=\"unique_gene_identifiers\", ascending=False)\n",
    "df = pd.concat([df_sorted,excluded])\n",
    "\n",
    "# Add columns summarizing information about the text descriptions in the dataset.\n",
    "df[\"total_sents\"] = df[\"species\"].map(lambda x: len(sent_lists[x]))\n",
    "df[\"total_words\"] = df[\"species\"].map(lambda x: len(token_lists[x]))\n",
    "df[\"unique_words\"] = df[\"species\"].map(lambda x: len(set(token_lists[x])))\n",
    "df[\"unique_stems\"] = df[\"species\"].map(lambda x: len(set(stems_lists[x])))\n",
    "df[\"total_lemmas\"] = df[\"species\"].map(lambda x: len(lemma_lists[x]))\n",
    "df[\"unique_lemmas\"] = df[\"species\"].map(lambda x: len(set(lemma_lists[x])))\n",
    "df[\"unique_lemmas_to_species\"] = df[\"species\"].map(lambda x: len(lemma_sets_unique_to_species[x]))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_words</th>\n",
       "      <th>num_sents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zma</td>\n",
       "      <td>115</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zma</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zma</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zma</td>\n",
       "      <td>33</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>105</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>zma</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>zma</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>zma</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>zma</td>\n",
       "      <td>115</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>zma</td>\n",
       "      <td>90</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>zma</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>zma</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>zma</td>\n",
       "      <td>86</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>zma</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>zma</td>\n",
       "      <td>39</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species  num_words  num_sents\n",
       "0      zma         14          1\n",
       "1      zma        115          7\n",
       "2      zma          6          2\n",
       "3      zma         13          3\n",
       "4      zma         33          3\n",
       "5      zma        105          9\n",
       "6      zma        100          7\n",
       "7      zma         43          3\n",
       "8      zma         43          3\n",
       "9      zma        115          7\n",
       "10     zma         90          7\n",
       "11     zma         13          3\n",
       "12     zma         14          1\n",
       "13     zma         13          3\n",
       "14     zma         86         10\n",
       "15     zma         25          5\n",
       "16     zma         14          1\n",
       "17     zma         14          1\n",
       "18     zma         14          1\n",
       "19     zma         39          6"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_distributions = pd.DataFrame(dists)\n",
    "text_distributions.to_csv(os.path.join(OUTPUT_DIR, \"word_sent_distributions.csv\"), index=False)\n",
    "text_distributions.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the ontology term annotations for each species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many of the genes in this dataset for each species are mapped to atleast one term from a given ontology?\n",
    "num_mapped_go = {}\n",
    "num_mapped_po = {}\n",
    "for species in data.get_species():\n",
    "    d = data.to_pandas()\n",
    "    subset = d[d[\"species\"]==species]    \n",
    "    num_mapped_po[species] = len([t for t in subset[\"annotations\"].values if \"PO\" in t])\n",
    "    num_mapped_go[species] = len([t for t in subset[\"annotations\"].values if \"GO\" in t])\n",
    "num_mapped_go[\"total\"] = sum(list(num_mapped_go.values()))   \n",
    "num_mapped_po[\"total\"] = sum(list(num_mapped_po.values()))\n",
    "df[\"go\"] = df[\"species\"].map(lambda x: num_mapped_go[x])\n",
    "df[\"po\"] = df[\"species\"].map(lambda x: num_mapped_po[x])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the biologically relevant groups like biochemical pathways and phenotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "      <td>1084</td>\n",
       "      <td>654</td>\n",
       "      <td>1570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>157</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "      <td>1263</td>\n",
       "      <td>796</td>\n",
       "      <td>1570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220  1084       654          1570\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111   157       133             0\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92     1         4             0\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65    18         3             0\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32     0         2             0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27     3         0             0\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547  1263       796          1570"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the groupings that we're interested in mapping to? Uses the paths defined at the top of the notebook.\n",
    "groupings_dict = {\n",
    "    \"kegg\":(kegg_pathways_path, kegg_pathways_names_path),\n",
    "    \"plantcyc\":(plantcyc_pathways_path, plantcyc_pathways_names_path),\n",
    "    \"lloyd_meinke\":(lloyd_meinke_subsets_path, lloyd_meinke_subsets_names_path)\n",
    "}\n",
    "\n",
    "\n",
    "for name,(filename,mapfile) in groupings_dict.items():\n",
    "    groups = Groupings(filename, {row.group_id:row.group_name for row in pd.read_csv(mapfile).itertuples()})\n",
    "    id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "    group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "    species_dict = data.get_species_dictionary()\n",
    "    num_mapped = {}\n",
    "    for species in data.get_species():\n",
    "        num_mapped[species] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "    num_mapped[\"total\"] = sum(list(num_mapped.values()))    \n",
    "    df[name] = df[\"species\"].map(lambda x: num_mapped[x])  \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the other biologically relevant information like orthologous genes and protein interactions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "      <th>panther</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "      <td>1084</td>\n",
       "      <td>654</td>\n",
       "      <td>1570</td>\n",
       "      <td>317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>157</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "      <td>1263</td>\n",
       "      <td>796</td>\n",
       "      <td>1570</td>\n",
       "      <td>858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke  panther\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220  1084       654          1570      317\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111   157       133             0      443\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92     1         4             0       86\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65    18         3             0       11\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32     0         2             0        0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27     3         0             0        1\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547  1263       796          1570      858"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PantherDB for plant orthologs.\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in ortholog_edgelist.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"panther\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "      <th>panther</th>\n",
       "      <th>stringdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "      <td>1084</td>\n",
       "      <td>654</td>\n",
       "      <td>1570</td>\n",
       "      <td>317</td>\n",
       "      <td>3173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>157</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "      <td>1263</td>\n",
       "      <td>796</td>\n",
       "      <td>1570</td>\n",
       "      <td>858</td>\n",
       "      <td>3476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke  panther  stringdb\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220  1084       654          1570      317      3173\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111   157       133             0      443       229\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92     1         4             0       86        45\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65    18         3             0       11        11\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32     0         2             0        0        13\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27     3         0             0        1         5\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547  1263       796          1570      858      3476"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# STRING DB for protein-protein interactions.\n",
    "naming_file = \"../../plant-data/databases/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../../plant-data/databases/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis\n",
    "    \"../../plant-data/databases/string/4577.protein.links.detailed.v11.0.txt\", # Maize\n",
    "    \"../../plant-data/databases/string/4530.protein.links.detailed.v11.0.txt\", # Tomato \n",
    "    \"../../plant-data/databases/string/4081.protein.links.detailed.v11.0.txt\", # Medicago\n",
    "    \"../../plant-data/databases/string/3880.protein.links.detailed.v11.0.txt\", # Rice \n",
    "    \"../../plant-data/databases/string/3847.protein.links.detailed.v11.0.txt\", # Soybean\n",
    "    \"../../plant-data/databases/string/9606.protein.links.detailed.v11.0.txt\", # Human\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in string_data.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"stringdb\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write that dataframe with all the information about datast to a file.\n",
    "df.to_csv(os.path.join(OUTPUT_DIR,\"full_dataset_composition.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the vocabularies used for different species compare?\n",
    "One of the things we are interested in is discovering or recovering phenotype similarity between different species in order to identify phenologs (phenotypes between species that share some underlying genetic cause). For this reason, we are interested in how the vocabularies used to describe phenotypes between different species vary, because this will impact how feasible it is to use a dataset like this to identify phenologs. Because the Arabidopsis and maize datasets are the largest in this case, we will compare the vocabularies used in describing the phenotypes associated with the genes from these species in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>effectively</td>\n",
       "      <td>RB</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001721</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fill</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>4</td>\n",
       "      <td>0.008166</td>\n",
       "      <td>-0.007736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ever</td>\n",
       "      <td>RB</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>live</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>frequent</td>\n",
       "      <td>NN</td>\n",
       "      <td>5</td>\n",
       "      <td>0.002151</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pc</td>\n",
       "      <td>NN</td>\n",
       "      <td>7</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>atrz</td>\n",
       "      <td>NN</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001721</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>reduction</td>\n",
       "      <td>NN</td>\n",
       "      <td>418</td>\n",
       "      <td>0.179827</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014291</td>\n",
       "      <td>0.165537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pointing</td>\n",
       "      <td>VB</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001721</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cellularizes</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000430</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "0   effectively             RB         4  0.001721         0  0.000000  0.001721\n",
       "1          fill             NN         1  0.000430         4  0.008166 -0.007736\n",
       "2          ever             RB         1  0.000430         0  0.000000  0.000430\n",
       "3          live             JJ         1  0.000430         0  0.000000  0.000430\n",
       "4      frequent             NN         5  0.002151         0  0.000000  0.002151\n",
       "5            pc             NN         7  0.003011         0  0.000000  0.003011\n",
       "6          atrz             NN         4  0.001721         0  0.000000  0.001721\n",
       "7     reduction             NN       418  0.179827         7  0.014291  0.165537\n",
       "8      pointing             VB         4  0.001721         0  0.000000  0.001721\n",
       "9  cellularizes             NN         1  0.000430         0  0.000000  0.000430"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using lemmas as the vocabulary components.\n",
    "vocabs = {s:set(lemma_list) for s,lemma_list in lemma_lists.items()}\n",
    "fdist_zma = FreqDist(lemma_lists[\"zma\"])\n",
    "fdist_ath = FreqDist(lemma_lists[\"ath\"])\n",
    "\n",
    "# Using word stems as the vocabulary components.\n",
    "#vocabs = {s:set(stems_list) for s,stems_list in stems_lists.items()}\n",
    "#fdist_zma = FreqDist(stems_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(stems_lists[\"ath\"])\n",
    "\n",
    "# Using tokens (full words) as the vocabulary components.\n",
    "#vocabs = {s:set(token_list) for s,token_list in token_lists.items()}\n",
    "#fdist_zma = FreqDist(token_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(token_lists[\"ath\"])\n",
    "\n",
    "union_vocab = vocabs[\"zma\"].union(vocabs[\"ath\"])\n",
    "table = pd.DataFrame({\"token\":list(union_vocab)})\n",
    "stops = set(stopwords.words('english'))\n",
    "table = table[~table.token.isin(stops)]\n",
    "table[\"part_of_speech\"] = table[\"token\"].map(lambda x: nltk.pos_tag([x])[0][1][:2])\n",
    "table[\"ath_freq\"] = table[\"token\"].map(lambda x: fdist_ath[x])\n",
    "table[\"ath_rate\"] = table[\"ath_freq\"]*100/len(token_lists[\"ath\"])\n",
    "table[\"zma_freq\"] = table[\"token\"].map(lambda x: fdist_zma[x])\n",
    "table[\"zma_rate\"] = table[\"zma_freq\"]*100/len(token_lists[\"zma\"])\n",
    "table[\"diff\"] = table[\"ath_rate\"]-table[\"zma_rate\"]\n",
    "table.to_csv(os.path.join(OUTPUT_DIR,\"token_frequencies.csv\"), index=False)\n",
    "table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2872</th>\n",
       "      <td>embryo</td>\n",
       "      <td>NN</td>\n",
       "      <td>4087</td>\n",
       "      <td>1.758265</td>\n",
       "      <td>150</td>\n",
       "      <td>0.306229</td>\n",
       "      <td>1.452037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4861</th>\n",
       "      <td>mutant</td>\n",
       "      <td>NN</td>\n",
       "      <td>4433</td>\n",
       "      <td>1.907118</td>\n",
       "      <td>254</td>\n",
       "      <td>0.518547</td>\n",
       "      <td>1.388571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4828</th>\n",
       "      <td>phenotype</td>\n",
       "      <td>NN</td>\n",
       "      <td>3393</td>\n",
       "      <td>1.459700</td>\n",
       "      <td>76</td>\n",
       "      <td>0.155156</td>\n",
       "      <td>1.304544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5112</th>\n",
       "      <td>type</td>\n",
       "      <td>NN</td>\n",
       "      <td>2364</td>\n",
       "      <td>1.017015</td>\n",
       "      <td>15</td>\n",
       "      <td>0.030623</td>\n",
       "      <td>0.986392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5501</th>\n",
       "      <td>wild</td>\n",
       "      <td>NN</td>\n",
       "      <td>2287</td>\n",
       "      <td>0.983889</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014291</td>\n",
       "      <td>0.969598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2930</th>\n",
       "      <td>root</td>\n",
       "      <td>NN</td>\n",
       "      <td>2487</td>\n",
       "      <td>1.069931</td>\n",
       "      <td>70</td>\n",
       "      <td>0.142907</td>\n",
       "      <td>0.927024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5078</th>\n",
       "      <td>stage</td>\n",
       "      <td>NN</td>\n",
       "      <td>1748</td>\n",
       "      <td>0.752006</td>\n",
       "      <td>55</td>\n",
       "      <td>0.112284</td>\n",
       "      <td>0.639722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6960</th>\n",
       "      <td>reduced</td>\n",
       "      <td>VB</td>\n",
       "      <td>2416</td>\n",
       "      <td>1.039386</td>\n",
       "      <td>215</td>\n",
       "      <td>0.438928</td>\n",
       "      <td>0.600458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4843</th>\n",
       "      <td>terminal</td>\n",
       "      <td>NN</td>\n",
       "      <td>1265</td>\n",
       "      <td>0.544215</td>\n",
       "      <td>4</td>\n",
       "      <td>0.008166</td>\n",
       "      <td>0.536049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4040</th>\n",
       "      <td>growth</td>\n",
       "      <td>NN</td>\n",
       "      <td>1590</td>\n",
       "      <td>0.684033</td>\n",
       "      <td>78</td>\n",
       "      <td>0.159239</td>\n",
       "      <td>0.524794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5182</th>\n",
       "      <td>cotyledon</td>\n",
       "      <td>NN</td>\n",
       "      <td>1148</td>\n",
       "      <td>0.493880</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.493880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1652</th>\n",
       "      <td>seed</td>\n",
       "      <td>NN</td>\n",
       "      <td>1593</td>\n",
       "      <td>0.685323</td>\n",
       "      <td>113</td>\n",
       "      <td>0.230692</td>\n",
       "      <td>0.454631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>level</td>\n",
       "      <td>NN</td>\n",
       "      <td>1132</td>\n",
       "      <td>0.486997</td>\n",
       "      <td>20</td>\n",
       "      <td>0.040830</td>\n",
       "      <td>0.446166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3891</th>\n",
       "      <td>increased</td>\n",
       "      <td>VB</td>\n",
       "      <td>1070</td>\n",
       "      <td>0.460324</td>\n",
       "      <td>11</td>\n",
       "      <td>0.022457</td>\n",
       "      <td>0.437867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915</th>\n",
       "      <td>cell</td>\n",
       "      <td>NN</td>\n",
       "      <td>1516</td>\n",
       "      <td>0.652197</td>\n",
       "      <td>106</td>\n",
       "      <td>0.216402</td>\n",
       "      <td>0.435796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5409</th>\n",
       "      <td>defective</td>\n",
       "      <td>JJ</td>\n",
       "      <td>2337</td>\n",
       "      <td>1.005399</td>\n",
       "      <td>285</td>\n",
       "      <td>0.581835</td>\n",
       "      <td>0.423565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>double</td>\n",
       "      <td>RB</td>\n",
       "      <td>881</td>\n",
       "      <td>0.379014</td>\n",
       "      <td>3</td>\n",
       "      <td>0.006125</td>\n",
       "      <td>0.372890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5497</th>\n",
       "      <td>wa</td>\n",
       "      <td>NN</td>\n",
       "      <td>841</td>\n",
       "      <td>0.361806</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002042</td>\n",
       "      <td>0.359764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>compared</td>\n",
       "      <td>VB</td>\n",
       "      <td>840</td>\n",
       "      <td>0.361376</td>\n",
       "      <td>9</td>\n",
       "      <td>0.018374</td>\n",
       "      <td>0.343002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3808</th>\n",
       "      <td>rosette</td>\n",
       "      <td>NN</td>\n",
       "      <td>796</td>\n",
       "      <td>0.342447</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002042</td>\n",
       "      <td>0.340405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3805</th>\n",
       "      <td>flower</td>\n",
       "      <td>NN</td>\n",
       "      <td>799</td>\n",
       "      <td>0.343737</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014291</td>\n",
       "      <td>0.329447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>silique</td>\n",
       "      <td>NN</td>\n",
       "      <td>723</td>\n",
       "      <td>0.311041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.311041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2315</th>\n",
       "      <td>visible</td>\n",
       "      <td>JJ</td>\n",
       "      <td>767</td>\n",
       "      <td>0.329971</td>\n",
       "      <td>11</td>\n",
       "      <td>0.022457</td>\n",
       "      <td>0.307514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3993</th>\n",
       "      <td>abnormal</td>\n",
       "      <td>JJ</td>\n",
       "      <td>805</td>\n",
       "      <td>0.346318</td>\n",
       "      <td>26</td>\n",
       "      <td>0.053080</td>\n",
       "      <td>0.293239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1423</th>\n",
       "      <td>development</td>\n",
       "      <td>NN</td>\n",
       "      <td>765</td>\n",
       "      <td>0.329110</td>\n",
       "      <td>19</td>\n",
       "      <td>0.038789</td>\n",
       "      <td>0.290321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6366</th>\n",
       "      <td>globular</td>\n",
       "      <td>NN</td>\n",
       "      <td>672</td>\n",
       "      <td>0.289101</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.289101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>sensitive</td>\n",
       "      <td>JJ</td>\n",
       "      <td>689</td>\n",
       "      <td>0.296414</td>\n",
       "      <td>6</td>\n",
       "      <td>0.012249</td>\n",
       "      <td>0.284165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4431</th>\n",
       "      <td>defect</td>\n",
       "      <td>NN</td>\n",
       "      <td>722</td>\n",
       "      <td>0.310611</td>\n",
       "      <td>20</td>\n",
       "      <td>0.040830</td>\n",
       "      <td>0.269781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4032</th>\n",
       "      <td>hypocotyl</td>\n",
       "      <td>NN</td>\n",
       "      <td>607</td>\n",
       "      <td>0.261137</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.261137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>short</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1173</td>\n",
       "      <td>0.504636</td>\n",
       "      <td>120</td>\n",
       "      <td>0.244983</td>\n",
       "      <td>0.259653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "2872       embryo             NN      4087  1.758265       150  0.306229  1.452037\n",
       "4861       mutant             NN      4433  1.907118       254  0.518547  1.388571\n",
       "4828    phenotype             NN      3393  1.459700        76  0.155156  1.304544\n",
       "5112         type             NN      2364  1.017015        15  0.030623  0.986392\n",
       "5501         wild             NN      2287  0.983889         7  0.014291  0.969598\n",
       "2930         root             NN      2487  1.069931        70  0.142907  0.927024\n",
       "5078        stage             NN      1748  0.752006        55  0.112284  0.639722\n",
       "6960      reduced             VB      2416  1.039386       215  0.438928  0.600458\n",
       "4843     terminal             NN      1265  0.544215         4  0.008166  0.536049\n",
       "4040       growth             NN      1590  0.684033        78  0.159239  0.524794\n",
       "5182    cotyledon             NN      1148  0.493880         0  0.000000  0.493880\n",
       "1652         seed             NN      1593  0.685323       113  0.230692  0.454631\n",
       "227         level             NN      1132  0.486997        20  0.040830  0.446166\n",
       "3891    increased             VB      1070  0.460324        11  0.022457  0.437867\n",
       "1915         cell             NN      1516  0.652197       106  0.216402  0.435796\n",
       "5409    defective             JJ      2337  1.005399       285  0.581835  0.423565\n",
       "580        double             RB       881  0.379014         3  0.006125  0.372890\n",
       "5497           wa             NN       841  0.361806         1  0.002042  0.359764\n",
       "427      compared             VB       840  0.361376         9  0.018374  0.343002\n",
       "3808      rosette             NN       796  0.342447         1  0.002042  0.340405\n",
       "3805       flower             NN       799  0.343737         7  0.014291  0.329447\n",
       "2141      silique             NN       723  0.311041         0  0.000000  0.311041\n",
       "2315      visible             JJ       767  0.329971        11  0.022457  0.307514\n",
       "3993     abnormal             JJ       805  0.346318        26  0.053080  0.293239\n",
       "1423  development             NN       765  0.329110        19  0.038789  0.290321\n",
       "6366     globular             NN       672  0.289101         0  0.000000  0.289101\n",
       "925     sensitive             JJ       689  0.296414         6  0.012249  0.284165\n",
       "4431       defect             NN       722  0.310611        20  0.040830  0.269781\n",
       "4032    hypocotyl             NN       607  0.261137         0  0.000000  0.261137\n",
       "155         short             JJ      1173  0.504636       120  0.244983  0.259653"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for Arabidopsis than maize descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=False, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3093</th>\n",
       "      <td>endosperm</td>\n",
       "      <td>NN</td>\n",
       "      <td>100</td>\n",
       "      <td>0.043021</td>\n",
       "      <td>1069</td>\n",
       "      <td>2.182390</td>\n",
       "      <td>-2.139369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>seedling</td>\n",
       "      <td>VB</td>\n",
       "      <td>1276</td>\n",
       "      <td>0.548947</td>\n",
       "      <td>1316</td>\n",
       "      <td>2.686646</td>\n",
       "      <td>-2.137699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1188</th>\n",
       "      <td>kernel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>756</td>\n",
       "      <td>1.543393</td>\n",
       "      <td>-1.543393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>leaf</td>\n",
       "      <td>NN</td>\n",
       "      <td>3224</td>\n",
       "      <td>1.386995</td>\n",
       "      <td>1401</td>\n",
       "      <td>2.860176</td>\n",
       "      <td>-1.473181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>yellow</td>\n",
       "      <td>NN</td>\n",
       "      <td>239</td>\n",
       "      <td>0.102820</td>\n",
       "      <td>766</td>\n",
       "      <td>1.563808</td>\n",
       "      <td>-1.460988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5398</th>\n",
       "      <td>green</td>\n",
       "      <td>JJ</td>\n",
       "      <td>743</td>\n",
       "      <td>0.319646</td>\n",
       "      <td>778</td>\n",
       "      <td>1.588306</td>\n",
       "      <td>-1.268661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1213</th>\n",
       "      <td>white</td>\n",
       "      <td>JJ</td>\n",
       "      <td>310</td>\n",
       "      <td>0.133365</td>\n",
       "      <td>642</td>\n",
       "      <td>1.310659</td>\n",
       "      <td>-1.177294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5266</th>\n",
       "      <td>albino</td>\n",
       "      <td>NN</td>\n",
       "      <td>182</td>\n",
       "      <td>0.078298</td>\n",
       "      <td>396</td>\n",
       "      <td>0.808444</td>\n",
       "      <td>-0.730146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1765</th>\n",
       "      <td>usually</td>\n",
       "      <td>RB</td>\n",
       "      <td>41</td>\n",
       "      <td>0.017639</td>\n",
       "      <td>342</td>\n",
       "      <td>0.698201</td>\n",
       "      <td>-0.680563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4734</th>\n",
       "      <td>color</td>\n",
       "      <td>NN</td>\n",
       "      <td>40</td>\n",
       "      <td>0.017208</td>\n",
       "      <td>326</td>\n",
       "      <td>0.665537</td>\n",
       "      <td>-0.648329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650</th>\n",
       "      <td>tassel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>300</td>\n",
       "      <td>0.612457</td>\n",
       "      <td>-0.612457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4365</th>\n",
       "      <td>ear</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>272</td>\n",
       "      <td>0.555295</td>\n",
       "      <td>-0.555295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>often</td>\n",
       "      <td>RB</td>\n",
       "      <td>136</td>\n",
       "      <td>0.058508</td>\n",
       "      <td>275</td>\n",
       "      <td>0.561419</td>\n",
       "      <td>-0.502911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5781</th>\n",
       "      <td>aleurone</td>\n",
       "      <td>NN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001291</td>\n",
       "      <td>246</td>\n",
       "      <td>0.502215</td>\n",
       "      <td>-0.500924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5020</th>\n",
       "      <td>opaque</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>244</td>\n",
       "      <td>0.498132</td>\n",
       "      <td>-0.497702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5341</th>\n",
       "      <td>class</td>\n",
       "      <td>NN</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003442</td>\n",
       "      <td>240</td>\n",
       "      <td>0.489966</td>\n",
       "      <td>-0.486524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4969</th>\n",
       "      <td>tissue</td>\n",
       "      <td>NN</td>\n",
       "      <td>213</td>\n",
       "      <td>0.091635</td>\n",
       "      <td>253</td>\n",
       "      <td>0.516506</td>\n",
       "      <td>-0.424871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6824</th>\n",
       "      <td>small</td>\n",
       "      <td>JJ</td>\n",
       "      <td>557</td>\n",
       "      <td>0.239627</td>\n",
       "      <td>321</td>\n",
       "      <td>0.655329</td>\n",
       "      <td>-0.415703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4297</th>\n",
       "      <td>pale</td>\n",
       "      <td>NN</td>\n",
       "      <td>524</td>\n",
       "      <td>0.225430</td>\n",
       "      <td>306</td>\n",
       "      <td>0.624707</td>\n",
       "      <td>-0.399277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6973</th>\n",
       "      <td>vp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190</td>\n",
       "      <td>0.387890</td>\n",
       "      <td>-0.387890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1419</th>\n",
       "      <td>palegreen</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>171</td>\n",
       "      <td>0.349101</td>\n",
       "      <td>-0.348670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6134</th>\n",
       "      <td>general</td>\n",
       "      <td>JJ</td>\n",
       "      <td>13</td>\n",
       "      <td>0.005593</td>\n",
       "      <td>171</td>\n",
       "      <td>0.349101</td>\n",
       "      <td>-0.343508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5114</th>\n",
       "      <td>term</td>\n",
       "      <td>NN</td>\n",
       "      <td>21</td>\n",
       "      <td>0.009034</td>\n",
       "      <td>172</td>\n",
       "      <td>0.351142</td>\n",
       "      <td>-0.342108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2006</th>\n",
       "      <td>texture</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>163</td>\n",
       "      <td>0.332769</td>\n",
       "      <td>-0.331908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5577</th>\n",
       "      <td>appearance</td>\n",
       "      <td>NN</td>\n",
       "      <td>121</td>\n",
       "      <td>0.052055</td>\n",
       "      <td>184</td>\n",
       "      <td>0.375641</td>\n",
       "      <td>-0.323585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6574</th>\n",
       "      <td>surface</td>\n",
       "      <td>NN</td>\n",
       "      <td>131</td>\n",
       "      <td>0.056357</td>\n",
       "      <td>185</td>\n",
       "      <td>0.377682</td>\n",
       "      <td>-0.321325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5381</th>\n",
       "      <td>otherwise</td>\n",
       "      <td>RB</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003872</td>\n",
       "      <td>157</td>\n",
       "      <td>0.320519</td>\n",
       "      <td>-0.316647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6054</th>\n",
       "      <td>pericarp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>139</td>\n",
       "      <td>0.283772</td>\n",
       "      <td>-0.283772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>sometimes</td>\n",
       "      <td>RB</td>\n",
       "      <td>37</td>\n",
       "      <td>0.015918</td>\n",
       "      <td>145</td>\n",
       "      <td>0.296021</td>\n",
       "      <td>-0.280103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1595</th>\n",
       "      <td>male</td>\n",
       "      <td>NN</td>\n",
       "      <td>354</td>\n",
       "      <td>0.152294</td>\n",
       "      <td>210</td>\n",
       "      <td>0.428720</td>\n",
       "      <td>-0.276426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "3093   endosperm             NN       100  0.043021      1069  2.182390 -2.139369\n",
       "188     seedling             VB      1276  0.548947      1316  2.686646 -2.137699\n",
       "1188      kernel             NN         0  0.000000       756  1.543393 -1.543393\n",
       "265         leaf             NN      3224  1.386995      1401  2.860176 -1.473181\n",
       "211       yellow             NN       239  0.102820       766  1.563808 -1.460988\n",
       "5398       green             JJ       743  0.319646       778  1.588306 -1.268661\n",
       "1213       white             JJ       310  0.133365       642  1.310659 -1.177294\n",
       "5266      albino             NN       182  0.078298       396  0.808444 -0.730146\n",
       "1765     usually             RB        41  0.017639       342  0.698201 -0.680563\n",
       "4734       color             NN        40  0.017208       326  0.665537 -0.648329\n",
       "1650      tassel             NN         0  0.000000       300  0.612457 -0.612457\n",
       "4365         ear             NN         0  0.000000       272  0.555295 -0.555295\n",
       "198        often             RB       136  0.058508       275  0.561419 -0.502911\n",
       "5781    aleurone             NN         3  0.001291       246  0.502215 -0.500924\n",
       "5020      opaque             NN         1  0.000430       244  0.498132 -0.497702\n",
       "5341       class             NN         8  0.003442       240  0.489966 -0.486524\n",
       "4969      tissue             NN       213  0.091635       253  0.516506 -0.424871\n",
       "6824       small             JJ       557  0.239627       321  0.655329 -0.415703\n",
       "4297        pale             NN       524  0.225430       306  0.624707 -0.399277\n",
       "6973          vp             NN         0  0.000000       190  0.387890 -0.387890\n",
       "1419   palegreen             NN         1  0.000430       171  0.349101 -0.348670\n",
       "6134     general             JJ        13  0.005593       171  0.349101 -0.343508\n",
       "5114        term             NN        21  0.009034       172  0.351142 -0.342108\n",
       "2006     texture             NN         2  0.000860       163  0.332769 -0.331908\n",
       "5577  appearance             NN       121  0.052055       184  0.375641 -0.323585\n",
       "6574     surface             NN       131  0.056357       185  0.377682 -0.321325\n",
       "5381   otherwise             RB         9  0.003872       157  0.320519 -0.316647\n",
       "6054    pericarp             NN         0  0.000000       139  0.283772 -0.283772\n",
       "653    sometimes             RB        37  0.015918       145  0.296021 -0.280103\n",
       "1595        male             NN       354  0.152294       210  0.428720 -0.276426"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for maize than Arabidopsis descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=True, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MD</td>\n",
       "      <td>0.059567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD</td>\n",
       "      <td>0.029173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IN</td>\n",
       "      <td>0.026193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JJ</td>\n",
       "      <td>0.023080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DT</td>\n",
       "      <td>0.019441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RB</td>\n",
       "      <td>0.016217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NN</td>\n",
       "      <td>0.014511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VB</td>\n",
       "      <td>0.012129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC</td>\n",
       "      <td>0.007736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WP</td>\n",
       "      <td>0.001611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  part_of_speech  abs_diff\n",
       "0             MD  0.059567\n",
       "1             CD  0.029173\n",
       "2             IN  0.026193\n",
       "3             JJ  0.023080\n",
       "4             DT  0.019441\n",
       "5             RB  0.016217\n",
       "6             NN  0.014511\n",
       "7             VB  0.012129\n",
       "8             CC  0.007736\n",
       "9             WP  0.001611"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is the mean absolute value of the rate differences different between the different parts of speech?\n",
    "table[\"abs_diff\"] = abs(table[\"diff\"])\n",
    "pos_table = table.groupby(\"part_of_speech\").mean()\n",
    "pos_table.sort_values(by=\"abs_diff\", inplace=True, ascending=False)\n",
    "pos_table = pos_table[[\"abs_diff\"]]\n",
    "pos_table.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Working on the Venn Diagram for this part, unused currently.\n",
    "#print(table.shape)\n",
    "#zma_only = table[table[\"ath_rate\"]==0]\n",
    "#ath_only = table[table[\"zma_rate\"]==0]\n",
    "#print(zma_only.shape)\n",
    "#print(ath_only.shape)\n",
    "#print(ath_only.shape[0]+zma_only.shape[0])\n",
    "#ath_only.head(10)\n",
    "# We need to create a mapping between stems and the words that were present for them.\n",
    "# This is because what we want is the stems that are exclusive to a species.\n",
    "# but then the words that are actually there for those stems, so that we can count their parts of speech."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at Term and Word Enrichment for Groups of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3547"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the dataset of phenotype descriptions and ontology annotations.\n",
    "plant_dataset = Dataset(plant_dataset_path)\n",
    "data = plant_dataset\n",
    "data.filter_has_description()\n",
    "#data.filter_has_annotation(\"GO\")\n",
    "data.filter_has_annotation(\"PO\")\n",
    "d = data.get_description_dictionary()\n",
    "texts = {i:\" \".join(simple_preprocess(t)) for i,t in d.items()}\n",
    "len(texts)                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create ontology objects for all the biological ontologies being used.\n",
    "go_pickle_path = \"../ontologies/go.pickle\"                                                                \n",
    "po_pickle_path = \"../ontologies/po.pickle\"                                                             \n",
    "pato_pickle_path = \"../ontologies/pato.pickle\"\n",
    "pato = load_from_pickle(pato_pickle_path)\n",
    "po = load_from_pickle(po_pickle_path)\n",
    "go = load_from_pickle(go_pickle_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "curated_go_annotations = data.get_annotations_dictionary(\"GO\")\n",
    "curated_po_annotations = data.get_annotations_dictionary(\"PO\")\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>freq</th>\n",
       "      <th>term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>665</td>\n",
       "      <td>protein binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>595</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GO:0009507</td>\n",
       "      <td>463</td>\n",
       "      <td>chloroplast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GO:0005886</td>\n",
       "      <td>414</td>\n",
       "      <td>plasma membrane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>349</td>\n",
       "      <td>cytosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>340</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>225</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>213</td>\n",
       "      <td>cytoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>192</td>\n",
       "      <td>regulation of transcription, DNA-templated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>GO:0009506</td>\n",
       "      <td>189</td>\n",
       "      <td>plasmodesma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>182</td>\n",
       "      <td>mitochondrion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>177</td>\n",
       "      <td>membrane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>GO:0010228</td>\n",
       "      <td>173</td>\n",
       "      <td>vegetative to reproductive phase transition of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>GO:0009570</td>\n",
       "      <td>171</td>\n",
       "      <td>chloroplast stroma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>GO:0009651</td>\n",
       "      <td>160</td>\n",
       "      <td>response to salt stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GO:0009941</td>\n",
       "      <td>156</td>\n",
       "      <td>chloroplast envelope</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>GO:0009737</td>\n",
       "      <td>139</td>\n",
       "      <td>response to abscisic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>133</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>GO:0005794</td>\n",
       "      <td>131</td>\n",
       "      <td>Golgi apparatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0009409</td>\n",
       "      <td>124</td>\n",
       "      <td>response to cold</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        term_id  freq                                          term_name\n",
       "9    GO:0005515   665                                    protein binding\n",
       "58   GO:0005634   595                                            nucleus\n",
       "19   GO:0009507   463                                        chloroplast\n",
       "18   GO:0005886   414                                    plasma membrane\n",
       "100  GO:0005829   349                                            cytosol\n",
       "2    GO:0003674   340                                 molecular_function\n",
       "1    GO:0008150   225                                 biological_process\n",
       "130  GO:0005737   213                                          cytoplasm\n",
       "57   GO:0006355   192         regulation of transcription, DNA-templated\n",
       "46   GO:0009506   189                                        plasmodesma\n",
       "175  GO:0005739   182                                      mitochondrion\n",
       "149  GO:0016020   177                                           membrane\n",
       "76   GO:0010228   173  vegetative to reproductive phase transition of...\n",
       "62   GO:0009570   171                                 chloroplast stroma\n",
       "201  GO:0009651   160                            response to salt stress\n",
       "61   GO:0009941   156                               chloroplast envelope\n",
       "77   GO:0009737   139                          response to abscisic acid\n",
       "11   GO:0005783   133                              endoplasmic reticulum\n",
       "144  GO:0005794   131                                    Golgi apparatus\n",
       "3    GO:0009409   124                                   response to cold"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Which GO terms are used to annotate the most genes in this dataset?\n",
    "term_id_to_ids = defaultdict(list)\n",
    "for i,term_id_list in curated_go_annotations.items():\n",
    "    for term_id in term_id_list:\n",
    "        term_id_to_ids[term_id].append(i)\n",
    "term_id_to_num_ids = {k:len(v) for k,v in term_id_to_ids.items()}\n",
    "terms_df = pd.DataFrame(term_id_to_num_ids.items(), columns=[\"term_id\", \"freq\"])\n",
    "\n",
    "def get_term_name(ont,i):\n",
    "    try:\n",
    "        return(ont[i].name)\n",
    "    except:\n",
    "        return(\"\")\n",
    "\n",
    "terms_df[\"term_name\"] = terms_df[\"term_id\"].map(lambda x: get_term_name(go,x))\n",
    "terms_df.sort_values(by=\"freq\", ascending=False, inplace=True)\n",
    "terms_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the group be ones that have that GO term anntation.\n",
    "#go_term_id_of_interest = \"GO:0009640\"\n",
    "#gene_ids_in_this_pathway = [k for k,v in curated_go_annotations.items() if go_term_id_of_interest in v]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>num_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>phenylpropanoid biosynthesis</td>\n",
       "      <td>PWY-361</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>indole-3-acetate biosynthesis II</td>\n",
       "      <td>PWY-581</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>suberin monomers biosynthesis</td>\n",
       "      <td>PWY-1121</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gluconeogenesis III</td>\n",
       "      <td>PWY66-399</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gluconeogenesis I</td>\n",
       "      <td>GLUCONEO-PWY</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>phosphatidylcholine acyl editing</td>\n",
       "      <td>PWY-6803</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sporopollenin precursors biosynthesis</td>\n",
       "      <td>PWY-6733</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>L-leucine biosynthesis</td>\n",
       "      <td>LEUSYN-PWY</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>glycolysis I (from glucose 6-phosphate)</td>\n",
       "      <td>GLYCOLYSIS</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>glycolysis II (from fructose 6-phosphate)</td>\n",
       "      <td>PWY-5484</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>palmitate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5971</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3-phosphoinositide biosynthesis</td>\n",
       "      <td>PWY-6352</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>jasmonic acid biosynthesis</td>\n",
       "      <td>PWY-735</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>very long chain fatty acid biosynthesis I</td>\n",
       "      <td>PWY-5080</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>photosynthesis light reactions</td>\n",
       "      <td>PWY-101</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       pathway_name    pathway_id  num_genes\n",
       "0                      phenylpropanoid biosynthesis       PWY-361         21\n",
       "1                  indole-3-acetate biosynthesis II       PWY-581         20\n",
       "2                     suberin monomers biosynthesis      PWY-1121         20\n",
       "3                               gluconeogenesis III     PWY66-399         19\n",
       "4                                 gluconeogenesis I  GLUCONEO-PWY         19\n",
       "5                  phosphatidylcholine acyl editing      PWY-6803         17\n",
       "6             sporopollenin precursors biosynthesis      PWY-6733         17\n",
       "7                            L-leucine biosynthesis    LEUSYN-PWY         16\n",
       "8           glycolysis I (from glucose 6-phosphate)    GLYCOLYSIS         16\n",
       "9         glycolysis II (from fructose 6-phosphate)      PWY-5484         15\n",
       "10  palmitate biosynthesis II (bacteria and plants)      PWY-5971         15\n",
       "11                  3-phosphoinositide biosynthesis      PWY-6352         15\n",
       "12                       jasmonic acid biosynthesis       PWY-735         15\n",
       "13        very long chain fatty acid biosynthesis I      PWY-5080         14\n",
       "14                   photosynthesis light reactions       PWY-101         14"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the mappings from this dataset to PlantCyc information.\n",
    "#pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "#groups = load_from_pickle(pmn_pathways_filename)\n",
    "#id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "\n",
    "\n",
    "# Reading in the dataset of groupings for pathways in PlantCyc.\n",
    "plantcyc_name_mapping = {row.group_id:row.group_name for row in pd.read_csv(plantcyc_pathways_names_path).itertuples()}\n",
    "plantcyc_grouping = Groupings(plantcyc_pathways_path, plantcyc_name_mapping)\n",
    "id_to_group_ids, group_id_to_ids = plantcyc_grouping.get_groupings_for_dataset(data)\n",
    "\n",
    "# Look at which pathways are best represented in this dataset.\n",
    "pathways_sorted = sorted(group_id_to_ids.items(), key=lambda item: len(item[1]), reverse=True)\n",
    "pathways_sorted_lengths = [(i,len(l)) for (i,l) in pathways_sorted]\n",
    "pathways_df = pd.DataFrame(pathways_sorted_lengths, columns=[\"pathway_id\",\"num_genes\"])\n",
    "pathways_df[\"pathway_name\"] = pathways_df[\"pathway_id\"].map(lambda x: plantcyc_grouping.get_long_name(x))\n",
    "pathways_df = pathways_df[[\"pathway_name\",\"pathway_id\",\"num_genes\"]]\n",
    "pathways_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[718, 860, 871, 955, 1228, 1833, 4387, 4399]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For some example pathway to use.\n",
    "#pathway_id = \"PWY-361\"\n",
    "pathway_id = \"PWY-581\"\n",
    "#pathway_id = \"PWY-1121\"\n",
    "pathway_id = \"PWY-695\"\n",
    "gene_ids_in_this_pathway = group_id_to_ids[pathway_id]\n",
    "gene_ids_in_this_pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "wordcloud = defaultdict(list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PO:0000258</td>\n",
       "      <td>root cortex</td>\n",
       "      <td>26</td>\n",
       "      <td>3521</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001627</td>\n",
       "      <td>0.032540</td>\n",
       "      <td>True</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PO:0007057</td>\n",
       "      <td>seed germination stage</td>\n",
       "      <td>94</td>\n",
       "      <td>3453</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000999</td>\n",
       "      <td>0.019971</td>\n",
       "      <td>True</td>\n",
       "      <td>3.633786</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PO:0020031</td>\n",
       "      <td>radicle</td>\n",
       "      <td>17</td>\n",
       "      <td>3530</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000744</td>\n",
       "      <td>0.014871</td>\n",
       "      <td>True</td>\n",
       "      <td>3.419565</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PO:0006203</td>\n",
       "      <td>pericycle</td>\n",
       "      <td>29</td>\n",
       "      <td>3518</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001995</td>\n",
       "      <td>0.039895</td>\n",
       "      <td>True</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PO:0000045</td>\n",
       "      <td>embryo root</td>\n",
       "      <td>23</td>\n",
       "      <td>3524</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001296</td>\n",
       "      <td>0.025913</td>\n",
       "      <td>True</td>\n",
       "      <td>2.362259</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PO:0005708</td>\n",
       "      <td>cortex</td>\n",
       "      <td>31</td>\n",
       "      <td>3516</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.002260</td>\n",
       "      <td>0.045197</td>\n",
       "      <td>True</td>\n",
       "      <td>2.015353</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      term_id              term_label  genes_with  genes_without  group_genes_with  group_genes_without   p_value  p_value_adj  significant  info_content significance\n",
       "3  PO:0000258             root cortex          26           3521                 2                    6  0.001627     0.032540         True      4.000000            *\n",
       "1  PO:0007057  seed germination stage          94           3453                 3                    5  0.000999     0.019971         True      3.633786            *\n",
       "0  PO:0020031                 radicle          17           3530                 2                    6  0.000744     0.014871         True      3.419565            *\n",
       "4  PO:0006203               pericycle          29           3518                 2                    6  0.001995     0.039895         True      3.000000            *\n",
       "2  PO:0000045             embryo root          23           3524                 2                    6  0.001296     0.025913         True      2.362259            *\n",
       "5  PO:0005708                  cortex          31           3516                 2                    6  0.002260     0.045197         True      2.015353            *"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_po_annotations, gene_ids_in_this_pathway, po).head(20)\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: po.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# ns   P > 0.05\n",
    "# *    P  0.05\n",
    "# **   P  0.01\n",
    "# ***  P  0.001\n",
    "# **** P  0.0001\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in results.itertuples():\n",
    "    wordcloud[\"Weight\"].append(int(1/row.p_value_adj))\n",
    "    wordcloud[\"Word\"].append(\"{} ({})\".format(row.term_id,row.term_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0009684</td>\n",
       "      <td>indoleacetic acid biosynthetic process</td>\n",
       "      <td>9</td>\n",
       "      <td>3538</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>6.605807e-11</td>\n",
       "      <td>1.321161e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>4.489831</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GO:0048825</td>\n",
       "      <td>cotyledon development</td>\n",
       "      <td>23</td>\n",
       "      <td>3524</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>2.928356e-07</td>\n",
       "      <td>5.856712e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0042435</td>\n",
       "      <td>indole-containing compound biosynthetic process</td>\n",
       "      <td>20</td>\n",
       "      <td>3527</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>5.716694e-13</td>\n",
       "      <td>1.143339e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>3.632673</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GO:0072330</td>\n",
       "      <td>monocarboxylic acid biosynthetic process</td>\n",
       "      <td>66</td>\n",
       "      <td>3481</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>2.411502e-09</td>\n",
       "      <td>4.823003e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>3.313892</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0009851</td>\n",
       "      <td>auxin biosynthetic process</td>\n",
       "      <td>17</td>\n",
       "      <td>3530</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>7.257074e-22</td>\n",
       "      <td>1.451415e-20</td>\n",
       "      <td>True</td>\n",
       "      <td>3.277093</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0009683</td>\n",
       "      <td>indoleacetic acid metabolic process</td>\n",
       "      <td>10</td>\n",
       "      <td>3537</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>1.053362e-10</td>\n",
       "      <td>2.106724e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>3.144588</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:0032787</td>\n",
       "      <td>monocarboxylic acid metabolic process</td>\n",
       "      <td>119</td>\n",
       "      <td>3428</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>1.823356e-07</td>\n",
       "      <td>3.646712e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>2.591912</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0042430</td>\n",
       "      <td>indole-containing compound metabolic process</td>\n",
       "      <td>25</td>\n",
       "      <td>3522</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>2.515474e-12</td>\n",
       "      <td>5.030948e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>2.546680</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GO:0046394</td>\n",
       "      <td>carboxylic acid biosynthetic process</td>\n",
       "      <td>120</td>\n",
       "      <td>3427</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>1.938932e-07</td>\n",
       "      <td>3.877864e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>2.368647</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0009850</td>\n",
       "      <td>auxin metabolic process</td>\n",
       "      <td>26</td>\n",
       "      <td>3521</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>3.715543e-20</td>\n",
       "      <td>7.431086e-19</td>\n",
       "      <td>True</td>\n",
       "      <td>2.165312</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0042446</td>\n",
       "      <td>hormone biosynthetic process</td>\n",
       "      <td>70</td>\n",
       "      <td>3477</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>1.515435e-17</td>\n",
       "      <td>3.030871e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>1.939366</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GO:0016053</td>\n",
       "      <td>organic acid biosynthetic process</td>\n",
       "      <td>120</td>\n",
       "      <td>3427</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>1.938932e-07</td>\n",
       "      <td>3.877864e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>1.878572</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GO:0034754</td>\n",
       "      <td>cellular hormone metabolic process</td>\n",
       "      <td>25</td>\n",
       "      <td>3522</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>9.205385e-09</td>\n",
       "      <td>1.841077e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>1.836909</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GO:0018130</td>\n",
       "      <td>heterocycle biosynthetic process</td>\n",
       "      <td>107</td>\n",
       "      <td>3440</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>8.348857e-08</td>\n",
       "      <td>1.669771e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>1.566799</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GO:0044271</td>\n",
       "      <td>cellular nitrogen compound biosynthetic process</td>\n",
       "      <td>114</td>\n",
       "      <td>3433</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>1.330192e-07</td>\n",
       "      <td>2.660385e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>1.553232</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0010817</td>\n",
       "      <td>regulation of hormone levels</td>\n",
       "      <td>140</td>\n",
       "      <td>3407</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>6.011298e-14</td>\n",
       "      <td>1.202260e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>1.312425</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0042445</td>\n",
       "      <td>hormone metabolic process</td>\n",
       "      <td>90</td>\n",
       "      <td>3457</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>2.954183e-16</td>\n",
       "      <td>5.908365e-15</td>\n",
       "      <td>True</td>\n",
       "      <td>1.080766</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0044249</td>\n",
       "      <td>cellular biosynthetic process</td>\n",
       "      <td>388</td>\n",
       "      <td>3159</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>8.851918e-10</td>\n",
       "      <td>1.770384e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.953573</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0065008</td>\n",
       "      <td>regulation of biological quality</td>\n",
       "      <td>273</td>\n",
       "      <td>3274</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>2.026121e-10</td>\n",
       "      <td>4.052241e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>0.617445</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GO:0009058</td>\n",
       "      <td>biosynthetic process</td>\n",
       "      <td>438</td>\n",
       "      <td>3109</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>4.291636e-09</td>\n",
       "      <td>8.583273e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.605750</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       term_id                                       term_label  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant  info_content significance\n",
       "7   GO:0009684           indoleacetic acid biosynthetic process           9           3538                 6                   14  6.605807e-11  1.321161e-09         True      4.489831         ****\n",
       "19  GO:0048825                            cotyledon development          23           3524                 5                   15  2.928356e-07  5.856712e-06         True      4.000000         ****\n",
       "5   GO:0042435  indole-containing compound biosynthetic process          20           3527                 8                   12  5.716694e-13  1.143339e-11         True      3.632673         ****\n",
       "11  GO:0072330         monocarboxylic acid biosynthetic process          66           3481                 8                   12  2.411502e-09  4.823003e-08         True      3.313892         ****\n",
       "0   GO:0009851                       auxin biosynthetic process          17           3530                12                    8  7.257074e-22  1.451415e-20         True      3.277093         ****\n",
       "8   GO:0009683              indoleacetic acid metabolic process          10           3537                 6                   14  1.053362e-10  2.106724e-09         True      3.144588         ****\n",
       "16  GO:0032787            monocarboxylic acid metabolic process         119           3428                 8                   12  1.823356e-07  3.646712e-06         True      2.591912         ****\n",
       "6   GO:0042430     indole-containing compound metabolic process          25           3522                 8                   12  2.515474e-12  5.030948e-11         True      2.546680         ****\n",
       "17  GO:0046394             carboxylic acid biosynthetic process         120           3427                 8                   12  1.938932e-07  3.877864e-06         True      2.368647         ****\n",
       "1   GO:0009850                          auxin metabolic process          26           3521                12                    8  3.715543e-20  7.431086e-19         True      2.165312         ****\n",
       "2   GO:0042446                     hormone biosynthetic process          70           3477                13                    7  1.515435e-17  3.030871e-16         True      1.939366         ****\n",
       "18  GO:0016053                organic acid biosynthetic process         120           3427                 8                   12  1.938932e-07  3.877864e-06         True      1.878572         ****\n",
       "13  GO:0034754               cellular hormone metabolic process          25           3522                 6                   14  9.205385e-09  1.841077e-07         True      1.836909         ****\n",
       "14  GO:0018130                 heterocycle biosynthetic process         107           3440                 8                   12  8.348857e-08  1.669771e-06         True      1.566799         ****\n",
       "15  GO:0044271  cellular nitrogen compound biosynthetic process         114           3433                 8                   12  1.330192e-07  2.660385e-06         True      1.553232         ****\n",
       "4   GO:0010817                     regulation of hormone levels         140           3407                13                    7  6.011298e-14  1.202260e-12         True      1.312425         ****\n",
       "3   GO:0042445                        hormone metabolic process          90           3457                13                    7  2.954183e-16  5.908365e-15         True      1.080766         ****\n",
       "10  GO:0044249                    cellular biosynthetic process         388           3159                14                    6  8.851918e-10  1.770384e-08         True      0.953573         ****\n",
       "9   GO:0065008                 regulation of biological quality         273           3274                13                    7  2.026121e-10  4.052241e-09         True      0.617445         ****\n",
       "12  GO:0009058                             biosynthetic process         438           3109                14                    6  4.291636e-09  8.583273e-08         True      0.605750         ****"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_go_annotations, gene_ids_in_this_pathway, go).head(20)\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "\n",
    "\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: go.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bps</td>\n",
       "      <td>9</td>\n",
       "      <td>3538</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>7.472458e-09</td>\n",
       "      <td>1.494492e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ga</td>\n",
       "      <td>49</td>\n",
       "      <td>3498</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2.951314e-06</td>\n",
       "      <td>5.902628e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>entirely</td>\n",
       "      <td>12</td>\n",
       "      <td>3535</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3.362705e-06</td>\n",
       "      <td>6.725410e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>inhibitor</td>\n",
       "      <td>56</td>\n",
       "      <td>3491</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4.883520e-06</td>\n",
       "      <td>9.767040e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>germinate</td>\n",
       "      <td>62</td>\n",
       "      <td>3485</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>7.178407e-06</td>\n",
       "      <td>1.435681e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>paclobutrazol</td>\n",
       "      <td>21</td>\n",
       "      <td>3526</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1.481661e-05</td>\n",
       "      <td>2.963323e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>aba</td>\n",
       "      <td>201</td>\n",
       "      <td>3346</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3.022841e-05</td>\n",
       "      <td>6.045683e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>become</td>\n",
       "      <td>43</td>\n",
       "      <td>3504</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1.085597e-04</td>\n",
       "      <td>2.171194e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>if</td>\n",
       "      <td>43</td>\n",
       "      <td>3504</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1.085597e-04</td>\n",
       "      <td>2.171194e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>neoxanthin</td>\n",
       "      <td>6</td>\n",
       "      <td>3541</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1.232688e-04</td>\n",
       "      <td>2.465377e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>suppressed</td>\n",
       "      <td>47</td>\n",
       "      <td>3500</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1.395742e-04</td>\n",
       "      <td>2.791484e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>synthesis</td>\n",
       "      <td>49</td>\n",
       "      <td>3498</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1.570425e-04</td>\n",
       "      <td>3.140850e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>violaxanthin</td>\n",
       "      <td>7</td>\n",
       "      <td>3540</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1.583100e-04</td>\n",
       "      <td>3.166200e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>far</td>\n",
       "      <td>51</td>\n",
       "      <td>3496</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1.758823e-04</td>\n",
       "      <td>3.517646e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>xanthoxin</td>\n",
       "      <td>8</td>\n",
       "      <td>3539</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1.976646e-04</td>\n",
       "      <td>3.953292e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>irregulare</td>\n",
       "      <td>9</td>\n",
       "      <td>3538</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2.413179e-04</td>\n",
       "      <td>4.826358e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sorbitol</td>\n",
       "      <td>9</td>\n",
       "      <td>3538</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2.413179e-04</td>\n",
       "      <td>4.826358e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>germination</td>\n",
       "      <td>318</td>\n",
       "      <td>3229</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2.672120e-04</td>\n",
       "      <td>5.344240e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>were</td>\n",
       "      <td>323</td>\n",
       "      <td>3224</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2.875900e-04</td>\n",
       "      <td>5.751800e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>single</td>\n",
       "      <td>336</td>\n",
       "      <td>3211</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3.462566e-04</td>\n",
       "      <td>6.925131e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant significance\n",
       "0             bps           9           3538                 4                    4  7.472458e-09  1.494492e-07         True         ****\n",
       "1              ga          49           3498                 4                    4  2.951314e-06  5.902628e-05         True         ****\n",
       "2        entirely          12           3535                 3                    5  3.362705e-06  6.725410e-05         True         ****\n",
       "3       inhibitor          56           3491                 4                    4  4.883520e-06  9.767040e-05         True         ****\n",
       "4       germinate          62           3485                 4                    4  7.178407e-06  1.435681e-04         True          ***\n",
       "5   paclobutrazol          21           3526                 3                    5  1.481661e-05  2.963323e-04         True          ***\n",
       "6             aba         201           3346                 5                    3  3.022841e-05  6.045683e-04         True          ***\n",
       "7          become          43           3504                 3                    5  1.085597e-04  2.171194e-03         True           **\n",
       "8              if          43           3504                 3                    5  1.085597e-04  2.171194e-03         True           **\n",
       "9      neoxanthin           6           3541                 2                    6  1.232688e-04  2.465377e-03         True           **\n",
       "10     suppressed          47           3500                 3                    5  1.395742e-04  2.791484e-03         True           **\n",
       "11      synthesis          49           3498                 3                    5  1.570425e-04  3.140850e-03         True           **\n",
       "12   violaxanthin           7           3540                 2                    6  1.583100e-04  3.166200e-03         True           **\n",
       "13            far          51           3496                 3                    5  1.758823e-04  3.517646e-03         True           **\n",
       "14      xanthoxin           8           3539                 2                    6  1.976646e-04  3.953292e-03         True           **\n",
       "15     irregulare           9           3538                 2                    6  2.413179e-04  4.826358e-03         True           **\n",
       "16       sorbitol           9           3538                 2                    6  2.413179e-04  4.826358e-03         True           **\n",
       "17    germination         318           3229                 5                    3  2.672120e-04  5.344240e-03         True           **\n",
       "18           were         323           3224                 5                    3  2.875900e-04  5.751800e-03         True           **\n",
       "19         single         336           3211                 5                    3  3.462566e-04  6.925131e-03         True           **"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = token_enrichment(texts, gene_ids_in_this_pathway).head(20)\n",
    "\n",
    "\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "\n",
    "# This lambda won't work if passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in results.itertuples():\n",
    "    wordcloud[\"Weight\"].append(int(1/row.p_value_adj))\n",
    "    wordcloud[\"Word\"].append(row.token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Weight</th>\n",
       "      <th>Word</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>91</td>\n",
       "      <td>PO:0000229 (flower meristem)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>47</td>\n",
       "      <td>PO:0006020 (lateral root apical meristem)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>113</td>\n",
       "      <td>PO:0006056 (cotyledon epidermis)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>81</td>\n",
       "      <td>PO:0008028 (reproductive shoot apical meristem)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>PO:0020147 (root apical meristem)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>839</td>\n",
       "      <td>PO:0006085 (root meristem)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>169</td>\n",
       "      <td>PO:0025277 (pollen sac)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>648</td>\n",
       "      <td>PO:0000016 (lateral root primordium)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5306</td>\n",
       "      <td>PO:0000056 (flower bud)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>173</td>\n",
       "      <td>PO:0020144 (apical meristem)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>97</td>\n",
       "      <td>PO:0025127 (primordium)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>265</td>\n",
       "      <td>PO:0025233 (portion of embryo plant tissue)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>70</td>\n",
       "      <td>PO:0003011 (root vascular system)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>26</td>\n",
       "      <td>PO:0008015 (hypocotyl vascular system)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>113</td>\n",
       "      <td>PO:0008011 (plant embryo vascular system)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>169</td>\n",
       "      <td>PO:0025202 (microsporangium)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>540</td>\n",
       "      <td>PO:0005029 (root primordium)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>4849</td>\n",
       "      <td>PO:0025084 (reproductive bud)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>26</td>\n",
       "      <td>PO:0025094 (sporangium)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>21320</td>\n",
       "      <td>PO:0000055 (bud)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>13328406</td>\n",
       "      <td>indole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1614353</td>\n",
       "      <td>hypocotyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1007166</td>\n",
       "      <td>acetonitrile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1007166</td>\n",
       "      <td>ian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>592088</td>\n",
       "      <td>iaa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>170085</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>144493</td>\n",
       "      <td>shade</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>99812</td>\n",
       "      <td>failure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>72283</td>\n",
       "      <td>changes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>71969</td>\n",
       "      <td>camalexin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>31317</td>\n",
       "      <td>this</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>17176</td>\n",
       "      <td>inflorescences</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>16753</td>\n",
       "      <td>propensity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>16753</td>\n",
       "      <td>monopteros</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>16753</td>\n",
       "      <td>segragegates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>15582</td>\n",
       "      <td>reduction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>13045</td>\n",
       "      <td>extremely</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>11837</td>\n",
       "      <td>apical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>9583</td>\n",
       "      <td>addition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>6702</td>\n",
       "      <td>phenotypes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>30</td>\n",
       "      <td>PO:0000258 (root cortex)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>50</td>\n",
       "      <td>PO:0007057 (seed germination stage)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>67</td>\n",
       "      <td>PO:0020031 (radicle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>25</td>\n",
       "      <td>PO:0006203 (pericycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>38</td>\n",
       "      <td>PO:0000045 (embryo root)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>22</td>\n",
       "      <td>PO:0005708 (cortex)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>6691238</td>\n",
       "      <td>bps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>16941</td>\n",
       "      <td>ga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>14868</td>\n",
       "      <td>entirely</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>10238</td>\n",
       "      <td>inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>6965</td>\n",
       "      <td>germinate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>3374</td>\n",
       "      <td>paclobutrazol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>1654</td>\n",
       "      <td>aba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>460</td>\n",
       "      <td>become</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>460</td>\n",
       "      <td>if</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>405</td>\n",
       "      <td>neoxanthin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>358</td>\n",
       "      <td>suppressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>318</td>\n",
       "      <td>synthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>315</td>\n",
       "      <td>violaxanthin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>284</td>\n",
       "      <td>far</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>252</td>\n",
       "      <td>xanthoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>207</td>\n",
       "      <td>irregulare</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>207</td>\n",
       "      <td>sorbitol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>187</td>\n",
       "      <td>germination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>173</td>\n",
       "      <td>were</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>144</td>\n",
       "      <td>single</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Weight                                             Word\n",
       "0         91                     PO:0000229 (flower meristem)\n",
       "1         47        PO:0006020 (lateral root apical meristem)\n",
       "2        113                 PO:0006056 (cotyledon epidermis)\n",
       "3         81  PO:0008028 (reproductive shoot apical meristem)\n",
       "4         27                PO:0020147 (root apical meristem)\n",
       "5        839                       PO:0006085 (root meristem)\n",
       "6        169                          PO:0025277 (pollen sac)\n",
       "7        648             PO:0000016 (lateral root primordium)\n",
       "8       5306                          PO:0000056 (flower bud)\n",
       "9        173                     PO:0020144 (apical meristem)\n",
       "10        97                          PO:0025127 (primordium)\n",
       "11       265      PO:0025233 (portion of embryo plant tissue)\n",
       "12        70                PO:0003011 (root vascular system)\n",
       "13        26           PO:0008015 (hypocotyl vascular system)\n",
       "14       113        PO:0008011 (plant embryo vascular system)\n",
       "15       169                     PO:0025202 (microsporangium)\n",
       "16       540                     PO:0005029 (root primordium)\n",
       "17      4849                    PO:0025084 (reproductive bud)\n",
       "18        26                          PO:0025094 (sporangium)\n",
       "19     21320                                 PO:0000055 (bud)\n",
       "20  13328406                                           indole\n",
       "21   1614353                                        hypocotyl\n",
       "22   1007166                                     acetonitrile\n",
       "23   1007166                                              ian\n",
       "24    592088                                              iaa\n",
       "25    170085                                        treatment\n",
       "26    144493                                            shade\n",
       "27     99812                                          failure\n",
       "28     72283                                          changes\n",
       "29     71969                                        camalexin\n",
       "30     31317                                             this\n",
       "31     17176                                   inflorescences\n",
       "32     16753                                       propensity\n",
       "33     16753                                       monopteros\n",
       "34     16753                                     segragegates\n",
       "35     15582                                        reduction\n",
       "36     13045                                        extremely\n",
       "37     11837                                           apical\n",
       "38      9583                                         addition\n",
       "39      6702                                       phenotypes\n",
       "40        30                         PO:0000258 (root cortex)\n",
       "41        50              PO:0007057 (seed germination stage)\n",
       "42        67                             PO:0020031 (radicle)\n",
       "43        25                           PO:0006203 (pericycle)\n",
       "44        38                         PO:0000045 (embryo root)\n",
       "45        22                              PO:0005708 (cortex)\n",
       "46   6691238                                              bps\n",
       "47     16941                                               ga\n",
       "48     14868                                         entirely\n",
       "49     10238                                        inhibitor\n",
       "50      6965                                        germinate\n",
       "51      3374                                    paclobutrazol\n",
       "52      1654                                              aba\n",
       "53       460                                           become\n",
       "54       460                                               if\n",
       "55       405                                       neoxanthin\n",
       "56       358                                       suppressed\n",
       "57       318                                        synthesis\n",
       "58       315                                     violaxanthin\n",
       "59       284                                              far\n",
       "60       252                                        xanthoxin\n",
       "61       207                                       irregulare\n",
       "62       207                                         sorbitol\n",
       "63       187                                      germination\n",
       "64       173                                             were\n",
       "65       144                                           single"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(wordcloud).to_csv(os.path.join(OUTPUT_DIR, \"{}_word_cloud.csv\".format(pathway_id)), index=False)\n",
    "pd.DataFrame(wordcloud)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
