{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at the Dataset\n",
    "The purpose of this notebook is to look closer at the dataset of genes, natural language descriptions, and ontology term annotations that are used in this work. As included in the preprocessing notebooks, these data are drawn from files from either publications supplements like Oellrich, Walls et al. (2015) or model species databases such as TAIR, MaizeGDB, and SGN. The datasets are already loaded and merged using classes available through the oats package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warming up PyWSD (takes ~10 secs)... took 5.330681085586548 secs.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.probability import FreqDist\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.annotation.annotation import term_enrichment\n",
    "from oats.nlp.vocabulary import token_enrichment\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>628658</td>\n",
       "      <td>9110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>156</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>342</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>772</td>\n",
       "      <td>389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>786</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>6526</td>\n",
       "      <td>998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>637240</td>\n",
       "      <td>11015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath     628658                 9110\n",
       "1     gmx        156                   49\n",
       "2     mtr        342                  155\n",
       "3     osa        772                  389\n",
       "4     sly        786                  314\n",
       "5     zma       6526                  998\n",
       "6   total     637240                11015"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = load_from_pickle(\"../data/pickles/gene_phenotype_dataset_all_text_and_annotations_unmerged.pickle\")\n",
    "data.to_pandas().head()\n",
    "data.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions\n",
       "0     ath       5851                 3527\n",
       "1     gmx         30                   24\n",
       "2     mtr         37                   36\n",
       "3     osa         92                   85\n",
       "4     sly         70                   70\n",
       "5     zma       1405                  811\n",
       "6   total       7485                 4553"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = load_from_pickle(\"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\")\n",
    "data.filter_has_description()\n",
    "data.to_pandas().head()\n",
    "data.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's there for each species?\n",
    "The previously loaded dataset contains all of the genes that across six plant species that have natural language description data for phenotype(s) related to that gene. Each gene can have multiple descriptions annotated to it, which were combined or concatenated when the datasets from multiple sources were merged in creating the pickled datasets. Arabidopsis has the highest number of genes that satisfy this criteria, followed by maize, and then followed by the other four species which have a relatively low number of genes that satisfy this criteria, atleast given the sources used for this work. Note that the number of unique descriptions is lower than the number of genes in call cases, because multiple genes can have the same phenotype description associated with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wnl = WordNetLemmatizer()\n",
    "lemmatize_doc = lambda d: [wnl.lemmatize(x) for x in simple_preprocess(d)]\n",
    "\n",
    "dists = defaultdict(list)\n",
    "\n",
    "sent_lists = {}\n",
    "token_lists = {}\n",
    "stems_lists = {}\n",
    "lemma_lists = {}\n",
    "\n",
    "\n",
    "# For each individual species.\n",
    "for species in data.get_species():\n",
    "    df = data.to_pandas()\n",
    "    subset = df[df[\"species\"]==species]\n",
    "    sentences = [sent_tokenize(d) for d in subset[\"description\"].values]\n",
    "    descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"description\"].values]\n",
    "    descriptions_stemmed = [preprocess_string(d) for d in subset[\"description\"].values]\n",
    "    descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"description\"].values]\n",
    "    sent_lists[species] = flatten(sentences)\n",
    "    token_lists[species] = flatten(descriptions_not_stemmed)\n",
    "    stems_lists[species] = flatten(descriptions_stemmed)    \n",
    "    lemma_lists[species] = flatten(descriptions_lemmatized)\n",
    "    \n",
    "    # What about the distributions of words per gene and sentences per gene?\n",
    "    dists[\"species\"].extend([species]*subset.shape[0])\n",
    "    dists[\"num_words\"].extend([len(word_tokenize(x)) for x in subset[\"description\"].values])\n",
    "    dists[\"num_sents\"].extend([len(sent_tokenize(x)) for x in subset[\"description\"].values])\n",
    "    \n",
    "# For the entire dataset including all of the species.\n",
    "df = data.to_pandas()\n",
    "subset = df\n",
    "sentences = [sent_tokenize(d) for d in subset[\"description\"].values]\n",
    "descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"description\"].values]\n",
    "descriptions_stemmed = [preprocess_string(d) for d in subset[\"description\"].values]\n",
    "descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"description\"].values]\n",
    "sent_lists[\"total\"] = flatten(sentences)\n",
    "token_lists[\"total\"] = flatten(descriptions_not_stemmed)\n",
    "stems_lists[\"total\"] = flatten(descriptions_stemmed)    \n",
    "lemma_lists[\"total\"] = flatten(descriptions_lemmatized)\n",
    "\n",
    "# What about lemmas that are uniquely used for a particular species?\n",
    "lemma_sets_unique_to_species = {}\n",
    "for species in data.get_species():\n",
    "    other_species = [s for s in data.get_species() if s != species]\n",
    "    lemmas_used_in_other_species = set(flatten([lemma_lists[s] for s in other_species]))\n",
    "    unique_lemmas = set(lemma_lists[species]).difference(lemmas_used_in_other_species)\n",
    "    lemma_sets_unique_to_species[species] = unique_lemmas\n",
    "lemma_sets_unique_to_species[\"total\"] = flatten([list(s) for s in lemma_sets_unique_to_species.values()])\n",
    "\n",
    "    \n",
    "# Create a dataframe to contain the summarizing information about this dataset, and sort it by number of genes.\n",
    "df = data.describe() \n",
    "condition = (df.species==\"total\")\n",
    "excluded = df[condition]\n",
    "included = df[~condition]\n",
    "df_sorted = included.sort_values(by=\"num_genes\", ascending=False)\n",
    "df = pd.concat([df_sorted,excluded])\n",
    "\n",
    "# Add columns summarizing information about the text descriptions in the dataset.\n",
    "df[\"total_sents\"] = df[\"species\"].map(lambda x: len(sent_lists[x]))\n",
    "df[\"total_words\"] = df[\"species\"].map(lambda x: len(token_lists[x]))\n",
    "df[\"unique_words\"] = df[\"species\"].map(lambda x: len(set(token_lists[x])))\n",
    "df[\"unique_stems\"] = df[\"species\"].map(lambda x: len(set(stems_lists[x])))\n",
    "df[\"total_lemmas\"] = df[\"species\"].map(lambda x: len(lemma_lists[x]))\n",
    "df[\"unique_lemmas\"] = df[\"species\"].map(lambda x: len(set(lemma_lists[x])))\n",
    "df[\"unique_lemmas_to_species\"] = df[\"species\"].map(lambda x: len(lemma_sets_unique_to_species[x]))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_words</th>\n",
       "      <th>num_sents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>157</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>22</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>182</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>104</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>127</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>343</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>49</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>169</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>212</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>31</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species  num_words  num_sents\n",
       "0      ath          8          2\n",
       "1      ath         17          4\n",
       "2      ath         16          2\n",
       "3      ath         30          3\n",
       "4      ath        157         11\n",
       "5      ath         22          4\n",
       "6      ath        182         33\n",
       "7      ath        104          7\n",
       "8      ath        127         13\n",
       "9      ath        343         30\n",
       "10     ath         49          6\n",
       "11     ath         40          3\n",
       "12     ath          3          1\n",
       "13     ath          3          1\n",
       "14     ath        169         15\n",
       "15     ath          5          1\n",
       "16     ath        212         16\n",
       "17     ath          5          1\n",
       "18     ath         23          3\n",
       "19     ath         31          7"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_distributions = pd.DataFrame(dists)\n",
    "text_distributions.to_csv(\"../data/scratch/word_sent_distributions.csv\", index=False)\n",
    "text_distributions.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the ontology term annotations for each species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many of the genes in this dataset for each species are mapped to atleast one term from a given ontology?\n",
    "num_mapped_go = {}\n",
    "num_mapped_po = {}\n",
    "for species in data.get_species():\n",
    "    d = data.to_pandas()\n",
    "    subset = d[d[\"species\"]==species]    \n",
    "    num_mapped_po[species] = len([t for t in subset[\"term_ids\"].values if \"PO\" in t])\n",
    "    num_mapped_go[species] = len([t for t in subset[\"term_ids\"].values if \"GO\" in t])\n",
    "num_mapped_go[\"total\"] = sum(list(num_mapped_go.values()))   \n",
    "num_mapped_po[\"total\"] = sum(list(num_mapped_po.values()))\n",
    "df[\"go\"] = df[\"species\"].map(lambda x: num_mapped_go[x])\n",
    "df[\"po\"] = df[\"species\"].map(lambda x: num_mapped_po[x])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the biologically relevant groups like biochemical pathways and phenotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "      <td>1345</td>\n",
       "      <td>807</td>\n",
       "      <td>2355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "      <td>1519</td>\n",
       "      <td>947</td>\n",
       "      <td>2355</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003  1345       807   2355\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111   156       133      0\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92     0         3      0\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65    17         2      0\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32     0         2      0\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27     1         0      0\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330  1519       947   2355"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the groupings that we're interested in mapping to?\n",
    "kegg_pathways_filename = \"../data/pickles/groupings_from_kegg_pathways.pickle\" \n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "lloyd_subsets_filename = \"../data/pickles/groupings_from_lloyd_subsets.pickle\"                     \n",
    "groupings_dict = {\"kegg\":kegg_pathways_filename,\"plantcyc\":pmn_pathways_filename,\"lloyd\":lloyd_subsets_filename}\n",
    "\n",
    "for name,filename in groupings_dict.items():\n",
    "    groups = load_from_pickle(filename)\n",
    "    id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "    group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "    species_dict = data.get_species_dictionary()\n",
    "    num_mapped = {}\n",
    "    for species in data.get_species():\n",
    "        num_mapped[species] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "    num_mapped[\"total\"] = sum(list(num_mapped.values()))    \n",
    "    df[name] = df[\"species\"].map(lambda x: num_mapped[x])  \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the other biologically relevant information like orthologous genes and protein interactions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd</th>\n",
       "      <th>panther</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "      <td>1345</td>\n",
       "      <td>807</td>\n",
       "      <td>2355</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "      <td>1519</td>\n",
       "      <td>947</td>\n",
       "      <td>2355</td>\n",
       "      <td>884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd  panther\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003  1345       807   2355      344\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111   156       133      0      443\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92     0         3      0       86\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65    17         2      0       11\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32     0         2      0        0\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27     1         0      0        0\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330  1519       947   2355      884"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PantherDB for plant orthologs.\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\"\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in ortholog_edgelist.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"panther\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_genes</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd</th>\n",
       "      <th>panther</th>\n",
       "      <th>stringdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5851</td>\n",
       "      <td>3527</td>\n",
       "      <td>32401</td>\n",
       "      <td>264099</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>264099</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>5387</td>\n",
       "      <td>4003</td>\n",
       "      <td>1345</td>\n",
       "      <td>807</td>\n",
       "      <td>2355</td>\n",
       "      <td>344</td>\n",
       "      <td>3898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>811</td>\n",
       "      <td>8069</td>\n",
       "      <td>50029</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>50029</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>533</td>\n",
       "      <td>3887</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3887</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>399</td>\n",
       "      <td>1810</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1810</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>281</td>\n",
       "      <td>2672</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2672</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>66</td>\n",
       "      <td>233</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>233</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7485</td>\n",
       "      <td>4553</td>\n",
       "      <td>41749</td>\n",
       "      <td>322730</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>322730</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5698</td>\n",
       "      <td>4330</td>\n",
       "      <td>1519</td>\n",
       "      <td>947</td>\n",
       "      <td>2355</td>\n",
       "      <td>884</td>\n",
       "      <td>4134</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  num_genes  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd  panther  stringdb\n",
       "0     ath       5851                 3527        32401       264099          7085          5116        264099           6561                      4864  5387  4003  1345       807   2355      344      3898\n",
       "5     zma       1405                  811         8069        50029          1846          1317         50029           1722                       503   184   111   156       133      0      443       168\n",
       "3     osa         92                   85          533         3887           826           586          3887            760                        99    46    92     0         3      0       86        45\n",
       "4     sly         70                   70          399         1810           577           438          1810            552                        99    23    65    17         2      0       11         8\n",
       "2     mtr         37                   36          281         2672           718           516          2672            671                       126    30    32     0         2      0        0        13\n",
       "1     gmx         30                   24           66          233            81            68           233             78                        12    28    27     1         0      0        0         2\n",
       "6   total       7485                 4553        41749       322730          8043          5802        322730           7443                      5703  5698  4330  1519       947   2355      884      4134"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# STRING DB for protein-protein interactions.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in string_data.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"stringdb\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the vocabularies used for different species compare?\n",
    "One of the things we are interested in is discovering or recovering phenotype similarity between different species in order to identify phenologs (phenotypes between species that share some underlying genetic cause). For this reason, we are interested in how the vocabularies used to describe phenotypes between different species vary, because this will impact how feasible it is to use a dataset like this to identify phenologs. Because the Arabidopsis and maize datasets are the largest in this case, we will compare the vocabularies used in describing the phenotypes associated with the genes from these species in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fast</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>invasion</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample</td>\n",
       "      <td>NN</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>unbend</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>oxygen</td>\n",
       "      <td>NN</td>\n",
       "      <td>36</td>\n",
       "      <td>0.013631</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>orobanche</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>0.005997</td>\n",
       "      <td>-0.005997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>novo</td>\n",
       "      <td>NN</td>\n",
       "      <td>7</td>\n",
       "      <td>0.002651</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>arose</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nm</td>\n",
       "      <td>NN</td>\n",
       "      <td>24</td>\n",
       "      <td>0.009088</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>supersensitive</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "0             fast             NN         2  0.000757         0  0.000000  0.000757\n",
       "1         invasion             NN         2  0.000757         0  0.000000  0.000757\n",
       "2           sample             NN         4  0.001515         0  0.000000  0.001515\n",
       "3           unbend             NN         1  0.000379         0  0.000000  0.000379\n",
       "4           oxygen             NN        36  0.013631         0  0.000000  0.013631\n",
       "5        orobanche             NN         0  0.000000         3  0.005997 -0.005997\n",
       "6             novo             NN         7  0.002651         0  0.000000  0.002651\n",
       "8            arose             NN         2  0.000757         0  0.000000  0.000757\n",
       "9               nm             NN        24  0.009088         0  0.000000  0.009088\n",
       "10  supersensitive             NN         2  0.000757         0  0.000000  0.000757"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using lemmas as the vocabulary components.\n",
    "vocabs = {s:set(lemma_list) for s,lemma_list in lemma_lists.items()}\n",
    "fdist_zma = FreqDist(lemma_lists[\"zma\"])\n",
    "fdist_ath = FreqDist(lemma_lists[\"ath\"])\n",
    "\n",
    "# Using word stems as the vocabulary components.\n",
    "#vocabs = {s:set(stems_list) for s,stems_list in stems_lists.items()}\n",
    "#fdist_zma = FreqDist(stems_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(stems_lists[\"ath\"])\n",
    "\n",
    "# Using tokens (full words) as the vocabulary components.\n",
    "#vocabs = {s:set(token_list) for s,token_list in token_lists.items()}\n",
    "#fdist_zma = FreqDist(token_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(token_lists[\"ath\"])\n",
    "\n",
    "union_vocab = vocabs[\"zma\"].union(vocabs[\"ath\"])\n",
    "table = pd.DataFrame({\"token\":list(union_vocab)})\n",
    "stops = set(stopwords.words('english'))\n",
    "table = table[~table.token.isin(stops)]\n",
    "table[\"part_of_speech\"] = table[\"token\"].map(lambda x: nltk.pos_tag([x])[0][1][:2])\n",
    "table[\"ath_freq\"] = table[\"token\"].map(lambda x: fdist_ath[x])\n",
    "table[\"ath_rate\"] = table[\"ath_freq\"]*100/len(token_lists[\"ath\"])\n",
    "table[\"zma_freq\"] = table[\"token\"].map(lambda x: fdist_zma[x])\n",
    "table[\"zma_rate\"] = table[\"zma_freq\"]*100/len(token_lists[\"zma\"])\n",
    "table[\"diff\"] = table[\"ath_rate\"]-table[\"zma_rate\"]\n",
    "table.to_csv(\"../data/scratch/token_frequencies.csv\")\n",
    "table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3051</th>\n",
       "      <td>embryo</td>\n",
       "      <td>NN</td>\n",
       "      <td>4987</td>\n",
       "      <td>1.888307</td>\n",
       "      <td>151</td>\n",
       "      <td>0.301825</td>\n",
       "      <td>1.586482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>mutant</td>\n",
       "      <td>NN</td>\n",
       "      <td>4778</td>\n",
       "      <td>1.809170</td>\n",
       "      <td>256</td>\n",
       "      <td>0.511703</td>\n",
       "      <td>1.297467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>phenotype</td>\n",
       "      <td>NN</td>\n",
       "      <td>3583</td>\n",
       "      <td>1.356688</td>\n",
       "      <td>78</td>\n",
       "      <td>0.155910</td>\n",
       "      <td>1.200779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>root</td>\n",
       "      <td>NN</td>\n",
       "      <td>2942</td>\n",
       "      <td>1.113976</td>\n",
       "      <td>72</td>\n",
       "      <td>0.143917</td>\n",
       "      <td>0.970060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2331</th>\n",
       "      <td>type</td>\n",
       "      <td>NN</td>\n",
       "      <td>2540</td>\n",
       "      <td>0.961761</td>\n",
       "      <td>15</td>\n",
       "      <td>0.029983</td>\n",
       "      <td>0.931778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2352</th>\n",
       "      <td>wild</td>\n",
       "      <td>NN</td>\n",
       "      <td>2456</td>\n",
       "      <td>0.929954</td>\n",
       "      <td>7</td>\n",
       "      <td>0.013992</td>\n",
       "      <td>0.915962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>defective</td>\n",
       "      <td>JJ</td>\n",
       "      <td>3332</td>\n",
       "      <td>1.261648</td>\n",
       "      <td>285</td>\n",
       "      <td>0.569670</td>\n",
       "      <td>0.691979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5367</th>\n",
       "      <td>reduced</td>\n",
       "      <td>VB</td>\n",
       "      <td>2857</td>\n",
       "      <td>1.081791</td>\n",
       "      <td>216</td>\n",
       "      <td>0.431750</td>\n",
       "      <td>0.650042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1314</th>\n",
       "      <td>stage</td>\n",
       "      <td>NN</td>\n",
       "      <td>1814</td>\n",
       "      <td>0.686864</td>\n",
       "      <td>63</td>\n",
       "      <td>0.125927</td>\n",
       "      <td>0.560937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4949</th>\n",
       "      <td>cotyledon</td>\n",
       "      <td>NN</td>\n",
       "      <td>1439</td>\n",
       "      <td>0.544871</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.544871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4991</th>\n",
       "      <td>growth</td>\n",
       "      <td>NN</td>\n",
       "      <td>1829</td>\n",
       "      <td>0.692543</td>\n",
       "      <td>79</td>\n",
       "      <td>0.157908</td>\n",
       "      <td>0.534635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>seed</td>\n",
       "      <td>NN</td>\n",
       "      <td>1890</td>\n",
       "      <td>0.715641</td>\n",
       "      <td>113</td>\n",
       "      <td>0.225869</td>\n",
       "      <td>0.489772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>887</th>\n",
       "      <td>increased</td>\n",
       "      <td>VB</td>\n",
       "      <td>1315</td>\n",
       "      <td>0.497919</td>\n",
       "      <td>11</td>\n",
       "      <td>0.021987</td>\n",
       "      <td>0.475932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783</th>\n",
       "      <td>terminal</td>\n",
       "      <td>NN</td>\n",
       "      <td>1268</td>\n",
       "      <td>0.480123</td>\n",
       "      <td>4</td>\n",
       "      <td>0.007995</td>\n",
       "      <td>0.472128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3953</th>\n",
       "      <td>level</td>\n",
       "      <td>NN</td>\n",
       "      <td>1336</td>\n",
       "      <td>0.505871</td>\n",
       "      <td>22</td>\n",
       "      <td>0.043974</td>\n",
       "      <td>0.461896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>cell</td>\n",
       "      <td>NN</td>\n",
       "      <td>1646</td>\n",
       "      <td>0.623251</td>\n",
       "      <td>112</td>\n",
       "      <td>0.223870</td>\n",
       "      <td>0.399381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3980</th>\n",
       "      <td>rosette</td>\n",
       "      <td>NN</td>\n",
       "      <td>966</td>\n",
       "      <td>0.365772</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>0.363773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6941</th>\n",
       "      <td>double</td>\n",
       "      <td>RB</td>\n",
       "      <td>956</td>\n",
       "      <td>0.361985</td>\n",
       "      <td>3</td>\n",
       "      <td>0.005997</td>\n",
       "      <td>0.355989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2543</th>\n",
       "      <td>globular</td>\n",
       "      <td>NN</td>\n",
       "      <td>905</td>\n",
       "      <td>0.342675</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.342675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5175</th>\n",
       "      <td>abnormal</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1048</td>\n",
       "      <td>0.396821</td>\n",
       "      <td>28</td>\n",
       "      <td>0.055968</td>\n",
       "      <td>0.340853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5480</th>\n",
       "      <td>wa</td>\n",
       "      <td>NN</td>\n",
       "      <td>886</td>\n",
       "      <td>0.335480</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>0.333481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>flower</td>\n",
       "      <td>NN</td>\n",
       "      <td>940</td>\n",
       "      <td>0.355927</td>\n",
       "      <td>14</td>\n",
       "      <td>0.027984</td>\n",
       "      <td>0.327943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>flowering</td>\n",
       "      <td>VB</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.390005</td>\n",
       "      <td>32</td>\n",
       "      <td>0.063963</td>\n",
       "      <td>0.326042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7077</th>\n",
       "      <td>short</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1504</td>\n",
       "      <td>0.569483</td>\n",
       "      <td>123</td>\n",
       "      <td>0.245857</td>\n",
       "      <td>0.323626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>compared</td>\n",
       "      <td>VB</td>\n",
       "      <td>882</td>\n",
       "      <td>0.333966</td>\n",
       "      <td>9</td>\n",
       "      <td>0.017990</td>\n",
       "      <td>0.315976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>980</th>\n",
       "      <td>silique</td>\n",
       "      <td>NN</td>\n",
       "      <td>798</td>\n",
       "      <td>0.302159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.302159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4093</th>\n",
       "      <td>sensitive</td>\n",
       "      <td>JJ</td>\n",
       "      <td>826</td>\n",
       "      <td>0.312762</td>\n",
       "      <td>6</td>\n",
       "      <td>0.011993</td>\n",
       "      <td>0.300768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5362</th>\n",
       "      <td>visible</td>\n",
       "      <td>JJ</td>\n",
       "      <td>838</td>\n",
       "      <td>0.317305</td>\n",
       "      <td>13</td>\n",
       "      <td>0.025985</td>\n",
       "      <td>0.291320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3044</th>\n",
       "      <td>development</td>\n",
       "      <td>NN</td>\n",
       "      <td>862</td>\n",
       "      <td>0.326393</td>\n",
       "      <td>19</td>\n",
       "      <td>0.037978</td>\n",
       "      <td>0.288415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5938</th>\n",
       "      <td>hypocotyl</td>\n",
       "      <td>NN</td>\n",
       "      <td>752</td>\n",
       "      <td>0.284742</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.284742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "3051       embryo             NN      4987  1.888307       151  0.301825  1.586482\n",
       "1460       mutant             NN      4778  1.809170       256  0.511703  1.297467\n",
       "51      phenotype             NN      3583  1.356688        78  0.155910  1.200779\n",
       "247          root             NN      2942  1.113976        72  0.143917  0.970060\n",
       "2331         type             NN      2540  0.961761        15  0.029983  0.931778\n",
       "2352         wild             NN      2456  0.929954         7  0.013992  0.915962\n",
       "402     defective             JJ      3332  1.261648       285  0.569670  0.691979\n",
       "5367      reduced             VB      2857  1.081791       216  0.431750  0.650042\n",
       "1314        stage             NN      1814  0.686864        63  0.125927  0.560937\n",
       "4949    cotyledon             NN      1439  0.544871         0  0.000000  0.544871\n",
       "4991       growth             NN      1829  0.692543        79  0.157908  0.534635\n",
       "745          seed             NN      1890  0.715641       113  0.225869  0.489772\n",
       "887     increased             VB      1315  0.497919        11  0.021987  0.475932\n",
       "4783     terminal             NN      1268  0.480123         4  0.007995  0.472128\n",
       "3953        level             NN      1336  0.505871        22  0.043974  0.461896\n",
       "1855         cell             NN      1646  0.623251       112  0.223870  0.399381\n",
       "3980      rosette             NN       966  0.365772         1  0.001999  0.363773\n",
       "6941       double             RB       956  0.361985         3  0.005997  0.355989\n",
       "2543     globular             NN       905  0.342675         0  0.000000  0.342675\n",
       "5175     abnormal             JJ      1048  0.396821        28  0.055968  0.340853\n",
       "5480           wa             NN       886  0.335480         1  0.001999  0.333481\n",
       "952        flower             NN       940  0.355927        14  0.027984  0.327943\n",
       "1524    flowering             VB      1030  0.390005        32  0.063963  0.326042\n",
       "7077        short             JJ      1504  0.569483       123  0.245857  0.323626\n",
       "208      compared             VB       882  0.333966         9  0.017990  0.315976\n",
       "980       silique             NN       798  0.302159         0  0.000000  0.302159\n",
       "4093    sensitive             JJ       826  0.312762         6  0.011993  0.300768\n",
       "5362      visible             JJ       838  0.317305        13  0.025985  0.291320\n",
       "3044  development             NN       862  0.326393        19  0.037978  0.288415\n",
       "5938    hypocotyl             NN       752  0.284742         0  0.000000  0.284742"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for Arabidopsis than maize descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=False, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4947</th>\n",
       "      <td>endosperm</td>\n",
       "      <td>NN</td>\n",
       "      <td>124</td>\n",
       "      <td>0.046952</td>\n",
       "      <td>1078</td>\n",
       "      <td>2.154750</td>\n",
       "      <td>-2.107798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6563</th>\n",
       "      <td>seedling</td>\n",
       "      <td>VB</td>\n",
       "      <td>1560</td>\n",
       "      <td>0.590688</td>\n",
       "      <td>1318</td>\n",
       "      <td>2.634472</td>\n",
       "      <td>-2.043784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3269</th>\n",
       "      <td>kernel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>766</td>\n",
       "      <td>1.531112</td>\n",
       "      <td>-1.531112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4002</th>\n",
       "      <td>yellow</td>\n",
       "      <td>NN</td>\n",
       "      <td>304</td>\n",
       "      <td>0.115108</td>\n",
       "      <td>775</td>\n",
       "      <td>1.549102</td>\n",
       "      <td>-1.433993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4349</th>\n",
       "      <td>leaf</td>\n",
       "      <td>NN</td>\n",
       "      <td>3876</td>\n",
       "      <td>1.467631</td>\n",
       "      <td>1445</td>\n",
       "      <td>2.888325</td>\n",
       "      <td>-1.420693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>green</td>\n",
       "      <td>JJ</td>\n",
       "      <td>883</td>\n",
       "      <td>0.334344</td>\n",
       "      <td>788</td>\n",
       "      <td>1.575086</td>\n",
       "      <td>-1.240742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6395</th>\n",
       "      <td>white</td>\n",
       "      <td>JJ</td>\n",
       "      <td>375</td>\n",
       "      <td>0.141992</td>\n",
       "      <td>642</td>\n",
       "      <td>1.283256</td>\n",
       "      <td>-1.141264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5261</th>\n",
       "      <td>albino</td>\n",
       "      <td>NN</td>\n",
       "      <td>222</td>\n",
       "      <td>0.084059</td>\n",
       "      <td>396</td>\n",
       "      <td>0.791541</td>\n",
       "      <td>-0.707482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5422</th>\n",
       "      <td>usually</td>\n",
       "      <td>RB</td>\n",
       "      <td>46</td>\n",
       "      <td>0.017418</td>\n",
       "      <td>353</td>\n",
       "      <td>0.705591</td>\n",
       "      <td>-0.688173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6069</th>\n",
       "      <td>color</td>\n",
       "      <td>NN</td>\n",
       "      <td>43</td>\n",
       "      <td>0.016282</td>\n",
       "      <td>330</td>\n",
       "      <td>0.659617</td>\n",
       "      <td>-0.643336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3755</th>\n",
       "      <td>tassel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>310</td>\n",
       "      <td>0.619641</td>\n",
       "      <td>-0.619641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5593</th>\n",
       "      <td>ear</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>284</td>\n",
       "      <td>0.567671</td>\n",
       "      <td>-0.567671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3694</th>\n",
       "      <td>often</td>\n",
       "      <td>RB</td>\n",
       "      <td>149</td>\n",
       "      <td>0.056418</td>\n",
       "      <td>278</td>\n",
       "      <td>0.555678</td>\n",
       "      <td>-0.499259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2726</th>\n",
       "      <td>aleurone</td>\n",
       "      <td>NN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001136</td>\n",
       "      <td>248</td>\n",
       "      <td>0.495712</td>\n",
       "      <td>-0.494577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4497</th>\n",
       "      <td>opaque</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>244</td>\n",
       "      <td>0.487717</td>\n",
       "      <td>-0.487338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741</th>\n",
       "      <td>class</td>\n",
       "      <td>NN</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003029</td>\n",
       "      <td>240</td>\n",
       "      <td>0.479722</td>\n",
       "      <td>-0.476693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5701</th>\n",
       "      <td>tissue</td>\n",
       "      <td>NN</td>\n",
       "      <td>236</td>\n",
       "      <td>0.089360</td>\n",
       "      <td>259</td>\n",
       "      <td>0.517700</td>\n",
       "      <td>-0.428339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1704</th>\n",
       "      <td>small</td>\n",
       "      <td>JJ</td>\n",
       "      <td>669</td>\n",
       "      <td>0.253314</td>\n",
       "      <td>323</td>\n",
       "      <td>0.645626</td>\n",
       "      <td>-0.392311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2798</th>\n",
       "      <td>pale</td>\n",
       "      <td>NN</td>\n",
       "      <td>651</td>\n",
       "      <td>0.246498</td>\n",
       "      <td>318</td>\n",
       "      <td>0.635631</td>\n",
       "      <td>-0.389133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4273</th>\n",
       "      <td>vp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190</td>\n",
       "      <td>0.379780</td>\n",
       "      <td>-0.379780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4830</th>\n",
       "      <td>palegreen</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>171</td>\n",
       "      <td>0.341802</td>\n",
       "      <td>-0.341423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5949</th>\n",
       "      <td>general</td>\n",
       "      <td>JJ</td>\n",
       "      <td>13</td>\n",
       "      <td>0.004922</td>\n",
       "      <td>171</td>\n",
       "      <td>0.341802</td>\n",
       "      <td>-0.336879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6691</th>\n",
       "      <td>term</td>\n",
       "      <td>NN</td>\n",
       "      <td>22</td>\n",
       "      <td>0.008330</td>\n",
       "      <td>172</td>\n",
       "      <td>0.343801</td>\n",
       "      <td>-0.335470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2676</th>\n",
       "      <td>surface</td>\n",
       "      <td>NN</td>\n",
       "      <td>138</td>\n",
       "      <td>0.052253</td>\n",
       "      <td>190</td>\n",
       "      <td>0.379780</td>\n",
       "      <td>-0.327527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4021</th>\n",
       "      <td>appearance</td>\n",
       "      <td>NN</td>\n",
       "      <td>131</td>\n",
       "      <td>0.049603</td>\n",
       "      <td>188</td>\n",
       "      <td>0.375782</td>\n",
       "      <td>-0.326179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6784</th>\n",
       "      <td>texture</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>163</td>\n",
       "      <td>0.325811</td>\n",
       "      <td>-0.325054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4981</th>\n",
       "      <td>otherwise</td>\n",
       "      <td>RB</td>\n",
       "      <td>10</td>\n",
       "      <td>0.003786</td>\n",
       "      <td>158</td>\n",
       "      <td>0.315817</td>\n",
       "      <td>-0.312030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>anther</td>\n",
       "      <td>NN</td>\n",
       "      <td>139</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>167</td>\n",
       "      <td>0.333806</td>\n",
       "      <td>-0.281175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6855</th>\n",
       "      <td>pericarp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>140</td>\n",
       "      <td>0.279838</td>\n",
       "      <td>-0.279838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2950</th>\n",
       "      <td>sometimes</td>\n",
       "      <td>RB</td>\n",
       "      <td>44</td>\n",
       "      <td>0.016660</td>\n",
       "      <td>148</td>\n",
       "      <td>0.295828</td>\n",
       "      <td>-0.279168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "4947   endosperm             NN       124  0.046952      1078  2.154750 -2.107798\n",
       "6563    seedling             VB      1560  0.590688      1318  2.634472 -2.043784\n",
       "3269      kernel             NN         0  0.000000       766  1.531112 -1.531112\n",
       "4002      yellow             NN       304  0.115108       775  1.549102 -1.433993\n",
       "4349        leaf             NN      3876  1.467631      1445  2.888325 -1.420693\n",
       "549        green             JJ       883  0.334344       788  1.575086 -1.240742\n",
       "6395       white             JJ       375  0.141992       642  1.283256 -1.141264\n",
       "5261      albino             NN       222  0.084059       396  0.791541 -0.707482\n",
       "5422     usually             RB        46  0.017418       353  0.705591 -0.688173\n",
       "6069       color             NN        43  0.016282       330  0.659617 -0.643336\n",
       "3755      tassel             NN         0  0.000000       310  0.619641 -0.619641\n",
       "5593         ear             NN         0  0.000000       284  0.567671 -0.567671\n",
       "3694       often             RB       149  0.056418       278  0.555678 -0.499259\n",
       "2726    aleurone             NN         3  0.001136       248  0.495712 -0.494577\n",
       "4497      opaque             NN         1  0.000379       244  0.487717 -0.487338\n",
       "2741       class             NN         8  0.003029       240  0.479722 -0.476693\n",
       "5701      tissue             NN       236  0.089360       259  0.517700 -0.428339\n",
       "1704       small             JJ       669  0.253314       323  0.645626 -0.392311\n",
       "2798        pale             NN       651  0.246498       318  0.635631 -0.389133\n",
       "4273          vp             NN         0  0.000000       190  0.379780 -0.379780\n",
       "4830   palegreen             NN         1  0.000379       171  0.341802 -0.341423\n",
       "5949     general             JJ        13  0.004922       171  0.341802 -0.336879\n",
       "6691        term             NN        22  0.008330       172  0.343801 -0.335470\n",
       "2676     surface             NN       138  0.052253       190  0.379780 -0.327527\n",
       "4021  appearance             NN       131  0.049603       188  0.375782 -0.326179\n",
       "6784     texture             NN         2  0.000757       163  0.325811 -0.325054\n",
       "4981   otherwise             RB        10  0.003786       158  0.315817 -0.312030\n",
       "476       anther             NN       139  0.052632       167  0.333806 -0.281175\n",
       "6855    pericarp             NN         0  0.000000       140  0.279838 -0.279838\n",
       "2950   sometimes             RB        44  0.016660       148  0.295828 -0.279168"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for maize than Arabidopsis descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=True, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MD</td>\n",
       "      <td>0.059431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD</td>\n",
       "      <td>0.027896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IN</td>\n",
       "      <td>0.026204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JJ</td>\n",
       "      <td>0.023686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DT</td>\n",
       "      <td>0.019075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RB</td>\n",
       "      <td>0.016302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NN</td>\n",
       "      <td>0.014463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VB</td>\n",
       "      <td>0.012130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC</td>\n",
       "      <td>0.007617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WP</td>\n",
       "      <td>0.001620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  part_of_speech  abs_diff\n",
       "0             MD  0.059431\n",
       "1             CD  0.027896\n",
       "2             IN  0.026204\n",
       "3             JJ  0.023686\n",
       "4             DT  0.019075\n",
       "5             RB  0.016302\n",
       "6             NN  0.014463\n",
       "7             VB  0.012130\n",
       "8             CC  0.007617\n",
       "9             WP  0.001620"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is the mean absolute value of the rate differences different between the different parts of speech?\n",
    "table[\"abs_diff\"] = abs(table[\"diff\"])\n",
    "pos_table = table.groupby(\"part_of_speech\").mean()\n",
    "pos_table.sort_values(by=\"abs_diff\", inplace=True, ascending=False)\n",
    "pos_table = pos_table[[\"abs_diff\"]]\n",
    "pos_table.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Working on the Venn Diagram for this part, unused currently.\n",
    "#print(table.shape)\n",
    "#zma_only = table[table[\"ath_rate\"]==0]\n",
    "#ath_only = table[table[\"zma_rate\"]==0]\n",
    "#print(zma_only.shape)\n",
    "#print(ath_only.shape)\n",
    "#print(ath_only.shape[0]+zma_only.shape[0])\n",
    "#ath_only.head(10)\n",
    "# We need to create a mapping between stems and the words that were present for them.\n",
    "# This is because what we want is the stems that are exclusive to a species.\n",
    "# but then the words that are actually there for those stems, so that we can count their parts of speech."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at Term and Word Enrichment for Groups of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3815"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the dataset of phenotype descriptions and ontology annotations.\n",
    "data = load_from_pickle(\"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\")\n",
    "data.filter_has_description()\n",
    "data.filter_has_annotation(\"GO\")\n",
    "data.filter_has_annotation(\"PO\")\n",
    "d = data.get_description_dictionary()\n",
    "texts = {i:\" \".join(simple_preprocess(t)) for i,t in d.items()}\n",
    "len(texts)                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create ontology objects for all the biological ontologies being used.\n",
    "go_filename = \"../ontologies/go.obo\"                                                                \n",
    "po_filename = \"../ontologies/po.obo\"                                                                 \n",
    "pato_filename = \"../ontologies/pato.obo\"   \n",
    "pato = Ontology(pato_filename)\n",
    "po = Ontology(po_filename)\n",
    "go = Ontology(go_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "curated_go_annotations = data.get_annotations_dictionary(\"GO\")\n",
    "curated_po_annotations = data.get_annotations_dictionary(\"PO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>num_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sporopollenin precursors biosynthesis</td>\n",
       "      <td>PWY-6733</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gluconeogenesis III</td>\n",
       "      <td>PWY66-399</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gluconeogenesis I</td>\n",
       "      <td>GLUCONEO-PWY</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>suberin monomers biosynthesis</td>\n",
       "      <td>PWY-1121</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phosphatidylcholine acyl editing</td>\n",
       "      <td>PWY-6803</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>indole-3-acetate biosynthesis II</td>\n",
       "      <td>PWY-581</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>glycolysis I (from glucose 6-phosphate)</td>\n",
       "      <td>GLYCOLYSIS</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>glycolysis II (from fructose 6-phosphate)</td>\n",
       "      <td>PWY-5484</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>palmitate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5971</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3-phosphoinositide biosynthesis</td>\n",
       "      <td>PWY-6352</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>phenylpropanoid biosynthesis</td>\n",
       "      <td>PWY-361</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>L-leucine biosynthesis</td>\n",
       "      <td>LEUSYN-PWY</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>starch biosynthesis</td>\n",
       "      <td>PWY-622</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>glycolysis IV (plant cytosol)</td>\n",
       "      <td>PWY-1042</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>stearate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5989</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       pathway_name    pathway_id  num_genes\n",
       "0             sporopollenin precursors biosynthesis      PWY-6733         21\n",
       "1                               gluconeogenesis III     PWY66-399         20\n",
       "2                                 gluconeogenesis I  GLUCONEO-PWY         20\n",
       "3                     suberin monomers biosynthesis      PWY-1121         20\n",
       "4                  phosphatidylcholine acyl editing      PWY-6803         19\n",
       "5                  indole-3-acetate biosynthesis II       PWY-581         18\n",
       "6           glycolysis I (from glucose 6-phosphate)    GLYCOLYSIS         16\n",
       "7         glycolysis II (from fructose 6-phosphate)      PWY-5484         15\n",
       "8   palmitate biosynthesis II (bacteria and plants)      PWY-5971         15\n",
       "9                   3-phosphoinositide biosynthesis      PWY-6352         15\n",
       "10                     phenylpropanoid biosynthesis       PWY-361         15\n",
       "11                           L-leucine biosynthesis    LEUSYN-PWY         14\n",
       "12                              starch biosynthesis       PWY-622         14\n",
       "13                    glycolysis IV (plant cytosol)      PWY-1042         14\n",
       "14   stearate biosynthesis II (bacteria and plants)      PWY-5989         14"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the mappings from this dataset to PlantCyc information.\n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "groups = load_from_pickle(pmn_pathways_filename)\n",
    "id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "\n",
    "# Look at which pathways are best represented in this dataset.\n",
    "pathways_sorted = sorted(group_id_to_ids.items(), key=lambda item: len(item[1]), reverse=True)\n",
    "pathways_sorted_lengths = [(i,len(l)) for (i,l) in pathways_sorted]\n",
    "pathways_df = pd.DataFrame(pathways_sorted_lengths, columns=[\"pathway_id\",\"num_genes\"])\n",
    "pathways_df[\"pathway_name\"] = pathways_df[\"pathway_id\"].map(lambda x: groups.get_long_name(x))\n",
    "pathways_df = pathways_df[[\"pathway_name\",\"pathway_id\",\"num_genes\"]]\n",
    "pathways_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For some example pathway to use.\n",
    "pathway_id = \"PWY-6733\"\n",
    "gene_ids_in_this_pathway = group_id_to_ids[pathway_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PO:0009071</td>\n",
       "      <td>anther wall tapetum</td>\n",
       "      <td>29</td>\n",
       "      <td>3786</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>2.452198e-05</td>\n",
       "      <td>4.904397e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PO:0005059</td>\n",
       "      <td>root endodermis</td>\n",
       "      <td>18</td>\n",
       "      <td>3797</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>1.765972e-04</td>\n",
       "      <td>3.531944e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PO:0025314</td>\n",
       "      <td>microsporangium tapetum</td>\n",
       "      <td>29</td>\n",
       "      <td>3786</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>2.452198e-05</td>\n",
       "      <td>4.904397e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>3.633786</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PO:0009088</td>\n",
       "      <td>seed coat</td>\n",
       "      <td>37</td>\n",
       "      <td>3778</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>1.226875e-03</td>\n",
       "      <td>2.453750e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PO:0025313</td>\n",
       "      <td>tapetum</td>\n",
       "      <td>33</td>\n",
       "      <td>3782</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>3.593559e-08</td>\n",
       "      <td>7.187118e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>2.450679</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PO:0000252</td>\n",
       "      <td>endodermis</td>\n",
       "      <td>25</td>\n",
       "      <td>3790</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>4.243818e-04</td>\n",
       "      <td>8.487637e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>2.290013</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PO:0025059</td>\n",
       "      <td>portion of ground tissue</td>\n",
       "      <td>173</td>\n",
       "      <td>3642</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>1.797079e-07</td>\n",
       "      <td>3.594158e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.748851</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PO:0020008</td>\n",
       "      <td>exine</td>\n",
       "      <td>9</td>\n",
       "      <td>3806</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>5.718655e-09</td>\n",
       "      <td>1.143731e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      term_id                term_label  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant  info_content significance\n",
       "4  PO:0009071       anther wall tapetum          29           3786                 4                   17  2.452198e-05  4.904397e-04         True      5.000000          ***\n",
       "5  PO:0005059           root endodermis          18           3797                 3                   18  1.765972e-04  3.531944e-03         True      4.000000           **\n",
       "3  PO:0025314   microsporangium tapetum          29           3786                 4                   17  2.452198e-05  4.904397e-04         True      3.633786          ***\n",
       "7  PO:0009088                 seed coat          37           3778                 3                   18  1.226875e-03  2.453750e-02         True      3.000000            *\n",
       "1  PO:0025313                   tapetum          33           3782                 6                   15  3.593559e-08  7.187118e-07         True      2.450679         ****\n",
       "6  PO:0000252                endodermis          25           3790                 3                   18  4.243818e-04  8.487637e-03         True      2.290013           **\n",
       "2  PO:0025059  portion of ground tissue         173           3642                 9                   12  1.797079e-07  3.594158e-06         True      0.748851         ****\n",
       "0  PO:0020008                     exine           9           3806                 5                   16  5.718655e-09  1.143731e-07         True      0.000000         ****"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_po_annotations, gene_ids_in_this_pathway, po).head(20)\n",
    "\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "\n",
    "\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: po.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# ns   P > 0.05\n",
    "# *    P  0.05\n",
    "# **   P  0.01\n",
    "# ***  P  0.001\n",
    "# **** P  0.0001\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GO:0050062</td>\n",
       "      <td>long-chain-fatty-acyl-CoA reductase activity</td>\n",
       "      <td>3</td>\n",
       "      <td>3812</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>2.799884e-06</td>\n",
       "      <td>5.599768e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0080110</td>\n",
       "      <td>sporopollenin biosynthetic process</td>\n",
       "      <td>7</td>\n",
       "      <td>3808</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>1.617948e-13</td>\n",
       "      <td>3.235897e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0010584</td>\n",
       "      <td>pollen exine formation</td>\n",
       "      <td>16</td>\n",
       "      <td>3799</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>1.122824e-11</td>\n",
       "      <td>2.245649e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0004467</td>\n",
       "      <td>long-chain fatty acid-CoA ligase activity</td>\n",
       "      <td>7</td>\n",
       "      <td>3808</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>1.617948e-13</td>\n",
       "      <td>3.235897e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>4.356238</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>154</td>\n",
       "      <td>3661</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>1.084597e-06</td>\n",
       "      <td>2.169193e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>4.167950</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0015645</td>\n",
       "      <td>fatty acid ligase activity</td>\n",
       "      <td>8</td>\n",
       "      <td>3807</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>3.023935e-13</td>\n",
       "      <td>6.047871e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>3.846069</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0010208</td>\n",
       "      <td>pollen wall assembly</td>\n",
       "      <td>21</td>\n",
       "      <td>3794</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>5.336455e-11</td>\n",
       "      <td>1.067291e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>3.742495</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GO:0045229</td>\n",
       "      <td>external encapsulating structure organization</td>\n",
       "      <td>57</td>\n",
       "      <td>3758</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>2.492753e-08</td>\n",
       "      <td>4.985505e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>2.611453</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0016405</td>\n",
       "      <td>CoA-ligase activity</td>\n",
       "      <td>16</td>\n",
       "      <td>3799</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>9.561414e-18</td>\n",
       "      <td>1.912283e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>2.511272</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0016878</td>\n",
       "      <td>acid-thiol ligase activity</td>\n",
       "      <td>17</td>\n",
       "      <td>3798</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>1.514595e-17</td>\n",
       "      <td>3.029189e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>2.406091</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GO:0016620</td>\n",
       "      <td>oxidoreductase activity, acting on the aldehyd...</td>\n",
       "      <td>14</td>\n",
       "      <td>3801</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>1.934351e-06</td>\n",
       "      <td>3.868701e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>2.285479</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GO:0006631</td>\n",
       "      <td>fatty acid metabolic process</td>\n",
       "      <td>70</td>\n",
       "      <td>3745</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>2.124112e-06</td>\n",
       "      <td>4.248225e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>2.240498</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0010927</td>\n",
       "      <td>cellular component assembly involved in morpho...</td>\n",
       "      <td>21</td>\n",
       "      <td>3794</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>5.336455e-11</td>\n",
       "      <td>1.067291e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>2.043203</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0016877</td>\n",
       "      <td>ligase activity, forming carbon-sulfur bonds</td>\n",
       "      <td>22</td>\n",
       "      <td>3793</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>1.143091e-16</td>\n",
       "      <td>2.286181e-15</td>\n",
       "      <td>True</td>\n",
       "      <td>1.743159</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0019438</td>\n",
       "      <td>aromatic compound biosynthetic process</td>\n",
       "      <td>130</td>\n",
       "      <td>3685</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>7.625822e-10</td>\n",
       "      <td>1.525164e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>1.575946</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0016874</td>\n",
       "      <td>ligase activity</td>\n",
       "      <td>42</td>\n",
       "      <td>3773</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>2.659235e-14</td>\n",
       "      <td>5.318469e-13</td>\n",
       "      <td>True</td>\n",
       "      <td>0.953336</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GO:0048646</td>\n",
       "      <td>anatomical structure formation involved in mor...</td>\n",
       "      <td>127</td>\n",
       "      <td>3688</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>2.646803e-07</td>\n",
       "      <td>5.293606e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.925652</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GO:0009058</td>\n",
       "      <td>biosynthetic process</td>\n",
       "      <td>469</td>\n",
       "      <td>3346</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>1.306664e-07</td>\n",
       "      <td>2.613329e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.605750</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GO:0008152</td>\n",
       "      <td>metabolic process</td>\n",
       "      <td>1146</td>\n",
       "      <td>2669</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>2.116911e-07</td>\n",
       "      <td>4.233823e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.186546</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GO:0003824</td>\n",
       "      <td>catalytic activity</td>\n",
       "      <td>940</td>\n",
       "      <td>2875</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>7.592595e-09</td>\n",
       "      <td>1.518519e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>0.170517</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       term_id                                         term_label  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant  info_content significance\n",
       "19  GO:0050062       long-chain-fatty-acyl-CoA reductase activity           3           3812                 3                   18  2.799884e-06  5.599768e-05         True      5.000000         ****\n",
       "4   GO:0080110                 sporopollenin biosynthetic process           7           3808                 7                   14  1.617948e-13  3.235897e-12         True      5.000000         ****\n",
       "7   GO:0010584                             pollen exine formation          16           3799                 7                   14  1.122824e-11  2.245649e-10         True      5.000000         ****\n",
       "5   GO:0004467          long-chain fatty acid-CoA ligase activity           7           3808                 7                   14  1.617948e-13  3.235897e-12         True      4.356238         ****\n",
       "16  GO:0005783                              endoplasmic reticulum         154           3661                 8                   13  1.084597e-06  2.169193e-05         True      4.167950         ****\n",
       "6   GO:0015645                         fatty acid ligase activity           8           3807                 7                   14  3.023935e-13  6.047871e-12         True      3.846069         ****\n",
       "9   GO:0010208                               pollen wall assembly          21           3794                 7                   14  5.336455e-11  1.067291e-09         True      3.742495         ****\n",
       "12  GO:0045229      external encapsulating structure organization          57           3758                 7                   14  2.492753e-08  4.985505e-07         True      2.611453         ****\n",
       "0   GO:0016405                                CoA-ligase activity          16           3799                10                   11  9.561414e-18  1.912283e-16         True      2.511272         ****\n",
       "1   GO:0016878                         acid-thiol ligase activity          17           3798                10                   11  1.514595e-17  3.029189e-16         True      2.406091         ****\n",
       "17  GO:0016620  oxidoreductase activity, acting on the aldehyd...          14           3801                 4                   17  1.934351e-06  3.868701e-05         True      2.285479         ****\n",
       "18  GO:0006631                       fatty acid metabolic process          70           3745                 6                   15  2.124112e-06  4.248225e-05         True      2.240498         ****\n",
       "8   GO:0010927  cellular component assembly involved in morpho...          21           3794                 7                   14  5.336455e-11  1.067291e-09         True      2.043203         ****\n",
       "2   GO:0016877       ligase activity, forming carbon-sulfur bonds          22           3793                10                   11  1.143091e-16  2.286181e-15         True      1.743159         ****\n",
       "10  GO:0019438             aromatic compound biosynthetic process         130           3685                10                   11  7.625822e-10  1.525164e-08         True      1.575946         ****\n",
       "3   GO:0016874                                    ligase activity          42           3773                10                   11  2.659235e-14  5.318469e-13         True      0.953336         ****\n",
       "15  GO:0048646  anatomical structure formation involved in mor...         127           3688                 8                   13  2.646803e-07  5.293606e-06         True      0.925652         ****\n",
       "13  GO:0009058                               biosynthetic process         469           3346                13                    8  1.306664e-07  2.613329e-06         True      0.605750         ****\n",
       "14  GO:0008152                                  metabolic process        1146           2669                18                    3  2.116911e-07  4.233823e-06         True      0.186546         ****\n",
       "11  GO:0003824                                 catalytic activity         940           2875                18                    3  7.592595e-09  1.518519e-07         True      0.170517         ****"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_go_annotations, gene_ids_in_this_pathway, go).head(20)\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "\n",
    "\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: go.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>layer</td>\n",
       "      <td>65</td>\n",
       "      <td>3750</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>5.499473e-11</td>\n",
       "      <td>1.099895e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exine</td>\n",
       "      <td>19</td>\n",
       "      <td>3796</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>2.045050e-09</td>\n",
       "      <td>4.090100e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>suberin</td>\n",
       "      <td>8</td>\n",
       "      <td>3807</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>2.269517e-05</td>\n",
       "      <td>4.539035e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lacs</td>\n",
       "      <td>2</td>\n",
       "      <td>3813</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.701698e-04</td>\n",
       "      <td>3.403396e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cutin</td>\n",
       "      <td>4</td>\n",
       "      <td>3811</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>4.226197e-04</td>\n",
       "      <td>8.452394e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>wax</td>\n",
       "      <td>32</td>\n",
       "      <td>3783</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>8.271913e-04</td>\n",
       "      <td>1.654383e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>permeability</td>\n",
       "      <td>7</td>\n",
       "      <td>3808</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.004279e-03</td>\n",
       "      <td>2.008557e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>flavonoids</td>\n",
       "      <td>8</td>\n",
       "      <td>3807</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.251207e-03</td>\n",
       "      <td>2.502414e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>flavonoid</td>\n",
       "      <td>9</td>\n",
       "      <td>3806</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>1.524210e-03</td>\n",
       "      <td>3.048420e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pollen</td>\n",
       "      <td>251</td>\n",
       "      <td>3564</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>1.964829e-03</td>\n",
       "      <td>3.929658e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>glossy</td>\n",
       "      <td>11</td>\n",
       "      <td>3804</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2.147377e-03</td>\n",
       "      <td>4.294755e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>composition</td>\n",
       "      <td>46</td>\n",
       "      <td>3769</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>2.216439e-03</td>\n",
       "      <td>4.432879e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>no</td>\n",
       "      <td>981</td>\n",
       "      <td>2834</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>2.269727e-03</td>\n",
       "      <td>4.539454e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cuticle</td>\n",
       "      <td>12</td>\n",
       "      <td>3803</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2.497017e-03</td>\n",
       "      <td>4.994034e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant significance\n",
       "0          layer          65           3750                 9                   12  5.499473e-11  1.099895e-09         True         ****\n",
       "1          exine          19           3796                 6                   15  2.045050e-09  4.090100e-08         True         ****\n",
       "2        suberin           8           3807                 3                   18  2.269517e-05  4.539035e-04         True          ***\n",
       "3           lacs           2           3813                 2                   19  1.701698e-04  3.403396e-03         True           **\n",
       "4          cutin           4           3811                 2                   19  4.226197e-04  8.452394e-03         True           **\n",
       "5            wax          32           3783                 3                   18  8.271913e-04  1.654383e-02         True            *\n",
       "6   permeability           7           3808                 2                   19  1.004279e-03  2.008557e-02         True            *\n",
       "7     flavonoids           8           3807                 2                   19  1.251207e-03  2.502414e-02         True            *\n",
       "8      flavonoid           9           3806                 2                   19  1.524210e-03  3.048420e-02         True            *\n",
       "9         pollen         251           3564                 6                   15  1.964829e-03  3.929658e-02         True            *\n",
       "10        glossy          11           3804                 2                   19  2.147377e-03  4.294755e-02         True            *\n",
       "11   composition          46           3769                 3                   18  2.216439e-03  4.432879e-02         True            *\n",
       "12            no         981           2834                12                    9  2.269727e-03  4.539454e-02         True            *\n",
       "13       cuticle          12           3803                 2                   19  2.497017e-03  4.994034e-02         True            *"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = token_enrichment(texts, gene_ids_in_this_pathway).head(20)\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
