{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at the Dataset\n",
    "The purpose of this notebook is to look closer at the dataset of genes, natural language descriptions, and ontology term annotations that are used in this work. As included in the preprocessing notebooks, these data are drawn from files from either publications supplements like Oellrich, Walls et al. (2015) or model species databases such as TAIR, MaizeGDB, and SGN. The datasets are already loaded and merged using classes available through the oats package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.probability import FreqDist\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, flatten, to_hms\n",
    "from oats.utils.utils import function_wrapper_with_duration, remove_duplicates_retain_order\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.annotation.annotation import annotate_using_noble_coder, term_enrichment\n",
    "from oats.distances import pairwise as pw\n",
    "from oats.nlp.vocabulary import get_overrepresented_tokens, get_vocab_from_tokens\n",
    "from oats.nlp.vocabulary import reduce_vocab_connected_components, reduce_vocab_linares_pontes, token_enrichment\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paths to the files that are used for this notebook.\n",
    "plant_dataset_path = \"../../plant-data/genes_texts_annots.csv\"\n",
    "\n",
    "# Paths to files with mappings to groups.\n",
    "kegg_pathways_path = \"../../plant-data/reshaped_data/kegg_pathways.csv\" \n",
    "plantcyc_pathways_path = \"../../plant-data/reshaped_data/plantcyc_pathways.csv\" \n",
    "lloyd_meinke_subsets_path = \"../../plant-data/reshaped_data/lloyd_meinke_subsets.csv\" \n",
    "lloyd_meinke_classes_path = \"../../plant-data/reshaped_data/lloyd_meinke_classes.csv\" \n",
    "\n",
    "# Paths to files that map group identifers to longer group names.\n",
    "kegg_pathways_names_path = \"../../plant-data/reshaped_data/kegg_pathways_name_map.csv\"\n",
    "plantcyc_pathways_names_path = \"../../plant-data/reshaped_data/plantcyc_pathways_name_map.csv\"\n",
    "lloyd_meinke_subsets_names_path = \"../../plant-data/reshaped_data/lloyd_meinke_subsets_name_map.csv\"\n",
    "lloyd_meinke_classes_names_path = \"../../plant-data/reshaped_data/lloyd_meinke_classes_name_map.csv\"\n",
    "\n",
    "# Path to file with plant ortholog mappings.\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create and name an output directory according to when the notebooks was run.\n",
    "OUTPUT_NAME = \"composition\"\n",
    "OUTPUT_DIR = os.path.join(\"../outputs\",\"{}_{}_{}\".format(OUTPUT_NAME,datetime.datetime.now().strftime('%m_%d_%Y_h%Hm%Ms%S'),random.randrange(1000,9999)))\n",
    "os.mkdir(OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions\n",
       "0     ath                     5066                 2975\n",
       "1     gmx                       30                   23\n",
       "2     mtr                       37                   36\n",
       "3     osa                       92                   85\n",
       "4     sly                       70                   70\n",
       "5     zma                     1405                  810\n",
       "6   total                     6700                 3999"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading in and describing the dataset of plant genes.\n",
    "plant_dataset = Dataset(plant_dataset_path)\n",
    "plant_dataset.filter_has_description()\n",
    "plant_dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's there for each species?\n",
    "The previously loaded dataset contains all of the genes that across six plant species that have natural language description data for phenotype(s) related to that gene. Each gene can have multiple descriptions annotated to it, which were combined or concatenated when the datasets from multiple sources were merged in creating the pickled datasets. Arabidopsis has the highest number of genes that satisfy this criteria, followed by maize, and then followed by the other four species which have a relatively low number of genes that satisfy this criteria, atleast given the sources used for this work. Note that the number of unique descriptions is lower than the number of genes in call cases, because multiple genes can have the same phenotype description associated with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = plant_dataset\n",
    "\n",
    "wnl = WordNetLemmatizer()\n",
    "lemmatize_doc = lambda d: [wnl.lemmatize(x) for x in simple_preprocess(d)]\n",
    "\n",
    "dists = defaultdict(list)\n",
    "\n",
    "sent_lists = {}\n",
    "token_lists = {}\n",
    "stems_lists = {}\n",
    "lemma_lists = {}\n",
    "\n",
    "\n",
    "# For each individual species.\n",
    "for species in data.get_species():\n",
    "    df = data.to_pandas()\n",
    "    subset = df[df[\"species\"]==species]\n",
    "    sentences = [sent_tokenize(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_stemmed = [preprocess_string(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"descriptions\"].values]\n",
    "    sent_lists[species] = flatten(sentences)\n",
    "    token_lists[species] = flatten(descriptions_not_stemmed)\n",
    "    stems_lists[species] = flatten(descriptions_stemmed)    \n",
    "    lemma_lists[species] = flatten(descriptions_lemmatized)\n",
    "    \n",
    "    # What about the distributions of words per gene and sentences per gene?\n",
    "    dists[\"species\"].extend([species]*subset.shape[0])\n",
    "    dists[\"num_words\"].extend([len(word_tokenize(x)) for x in subset[\"descriptions\"].values])\n",
    "    dists[\"num_sents\"].extend([len(sent_tokenize(x)) for x in subset[\"descriptions\"].values])\n",
    "    \n",
    "# For the entire dataset including all of the species.\n",
    "df = data.to_pandas()\n",
    "subset = df\n",
    "sentences = [sent_tokenize(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_stemmed = [preprocess_string(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"descriptions\"].values]\n",
    "sent_lists[\"total\"] = flatten(sentences)\n",
    "token_lists[\"total\"] = flatten(descriptions_not_stemmed)\n",
    "stems_lists[\"total\"] = flatten(descriptions_stemmed)    \n",
    "lemma_lists[\"total\"] = flatten(descriptions_lemmatized)\n",
    "\n",
    "# What about lemmas that are uniquely used for a particular species?\n",
    "lemma_sets_unique_to_species = {}\n",
    "for species in data.get_species():\n",
    "    other_species = [s for s in data.get_species() if s != species]\n",
    "    lemmas_used_in_other_species = set(flatten([lemma_lists[s] for s in other_species]))\n",
    "    unique_lemmas = set(lemma_lists[species]).difference(lemmas_used_in_other_species)\n",
    "    lemma_sets_unique_to_species[species] = unique_lemmas\n",
    "lemma_sets_unique_to_species[\"total\"] = flatten([list(s) for s in lemma_sets_unique_to_species.values()])\n",
    "\n",
    "    \n",
    "# Create a dataframe to contain the summarizing information about this dataset, and sort it by number of genes.\n",
    "# Unique gene identifiers is just the total number of genes, this column name should be changed in the class...\n",
    "df = data.describe() \n",
    "condition = (df.species==\"total\")\n",
    "excluded = df[condition]\n",
    "included = df[~condition]\n",
    "df_sorted = included.sort_values(by=\"unique_gene_identifiers\", ascending=False)\n",
    "df = pd.concat([df_sorted,excluded])\n",
    "\n",
    "# Add columns summarizing information about the text descriptions in the dataset.\n",
    "df[\"total_sents\"] = df[\"species\"].map(lambda x: len(sent_lists[x]))\n",
    "df[\"total_words\"] = df[\"species\"].map(lambda x: len(token_lists[x]))\n",
    "df[\"unique_words\"] = df[\"species\"].map(lambda x: len(set(token_lists[x])))\n",
    "df[\"unique_stems\"] = df[\"species\"].map(lambda x: len(set(stems_lists[x])))\n",
    "df[\"total_lemmas\"] = df[\"species\"].map(lambda x: len(lemma_lists[x]))\n",
    "df[\"unique_lemmas\"] = df[\"species\"].map(lambda x: len(set(lemma_lists[x])))\n",
    "df[\"unique_lemmas_to_species\"] = df[\"species\"].map(lambda x: len(lemma_sets_unique_to_species[x]))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_words</th>\n",
       "      <th>num_sents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zma</td>\n",
       "      <td>115</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zma</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zma</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zma</td>\n",
       "      <td>33</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>105</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>zma</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>zma</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>zma</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>zma</td>\n",
       "      <td>115</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>zma</td>\n",
       "      <td>90</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>zma</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>zma</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>zma</td>\n",
       "      <td>86</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>zma</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>zma</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>zma</td>\n",
       "      <td>39</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species  num_words  num_sents\n",
       "0      zma         14          1\n",
       "1      zma        115          7\n",
       "2      zma          6          2\n",
       "3      zma         13          3\n",
       "4      zma         33          3\n",
       "5      zma        105          9\n",
       "6      zma        100          7\n",
       "7      zma         43          3\n",
       "8      zma         43          3\n",
       "9      zma        115          7\n",
       "10     zma         90          7\n",
       "11     zma         13          3\n",
       "12     zma         14          1\n",
       "13     zma         13          3\n",
       "14     zma         86         10\n",
       "15     zma         25          5\n",
       "16     zma         14          1\n",
       "17     zma         14          1\n",
       "18     zma         14          1\n",
       "19     zma         39          6"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_distributions = pd.DataFrame(dists)\n",
    "text_distributions.to_csv(os.path.join(OUTPUT_DIR, \"word_sent_distributions.csv\"), index=False)\n",
    "text_distributions.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the ontology term annotations for each species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many of the genes in this dataset for each species are mapped to atleast one term from a given ontology?\n",
    "num_mapped_go = {}\n",
    "num_mapped_po = {}\n",
    "for species in data.get_species():\n",
    "    d = data.to_pandas()\n",
    "    subset = d[d[\"species\"]==species]    \n",
    "    num_mapped_po[species] = len([t for t in subset[\"annotations\"].values if \"PO\" in t])\n",
    "    num_mapped_go[species] = len([t for t in subset[\"annotations\"].values if \"GO\" in t])\n",
    "num_mapped_go[\"total\"] = sum(list(num_mapped_go.values()))   \n",
    "num_mapped_po[\"total\"] = sum(list(num_mapped_po.values()))\n",
    "df[\"go\"] = df[\"species\"].map(lambda x: num_mapped_go[x])\n",
    "df[\"po\"] = df[\"species\"].map(lambda x: num_mapped_po[x])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the biologically relevant groups like biochemical pathways and phenotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "      <td>1084</td>\n",
       "      <td>654</td>\n",
       "      <td>1570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>157</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "      <td>1263</td>\n",
       "      <td>796</td>\n",
       "      <td>1570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220  1084       654          1570\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111   157       133             0\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92     1         4             0\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65    18         3             0\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32     0         2             0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27     3         0             0\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547  1263       796          1570"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the groupings that we're interested in mapping to? Uses the paths defined at the top of the notebook.\n",
    "groupings_dict = {\n",
    "    \"kegg\":(kegg_pathways_path, kegg_pathways_names_path),\n",
    "    \"plantcyc\":(plantcyc_pathways_path, plantcyc_pathways_names_path),\n",
    "    \"lloyd_meinke\":(lloyd_meinke_subsets_path, lloyd_meinke_subsets_names_path)\n",
    "}\n",
    "\n",
    "\n",
    "for name,(filename,mapfile) in groupings_dict.items():\n",
    "    groups = Groupings(filename, {row.group_id:row.group_name for row in pd.read_csv(mapfile).itertuples()})\n",
    "    id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "    group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "    species_dict = data.get_species_dictionary()\n",
    "    num_mapped = {}\n",
    "    for species in data.get_species():\n",
    "        num_mapped[species] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "    num_mapped[\"total\"] = sum(list(num_mapped.values()))    \n",
    "    df[name] = df[\"species\"].map(lambda x: num_mapped[x])  \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the other biologically relevant information like orthologous genes and protein interactions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "      <th>panther</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>5066</td>\n",
       "      <td>2975</td>\n",
       "      <td>26043</td>\n",
       "      <td>232445</td>\n",
       "      <td>7085</td>\n",
       "      <td>5116</td>\n",
       "      <td>232445</td>\n",
       "      <td>6561</td>\n",
       "      <td>4864</td>\n",
       "      <td>4691</td>\n",
       "      <td>3220</td>\n",
       "      <td>1084</td>\n",
       "      <td>654</td>\n",
       "      <td>1570</td>\n",
       "      <td>317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>5475</td>\n",
       "      <td>48983</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>48983</td>\n",
       "      <td>1722</td>\n",
       "      <td>503</td>\n",
       "      <td>184</td>\n",
       "      <td>111</td>\n",
       "      <td>157</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>99</td>\n",
       "      <td>46</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>65</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>126</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>6700</td>\n",
       "      <td>3999</td>\n",
       "      <td>32680</td>\n",
       "      <td>289464</td>\n",
       "      <td>8043</td>\n",
       "      <td>5802</td>\n",
       "      <td>289464</td>\n",
       "      <td>7443</td>\n",
       "      <td>5703</td>\n",
       "      <td>5002</td>\n",
       "      <td>3547</td>\n",
       "      <td>1263</td>\n",
       "      <td>796</td>\n",
       "      <td>1570</td>\n",
       "      <td>858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke  panther\n",
       "0     ath                     5066                 2975        26043       232445          7085          5116        232445           6561                      4864  4691  3220  1084       654          1570      317\n",
       "5     zma                     1405                  810         5475        48983          1846          1317         48983           1722                       503   184   111   157       133             0      443\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        99    46    92     1         4             0       86\n",
       "4     sly                       70                   70          359         1678           577           438          1678            552                        99    23    65    18         3             0       11\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       126    30    32     0         2             0        0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12    28    27     3         0             0        1\n",
       "6   total                     6700                 3999        32680       289464          8043          5802        289464           7443                      5703  5002  3547  1263       796          1570      858"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PantherDB for plant orthologs.\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in ortholog_edgelist.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"panther\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STRING DB for protein-protein interactions.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in string_data.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"stringdb\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write that dataframe with all the information about datast to a file.\n",
    "df.to_csv(os.path.join(OUTPUT_DIR,\"full_dataset_composition.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the vocabularies used for different species compare?\n",
    "One of the things we are interested in is discovering or recovering phenotype similarity between different species in order to identify phenologs (phenotypes between species that share some underlying genetic cause). For this reason, we are interested in how the vocabularies used to describe phenotypes between different species vary, because this will impact how feasible it is to use a dataset like this to identify phenologs. Because the Arabidopsis and maize datasets are the largest in this case, we will compare the vocabularies used in describing the phenotypes associated with the genes from these species in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>twenty</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seedligs</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atpase</td>\n",
       "      <td>NN</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003872</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>polynucleate</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hub</td>\n",
       "      <td>NN</td>\n",
       "      <td>13</td>\n",
       "      <td>0.005593</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>indicate</td>\n",
       "      <td>NN</td>\n",
       "      <td>7</td>\n",
       "      <td>0.003011</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>blade</td>\n",
       "      <td>NN</td>\n",
       "      <td>67</td>\n",
       "      <td>0.028824</td>\n",
       "      <td>114</td>\n",
       "      <td>0.232734</td>\n",
       "      <td>-0.203910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>galactose</td>\n",
       "      <td>NN</td>\n",
       "      <td>10</td>\n",
       "      <td>0.004302</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cacl</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>evaluation</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002042</td>\n",
       "      <td>-0.002042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "0        twenty             NN         1  0.000430         0  0.000000  0.000430\n",
       "1      seedligs             NN         2  0.000860         0  0.000000  0.000860\n",
       "2        atpase             NN         9  0.003872         0  0.000000  0.003872\n",
       "3  polynucleate             NN         2  0.000860         0  0.000000  0.000860\n",
       "4           hub             NN        13  0.005593         0  0.000000  0.005593\n",
       "5      indicate             NN         7  0.003011         0  0.000000  0.003011\n",
       "6         blade             NN        67  0.028824       114  0.232734 -0.203910\n",
       "7     galactose             NN        10  0.004302         0  0.000000  0.004302\n",
       "8          cacl             NN         1  0.000430         0  0.000000  0.000430\n",
       "9    evaluation             NN         0  0.000000         1  0.002042 -0.002042"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using lemmas as the vocabulary components.\n",
    "vocabs = {s:set(lemma_list) for s,lemma_list in lemma_lists.items()}\n",
    "fdist_zma = FreqDist(lemma_lists[\"zma\"])\n",
    "fdist_ath = FreqDist(lemma_lists[\"ath\"])\n",
    "\n",
    "# Using word stems as the vocabulary components.\n",
    "#vocabs = {s:set(stems_list) for s,stems_list in stems_lists.items()}\n",
    "#fdist_zma = FreqDist(stems_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(stems_lists[\"ath\"])\n",
    "\n",
    "# Using tokens (full words) as the vocabulary components.\n",
    "#vocabs = {s:set(token_list) for s,token_list in token_lists.items()}\n",
    "#fdist_zma = FreqDist(token_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(token_lists[\"ath\"])\n",
    "\n",
    "union_vocab = vocabs[\"zma\"].union(vocabs[\"ath\"])\n",
    "table = pd.DataFrame({\"token\":list(union_vocab)})\n",
    "stops = set(stopwords.words('english'))\n",
    "table = table[~table.token.isin(stops)]\n",
    "table[\"part_of_speech\"] = table[\"token\"].map(lambda x: nltk.pos_tag([x])[0][1][:2])\n",
    "table[\"ath_freq\"] = table[\"token\"].map(lambda x: fdist_ath[x])\n",
    "table[\"ath_rate\"] = table[\"ath_freq\"]*100/len(token_lists[\"ath\"])\n",
    "table[\"zma_freq\"] = table[\"token\"].map(lambda x: fdist_zma[x])\n",
    "table[\"zma_rate\"] = table[\"zma_freq\"]*100/len(token_lists[\"zma\"])\n",
    "table[\"diff\"] = table[\"ath_rate\"]-table[\"zma_rate\"]\n",
    "table.to_csv(os.path.join(OUTPUT_DIR,\"token_frequencies.csv\"), index=False)\n",
    "table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5437</th>\n",
       "      <td>embryo</td>\n",
       "      <td>NN</td>\n",
       "      <td>4087</td>\n",
       "      <td>1.758265</td>\n",
       "      <td>150</td>\n",
       "      <td>0.306229</td>\n",
       "      <td>1.452037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882</th>\n",
       "      <td>mutant</td>\n",
       "      <td>NN</td>\n",
       "      <td>4433</td>\n",
       "      <td>1.907118</td>\n",
       "      <td>254</td>\n",
       "      <td>0.518547</td>\n",
       "      <td>1.388571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4767</th>\n",
       "      <td>phenotype</td>\n",
       "      <td>NN</td>\n",
       "      <td>3393</td>\n",
       "      <td>1.459700</td>\n",
       "      <td>76</td>\n",
       "      <td>0.155156</td>\n",
       "      <td>1.304544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5715</th>\n",
       "      <td>type</td>\n",
       "      <td>NN</td>\n",
       "      <td>2364</td>\n",
       "      <td>1.017015</td>\n",
       "      <td>15</td>\n",
       "      <td>0.030623</td>\n",
       "      <td>0.986392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5801</th>\n",
       "      <td>wild</td>\n",
       "      <td>NN</td>\n",
       "      <td>2287</td>\n",
       "      <td>0.983889</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014291</td>\n",
       "      <td>0.969598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>root</td>\n",
       "      <td>NN</td>\n",
       "      <td>2487</td>\n",
       "      <td>1.069931</td>\n",
       "      <td>70</td>\n",
       "      <td>0.142907</td>\n",
       "      <td>0.927024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2612</th>\n",
       "      <td>stage</td>\n",
       "      <td>NN</td>\n",
       "      <td>1748</td>\n",
       "      <td>0.752006</td>\n",
       "      <td>55</td>\n",
       "      <td>0.112284</td>\n",
       "      <td>0.639722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6370</th>\n",
       "      <td>reduced</td>\n",
       "      <td>VB</td>\n",
       "      <td>2416</td>\n",
       "      <td>1.039386</td>\n",
       "      <td>215</td>\n",
       "      <td>0.438928</td>\n",
       "      <td>0.600458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311</th>\n",
       "      <td>terminal</td>\n",
       "      <td>NN</td>\n",
       "      <td>1265</td>\n",
       "      <td>0.544215</td>\n",
       "      <td>4</td>\n",
       "      <td>0.008166</td>\n",
       "      <td>0.536049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2079</th>\n",
       "      <td>growth</td>\n",
       "      <td>NN</td>\n",
       "      <td>1590</td>\n",
       "      <td>0.684033</td>\n",
       "      <td>78</td>\n",
       "      <td>0.159239</td>\n",
       "      <td>0.524794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5571</th>\n",
       "      <td>cotyledon</td>\n",
       "      <td>NN</td>\n",
       "      <td>1148</td>\n",
       "      <td>0.493880</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.493880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3626</th>\n",
       "      <td>seed</td>\n",
       "      <td>NN</td>\n",
       "      <td>1593</td>\n",
       "      <td>0.685323</td>\n",
       "      <td>113</td>\n",
       "      <td>0.230692</td>\n",
       "      <td>0.454631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1610</th>\n",
       "      <td>level</td>\n",
       "      <td>NN</td>\n",
       "      <td>1132</td>\n",
       "      <td>0.486997</td>\n",
       "      <td>20</td>\n",
       "      <td>0.040830</td>\n",
       "      <td>0.446166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3497</th>\n",
       "      <td>increased</td>\n",
       "      <td>VB</td>\n",
       "      <td>1070</td>\n",
       "      <td>0.460324</td>\n",
       "      <td>11</td>\n",
       "      <td>0.022457</td>\n",
       "      <td>0.437867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4451</th>\n",
       "      <td>cell</td>\n",
       "      <td>NN</td>\n",
       "      <td>1516</td>\n",
       "      <td>0.652197</td>\n",
       "      <td>106</td>\n",
       "      <td>0.216402</td>\n",
       "      <td>0.435796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6526</th>\n",
       "      <td>defective</td>\n",
       "      <td>JJ</td>\n",
       "      <td>2337</td>\n",
       "      <td>1.005399</td>\n",
       "      <td>285</td>\n",
       "      <td>0.581835</td>\n",
       "      <td>0.423565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5883</th>\n",
       "      <td>double</td>\n",
       "      <td>RB</td>\n",
       "      <td>881</td>\n",
       "      <td>0.379014</td>\n",
       "      <td>3</td>\n",
       "      <td>0.006125</td>\n",
       "      <td>0.372890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049</th>\n",
       "      <td>wa</td>\n",
       "      <td>NN</td>\n",
       "      <td>841</td>\n",
       "      <td>0.361806</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002042</td>\n",
       "      <td>0.359764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>compared</td>\n",
       "      <td>VB</td>\n",
       "      <td>840</td>\n",
       "      <td>0.361376</td>\n",
       "      <td>9</td>\n",
       "      <td>0.018374</td>\n",
       "      <td>0.343002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5755</th>\n",
       "      <td>rosette</td>\n",
       "      <td>NN</td>\n",
       "      <td>796</td>\n",
       "      <td>0.342447</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002042</td>\n",
       "      <td>0.340405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>flower</td>\n",
       "      <td>NN</td>\n",
       "      <td>799</td>\n",
       "      <td>0.343737</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014291</td>\n",
       "      <td>0.329447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1152</th>\n",
       "      <td>silique</td>\n",
       "      <td>NN</td>\n",
       "      <td>723</td>\n",
       "      <td>0.311041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.311041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5296</th>\n",
       "      <td>visible</td>\n",
       "      <td>JJ</td>\n",
       "      <td>767</td>\n",
       "      <td>0.329971</td>\n",
       "      <td>11</td>\n",
       "      <td>0.022457</td>\n",
       "      <td>0.307514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2188</th>\n",
       "      <td>abnormal</td>\n",
       "      <td>JJ</td>\n",
       "      <td>805</td>\n",
       "      <td>0.346318</td>\n",
       "      <td>26</td>\n",
       "      <td>0.053080</td>\n",
       "      <td>0.293239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3250</th>\n",
       "      <td>development</td>\n",
       "      <td>NN</td>\n",
       "      <td>765</td>\n",
       "      <td>0.329110</td>\n",
       "      <td>19</td>\n",
       "      <td>0.038789</td>\n",
       "      <td>0.290321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6409</th>\n",
       "      <td>globular</td>\n",
       "      <td>NN</td>\n",
       "      <td>672</td>\n",
       "      <td>0.289101</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.289101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3178</th>\n",
       "      <td>sensitive</td>\n",
       "      <td>JJ</td>\n",
       "      <td>689</td>\n",
       "      <td>0.296414</td>\n",
       "      <td>6</td>\n",
       "      <td>0.012249</td>\n",
       "      <td>0.284165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4849</th>\n",
       "      <td>defect</td>\n",
       "      <td>NN</td>\n",
       "      <td>722</td>\n",
       "      <td>0.310611</td>\n",
       "      <td>20</td>\n",
       "      <td>0.040830</td>\n",
       "      <td>0.269781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5486</th>\n",
       "      <td>hypocotyl</td>\n",
       "      <td>NN</td>\n",
       "      <td>607</td>\n",
       "      <td>0.261137</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.261137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>short</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1173</td>\n",
       "      <td>0.504636</td>\n",
       "      <td>120</td>\n",
       "      <td>0.244983</td>\n",
       "      <td>0.259653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "5437       embryo             NN      4087  1.758265       150  0.306229  1.452037\n",
       "2882       mutant             NN      4433  1.907118       254  0.518547  1.388571\n",
       "4767    phenotype             NN      3393  1.459700        76  0.155156  1.304544\n",
       "5715         type             NN      2364  1.017015        15  0.030623  0.986392\n",
       "5801         wild             NN      2287  0.983889         7  0.014291  0.969598\n",
       "795          root             NN      2487  1.069931        70  0.142907  0.927024\n",
       "2612        stage             NN      1748  0.752006        55  0.112284  0.639722\n",
       "6370      reduced             VB      2416  1.039386       215  0.438928  0.600458\n",
       "1311     terminal             NN      1265  0.544215         4  0.008166  0.536049\n",
       "2079       growth             NN      1590  0.684033        78  0.159239  0.524794\n",
       "5571    cotyledon             NN      1148  0.493880         0  0.000000  0.493880\n",
       "3626         seed             NN      1593  0.685323       113  0.230692  0.454631\n",
       "1610        level             NN      1132  0.486997        20  0.040830  0.446166\n",
       "3497    increased             VB      1070  0.460324        11  0.022457  0.437867\n",
       "4451         cell             NN      1516  0.652197       106  0.216402  0.435796\n",
       "6526    defective             JJ      2337  1.005399       285  0.581835  0.423565\n",
       "5883       double             RB       881  0.379014         3  0.006125  0.372890\n",
       "1049           wa             NN       841  0.361806         1  0.002042  0.359764\n",
       "228      compared             VB       840  0.361376         9  0.018374  0.343002\n",
       "5755      rosette             NN       796  0.342447         1  0.002042  0.340405\n",
       "858        flower             NN       799  0.343737         7  0.014291  0.329447\n",
       "1152      silique             NN       723  0.311041         0  0.000000  0.311041\n",
       "5296      visible             JJ       767  0.329971        11  0.022457  0.307514\n",
       "2188     abnormal             JJ       805  0.346318        26  0.053080  0.293239\n",
       "3250  development             NN       765  0.329110        19  0.038789  0.290321\n",
       "6409     globular             NN       672  0.289101         0  0.000000  0.289101\n",
       "3178    sensitive             JJ       689  0.296414         6  0.012249  0.284165\n",
       "4849       defect             NN       722  0.310611        20  0.040830  0.269781\n",
       "5486    hypocotyl             NN       607  0.261137         0  0.000000  0.261137\n",
       "3113        short             JJ      1173  0.504636       120  0.244983  0.259653"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for Arabidopsis than maize descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=False, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4478</th>\n",
       "      <td>endosperm</td>\n",
       "      <td>NN</td>\n",
       "      <td>100</td>\n",
       "      <td>0.043021</td>\n",
       "      <td>1069</td>\n",
       "      <td>2.182390</td>\n",
       "      <td>-2.139369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4888</th>\n",
       "      <td>seedling</td>\n",
       "      <td>VB</td>\n",
       "      <td>1276</td>\n",
       "      <td>0.548947</td>\n",
       "      <td>1316</td>\n",
       "      <td>2.686646</td>\n",
       "      <td>-2.137699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182</th>\n",
       "      <td>kernel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>756</td>\n",
       "      <td>1.543393</td>\n",
       "      <td>-1.543393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3835</th>\n",
       "      <td>leaf</td>\n",
       "      <td>NN</td>\n",
       "      <td>3224</td>\n",
       "      <td>1.386995</td>\n",
       "      <td>1401</td>\n",
       "      <td>2.860176</td>\n",
       "      <td>-1.473181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4981</th>\n",
       "      <td>yellow</td>\n",
       "      <td>NN</td>\n",
       "      <td>239</td>\n",
       "      <td>0.102820</td>\n",
       "      <td>766</td>\n",
       "      <td>1.563808</td>\n",
       "      <td>-1.460988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3900</th>\n",
       "      <td>green</td>\n",
       "      <td>JJ</td>\n",
       "      <td>743</td>\n",
       "      <td>0.319646</td>\n",
       "      <td>778</td>\n",
       "      <td>1.588306</td>\n",
       "      <td>-1.268661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1735</th>\n",
       "      <td>white</td>\n",
       "      <td>JJ</td>\n",
       "      <td>310</td>\n",
       "      <td>0.133365</td>\n",
       "      <td>642</td>\n",
       "      <td>1.310659</td>\n",
       "      <td>-1.177294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6019</th>\n",
       "      <td>albino</td>\n",
       "      <td>NN</td>\n",
       "      <td>182</td>\n",
       "      <td>0.078298</td>\n",
       "      <td>396</td>\n",
       "      <td>0.808444</td>\n",
       "      <td>-0.730146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2680</th>\n",
       "      <td>usually</td>\n",
       "      <td>RB</td>\n",
       "      <td>41</td>\n",
       "      <td>0.017639</td>\n",
       "      <td>342</td>\n",
       "      <td>0.698201</td>\n",
       "      <td>-0.680563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4332</th>\n",
       "      <td>color</td>\n",
       "      <td>NN</td>\n",
       "      <td>40</td>\n",
       "      <td>0.017208</td>\n",
       "      <td>326</td>\n",
       "      <td>0.665537</td>\n",
       "      <td>-0.648329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2565</th>\n",
       "      <td>tassel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>300</td>\n",
       "      <td>0.612457</td>\n",
       "      <td>-0.612457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4851</th>\n",
       "      <td>ear</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>272</td>\n",
       "      <td>0.555295</td>\n",
       "      <td>-0.555295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2990</th>\n",
       "      <td>often</td>\n",
       "      <td>RB</td>\n",
       "      <td>136</td>\n",
       "      <td>0.058508</td>\n",
       "      <td>275</td>\n",
       "      <td>0.561419</td>\n",
       "      <td>-0.502911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1963</th>\n",
       "      <td>aleurone</td>\n",
       "      <td>NN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001291</td>\n",
       "      <td>246</td>\n",
       "      <td>0.502215</td>\n",
       "      <td>-0.500924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3099</th>\n",
       "      <td>opaque</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>244</td>\n",
       "      <td>0.498132</td>\n",
       "      <td>-0.497702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078</th>\n",
       "      <td>class</td>\n",
       "      <td>NN</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003442</td>\n",
       "      <td>240</td>\n",
       "      <td>0.489966</td>\n",
       "      <td>-0.486524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4597</th>\n",
       "      <td>tissue</td>\n",
       "      <td>NN</td>\n",
       "      <td>213</td>\n",
       "      <td>0.091635</td>\n",
       "      <td>253</td>\n",
       "      <td>0.516506</td>\n",
       "      <td>-0.424871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>small</td>\n",
       "      <td>JJ</td>\n",
       "      <td>557</td>\n",
       "      <td>0.239627</td>\n",
       "      <td>321</td>\n",
       "      <td>0.655329</td>\n",
       "      <td>-0.415703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5396</th>\n",
       "      <td>pale</td>\n",
       "      <td>NN</td>\n",
       "      <td>524</td>\n",
       "      <td>0.225430</td>\n",
       "      <td>306</td>\n",
       "      <td>0.624707</td>\n",
       "      <td>-0.399277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5990</th>\n",
       "      <td>vp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190</td>\n",
       "      <td>0.387890</td>\n",
       "      <td>-0.387890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>palegreen</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>171</td>\n",
       "      <td>0.349101</td>\n",
       "      <td>-0.348670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329</th>\n",
       "      <td>general</td>\n",
       "      <td>JJ</td>\n",
       "      <td>13</td>\n",
       "      <td>0.005593</td>\n",
       "      <td>171</td>\n",
       "      <td>0.349101</td>\n",
       "      <td>-0.343508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3941</th>\n",
       "      <td>term</td>\n",
       "      <td>NN</td>\n",
       "      <td>21</td>\n",
       "      <td>0.009034</td>\n",
       "      <td>172</td>\n",
       "      <td>0.351142</td>\n",
       "      <td>-0.342108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343</th>\n",
       "      <td>texture</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>163</td>\n",
       "      <td>0.332769</td>\n",
       "      <td>-0.331908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5443</th>\n",
       "      <td>appearance</td>\n",
       "      <td>NN</td>\n",
       "      <td>121</td>\n",
       "      <td>0.052055</td>\n",
       "      <td>184</td>\n",
       "      <td>0.375641</td>\n",
       "      <td>-0.323585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>surface</td>\n",
       "      <td>NN</td>\n",
       "      <td>131</td>\n",
       "      <td>0.056357</td>\n",
       "      <td>185</td>\n",
       "      <td>0.377682</td>\n",
       "      <td>-0.321325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2530</th>\n",
       "      <td>otherwise</td>\n",
       "      <td>RB</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003872</td>\n",
       "      <td>157</td>\n",
       "      <td>0.320519</td>\n",
       "      <td>-0.316647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>pericarp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>139</td>\n",
       "      <td>0.283772</td>\n",
       "      <td>-0.283772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2608</th>\n",
       "      <td>sometimes</td>\n",
       "      <td>RB</td>\n",
       "      <td>37</td>\n",
       "      <td>0.015918</td>\n",
       "      <td>145</td>\n",
       "      <td>0.296021</td>\n",
       "      <td>-0.280103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698</th>\n",
       "      <td>male</td>\n",
       "      <td>NN</td>\n",
       "      <td>354</td>\n",
       "      <td>0.152294</td>\n",
       "      <td>210</td>\n",
       "      <td>0.428720</td>\n",
       "      <td>-0.276426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "4478   endosperm             NN       100  0.043021      1069  2.182390 -2.139369\n",
       "4888    seedling             VB      1276  0.548947      1316  2.686646 -2.137699\n",
       "6182      kernel             NN         0  0.000000       756  1.543393 -1.543393\n",
       "3835        leaf             NN      3224  1.386995      1401  2.860176 -1.473181\n",
       "4981      yellow             NN       239  0.102820       766  1.563808 -1.460988\n",
       "3900       green             JJ       743  0.319646       778  1.588306 -1.268661\n",
       "1735       white             JJ       310  0.133365       642  1.310659 -1.177294\n",
       "6019      albino             NN       182  0.078298       396  0.808444 -0.730146\n",
       "2680     usually             RB        41  0.017639       342  0.698201 -0.680563\n",
       "4332       color             NN        40  0.017208       326  0.665537 -0.648329\n",
       "2565      tassel             NN         0  0.000000       300  0.612457 -0.612457\n",
       "4851         ear             NN         0  0.000000       272  0.555295 -0.555295\n",
       "2990       often             RB       136  0.058508       275  0.561419 -0.502911\n",
       "1963    aleurone             NN         3  0.001291       246  0.502215 -0.500924\n",
       "3099      opaque             NN         1  0.000430       244  0.498132 -0.497702\n",
       "1078       class             NN         8  0.003442       240  0.489966 -0.486524\n",
       "4597      tissue             NN       213  0.091635       253  0.516506 -0.424871\n",
       "628        small             JJ       557  0.239627       321  0.655329 -0.415703\n",
       "5396        pale             NN       524  0.225430       306  0.624707 -0.399277\n",
       "5990          vp             NN         0  0.000000       190  0.387890 -0.387890\n",
       "581    palegreen             NN         1  0.000430       171  0.349101 -0.348670\n",
       "1329     general             JJ        13  0.005593       171  0.349101 -0.343508\n",
       "3941        term             NN        21  0.009034       172  0.351142 -0.342108\n",
       "2343     texture             NN         2  0.000860       163  0.332769 -0.331908\n",
       "5443  appearance             NN       121  0.052055       184  0.375641 -0.323585\n",
       "1375     surface             NN       131  0.056357       185  0.377682 -0.321325\n",
       "2530   otherwise             RB         9  0.003872       157  0.320519 -0.316647\n",
       "793     pericarp             NN         0  0.000000       139  0.283772 -0.283772\n",
       "2608   sometimes             RB        37  0.015918       145  0.296021 -0.280103\n",
       "1698        male             NN       354  0.152294       210  0.428720 -0.276426"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for maize than Arabidopsis descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=True, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MD</td>\n",
       "      <td>0.059567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD</td>\n",
       "      <td>0.029173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IN</td>\n",
       "      <td>0.026193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JJ</td>\n",
       "      <td>0.023080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DT</td>\n",
       "      <td>0.019441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RB</td>\n",
       "      <td>0.016217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NN</td>\n",
       "      <td>0.014511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VB</td>\n",
       "      <td>0.012129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC</td>\n",
       "      <td>0.007736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WP</td>\n",
       "      <td>0.001611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  part_of_speech  abs_diff\n",
       "0             MD  0.059567\n",
       "1             CD  0.029173\n",
       "2             IN  0.026193\n",
       "3             JJ  0.023080\n",
       "4             DT  0.019441\n",
       "5             RB  0.016217\n",
       "6             NN  0.014511\n",
       "7             VB  0.012129\n",
       "8             CC  0.007736\n",
       "9             WP  0.001611"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is the mean absolute value of the rate differences different between the different parts of speech?\n",
    "table[\"abs_diff\"] = abs(table[\"diff\"])\n",
    "pos_table = table.groupby(\"part_of_speech\").mean()\n",
    "pos_table.sort_values(by=\"abs_diff\", inplace=True, ascending=False)\n",
    "pos_table = pos_table[[\"abs_diff\"]]\n",
    "pos_table.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Working on the Venn Diagram for this part, unused currently.\n",
    "#print(table.shape)\n",
    "#zma_only = table[table[\"ath_rate\"]==0]\n",
    "#ath_only = table[table[\"zma_rate\"]==0]\n",
    "#print(zma_only.shape)\n",
    "#print(ath_only.shape)\n",
    "#print(ath_only.shape[0]+zma_only.shape[0])\n",
    "#ath_only.head(10)\n",
    "# We need to create a mapping between stems and the words that were present for them.\n",
    "# This is because what we want is the stems that are exclusive to a species.\n",
    "# but then the words that are actually there for those stems, so that we can count their parts of speech."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at Term and Word Enrichment for Groups of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3114"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the dataset of phenotype descriptions and ontology annotations.\n",
    "data = plant_dataset\n",
    "data.filter_has_description()\n",
    "data.filter_has_annotation(\"GO\")\n",
    "data.filter_has_annotation(\"PO\")\n",
    "d = data.get_description_dictionary()\n",
    "texts = {i:\" \".join(simple_preprocess(t)) for i,t in d.items()}\n",
    "len(texts)                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create ontology objects for all the biological ontologies being used.\n",
    "go_pickle_path = \"../ontologies/go.pickle\"                                                                \n",
    "po_pickle_path = \"../ontologies/po.pickle\"                                                             \n",
    "pato_pickle_path = \"../ontologies/pato.pickle\"\n",
    "pato = load_from_pickle(pato_pickle_path)\n",
    "po = load_from_pickle(po_pickle_path)\n",
    "go = load_from_pickle(go_pickle_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "curated_go_annotations = data.get_annotations_dictionary(\"GO\")\n",
    "curated_po_annotations = data.get_annotations_dictionary(\"PO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>num_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>suberin monomers biosynthesis</td>\n",
       "      <td>PWY-1121</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gluconeogenesis I</td>\n",
       "      <td>GLUCONEO-PWY</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gluconeogenesis III</td>\n",
       "      <td>PWY66-399</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>indole-3-acetate biosynthesis II</td>\n",
       "      <td>PWY-581</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phosphatidylcholine acyl editing</td>\n",
       "      <td>PWY-6803</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sporopollenin precursors biosynthesis</td>\n",
       "      <td>PWY-6733</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>glycolysis I (from glucose 6-phosphate)</td>\n",
       "      <td>GLYCOLYSIS</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3-phosphoinositide biosynthesis</td>\n",
       "      <td>PWY-6352</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>palmitate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5971</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>glycolysis II (from fructose 6-phosphate)</td>\n",
       "      <td>PWY-5484</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>phenylpropanoid biosynthesis</td>\n",
       "      <td>PWY-361</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>oleate biosynthesis I (plants)</td>\n",
       "      <td>PWY-5147</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>stearate biosynthesis II (bacteria and plants)</td>\n",
       "      <td>PWY-5989</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>glycolysis IV (plant cytosol)</td>\n",
       "      <td>PWY-1042</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>photosynthesis light reactions</td>\n",
       "      <td>PWY-101</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       pathway_name    pathway_id  num_genes\n",
       "0                     suberin monomers biosynthesis      PWY-1121         19\n",
       "1                                 gluconeogenesis I  GLUCONEO-PWY         19\n",
       "2                               gluconeogenesis III     PWY66-399         19\n",
       "3                  indole-3-acetate biosynthesis II       PWY-581         18\n",
       "4                  phosphatidylcholine acyl editing      PWY-6803         17\n",
       "5             sporopollenin precursors biosynthesis      PWY-6733         16\n",
       "6           glycolysis I (from glucose 6-phosphate)    GLYCOLYSIS         16\n",
       "7                   3-phosphoinositide biosynthesis      PWY-6352         15\n",
       "8   palmitate biosynthesis II (bacteria and plants)      PWY-5971         15\n",
       "9         glycolysis II (from fructose 6-phosphate)      PWY-5484         15\n",
       "10                     phenylpropanoid biosynthesis       PWY-361         15\n",
       "11                   oleate biosynthesis I (plants)      PWY-5147         14\n",
       "12   stearate biosynthesis II (bacteria and plants)      PWY-5989         14\n",
       "13                    glycolysis IV (plant cytosol)      PWY-1042         14\n",
       "14                   photosynthesis light reactions       PWY-101         14"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the mappings from this dataset to PlantCyc information.\n",
    "#pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "#groups = load_from_pickle(pmn_pathways_filename)\n",
    "#id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "\n",
    "\n",
    "# Reading in the dataset of groupings for pathways in PlantCyc.\n",
    "plantcyc_name_mapping = {row.group_id:row.group_name for row in pd.read_csv(plantcyc_pathways_names_path).itertuples()}\n",
    "plantcyc_grouping = Groupings(plantcyc_pathways_path, plantcyc_name_mapping)\n",
    "id_to_group_ids, group_id_to_ids = plantcyc_grouping.get_groupings_for_dataset(data)\n",
    "\n",
    "# Look at which pathways are best represented in this dataset.\n",
    "pathways_sorted = sorted(group_id_to_ids.items(), key=lambda item: len(item[1]), reverse=True)\n",
    "pathways_sorted_lengths = [(i,len(l)) for (i,l) in pathways_sorted]\n",
    "pathways_df = pd.DataFrame(pathways_sorted_lengths, columns=[\"pathway_id\",\"num_genes\"])\n",
    "pathways_df[\"pathway_name\"] = pathways_df[\"pathway_id\"].map(lambda x: plantcyc_grouping.get_long_name(x))\n",
    "pathways_df = pathways_df[[\"pathway_name\",\"pathway_id\",\"num_genes\"]]\n",
    "pathways_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For some example pathway to use.\n",
    "pathway_id = \"PWY-6733\"\n",
    "gene_ids_in_this_pathway = group_id_to_ids[pathway_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PO:0009071</td>\n",
       "      <td>anther wall tapetum</td>\n",
       "      <td>29</td>\n",
       "      <td>3085</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000497</td>\n",
       "      <td>0.009938</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PO:0009067</td>\n",
       "      <td>filament</td>\n",
       "      <td>52</td>\n",
       "      <td>3062</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.048914</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PO:0009073</td>\n",
       "      <td>stigma</td>\n",
       "      <td>47</td>\n",
       "      <td>3067</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.001856</td>\n",
       "      <td>0.037119</td>\n",
       "      <td>True</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PO:0025314</td>\n",
       "      <td>microsporangium tapetum</td>\n",
       "      <td>29</td>\n",
       "      <td>3085</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000497</td>\n",
       "      <td>0.009938</td>\n",
       "      <td>True</td>\n",
       "      <td>3.633786</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PO:0025313</td>\n",
       "      <td>tapetum</td>\n",
       "      <td>32</td>\n",
       "      <td>3082</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000487</td>\n",
       "      <td>True</td>\n",
       "      <td>2.450679</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PO:0005052</td>\n",
       "      <td>plant callus</td>\n",
       "      <td>141</td>\n",
       "      <td>2973</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>0.011869</td>\n",
       "      <td>True</td>\n",
       "      <td>2.450679</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PO:0025059</td>\n",
       "      <td>portion of ground tissue</td>\n",
       "      <td>156</td>\n",
       "      <td>2958</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.018618</td>\n",
       "      <td>True</td>\n",
       "      <td>0.748851</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PO:0020008</td>\n",
       "      <td>exine</td>\n",
       "      <td>7</td>\n",
       "      <td>3107</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      term_id                term_label  genes_with  genes_without  group_genes_with  group_genes_without   p_value  p_value_adj  significant  info_content significance\n",
       "2  PO:0009071       anther wall tapetum          29           3085                 3                   13  0.000497     0.009938         True      5.000000           **\n",
       "7  PO:0009067                  filament          52           3062                 3                   13  0.002446     0.048914         True      5.000000            *\n",
       "6  PO:0009073                    stigma          47           3067                 3                   13  0.001856     0.037119         True      4.000000            *\n",
       "3  PO:0025314   microsporangium tapetum          29           3085                 3                   13  0.000497     0.009938         True      3.633786           **\n",
       "1  PO:0025313                   tapetum          32           3082                 4                   12  0.000024     0.000487         True      2.450679          ***\n",
       "4  PO:0005052              plant callus         141           2973                 5                   11  0.000593     0.011869         True      2.450679            *\n",
       "5  PO:0025059  portion of ground tissue         156           2958                 5                   11  0.000931     0.018618         True      0.748851            *\n",
       "0  PO:0020008                     exine           7           3107                 3                   13  0.000013     0.000258         True      0.000000          ***"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_po_annotations, gene_ids_in_this_pathway, po).head(20)\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: po.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# ns   P > 0.05\n",
    "# *    P  0.05\n",
    "# **   P  0.01\n",
    "# ***  P  0.001\n",
    "# **** P  0.0001\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GO:0080110</td>\n",
       "      <td>sporopollenin biosynthetic process</td>\n",
       "      <td>4</td>\n",
       "      <td>3110</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>3.152767e-08</td>\n",
       "      <td>6.305535e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0004467</td>\n",
       "      <td>long-chain fatty acid-CoA ligase activity</td>\n",
       "      <td>7</td>\n",
       "      <td>3107</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>6.650099e-14</td>\n",
       "      <td>1.330020e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>4.356238</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>133</td>\n",
       "      <td>2981</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1.321393e-07</td>\n",
       "      <td>2.642786e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>4.167950</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0015645</td>\n",
       "      <td>fatty acid ligase activity</td>\n",
       "      <td>8</td>\n",
       "      <td>3106</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>1.243744e-13</td>\n",
       "      <td>2.487488e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>3.846069</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GO:1901570</td>\n",
       "      <td>fatty acid derivative biosynthetic process</td>\n",
       "      <td>9</td>\n",
       "      <td>3105</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>3.171141e-07</td>\n",
       "      <td>6.342281e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>2.775594</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0032787</td>\n",
       "      <td>monocarboxylic acid metabolic process</td>\n",
       "      <td>119</td>\n",
       "      <td>2995</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>1.527934e-12</td>\n",
       "      <td>3.055868e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>2.591912</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0016405</td>\n",
       "      <td>CoA-ligase activity</td>\n",
       "      <td>16</td>\n",
       "      <td>3098</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1.687120e-18</td>\n",
       "      <td>3.374240e-17</td>\n",
       "      <td>True</td>\n",
       "      <td>2.511272</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0016878</td>\n",
       "      <td>acid-thiol ligase activity</td>\n",
       "      <td>16</td>\n",
       "      <td>3098</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1.687120e-18</td>\n",
       "      <td>3.374240e-17</td>\n",
       "      <td>True</td>\n",
       "      <td>2.406091</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0006631</td>\n",
       "      <td>fatty acid metabolic process</td>\n",
       "      <td>67</td>\n",
       "      <td>3047</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>4.423822e-15</td>\n",
       "      <td>8.847644e-14</td>\n",
       "      <td>True</td>\n",
       "      <td>2.240498</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0019752</td>\n",
       "      <td>carboxylic acid metabolic process</td>\n",
       "      <td>202</td>\n",
       "      <td>2912</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>1.107028e-11</td>\n",
       "      <td>2.214055e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>1.778929</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GO:1901568</td>\n",
       "      <td>fatty acid derivative metabolic process</td>\n",
       "      <td>9</td>\n",
       "      <td>3105</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>3.171141e-07</td>\n",
       "      <td>6.342281e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>1.762164</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0016877</td>\n",
       "      <td>ligase activity, forming carbon-sulfur bonds</td>\n",
       "      <td>21</td>\n",
       "      <td>3093</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1.396386e-17</td>\n",
       "      <td>2.792773e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>1.743159</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0043436</td>\n",
       "      <td>oxoacid metabolic process</td>\n",
       "      <td>223</td>\n",
       "      <td>2891</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>3.411042e-11</td>\n",
       "      <td>6.822085e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>1.407934</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0044255</td>\n",
       "      <td>cellular lipid metabolic process</td>\n",
       "      <td>156</td>\n",
       "      <td>2958</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>2.504877e-11</td>\n",
       "      <td>5.009754e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>1.231455</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GO:0006629</td>\n",
       "      <td>lipid metabolic process</td>\n",
       "      <td>176</td>\n",
       "      <td>2938</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>8.743582e-11</td>\n",
       "      <td>1.748716e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>1.185247</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GO:0006082</td>\n",
       "      <td>organic acid metabolic process</td>\n",
       "      <td>224</td>\n",
       "      <td>2890</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>3.589226e-11</td>\n",
       "      <td>7.178453e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>1.045243</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0016874</td>\n",
       "      <td>ligase activity</td>\n",
       "      <td>41</td>\n",
       "      <td>3073</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>3.883425e-15</td>\n",
       "      <td>7.766849e-14</td>\n",
       "      <td>True</td>\n",
       "      <td>0.953336</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GO:0044281</td>\n",
       "      <td>small molecule metabolic process</td>\n",
       "      <td>296</td>\n",
       "      <td>2818</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>8.547271e-10</td>\n",
       "      <td>1.709454e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.559443</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GO:0044237</td>\n",
       "      <td>cellular metabolic process</td>\n",
       "      <td>989</td>\n",
       "      <td>2125</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>4.096988e-07</td>\n",
       "      <td>8.193976e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.395226</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:0071704</td>\n",
       "      <td>organic substance metabolic process</td>\n",
       "      <td>962</td>\n",
       "      <td>2152</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>2.745401e-07</td>\n",
       "      <td>5.490801e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.383907</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       term_id                                    term_label  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant  info_content significance\n",
       "14  GO:0080110            sporopollenin biosynthetic process           4           3110                 4                   12  3.152767e-08  6.305535e-07         True      5.000000         ****\n",
       "5   GO:0004467     long-chain fatty acid-CoA ligase activity           7           3107                 7                    9  6.650099e-14  1.330020e-12         True      4.356238         ****\n",
       "15  GO:0005783                         endoplasmic reticulum         133           2981                 8                    8  1.321393e-07  2.642786e-06         True      4.167950         ****\n",
       "6   GO:0015645                    fatty acid ligase activity           8           3106                 7                    9  1.243744e-13  2.487488e-12         True      3.846069         ****\n",
       "18  GO:1901570    fatty acid derivative biosynthetic process           9           3105                 4                   12  3.171141e-07  6.342281e-06         True      2.775594         ****\n",
       "7   GO:0032787         monocarboxylic acid metabolic process         119           2995                11                    5  1.527934e-12  3.055868e-11         True      2.591912         ****\n",
       "0   GO:0016405                           CoA-ligase activity          16           3098                10                    6  1.687120e-18  3.374240e-17         True      2.511272         ****\n",
       "1   GO:0016878                    acid-thiol ligase activity          16           3098                10                    6  1.687120e-18  3.374240e-17         True      2.406091         ****\n",
       "4   GO:0006631                  fatty acid metabolic process          67           3047                11                    5  4.423822e-15  8.847644e-14         True      2.240498         ****\n",
       "8   GO:0019752             carboxylic acid metabolic process         202           2912                12                    4  1.107028e-11  2.214055e-10         True      1.778929         ****\n",
       "17  GO:1901568       fatty acid derivative metabolic process           9           3105                 4                   12  3.171141e-07  6.342281e-06         True      1.762164         ****\n",
       "2   GO:0016877  ligase activity, forming carbon-sulfur bonds          21           3093                10                    6  1.396386e-17  2.792773e-16         True      1.743159         ****\n",
       "10  GO:0043436                     oxoacid metabolic process         223           2891                12                    4  3.411042e-11  6.822085e-10         True      1.407934         ****\n",
       "9   GO:0044255              cellular lipid metabolic process         156           2958                11                    5  2.504877e-11  5.009754e-10         True      1.231455         ****\n",
       "12  GO:0006629                       lipid metabolic process         176           2938                11                    5  8.743582e-11  1.748716e-09         True      1.185247         ****\n",
       "11  GO:0006082                organic acid metabolic process         224           2890                12                    4  3.589226e-11  7.178453e-10         True      1.045243         ****\n",
       "3   GO:0016874                               ligase activity          41           3073                10                    6  3.883425e-15  7.766849e-14         True      0.953336         ****\n",
       "13  GO:0044281              small molecule metabolic process         296           2818                12                    4  8.547271e-10  1.709454e-08         True      0.559443         ****\n",
       "19  GO:0044237                    cellular metabolic process         989           2125                15                    1  4.096988e-07  8.193976e-06         True      0.395226         ****\n",
       "16  GO:0071704           organic substance metabolic process         962           2152                15                    1  2.745401e-07  5.490801e-06         True      0.383907         ****"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_go_annotations, gene_ids_in_this_pathway, go).head(20)\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "\n",
    "\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: go.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>layer</td>\n",
       "      <td>56</td>\n",
       "      <td>3058</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>9.464101e-09</td>\n",
       "      <td>1.892820e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cuticle</td>\n",
       "      <td>10</td>\n",
       "      <td>3104</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>4.425946e-07</td>\n",
       "      <td>8.851892e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exine</td>\n",
       "      <td>17</td>\n",
       "      <td>3097</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>2.589817e-06</td>\n",
       "      <td>5.179633e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>permeability</td>\n",
       "      <td>7</td>\n",
       "      <td>3107</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>1.287683e-05</td>\n",
       "      <td>2.575365e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>visible</td>\n",
       "      <td>589</td>\n",
       "      <td>2525</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>2.017943e-05</td>\n",
       "      <td>4.035885e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>humidity</td>\n",
       "      <td>14</td>\n",
       "      <td>3100</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>7.139017e-05</td>\n",
       "      <td>1.427803e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>phenotype</td>\n",
       "      <td>1017</td>\n",
       "      <td>2097</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>9.587024e-05</td>\n",
       "      <td>1.917405e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>lacs</td>\n",
       "      <td>2</td>\n",
       "      <td>3112</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>1.461562e-04</td>\n",
       "      <td>2.923124e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>extracellular</td>\n",
       "      <td>4</td>\n",
       "      <td>3110</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>3.632139e-04</td>\n",
       "      <td>7.264279e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cutin</td>\n",
       "      <td>4</td>\n",
       "      <td>3110</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>3.632139e-04</td>\n",
       "      <td>7.264279e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>wax</td>\n",
       "      <td>26</td>\n",
       "      <td>3088</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>3.694953e-04</td>\n",
       "      <td>7.389906e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>derivatives</td>\n",
       "      <td>5</td>\n",
       "      <td>3109</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>5.069831e-04</td>\n",
       "      <td>1.013966e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>glossy</td>\n",
       "      <td>6</td>\n",
       "      <td>3108</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>6.739618e-04</td>\n",
       "      <td>1.347924e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>no</td>\n",
       "      <td>912</td>\n",
       "      <td>2202</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>1.301487e-03</td>\n",
       "      <td>2.602974e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mainly</td>\n",
       "      <td>9</td>\n",
       "      <td>3105</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>1.312050e-03</td>\n",
       "      <td>2.624100e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>angustifolia</td>\n",
       "      <td>11</td>\n",
       "      <td>3103</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>1.849652e-03</td>\n",
       "      <td>3.699304e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>inoculation</td>\n",
       "      <td>11</td>\n",
       "      <td>3103</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>1.849652e-03</td>\n",
       "      <td>3.699304e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cuticular</td>\n",
       "      <td>11</td>\n",
       "      <td>3103</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>1.849652e-03</td>\n",
       "      <td>3.699304e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>epicuticular</td>\n",
       "      <td>11</td>\n",
       "      <td>3103</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>1.849652e-03</td>\n",
       "      <td>3.699304e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>sites</td>\n",
       "      <td>13</td>\n",
       "      <td>3101</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>2.475103e-03</td>\n",
       "      <td>4.950206e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant significance\n",
       "0           layer          56           3058                 7                    9  9.464101e-09  1.892820e-07         True         ****\n",
       "1         cuticle          10           3104                 4                   12  4.425946e-07  8.851892e-06         True         ****\n",
       "2           exine          17           3097                 4                   12  2.589817e-06  5.179633e-05         True         ****\n",
       "3    permeability           7           3107                 3                   13  1.287683e-05  2.575365e-04         True          ***\n",
       "4         visible         589           2525                11                    5  2.017943e-05  4.035885e-04         True          ***\n",
       "5        humidity          14           3100                 3                   13  7.139017e-05  1.427803e-03         True           **\n",
       "6       phenotype        1017           2097                13                    3  9.587024e-05  1.917405e-03         True           **\n",
       "7            lacs           2           3112                 2                   14  1.461562e-04  2.923124e-03         True           **\n",
       "8   extracellular           4           3110                 2                   14  3.632139e-04  7.264279e-03         True           **\n",
       "9           cutin           4           3110                 2                   14  3.632139e-04  7.264279e-03         True           **\n",
       "10            wax          26           3088                 3                   13  3.694953e-04  7.389906e-03         True           **\n",
       "11    derivatives           5           3109                 2                   14  5.069831e-04  1.013966e-02         True            *\n",
       "12         glossy           6           3108                 2                   14  6.739618e-04  1.347924e-02         True            *\n",
       "13             no         912           2202                11                    5  1.301487e-03  2.602974e-02         True            *\n",
       "14         mainly           9           3105                 2                   14  1.312050e-03  2.624100e-02         True            *\n",
       "15   angustifolia          11           3103                 2                   14  1.849652e-03  3.699304e-02         True            *\n",
       "16    inoculation          11           3103                 2                   14  1.849652e-03  3.699304e-02         True            *\n",
       "17      cuticular          11           3103                 2                   14  1.849652e-03  3.699304e-02         True            *\n",
       "18   epicuticular          11           3103                 2                   14  1.849652e-03  3.699304e-02         True            *\n",
       "19          sites          13           3101                 2                   14  2.475103e-03  4.950206e-02         True            *"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = token_enrichment(texts, gene_ids_in_this_pathway).head(20)\n",
    "\n",
    "\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "\n",
    "# This lambda won't work if passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
