{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from itertools import combinations\n",
    "import random\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASETS_DIR = \"../data/pickles/\"\n",
    "INPUT_DATASET_PATH = \"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\"\n",
    "OUTPUT_TABLE = defaultdict(dict)\n",
    "OUTPUT_PAIR_TUPLES = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes share a phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       2868                 2365\n",
      "1   total       2868                 2365\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "# Find and save the dataset that is relevant for predicting whether two genes share a phenotype\n",
    "lloyd_subsets_filename = \"../data/pickles/groupings_from_lloyd_subsets.pickle\"  \n",
    "groups = load_from_pickle(lloyd_subsets_filename)\n",
    "id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "data.filter_with_ids(group_mapped_ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_phenotypes.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "shared = lambda x,y: len(set(id_to_group_ids[x]).intersection(set(id_to_group_ids[y])))>0\n",
    "pair_list = [shared(x,y) for x,y in list(combinations(group_mapped_ids, 2))]\n",
    "OUTPUT_TABLE[\"phenotype\"][\"relevant_genes\"] = len(group_mapped_ids)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"positive\"] = pair_list.count(True)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"negative\"] = pair_list.count(False)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the pairs of phenotype descriptions to be used for curation.\n",
    "pos_pair_list = [(x,y) for x,y in list(combinations(group_mapped_ids, 2)) if shared(x,y)]\n",
    "sampling = random.choices(pos_pair_list, k=50)\n",
    "to_desc = data.get_description_dictionary()\n",
    "for (x,y) in sampling:\n",
    "    OUTPUT_PAIR_TUPLES.append((to_desc[x],to_desc[y]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes share a biochemical pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       1769                 1469\n",
      "1     gmx          1                    1\n",
      "2     mtr          2                    2\n",
      "3     osa          3                    3\n",
      "4     sly         18                   18\n",
      "5     zma        185                  160\n",
      "6   total       1978                 1653\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "# Using the KEGG pathways from the saved groupings object.\n",
    "kegg_pathways_filename = \"../data/pickles/groupings_from_kegg_pathways.pickle\"                         \n",
    "kegg_groups = load_from_pickle(kegg_pathways_filename)\n",
    "id_to_kegg_group_ids, kegg_group_id_to_ids = kegg_groups.get_groupings_for_dataset(data)\n",
    "# Using the PlantCyc pathways from the saved groupings object.\n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"  \n",
    "pmn_groups = load_from_pickle(pmn_pathways_filename)\n",
    "id_to_pmn_group_ids, pmn_group_id_to_ids = pmn_groups.get_groupings_for_dataset(data)\n",
    "# Find and save the dataset that is relevant for predicting whether two genes share a phenotype\n",
    "group_mapped_ids = []\n",
    "group_mapped_ids.extend([k for (k,v) in id_to_kegg_group_ids.items() if len(v)>0])\n",
    "group_mapped_ids.extend([k for (k,v) in id_to_pmn_group_ids.items() if len(v)>0])\n",
    "group_mapped_ids = list(set(group_mapped_ids))\n",
    "data.filter_with_ids(group_mapped_ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_pathways.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "shared_kegg = lambda x,y: len(set(id_to_kegg_group_ids[x]).intersection(set(id_to_kegg_group_ids[y])))>0\n",
    "shared_pmn = lambda x,y: len(set(id_to_pmn_group_ids[x]).intersection(set(id_to_pmn_group_ids[y])))>0\n",
    "pair_list = [(shared_kegg(x,y) or shared_pmn(x,y)) for x,y in list(combinations(group_mapped_ids, 2))]\n",
    "OUTPUT_TABLE[\"pathways\"][\"relevant_genes\"] = len(group_mapped_ids)\n",
    "OUTPUT_TABLE[\"pathways\"][\"positive\"] = pair_list.count(True)\n",
    "OUTPUT_TABLE[\"pathways\"][\"negative\"] = pair_list.count(False)\n",
    "OUTPUT_TABLE[\"pathways\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the pairs of phenotype descriptions to be used for curation.\n",
    "pos_pair_list = [(x,y) for x,y in list(combinations(group_mapped_ids, 2)) if shared(x,y)]\n",
    "sampling = random.choices(pos_pair_list, k=50)\n",
    "to_desc = data.get_description_dictionary()\n",
    "for (x,y) in sampling:\n",
    "    OUTPUT_PAIR_TUPLES.append((to_desc[x],to_desc[y]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether a protein-protein interaction exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       3674                 2950\n",
      "1     mtr         10                   10\n",
      "2     osa         44                   42\n",
      "3     sly          3                    3\n",
      "4     zma        160                  147\n",
      "5   total       3891                 3152\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "\n",
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "data.filter_with_ids(string_data.ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_interactions.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the ProteinInteraction class guarantees the dataframe contains the reverse of each edge as well.\n",
    "#interaction = lambda x,y: ((string_data.df[\"from\"] == x) & (string_data.df[\"to\"] == y)).any()\n",
    "#interaction_strings = set([\"{}:{}\".format(int(i),int(j)) for i,j in zip(string_data.df[\"from\"].values, string_data.df[\"to\"].values)])\n",
    "#has_interaction = lambda x,y: \"{}:{}\".format(int(x),int(y)) in interaction_strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "num_ids = len(data.get_ids())\n",
    "num_total_pairs = ((num_ids**2)/2)-num_ids\n",
    "num_positive = string_data.df.shape[0]/2\n",
    "num_negative = num_total_pairs - num_positive\n",
    "OUTPUT_TABLE[\"interactions\"][\"relevant_genes\"] = num_ids\n",
    "OUTPUT_TABLE[\"interactions\"][\"positive\"] = num_positive\n",
    "OUTPUT_TABLE[\"interactions\"][\"negative\"] = num_negative\n",
    "OUTPUT_TABLE[\"interactions\"][\"class_ratio\"] = \"{:0.4f}\".format(num_positive/num_negative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes are orthologous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath        350                  303\n",
      "1     osa         86                   80\n",
      "2     sly          7                    7\n",
      "3     zma        443                  337\n",
      "4   total        886                  727\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "\n",
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\"\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "data.filter_with_ids(ortholog_edgelist.ids)\n",
    "print(\"\\n\", data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_orthologs.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the AnyInteraction class guarantees the dataframe contains the reverse of each edge as well.\n",
    "#ortholog_strings = set([\"{}:{}\".format(int(i),int(j)) for i,j in zip(ortholog_edgelist.df[\"from\"].values, ortholog_edgelist.df[\"to\"].values)])\n",
    "#are_orthologs = lambda x,y: \"{}:{}\".format(int(x),int(y)) in ortholog_strings\n",
    "#pair_list = [are_orthologs(x,y) for x,y in list(combinations(data.get_ids(), 2))]\n",
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"positive\"] = pair_list.count(True)\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"negative\"] = pair_list.count(False)\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes, without enumerating.\n",
    "num_ids = len(data.get_ids())\n",
    "num_total_pairs = ((num_ids**2)/2)-num_ids\n",
    "num_positive = ortholog_edgelist.df.shape[0]/2\n",
    "num_negative = num_total_pairs - num_positive\n",
    "OUTPUT_TABLE[\"orthologs\"][\"relevant_genes\"] = num_ids\n",
    "OUTPUT_TABLE[\"orthologs\"][\"positive\"] = num_positive\n",
    "OUTPUT_TABLE[\"orthologs\"][\"negative\"] = num_negative\n",
    "OUTPUT_TABLE[\"orthologs\"][\"class_ratio\"] = \"{:0.4f}\".format(num_positive/num_negative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarzing table of the classes for each prediction problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "      <th>class_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>phenotype</th>\n",
       "      <td>458461</td>\n",
       "      <td>3652817</td>\n",
       "      <td>0.1255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathways</th>\n",
       "      <td>52530</td>\n",
       "      <td>1902723</td>\n",
       "      <td>0.0276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>interactions</th>\n",
       "      <td>450298</td>\n",
       "      <td>7.11575e+06</td>\n",
       "      <td>0.0633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orthologs</th>\n",
       "      <td>77</td>\n",
       "      <td>391535</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orthologs2</th>\n",
       "      <td>77</td>\n",
       "      <td>391535</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             positive     negative class_ratio\n",
       "phenotype      458461      3652817      0.1255\n",
       "pathways        52530      1902723      0.0276\n",
       "interactions   450298  7.11575e+06      0.0633\n",
       "orthologs          77       391535      0.0002\n",
       "orthologs2         77       391535      0.0002"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(OUTPUT_TABLE).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phenotype 1</th>\n",
       "      <th>Phenotype 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>long hypocotyl under far-red light; partially ...</td>\n",
       "      <td>sensitive to uv-b light</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>embryo defective; seedling defective</td>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>increased lateral root number; late flowering</td>\n",
       "      <td>insensitive to cytokinin; short roots</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo defective-transition; embryo defective;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>abnormal microtubule development; abnormal pav...</td>\n",
       "      <td>albino seedlings; pale green seedlings</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>albino; dwarf</td>\n",
       "      <td>homozygous plants have small and hyponastic co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pale green leaves; reduced rosette size</td>\n",
       "      <td>blunt siliques; short inflorescence stems</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pale green seedlings; pale green seeds</td>\n",
       "      <td>anthocyanin accumulation; dark-grown seedlings...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>embryo defective; seedling defective</td>\n",
       "      <td>abnormal rosette leaf morphology; decreased le...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>large rosette leaves</td>\n",
       "      <td>abnormal leaf morphology; dwarf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chlorotic leaves; dwarf</td>\n",
       "      <td>reduced lateral root formation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>abnormal cotyledon morphology; collapsed root tip</td>\n",
       "      <td>abnormal chloroplast morphology; albino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dwarf; fragile inflorescence stems</td>\n",
       "      <td>altered response to red:far-red light; decreas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>anther defects; dwarf</td>\n",
       "      <td>fewer lateral roots; semi-dwarf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>curled leaves; increased rosette leaf number</td>\n",
       "      <td>abnormal flower morphology; blister-like outgr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>dark green; dwarf</td>\n",
       "      <td>elongated hypocotyl in far-red light</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>resistant to anthranilate analogs (herbicide);...</td>\n",
       "      <td>dwarf; pale green rosette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>elevated cobalt levels in shoots; sensitive to...</td>\n",
       "      <td>this mutant has about 1/10 of the ornithine am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>dwarf; sensitive to oxidative stress</td>\n",
       "      <td>insensitive to brassinosteroids; semi-dwarf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>altered response to wounding; insensitive to j...</td>\n",
       "      <td>defects in meiosis; severely reduced fertility</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sensitive to zinc</td>\n",
       "      <td>mutants did not grow on medium  containing 2.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>embryo defective (inferred); female gametophyt...</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>pale green; slow growth</td>\n",
       "      <td>pigment defective embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>complete female gametophyte defective; male ga...</td>\n",
       "      <td>female gametophyte defective; rare embryo defe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>abnormal flower morphology; abnormal leaf morp...</td>\n",
       "      <td>embryo defective; root meristemless; the mutan...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>right-handed helical growth in the root and ot...</td>\n",
       "      <td>delayed germination; germination of the mutant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>late flowering; late flowering; strong influen...</td>\n",
       "      <td>the morphology of mutant plants was similar an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>sensitive to avirulent pseudomonas syringae; s...</td>\n",
       "      <td>sensitive to oxidative stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>abnormal flower morphology; abnormal leaf morp...</td>\n",
       "      <td>elevated threonine levels; short roots</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>decreased potassium permeability of the plasma...</td>\n",
       "      <td>chlorate resistance due to a reduced uptake of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>altered fatty acid composition; altered system...</td>\n",
       "      <td>late flowering; short leaf blades</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>abnormal flower morphology; abnormal leaf morp...</td>\n",
       "      <td>slow growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>elevated starch levels</td>\n",
       "      <td>low hydroxyl fatty acid levels in seeds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>embryo defective-preglobular / globular; embry...</td>\n",
       "      <td>altered seed storage composition, reduced oil ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>abnormal mitochondria morphology; abnormal per...</td>\n",
       "      <td>abnormal mitochondria morphology; abnormal per...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>complete sterility; defects in meiosis</td>\n",
       "      <td>chromosomes do not synapse during both male an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>albino seedlings; albino seeds</td>\n",
       "      <td>pale green; seedling lethal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "      <td>embryo defective; leafy cotyledons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>homozygotes are embryo lethal; homozygotes are...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>embryo defective-cotyledon; embryo defective; ...</td>\n",
       "      <td>embryo defective-globular; embryo defective; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>sensitive to manganese starvation; under mn-de...</td>\n",
       "      <td>etiolated seedlings of both wild type and mtk ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>completely sterile; dwarf</td>\n",
       "      <td>dwarf; pale green leaves</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>decreased stomatal closure; decreased stomatal...</td>\n",
       "      <td>decreased stomatal density; reduced number of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>reduced nitrate uptake; resistant to chlorate</td>\n",
       "      <td>resistant to tunicamycin; small seedlings</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>abnormal cell plates; abnormal cell walls</td>\n",
       "      <td>resistant to tunicamycin; small seedlings</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>embryo defective-preglobular; embryo defective...</td>\n",
       "      <td>crinkled leaves; embryo lethal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>abnormal pith cell morphology; dwarf</td>\n",
       "      <td>dwarf in the dark; no apical hook in the dark</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>low penetrance of heart-shaped cotyledons; sem...</td>\n",
       "      <td>embryo defective; variegated seedlings</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Phenotype 1                                        Phenotype 2\n",
       "0   long hypocotyl under far-red light; partially ...                            sensitive to uv-b light\n",
       "1                embryo defective; seedling defective  embryo defective-cotyledon; embryo defective; ...\n",
       "2       increased lateral root number; late flowering              insensitive to cytokinin; short roots\n",
       "3   embryo defective-globular; embryo defective; g...  embryo defective-transition; embryo defective;...\n",
       "4   abnormal microtubule development; abnormal pav...             albino seedlings; pale green seedlings\n",
       "5                                       albino; dwarf  homozygous plants have small and hyponastic co...\n",
       "6             pale green leaves; reduced rosette size          blunt siliques; short inflorescence stems\n",
       "7              pale green seedlings; pale green seeds  anthocyanin accumulation; dark-grown seedlings...\n",
       "8                embryo defective; seedling defective  abnormal rosette leaf morphology; decreased le...\n",
       "9                                large rosette leaves                    abnormal leaf morphology; dwarf\n",
       "10                            chlorotic leaves; dwarf                     reduced lateral root formation\n",
       "11  abnormal cotyledon morphology; collapsed root tip            abnormal chloroplast morphology; albino\n",
       "12                 dwarf; fragile inflorescence stems  altered response to red:far-red light; decreas...\n",
       "13                              anther defects; dwarf                    fewer lateral roots; semi-dwarf\n",
       "14       curled leaves; increased rosette leaf number  abnormal flower morphology; blister-like outgr...\n",
       "15                                  dark green; dwarf               elongated hypocotyl in far-red light\n",
       "16  resistant to anthranilate analogs (herbicide);...                          dwarf; pale green rosette\n",
       "17  elevated cobalt levels in shoots; sensitive to...  this mutant has about 1/10 of the ornithine am...\n",
       "18               dwarf; sensitive to oxidative stress        insensitive to brassinosteroids; semi-dwarf\n",
       "19  altered response to wounding; insensitive to j...     defects in meiosis; severely reduced fertility\n",
       "20                                  sensitive to zinc  mutants did not grow on medium  containing 2.5...\n",
       "21  embryo defective (inferred); female gametophyt...  embryo defective; developmental arrest of muta...\n",
       "22                            pale green; slow growth                           pigment defective embryo\n",
       "23  complete female gametophyte defective; male ga...  female gametophyte defective; rare embryo defe...\n",
       "24  abnormal flower morphology; abnormal leaf morp...  embryo defective; root meristemless; the mutan...\n",
       "25  right-handed helical growth in the root and ot...  delayed germination; germination of the mutant...\n",
       "26  late flowering; late flowering; strong influen...  the morphology of mutant plants was similar an...\n",
       "27  sensitive to avirulent pseudomonas syringae; s...                      sensitive to oxidative stress\n",
       "28  abnormal flower morphology; abnormal leaf morp...             elevated threonine levels; short roots\n",
       "29  decreased potassium permeability of the plasma...  chlorate resistance due to a reduced uptake of...\n",
       "30  altered fatty acid composition; altered system...                  late flowering; short leaf blades\n",
       "31  abnormal flower morphology; abnormal leaf morp...                                        slow growth\n",
       "32  embryo defective-globular; embryo defective; g...  embryo defective; developmental arrest of muta...\n",
       "33                             elevated starch levels            low hydroxyl fatty acid levels in seeds\n",
       "34  embryo defective-cotyledon; embryo defective; ...  embryo defective-globular; embryo defective; g...\n",
       "35  embryo defective-preglobular / globular; embry...  altered seed storage composition, reduced oil ...\n",
       "36  abnormal mitochondria morphology; abnormal per...  abnormal mitochondria morphology; abnormal per...\n",
       "37             complete sterility; defects in meiosis  chromosomes do not synapse during both male an...\n",
       "38                     albino seedlings; albino seeds                        pale green; seedling lethal\n",
       "39  embryo defective-globular; embryo defective; g...                 embryo defective; leafy cotyledons\n",
       "40  embryo defective-cotyledon; embryo defective; ...  homozygotes are embryo lethal; homozygotes are...\n",
       "41  embryo defective-cotyledon; embryo defective; ...  embryo defective-globular; embryo defective; g...\n",
       "42  sensitive to manganese starvation; under mn-de...  etiolated seedlings of both wild type and mtk ...\n",
       "43                          completely sterile; dwarf                           dwarf; pale green leaves\n",
       "44  decreased stomatal closure; decreased stomatal...  decreased stomatal density; reduced number of ...\n",
       "45      reduced nitrate uptake; resistant to chlorate          resistant to tunicamycin; small seedlings\n",
       "46          abnormal cell plates; abnormal cell walls          resistant to tunicamycin; small seedlings\n",
       "47  embryo defective-preglobular; embryo defective...                     crinkled leaves; embryo lethal\n",
       "48               abnormal pith cell morphology; dwarf      dwarf in the dark; no apical hook in the dark\n",
       "49  low penetrance of heart-shaped cotyledons; sem...             embryo defective; variegated seedlings"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Methodology\n",
    "# These are phenotype pairs drawn from the \n",
    "\n",
    "pairs = pd.DataFrame(OUTPUT_PAIR_TUPLES)\n",
    "pairs.columns = [\"Phenotype 1\",\"Phenotype 2\"]\n",
    "pairs[\"Phenotype 1\"] = pairs[\"Phenotype 1\"].str.casefold()\n",
    "pairs[\"Phenotype 2\"] = pairs[\"Phenotype 2\"].str.casefold()\n",
    "pairs[\"Phenotype 1\"] = pairs[\"Phenotype 1\"].map(lambda x: \"; \".join([s[:-1] for s in sent_tokenize(x)[0:2]]))\n",
    "pairs[\"Phenotype 2\"] = pairs[\"Phenotype 2\"].map(lambda x: \"; \".join([s[:-1] for s in sent_tokenize(x)[0:2]]))\n",
    "pairs.to_csv(\"../data/corpus_related_files/phenotype_pairs/unscored.csv\",index=False)\n",
    "pairs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
