{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from nltk.tokenize import word_tokenize\n",
    "from itertools import combinations\n",
    "import random\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, merge_list_dicts, flatten, to_hms\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASETS_DIR = \"../data/pickles/\"\n",
    "INPUT_DATASET_PATH = \"../data/pickles/gene_phenotype_dataset_all_text_and_annotations.pickle\"\n",
    "OUTPUT_TABLE = defaultdict(dict)\n",
    "OUTPUT_PAIR_TUPLES = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes share a phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       2868                 2365\n",
      "1   total       2868                 2365\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "# Find and save the dataset that is relevant for predicting whether two genes share a phenotype\n",
    "lloyd_subsets_filename = \"../data/pickles/groupings_from_lloyd_subsets.pickle\"  \n",
    "groups = load_from_pickle(lloyd_subsets_filename)\n",
    "id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "data.filter_with_ids(group_mapped_ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_phenotypes.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "shared = lambda x,y: len(set(id_to_group_ids[x]).intersection(set(id_to_group_ids[y])))>0\n",
    "pair_list = [shared(x,y) for x,y in list(combinations(group_mapped_ids, 2))]\n",
    "OUTPUT_TABLE[\"phenotype\"][\"relevant_genes\"] = len(group_mapped_ids)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"positive\"] = pair_list.count(True)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"negative\"] = pair_list.count(False)\n",
    "OUTPUT_TABLE[\"phenotype\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the pairs of phenotype descriptions to be used for curation.\n",
    "pos_pair_list = [(x,y) for x,y in list(combinations(group_mapped_ids, 2)) if shared(x,y)]\n",
    "sampling = random.choices(pos_pair_list, k=50)\n",
    "to_desc = data.get_description_dictionary()\n",
    "for (x,y) in sampling:\n",
    "    OUTPUT_PAIR_TUPLES.append((to_desc[x],to_desc[y]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes share a biochemical pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       1769                 1469\n",
      "1     gmx          1                    1\n",
      "2     mtr          2                    2\n",
      "3     osa          3                    3\n",
      "4     sly         18                   18\n",
      "5     zma        185                  160\n",
      "6   total       1978                 1653\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "# Using the KEGG pathways from the saved groupings object.\n",
    "kegg_pathways_filename = \"../data/pickles/groupings_from_kegg_pathways.pickle\"                         \n",
    "kegg_groups = load_from_pickle(kegg_pathways_filename)\n",
    "id_to_kegg_group_ids, kegg_group_id_to_ids = kegg_groups.get_groupings_for_dataset(data)\n",
    "# Using the PlantCyc pathways from the saved groupings object.\n",
    "pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"  \n",
    "pmn_groups = load_from_pickle(pmn_pathways_filename)\n",
    "id_to_pmn_group_ids, pmn_group_id_to_ids = pmn_groups.get_groupings_for_dataset(data)\n",
    "# Find and save the dataset that is relevant for predicting whether two genes share a phenotype\n",
    "group_mapped_ids = []\n",
    "group_mapped_ids.extend([k for (k,v) in id_to_kegg_group_ids.items() if len(v)>0])\n",
    "group_mapped_ids.extend([k for (k,v) in id_to_pmn_group_ids.items() if len(v)>0])\n",
    "group_mapped_ids = list(set(group_mapped_ids))\n",
    "data.filter_with_ids(group_mapped_ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_pathways.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "shared_kegg = lambda x,y: len(set(id_to_kegg_group_ids[x]).intersection(set(id_to_kegg_group_ids[y])))>0\n",
    "shared_pmn = lambda x,y: len(set(id_to_pmn_group_ids[x]).intersection(set(id_to_pmn_group_ids[y])))>0\n",
    "pair_list = [(shared_kegg(x,y) or shared_pmn(x,y)) for x,y in list(combinations(group_mapped_ids, 2))]\n",
    "OUTPUT_TABLE[\"pathways\"][\"relevant_genes\"] = len(group_mapped_ids)\n",
    "OUTPUT_TABLE[\"pathways\"][\"positive\"] = pair_list.count(True)\n",
    "OUTPUT_TABLE[\"pathways\"][\"negative\"] = pair_list.count(False)\n",
    "OUTPUT_TABLE[\"pathways\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the pairs of phenotype descriptions to be used for curation.\n",
    "pos_pair_list = [(x,y) for x,y in list(combinations(group_mapped_ids, 2)) if shared(x,y)]\n",
    "sampling = random.choices(pos_pair_list, k=50)\n",
    "to_desc = data.get_description_dictionary()\n",
    "for (x,y) in sampling:\n",
    "    OUTPUT_PAIR_TUPLES.append((to_desc[x],to_desc[y]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether a protein-protein interaction exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath       3674                 2950\n",
      "1     mtr         10                   10\n",
      "2     osa         44                   42\n",
      "3     sly          3                    3\n",
      "4     zma        160                  147\n",
      "5   total       3891                 3152\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "\n",
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "naming_file = \"../data/group_related_files/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../data/group_related_files/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis thaliana\n",
    "    \"../data/group_related_files/string/4577.protein.links.detailed.v11.0.txt\", # maize\n",
    "    \"../data/group_related_files/string/4530.protein.links.detailed.v11.0.txt\", # tomato \n",
    "    \"../data/group_related_files/string/4081.protein.links.detailed.v11.0.txt\", # medicago\n",
    "    \"../data/group_related_files/string/3880.protein.links.detailed.v11.0.txt\", # rice \n",
    "    \"../data/group_related_files/string/3847.protein.links.detailed.v11.0.txt\", # soybean\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "data.filter_with_ids(string_data.ids)\n",
    "print(\"\\n\",data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_interactions.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the ProteinInteraction class guarantees the dataframe contains the reverse of each edge as well.\n",
    "#interaction = lambda x,y: ((string_data.df[\"from\"] == x) & (string_data.df[\"to\"] == y)).any()\n",
    "#interaction_strings = set([\"{}:{}\".format(int(i),int(j)) for i,j in zip(string_data.df[\"from\"].values, string_data.df[\"to\"].values)])\n",
    "#has_interaction = lambda x,y: \"{}:{}\".format(int(x),int(y)) in interaction_strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "num_ids = len(data.get_ids())\n",
    "num_total_pairs = ((num_ids**2)/2)-num_ids\n",
    "num_positive = string_data.df.shape[0]/2\n",
    "num_negative = num_total_pairs - num_positive\n",
    "OUTPUT_TABLE[\"interactions\"][\"relevant_genes\"] = num_ids\n",
    "OUTPUT_TABLE[\"interactions\"][\"positive\"] = num_positive\n",
    "OUTPUT_TABLE[\"interactions\"][\"negative\"] = num_negative\n",
    "OUTPUT_TABLE[\"interactions\"][\"class_ratio\"] = \"{:0.4f}\".format(num_positive/num_negative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset that can be used to predict whether two genes are orthologous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  species  num_genes  unique_descriptions\n",
      "0     ath       6364                 3813\n",
      "1     gmx         30                   24\n",
      "2     mtr         37                   36\n",
      "3     osa         92                   85\n",
      "4     sly         70                   70\n",
      "5     zma       1406                  811\n",
      "6   total       7999                 4839\n",
      "\n",
      "   species  num_genes  unique_descriptions\n",
      "0     ath        350                  303\n",
      "1     osa         86                   80\n",
      "2     sly          7                    7\n",
      "3     zma        443                  337\n",
      "4   total        886                  727\n"
     ]
    }
   ],
   "source": [
    "# Starting off with the full dataset that is available.\n",
    "data = load_from_pickle(INPUT_DATASET_PATH)\n",
    "data.filter_has_description()\n",
    "print(data.describe())\n",
    "\n",
    "# Filter the dataset based on whether or not the genes were successfully mapped to an interaction.\n",
    "# Reduce size of the dataset by removing genes not mentioned in the STRING.\n",
    "ortholog_file_path = \"../data/orthology_related_files/pantherdb/PlantGenomeOrthologs_IRB_Modified.txt\"\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "data.filter_with_ids(ortholog_edgelist.ids)\n",
    "print(\"\\n\", data.describe())\n",
    "save_to_pickle(obj=data, path=os.path.join(DATASETS_DIR, \"gene_phenotype_dataset_for_predicting_orthologs.pickle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the AnyInteraction class guarantees the dataframe contains the reverse of each edge as well.\n",
    "#ortholog_strings = set([\"{}:{}\".format(int(i),int(j)) for i,j in zip(ortholog_edgelist.df[\"from\"].values, ortholog_edgelist.df[\"to\"].values)])\n",
    "#are_orthologs = lambda x,y: \"{}:{}\".format(int(x),int(y)) in ortholog_strings\n",
    "#pair_list = [are_orthologs(x,y) for x,y in list(combinations(data.get_ids(), 2))]\n",
    "# Update an output table to show the size of the datasets and the class sizes.\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"positive\"] = pair_list.count(True)\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"negative\"] = pair_list.count(False)\n",
    "#OUTPUT_TABLE[\"orthologs\"][\"class_ratio\"] = \"{:0.4f}\".format(pair_list.count(True)/pair_list.count(False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update an output table to show the size of the datasets and the class sizes, without enumerating.\n",
    "num_ids = len(data.get_ids())\n",
    "num_total_pairs = ((num_ids**2)/2)-num_ids\n",
    "num_positive = ortholog_edgelist.df.shape[0]/2\n",
    "num_negative = num_total_pairs - num_positive\n",
    "OUTPUT_TABLE[\"orthologs\"][\"relevant_genes\"] = num_ids\n",
    "OUTPUT_TABLE[\"orthologs\"][\"positive\"] = num_positive\n",
    "OUTPUT_TABLE[\"orthologs\"][\"negative\"] = num_negative\n",
    "OUTPUT_TABLE[\"orthologs\"][\"class_ratio\"] = \"{:0.4f}\".format(num_positive/num_negative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarzing table of the classes for each prediction problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "      <th>class_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>phenotype</th>\n",
       "      <td>458461</td>\n",
       "      <td>3652817</td>\n",
       "      <td>0.1255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathways</th>\n",
       "      <td>52530</td>\n",
       "      <td>1902723</td>\n",
       "      <td>0.0276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>interactions</th>\n",
       "      <td>450298</td>\n",
       "      <td>7.11575e+06</td>\n",
       "      <td>0.0633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orthologs</th>\n",
       "      <td>77</td>\n",
       "      <td>391535</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orthologs2</th>\n",
       "      <td>77</td>\n",
       "      <td>391535</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             positive     negative class_ratio\n",
       "phenotype      458461      3652817      0.1255\n",
       "pathways        52530      1902723      0.0276\n",
       "interactions   450298  7.11575e+06      0.0633\n",
       "orthologs          77       391535      0.0002\n",
       "orthologs2         77       391535      0.0002"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(OUTPUT_TABLE).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Long roots.</td>\n",
       "      <td>Abnormal inflorescence stem architecture. Narr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Embryo defective-Globular. Embryo defective; G...</td>\n",
       "      <td>embryo defective.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Altered leaf morphology. Altered leaf morpholo...</td>\n",
       "      <td>Abnormal flower morphology. Abnormal leaf morp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Embryo defective-Cotyledon. Low vitamin C leve...</td>\n",
       "      <td>Male and female sterile.  Defects in meiosis b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Elevated 16:0 fatty acid levels. Embryo lethal...</td>\n",
       "      <td>Contain Reduced Levels of  Aliphatic Glucosino...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Complete male gametophyte defective. Embryo de...</td>\n",
       "      <td>Delayed germination. Embryo defective-Globular...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Abnormal cotyledon morphology. Collapsed root ...</td>\n",
       "      <td>Abnormal leaf shape. Pale green leaves. Reduce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Dwarf. Pale green leaves. Dwarf; Pale green le...</td>\n",
       "      <td>Pigment defective embryo.  seedling-lethal whe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seedling lethal (inferred from pigment defect)...</td>\n",
       "      <td>Delayed senescence and yellowing of leaves.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Abnormal seedling growth in the dark. Insensit...</td>\n",
       "      <td>Early flowering. Long hypocotyl under blue lig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Abnormal shoot gravitropism. Dwarf. Embryo def...</td>\n",
       "      <td>Resistant to fungal infection. Semi-dwarf. Sem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Embryo defective-Cotyledon. Embryo defective; ...</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Dwarf. Embryo defective-Cotyledon. Late flower...</td>\n",
       "      <td>embryo defective; cotyledon. developmental arr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Early flowering. Germination sensitive to ABA....</td>\n",
       "      <td>Sensitive to ABA. Sensitive to salt. Sensitive...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Abnormal leaf morphology. Base of rosette divi...</td>\n",
       "      <td>Reduced organ growth.  Reduced cell proliferat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Complete loss of stomata formation. Dwarf. Pal...</td>\n",
       "      <td>angustifolia, apetala, pyrimidine requiring, e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Abnormal cotyledon venation. Abnormal leaf ven...</td>\n",
       "      <td>Purple leaves. Slow growth. Upward-bending lea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Pale green seedlings.</td>\n",
       "      <td>Yellow seed coat.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Embryo defective-Cotyledon. Embryo defective; ...</td>\n",
       "      <td>Yellow/pale-green phenotype and reduced growth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Abnormal seeds. Curled leaves. Dwarf. Embryo d...</td>\n",
       "      <td>Serrated leaves.  Increased endoreduplication ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Reduced nitrate uptake. Resistant to chlorate....</td>\n",
       "      <td>Slightly decreased chlorophyll fluorescence. S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Resistant to 2, 4-DB. Insensitive to IBA. Inse...</td>\n",
       "      <td>Leaf senescence insensitive to ethylene.  Enha...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Dwarf. Elongated mitochondria. Slow growth. Dw...</td>\n",
       "      <td>Increased growth in response to sucrose. Reduc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Embryo defective-Preglobular. Embryo defective...</td>\n",
       "      <td>Embryo defective-Preglobular. Embryo defective...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Embryo defective-Preglobular / Globular. Embry...</td>\n",
       "      <td>Shriveled seeds, rescued seedlings are albino ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Abnormal root gravitropism. Root growth insens...</td>\n",
       "      <td>Early flowering. Insensitive to cytokinin. Ins...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Dwarf. Increased branching. Thin inflorescence...</td>\n",
       "      <td>Slightly slower growth. Small rosettes under h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Dwarf under low light. Dwarf under sucrose. Le...</td>\n",
       "      <td>Large seeds. Seedling lethal. Null: Seedling l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Long hypocotyl. Long roots in response to MeIA...</td>\n",
       "      <td>Short hypocotyl. Small cotyledons. Short hypoc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Long hypocotyl in the dark. Sensitive to ABA. ...</td>\n",
       "      <td>Dwarf in the dark. No apical hook in the dark....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Rosette leaves are lobed and curled downwards ...</td>\n",
       "      <td>Reduced stature, early flowering, and altered ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Abnormal cell specification. Embryo defective-...</td>\n",
       "      <td>Embryo lethal. Embryo development was arrested...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Embryo defective-Cotyledon. Embryo defective; ...</td>\n",
       "      <td>Embryo defective-Globular. Embryo defective; G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Lethal at high temperature.  pmm-12 plants are...</td>\n",
       "      <td>Complete loss of  fructose accumulation at low...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Curled leaves.</td>\n",
       "      <td>GA responsive dwarf, transparent testa, yellow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Cotyledons do not turn purple in response to s...</td>\n",
       "      <td>Dwarf under limited phosphate. Elevated anthoc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Abnormal floral development. Abnormal leaf mor...</td>\n",
       "      <td>Dwarf. Long telomeres. Sensitive to MMS (induc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Necrotic lesions on leaves. Susceptible to bac...</td>\n",
       "      <td>Sensitive to blue  light. Sensitive to far-red...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Dwarf. Increased floral organ number. Large sh...</td>\n",
       "      <td>Curved leaves. Elongated leaves. Incomplete pe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Early flowering. Slightly elevated sulfate lev...</td>\n",
       "      <td>Low phosphate levels in response to elevated p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Complete female gametophyte defective. Male ga...</td>\n",
       "      <td>Delayed senescence. Few flowers. Insensitive t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Increased lateral root number. Sensitive to sa...</td>\n",
       "      <td>Early flowering. Long petioles. Reduced fertil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Completely female sterile. Completely male ste...</td>\n",
       "      <td>Reduced fertility.  Semi-sterile, reduced fert...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Late flowering.  late flowering; not later flo...</td>\n",
       "      <td>Complete loss of cotyledonary petiole. Large S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Embryo defective-Cotyledon. Embryo defective; ...</td>\n",
       "      <td>Disruption in ER-derived thylakoid lipids. Com...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Embryo defective-Globular. Embryo defective; G...</td>\n",
       "      <td>Embryo defective-Preglobular. Embryo defective...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Large cotyledons under red light. Short hypoco...</td>\n",
       "      <td>strong allele; long hypocotyl and unexpanded c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Sensitive to high temperature.  embryo defecti...</td>\n",
       "      <td>Early flowering. Enhanced ethylene production....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Incomplete penetrance of decreased stamen numb...</td>\n",
       "      <td>Severely reduced fertility. Small anthers. Ste...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Abnormal floral morphology. Abnormal rosette g...</td>\n",
       "      <td>Slight reduction in root and shoot growth, exa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0                                                  1\n",
       "0                                         Long roots.  Abnormal inflorescence stem architecture. Narr...\n",
       "1   Embryo defective-Globular. Embryo defective; G...                                  embryo defective.\n",
       "2   Altered leaf morphology. Altered leaf morpholo...  Abnormal flower morphology. Abnormal leaf morp...\n",
       "3   Embryo defective-Cotyledon. Low vitamin C leve...  Male and female sterile.  Defects in meiosis b...\n",
       "4   Elevated 16:0 fatty acid levels. Embryo lethal...  Contain Reduced Levels of  Aliphatic Glucosino...\n",
       "5   Complete male gametophyte defective. Embryo de...  Delayed germination. Embryo defective-Globular...\n",
       "6   Abnormal cotyledon morphology. Collapsed root ...  Abnormal leaf shape. Pale green leaves. Reduce...\n",
       "7   Dwarf. Pale green leaves. Dwarf; Pale green le...  Pigment defective embryo.  seedling-lethal whe...\n",
       "8   Seedling lethal (inferred from pigment defect)...        Delayed senescence and yellowing of leaves.\n",
       "9   Abnormal seedling growth in the dark. Insensit...  Early flowering. Long hypocotyl under blue lig...\n",
       "10  Abnormal shoot gravitropism. Dwarf. Embryo def...  Resistant to fungal infection. Semi-dwarf. Sem...\n",
       "11  Embryo defective-Cotyledon. Embryo defective; ...  embryo defective; developmental arrest of muta...\n",
       "12  Dwarf. Embryo defective-Cotyledon. Late flower...  embryo defective; cotyledon. developmental arr...\n",
       "13  Early flowering. Germination sensitive to ABA....  Sensitive to ABA. Sensitive to salt. Sensitive...\n",
       "14  Abnormal leaf morphology. Base of rosette divi...  Reduced organ growth.  Reduced cell proliferat...\n",
       "15  Complete loss of stomata formation. Dwarf. Pal...  angustifolia, apetala, pyrimidine requiring, e...\n",
       "16  Abnormal cotyledon venation. Abnormal leaf ven...  Purple leaves. Slow growth. Upward-bending lea...\n",
       "17                              Pale green seedlings.                                  Yellow seed coat.\n",
       "18  Embryo defective-Cotyledon. Embryo defective; ...  Yellow/pale-green phenotype and reduced growth...\n",
       "19  Abnormal seeds. Curled leaves. Dwarf. Embryo d...  Serrated leaves.  Increased endoreduplication ...\n",
       "20  Reduced nitrate uptake. Resistant to chlorate....  Slightly decreased chlorophyll fluorescence. S...\n",
       "21  Resistant to 2, 4-DB. Insensitive to IBA. Inse...  Leaf senescence insensitive to ethylene.  Enha...\n",
       "22  Dwarf. Elongated mitochondria. Slow growth. Dw...  Increased growth in response to sucrose. Reduc...\n",
       "23  Embryo defective-Preglobular. Embryo defective...  Embryo defective-Preglobular. Embryo defective...\n",
       "24  Embryo defective-Preglobular / Globular. Embry...  Shriveled seeds, rescued seedlings are albino ...\n",
       "25  Abnormal root gravitropism. Root growth insens...  Early flowering. Insensitive to cytokinin. Ins...\n",
       "26  Dwarf. Increased branching. Thin inflorescence...  Slightly slower growth. Small rosettes under h...\n",
       "27  Dwarf under low light. Dwarf under sucrose. Le...  Large seeds. Seedling lethal. Null: Seedling l...\n",
       "28  Long hypocotyl. Long roots in response to MeIA...  Short hypocotyl. Small cotyledons. Short hypoc...\n",
       "29  Long hypocotyl in the dark. Sensitive to ABA. ...  Dwarf in the dark. No apical hook in the dark....\n",
       "30  Rosette leaves are lobed and curled downwards ...  Reduced stature, early flowering, and altered ...\n",
       "31  Abnormal cell specification. Embryo defective-...  Embryo lethal. Embryo development was arrested...\n",
       "32  Embryo defective-Cotyledon. Embryo defective; ...  Embryo defective-Globular. Embryo defective; G...\n",
       "33  Lethal at high temperature.  pmm-12 plants are...  Complete loss of  fructose accumulation at low...\n",
       "34                                     Curled leaves.  GA responsive dwarf, transparent testa, yellow...\n",
       "35  Cotyledons do not turn purple in response to s...  Dwarf under limited phosphate. Elevated anthoc...\n",
       "36  Abnormal floral development. Abnormal leaf mor...  Dwarf. Long telomeres. Sensitive to MMS (induc...\n",
       "37  Necrotic lesions on leaves. Susceptible to bac...  Sensitive to blue  light. Sensitive to far-red...\n",
       "38  Dwarf. Increased floral organ number. Large sh...  Curved leaves. Elongated leaves. Incomplete pe...\n",
       "39  Early flowering. Slightly elevated sulfate lev...  Low phosphate levels in response to elevated p...\n",
       "40  Complete female gametophyte defective. Male ga...  Delayed senescence. Few flowers. Insensitive t...\n",
       "41  Increased lateral root number. Sensitive to sa...  Early flowering. Long petioles. Reduced fertil...\n",
       "42  Completely female sterile. Completely male ste...  Reduced fertility.  Semi-sterile, reduced fert...\n",
       "43  Late flowering.  late flowering; not later flo...  Complete loss of cotyledonary petiole. Large S...\n",
       "44  Embryo defective-Cotyledon. Embryo defective; ...  Disruption in ER-derived thylakoid lipids. Com...\n",
       "45  Embryo defective-Globular. Embryo defective; G...  Embryo defective-Preglobular. Embryo defective...\n",
       "46  Large cotyledons under red light. Short hypoco...  strong allele; long hypocotyl and unexpanded c...\n",
       "47  Sensitive to high temperature.  embryo defecti...  Early flowering. Enhanced ethylene production....\n",
       "48  Incomplete penetrance of decreased stamen numb...  Severely reduced fertility. Small anthers. Ste...\n",
       "49  Abnormal floral morphology. Abnormal rosette g...  Slight reduction in root and shoot growth, exa..."
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(OUTPUT_PAIR_TUPLES)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
