{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description\n",
    "This notebook is for generating tables and corresponding files that contain mappings between plant gene identifiers from the dataset of interest, the gene symbols of orthologous human genes, and a description of the disease phenotype that is associated with that human gene symbol, where this information is known. The mappings between plant gene identifiers and human orthologs are identfied using [PantherDB](http://www.pantherdb.org/), and the associations between human gene symbols and disease phenotypes are identified using [OMIM](https://omim.org/). The purpose of mapping the dataset of plant gene identifiers to this information is to test if there are disease phenotypes that are phenologs or highly associated with groups of plant genes based on their phenotype descriptions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import urllib.request\n",
    "import json\n",
    "import time\n",
    "import sys\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import flatten\n",
    "from _hidden import omim_api_key\n",
    "\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Script needs this information in the shape dict[organism name]-->[gene1, gene2, gene3, ...]\n",
    "input_data = pd.read_table(\"/Users/irbraun/Desktop/Locus_Germplasm_Phenotype_20180702.txt\")\n",
    "gene_names = pd.unique(input_data[\"LOCUS_NAME\"].values)\n",
    "\n",
    "# Subset the gene names to not query PantherDB for the entire dataset at once, and name file accordingly.\n",
    "lower_index = 4800\n",
    "upper_index = 6000\n",
    "gene_names = gene_names[lower_index:upper_index]\n",
    "input_data = {\"ARATH\":gene_names}\n",
    "mim2gene_filename = \"../data/gene_related_files/omim/mim2gene_irb_cleaned.txt\"\n",
    "temp_df_filename = \"../data/scratch/arabidopsis_orthologs_df_{}_to_{}.csv\".format(lower_index,upper_index)\n",
    "full_df_filename = \"../data/orthology_related_files/arabidopsis_panther_omim_irb_{}_to_{}.csv\".format(lower_index,upper_index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying PantherDB for Human Orthologs of Plant Genes\n",
    "\n",
    "This text is copied from [this page](http://www.pantherdb.org/help/PANTHERhelp.jsp#V.A.).\n",
    "\n",
    "Search via a URL\n",
    "\n",
    "The following parameters are required.\n",
    "1. type - This refers to the search type and should be specified as \"matchingOrtholog\"\n",
    "2. inputOrganism - The organism/genome being queried by the search term(s)\n",
    "3. targetOrganism -  Target organism for ortholgos from search.  Multiple organisms can be specified, separated by commas\n",
    "4. orthologType - LDO or all - LDO will return only least diverged ortholog for each gene (single \"best\" ortholog), and all will return all orthologs if more than one\n",
    "5. searchTerm - query terms, these should optimally be Uniprot of MOD gene identifiers, but, other identifiers are supported such as gene symbols.  Maximum of 10 query terms can be submitted, separated by commas.\n",
    "\n",
    "For the inputOrganism and targetOrganism, the 5-letter Uniprot code is used, see the Short Name field on the [summaryStats](http://www.pantherdb.org/panther/summaryStats.jsp) page for a list of available organisms and the associated codes.\n",
    "Example - http://www.pantherdb.org/webservices/ortholog.jsp?type=matchingOrtholog&inputOrganism=MOUSE&targetOrganism=HUMAN&orthologType=all&searchTerm=101816\n",
    "\n",
    "For each match, the following data are returned:\n",
    "\n",
    "If search term has no match in the input organism:\n",
    "<searchTerm>    \"Search term not found in input organism <inputOrganism\"\n",
    "If gene is found, but it has no ortholog in the database or orthologType specified is not found in the database:\n",
    "<searchTerm>     <matchedGene>    \"No ortholog found in target organism <targetOrganism>\"\n",
    "\n",
    "If one or more hit, example results, one line per match (each field is tab-separated):\n",
    "101816 MOUSE|MGI=MGI=101816|UniProtKB=P43247 HUMAN|ENSEMBL=ENSG00000095002|UniProtKB=P43246 MSH2 LDO \n",
    "\n",
    "1. searchTerm that was matched\n",
    "\n",
    "2. Information about matched gene in input organism:organism|gene_id(database=id)|protein_id(database=id)\n",
    "\n",
    "2. information about ortholog in target organism:    organism|gene_id(database=id)|protein_id(database=id)\n",
    "\n",
    "4. gene symbol of the matched gene in the target organism\n",
    "\n",
    "5. Type of ortholog, LDO (least diverged) or O (other ortholog, if more than one) or P for paralog or X for horizontal gene transfer or LDX for least diverged horizontal gene transfer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "completed queries for 10/503 arath genes\n",
      "completed queries for 20/503 arath genes\n",
      "completed queries for 30/503 arath genes\n",
      "completed queries for 40/503 arath genes\n",
      "completed queries for 50/503 arath genes\n",
      "completed queries for 60/503 arath genes\n",
      "completed queries for 70/503 arath genes\n",
      "completed queries for 80/503 arath genes\n",
      "completed queries for 90/503 arath genes\n",
      "completed queries for 100/503 arath genes\n",
      "completed queries for 110/503 arath genes\n",
      "completed queries for 120/503 arath genes\n",
      "completed queries for 130/503 arath genes\n",
      "completed queries for 140/503 arath genes\n",
      "completed queries for 150/503 arath genes\n",
      "completed queries for 160/503 arath genes\n",
      "completed queries for 170/503 arath genes\n",
      "completed queries for 180/503 arath genes\n",
      "completed queries for 190/503 arath genes\n",
      "completed queries for 200/503 arath genes\n",
      "completed queries for 210/503 arath genes\n",
      "completed queries for 220/503 arath genes\n",
      "completed queries for 230/503 arath genes\n",
      "completed queries for 240/503 arath genes\n",
      "completed queries for 250/503 arath genes\n",
      "completed queries for 260/503 arath genes\n",
      "completed queries for 270/503 arath genes\n",
      "completed queries for 280/503 arath genes\n",
      "completed queries for 290/503 arath genes\n",
      "completed queries for 300/503 arath genes\n",
      "completed queries for 310/503 arath genes\n",
      "completed queries for 320/503 arath genes\n",
      "completed queries for 330/503 arath genes\n",
      "completed queries for 340/503 arath genes\n",
      "completed queries for 350/503 arath genes\n",
      "completed queries for 360/503 arath genes\n",
      "completed queries for 370/503 arath genes\n",
      "completed queries for 380/503 arath genes\n",
      "completed queries for 390/503 arath genes\n",
      "completed queries for 400/503 arath genes\n",
      "completed queries for 410/503 arath genes\n",
      "completed queries for 420/503 arath genes\n",
      "completed queries for 430/503 arath genes\n",
      "completed queries for 440/503 arath genes\n",
      "completed queries for 450/503 arath genes\n",
      "completed queries for 460/503 arath genes\n",
      "completed queries for 470/503 arath genes\n",
      "completed queries for 480/503 arath genes\n",
      "completed queries for 490/503 arath genes\n",
      "completed queries for 500/503 arath genes\n",
      "completed queries for 503/503 arath genes\n",
      "done with all queries\n"
     ]
    }
   ],
   "source": [
    "# Constructing a query for orthologs, the search terms should be a gene symbols or UniProt IDs.\n",
    "query_format = \"http://pantherdb.org/webservices/ortholog.jsp?type=matchingOrtholog&inputOrganism={}&targetOrganism={}&orthologType=all&searchTerm={}\"\n",
    "target_organism = \"HUMAN\"\n",
    "query_term_limit = 10\n",
    "wait_time_s = 2\n",
    "\n",
    "\n",
    "# A class for a single line of the PantherDB query results, not all this information is currently used.\n",
    "class Result:\n",
    "    def __init__(self, result_str):\n",
    "        fields = result_str.split(r\"\\t\")\n",
    "        self.search_term = fields[0].strip()\n",
    "        self.input_info = fields[1].strip()\n",
    "        self.target_info = fields[2].strip()\n",
    "        self.target_gene_symbol = fields[3].strip()\n",
    "        self.ortholog_type = fields[4].strip()\n",
    "\n",
    "\n",
    "# Given an input dict[organism]-->[g1, g2, g3..], make all the queries with a maximum of ten at a time \n",
    "# to ahdere to the instructions in the guidelines for the PantherDB guidelines listed above.\n",
    "tuples = []\n",
    "for input_organism,gene_terms in input_data.items():\n",
    "    for lower_idx in range(0,len(gene_terms),query_term_limit):\n",
    "        upper_idx = min(lower_idx+query_term_limit-1,len(gene_terms)-1)\n",
    "        gene_terms_sublist = gene_terms[lower_idx:upper_idx+1]\n",
    "        query = query_format.format(input_organism,target_organism,\",\".join(gene_terms_sublist))\n",
    "        result = urllib.request.urlopen(query).read()\n",
    "        all_results = str(result).replace(r\"b'\\r\\n\\r\\n\",\"\").replace(r\"\\n\\r\\n'\",\"\")\n",
    "        results = all_results.split(r\"\\n\")\n",
    "        results = [Result(r) for r in results if len(r.split(r\"\\t\"))==5]    \n",
    "        print(\"completed queries for {}/{} {} genes\".format(upper_idx+1, len(gene_terms), input_organism.lower()))\n",
    "        tuples.extend([(input_organism, result.search_term, result.target_gene_symbol) for result in results])\n",
    "        time.sleep(wait_time_s)\n",
    "print(\"done with all queries\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a dataframe containing all of the PantherDB query results.\n",
    "ortholog_df = pd.DataFrame(tuples, columns=[\"species\", \"gene_identifier\", \"human_ortholog_gene_symbol\"])\n",
    "ortholog_df.head()\n",
    "ortholog_df.to_csv(temp_df_filename, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying OMIM for the Asscoiated Disease Phenotypes\n",
    "See [this page](https://www.omim.org/help/api) for structuring queries to the OMIM API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in the table provided by OMIM for mapping between gene symbols and MIM numbers.\n",
    "mim2gene_df = pd.read_table(mim2gene_filename)\n",
    "gene2mim_dict = {gene_symbol:mim_number for gene_symbol,mim_number in zip(mim2gene_df[\"Approved Gene Symbol (HGNC)\"].values, mim2gene_df[\"MIM Number\"].values)}\n",
    "# What are the lists of human gene symboles and corresponding MIM numbers?\n",
    "gene_symbols = pd.unique(ortholog_df[\"human_ortholog_gene_symbol\"].values)\n",
    "gene_mim_numbers = [gene2mim_dict.get(symbol,None) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "      <th>gene_mim_number</th>\n",
       "      <th>phenotype_mim_number</th>\n",
       "      <th>phenotype_mim_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [human_ortholog_gene_symbol, gene_mim_number, phenotype_mim_number, phenotype_mim_name]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Queryinq OMIM for the phenotypes which are associated with the gene MIM numbers of these orthologs.\n",
    "query_format = \"https://api.omim.org/api/geneMap?mimNumber={}&format=json&phenotypeExists=true&start=0&limit=10&apiKey={}\"\n",
    "api_key = omim_api_key\n",
    "wait_time_s = 1\n",
    "tuples = []\n",
    "for gene_symbol, gene_mim_number in zip(gene_symbols, gene_mim_numbers):\n",
    "    if gene_mim_number is not None:\n",
    "        try:\n",
    "            query = query_format.format(gene_mim_number, api_key)\n",
    "            result_dict = json.loads(urllib.request.urlopen(query).read())\n",
    "            for phenotype_map_list_item in result_dict[\"omim\"][\"listResponse\"][\"geneMapList\"][0][\"geneMap\"][\"phenotypeMapList\"]:\n",
    "                phenotype_mim_number = phenotype_map_list_item[\"phenotypeMap\"][\"phenotypeMimNumber\"]\n",
    "                phenotype_mim_name = phenotype_map_list_item[\"phenotypeMap\"][\"phenotype\"]\n",
    "                tuples.append((gene_symbol, gene_mim_number, phenotype_mim_number, phenotype_mim_name))  \n",
    "        except KeyError:\n",
    "            continue\n",
    "        except IndexError:\n",
    "            continue\n",
    "        time.sleep(wait_time_s)   \n",
    "        \n",
    "gene_mim_df = pd.DataFrame(tuples, columns=[\"human_ortholog_gene_symbol\",\"gene_mim_number\",\"phenotype_mim_number\",\"phenotype_mim_name\"])\n",
    "gene_mim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_identifier</th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "      <th>gene_mim_number</th>\n",
       "      <th>phenotype_mim_number</th>\n",
       "      <th>phenotype_mim_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>GR2</td>\n",
       "      <td>GLYR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>PUP3</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>PUP3</td>\n",
       "      <td>MTG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>RE</td>\n",
       "      <td>MTG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species gene_identifier human_ortholog_gene_symbol  gene_mim_number  phenotype_mim_number phenotype_mim_name\n",
       "0   ARATH             GR2                      GLYR1              NaN                   NaN                NaN\n",
       "1   ARATH            PUP3                          -              NaN                   NaN                NaN\n",
       "2   ARATH            PUP3                       MTG1              NaN                   NaN                NaN\n",
       "3   ARATH              RE                          -              NaN                   NaN                NaN\n",
       "4   ARATH              RE                       MTG1              NaN                   NaN                NaN"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Producing a dataframe and CSV file contains the merged information from both dataframes created.\n",
    "full_df = pd.merge(left=ortholog_df, right=gene_mim_df, how=\"left\")\n",
    "full_df[\"gene_mim_number\"] = full_df[\"gene_mim_number\"].astype(\"Int64\")\n",
    "full_df[\"phenotype_mim_number\"] = full_df[\"phenotype_mim_number\"].astype(\"Int64\")\n",
    "full_df.to_csv(full_df_filename, index=False)\n",
    "full_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
