{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import urllib.request\n",
    "import json\n",
    "import time\n",
    "import sys\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import flatten\n",
    "from _hidden import omim_api_key\n",
    "\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32\n"
     ]
    }
   ],
   "source": [
    "input_data = pd.read_table(\"/Users/irbraun/Desktop/Locus_Germplasm_Phenotype_20180702.txt\")\n",
    "gene_names = pd.unique(input_data[\"LOCUS_NAME\"].values)\n",
    "gene_names = gene_names[50:82]\n",
    "\n",
    "# Want to get this information in the shape dict[organism]-->[gene1, gene2, gene3, ...]\n",
    "input_data = {\"ARATH\":gene_names}\n",
    "print(len(input_data[\"ARATH\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying PantherDB for Orthologs, Copied From [This Page](http://www.pantherdb.org/help/PANTHERhelp.jsp#V.A.)\n",
    "\n",
    "Search via a URL\n",
    "\n",
    "The following parameters are required.\n",
    "1. type - This refers to the search type and should be specified as \"matchingOrtholog\"\n",
    "2. inputOrganism - The organism/genome being queried by the search term(s)\n",
    "3. targetOrganism -  Target organism for ortholgos from search.  Multiple organisms can be specified, separated by commas\n",
    "4. orthologType - LDO or all - LDO will return only least diverged ortholog for each gene (single \"best\" ortholog), and all will return all orthologs if more than one\n",
    "5. searchTerm - query terms, these should optimally be Uniprot of MOD gene identifiers, but, other identifiers are supported such as gene symbols.  Maximum of 10 query terms can be submitted, separated by commas.\n",
    "\n",
    "For the inputOrganism and targetOrganism, the 5-letter Uniprot code is used, see the Short Name field on the [summaryStats](http://www.pantherdb.org/panther/summaryStats.jsp) page for a list of available organisms and the associated codes.\n",
    "Example - http://www.pantherdb.org/webservices/ortholog.jsp?type=matchingOrtholog&inputOrganism=MOUSE&targetOrganism=HUMAN&orthologType=all&searchTerm=101816\n",
    "\n",
    "For each match, the following data are returned:\n",
    "\n",
    "If search term has no match in the input organism:\n",
    "<searchTerm>    \"Search term not found in input organism <inputOrganism\"\n",
    "If gene is found, but it has no ortholog in the database or orthologType specified is not found in the database:\n",
    "<searchTerm>     <matchedGene>    \"No ortholog found in target organism <targetOrganism>\"\n",
    "\n",
    "If one or more hit, example results, one line per match (each field is tab-separated):\n",
    "101816 MOUSE|MGI=MGI=101816|UniProtKB=P43247 HUMAN|ENSEMBL=ENSG00000095002|UniProtKB=P43246 MSH2 LDO \n",
    "\n",
    "1. searchTerm that was matched\n",
    "\n",
    "2. Information about matched gene in input organism:organism|gene_id(database=id)|protein_id(database=id)\n",
    "\n",
    "2. information about ortholog in target organism:    organism|gene_id(database=id)|protein_id(database=id)\n",
    "\n",
    "4. gene symbol of the matched gene in the target organism\n",
    "\n",
    "5. Type of ortholog, LDO (least diverged) or O (other ortholog, if more than one) or P for paralog or X for horizontal gene transfer or LDX for least diverged horizontal gene transfer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "completed queries for 10/32 arath genes\n",
      "completed queries for 20/32 arath genes\n",
      "completed queries for 30/32 arath genes\n",
      "completed queries for 32/32 arath genes\n",
      "done with all queries\n"
     ]
    }
   ],
   "source": [
    "# Constructing a query for orthologs, the search terms should be a gene symbols or UniProt IDs.\n",
    "query_format = \"http://pantherdb.org/webservices/ortholog.jsp?type=matchingOrtholog&inputOrganism={}&targetOrganism={}&orthologType=all&searchTerm={}\"\n",
    "target_organism = \"HUMAN\"\n",
    "query_term_limit = 10\n",
    "wait_time_s = 2\n",
    "\n",
    "\n",
    "# A class for a single line of the PantherDB query results, not all this information is currently used.\n",
    "class Result:\n",
    "    def __init__(self, result_str):\n",
    "        fields = result_str.split(r\"\\t\")\n",
    "        self.search_term = fields[0].strip()\n",
    "        self.input_info = fields[1].strip()\n",
    "        self.target_info = fields[2].strip()\n",
    "        self.target_gene_symbol = fields[3].strip()\n",
    "        self.ortholog_type = fields[4].strip()\n",
    "\n",
    "\n",
    "# Given an input dict[organism]-->[g1, g2, g3..], make all the queries with a maximum of ten at a time \n",
    "# to ahdere to the instructions in the guidelines for the PantherDB guidelines listed above.\n",
    "tuples = []\n",
    "for input_organism,gene_terms in input_data.items():\n",
    "    for lower_idx in range(0,len(gene_terms),query_term_limit):\n",
    "        upper_idx = min(lower_idx+query_term_limit-1,len(gene_terms)-1)\n",
    "        gene_terms_sublist = gene_terms[lower_idx:upper_idx+1]\n",
    "        query = query_format.format(input_organism,target_organism,\",\".join(gene_terms_sublist))\n",
    "        result = urllib.request.urlopen(query).read()\n",
    "        all_results = str(result).replace(r\"b'\\r\\n\\r\\n\",\"\").replace(r\"\\n\\r\\n'\",\"\")\n",
    "        results = all_results.split(r\"\\n\")\n",
    "        results = [Result(r) for r in results if len(r.split(r\"\\t\"))==5]    \n",
    "        print(\"completed queries for {}/{} {} genes\".format(upper_idx+1, len(gene_terms), input_organism.lower()))\n",
    "        tuples.extend([(input_organism, result.search_term, result.target_gene_symbol) for result in results])\n",
    "        time.sleep(wait_time_s)\n",
    "print(\"done with all queries\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_identifier</th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>MCCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>ERCC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>CPOX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>SOHLH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>SOHLH2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species gene_identifier human_ortholog_gene_symbol\n",
       "0   ARATH       AT1G03430                      MCCC1\n",
       "1   ARATH       AT1G03430                      ERCC2\n",
       "2   ARATH       AT1G03430                       CPOX\n",
       "3   ARATH       AT1G03430                     SOHLH1\n",
       "4   ARATH       AT1G03430                     SOHLH2"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating a dataframe containing all of the PantherDB query results.\n",
    "ortholog_df = pd.DataFrame(tuples, columns=[\"species\", \"gene_identifier\", \"human_ortholog_gene_symbol\"])\n",
    "ortholog_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying OMIM for the Asscoiated Disease Phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in the table provided by OMIM for mapping between \n",
    "mim2gene_df = pd.read_table(\"../data/gene_related_files/omim/mim2gene_irb_cleaned.txt\")\n",
    "gene2mim_dict = {gene_symbol:mim_number for gene_symbol,mim_number in zip(mim2gene_df[\"Approved Gene Symbol (HGNC)\"].values, mim2gene_df[\"MIM Number\"].values)}\n",
    "# What are the lists of human gene symboles and corresponding MIM numbers?\n",
    "gene_symbols = pd.unique(ortholog_df[\"human_ortholog_gene_symbol\"].values)\n",
    "gene_mim_numbers = [gene2mim_dict.get(symbol,None) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "      <th>gene_mim_number</th>\n",
       "      <th>phenotype_mim_number</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MCCC1</td>\n",
       "      <td>609010</td>\n",
       "      <td>210200</td>\n",
       "      <td>3-Methylcrotonyl-CoA carboxylase 1 deficiency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126340</td>\n",
       "      <td>610756</td>\n",
       "      <td>?Cerebrooculofacioskeletal syndrome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126340</td>\n",
       "      <td>601675</td>\n",
       "      <td>Trichothiodystrophy 1, photosensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126340</td>\n",
       "      <td>278730</td>\n",
       "      <td>Xeroderma pigmentosum, group D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CPOX</td>\n",
       "      <td>612732</td>\n",
       "      <td>121300</td>\n",
       "      <td>Coproporphyria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  human_ortholog_gene_symbol  gene_mim_number  phenotype_mim_number                                      phenotype\n",
       "0                      MCCC1           609010                210200  3-Methylcrotonyl-CoA carboxylase 1 deficiency\n",
       "1                      ERCC2           126340                610756          ?Cerebrooculofacioskeletal syndrome 2\n",
       "2                      ERCC2           126340                601675          Trichothiodystrophy 1, photosensitive\n",
       "3                      ERCC2           126340                278730                 Xeroderma pigmentosum, group D\n",
       "4                       CPOX           612732                121300                                 Coproporphyria"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Queryinq OMIM for the phenotypes which are associated with the gene MIM numbers of these orthologs.\n",
    "query_format = \"https://api.omim.org/api/geneMap?mimNumber={}&format=json&phenotypeExists=true&start=0&limit=10&apiKey={}\"\n",
    "api_key = omim_api_key\n",
    "tuples = []\n",
    "for gene_symbol, gene_mim_number in zip(gene_symbols, gene_mim_numbers):\n",
    "    if gene_mim_number is not None:\n",
    "        try:\n",
    "            query = query_format.format(gene_mim_number, api_key)\n",
    "            result_dict = json.loads(urllib.request.urlopen(query).read())\n",
    "            for phenotype_map_list_item in result_dict[\"omim\"][\"listResponse\"][\"geneMapList\"][0][\"geneMap\"][\"phenotypeMapList\"]:\n",
    "                phenotype_mim_number = phenotype_map_list_item[\"phenotypeMap\"][\"phenotypeMimNumber\"]\n",
    "                phenotype_mim_name = phenotype_map_list_item[\"phenotypeMap\"][\"phenotype\"]\n",
    "                tuples.append((gene_symbol, gene_mim_number, phenotype_mim_number, phenotype_mim_name))\n",
    "        except KeyError:\n",
    "            continue\n",
    "        except IndexError:\n",
    "            continue\n",
    "gene_mim_df = pd.DataFrame(tuples, columns=[\"human_ortholog_gene_symbol\",\"gene_mim_number\",\"phenotype_mim_number\",\"phenotype\"])\n",
    "gene_mim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_identifier</th>\n",
       "      <th>human_ortholog_gene_symbol</th>\n",
       "      <th>gene_mim_number</th>\n",
       "      <th>phenotype_mim_number</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>MCCC1</td>\n",
       "      <td>609010</td>\n",
       "      <td>210200</td>\n",
       "      <td>3-Methylcrotonyl-CoA carboxylase 1 deficiency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126340</td>\n",
       "      <td>610756</td>\n",
       "      <td>?Cerebrooculofacioskeletal syndrome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126340</td>\n",
       "      <td>601675</td>\n",
       "      <td>Trichothiodystrophy 1, photosensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126340</td>\n",
       "      <td>278730</td>\n",
       "      <td>Xeroderma pigmentosum, group D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>AT1G03430</td>\n",
       "      <td>CPOX</td>\n",
       "      <td>612732</td>\n",
       "      <td>121300</td>\n",
       "      <td>Coproporphyria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species gene_identifier human_ortholog_gene_symbol  gene_mim_number  phenotype_mim_number                                      phenotype\n",
       "0   ARATH       AT1G03430                      MCCC1           609010                210200  3-Methylcrotonyl-CoA carboxylase 1 deficiency\n",
       "1   ARATH       AT1G03430                      ERCC2           126340                610756          ?Cerebrooculofacioskeletal syndrome 2\n",
       "2   ARATH       AT1G03430                      ERCC2           126340                601675          Trichothiodystrophy 1, photosensitive\n",
       "3   ARATH       AT1G03430                      ERCC2           126340                278730                 Xeroderma pigmentosum, group D\n",
       "4   ARATH       AT1G03430                       CPOX           612732                121300                                 Coproporphyria"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_df = pd.merge(left=ortholog_df, right=gene_mim_df, how=\"left\")\n",
    "full_df[\"gene_mim_number\"] = full_df[\"gene_mim_number\"].astype(\"Int64\")\n",
    "full_df[\"phenotype_mim_number\"] = full_df[\"phenotype_mim_number\"].astype(\"Int64\")\n",
    "full_df.to_csv(\"/Users/irbraun/Desktop/a.csv\", index=False)\n",
    "full_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
