{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in data available through The Arabidopsis Information Resource (TAIR) database\n",
    "The purpose of this notebook is to read in and do a preliminary survey of the data related to text and ontology annotations of text that is available through TAIR. The datasets need to be organized and also restructured into a standard format that will allow it be combined with datasets from other resources. This notebook takes the following input files that were downloaded from the TAIR and produces a set of files that have standard columns, including the species name, gene names, gene synonyms, text descriptions, and ontology term annotations. The gene names column includes unique gene accessions, names, symbols, or identifiers. The gene synonyms column is included for strings that are not necessary unique identifiers for a particular gene but still refer to that gene or describe its function.\n",
    "\n",
    "### Files read\n",
    "```\n",
    "phenologs-with-oats/data/gene_related_files/tair/Locus_Germplasm_Phenotype_20180702.txt\n",
    "phenologs-with-oats/data/gene_related_files/tair/po_anatomy_gene_arabidopsis_tair.assoc\n",
    "phenologs-with-oats/data/gene_related_files/tair/po_temporal_gene_arabidopsis_tair.assoc\n",
    "phenologs-with-oats/data/gene_related_files/tair/ATH_GO_GOSLIM.txt\n",
    "```\n",
    "\n",
    "### Files created\n",
    "```\n",
    "phenologs-with-oats/data/reshaped_files/ath_phenotypes.csv\n",
    "phenologs-with-oats/data/reshaped_files/ath_all_go_annotations.csv\n",
    "phenologs-with-oats/data/reshaped_files/ath_high_confidence_go_annotations.csv\n",
    "phenologs-with-oats/data/reshaped_files/ath_high_confidence_po_annotations.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.constants import NCBI_TAG\n",
    "from oats.utils.constants import EVIDENCE_CODES\n",
    "from oats.utils.utils import to_abbreviation\n",
    "from oats.nlp.preprocess import concatenate_with_bar_delim\n",
    "from oats.nlp.preprocess import other_delim_to_bar_delim\n",
    "from oats.nlp.preprocess import remove_punctuation\n",
    "from oats.nlp.preprocess import remove_enclosing_brackets\n",
    "from oats.nlp.preprocess import concatenate_descriptions\n",
    "from oats.nlp.preprocess import add_prefix\n",
    "\n",
    "OUTPUT_DIR = \"../data/reshaped_files\"\n",
    "mpl.rcParams[\"figure.dpi\"] = 200\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with genes and phenotype descriptions (Locus_Germplasm_Phenotype_20180702.txt)\n",
    "Reading in the dataset of phenotypic descriptions. There is only one value specified as the gene name (locus name) in the original dataset so this column does not need to be parsed further. The descriptions commmonly use semi-colons to separate phrases. The next cell gets the distribution of the number of phrases in each description field for the dataset of text descriptions, as determined by a sentence parser. The majority of the descriptions are a single sentence or phrase, but some contain more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10347</th>\n",
       "      <td>AT5G20730</td>\n",
       "      <td>Hypocotyl growth resistant (but not root growt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9397</th>\n",
       "      <td>AT5G02810</td>\n",
       "      <td>The prr7 mutant phenotype in Rc was more drama...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9935</th>\n",
       "      <td>AT5G13930</td>\n",
       "      <td>yellow seeds (lacks brown pigment in seed coat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3610</th>\n",
       "      <td>AT2G17430</td>\n",
       "      <td>female gametophytic defects; reduced fertility...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1930</th>\n",
       "      <td>AT1G51220</td>\n",
       "      <td>Rootless phenotype. nww mutant embryos diverge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4759</th>\n",
       "      <td>AT2G39200</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>no dramatic change in phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>AT1G01550</td>\n",
       "      <td>In double mutant seedlings grown on CPTA, a pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>AT1G08260</td>\n",
       "      <td>developmental arrest of mutant embryos occurs ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13596</th>\n",
       "      <td>EMB2995</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4915</th>\n",
       "      <td>AT2G41370</td>\n",
       "      <td>Whereas the gynoecium of wild-type flowers con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11158</th>\n",
       "      <td>AT5G47910</td>\n",
       "      <td>ABA fails to induce significant nitric acid (N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5969</th>\n",
       "      <td>AT3G16857</td>\n",
       "      <td>The development and size of the double mutant ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13196</th>\n",
       "      <td>EMB2233</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12841</th>\n",
       "      <td>EMB1732</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8097</th>\n",
       "      <td>AT4G15802</td>\n",
       "      <td>Early flowering, reduced fertility, aborted se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4373</th>\n",
       "      <td>AT2G32660</td>\n",
       "      <td>No visible phenotype.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13274</th>\n",
       "      <td>EMB2342</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>AT1G09730</td>\n",
       "      <td>Mutant exhibits abnormal floral structures and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3620</th>\n",
       "      <td>AT2G17790</td>\n",
       "      <td>No visible phenotype.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_names                                        description\n",
       "10347  AT5G20730  Hypocotyl growth resistant (but not root growt...\n",
       "9397   AT5G02810  The prr7 mutant phenotype in Rc was more drama...\n",
       "9935   AT5G13930  yellow seeds (lacks brown pigment in seed coat...\n",
       "3610   AT2G17430  female gametophytic defects; reduced fertility...\n",
       "1930   AT1G51220  Rootless phenotype. nww mutant embryos diverge...\n",
       "4759   AT2G39200                                               none\n",
       "70     AT1G01480                    no dramatic change in phenotype\n",
       "149    AT1G01550  In double mutant seedlings grown on CPTA, a pa...\n",
       "491    AT1G08260  developmental arrest of mutant embryos occurs ...\n",
       "13596    EMB2995  embryo defective; terminal embryo phenotype: c...\n",
       "4915   AT2G41370  Whereas the gynoecium of wild-type flowers con...\n",
       "11158  AT5G47910  ABA fails to induce significant nitric acid (N...\n",
       "5969   AT3G16857  The development and size of the double mutant ...\n",
       "13196    EMB2233  embryo defective; terminal embryo phenotype: p...\n",
       "12841    EMB1732  embryo defective; terminal embryo phenotype: g...\n",
       "8097   AT4G15802  Early flowering, reduced fertility, aborted se...\n",
       "4373   AT2G32660                              No visible phenotype.\n",
       "13274    EMB2342  embryo defective; terminal embryo phenotype: g...\n",
       "627    AT1G09730  Mutant exhibits abnormal floral structures and...\n",
       "3620   AT2G17790                              No visible phenotype."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/tair/Locus_Germplasm_Phenotype_20180702.txt\"\n",
    "usecols = [\"LOCUS_NAME\", \"PHENOTYPE\"]\n",
    "usenames = [\"gene_names\", \"description\"]\n",
    "renamed = {k:v for k,v in zip(usecols,usenames)}\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.rename(columns=renamed, inplace=True)\n",
    "df.sample(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting distributions of number of phrases in each description.\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.set_title(\"Phenotype Descriptions\")\n",
    "ax2.set_title(\"Phenotype Descriptions\")\n",
    "ax1.set_xlabel(\"Number of phrases\")\n",
    "ax2.set_xlabel(\"Number of words\")\n",
    "x1 = [len(sent_tokenize(x)) for x in df[\"description\"].values]\n",
    "x2 = [len(word_tokenize(x)) for x in df[\"description\"].values]\n",
    "ax1.hist(x1, bins=15, range=(0,15), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "ax2.hist(x2, bins=30, range=(0,200), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "fig.set_size_inches(15,4)\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>gene_synonyms</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ACL2</td>\n",
       "      <td></td>\n",
       "      <td>flower stems are much reduced in length, semi-...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>ADE1</td>\n",
       "      <td></td>\n",
       "      <td>sustained and enhanced levels of ABA-regulated...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALB2</td>\n",
       "      <td></td>\n",
       "      <td>white embryo and seedling (albino), lethal.</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-104</td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-128</td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS4</td>\n",
       "      <td></td>\n",
       "      <td>increased aluminum sensitivity; poor root grow...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS5</td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum sens...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC10</td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC2</td>\n",
       "      <td></td>\n",
       "      <td>mesophyll cells contain approximately half the...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC4</td>\n",
       "      <td></td>\n",
       "      <td>fewer and larger chloroplasts than wild type i...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC7</td>\n",
       "      <td></td>\n",
       "      <td>pale first leaves in early seedling growth; me...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC8</td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect;  mea...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC9</td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARS1</td>\n",
       "      <td></td>\n",
       "      <td>Semi-dominant arsenic resistance phenotype. In...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species gene_names gene_synonyms                                        description term_ids sources\n",
       "0      ath       ACL2                flower stems are much reduced in length, semi-...             TAIR\n",
       "1      ath       ADE1                sustained and enhanced levels of ABA-regulated...             TAIR\n",
       "2      ath       ALB2                      white embryo and seedling (albino), lethal.             TAIR\n",
       "3      ath    ALR-104                incomplete penetrance; increased aluminum resi...             TAIR\n",
       "4      ath    ALR-128                incomplete penetrance; increased aluminum resi...             TAIR\n",
       "5      ath       ALS4                increased aluminum sensitivity; poor root grow...             TAIR\n",
       "6      ath       ALS5                incomplete penetrance; increased aluminum sens...             TAIR\n",
       "7      ath      ARC10                no visible whole plant phenotypic effect; mean...             TAIR\n",
       "8      ath       ARC2                mesophyll cells contain approximately half the...             TAIR\n",
       "9      ath       ARC4                fewer and larger chloroplasts than wild type i...             TAIR\n",
       "10     ath       ARC7                pale first leaves in early seedling growth; me...             TAIR\n",
       "11     ath       ARC8                no visible whole plant phenotypic effect;  mea...             TAIR\n",
       "12     ath       ARC9                no visible whole plant phenotypic effect; mean...             TAIR\n",
       "13     ath       ARS1                Semi-dominant arsenic resistance phenotype. In...             TAIR\n",
       "14     ath  AT1G01040                embryo defective; developmental arrest of muta...             TAIR\n",
       "15     ath  AT1G01040                embryo defective; developmental arrest of muta...             TAIR\n",
       "16     ath  AT1G01040                embryo defective; developmental arrest of muta...             TAIR\n",
       "17     ath  AT1G01040                embryo defective; developmental arrest of muta...             TAIR\n",
       "18     ath  AT1G01040                                       embryo defective; globular             TAIR\n",
       "19     ath  AT1G01040                                       embryo defective; globular             TAIR"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all expected column names.\n",
    "df[\"species\"] = \"ath\"\n",
    "df[\"term_ids\"] = \"\"\n",
    "df[\"gene_synonyms\"] = \"\"\n",
    "df[\"sources\"] = \"TAIR\"\n",
    "df = df[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\", \"sources\"]]\n",
    "\n",
    "# Outputting the dataset of phenotype descriptions to csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_phenotypes.csv\")\n",
    "df.to_csv(path, index=False)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with gene ontology annotations (ATH_GO_GOSLIM.txt)\n",
    "Read in the file containing names of loci and corresponding information relating to gene ontology term annotation. Not all of the columns are used here, only a subset of them are read in. The relationship column refers to the relationships between the gene for that loci and the term mentioned on that given line. Evidence refer to the method of acquiring and the confidence in the annotation itself. This is retained so that we can subset that dataset based on whether the annotations are experimentally confirmed or simply predicted annotations. This section also looks at how many unique values are present for each field.\n",
    "\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. About half of the term annotations were made computationally, but there are also a high number of annotations available from high confidence annotations such as experimentally validated, curator statements, and author statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       locus     term_id evidence_code                    reference\n",
      "0  AT1G01010  GO:0007275           ISS      Communication:501714663\n",
      "1  AT1G01010  GO:0005634           ISM  AnalysisReference:501750651\n",
      "2  AT1G01010  GO:0003700           ISS          Publication:1345963\n",
      "3  AT1G01010  GO:0043090           RCA        Publication:501748728\n",
      "4  AT1G01010  GO:0006888           RCA        Publication:501748728\n",
      "5  AT1G01020  GO:0006665           IMP        Publication:501719026\n",
      "6  AT1G01020  GO:0016125           IMP        Publication:501719026\n",
      "7  AT1G01020  GO:0005783           IDA        Publication:501719026\n",
      "8  AT1G01020  GO:0016020           ISS      Communication:501714663\n",
      "9  AT1G01020  GO:0005739           ISM  AnalysisReference:501750651\n",
      "(251297, 6)\n",
      "locus                30251\n",
      "relationship            19\n",
      "term_label            4924\n",
      "term_id               4924\n",
      "evidence_code           13\n",
      "reference             9149\n"
     ]
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/tair/ATH_GO_GOSLIM.txt\"\n",
    "df_go = pd.read_table(filename, header=None, usecols=[0,3,4,5,9,12])\n",
    "df_go.columns = [\"locus\",\"relationship\",\"term_label\",\"term_id\",\"evidence_code\",\"reference\"]\n",
    "unique_values = {col:len(pd.unique(df_go[col].values)) for col in df_go.columns}\n",
    "print(df_go[[\"locus\",\"term_id\",\"evidence_code\",\"reference\"]].head(10))\n",
    "print(df_go.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "electronic_annotation           0\n",
      "computational              125915\n",
      "curator_statement           24550\n",
      "author_statement            13460\n",
      "high_throughput                 0\n",
      "experimental                85701\n",
      "phylogenetics                1671\n"
     ]
    }
   ],
   "source": [
    "code_quantities = {c:len([x for x in df_go[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>gene_synonyms</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0007275</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0043090</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0006888</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0006665</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0006665</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0048366</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GO:0009908</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species gene_names gene_synonyms description    term_ids sources\n",
       "0      ath  AT1G01010                            GO:0007275    TAIR\n",
       "1      ath  AT1G01010                            GO:0005634    TAIR\n",
       "2      ath  AT1G01010                            GO:0003700    TAIR\n",
       "3      ath  AT1G01010                            GO:0043090    TAIR\n",
       "4      ath  AT1G01010                            GO:0006888    TAIR\n",
       "5      ath  AT1G01020                            GO:0006665    TAIR\n",
       "6      ath  AT1G01020                            GO:0016125    TAIR\n",
       "7      ath  AT1G01020                            GO:0005783    TAIR\n",
       "8      ath  AT1G01020                            GO:0016020    TAIR\n",
       "9      ath  AT1G01020                            GO:0005739    TAIR\n",
       "10     ath  AT1G01020                            GO:0005739    TAIR\n",
       "11     ath  AT1G01020                            GO:0016125    TAIR\n",
       "12     ath  AT1G01020                            GO:0003674    TAIR\n",
       "13     ath  AT1G01020                            GO:0006665    TAIR\n",
       "14     ath  AT1G01030                            GO:0006355    TAIR\n",
       "15     ath  AT1G01030                            GO:0003700    TAIR\n",
       "16     ath  AT1G01030                            GO:0048366    TAIR\n",
       "17     ath  AT1G01030                            GO:0006355    TAIR\n",
       "18     ath  AT1G01030                            GO:0006355    TAIR\n",
       "19     ath  AT1G01030                            GO:0009908    TAIR"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df_go[\"species\"] = \"ath\"\n",
    "df_go[\"gene_names\"] = df_go[\"locus\"]\n",
    "df_go[\"gene_synonyms\"] = \"\"\n",
    "df_go[\"description\"] = \"\"\n",
    "df_go[\"term_ids\"] = df_go[\"term_id\"]\n",
    "df_go[\"sources\"] = \"TAIR\"\n",
    "high_confidence_categories = [\"experimental\",\"other_statement\",\"curator_statement\"]\n",
    "df_go[\"high_confidence\"] = df_go[\"evidence_code\"].apply(lambda x: EVIDENCE_CODES[x] in high_confidence_categories)\n",
    "\n",
    "# Subset to only include the high-quality GO annotations and ouptut the dataset to a csv file.\n",
    "df_go_high_confidence = df_go[df_go[\"high_confidence\"]==True]\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_high_confidence_go_annotations.csv\")\n",
    "df_go_high_confidence = df_go_high_confidence[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\", \"sources\"]]\n",
    "df_go_high_confidence.to_csv(path, index=False)\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "df_go = df_go[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\", \"sources\"]]\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_all_go_annotations.csv\")\n",
    "df_go.to_csv(path, index=False)\n",
    "df_go.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with plant ontology term annotations (po_[termporal/anatomy]_gene_arabidopsis_tair.assoc)\n",
    "There are two separate files available that include annotations of PO terms. The files do not have headers so column names are added based on how the columns are described in the accompanying available readme files. One of the files contains annotations for PO terms that are spatial, or describe a specific part of plant anatomy or plant molecular structures. The other file contains annotations for PO terms that are temporal, or refer to a specific process or stage of development. These files are each read in separately, and the next cells look at the quantity of unique values in the columns of each dataset. There are more spatial annotations than temporal annotations, and a greater number of terms used to describe the spatial annotations.\n",
    "\n",
    "The next field combines the two datasets of PO annotations and looks at the number of unique values for each column in the resulting dataset. Because there is no overlap in the terms between the two, the datasets are simply appended to one another and the total unique terms are a sum of the individual datasets.\n",
    "\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. Almost all of the PO term annotations are high confidence, they are experimentally validated, and only a few of them are derived from author statements.\n",
    "\n",
    "The strings which are described in the synonyms column are included as references to each gene, and are combined with the gene name mentioned in the symbol column into a single bar delimited list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(366398, 7)\n",
      "symbol               24123\n",
      "term_id                365\n",
      "references            2600\n",
      "evidence_code            6\n",
      "name                 25269\n",
      "synonyms             28005\n",
      "type                     6\n"
     ]
    }
   ],
   "source": [
    "# Reading in the dataset of spatial PO term annotations.\n",
    "filename = \"../data/gene_related_files/tair/po_anatomy_gene_arabidopsis_tair.assoc\"\n",
    "df_po_spatial = pd.read_table(filename, header=None, skiprows=1, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_spatial.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_spatial[col].values)) for col in df_po_spatial.columns}\n",
    "print(df_po_spatial.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186740, 7)\n",
      "symbol               18015\n",
      "term_id                 84\n",
      "references             447\n",
      "evidence_code            4\n",
      "name                 18506\n",
      "synonyms             18960\n",
      "type                     5\n"
     ]
    }
   ],
   "source": [
    "# Reading in the dataset of temporal PO term annotations.\n",
    "filename = \"../data/gene_related_files/tair/po_temporal_gene_arabidopsis_tair.assoc\"\n",
    "df_po_temporal = pd.read_table(filename, header=None, skiprows=1, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_temporal.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_temporal[col].values)) for col in df_po_temporal.columns}\n",
    "print(df_po_temporal.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   symbol                                           synonyms evidence_code\n",
      "0  ATARCA  AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...           IEP\n",
      "1    HMG1  AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...           TAS\n",
      "2    BAS1  AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...           TAS\n",
      "3    BAK1  AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...           TAS\n",
      "4    BRI1  AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...           TAS\n",
      "5    ATJ6    AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8           IEP\n",
      "6    XPB2  AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...           NAS\n",
      "7    XPB1  AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...           NAS\n",
      "8   PPCK2  AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...           IEP\n",
      "9    MYC2  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...           NAS\n",
      "(553138, 7)\n",
      "symbol               24188\n",
      "term_id                449\n",
      "references            2671\n",
      "evidence_code            6\n",
      "name                 25350\n",
      "synonyms             28150\n",
      "type                     7\n"
     ]
    }
   ],
   "source": [
    "# Looking at how many unique values each column has.\n",
    "df_po = df_po_spatial.append(df_po_temporal, ignore_index=True)\n",
    "unique_values = {col:len(pd.unique(df_po[col].values)) for col in df_po.columns}\n",
    "print(df_po[[\"symbol\",\"synonyms\",\"evidence_code\"]].head(10))\n",
    "print(df_po.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "electronic_annotation           0\n",
      "computational                   0\n",
      "curator_statement               0\n",
      "author_statement              118\n",
      "high_throughput                 0\n",
      "experimental               553020\n",
      "phylogenetics                   0\n"
     ]
    }
   ],
   "source": [
    "# Quantifying the number of annotations of each type.\n",
    "code_quantities = {c:len([x for x in df_po[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol           object\n",
       "term_id          object\n",
       "references       object\n",
       "evidence_code    object\n",
       "name             object\n",
       "synonyms         object\n",
       "type             object\n",
       "dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_po['name'] = df_po['name'].astype(str)\n",
    "df_po.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>gene_synonyms</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATARCA|AT1G18080</td>\n",
       "      <td>AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>HMG1|AT1G76490</td>\n",
       "      <td>AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>BAS1|AT2G26710</td>\n",
       "      <td>AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>BAK1|AT4G33430</td>\n",
       "      <td>AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>BRI1|AT4G39400</td>\n",
       "      <td>AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATJ6|AT5G06910</td>\n",
       "      <td>AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>XPB2|AT5G41360</td>\n",
       "      <td>AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>XPB1|AT5G41370</td>\n",
       "      <td>AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>PPCK2|AT3G04530</td>\n",
       "      <td>AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000013</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>MYC2|AT1G32640</td>\n",
       "      <td>AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000017</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATSBT5.2|AT1G20160</td>\n",
       "      <td>AT1G20160|ATSBT5.2|CRSP|SBT5.2(a)|SBT5.2(b)|CO...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>AT1G32960|ATSBT3.3|SBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.12|AT4G21326</td>\n",
       "      <td>AT4G21326|ATSBT3.12|SBT3.12|subtilase 3.12</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>AT5G51750|ATSBT1.3|SBT1.3|subtilase 1.3|MIO24....</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWO1|AT3G03140</td>\n",
       "      <td>AT3G03140|PWO1|WWP-DOMAIN INTERACTOR OF POLYCO...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000037</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>AT5G51750|ATSBT1.3|SBT1.3|subtilase 1.3|MIO24....</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000037</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>SEP3|AT1G24260</td>\n",
       "      <td>AT1G24260|SEP3|AGL9|SEPALLATA3|AGAMOUS-like 9|...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000229</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>RABA1b|AT1G16920</td>\n",
       "      <td>AT1G16920|RAB11|ATRABA1B|RABA1b|BEX5|RAB GTPas...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000256</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>MYC2|AT1G32640</td>\n",
       "      <td>AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000256</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>AT1G32960|ATSBT3.3|SBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ath</td>\n",
       "      <td>EXPA1|AT1G69530</td>\n",
       "      <td>AT1G69530|ATEXPA1|EXP1|AT-EXP1|ATEXP1|ATHEXP A...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>AT5G51750|ATSBT1.3|SBT1.3|subtilase 1.3|MIO24....</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.12|AT5G59090</td>\n",
       "      <td>AT5G59090|ATSBT4.12|SBT4.12|subtilase 4.12|K18...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.13|AT5G59120</td>\n",
       "      <td>AT5G59120|ATSBT4.13|SBT4.13|subtilase 4.13|MNC...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT5.4|AT5G59810</td>\n",
       "      <td>AT5G59810|ATSBT5.4|SBT5.4|MMN10.6|MMN10_6</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ath</td>\n",
       "      <td>UBP19|AT2G24640</td>\n",
       "      <td>AT2G24640|UBP19|ubiquitin-specific protease 19...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0002000</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATSBT5.2|AT1G20160</td>\n",
       "      <td>AT1G20160|ATSBT5.2|CRSP|SBT5.2(a)|SBT5.2(b)|CO...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>AT1G32960|ATSBT3.3|SBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.12|AT5G59090</td>\n",
       "      <td>AT5G59090|ATSBT4.12|SBT4.12|subtilase 4.12|K18...</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT5.4|AT5G59810</td>\n",
       "      <td>AT5G59810|ATSBT5.4|SBT5.4|MMN10.6|MMN10_6</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species          gene_names                                      gene_synonyms description    term_ids sources\n",
       "0      ath    ATARCA|AT1G18080  AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...              PO:0000003    TAIR\n",
       "1      ath      HMG1|AT1G76490  AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...              PO:0000003    TAIR\n",
       "2      ath      BAS1|AT2G26710  AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...              PO:0000003    TAIR\n",
       "3      ath      BAK1|AT4G33430  AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...              PO:0000003    TAIR\n",
       "4      ath      BRI1|AT4G39400  AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...              PO:0000003    TAIR\n",
       "5      ath      ATJ6|AT5G06910    AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8              PO:0000003    TAIR\n",
       "6      ath      XPB2|AT5G41360  AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...              PO:0000003    TAIR\n",
       "7      ath      XPB1|AT5G41370  AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...              PO:0000003    TAIR\n",
       "8      ath     PPCK2|AT3G04530  AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...              PO:0000013    TAIR\n",
       "9      ath      MYC2|AT1G32640  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...              PO:0000017    TAIR\n",
       "10     ath  ATSBT5.2|AT1G20160  AT1G20160|ATSBT5.2|CRSP|SBT5.2(a)|SBT5.2(b)|CO...              PO:0000036    TAIR\n",
       "11     ath    SBT3.3|AT1G32960        AT1G32960|ATSBT3.3|SBT3.3|F9L11.13|F9L11_13              PO:0000036    TAIR\n",
       "12     ath   SBT3.12|AT4G21326         AT4G21326|ATSBT3.12|SBT3.12|subtilase 3.12              PO:0000036    TAIR\n",
       "13     ath    SBT1.3|AT5G51750  AT5G51750|ATSBT1.3|SBT1.3|subtilase 1.3|MIO24....              PO:0000036    TAIR\n",
       "14     ath      PWO1|AT3G03140  AT3G03140|PWO1|WWP-DOMAIN INTERACTOR OF POLYCO...              PO:0000037    TAIR\n",
       "15     ath    SBT1.3|AT5G51750  AT5G51750|ATSBT1.3|SBT1.3|subtilase 1.3|MIO24....              PO:0000037    TAIR\n",
       "16     ath      SEP3|AT1G24260  AT1G24260|SEP3|AGL9|SEPALLATA3|AGAMOUS-like 9|...              PO:0000229    TAIR\n",
       "17     ath    RABA1b|AT1G16920  AT1G16920|RAB11|ATRABA1B|RABA1b|BEX5|RAB GTPas...              PO:0000256    TAIR\n",
       "18     ath      MYC2|AT1G32640  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...              PO:0000256    TAIR\n",
       "19     ath    SBT3.3|AT1G32960        AT1G32960|ATSBT3.3|SBT3.3|F9L11.13|F9L11_13              PO:0000293    TAIR\n",
       "20     ath     EXPA1|AT1G69530  AT1G69530|ATEXPA1|EXP1|AT-EXP1|ATEXP1|ATHEXP A...              PO:0000293    TAIR\n",
       "21     ath    SBT1.3|AT5G51750  AT5G51750|ATSBT1.3|SBT1.3|subtilase 1.3|MIO24....              PO:0000293    TAIR\n",
       "22     ath   SBT4.12|AT5G59090  AT5G59090|ATSBT4.12|SBT4.12|subtilase 4.12|K18...              PO:0000293    TAIR\n",
       "23     ath   SBT4.13|AT5G59120  AT5G59120|ATSBT4.13|SBT4.13|subtilase 4.13|MNC...              PO:0000293    TAIR\n",
       "24     ath    SBT5.4|AT5G59810          AT5G59810|ATSBT5.4|SBT5.4|MMN10.6|MMN10_6              PO:0000293    TAIR\n",
       "25     ath     UBP19|AT2G24640  AT2G24640|UBP19|ubiquitin-specific protease 19...              PO:0002000    TAIR\n",
       "26     ath  ATSBT5.2|AT1G20160  AT1G20160|ATSBT5.2|CRSP|SBT5.2(a)|SBT5.2(b)|CO...              PO:0003011    TAIR\n",
       "27     ath    SBT3.3|AT1G32960        AT1G32960|ATSBT3.3|SBT3.3|F9L11.13|F9L11_13              PO:0003011    TAIR\n",
       "28     ath   SBT4.12|AT5G59090  AT5G59090|ATSBT4.12|SBT4.12|subtilase 4.12|K18...              PO:0003011    TAIR\n",
       "29     ath    SBT5.4|AT5G59810          AT5G59810|ATSBT5.4|SBT5.4|MMN10.6|MMN10_6              PO:0003011    TAIR"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df_po[\"species\"] = \"ath\"\n",
    "df_po[\"gene_names\"] = np.vectorize(concatenate_with_bar_delim)(df_po[\"symbol\"], df_po[\"name\"])\n",
    "df_po[\"gene_synonyms\"] = df_po[\"synonyms\"]\n",
    "df_po[\"description\"] = \"\"\n",
    "df_po[\"term_ids\"] = df_po[\"term_id\"]\n",
    "df_po[\"sources\"] = \"TAIR\"\n",
    "df_po = df_po[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\", \"sources\"]]\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_high_confidence_po_annotations.csv\")\n",
    "df_po.to_csv(path, index=False)\n",
    "df_po.head(30)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
