{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in data available from the Sol Genomics Network (SGN) database\n",
    "The purpose of this notebook is to read in and do a preliminary survey of the data related to text descriptions obtained from the Sol Genomics Network. The data was provided in the form of the input file by request, rather than obtained through an already available file from the database. The data needs to be organized and restructured into a standard format that will allow it to be easily combined with datasets from other resources. This notebook takes the following input files that were obtained from SGN and produces a set of files that have standard columns, including the species name, gene names, gene synonyms, text descriptions, and ontology term annotations. The gene names column includes unique gene accessions, names, symbols, or identifiers. The gene synonyms column is included for strings that are not necessary unique identifiers for a particular gene but still refer to that gene or describe its function.\n",
    "\n",
    "### Files read\n",
    "```\n",
    "phenologs-with-oats/data/gene_related_files/sgn/sgn_tomato_phenotyped_loci.txt\n",
    "```\n",
    "\n",
    "\n",
    "### Files created\n",
    "```\n",
    "phenologs-with-oats/data/reshaped_files/sly_phenotypes.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import to_abbreviation\n",
    "from oats.nlp.preprocess import concatenate_with_bar_delim\n",
    "from oats.nlp.preprocess import other_delim_to_bar_delim\n",
    "from oats.nlp.preprocess import remove_punctuation\n",
    "from oats.nlp.preprocess import remove_enclosing_brackets\n",
    "\n",
    "OUTPUT_DIR = \"../data/reshaped_files\"\n",
    "mpl.rcParams[\"figure.dpi\"] = 200\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with genes and phenotype descriptions (sgn_tomato_phenotyped_loci.txt)\n",
    "Note that fillna is being used here to replace missing values with an empty string. This is done so that the missing string will be quantified when checking for the number of occurences of unique values from different columns, see the analysis below. However this is not necessary as a preprocessing step because when the data is read in and appended to a dataset object later, any missing values or empty strings will be handled at that step.\n",
    "\n",
    "There are several columns that contain information about gene names and accessions. We need to know what type of information is in each in order to know which should be retained in the dataset we are preparing. We are interested in both gene names that should map to a specific accession (like cyp716A12 or Medtr3g021350) as well as gene names that are enzyme descriptions (like Ubiquitin-Specific Protease) that could map to more than one gene in a particular species. Each type of information is valuable, but needs to be differentiated so that when comparing whether two rows are specifying the same gene, this is not confused with specifying two different genes that have the same function. In the case of this dataset, we only want to considered the locus names in a single column, the rest of the columns are more ambiguous and as long as all the mapping can be done with the locus names the other names can be ignored for downstream analysis.\n",
    "\n",
    "This section creates a set of columns that have standardized names and include data in a standardized format that other functions within the package expect. The species column contains strings which are KEGG abbreviations for particular species. The gene names column contains any strings we want to consider to be uniquely mapped to some particular gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>locus_symbol</th>\n",
       "      <th>locus_name</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>allele_name</th>\n",
       "      <th>allele_phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Solyc01g008710</td>\n",
       "      <td>man4</td>\n",
       "      <td>mannan endo-1,4-beta-mannosidase</td>\n",
       "      <td>man4i</td>\n",
       "      <td>man4i</td>\n",
       "      <td>Inactive allele of mannosidase that occurs in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Solyc01g009070</td>\n",
       "      <td>gamybl1</td>\n",
       "      <td>gamyb-like 1</td>\n",
       "      <td>gib-1</td>\n",
       "      <td>gib-1</td>\n",
       "      <td>deficient in gibberelin (GA) and their seeds r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Solyc01g056340</td>\n",
       "      <td>det1</td>\n",
       "      <td>de-etiolated 1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>Similar to hp-1; enhances all pigments of gree...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Solyc01g056340</td>\n",
       "      <td>det1</td>\n",
       "      <td>de-etiolated 1</td>\n",
       "      <td>dg</td>\n",
       "      <td>dark green</td>\n",
       "      <td>Immature fruit color normal; darker green colo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Solyc01g056340</td>\n",
       "      <td>det1</td>\n",
       "      <td>de-etiolated 1</td>\n",
       "      <td>j</td>\n",
       "      <td>Jones</td>\n",
       "      <td>Similar to hp-2.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Solyc01g079620</td>\n",
       "      <td>y</td>\n",
       "      <td>colorless fruit epidermis</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>Fruit epidermis lacks pigmentation.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Solyc01g091630</td>\n",
       "      <td>CD2</td>\n",
       "      <td>cutin deficient 2</td>\n",
       "      <td>cd2</td>\n",
       "      <td>cutin deficient 2</td>\n",
       "      <td>Extremely low levels (less than 2%) of cutin i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Solyc01g100490</td>\n",
       "      <td>chln</td>\n",
       "      <td>chloronerva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>Upper leaves smaller and chlorotic interveinal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Solyc01g100490</td>\n",
       "      <td>chln</td>\n",
       "      <td>chloronerva</td>\n",
       "      <td>chln-1</td>\n",
       "      <td>chloronerva</td>\n",
       "      <td>This mutant is an auxotroph for nicotianmine, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Solyc01g104340</td>\n",
       "      <td>Gr</td>\n",
       "      <td>green ripe</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>Resembles gf, except that center of fruit turn...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            locus locus_symbol                        locus_name allele_symbol        allele_name                                   allele_phenotype\n",
       "0  Solyc01g008710         man4  mannan endo-1,4-beta-mannosidase         man4i              man4i  Inactive allele of mannosidase that occurs in ...\n",
       "1  Solyc01g009070      gamybl1                      gamyb-like 1         gib-1              gib-1  deficient in gibberelin (GA) and their seeds r...\n",
       "2  Solyc01g056340         det1                    de-etiolated 1             1                     Similar to hp-1; enhances all pigments of gree...\n",
       "3  Solyc01g056340         det1                    de-etiolated 1            dg         dark green  Immature fruit color normal; darker green colo...\n",
       "4  Solyc01g056340         det1                    de-etiolated 1             j              Jones                                   Similar to hp-2.\n",
       "5  Solyc01g079620            y         colorless fruit epidermis             1                                   Fruit epidermis lacks pigmentation.\n",
       "6  Solyc01g091630          CD2                 cutin deficient 2           cd2  cutin deficient 2  Extremely low levels (less than 2%) of cutin i...\n",
       "7  Solyc01g100490         chln                       chloronerva             1                     Upper leaves smaller and chlorotic interveinal...\n",
       "8  Solyc01g100490         chln                       chloronerva        chln-1        chloronerva  This mutant is an auxotroph for nicotianmine, ...\n",
       "9  Solyc01g104340           Gr                        green ripe             1                     Resembles gf, except that center of fruit turn..."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/sgn/sgn_tomato_phenotyped_loci.txt\"\n",
    "df = pd.read_table(filename)\n",
    "df.fillna(\"\", inplace=True)\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removing rows that have missing inforrmation in the columns we want to keep.\n",
    "df = df[(df[\"locus\"] != \"\") & (df[\"allele_phenotype\"] != \"\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "locus                         71\n",
      "locus_symbol                  68\n",
      "locus_name                    68\n",
      "allele_symbol                 76\n",
      "allele_name                   65\n",
      "allele_phenotype             107\n"
     ]
    }
   ],
   "source": [
    "# Finding out how many unique values there are for each column.\n",
    "unique_values = {col:len(pd.unique(df[col].values)) for col in df.columns}\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:24}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting distributions of number of word and phrases in each description.\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.set_title(\"Phenotype Descriptions\")\n",
    "ax2.set_title(\"Phenotype Descriptions\")\n",
    "ax1.set_xlabel(\"Number of phrases\")\n",
    "ax2.set_xlabel(\"Number of words\")\n",
    "x1 = [len(sent_tokenize(x)) for x in df[\"allele_phenotype\"].values]\n",
    "x2 = [len(word_tokenize(x)) for x in df[\"allele_phenotype\"].values]\n",
    "ax1.hist(x1, bins=10, range=(0,10), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "ax2.hist(x2, bins=25, range=(0,50), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "fig.set_size_inches(15,4)\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>gene_synonyms</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g008710|man4</td>\n",
       "      <td>mannan endo-1,4-beta-mannosidase|man4i</td>\n",
       "      <td>Inactive allele of mannosidase that occurs in ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g009070|gamybl1</td>\n",
       "      <td>gamyb-like 1|gib-1</td>\n",
       "      <td>deficient in gibberelin (GA) and their seeds r...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g056340|det1</td>\n",
       "      <td>de-etiolated 1|1</td>\n",
       "      <td>Similar to hp-1; enhances all pigments of gree...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g056340|det1</td>\n",
       "      <td>de-etiolated 1|dg|dark green</td>\n",
       "      <td>Immature fruit color normal; darker green colo...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g056340|det1</td>\n",
       "      <td>de-etiolated 1|j|Jones</td>\n",
       "      <td>Similar to hp-2.</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g079620|y</td>\n",
       "      <td>colorless fruit epidermis|1</td>\n",
       "      <td>Fruit epidermis lacks pigmentation.</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g091630|CD2</td>\n",
       "      <td>cutin deficient 2|cd2</td>\n",
       "      <td>Extremely low levels (less than 2%) of cutin i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g100490|chln</td>\n",
       "      <td>chloronerva|1</td>\n",
       "      <td>Upper leaves smaller and chlorotic interveinal...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g100490|chln</td>\n",
       "      <td>chloronerva|chln-1</td>\n",
       "      <td>This mutant is an auxotroph for nicotianmine, ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sly</td>\n",
       "      <td>Solyc01g104340|Gr</td>\n",
       "      <td>green ripe|1</td>\n",
       "      <td>Resembles gf, except that center of fruit turn...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species              gene_names                           gene_synonyms                                        description term_ids\n",
       "0     sly     Solyc01g008710|man4  mannan endo-1,4-beta-mannosidase|man4i  Inactive allele of mannosidase that occurs in ...         \n",
       "1     sly  Solyc01g009070|gamybl1                      gamyb-like 1|gib-1  deficient in gibberelin (GA) and their seeds r...         \n",
       "2     sly     Solyc01g056340|det1                        de-etiolated 1|1  Similar to hp-1; enhances all pigments of gree...         \n",
       "3     sly     Solyc01g056340|det1            de-etiolated 1|dg|dark green  Immature fruit color normal; darker green colo...         \n",
       "4     sly     Solyc01g056340|det1                  de-etiolated 1|j|Jones                                   Similar to hp-2.         \n",
       "5     sly        Solyc01g079620|y             colorless fruit epidermis|1                Fruit epidermis lacks pigmentation.         \n",
       "6     sly      Solyc01g091630|CD2                   cutin deficient 2|cd2  Extremely low levels (less than 2%) of cutin i...         \n",
       "7     sly     Solyc01g100490|chln                           chloronerva|1  Upper leaves smaller and chlorotic interveinal...         \n",
       "8     sly     Solyc01g100490|chln                      chloronerva|chln-1  This mutant is an auxotroph for nicotianmine, ...         \n",
       "9     sly       Solyc01g104340|Gr                            green ripe|1  Resembles gf, except that center of fruit turn...         "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Organizing the desired information into a standard set of column headers.\n",
    "df[\"species\"] = \"sly\"\n",
    "df[\"description\"] = df[\"allele_phenotype\"]\n",
    "df[\"gene_names\"] = np.vectorize(concatenate_with_bar_delim)(df[\"locus\"], df[\"locus_symbol\"])\n",
    "df[\"gene_synonyms\"] = np.vectorize(concatenate_with_bar_delim)(df[\"locus_name\"], df[\"allele_symbol\"], df[\"allele_name\"])\n",
    "df[\"term_ids\"] = \"\"\n",
    "df = df[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\"]]\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = os.path.join(OUTPUT_DIR,\"sly_phenotypes.csv\")\n",
    "df.to_csv(path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
