{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in data available through Maize GDB (Maize Genetics and Genomics Database)\n",
    "The purpose of this notebook is to read in and do a preliminary analysis of the data related to text descriptions that are available through Maize GDB. The data was provided in the form of the input file by a request through Maize GDB curators, rather than obtained through an already available file from the database. The data needs to be organized and also restructured into a standard format that will allow it to be easily combined with datasets from other resources. This notebook takes the following input files that were obtained from MaizeGDB and produces a set of files that have standard columns, including the species name, gene names, gene synonyms, text descriptions, and ontology term annotations. The gene names column includes unique gene accessions, names, symbols, or identifiers. The gene synonyms column is included for strings that are not necessary unique identifiers for a particular gene but still refer to that gene or describe its function.\n",
    "\n",
    "### Files read\n",
    "```\n",
    "phenologs-with-oats/data/gene_related_files/maizegdb/pheno_genes.txt\n",
    "phenologs-with-oats/data/gene_related_files/maizegdb/maize_v3.gold.gaf\n",
    "```\n",
    "\n",
    "\n",
    "### Files created\n",
    "```\n",
    "phenologs-with-oats/data/reshaped_files/zma_phenotypes.csv\n",
    "phenologs-with-oats/data/reshaped_files/zma_high_confidence_go_annotations.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.constants import NCBI_TAG, UNIPROT_TAG\n",
    "from oats.utils.constants import EVIDENCE_CODES\n",
    "from oats.utils.utils import to_abbreviation\n",
    "from oats.nlp.preprocess import concatenate_with_bar_delim\n",
    "from oats.nlp.preprocess import other_delim_to_bar_delim\n",
    "from oats.nlp.preprocess import remove_punctuation\n",
    "from oats.nlp.preprocess import remove_enclosing_brackets\n",
    "from oats.nlp.preprocess import concatenate_descriptions\n",
    "from oats.nlp.preprocess import add_prefix\n",
    "\n",
    "mpl.rcParams[\"figure.dpi\"] = 200\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTPUT_DIR = \"../data/reshaped_files\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with genes and phenotype descriptions (pheno_genes.txt)\n",
    "Note that fillna is being used here to replace missing values with an empty string. This is done so that the missing string will be quantified when checking for the number of occurences of unique values from different columns, see the analysis below. However this is not necessary as a preprocessing step because when the data is read in and appended to a dataset object later, any missing values or empty strings will be handled at that step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        phenotype_name                              phenotype_description\n",
      "0     2-seeded kernels                                                   \n",
      "1   A1 null transcript  No color in the aleurone, specifically no colo...\n",
      "2    aberrant seedling  first leaf is small, round and flat; second le...\n",
      "3    aberrant seedling  first leaf is small, round and flat; second le...\n",
      "4  abnormal root hairs                        root hairs fail to elongate\n",
      "5  abnormal root hairs                        root hairs fail to elongate\n",
      "6  abnormal root hairs                        root hairs fail to elongate\n",
      "7  abnormal root hairs                        root hairs fail to elongate\n",
      "8  abnormal root hairs                        root hairs fail to elongate\n",
      "9  abnormal root hairs                        root hairs fail to elongate\n",
      "(3616, 9)\n"
     ]
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/maizegdb/pheno_genes.txt\"\n",
    "usecols = [\"phenotype_name\", \"phenotype_description\", \"locus_name\", \"alleles\", \"locus_synonyms\", \"v3_gene_model\", \"v4_gene_model\", \"uniprot_id\", \"ncbi_gene\"]\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.fillna(\"\", inplace=True)\n",
    "print(df[[\"phenotype_name\",\"phenotype_description\"]].head(10))\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_name</th>\n",
       "      <th>phenotype_description</th>\n",
       "      <th>locus_name</th>\n",
       "      <th>alleles</th>\n",
       "      <th>locus_synonyms</th>\n",
       "      <th>v3_gene_model</th>\n",
       "      <th>v4_gene_model</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>ncbi_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2-seeded kernels</td>\n",
       "      <td></td>\n",
       "      <td>wcr1</td>\n",
       "      <td>Wcr1-reference</td>\n",
       "      <td>wandering carpel1|wcr1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1 null transcript</td>\n",
       "      <td>No color in the aleurone, specifically no colo...</td>\n",
       "      <td>a1</td>\n",
       "      <td>a1-m2-8004::dSpm</td>\n",
       "      <td>a1|anthocyaninless1|bnl(a1)|DFR|FNR|gsy38(A1)|...</td>\n",
       "      <td>GRMZM2G026930</td>\n",
       "      <td>Zm00001d044122</td>\n",
       "      <td>P51108</td>\n",
       "      <td>100286107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aberrant seedling</td>\n",
       "      <td>first leaf is small, round and flat; second le...</td>\n",
       "      <td>pve1</td>\n",
       "      <td>pve1-M2|pve1-R</td>\n",
       "      <td>AC211276.4_FG008|cl12053_1|cl12053_1(383)|inve...</td>\n",
       "      <td>AC211276.4_FG008</td>\n",
       "      <td>Zm00001d013672</td>\n",
       "      <td></td>\n",
       "      <td>103626037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aberrant seedling</td>\n",
       "      <td>first leaf is small, round and flat; second le...</td>\n",
       "      <td>ubl1</td>\n",
       "      <td>ubl1-1</td>\n",
       "      <td>si945031h05(438)|si945031h05a|ubl1</td>\n",
       "      <td>GRMZM2G156575</td>\n",
       "      <td>Zm00001d017432</td>\n",
       "      <td></td>\n",
       "      <td>100275088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>abnormal root hairs</td>\n",
       "      <td>root hairs fail to elongate</td>\n",
       "      <td>bhlh10</td>\n",
       "      <td>bhlh10-1|bhlh10-2</td>\n",
       "      <td>Lotus japonicus roothairless1-like|lrl5|transc...</td>\n",
       "      <td>GRMZM2G067654</td>\n",
       "      <td>Zm00001d045107</td>\n",
       "      <td></td>\n",
       "      <td>103637976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611</th>\n",
       "      <td>zebra necrotic leaf</td>\n",
       "      <td>Necrotic tissue appears between veins in regul...</td>\n",
       "      <td>zb1</td>\n",
       "      <td>zb1</td>\n",
       "      <td>zb1|zebra crossbands1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3612</th>\n",
       "      <td>zebra necrotic leaf</td>\n",
       "      <td>Necrotic tissue appears between veins in regul...</td>\n",
       "      <td>zn1</td>\n",
       "      <td>zn1|zn1-N25</td>\n",
       "      <td>zebra necrotic1|zn1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3613</th>\n",
       "      <td>zebra necrotic leaf</td>\n",
       "      <td>Necrotic tissue appears between veins in regul...</td>\n",
       "      <td>zn2</td>\n",
       "      <td>zn2|zn2-4-6(4461)|zn2-56-3012-10|zn2-94-234|zn...</td>\n",
       "      <td>zebra necrotic2|zn2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>zebra necrotic leaf</td>\n",
       "      <td>Necrotic tissue appears between veins in regul...</td>\n",
       "      <td>zn*-N571D</td>\n",
       "      <td>zn*-N571D</td>\n",
       "      <td>zebra necroticN571D|zn*-571D|zn*-N571D</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>zebra striped seedling</td>\n",
       "      <td>Diurnal crossbands of pale green tissue on lea...</td>\n",
       "      <td>lw1</td>\n",
       "      <td>lw1-97-2306|lw1-N101</td>\n",
       "      <td>4-hydroxy-3-methylbut-2-enyl diphosphate reduc...</td>\n",
       "      <td>GRMZM2G027059</td>\n",
       "      <td>Zm00001d033896</td>\n",
       "      <td></td>\n",
       "      <td>100383164</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3616 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              phenotype_name                              phenotype_description locus_name                                            alleles                                     locus_synonyms     v3_gene_model   v4_gene_model uniprot_id  ncbi_gene\n",
       "0           2-seeded kernels                                                          wcr1                                     Wcr1-reference                             wandering carpel1|wcr1                                                        \n",
       "1         A1 null transcript  No color in the aleurone, specifically no colo...         a1                                   a1-m2-8004::dSpm  a1|anthocyaninless1|bnl(a1)|DFR|FNR|gsy38(A1)|...     GRMZM2G026930  Zm00001d044122     P51108  100286107\n",
       "2          aberrant seedling  first leaf is small, round and flat; second le...       pve1                                     pve1-M2|pve1-R  AC211276.4_FG008|cl12053_1|cl12053_1(383)|inve...  AC211276.4_FG008  Zm00001d013672             103626037\n",
       "3          aberrant seedling  first leaf is small, round and flat; second le...       ubl1                                             ubl1-1                 si945031h05(438)|si945031h05a|ubl1     GRMZM2G156575  Zm00001d017432             100275088\n",
       "4        abnormal root hairs                        root hairs fail to elongate     bhlh10                                  bhlh10-1|bhlh10-2  Lotus japonicus roothairless1-like|lrl5|transc...     GRMZM2G067654  Zm00001d045107             103637976\n",
       "...                      ...                                                ...        ...                                                ...                                                ...               ...             ...        ...        ...\n",
       "3611     zebra necrotic leaf  Necrotic tissue appears between veins in regul...        zb1                                                zb1                              zb1|zebra crossbands1                                                        \n",
       "3612     zebra necrotic leaf  Necrotic tissue appears between veins in regul...        zn1                                        zn1|zn1-N25                                zebra necrotic1|zn1                                                        \n",
       "3613     zebra necrotic leaf  Necrotic tissue appears between veins in regul...        zn2  zn2|zn2-4-6(4461)|zn2-56-3012-10|zn2-94-234|zn...                                zebra necrotic2|zn2                                                        \n",
       "3614     zebra necrotic leaf  Necrotic tissue appears between veins in regul...  zn*-N571D                                          zn*-N571D             zebra necroticN571D|zn*-571D|zn*-N571D                                                        \n",
       "3615  zebra striped seedling  Diurnal crossbands of pale green tissue on lea...        lw1                               lw1-97-2306|lw1-N101  4-hydroxy-3-methylbut-2-enyl diphosphate reduc...     GRMZM2G027059  Zm00001d033896             100383164\n",
       "\n",
       "[3616 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Text information about the phenotypes are contained in both the phenotype name and phenotype description for these data. The can be concatenated and retained together in a new description column that contains all this information, or just the phenotype description could be retained, depending on which data should be used downstream for making similarity comparisons. This is different than for most of the other sources of text used. The next cell looks at how many unique values there are in this data for each column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phenotype_name               648\n",
      "phenotype_description        379\n",
      "locus_name                  1410\n",
      "alleles                     2088\n",
      "locus_synonyms              1408\n",
      "v3_gene_model                482\n",
      "v4_gene_model                469\n",
      "uniprot_id                   140\n",
      "ncbi_gene                    503\n"
     ]
    }
   ],
   "source": [
    "# Finding out how many unique values there are for each column.\n",
    "unique_values = {col:len(pd.unique(df[col].values)) for col in df.columns}\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:24}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are a fairly small number of distinct phenotype descriptions (379) compared to the number of lines that are in the complete dataset (3,616). This means that the same descriptions is occuring many times. Look at which descriptions are occuring most often."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   892                        \n",
      "    90    Anthers shriveled, not usually exerted from glum. Pollen absent or abo\n",
      "    87    seedling white, yellow or palegreen, becomes green, often normal\n",
      "    81    Two classes of albino seedlings: Class I characterized by white or pal\n",
      "    81    Seedling leaves are white.\n",
      "    77    endosperm is opaque and firm, not chalky and not waxy\n",
      "    73    a general term describing an improperly developed endosperm that appea\n",
      "    73    small kernel is the consistent characteristic but variable in other as\n",
      "    71    endosperm has a soft, chalk-like texture, usually a reduced yellow col\n",
      "    65    lighter green seedlings or plants; less yellow than yellow green\n"
     ]
    }
   ],
   "source": [
    "# Get a list sorted by number of occurences for each phenotype description.\n",
    "description_counts = df[\"phenotype_description\"].value_counts().to_dict()\n",
    "sorted_tuples = sorted(description_counts.items(), key = lambda x: x[1], reverse=True)\n",
    "for t in sorted_tuples[0:10]:\n",
    "    print(\"{:6}    {:20}\".format(t[1],t[0][:70]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The only description that occurs far more often than the next is an empty string, where this information is missing entirely. The next cell looks at how many phrases are included in the phenotype description values. Most have a single phrase, some have multiple. These look like they are mainly separated with semicolons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting distributions of number of phrases in each description.\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.set_title(\"Phenotype Descriptions\")\n",
    "ax2.set_title(\"Phenotype Descriptions\")\n",
    "ax1.set_xlabel(\"Number of phrases\")\n",
    "ax2.set_xlabel(\"Number of words\")\n",
    "x1 = [len(sent_tokenize(x)) for x in df[\"phenotype_description\"].values]\n",
    "x2 = [len(word_tokenize(x)) for x in df[\"phenotype_description\"].values]\n",
    "ax1.hist(x1, bins=15, range=(0,15), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "ax2.hist(x2, bins=30, range=(0,150), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "fig.set_size_inches(15,4)\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>gene_synonyms</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zma</td>\n",
       "      <td>wcr1</td>\n",
       "      <td>Wcr1-reference|wandering carpel1|wcr1</td>\n",
       "      <td>2-seeded kernels.</td>\n",
       "      <td></td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zma</td>\n",
       "      <td>a1|GRMZM2G026930|Zm00001d044122|uniprot=P51108...</td>\n",
       "      <td>a1-m2-8004::dSpm|a1|anthocyaninless1|bnl(a1)|D...</td>\n",
       "      <td>A1 null transcript. No color in the aleurone, ...</td>\n",
       "      <td></td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zma</td>\n",
       "      <td>pve1|AC211276.4_FG008|Zm00001d013672|ncbi=1036...</td>\n",
       "      <td>pve1-M2|pve1-R|AC211276.4_FG008|cl12053_1|cl12...</td>\n",
       "      <td>aberrant seedling. first leaf is small, round ...</td>\n",
       "      <td></td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zma</td>\n",
       "      <td>ubl1|GRMZM2G156575|Zm00001d017432|ncbi=100275088</td>\n",
       "      <td>ubl1-1|si945031h05(438)|si945031h05a|ubl1</td>\n",
       "      <td>aberrant seedling. first leaf is small, round ...</td>\n",
       "      <td></td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zma</td>\n",
       "      <td>bhlh10|GRMZM2G067654|Zm00001d045107|ncbi=10363...</td>\n",
       "      <td>bhlh10-1|bhlh10-2|Lotus japonicus roothairless...</td>\n",
       "      <td>abnormal root hairs. root hairs fail to elongate.</td>\n",
       "      <td></td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species                                         gene_names                                      gene_synonyms                                        description term_ids   sources\n",
       "0     zma                                               wcr1              Wcr1-reference|wandering carpel1|wcr1                                  2-seeded kernels.           MaizeGDB\n",
       "1     zma  a1|GRMZM2G026930|Zm00001d044122|uniprot=P51108...  a1-m2-8004::dSpm|a1|anthocyaninless1|bnl(a1)|D...  A1 null transcript. No color in the aleurone, ...           MaizeGDB\n",
       "2     zma  pve1|AC211276.4_FG008|Zm00001d013672|ncbi=1036...  pve1-M2|pve1-R|AC211276.4_FG008|cl12053_1|cl12...  aberrant seedling. first leaf is small, round ...           MaizeGDB\n",
       "3     zma   ubl1|GRMZM2G156575|Zm00001d017432|ncbi=100275088          ubl1-1|si945031h05(438)|si945031h05a|ubl1  aberrant seedling. first leaf is small, round ...           MaizeGDB\n",
       "4     zma  bhlh10|GRMZM2G067654|Zm00001d045107|ncbi=10363...  bhlh10-1|bhlh10-2|Lotus japonicus roothairless...  abnormal root hairs. root hairs fail to elongate.           MaizeGDB"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df[\"description\"] = np.vectorize(concatenate_descriptions)(df[\"phenotype_name\"], df[\"phenotype_description\"])\n",
    "df[\"uniprot_id\"] = df[\"uniprot_id\"].apply(add_prefix, prefix=UNIPROT_TAG)\n",
    "df[\"ncbi_gene\"] = df[\"ncbi_gene\"].apply(add_prefix, prefix=NCBI_TAG)\n",
    "df[\"gene_names\"] = np.vectorize(concatenate_with_bar_delim)(df[\"locus_name\"], df[\"v3_gene_model\"], df[\"v4_gene_model\"], df[\"uniprot_id\"], df[\"ncbi_gene\"])\n",
    "df[\"gene_synonyms\"] = np.vectorize(concatenate_with_bar_delim)(df[\"alleles\"], df[\"locus_synonyms\"])\n",
    "df[\"species\"] = \"zma\"\n",
    "df[\"term_ids\"] = \"\"\n",
    "df[\"sources\"] = \"MaizeGDB\"\n",
    "df = df[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\", \"sources\"]]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Outputting the dataset of descriptions to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"zma_phenotypes.csv\")\n",
    "df.to_csv(path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with high confidence gene ontology annotations (maize_v3.gold.gaf)\n",
    "This file was generated as part of the [Maize GAMER](https://onlinelibrary.wiley.com/doi/full/10.1002/pld3.52)  publication (Wimalanathan et al., 2018). The annotations include all of the associations between maize genes and ontology terms from GO where the terms have been experimentally confirmed to represent correct functional annotations for those genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>term_accession</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td>bx1</td>\n",
       "      <td>GRMZM2G085381</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000162</td>\n",
       "      <td>PMID:9235894</td>\n",
       "      <td>IMP</td>\n",
       "      <td></td>\n",
       "      <td>P</td>\n",
       "      <td>benzoxazinless1</td>\n",
       "      <td>MaizeGDB:GRMZM2G085381</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:4577</td>\n",
       "      <td>20151010</td>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td>mpk2</td>\n",
       "      <td>GRMZM2G020216</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000302</td>\n",
       "      <td>PMID:20693409</td>\n",
       "      <td>IDA</td>\n",
       "      <td></td>\n",
       "      <td>P</td>\n",
       "      <td>MAP kinase2</td>\n",
       "      <td>MaizeGDB:GRMZM2G020216</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:4577</td>\n",
       "      <td>20151010</td>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td>crs1</td>\n",
       "      <td>GRMZM2G078412</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000373</td>\n",
       "      <td>PMID:15598799</td>\n",
       "      <td>IDA</td>\n",
       "      <td></td>\n",
       "      <td>P</td>\n",
       "      <td>chloroplast RNA splicing1</td>\n",
       "      <td>MaizeGDB:GRMZM2G078412</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:4577</td>\n",
       "      <td>20151010</td>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td>crs2</td>\n",
       "      <td>GRMZM2G132021</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000373</td>\n",
       "      <td>PMID:11179231</td>\n",
       "      <td>IDA</td>\n",
       "      <td></td>\n",
       "      <td>P</td>\n",
       "      <td>chloroplast RNA splicing2</td>\n",
       "      <td>MaizeGDB:GRMZM2G132021</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:4577</td>\n",
       "      <td>20151010</td>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td>gams1</td>\n",
       "      <td>ga-ms1</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000775</td>\n",
       "      <td>MaizeGDB:123623</td>\n",
       "      <td>IMP</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>gametophytic male sterile1</td>\n",
       "      <td>MaizeGDB:ga-ms1</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:4577</td>\n",
       "      <td>20151010</td>\n",
       "      <td>MaizeGDB</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        !db db_object_id db_object_symbol qualifier term_accession     db_reference evidence_code with aspect              db_object_name       db_object_synonym db_object_type       taxon      date assigned_by annotation_extension gene_product_form_id\n",
       "0  MaizeGDB          bx1    GRMZM2G085381               GO:0000162     PMID:9235894           IMP           P             benzoxazinless1  MaizeGDB:GRMZM2G085381        protein  taxon:4577  20151010    MaizeGDB                                          \n",
       "1  MaizeGDB         mpk2    GRMZM2G020216               GO:0000302    PMID:20693409           IDA           P                 MAP kinase2  MaizeGDB:GRMZM2G020216        protein  taxon:4577  20151010    MaizeGDB                                          \n",
       "2  MaizeGDB         crs1    GRMZM2G078412               GO:0000373    PMID:15598799           IDA           P   chloroplast RNA splicing1  MaizeGDB:GRMZM2G078412        protein  taxon:4577  20151010    MaizeGDB                                          \n",
       "3  MaizeGDB         crs2    GRMZM2G132021               GO:0000373    PMID:11179231           IDA           P   chloroplast RNA splicing2  MaizeGDB:GRMZM2G132021        protein  taxon:4577  20151010    MaizeGDB                                          \n",
       "4  MaizeGDB        gams1           ga-ms1               GO:0000775  MaizeGDB:123623           IMP           C  gametophytic male sterile1         MaizeGDB:ga-ms1        protein  taxon:4577  20151010    MaizeGDB                                          "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/maizegdb/maize_v3.gold.gaf\"\n",
    "df = pd.read_table(filename, skiprows=1)\n",
    "df.fillna(\"\", inplace=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>gene_synonyms</th>\n",
       "      <th>description</th>\n",
       "      <th>term_ids</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zma</td>\n",
       "      <td>bx1|GRMZM2G085381</td>\n",
       "      <td>MaizeGDB:GRMZM2G085381</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000162</td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zma</td>\n",
       "      <td>mpk2|GRMZM2G020216</td>\n",
       "      <td>MaizeGDB:GRMZM2G020216</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000302</td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zma</td>\n",
       "      <td>crs1|GRMZM2G078412</td>\n",
       "      <td>MaizeGDB:GRMZM2G078412</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000373</td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zma</td>\n",
       "      <td>crs2|GRMZM2G132021</td>\n",
       "      <td>MaizeGDB:GRMZM2G132021</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000373</td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zma</td>\n",
       "      <td>gams1|ga-ms1</td>\n",
       "      <td>MaizeGDB:ga-ms1</td>\n",
       "      <td></td>\n",
       "      <td>GO:0000775</td>\n",
       "      <td>MaizeGDB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species          gene_names           gene_synonyms description    term_ids   sources\n",
       "0     zma   bx1|GRMZM2G085381  MaizeGDB:GRMZM2G085381              GO:0000162  MaizeGDB\n",
       "1     zma  mpk2|GRMZM2G020216  MaizeGDB:GRMZM2G020216              GO:0000302  MaizeGDB\n",
       "2     zma  crs1|GRMZM2G078412  MaizeGDB:GRMZM2G078412              GO:0000373  MaizeGDB\n",
       "3     zma  crs2|GRMZM2G132021  MaizeGDB:GRMZM2G132021              GO:0000373  MaizeGDB\n",
       "4     zma        gams1|ga-ms1         MaizeGDB:ga-ms1              GO:0000775  MaizeGDB"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df[\"description\"] = \"\"\n",
    "df[\"gene_names\"] = np.vectorize(concatenate_with_bar_delim)(df[\"db_object_id\"], df[\"db_object_symbol\"])\n",
    "df[\"gene_synonyms\"] = df[\"db_object_synonym\"]\n",
    "df[\"species\"] = \"zma\"\n",
    "df[\"term_ids\"] = df[\"term_accession\"]\n",
    "df[\"sources\"] = \"MaizeGDB\"\n",
    "df = df[[\"species\", \"gene_names\", \"gene_synonyms\", \"description\", \"term_ids\", \"sources\"]]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"zma_high_confidence_go_annotations.csv\")\n",
    "df.to_csv(path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
