{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data from TAIR (The Arabidopsis Information Resource)\n",
    "The purpose of this notebook is to read in and do a preliminary analysis of the data related to text and ontology annotations of text that is available through TAIR. The datasets need to be organized and also restructured into a standard format that will allow it be combined with datasets from other resources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.constants import NCBI_TAG\n",
    "from oats.utils.constants import EVIDENCE_CODES\n",
    "from oats.utils.utils import to_abbreviation\n",
    "from oats.nlp.preprocess import concatenate_with_bar_delim\n",
    "from oats.nlp.preprocess import other_delim_to_bar_delim\n",
    "from oats.nlp.preprocess import remove_punctuation\n",
    "from oats.nlp.preprocess import remove_enclosing_brackets\n",
    "\n",
    "OUTPUT_DIR = \"../data/reshaped_files\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading in the dataset of phenotypic descriptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   gene_names                                        description  pmid\n",
      "0        ACL2  flower stems are much reduced in length, semi-...   NaN\n",
      "1        ADE1  sustained and enhanced levels of ABA-regulated...   NaN\n",
      "2        ALB2        white embryo and seedling (albino), lethal.   NaN\n",
      "3     ALR-104  incomplete penetrance; increased aluminum resi...   NaN\n",
      "4     ALR-128  incomplete penetrance; increased aluminum resi...   NaN\n",
      "5        ALS4  increased aluminum sensitivity; poor root grow...   NaN\n",
      "6        ALS5  incomplete penetrance; increased aluminum sens...   NaN\n",
      "7       ARC10  no visible whole plant phenotypic effect; mean...   NaN\n",
      "8        ARC2  mesophyll cells contain approximately half the...   NaN\n",
      "9        ARC4  fewer and larger chloroplasts than wild type i...   NaN\n",
      "10       ARC7  pale first leaves in early seedling growth; me...   NaN\n",
      "11       ARC8  no visible whole plant phenotypic effect;  mea...   NaN\n",
      "12       ARC9  no visible whole plant phenotypic effect; mean...   NaN\n",
      "13       ARS1  Semi-dominant arsenic resistance phenotype. In...   NaN\n",
      "14  AT1G01040  embryo defective; developmental arrest of muta...   NaN\n",
      "15  AT1G01040  embryo defective; developmental arrest of muta...   NaN\n",
      "16  AT1G01040  embryo defective; developmental arrest of muta...   NaN\n",
      "17  AT1G01040  embryo defective; developmental arrest of muta...   NaN\n",
      "18  AT1G01040                         embryo defective; globular   NaN\n",
      "19  AT1G01040                         embryo defective; globular   NaN\n",
      "(14013, 3)\n"
     ]
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/tair/Locus_Germplasm_Phenotype_20180702.txt\"\n",
    "usecols = [\"LOCUS_NAME\", \"PHENOTYPE\", \"PUBMED_ID\"]\n",
    "usenames = [\"gene_names\", \"description\", \"pmid\"]\n",
    "renamed = {k:v for k,v in zip(usecols,usenames)}\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.rename(columns=renamed, inplace=True)\n",
    "print(df.head(20))\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is only one value specified as the gene name (locus name) in the original dataset so this column does not need to be parsed further. The descriptions commmonly use semi-colons to separate phrases. The next cell gets the distribution of the number of phrases in each description field for the dataset of text descriptions, as determined by a sentence parser. The majority of the descriptions are a single sentence or phrase, but some contain more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAaJklEQVR4nO3df7RdZX3n8feniSBIUZDAwiQ1VFMr0BE1UlptS4sdqLUN0ymdsFCoQ5tq8Ue7bB1xtdpameLYOmpXxYWggFIxg7bQKlUmRR2VgldE+SUlBSSRSKKIgq1I4Dt/7Ofa481NcskNnOfmvl9rnXX2fvZ+9vnufZN87rP3zj6pKiRJ6s0PjbsASZKmY0BJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRAqXtJPpHkt8Zdx3yU5F1J/ngW/W9IcvQuLEnziAGlLiS5Pcm/J7kvyV1J3ptknw5qev4YPndZkmrHYvJ4/EOSX3y0a6mql1bVn81k3STnJXnTlP6HVdUnHpHitNszoNSTX6mqfYBnAc8B/mjM9YzbE9rxeAZwOfC3SX7z0frwJAserc+SpmNAqTtV9VXgMuDwkeYnJ/lMknuTfDzJAZMLkhyV5LNJ7knyxdFTSu304J9tp++vttNQ97R1n97a3wf8CPD3bRTzmiQfSfKK0VqTfCnJ8W26krwyya1Jvp7kLUl+aGTd/57kpiTfTPKxJE+e4fH4WlW9HfgT4M2T20zypCQfSrI5yW1JXjnyWUcmmUjy7TYCe+vIsueNHK/1k6HXRkBnJfloku8APz86KkpydJINSV7X9u/2JCe1ZauBk4DXtOP19639+6PQJHsmeVuSO9vrbUn2nLLtVyfZlGRjkpeM1PyCJDe2n+FXk/zBTI6d5riq8uVr7C/gduD5bXopcAPwZ23+E8C/Aj8G7NXmz2zLFgPfAF7A8AvXL7b5RTPo+2PAd1qfxwCvAdYBe0ytqc3/BnDVyPwz2mdNrl/AFcD+DOH2L8BvtWXHt20/HVjIMDr87DaOxbK2rYVT2n+0tT+97evngdcDe7RltwLHtnWvBF7cpvcBjmrTPwLcC5zY9vmJwBFt2XnAt4Dntu0/trW9qS0/GtgCvBXYE/i5dvyeNtL/Tdv5ub4R+GfgQGAR8NmRn/Hktt/Y6noB8G/Afm35RuBn2vR+wLPG/WfW1yP/cgSlnvxdknuATwOfBP7nyLL3VtW/VNW/A2uAI1r7i4CPVtVHq+qhqrocmGD4B25Hff8b8JGquryqHgD+giHEfnob9V0CLE+yvM2/GPhgVX1vZJ03V9XdVXUH8DaGIAD4HeDPq+qmqtrS9u2ImY6imjvb+/4Mp0AXVdUbq+p7VXUr8G5gVVvnAeCpSQ6oqvuq6p9b+0nA/62qD1TVA1X1jaq6dnQfq+oz7Vh+dxt1/HFV3V9VnwQ+whDcM3ES8Maq2lRVm4E/ZTiGkx5oyx+oqo8C9wFPG1l2aJJ9q+qbVXXNDD9Tc5gBpZ4cX1VPqKonV9XvtkCZ9LWR6X9jGBUAPBk4oZ2uuqcF3POAg2fQ90nAVyYXVNVDwHqGUdlWqup+hoB7UTvNdiLwvimrrR+Z/kr7jMk63z5S491AtvVZ2zC57t1te0+ast+vAw5q65zKMEL8cpLPJXlha1/KMKLclvXbWQbwzar6zsj86D7uyA8c72n6fqOF96TRn9V/Zfil4ytJPpnkp2b4mZrDFo67AGmW1gPvq6rf3om+dwI/MTmTJAz/gH+1NU33qP/zGULp08C/VdWVU5ZPnp6E4XTa5KhnPXBGVV24E3VO+i/AJuBm4AnAbVW1fLoVq+oW4MQWpL8GXJzkia2OI7fzGTv6eoP9kjxuJKR+BLh+hn3vZAjW6Y7PdlXV54CVSR4DvJzhF4WlM+mrucsRlOa69wO/kuTYJAuSPLZdcF8yg75rgF9Ockz7h+/VwP0M10YA7mK4tvN9LZAeAv6SrUdPAH+YZL8kS4FXAR9s7e8CTk9yGECSxyc5YSY7mOSgJC8H3gCc3kZ6VwPfTvI/kuzV9v3wJM9pfV6UZFFb9562qQeBC4HnJ/mNJAuTPDHJEdN97nb8aZI9kvwM8ELg/7T2rY7XFB8A/ijJogw3qrye4ee3o/3fI8lJSR7fTsV+u+2LdnMGlOa0qloPrGQ4vbWZYYTwh8zgz3ZV3cxwDeuvgK8Dv8Jwq/vkNaU/Z/gH9Z4pd41dwDDymu4f10sYbl64luH6zLnts/4WeDNwUZJvM4w6fmkHJd7T7qa7juH01glV9Z62vQdbvUcAt7X6zwEe3/oeB9yQ5D7g7cCqqvpuuzb2AoYwvrvV+Ywd1DHqa8A3GUY+FwIvraovt2XnMlwnuifJ303T900M1we/1PbpmtY2Ey8Gbm/H7qUMPzft5lLlFxZKD0eSk4HVVfW8Ke0FLK+qdeOp7JGV4fb991fVTEan0qw5gpIehiR7A78LnD3uWqTdnQElzVCSYxlOI94F/M2Yy5F2e57ikyR1yRGUJKlLc/b/QR1wwAG1bNmycZchSZqlz3/+81+vqkVT2+dsQC1btoyJiYlxlyFJmqUkX5mu3VN8kqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLs3ZJ0nsCitWrBh3CQA+EUOSpuEISpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUpR0GVJL3JNmU5PqRtv2TXJ7klva+38iy05OsS3JzkmNH2p+d5Lq27B1J0tr3TPLB1n5VkmW7dhclSXPRTEZQ5wHHTWl7LbC2qpYDa9s8SQ4FVgGHtT7vTLKg9TkLWA0sb6/JbZ4KfLOqngr8b+DNO7szkqTdxw4Dqqo+Bdw9pXklcH6bPh84fqT9oqq6v6puA9YBRyY5GNi3qq6sqgIumNJnclsXA8dMjq4kSfPXzl6DOqiqNgK09wNb+2Jg/ch6G1rb4jY9tf0H+lTVFuBbwBOn+9Akq5NMJJnYvHnzTpYuSZoLdvVNEtONfGo77dvrs3Vj1dlVtaKqVixatGgnS5QkzQU7G1B3tdN2tPdNrX0DsHRkvSXAna19yTTtP9AnyULg8Wx9SlGSNM/sbEBdCpzSpk8BLhlpX9XuzDuE4WaIq9tpwHuTHNWuL508pc/ktn4d+Kd2nUqSNI/t8Bt1k3wAOBo4IMkG4A3AmcCaJKcCdwAnAFTVDUnWADcCW4DTqurBtqmXMdwRuBdwWXsBnAu8L8k6hpHTql2yZ5KkOW2HAVVVJ25j0THbWP8M4Ixp2ieAw6dp/y4t4CRJmuSTJCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqAkSV2aVUAl+f0kNyS5PskHkjw2yf5JLk9yS3vfb2T905OsS3JzkmNH2p+d5Lq27B1JMpu6JElz304HVJLFwCuBFVV1OLAAWAW8FlhbVcuBtW2eJIe25YcBxwHvTLKgbe4sYDWwvL2O29m6JEm7h9me4lsI7JVkIbA3cCewEji/LT8fOL5NrwQuqqr7q+o2YB1wZJKDgX2r6sqqKuCCkT6SpHlqpwOqqr4K/AVwB7AR+FZVfRw4qKo2tnU2Age2LouB9SOb2NDaFrfpqe1bSbI6yUSSic2bN+9s6ZKkOWA2p/j2YxgVHQI8CXhckhdtr8s0bbWd9q0bq86uqhVVtWLRokUPt2RJ0hwym1N8zwduq6rNVfUA8GHgp4G72mk72vumtv4GYOlI/yUMpwQ3tOmp7ZKkeWw2AXUHcFSSvdtdd8cANwGXAqe0dU4BLmnTlwKrkuyZ5BCGmyGubqcB701yVNvOySN9JEnz1MKd7VhVVyW5GLgG2AJ8ATgb2AdYk+RUhhA7oa1/Q5I1wI1t/dOq6sG2uZcB5wF7AZe1lyRpHtvpgAKoqjcAb5jSfD/DaGq69c8AzpimfQI4fDa1SJJ2Lz5JQpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktQlA0qS1KVZBVSSJyS5OMmXk9yU5KeS7J/k8iS3tPf9RtY/Pcm6JDcnOXak/dlJrmvL3pEks6lLkjT3zXYE9XbgH6vqx4FnADcBrwXWVtVyYG2bJ8mhwCrgMOA44J1JFrTtnAWsBpa313GzrEuSNMftdEAl2Rf4WeBcgKr6XlXdA6wEzm+rnQ8c36ZXAhdV1f1VdRuwDjgyycHAvlV1ZVUVcMFIH0nSPDWbEdSPApuB9yb5QpJzkjwOOKiqNgK09wPb+ouB9SP9N7S2xW16avtWkqxOMpFkYvPmzbMoXZLUu9kE1ELgWcBZVfVM4Du003nbMN11pdpO+9aNVWdX1YqqWrFo0aKHW68kaQ6ZTUBtADZU1VVt/mKGwLqrnbajvW8aWX/pSP8lwJ2tfck07ZKkeWynA6qqvgasT/K01nQMcCNwKXBKazsFuKRNXwqsSrJnkkMYboa4up0GvDfJUe3uvZNH+kiS5qmFs+z/CuDCJHsAtwIvYQi9NUlOBe4ATgCoqhuSrGEIsS3AaVX1YNvOy4DzgL2Ay9pLkjSPzSqgqupaYMU0i47ZxvpnAGdM0z4BHD6bWiRJuxefJCFJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSeqSASVJ6pIBJUnqkgElSerSrAMqyYIkX0jyD21+/ySXJ7mlve83su7pSdYluTnJsSPtz05yXVv2jiSZbV2SpLltV4ygXgXcNDL/WmBtVS0H1rZ5khwKrAIOA44D3plkQetzFrAaWN5ex+2CuiRJc9isAirJEuCXgXNGmlcC57fp84HjR9ovqqr7q+o2YB1wZJKDgX2r6sqqKuCCkT6SpHlqtiOotwGvAR4aaTuoqjYCtPcDW/tiYP3Iehta2+I2PbV9K0lWJ5lIMrF58+ZZli5J6tlOB1SSFwKbqurzM+0yTVttp33rxqqzq2pFVa1YtGjRDD9WkjQXLZxF3+cCv5rkBcBjgX2TvB+4K8nBVbWxnb7b1NbfACwd6b8EuLO1L5mmXZI0j+30CKqqTq+qJVW1jOHmh3+qqhcBlwKntNVOAS5p05cCq5LsmeQQhpshrm6nAe9NclS7e+/kkT6SpHlqNiOobTkTWJPkVOAO4ASAqrohyRrgRmALcFpVPdj6vAw4D9gLuKy9JEnz2C4JqKr6BPCJNv0N4JhtrHcGcMY07RPA4buiFknS7sEnSUiSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrq00wGVZGmSK5LclOSGJK9q7fsnuTzJLe19v5E+pydZl+TmJMeOtD87yXVt2TuSZHa7JUma62YzgtoCvLqqng4cBZyW5FDgtcDaqloOrG3ztGWrgMOA44B3JlnQtnUWsBpY3l7HzaIuSdJuYKcDqqo2VtU1bfpe4CZgMbASOL+tdj5wfJteCVxUVfdX1W3AOuDIJAcD+1bVlVVVwAUjfSRJ89QuuQaVZBnwTOAq4KCq2ghDiAEHttUWA+tHum1obYvb9NT26T5ndZKJJBObN2/eFaVLkjo164BKsg/wIeD3qurb21t1mrbaTvvWjVVnV9WKqlqxaNGih1+sJGnOWDibzkkewxBOF1bVh1vzXUkOrqqN7fTdpta+AVg60n0JcGdrXzJN+7yxYsWKcZcAwMTExLhLkKTvm81dfAHOBW6qqreOLLoUOKVNnwJcMtK+KsmeSQ5huBni6nYa8N4kR7VtnjzSR5I0T81mBPVc4MXAdUmubW2vA84E1iQ5FbgDOAGgqm5Isga4keEOwNOq6sHW72XAecBewGXtJUmax3Y6oKrq00x//QjgmG30OQM4Y5r2CeDwna1FkrT78UkSkqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLhlQkqQuGVCSpC4ZUJKkLhlQkqQuLRx3AerHihUrxl0CExMT4y5BUiccQUmSumRASZK6ZEBJkrpkQEmSumRASZK6ZEBJkrrkbebqSg+3uoO3u0s9cAQlSeqSASVJ6pIBJUnqUjfXoJIcB7wdWACcU1VnjrkkzWNeC5PGr4sRVJIFwF8DvwQcCpyY5NDxViVJGqdeRlBHAuuq6laAJBcBK4Ebx1qVNGa9jOT0HxzVPnp6CajFwPqR+Q3AT05dKclqYHWbvS/JzY9CbY+2A4Cvj7uIOcTjNXMeq5nb5rFK8iiX0r1d8efqydM19hJQ0/3Ea6uGqrOBsx/5csYnyURV+WvzDHm8Zs5jNXMeq5l7JI9VF9egGEZMS0fmlwB3jqkWSVIHegmozwHLkxySZA9gFXDpmGuSJI1RF6f4qmpLkpcDH2O4zfw9VXXDmMsal936FOYjwOM1cx6rmfNYzdwjdqxStdWlHkmSxq6XU3ySJP0AA0qS1CUDqhNJlia5IslNSW5I8qpx19S7JAuSfCHJP4y7lp4leUKSi5N8uf35+qlx19SrJL/f/v5dn+QDSR477pp6keQ9STYluX6kbf8klye5pb3vtys/04Dqxxbg1VX1dOAo4DQf97RDrwJuGncRc8DbgX+sqh8HnoHHbFpJFgOvBFZU1eEMN2ytGm9VXTkPOG5K22uBtVW1HFjb5ncZA6oTVbWxqq5p0/cy/COyeLxV9SvJEuCXgXPGXUvPkuwL/CxwLkBVfa+q7hlvVV1bCOyVZCGwN/5/zO+rqk8Bd09pXgmc36bPB47flZ9pQHUoyTLgmcBV462ka28DXgM8NO5COvejwGbgve106DlJHjfuonpUVV8F/gK4A9gIfKuqPj7eqrp3UFVthOGXbODAXblxA6ozSfYBPgT8XlV9e9z19CjJC4FNVfX5cdcyBywEngWcVVXPBL7DLj4Ns7to109WAocATwIel+RF461qfjOgOpLkMQzhdGFVfXjc9XTsucCvJrkduAj4hSTvH29J3doAbKiqydH4xQyBpa09H7itqjZX1QPAh4GfHnNNvbsrycEA7X3Trty4AdWJDI9IPhe4qareOu56elZVp1fVkqpaxnAR+5+qyt90p1FVXwPWJ3laazoGv8ZmW+4Ajkqyd/v7eAzeULIjlwKntOlTgEt25ca7eNSRgGFU8GLguiTXtrbXVdVHx1iTdg+vAC5sz7m8FXjJmOvpUlVdleRi4BqGu2q/gI88+r4kHwCOBg5IsgF4A3AmsCbJqQwBf8Iu/UwfdSRJ6pGn+CRJXTKgJEldMqAkSV0yoCRJXTKgJEldMqC020tSSf5yZP4PkvzJLtr2eUl+fVdsawefc0J7EvkVU9qXJfn3JNcmuTHJu5L8UJKjfcq75joDSvPB/cCvJTlg3IWMSrLgYax+KvC7VfXz0yz716o6AvhPwKE8jAd2PswapEeVAaX5YAvDf7j8/akLpo6AktzX3o9O8skka5L8S5Izk5yU5Ook1yV5yshmnp/k/7X1Xtj6L0jyliSfS/KlJL8zst0rkvwNcN009ZzYtn99kje3ttcDzwPeleQt29rJqtoCfBZ4amvaZ+R7oC5sT0cgye1JXp/k08AJSX671fnFJB9Ksndb74RWxxeTfGoH+3Vwkk+1kdz1SX5mJj8YaXt8koTmi78GvpTkfz2MPs8Ans7wFQO3AudU1ZEZvkzyFcDvtfWWAT8HPAW4IslTgZMZnob9nCR7Ap9JMvlk7COBw6vqttEPS/Ik4M3As4FvAh9PcnxVvTHJLwB/UFUT2yq2BcsxwOtb0zOBwxi+MuIzDE8r+XRb9t2qel7r98SqenebfhPDaO2v2naOraqvJnlC63fqNvbr14CPVdUZbVS2944Pr7R9jqA0L7Qnw1/A8IV0M/W59j1d9wP/CkwGzHUMoTRpTVU9VFW3MATZjwP/GTi5PbbqKuCJwPK2/tVTw6l5DvCJ9rDSLcCFDN/ltCNPaZ/zGeAjVXXZyOdsqKqHgGun1PzBkenD2wjwOuAkhlCjbe+8JL/N8OV9bGe/Pge8pF3b+4n2nWbSrDiC0nzyNobnrL13pG0L7Re1dgpsj5Fl949MPzQy/xA/+Hdn6vPCCgjwiqr62OiCJEczfOXFdLLDPZje5DWoqUbrf5AfrHm0hvOA46vqi0l+k+F5a1TVS5P8JMMXQ16b5Ai2sV8ASX62rfu+JG+pqgt2cn8kwBGU5pGquhtYw3CaatLtDKfUYPguoMfsxKZPaHfOPYXhCwJvBj4GvKx9hQpJfiw7/qLAq4CfS3JAO012IvDJnajn4fphYGOr9aTJxiRPqaqrqur1wNeBpWxjv5I8meE7ut7N8FR+v9JDs+YISvPNXwIvH5l/N3BJkquBtWx7dLM9NzMEyUHAS6vqu0nOYTildk0bmW1mB3fXVdXGJKcDVzCMVD5aVbv06wu24Y8ZwvErDKcvf7i1vyXJ8lbLWuCLwJeYfr+OBv4wyQPAfQzX4KRZ8WnmkqQueYpPktQlA0qS1CUDSpLUJQNKktQlA0qS1CUDSpLUJQNKktSl/w9rYC6LxlIfdAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plotting distributions of number of phrases in each description.\n",
    "x = [len(sent_tokenize(x)) for x in df[\"description\"].values]\n",
    "plt.hist(x, bins=10, range=(1,10), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "plt.title(\"Phenotype Descriptions\")\n",
    "plt.xlabel(\"Number of Phrases\")\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Restructing the data and saving to a csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Restructuring the dataset to include all expected column names.\n",
    "df[\"species\"] = \"ath\"\n",
    "df[\"term_ids\"] = \"\"\n",
    "\n",
    "# Outputting the dataset of phenotype descriptions to csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_tair_gene_text.csv\")\n",
    "df.to_csv(path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading in the dataset of GO annotations\n",
    "Read in the file containing names of loci and corresponding information relating to gene ontology term annotation. Not all of the columns are used here, only a subset of them are read in. The relationship column refers to the relationships between the gene for that loci and the term mentioned on that given line. Evidence refer to the method of acquiring and the confidence in the annotation itself. This is retained so that we can subset that dataset based on whether the annotations are experimentally confirmed or simply predicted annotations. This section also looks at how many unique values are present for each field."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       locus     term_id evidence_code                    reference\n",
      "0  AT1G01010  GO:0007275           ISS      Communication:501714663\n",
      "1  AT1G01010  GO:0005634           ISM  AnalysisReference:501750651\n",
      "2  AT1G01010  GO:0003700           ISS          Publication:1345963\n",
      "3  AT1G01010  GO:0043090           RCA        Publication:501748728\n",
      "4  AT1G01010  GO:0006888           RCA        Publication:501748728\n",
      "5  AT1G01020  GO:0006665           IMP        Publication:501719026\n",
      "6  AT1G01020  GO:0016125           IMP        Publication:501719026\n",
      "7  AT1G01020  GO:0005783           IDA        Publication:501719026\n",
      "8  AT1G01020  GO:0016020           ISS      Communication:501714663\n",
      "9  AT1G01020  GO:0005739           ISM  AnalysisReference:501750651\n",
      "(251297, 6)\n",
      "locus                30251\n",
      "relationship            19\n",
      "term_label            4924\n",
      "term_id               4924\n",
      "evidence_code           13\n",
      "reference             9149\n"
     ]
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/tair/ATH_GO_GOSLIM.txt\"\n",
    "df_go = pd.read_table(filename, header=None, usecols=[0,3,4,5,9,12])\n",
    "df_go.columns = [\"locus\",\"relationship\",\"term_label\",\"term_id\",\"evidence_code\",\"reference\"]\n",
    "unique_values = {col:len(pd.unique(df_go[col].values)) for col in df_go.columns}\n",
    "print(df_go[[\"locus\",\"term_id\",\"evidence_code\",\"reference\"]].head(10))\n",
    "print(df_go.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at the confidence level for the GO annotations\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. About half of the term annotations were made computationally, but there are also a high number of annotations available from high confidence annotations such as experimentally validated, curator statements, and author statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "curator_statement           24550\n",
      "experimental                85701\n",
      "phylogenetics                1671\n",
      "author_statement            13460\n",
      "high_throughput                 0\n",
      "electronic_annotation           0\n",
      "computational              125915\n"
     ]
    }
   ],
   "source": [
    "code_quantities = {c:len([x for x in df_go[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Restructing the GO annotations data and saving to a csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df_go[\"species\"] = \"ath\"\n",
    "df_go[\"gene_names\"] = df_go[\"locus\"]\n",
    "df_go[\"description\"] = \"\"\n",
    "df_go[\"term_ids\"] = df_go[\"term_id\"]\n",
    "df_go[\"pmid\"] = df_go[\"reference\"]\n",
    "df_go = df_go[[\"species\", \"gene_names\", \"description\", \"term_ids\", \"pmid\"]]\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_tair_gene_annot_go.csv\")\n",
    "df_go.to_csv(path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading in the datasets of PO annotations\n",
    "There are two separate files available that include annotations of PO terms. The files do not have headers so column names are added based on how the columns are described in the accompanying available readme files. One of the files contains annotations for PO terms that are spatial, or describe a specific part of plant anatomy or plant molecular structures. The other file contains annotations for PO terms that are temporal, or refer to a specific process or stage of development. These files are each read in separately, and the next cells look at the quantity of unique values in the columns of each dataset. There are more spatial annotations than temporal annotations, and a greater number of terms used to describe the spatial annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(366398, 7)\n",
      "symbol               24123\n",
      "term_id                365\n",
      "references            2600\n",
      "evidence_code            6\n",
      "name                 25269\n",
      "synonyms             28005\n",
      "type                     6\n"
     ]
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/tair/po_anatomy_gene_arabidopsis_tair.assoc\"\n",
    "df_po_spatial = pd.read_table(filename, header=None, skiprows=1, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_spatial.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_spatial[col].values)) for col in df_po_spatial.columns}\n",
    "print(df_po_spatial.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186740, 7)\n",
      "symbol               18015\n",
      "term_id                 84\n",
      "references             447\n",
      "evidence_code            4\n",
      "name                 18506\n",
      "synonyms             18960\n",
      "type                     5\n"
     ]
    }
   ],
   "source": [
    "filename = \"../data/gene_related_files/tair/po_temporal_gene_arabidopsis_tair.assoc\"\n",
    "df_po_temporal = pd.read_table(filename, header=None, skiprows=1, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_temporal.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_temporal[col].values)) for col in df_po_temporal.columns}\n",
    "print(df_po_temporal.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combining each type of PO annotations into single dataset\n",
    "The next field combines the two datasets of PO annotations and looks at the number of unique values for each column in the resulting dataset. Because there is no overlap in the terms between the two, the datasets are simply appended to one another and the total unique terms are a sum of the individual datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   symbol                                           synonyms evidence_code\n",
      "0  ATARCA  AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...           IEP\n",
      "1    HMG1  AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...           TAS\n",
      "2    BAS1  AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...           TAS\n",
      "3    BAK1  AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...           TAS\n",
      "4    BRI1  AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...           TAS\n",
      "5    ATJ6    AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8           IEP\n",
      "6    XPB2  AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...           NAS\n",
      "7    XPB1  AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...           NAS\n",
      "8   PPCK2  AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...           IEP\n",
      "9    MYC2  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...           NAS\n",
      "(553138, 7)\n",
      "symbol               24188\n",
      "term_id                449\n",
      "references            2671\n",
      "evidence_code            6\n",
      "name                 25350\n",
      "synonyms             28150\n",
      "type                     7\n"
     ]
    }
   ],
   "source": [
    "df_po = df_po_spatial.append(df_po_temporal, ignore_index=True)\n",
    "unique_values = {col:len(pd.unique(df_po[col].values)) for col in df_po.columns}\n",
    "print(df_po[[\"symbol\",\"synonyms\",\"evidence_code\"]].head(10))\n",
    "print(df_po.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at the confidence level for the PO annotations\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. Almost all of the PO term annotations are high confidence, they are experimentally validated, and only a few of them are derived from author statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "curator_statement               0\n",
      "experimental               553020\n",
      "phylogenetics                   0\n",
      "author_statement              118\n",
      "high_throughput                 0\n",
      "electronic_annotation           0\n",
      "computational                   0\n"
     ]
    }
   ],
   "source": [
    "code_quantities = {c:len([x for x in df_po[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Restructing the PO annotations data and saving to a csv file\n",
    "The strings which are described in the synonyms column are included as references to each gene, and are combined with the gene name mentioned in the symbol column into a single bar delimited list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df_po[\"species\"] = \"ath\"\n",
    "df_po[\"gene_names\"] = np.vectorize(concatenate_with_bar_delim)(df_po[\"symbol\"], df_po[\"synonyms\"])\n",
    "df_po[\"description\"] = \"\"\n",
    "df_po[\"term_ids\"] = df_po[\"term_id\"]\n",
    "df_po[\"pmid\"] = \"\"\n",
    "df_po = df_po[[\"species\", \"gene_names\", \"description\", \"term_ids\", \"pmid\"]]\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"ath_tair_gene_annot_po.csv\")\n",
    "df_po.to_csv(path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
